{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8dab95dd-2e8e-472e-b1ef-e9566cd08bfb",
   "metadata": {},
   "source": [
    "# Differential exprssion analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75fd7b70-81de-41ed-b00b-54dd26951307",
   "metadata": {},
   "outputs": [],
   "source": [
    "library_load <- suppressMessages(\n",
    "    \n",
    "    suppressWarnings(\n",
    "        \n",
    "        list(\n",
    "        \n",
    "            # DEA \n",
    "            library(DESeq2), \n",
    "            library(IHW), # Independent Hypothesis Weighting\n",
    "            library(edgeR), # For CPM \n",
    "            \n",
    "            # Bioconductor tools\n",
    "            library(rtracklayer), # readGFF, export\n",
    "            library(Rsubread), # featureCounts\n",
    "            \n",
    "            # Data\n",
    "            library(dplyr), \n",
    "            library(rtracklayer), \n",
    "            \n",
    "            # Plotting \n",
    "            library(ggplot2), \n",
    "            library(patchwork), \n",
    "            library(ComplexHeatmap), \n",
    "            library(grid) # For gpar\n",
    "            \n",
    "        )\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7ee4d36-c8a2-4cf8-a1a0-bfa6233f4b12",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c965fc68-e06b-404d-b9b7-6da6d498a9f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set working directory to project root\n",
    "setwd(\"/research/peer/fdeckert/FD20200109SPLENO\") # You need to adjust this path. I will help. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d5d13bf-d70f-405b-b7b8-a96b73cbd542",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting Theme\n",
    "source(\"plotting_global.R\")\n",
    "ggplot2::theme_set(theme_global_set(size_select=1)) # From project global source()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f33ed0c1-9eb3-4aba-89e4-d71fad43df60",
   "metadata": {},
   "source": [
    "# Convert from ENSEML ID to MGI Symbol "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19a12e7b-ffce-48cc-b942-8da3549ccc64",
   "metadata": {},
   "outputs": [],
   "source": [
    "gft_file <- \"/research/peer/fdeckert/reference/genome/GRCm39/gencode.vM33.annotation.gtf.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e06078c-3ae9-4706-b7e0-60ba70aa05f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import_convert <- function(gtf_path=gft_file) {\n",
    "    \n",
    "    # Load reference gtf \n",
    "    gtf <- rtracklayer::import(gtf_path)\n",
    "    gtf <- as.data.frame(gtf)\n",
    "    # Subset data \n",
    "    convert <- gtf[, c(\"gene_id\", \"gene_name\")] \n",
    "    # Rename columns \n",
    "    colnames(convert) <- c(\"ensembl_gene_id\", \"mgi_symbol\")\n",
    "    # Select unique entries \n",
    "    convert <- unique(convert) \n",
    "    convert <- na.omit(convert)\n",
    "\n",
    "    message(paste(\"Total number of ENSEMBL ID matched with MGI SYMBOL:\", nrow(convert)))\n",
    "    \n",
    "    # Filter convert table for unique entries\n",
    "    ensembl_gene_id_multiple <- table(convert$ensembl_gene_id)\n",
    "    ensembl_gene_id_multiple <- ensembl_gene_id_multiple[ensembl_gene_id_multiple>1]\n",
    "    convert <- convert[!convert$ensembl_gene_id %in% names(ensembl_gene_id_multiple), ]\n",
    "    message(paste(\"Unique ENSEMBL ID:\", nrow(convert)))\n",
    "\n",
    "    mgi_symbol_multiple <- table(convert$mgi_symbol)\n",
    "    mgi_symbol_multiple <- mgi_symbol_multiple[mgi_symbol_multiple>1]\n",
    "    convert <- convert[!convert$mgi_symbol %in% names(mgi_symbol_multiple), ]\n",
    "    message(paste(\"Unique MGI SYMBOL:\", nrow(convert)))\n",
    "    \n",
    "    return(convert)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc66c9e0-6291-4450-b0b4-924303055c75",
   "metadata": {},
   "outputs": [],
   "source": [
    "convert <- import_convert()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55ef5b4f-b4e6-4909-a2e6-104917c86b35",
   "metadata": {},
   "outputs": [],
   "source": [
    "ensembl_to_mgi <- function(data, convert) {\n",
    "     \n",
    "    # Filter data and convert by shared ensemble_gene_id \n",
    "    rownames(convert) <- convert$ensembl_gene_id\n",
    "    data <- data[rownames(data) %in% rownames(convert), ]\n",
    "    convert <- convert[rownames(convert) %in% rownames(data), ]\n",
    "    convert <- convert[rownames(data), ]\n",
    "\n",
    "    # Convert ensemble_gene_id to mgi_symbol\n",
    "    rownames(data) <- convert$mgi_symbol\n",
    "    \n",
    "    # Add full annotation \n",
    "    data$mgi_symbol <- convert$mgi_symbol\n",
    "    data$ensembl_gene_id <- rownames(convert)\n",
    "    \n",
    "    return(data)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc9e0ecb-9577-404f-a33d-3fe305352859",
   "metadata": {},
   "source": [
    "# Load pre-processed data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e659875-50e7-47e2-8953-485bfe9114b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta <- readRDS(\"data/bulkRNAseq/object/meta.rds\")\n",
    "cnt <- readRDS(\"data/bulkRNAseq/object/cnt.rds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc326455-b9b4-4759-88d4-09c1de1b0b76",
   "metadata": {},
   "source": [
    "# Differential expression analysis (DEA) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3292a226-22e0-435a-a40c-f97cbef28893",
   "metadata": {},
   "source": [
    "## Helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2da63054-e40a-4203-b189-b46ebcbfa997",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter gene matrix by group expression \n",
    "expression_filter <- function(cnt, group=NULL, min_count=10, min_sample=3, min_group=1) {\n",
    "    \n",
    "    message(paste0(capture.output(table(group)), collapse=\"\\n\"))\n",
    "    \n",
    "    # Print number of input genes \n",
    "    message(\"Number genes input genes \", nrow(cnt))\n",
    "    \n",
    "    # Remove zero expression genes \n",
    "    cnt <- cnt[rowSums(cnt)>0, ]\n",
    "    \n",
    "    # Print number of genes after zero removal \n",
    "    message(\"Number of genes after zero removal \", nrow(cnt))\n",
    "    \n",
    "    x <- cnt >= min_count\n",
    "    x <- as.data.frame(t(x))\n",
    "    x <- split(x, f=group)\n",
    "    x <- lapply(x, colSums)\n",
    "    x <- do.call(rbind, x)\n",
    "    x <- x >= min_sample\n",
    "    x <- x[, colSums(x) >= min_group]\n",
    "    genes <- colnames(x)\n",
    "        \n",
    "    message(\"Number genes after count filter \", length(genes))\n",
    "        \n",
    "    return(genes)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8e39fbf-4da7-4b9f-b1ff-b66f7c97b837",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset meta data by group and filter expression \n",
    "data_subset <- function(meta, cnt, group_by, group, min_count=10, min_sample=3, min_group=1) {\n",
    "    \n",
    "    # Select cell type and subset data \n",
    "    meta <- meta[meta[group_by]==group, ]\n",
    "    cnt <- cnt[, colnames(cnt) %in% rownames(meta)]\n",
    "    \n",
    "    # Reset order of samples in cnt and meta\n",
    "    meta <- meta[intersect(colnames(cnt), rownames(meta)), ]\n",
    "    cnt <- cnt[, intersect(colnames(cnt), rownames(meta))]\n",
    "    \n",
    "    # Drop emtyp factor levels \n",
    "    meta <- droplevels(meta)\n",
    "\n",
    "    # Filter genes\n",
    "    genes_filter <- expression_filter(cnt, min_count=min_count, min_sample=min_sample, min_group=min_group, group=paste(meta$genotype, meta$dpi))\n",
    "    cnt <- cnt[genes_filter, ]\n",
    "    \n",
    "    return(list(meta=meta, cnt=cnt))\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8f185c4-ee4e-446f-a0c5-893da3cb9556",
   "metadata": {},
   "outputs": [],
   "source": [
    "lrt <- function(data) {\n",
    "    \n",
    "    # lrt result list \n",
    "    res <- list()\n",
    "    \n",
    "    # LRT WT\n",
    "    message(\"LRT WT\")\n",
    "    \n",
    "    dds <- DESeqDataSetFromMatrix(countData=data[[\"cnt\"]][, data[[\"meta\"]]$genotype==\"IFNAR_fl\"], colData=data[[\"meta\"]][data[[\"meta\"]]$genotype==\"IFNAR_fl\", ], design=~dpi)\n",
    "    dds <- DESeq(dds, test=\"LRT\", full=~dpi, reduced=~1)\n",
    "    \n",
    "    res[[\"dpi_wt_lrt\"]] <- results(dds) %>% as.data.frame() %>% dplyr::mutate(ihw=ihw(pvalue ~ baseMean,  data=., alpha=0.1) %>% adj_pvalues(.))\n",
    "    \n",
    "    # Wald WT\n",
    "    message(\"Wald WT\")\n",
    "    dds <- DESeqDataSetFromMatrix(countData=data[[\"cnt\"]][, data[[\"meta\"]]$genotype==\"IFNAR_fl\"], colData=data[[\"meta\"]][data[[\"meta\"]]$genotype==\"IFNAR_fl\", ], design=~dpi)\n",
    "    dds <- DESeq(dds, test=\"Wald\")\n",
    "    \n",
    "    res[[\"dpi_wt_wald\"]] <- list(\n",
    "\n",
    "        results(dds, name=\"dpi_D1_vs_D0\", test=\"Wald\") %>% as.data.frame() %>% dplyr::mutate(ihw=ihw(pvalue ~ baseMean,  data=., alpha=0.1) %>% adj_pvalues(.)),\n",
    "        results(dds, name=\"dpi_D3_vs_D0\", test=\"Wald\") %>% as.data.frame() %>% dplyr::mutate(ihw=ihw(pvalue ~ baseMean,  data=., alpha=0.1) %>% adj_pvalues(.)),\n",
    "        results(dds, name=\"dpi_D6_vs_D0\", test=\"Wald\") %>% as.data.frame() %>% dplyr::mutate(ihw=ihw(pvalue ~ baseMean,  data=., alpha=0.1) %>% adj_pvalues(.))\n",
    "\n",
    "    )\n",
    "    \n",
    "    # LRT IFNARKO\n",
    "    message(\"LRT KO\")\n",
    "    \n",
    "    dds <- DESeqDataSetFromMatrix(countData=data[[\"cnt\"]][, data[[\"meta\"]]$genotype==\"IFNAR_fl_LysM_cre\"], colData=data[[\"meta\"]][data[[\"meta\"]]$genotype==\"IFNAR_fl_LysM_cre\", ], design=~dpi)\n",
    "    dds <- DESeq(dds, test=\"LRT\", full=~dpi, reduced=~1)\n",
    "    \n",
    "    res[[\"dpi_ko_lrt\"]] <- results(dds) %>% as.data.frame() %>% dplyr::mutate(ihw=ihw(pvalue ~ baseMean,  data=., alpha=0.1) %>% adj_pvalues(.))\n",
    "    \n",
    "    # Wald IFNARKO\n",
    "    message(\"Wald IFNARKO\")\n",
    "    \n",
    "    dds <- DESeqDataSetFromMatrix(countData=data[[\"cnt\"]][, data[[\"meta\"]]$genotype==\"IFNAR_fl_LysM_cre\"], colData=data[[\"meta\"]][data[[\"meta\"]]$genotype==\"IFNAR_fl_LysM_cre\", ], design=~dpi)\n",
    "    dds <- DESeq(dds, test=\"Wald\")\n",
    "    \n",
    "    res[[\"dpi_ko_wald\"]] <- list(\n",
    "\n",
    "        results(dds, name=\"dpi_D1_vs_D0\", test=\"Wald\") %>% as.data.frame() %>% dplyr::mutate(ihw=ihw(pvalue ~ baseMean,  data=., alpha=0.1) %>% adj_pvalues(.)),\n",
    "        results(dds, name=\"dpi_D3_vs_D0\", test=\"Wald\") %>% as.data.frame() %>% dplyr::mutate(ihw=ihw(pvalue ~ baseMean,  data=., alpha=0.1) %>% adj_pvalues(.)),\n",
    "        results(dds, name=\"dpi_D6_vs_D0\", test=\"Wald\") %>% as.data.frame() %>% dplyr::mutate(ihw=ihw(pvalue ~ baseMean,  data=., alpha=0.1) %>% adj_pvalues(.))\n",
    "\n",
    "    )\n",
    "    \n",
    "    # LRT WT vs KO\n",
    "    message(\"LRT WT vs KO\")\n",
    "    \n",
    "    dds <- DESeqDataSetFromMatrix(countData=data[[\"cnt\"]], colData=data[[\"meta\"]], design=~genotype + dpi + genotype:dpi)\n",
    "    dds <- DESeq(dds, test=\"LRT\", full=~genotype + dpi + genotype:dpi, reduced=~genotype+dpi)\n",
    "    \n",
    "    res[[\"dpi_wt_vs_ko_lrt\"]] <- results(dds) %>% as.data.frame() %>% dplyr::mutate(ihw=ihw(pvalue ~ baseMean,  data=., alpha=0.1) %>% adj_pvalues(.))\n",
    "    \n",
    "    # Wald WT vs KO\n",
    "    message(\"Wald WT vs KO\")\n",
    "    \n",
    "    dds <- DESeqDataSetFromMatrix(countData=data[[\"cnt\"]], colData=data[[\"meta\"]], design=~genotype + dpi + genotype:dpi)\n",
    "    dds$group <- factor(paste(dds$genotype, dds$dpi, sep = \"_\"))\n",
    "    design(dds) <- ~ group\n",
    "    dds$group <- relevel(dds$group, ref = \"IFNAR_fl_D0\")\n",
    "    dds <- DESeq(dds, test=\"Wald\")\n",
    "    \n",
    "    res[[\"dpi_wt_vs_ko_wald\"]] <- list(\n",
    "        \n",
    "        results(dds, contrast = c(\"group\", \"IFNAR_fl_LysM_cre_D0\", \"IFNAR_fl_D0\"), test=\"Wald\") %>% as.data.frame() %>% dplyr::mutate(ihw=ihw(pvalue ~ baseMean,  data=., alpha=0.1) %>% adj_pvalues(.)), \n",
    "        results(dds, contrast = c(\"group\", \"IFNAR_fl_LysM_cre_D1\", \"IFNAR_fl_D1\"), test=\"Wald\") %>% as.data.frame() %>% dplyr::mutate(ihw=ihw(pvalue ~ baseMean,  data=., alpha=0.1) %>% adj_pvalues(.)),\n",
    "        results(dds, contrast = c(\"group\", \"IFNAR_fl_LysM_cre_D3\", \"IFNAR_fl_D3\"), test=\"Wald\") %>% as.data.frame() %>% dplyr::mutate(ihw=ihw(pvalue ~ baseMean,  data=., alpha=0.1) %>% adj_pvalues(.)),\n",
    "        results(dds, contrast = c(\"group\", \"IFNAR_fl_LysM_cre_D6\", \"IFNAR_fl_D6\"), test=\"Wald\") %>% as.data.frame() %>% dplyr::mutate(ihw=ihw(pvalue ~ baseMean,  data=., alpha=0.1) %>% adj_pvalues(.))\n",
    "\n",
    "    )\n",
    "    \n",
    "    return(res)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14f92344-cf18-45c6-b938-fce13ab2badd",
   "metadata": {},
   "outputs": [],
   "source": [
    "response_class <- function(res_lrt, res_wald, padj_thr) {\n",
    "    \n",
    "    # Annotate response genes from LRT \n",
    "    res_lrt <- res_lrt %>% as.data.frame() %>% \n",
    "        dplyr::mutate(response_gene_lrt=ifelse(ihw <= padj_thr, TRUE, FALSE)) %>% \n",
    "        dplyr::rename(ihw_lrt=ihw, log2FoldChange_lrt=log2FoldChange) %>% \n",
    "        dplyr::select(ihw_lrt, log2FoldChange_lrt, response_gene_lrt) %>% \n",
    "        tibble::rownames_to_column(\"gene\")\n",
    "    \n",
    "    # Prepare data to annotate response class    \n",
    "    res_wald <- rbind(\n",
    "        \n",
    "        res[[\"dpi_wt_wald\"]][[1]] %>% as.data.frame() %>% dplyr::mutate(dpi=\"D1\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "        res[[\"dpi_wt_wald\"]][[2]] %>% as.data.frame() %>% dplyr::mutate(dpi=\"D3\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "        res[[\"dpi_wt_wald\"]][[3]] %>% as.data.frame() %>% dplyr::mutate(dpi=\"D6\") %>% tibble::rownames_to_column(\"gene\")\n",
    "        \n",
    "    ) \n",
    "    \n",
    "    res_wald <- res_wald %>% \n",
    "        dplyr::mutate(dpi=factor(dpi, levels=c(\"D1\", \"D3\", \"D6\"))) \n",
    "    \n",
    "    # Set log2FoldChanges for Wald test\n",
    "    res_wald <- res_wald %>% \n",
    "        dplyr::group_by(gene) %>% \n",
    "        dplyr::mutate(response_gene_wald=ifelse(any(ihw<=padj_thr), TRUE, FALSE)) %>%\n",
    "        dplyr::mutate(log2FoldChange=ifelse(response_gene_wald & ihw<=padj_thr, log2FoldChange, ifelse(ihw==min(padj), log2FoldChange, 0))) %>%  \n",
    "        dplyr::ungroup() %>% as.data.frame()\n",
    "    \n",
    "    # Annotate response class from Wald test\n",
    "    response_class_wald <- res_wald %>% \n",
    "        reshape2::dcast(., gene ~ dpi, value.var=\"log2FoldChange\") %>%\n",
    "        dplyr::mutate(response_class=NA) %>% \n",
    "        dplyr::mutate(response_class=ifelse(rowSums(.[2:4]>=0)==3, \"Activated\", response_class)) %>% \n",
    "        dplyr::mutate(response_class=ifelse(rowSums(.[2:4]<=0)==3, \"Repressed\", response_class)) %>%\n",
    "        dplyr::mutate(response_class=ifelse(rowSums(.[2:4]>=0)>0 & rowSums(.[2:4]<=0)>0 & is.na(response_class), \"Periodical\", response_class)) %>% \n",
    "        # dplyr::mutate(response_class=ifelse(.[4, drop=FALSE]!=0, paste(response_class, \"(monoton)\"), paste(response_class, \"(transient)\"))) %>% \n",
    "        dplyr::mutate(response_class=as.character(response_class))\n",
    "    \n",
    "    # Combine results \n",
    "    result <- list(response_class_wald, res_wald[, c(\"gene\", \"response_gene_wald\")], res_lrt[, c(\"gene\", \"response_gene_lrt\")]) %>% \n",
    "        purrr::reduce(full_join, by=\"gene\") %>% unique() \n",
    "    \n",
    "    # Set response gene \n",
    "    result <- result %>% \n",
    "        dplyr::mutate(response_gene=ifelse(response_gene_lrt & response_gene_wald, TRUE, FALSE))\n",
    "    \n",
    "    # Set rownames to gene column and remove  \n",
    "    rownames(result) <- result$gene\n",
    "    result$gene <- NULL\n",
    "    \n",
    "    return(result)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e44eeb24-5a62-41f0-a6f1-b423c475fae9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "perturbation_class <- function(res_wald, rc_wt, rc_ko, padj_thr) {\n",
    "    \n",
    "    rc_wt <- rc_wt %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, response_class, response_gene) %>% dplyr::rename(response_class_wt=response_class, response_gene_wt=response_gene)\n",
    "    rc_ko <- rc_ko %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, response_class, response_gene) %>% dplyr::rename(response_class_ko=response_class, response_gene_ko=response_gene)\n",
    "    \n",
    "    # Prepare data to annotate response class    \n",
    "    res_wald <- rbind(\n",
    "        \n",
    "        res_wald[[1]] %>% as.data.frame() %>% dplyr::mutate(dpi=\"D0\") %>% tibble::rownames_to_column(\"gene\"),\n",
    "        res_wald[[2]] %>% as.data.frame() %>% dplyr::mutate(dpi=\"D1\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "        res_wald[[3]] %>% as.data.frame() %>% dplyr::mutate(dpi=\"D3\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "        res_wald[[4]] %>% as.data.frame() %>% dplyr::mutate(dpi=\"D6\") %>% tibble::rownames_to_column(\"gene\")\n",
    "        \n",
    "    )\n",
    "    \n",
    "    res_wald <- res_wald %>% \n",
    "        dplyr::mutate(dpi=factor(dpi, levels=c(\"D0\", \"D1\", \"D3\", \"D6\")))\n",
    "    \n",
    "    # Set log2FoldChanges for Wald test\n",
    "    res_wald <- res_wald %>% \n",
    "        dplyr::group_by(gene) %>% \n",
    "        dplyr::mutate(perturbation_gene=ifelse(any(ihw<=padj_thr), TRUE, FALSE)) %>% \n",
    "        dplyr::mutate(log2FoldChange=ifelse(perturbation_gene & ihw<=padj_thr, log2FoldChange, ifelse(ihw==min(ihw), log2FoldChange, 0))) %>% \n",
    "        dplyr::ungroup() %>% as.data.frame() %>% \n",
    "        reshape2::dcast(., perturbation_gene + gene ~ dpi, value.var=\"log2FoldChange\") \n",
    "    \n",
    "    result <- list(res_wald, rc_wt, rc_ko) %>% purrr::reduce(full_join, by=\"gene\") %>% dplyr::distinct() %>%\n",
    "        dplyr::mutate(response_gene_wt=ifelse(is.na(response_gene_wt), FALSE, response_gene_wt)) %>% \n",
    "        dplyr::mutate(response_gene_ko=ifelse(is.na(response_gene_ko), FALSE, response_gene_ko)) %>% \n",
    "        relocate(perturbation_gene, .after=response_gene_ko)\n",
    "    \n",
    "    # Annotate perturbation \n",
    "    result <- result %>% \n",
    "        dplyr::mutate(perturbation_class=ifelse(perturbation_gene, \"Perturbed\", \"Canonical\")) %>% \n",
    "        tibble::column_to_rownames(\"gene\")\n",
    "    \n",
    "    return(result)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d96e9be7-3f88-403a-8886-f1cf9b2021b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_class <- function(res, padj_rc_thr, padj_pc_thr) {\n",
    "    \n",
    "    rc_wt <- response_class(res[[\"dpi_wt_lrt\"]], res[[\"dpi_wt_wald\"]], padj_thr=padj_rc_thr)\n",
    "    rc_ko <- response_class(res[[\"dpi_ko_lrt\"]], res[[\"dpi_ko_wald\"]], padj_thr=padj_rc_thr)\n",
    "    pc_wt_vs_ko <- perturbation_class(res[[\"dpi_wt_vs_ko_wald\"]], rc_wt, rc_ko, padj_thr=padj_pc_thr)\n",
    "    \n",
    "    return(list(\"rc_wt\"=rc_wt, \"rc_ko\"=rc_ko, \"pc_wt_vs_ko\"=pc_wt_vs_ko))\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff2f1cf5-0254-4d2b-a037-f5df86790fa2",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_response_class <- function(meta, cnt, result, top=NULL, response_class_n=NULL, title=NULL) {\n",
    "\n",
    "    # Remove failed tests and filter for response genes \n",
    "    result <- na.omit(result)\n",
    "    result <- result[result$response_gene, ]    \n",
    "    \n",
    "    # Subset count data by genes in result \n",
    "    cnt <- cnt[rownames(cnt) %in% rownames(result), ]\n",
    "\n",
    "    # Get count data for top results\n",
    "    mat <- cnt[rownames(result), ]\n",
    "\n",
    "    # Normalize and scale data\n",
    "    mat <- edgeR::cpm(mat, normalized.lib.sizes=TRUE, log=FALSE)\n",
    "    mat <- t(scale(t(mat), center=TRUE, scale=TRUE))\n",
    "\n",
    "    # Order meta by factor levels \n",
    "    meta <- dplyr::arrange(meta, dpi, genotype)\n",
    "\n",
    "    # Order mat by meta \n",
    "    mat <- mat[, rownames(meta)]\n",
    "    \n",
    "    # Subset and order mat by response class \n",
    "    result$response_class <- factor(result$response_class, levels=c(\"Activated\", \"Repressed\", \"Periodical\"))\n",
    "    \n",
    "    # Color heatmap cnt\n",
    "    color_ramp_mat <- c(rev(brewer.pal(11,\"RdBu\"))[1:5], \"#ffffff\", rev(brewer.pal(11,\"RdBu\"))[7:11])\n",
    "    breaks_mat <- seq(-max(abs(mat)), max(abs(mat)), length.out=length(color_ramp_mat))\n",
    "    color_function_mat <- circlize::colorRamp2(breaks_mat, color_ramp_mat) \n",
    "    \n",
    "    # Top annotation group \n",
    "    top_annotation <- HeatmapAnnotation(\n",
    "\n",
    "        df=data.frame(\n",
    "            \n",
    "            top_annotation_dpi=meta$dpi, \n",
    "            top_annotation_genotype=meta$genotype\n",
    "            \n",
    "\n",
    "        ), \n",
    "\n",
    "        annotation_legend_param=list(\n",
    "\n",
    "            top_annotation_dpi=list(title=\"DPI\", title_gp=gpar(fontsize=18, fontface=\"plain\"), labels_gp=gpar(fontsize=16), grid_width=unit(6, \"mm\"), grid_height=unit(6, \"mm\")), \n",
    "            top_annotation_genotype=list(title=\"Infection\", title_gp=gpar(fontsize=18, fontface=\"plain\"), labels_gp=gpar(fontsize=16), grid_width=unit(6, \"mm\"), grid_height=unit(6, \"mm\"))\n",
    "\n",
    "        ), \n",
    "\n",
    "        col=list(\n",
    "\n",
    "            top_annotation_dpi=color$dpi, \n",
    "            top_annotation_genotype=color$genotype \n",
    "\n",
    "        ), \n",
    "\n",
    "        simple_anno_size=unit(6, \"mm\"), \n",
    "        show_annotation_name=FALSE, \n",
    "        show_legend=TRUE\n",
    "\n",
    "    )\n",
    "\n",
    "    hm <- ComplexHeatmap::Heatmap(\n",
    "\n",
    "        matrix=mat, \n",
    "        name=\"z-score\",  \n",
    "\n",
    "        col=color_function_mat, \n",
    "        na_col=\"white\", \n",
    "\n",
    "        width=unit(5*ncol(mat), \"mm\"), \n",
    "        height=unit(5*100, \"mm\"), \n",
    "\n",
    "        column_title=title, \n",
    "        column_title_gp=gpar(fontsize=18, fontface=\"bold\"), \n",
    "\n",
    "        row_names_gp=gpar(fontsize=16, fontface=\"plain\"), \n",
    "        column_names_gp=gpar(fontsize=18, fontface=\"plain\"), \n",
    "\n",
    "        cluster_rows=TRUE, \n",
    "        cluster_row_slices =FALSE, \n",
    "        show_row_dend=FALSE,    \n",
    "        row_split=result$response_class,\n",
    "        row_title_gp=gpar(fontsize=18, fontface=\"plain\"),\n",
    "        row_gap=unit(5, \"mm\"),\n",
    "        show_row_names=ifelse(is.null(top), FALSE, TRUE),\n",
    "        row_names_side=\"left\",\n",
    "\n",
    "        cluster_columns=FALSE,\n",
    "        show_column_dend=FALSE, \n",
    "        column_split=factor(meta$dpi),\n",
    "        column_gap=unit(2, \"mm\"), \n",
    "        show_column_names=FALSE, \n",
    "\n",
    "        top_annotation=top_annotation, \n",
    "\n",
    "        rect_gp=gpar(col=\"white\", lwd=0, alpha=1), \n",
    "\n",
    "        heatmap_legend_param=list(title_gp=gpar(fontsize=18, fontface=\"plain\"), labels_gp=gpar(fontsize=16), legend_height=unit(5*6, \"mm\"), grid_width=unit(6, \"mm\")), \n",
    "\n",
    "        border=TRUE, \n",
    "        border_gp=gpar(col=\"black\", lwd=unit(0.5, \"mm\")), \n",
    "\n",
    "        use_raster=FALSE\n",
    "\n",
    "    )\n",
    "    \n",
    "    return(hm)\n",
    "       \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1e90a96-2ff4-46a8-bd4e-38d091cee98d",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_perturbation <- function(meta, cnt, result, title=NULL) {\n",
    "    \n",
    "    # Subset to genes that are resonse genes in either group \n",
    "    result <- result[result$response_gene_wt | result$response_gene_ko, ]\n",
    "    \n",
    "    # Subset count data by genes in result \n",
    "    cnt <- cnt[rownames(cnt) %in% rownames(result), ]\n",
    "    \n",
    "    # Get count data for top results\n",
    "    mat <- cnt[rownames(result), ]\n",
    "\n",
    "    # Normalize and scale data\n",
    "    mat <- edgeR::cpm(mat, normalized.lib.sizes=TRUE, log=FALSE)\n",
    "    mat <- t(scale(t(mat), center=TRUE, scale=TRUE))\n",
    "    \n",
    "    # Order meta by factor levels \n",
    "    meta <- dplyr::arrange(meta, dpi, genotype)\n",
    "    \n",
    "    # Order mat by meta \n",
    "    mat <- mat[, rownames(meta)]\n",
    "    \n",
    "    # Subset and order mat by response class \n",
    "    result$perturbation_class <- factor(result$perturbation_class, levels=c(\"Canonical\", \"Perturbed\"))\n",
    "    \n",
    "    # Color heatmap cnt\n",
    "    color_ramp_mat <- c(rev(brewer.pal(11,\"RdBu\"))[1:5], \"#ffffff\", rev(brewer.pal(11,\"RdBu\"))[7:11])\n",
    "    breaks_mat <- seq(-max(abs(mat)), max(abs(mat)), length.out=length(color_ramp_mat))\n",
    "    color_function_mat <- circlize::colorRamp2(breaks_mat, color_ramp_mat) \n",
    "    \n",
    "    # Top annotation group \n",
    "    top_annotation <- HeatmapAnnotation(\n",
    "\n",
    "        df=data.frame(\n",
    "            \n",
    "            top_annotation_dpi=meta$dpi, \n",
    "            top_annotation_genotype=meta$genotype\n",
    "            \n",
    "\n",
    "        ), \n",
    "\n",
    "        annotation_legend_param=list(\n",
    "\n",
    "            top_annotation_dpi=list(title=\"DPI\", title_gp=gpar(fontsize=18, fontface=\"plain\"), labels_gp=gpar(fontsize=16), grid_width=unit(6, \"mm\"), grid_height=unit(6, \"mm\")), \n",
    "            top_annotation_genotype=list(title=\"Infection\", title_gp=gpar(fontsize=18, fontface=\"plain\"), labels_gp=gpar(fontsize=16), grid_width=unit(6, \"mm\"), grid_height=unit(6, \"mm\"))\n",
    "\n",
    "        ), \n",
    "\n",
    "        col=list(\n",
    "\n",
    "            top_annotation_dpi=color$dpi, \n",
    "            top_annotation_genotype=color$genotype \n",
    "\n",
    "        ), \n",
    "\n",
    "        simple_anno_size=unit(6, \"mm\"), \n",
    "        show_annotation_name=FALSE, \n",
    "        show_legend=TRUE\n",
    "\n",
    "    )\n",
    "\n",
    "    hm <- ComplexHeatmap::Heatmap(\n",
    "\n",
    "        matrix=mat, \n",
    "        name=\"z-score\",  \n",
    "\n",
    "        col=color_function_mat, \n",
    "        na_col=\"white\", \n",
    "\n",
    "        width=unit(5*ncol(mat), \"mm\"), \n",
    "        height=unit(5*100, \"mm\"), \n",
    "\n",
    "        column_title=title, \n",
    "        column_title_gp=gpar(fontsize=18, fontface=\"bold\"), \n",
    "\n",
    "        row_names_gp=gpar(fontsize=16, fontface=\"plain\"), \n",
    "        column_names_gp=gpar(fontsize=18, fontface=\"plain\"), \n",
    "\n",
    "        cluster_rows=TRUE, \n",
    "        show_row_dend=FALSE,    \n",
    "        row_split=result$perturbation_class,\n",
    "        row_title_gp=gpar(fontsize=18, fontface=\"plain\"),\n",
    "        row_gap=unit(5, \"mm\"),\n",
    "        show_row_names=FALSE,\n",
    "        row_names_side=\"left\",\n",
    "\n",
    "        cluster_columns=FALSE,\n",
    "        show_column_dend=FALSE, \n",
    "        column_split=factor(meta$dpi),\n",
    "        column_gap=unit(2, \"mm\"), \n",
    "        show_column_names=FALSE, \n",
    "\n",
    "        top_annotation=top_annotation, \n",
    "        # right_annotation=row_annotation, \n",
    "\n",
    "        rect_gp=gpar(col=\"white\", lwd=0, alpha=1), \n",
    "\n",
    "        heatmap_legend_param=list(title_gp=gpar(fontsize=18, fontface=\"plain\"), labels_gp=gpar(fontsize=16), legend_height=unit(5*6, \"mm\"), grid_width=unit(6, \"mm\")), \n",
    "\n",
    "        border=TRUE, \n",
    "        border_gp=gpar(col=\"black\", lwd=unit(0.5, \"mm\")), \n",
    "\n",
    "        use_raster=FALSE\n",
    "\n",
    "    )\n",
    "    \n",
    "    return(hm)\n",
    "       \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed044a4c-fc15-4e4a-865f-eba5083a1757",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "gsea <- function(result, category=\"H\", subcategory=NULL) {\n",
    "    \n",
    "    # Get gene set\n",
    "    gene_set <- msigdbr::msigdbr(species=\"mouse\", category=category, subcategory=subcategory)\n",
    "    gene_set <- split(gene_set, x=gene_set$gene_symbol, f=gene_set$gs_name)\n",
    "    \n",
    "    # Remove failed comparison \n",
    "    result <- na.omit(result)\n",
    "    \n",
    "    # Set gene names \n",
    "    result$mgi_symbol <- rownames(result)\n",
    "    \n",
    "    # Make ranks \n",
    "    result$ihw <- ifelse(result$ihw==0, min(result$ihw[result$ihw>0]), result$ihw)\n",
    "    result$sign_log_adj_p_values <- -log10(result$ihw) * sign(result$log2FoldChange)\n",
    "    \n",
    "    ranks <- result$sign_log_adj_p_values\n",
    "    names(ranks) <- result$mgi_symbol\n",
    "    ranks <- ranks[order(ranks)]\n",
    "    ranks <- rev(ranks)\n",
    "    \n",
    "    # Retain only pathways that overlap with result lsit\n",
    "    gene_set_filter <- lapply(gene_set, function(x) {sum(result$mgi_symbol %in% x)>=1})\n",
    "    gene_set <- gene_set[unlist(gene_set_filter)]\n",
    "\n",
    "    gsea <- fgsea::fgsea(\n",
    "        \n",
    "        pathways=gene_set,\n",
    "        stats=ranks,\n",
    "        nperm=100000, \n",
    "        minSize=5,\n",
    "        maxSize=500\n",
    "        \n",
    "    )\n",
    "    \n",
    "    return(gsea)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0272459d-33b8-4d71-86e3-b00c641ce6bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_plot <- function(gsea, pval_thr, title=NULL, color=c(RColorBrewer::brewer.pal(8, \"Set1\")[1], RColorBrewer::brewer.pal(8, \"Set1\")[2]), color_names=c(\"Pos\", \"Neg\"), size_range=5, pathway_suffix=NULL, top=20) {\n",
    "    \n",
    "    # Set GSEA data frame \n",
    "    gsea <- as.data.frame(gsea)\n",
    "    gsea <- na.omit(gsea) \n",
    "    \n",
    "    # Set color names \n",
    "    color <- setNames(color, color_names)\n",
    "    \n",
    "    # Fix pathway names\n",
    "    if(!is.null(pathway_suffix)) {gsea$pathway <- gsub(pathway_suffix, \"\", gsea$pathway)}\n",
    "    gsea$pathway <- gsub(\"_\", \" \", gsea$pathway)\n",
    "    \n",
    "    # Filter hits \n",
    "    gsea_up <- gsea[sign(gsea$NES)==+1, ]\n",
    "    gsea_down <- gsea[sign(gsea$NES)==-1, ]\n",
    "    \n",
    "    gsea_up <- gsea_up[order(gsea_up$pval), ][1:top, ]\n",
    "    gsea_down <- gsea_down[order(gsea_down$pval), ][1:top, ]\n",
    "    \n",
    "    gsea <- rbind(gsea_up, gsea_down)\n",
    "    gsea <- na.omit(gsea)\n",
    "    gsea <- distinct(gsea)\n",
    "    \n",
    "    # Add color \n",
    "    gsea$color <- ifelse(sign(gsea$ES)==1, names(color)[1], names(color)[2])\n",
    "    gsea$color <- ifelse(gsea$pval<=pval_thr, gsea$color, NA)\n",
    "\n",
    "    gsea$sign_log_pval_values <- -log10(gsea$pval) * sign(gsea$ES)\n",
    "    \n",
    "    # Order hits \n",
    "    gsea <- gsea[rev(order(gsea$sign_log_pval_values)), ]\n",
    "    gsea$pathway <- factor(gsea$pathway, levels=rev(gsea$pathway))\n",
    "\n",
    "    x_max <- max(abs(gsea$sign_log_pval_values))\n",
    "    if(x_max<abs(log10(pval_thr))) {x_max <- abs(log10(pval_thr))}\n",
    "    x_max <- ceiling(ceiling(x_max))\n",
    "    \n",
    "    # Plot \n",
    "    plot <- ggplot(gsea, aes(x=sign_log_pval_values, y=pathway, color=color)) + \n",
    "        \n",
    "        geom_vline(xintercept=-log10(pval_thr), linetype=\"dashed\") + \n",
    "        geom_vline(xintercept=log10(pval_thr), linetype=\"dashed\") +\n",
    "    \n",
    "        geom_point(aes(size=abs(NES))) +\n",
    "\n",
    "        ggtitle(title) +\n",
    "        xlab(\"Signed -log10 adj. p-value\") + ylab(\"\") + \n",
    "        scale_x_continuous(breaks=c(-x_max, 0, x_max), limits=c(-x_max, x_max)) +\n",
    "        scale_color_manual(values=color, na.value=\"black\", drop=FALSE) +\n",
    "        scale_size(range=c(0, size_range)) + \n",
    "        guides(\n",
    "            \n",
    "            color=guide_legend(order=1, title=\"Agent\", keywidth=0.75, keyheight=0.75, override.aes=list(size=4)), \n",
    "            size=guide_legend(order=2, title=\"Abs. (NES)\", keywidth=0.75, keyheight=0.75)\n",
    "            \n",
    "        ) +\n",
    "    \n",
    "        theme(\n",
    "            \n",
    "            legend.position=\"right\", \n",
    "            legend.justification=\"top\", \n",
    "            axis.text.y=element_text(size=14, hjust=1, vjust=0.5, face=\"plain\", margin=margin(t=0, r=2, b=0, l=0), color=\"black\")\n",
    "            \n",
    "        ) \n",
    "    \n",
    "    return(plot)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3f7b37d-ad20-4bce-9b3d-6025ffa1bacb",
   "metadata": {},
   "outputs": [],
   "source": [
    "dea_vp <- function(dea, log2fc_thr=0, p_adj_thr=0.05, top_label=25, label=NULL, label_merge=FALSE, title=NULL, color_neg=RColorBrewer::brewer.pal(8, \"Set1\")[1], color_pos=RColorBrewer::brewer.pal(8, \"Set1\")[2]) {\n",
    "\n",
    "    label <- c(\"Icam1\", \"Itga4\", \"Itgb1\", \"Vcam1\", \"Gdf15\", \"Ccl2\", \"Tnf\", \"Il6\", \"Tgfb1\", \"Stat1\", \"Stat2\", \"Spic\", \"Irf8\", \"Siglec1\", \"Slc40a1\", \"Hmox1\", \"Blvrb\", \"Treml4\")\n",
    "    \n",
    "    # Set class\n",
    "    dea <- as.data.frame(dea)\n",
    "    \n",
    "    # Set rownames to genes\n",
    "    if(\"gene\" %in% colnames(dea)) {rownames(dea) <- dea$gene}\n",
    "    \n",
    "    dea <- na.omit(dea)\n",
    "    \n",
    "    # Annotate entries significance by log2fc_thr and p_adj_thr\n",
    "    dea$ihw <- ifelse(dea$ihw == 0, .Machine$double.xmin, dea$ihw)\n",
    "    dea$sig <- ifelse(abs(dea$log2FoldChange) >= log2fc_thr & -log10(dea$ihw) >= -log10(p_adj_thr), \"s\", \"ns\")\n",
    "    \n",
    "    # Set color based on significance and direction of dea e.g. positive and negative \n",
    "    dea$color <- ifelse(dea$sig == \"s\" & dea$log2FoldChange > 0, \"s_pos\", \"ns\")\n",
    "    dea$color <- ifelse(dea$sig == \"s\" & dea$log2FoldChange < 0, \"s_neg\", dea$color)\n",
    "    \n",
    "    color <- c(color_neg, \"gray\", \"black\", color_pos)\n",
    "    names(color) <- c(\"s_pos\", \"ns\", \"black\", \"s_neg\")\n",
    "    \n",
    "    # Create labels based log2FC and p_val_adj\n",
    "    dea_pos <- dea[dea$log2FoldChange > 0 & dea$sig == \"s\", ]\n",
    "    dea_neg <- dea[dea$log2FoldChange < 0 & dea$sig == \"s\", ]\n",
    "\n",
    "    pos_labels <- dea_pos[rev(order(dea_pos$log2FoldChange)), ] %>% rownames()\n",
    "    neg_labels <- dea_neg[order(dea_neg$log2FoldChange), ] %>% rownames()\n",
    "    \n",
    "    pos_top_labels <- dea_pos[rev(order(dea_pos$log2FoldChange)), ][1:top_label, ] %>% rownames()\n",
    "    neg_top_labels <- dea_neg[order(dea_neg$log2FoldChange), ][1:top_label, ] %>% rownames()\n",
    "    \n",
    "    # Set labels \n",
    "    if(is.null(label)) {\n",
    "\n",
    "        label_select <- c(pos_top_labels, neg_top_labels)\n",
    "        \n",
    "        \n",
    "    } else if (!is.null(label) & !label_merge) {\n",
    "\n",
    "        label_select <- label[label %in% c(pos_labels, neg_labels)]\n",
    "        \n",
    "    } else if (!is.null(label) & label_merge) {\n",
    "\n",
    "        label_select <- c(label[label %in% c(pos_labels, neg_labels)], c(pos_top_labels, neg_top_labels)) %>% unique()\n",
    "        \n",
    "    }\n",
    "\n",
    "    dea$label <- ifelse(rownames(dea) %in% label_select, rownames(dea), NA)\n",
    "\n",
    "    # Plot\n",
    "    vp <- ggplot(dea, aes(x=log10(baseMean), y=log2FoldChange, fill=dea$color, label=label), alpha=1) + \n",
    "    \n",
    "        geom_point(size=4, shape=21, color=\"white\") + \n",
    "        geom_hline(aes(yintercept=0), linetype=\"dotted\", colour=\"black\") +\n",
    "        ggrepel::geom_text_repel(segment.color=\"black\", force=20, force_pull=1, max.overlaps=getOption(\"ggrepel.max.overlaps\", default=100), size=8, alpha=1, guide=\"none\", segment.size=0.1, color=\"black\") + \n",
    "        ylim(-max(abs(dea$log2FoldChange)), max(abs(dea$log2FoldChange))) + \n",
    "        ggtitle(title) + xlab(\"log10(base expression)\") + ylab(\"log2FC\") + \n",
    "        scale_fill_manual(values=color) + \n",
    "    \n",
    "        guides(\n",
    "            \n",
    "            color=guide_legend(order=1, title=\"Group\", size=2, keywidth=0.75, keyheight=0.75), \n",
    "            alpha=\"none\"\n",
    "            \n",
    "        ) + \n",
    "    \n",
    "        theme(\n",
    "\n",
    "            legend.position=\"none\", \n",
    "            aspect.ratio=1\n",
    "\n",
    "        )\n",
    "    \n",
    "    return(vp)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22c6ded7-5444-4f34-9b24-fb1118d8509f",
   "metadata": {},
   "source": [
    "# DEA cMo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b498fa71-bb62-4594-9d7e-e4a460f9c313",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset data \n",
    "data <- data_subset(meta, cnt, group_by=\"celltype\", group=\"cMo\", min_count=10, min_sample=3, min_group=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33095af9-a09e-4adf-bc8d-5f4406fd70cf",
   "metadata": {},
   "source": [
    "## DDS and results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50d791a4-d79a-4159-b138-4d9ac190cebf",
   "metadata": {},
   "outputs": [],
   "source": [
    "res <- suppressMessages(lrt(data))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b18870fa-d87d-40a4-abfa-983d763b169e",
   "metadata": {},
   "source": [
    "## Result class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a2cc99a-8716-448a-ab50-bf73a8207e71",
   "metadata": {},
   "outputs": [],
   "source": [
    "resc <- result_class(res, padj_rc_thr=0.01, padj_pc_thr=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aed28f1d-aeb3-41ec-8c39-587cf65a4aa9",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=30, repr.plot.height=25)\n",
    "\n",
    "hm_1 <- hm_response_class(data[[\"meta\"]][data[[\"meta\"]]$genotype==\"IFNAR_fl\", ], data[[\"cnt\"]][, data[[\"meta\"]]$genotype==\"IFNAR_fl\"], resc[[\"rc_wt\"]], response_class_n=10, title=\"Response genes\\n cMo (IFNAR1 fl/fl)\") %>% ggplotify::as.ggplot()\n",
    "hm_2 <- hm_response_class(data[[\"meta\"]][data[[\"meta\"]]$genotype==\"IFNAR_fl_LysM_cre\", ], data[[\"cnt\"]][, data[[\"meta\"]]$genotype==\"IFNAR_fl_LysM_cre\"], resc[[\"rc_ko\"]], response_class_n=10, title=\"Response genes\\n cMo (IFNAR1 fl/fl LyzMCre wt/+)\") %>% ggplotify::as.ggplot()\n",
    "hm_3 <- hm_perturbation(data[[\"meta\"]], data[[\"cnt\"]], resc[[\"pc_wt_vs_ko\"]], title=\"Perturbation genes\\n cMo\") %>% ggplotify::as.ggplot()\n",
    "\n",
    "hm_1 + hm_2 + hm_3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e647d2a-6f01-4430-b3bf-580b47389a37",
   "metadata": {},
   "source": [
    "## GSEA DPI between infection types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "705e65bd-ead7-4be7-9286-2594466e2bee",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=4*10, repr.plot.height=12)\n",
    "\n",
    "gsea_1 <- lapply(res[[\"dpi_wt_vs_ko_wald\"]], function(result) {gsea(result)})\n",
    "gsea_plot_1 <- lapply(1:length(gsea_1), function(i) {gsea_plot(gsea_1[[i]], pathway_suffix=\"HALLMARK_\", color_names=c(\"IFNAR_fl_LysM_cre\", \"IFNAR_fl\"), title=c(\"D0\", \"D1\", \"D3\", \"D6\")[i], pval_thr=0.05)})\n",
    "\n",
    "patchwork::wrap_plots(gsea_plot_1, ncol=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "239ddb44-d7d9-4f4d-a1c4-a8036b8ea227",
   "metadata": {},
   "source": [
    "## Volcano DPI between infection types "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b90d77e-7a2e-4700-9785-1d0407851eac",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=4*9, repr.plot.height=12)\n",
    "\n",
    "v_plot <- lapply(1:length(res[[\"dpi_wt_vs_ko_wald\"]]), function(i) {dea_vp(res[[\"dpi_wt_vs_ko_wald\"]][[i]], title=c(\"D0\", \"D1\", \"D3\", \"D6\")[i], p_adj_thr=0.05)})\n",
    "patchwork::wrap_plots(v_plot, ncol=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "927da728-6258-4820-92c2-ff6a62ef90f1",
   "metadata": {},
   "source": [
    "## Save results "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd091097-34de-4c23-99b0-d339e429a297",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(list(data=data, res=res, resc=resc), \"result/dea/bulkRNAseq/res_dea_cmo.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "890cce1b-5e61-4b26-8550-208c3ec89f61",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Wald test genotype \n",
    "res[[\"dpi_wt_vs_ko_wald\"]] <- lapply(res[[\"dpi_wt_vs_ko_wald\"]], function(x) {x %>% tibble::rownames_to_column(\"gene\")})\n",
    "names(res[[\"dpi_wt_vs_ko_wald\"]]) <- c(\"D0\", \"D1\", \"D3\", \"D6\")\n",
    "openxlsx::write.xlsx(res[[\"dpi_wt_vs_ko_wald\"]], \"result/dea/bulkRNAseq/res_dea_genotpye_cmo.xlsx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55d115af-378c-46ad-a1b2-ac80de305ab4",
   "metadata": {},
   "source": [
    "# DEA PreRMP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc8e250a-1340-44c3-a630-7d2b5837b111",
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- data_subset(meta, cnt, group_by=\"celltype\", group=\"PreRPM\", min_count=10, min_sample=3, min_group=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be0045b8-707d-4d0d-bb1f-53d3e3db56bc",
   "metadata": {},
   "source": [
    "## DDS and results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4040d523-824e-414b-aa8b-a0da62eeb08b",
   "metadata": {},
   "outputs": [],
   "source": [
    "res <- suppressMessages(lrt(data))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a1ab38f-370e-40ed-a1d4-ee8ec43efa26",
   "metadata": {},
   "source": [
    "## Result class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8c4a603-69e5-4b77-8117-fd38360ac28e",
   "metadata": {},
   "outputs": [],
   "source": [
    "resc <- result_class(res, padj_rc_thr=0.01, padj_pc_thr=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2ae073b-cf31-47df-a0e0-8c351da75f4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=30, repr.plot.height=25)\n",
    "\n",
    "hm_1 <- hm_response_class(data[[\"meta\"]][data[[\"meta\"]]$genotype==\"IFNAR_fl\", ], data[[\"cnt\"]][, data[[\"meta\"]]$genotype==\"IFNAR_fl\"], resc[[\"rc_wt\"]], response_class_n=10, title=\"Response genes\\n PreRPM (WT)\") %>% ggplotify::as.ggplot()\n",
    "hm_2 <- hm_response_class(data[[\"meta\"]][data[[\"meta\"]]$genotype==\"IFNAR_fl_LysM_cre\", ], data[[\"cnt\"]][, data[[\"meta\"]]$genotype==\"IFNAR_fl_LysM_cre\"], resc[[\"rc_ko\"]], response_class_n=10, title=\"Response genes\\n PreRPM (IFNARKO)\") %>% ggplotify::as.ggplot()\n",
    "hm_3 <- hm_perturbation(data[[\"meta\"]], data[[\"cnt\"]], resc[[\"pc_wt_vs_ko\"]], title=\"Perturbation genes\\n PreRPM\") %>% ggplotify::as.ggplot()\n",
    "\n",
    "hm_1 + hm_2 + hm_3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd6830d4-74cb-4b0b-b8e1-87be3304775c",
   "metadata": {},
   "source": [
    "## GSEA DPI between infection types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22337236-fe94-4ecb-9d24-6b3307dcea8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=4*10, repr.plot.height=12)\n",
    "\n",
    "gsea_1 <- lapply(res[[\"dpi_wt_vs_ko_wald\"]], function(result) {gsea(result)})\n",
    "gsea_plot_1 <- lapply(1:length(gsea_1), function(i) {gsea_plot(gsea_1[[i]], pathway_suffix=\"HALLMARK_\", color_names=c(\"IFNAR_fl_LysM_cre\", \"IFNAR_fl\"), title=c(\"D0\", \"D1\", \"D3\", \"D6\")[i], pval_thr=0.05)})\n",
    "\n",
    "patchwork::wrap_plots(gsea_plot_1, ncol=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2800299-b386-4e19-9c74-110a86f7320e",
   "metadata": {},
   "source": [
    "## Volcano DPI between infection types "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a89e2ce-1073-4a8b-93ba-643ff0b298c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=4*9, repr.plot.height=12)\n",
    "\n",
    "v_plot <- lapply(1:length(res[[\"dpi_wt_vs_ko_wald\"]]), function(i) {dea_vp(res[[\"dpi_wt_vs_ko_wald\"]][[i]], title=c(\"D0\", \"D1\", \"D3\", \"D6\")[i], p_adj_thr=0.05)})\n",
    "patchwork::wrap_plots(v_plot, ncol=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f217ddb2-7f6f-4761-9219-37198a0aed49",
   "metadata": {},
   "source": [
    "## Save results "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e5fc429-7a14-474b-af1f-cd277855995f",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(list(data=data, res=res, resc=resc), \"result/dea/bulkRNAseq/res_dea_prerpm.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bf9b3f1-a31f-4315-9af8-60cb22a28089",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Wald test genotype \n",
    "res[[\"dpi_wt_vs_ko_wald\"]] <- lapply(res[[\"dpi_wt_vs_ko_wald\"]], function(x) {x %>% tibble::rownames_to_column(\"gene\")})\n",
    "names(res[[\"dpi_wt_vs_ko_wald\"]]) <- c(\"D0\", \"D1\", \"D3\", \"D6\")\n",
    "openxlsx::write.xlsx(res[[\"dpi_wt_vs_ko_wald\"]], \"result/dea/bulkRNAseq/res_dea_genotype_prerpm.xlsx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6b32668-d43e-4569-b561-15f500c66e85",
   "metadata": {},
   "source": [
    "# DEA RPM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e82a8ff2-83fd-4c39-883a-6a30fa038c1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- data_subset(meta, cnt, group_by=\"celltype\", group=\"RPM\", min_count=10, min_sample=3, min_group=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a47777db-380f-4559-a209-18789655e63d",
   "metadata": {},
   "source": [
    "## DDS and results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a371021-0074-4e52-85c6-65b3654f3374",
   "metadata": {},
   "outputs": [],
   "source": [
    "res <- suppressMessages(lrt(data))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf6cc66d-8833-4c1d-8b6c-ebbd5a2506c3",
   "metadata": {},
   "source": [
    "## Result class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ea89573-1bb9-4b5d-880d-6ef071e0685a",
   "metadata": {},
   "outputs": [],
   "source": [
    "resc <- result_class(res, padj_rc_thr=0.01, padj_pc_thr=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9d8b1cd-1691-4a8c-9231-4661d9ee29d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=30, repr.plot.height=25)\n",
    "\n",
    "hm_1 <- hm_response_class(data[[\"meta\"]][data[[\"meta\"]]$genotype==\"IFNAR_fl\", ], data[[\"cnt\"]][, data[[\"meta\"]]$genotype==\"IFNAR_fl\"], resc[[\"rc_wt\"]], response_class_n=10, title=\"Response genes\\n RPM (WT)\") %>% ggplotify::as.ggplot()\n",
    "hm_2 <- hm_response_class(data[[\"meta\"]][data[[\"meta\"]]$genotype==\"IFNAR_fl_LysM_cre\", ], data[[\"cnt\"]][, data[[\"meta\"]]$genotype==\"IFNAR_fl_LysM_cre\"], resc[[\"rc_ko\"]], response_class_n=10, title=\"Response genes\\n RPM (IFNARKO)\") %>% ggplotify::as.ggplot()\n",
    "hm_3 <- hm_perturbation(data[[\"meta\"]], data[[\"cnt\"]], resc[[\"pc_wt_vs_ko\"]], title=\"Perturbation genes\\n RPM\") %>% ggplotify::as.ggplot()\n",
    "\n",
    "hm_1 + hm_2 + hm_3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5675d7b5-e011-44ea-89e6-ebeb008641ad",
   "metadata": {},
   "source": [
    "## GSEA DPI between infection types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a906841-5ef6-4dca-a718-90604ff33d0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=4*10, repr.plot.height=12)\n",
    "\n",
    "gsea_1 <- lapply(res[[\"dpi_wt_vs_ko_wald\"]], function(result) {gsea(result)})\n",
    "gsea_plot_1 <- lapply(1:length(gsea_1), function(i) {gsea_plot(gsea_1[[i]], pathway_suffix=\"HALLMARK_\", color_names=c(\"IFNAR_fl_LysM_cre\", \"IFNAR_fl\"), title=c(\"D0\", \"D1\", \"D3\", \"D6\")[i], pval_thr=0.05)})\n",
    "\n",
    "patchwork::wrap_plots(gsea_plot_1, ncol=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c5933c6-b06f-41cd-9ca5-3ef6d8490d25",
   "metadata": {},
   "source": [
    "## Volcano DPI between infection types "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8931ff17-4d45-4c0d-86b9-09e6e6421949",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=4*9, repr.plot.height=12)\n",
    "\n",
    "v_plot <- lapply(1:length(res[[\"dpi_wt_vs_ko_wald\"]]), function(i) {dea_vp(res[[\"dpi_wt_vs_ko_wald\"]][[i]], title=c(\"D0\", \"D1\", \"D3\", \"D6\")[i], p_adj_thr=0.05)})\n",
    "patchwork::wrap_plots(v_plot, ncol=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3e5a2be-1139-4c54-85cd-e9da7e11a469",
   "metadata": {},
   "source": [
    "## Save results "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4140488-8fa9-453b-b9a9-1db7c3dd15b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(list(data=data, res=res, resc=resc), \"result/dea/bulkRNAseq/res_dea_rpm.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3032735b-a137-407a-9833-975de1b51015",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Wald test genotype \n",
    "res[[\"dpi_wt_vs_ko_wald\"]] <- lapply(res[[\"dpi_wt_vs_ko_wald\"]], function(x) {x %>% tibble::rownames_to_column(\"gene\")})\n",
    "names(res[[\"dpi_wt_vs_ko_wald\"]]) <- c(\"D0\", \"D1\", \"D3\", \"D6\")\n",
    "openxlsx::write.xlsx(res[[\"dpi_wt_vs_ko_wald\"]], \"result/dea/bulkRNAseq/res_dea_genotype_rpm.xlsx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a03ac96-74fb-4ee5-a2be-7065ad30085b",
   "metadata": {},
   "source": [
    "# DEA EB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ad442b6-99e2-4470-a73e-d3d3c570bcfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset data \n",
    "data <- data_subset(meta, cnt, group_by=\"celltype\", group=\"EB\", min_count=10, min_sample=3, min_group=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c25e02a-0e3e-4406-920a-33970427ad64",
   "metadata": {},
   "source": [
    "## DDS and results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "837f1a8c-04b6-4548-ae1b-20e09fad5ab9",
   "metadata": {},
   "outputs": [],
   "source": [
    "res <- suppressMessages(lrt(data))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15363926-d729-4546-8939-ffca965b5357",
   "metadata": {},
   "source": [
    "## Result class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff5011cf-e561-4842-afda-766536b69a97",
   "metadata": {},
   "outputs": [],
   "source": [
    "resc <- result_class(res, padj_rc_thr=0.01, padj_pc_thr=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "694a02bf-3b4a-41a5-b4f1-6065eaba4ba1",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=30, repr.plot.height=25)\n",
    "\n",
    "hm_1 <- hm_response_class(data[[\"meta\"]][data[[\"meta\"]]$genotype==\"IFNAR_fl\", ], data[[\"cnt\"]][, data[[\"meta\"]]$genotype==\"IFNAR_fl\"], resc[[\"rc_wt\"]], response_class_n=10, title=\"Response genes\\n cMo (WT)\") %>% ggplotify::as.ggplot()\n",
    "hm_2 <- hm_response_class(data[[\"meta\"]][data[[\"meta\"]]$genotype==\"IFNAR_fl_LysM_cre\", ], data[[\"cnt\"]][, data[[\"meta\"]]$genotype==\"IFNAR_fl_LysM_cre\"], resc[[\"rc_ko\"]], response_class_n=10, title=\"Response genes\\n cMo (IFNARKO)\") %>% ggplotify::as.ggplot()\n",
    "hm_3 <- hm_perturbation(data[[\"meta\"]], data[[\"cnt\"]], resc[[\"pc_wt_vs_ko\"]], title=\"Perturbation genes\\n cMo\") %>% ggplotify::as.ggplot()\n",
    "\n",
    "hm_1 + hm_2 + hm_3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8f3394f-010e-4235-a350-f5937fea670a",
   "metadata": {},
   "source": [
    "## GSEA DPI between infection types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a34612df-e28a-4b88-a147-17fca1611b21",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=4*10, repr.plot.height=12)\n",
    "\n",
    "gsea_1 <- lapply(res[[\"dpi_wt_vs_ko_wald\"]], function(result) {gsea(result)})\n",
    "gsea_plot_1 <- lapply(1:length(gsea_1), function(i) {gsea_plot(gsea_1[[i]], pathway_suffix=\"HALLMARK_\", color_names=c(\"IFNAR_fl_LysM_cre\", \"IFNAR_fl\"), title=c(\"D0\", \"D1\", \"D3\", \"D6\")[i], pval_thr=0.05)})\n",
    "\n",
    "patchwork::wrap_plots(gsea_plot_1, ncol=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1eb1c603-8fc8-4790-bcd1-38d833cd2e00",
   "metadata": {},
   "source": [
    "## Volcano DPI between infection types "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d248a36-6f69-44a0-b78f-c0c1b355d295",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=4*9, repr.plot.height=12)\n",
    "\n",
    "v_plot <- lapply(1:length(res[[\"dpi_wt_vs_ko_wald\"]]), function(i) {dea_vp(res[[\"dpi_wt_vs_ko_wald\"]][[i]], title=c(\"D0\", \"D1\", \"D3\", \"D6\")[i], p_adj_thr=0.05)})\n",
    "patchwork::wrap_plots(v_plot, ncol=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "978aa6fc-d0fb-4585-a0ba-2323aeb3e45f",
   "metadata": {},
   "source": [
    "## Save results "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c2e5f7f-e9cb-426c-ae57-21f54caa1b97",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(list(data=data, res=res, resc=resc), \"result/dea/bulkRNAseq/res_dea_ery.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71d10fab-2752-49a9-99dd-8dca4da8d604",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Wald test genotype \n",
    "res[[\"dpi_wt_vs_ko_wald\"]] <- lapply(res[[\"dpi_wt_vs_ko_wald\"]], function(x) {x %>% tibble::rownames_to_column(\"gene\")})\n",
    "names(res[[\"dpi_wt_vs_ko_wald\"]]) <- c(\"D0\", \"D1\", \"D3\", \"D6\")\n",
    "openxlsx::write.xlsx(res[[\"dpi_wt_vs_ko_wald\"]], \"result/dea/bulkRNAseq/res_dea_genotype_ery.xlsx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e3a6a2b-eabc-4236-98bc-811bac2d5698",
   "metadata": {},
   "source": [
    "# Session Info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfa8ea83-1199-4f04-94e6-646773902ff7",
   "metadata": {},
   "outputs": [],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "r.4.4.1-FD20200109SPLENO",
   "language": "R",
   "name": "r.4.4.1-fd20200109spleno"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
