{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3e57bbb1-f1d1-4c7a-8248-430750ec09f6",
   "metadata": {},
   "source": [
    "# Integration\n",
    "\n",
    "**[Sanorama Hie et al., 2019](https://www.nature.com/articles/s41587-019-0113-3)**  \n",
    "[GitHub](https://github.com/brianhie/scanorama)  \n",
    "[Tutorial external API](https://scanpy-tutorials.readthedocs.io/en/latest/spatial/integration-scanorama.html)  \n",
    "[External external API tutorial](https://scanpy.readthedocs.io/en/stable/generated/scanpy.external.pp.scanorama_integrate.html)\n",
    "\n",
    "A [fix](https://github.com/theislab/single-cell-tutorial/issues/60) to run [scran pooling normalization](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0947-7) computeSumFactors in current python environment. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d2b0fedc-315f-4a1d-9ccc-732ab77d0481",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import scanorama\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "from IPython.display import display, HTML\n",
    "from matplotlib.pyplot import rc_context\n",
    "\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "69fe0382-4510-449c-b271-0e16bb135384",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Working directory \n",
    "os.chdir('/research/peer/fdeckert/FD20200109SPLENO')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5cd3e203-b6ce-4bbd-a703-aa6bfe43a6b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# rpy2 \n",
    "os.environ['R_HOME'] = '/home/fdeckert/bin/miniconda3/envs/p.3.8.12-FD20200109SPLENO/lib/R'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7b50e162-f6b9-4d41-baff-b6b1967ebe45",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting \n",
    "import rpy2.robjects as robjects\n",
    "color_load = robjects.r.source('plotting_global.R')\n",
    "color = dict()\n",
    "for i in range(len(color_load[0])):\n",
    "    color[color_load[0].names[i]] = {key : color_load[0][i].rx2(key)[0] for key in color_load[0][i].names}\n",
    "\n",
    "sc.set_figure_params(figsize=(5, 5))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f835b5b-8f4d-4778-b543-7284a014d973",
   "metadata": {},
   "source": [
    "# Parameter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "99d7ea21-b20b-4844-8b7b-abd04413b764",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Scanorama \n",
    "dimred=100\n",
    "knn=20\n",
    "sigma=15\n",
    "alpha=0.1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eef2cda5-9a3d-4c76-8f70-c8bc72cff9c1",
   "metadata": {},
   "source": [
    "# Scanorama"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b676c2ba-7a39-4132-a4b9-22e4279d4494",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad('data/object/so_sct.h5ad')\n",
    "hvg = list(adata.uns['hvg_int_8000'])\n",
    "adata = adata.raw.to_adata()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e489dec4-65eb-4014-921d-9495b07c556a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_color(categories): \n",
    "    \n",
    "    categories = [x for x in categories if x in list(adata.obs.columns)]\n",
    "\n",
    "    for category in categories: \n",
    "        \n",
    "        adata.obs[category] = pd.Series(adata.obs[category], dtype='category')\n",
    "        \n",
    "        keys = list(color[category].keys())\n",
    "        keys = [x for x in keys if x in list(adata.obs[category])]\n",
    "\n",
    "        adata.obs[category] = adata.obs[category].cat.reorder_categories(keys)\n",
    "        adata.uns[category+'_colors'] = np.array([color[category].get(key) for key in keys], dtype=object)\n",
    "        \n",
    "# Set colors\n",
    "set_color(list(color.keys()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "43773bcc-63b4-42c4-b1e9-f0862d9bbe11",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(adata)\n",
    "sc.pp.log1p(adata)\n",
    "adata = adata[:,hvg]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9bcc04cb-c4a6-4188-840d-e73343e2ae30",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1146226/3130535487.py:1: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata.obs['integrate'] = adata.obs['treatment'].astype(str)+adata.obs['sample_rep'].astype(str)\n"
     ]
    }
   ],
   "source": [
    "adata.obs['integrate'] = adata.obs['treatment'].astype(str)+adata.obs['sample_rep'].astype(str)\n",
    "adata_sub = dict()\n",
    "for sample_group in adata.obs['integrate'].unique():\n",
    "    adata_tmp = adata[adata.obs['integrate']==sample_group].copy()\n",
    "    sc.pp.scale(adata_tmp)\n",
    "    adata_sub[sample_group] = adata_tmp\n",
    "adata_sub = list(adata_sub.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b36d6cf1-ba46-4f1a-8d1a-0dbd1f41a941",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 8000 genes among all datasets\n",
      "Processing datasets (2, 3)\n",
      "Processing datasets (1, 3)\n",
      "Processing datasets (0, 1)\n",
      "Processing datasets (0, 2)\n",
      "Processing datasets (0, 3)\n",
      "Processing datasets (1, 2)\n"
     ]
    }
   ],
   "source": [
    "# Run Scanorama\n",
    "scanorama.integrate_scanpy(adata_sub, dimred=dimred, knn=knn, sigma=sigma, alpha=alpha, verbose=True)\n",
    "\n",
    "# Concatenate scanorama output \n",
    "X_scanorama = [ad.obsm['X_scanorama'] for ad in adata_sub]\n",
    "X_scanorama = np.concatenate(X_scanorama)\n",
    "\n",
    "obs_names = [ad.obs_names for ad in adata_sub]\n",
    "obs_names = np.concatenate(obs_names)\n",
    "all(obs_names==adata.obs_names)\n",
    "\n",
    "# Add X_scanorama integration to adata \n",
    "adata.obsm[\"X_scanorama\"] = X_scanorama"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a48dea98-e7c1-4b2c-a2a2-5b543501298c",
   "metadata": {},
   "source": [
    "# Control cluster composition on high resolution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed1ec94b-a923-4efb-aa37-ace90726cf0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dimensional reduction and clustering \n",
    "sc.pp.neighbors(adata, n_neighbors=10, metric='euclidean', n_pcs=dimred, use_rep='X_scanorama')\n",
    "sc.tl.leiden(adata, resolution=2)\n",
    "sc.tl.louvain(adata, resolution=2)\n",
    "sc.tl.umap(adata, min_dist=0.3, spread=1.0)\n",
    "\n",
    "# Plot \n",
    "sc.pl.umap(adata, color=['louvain', 'leiden', 'tissue', 'treatment', 'label_fine_haemosphere', 'sample_rep', 'cc_phase_class', 'pMt_RNA', 'pHb_RNA', 'pRb_RNA', 'nCount_RNA', 'nFeature_RNA'], wspace=0.5, ncols=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7f4ff4a-62f9-4c1b-bd57-9d4070890de9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = adata[adata.obs['treatment']=='NaCl'].copy().obs\n",
    "df = df[['leiden', 'treatment', 'sample_rep']].groupby(['leiden']).value_counts('sample_rep').to_frame().reset_index().pivot(index='sample_rep', columns='leiden', values=0)\n",
    "display(HTML(df.to_html()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43fe7e2a-2079-4322-8684-ffb27e5a19a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = adata[adata.obs['treatment']=='CpG'].copy().obs\n",
    "df = df[['leiden', 'treatment', 'sample_rep']].groupby(['leiden']).value_counts('sample_rep').to_frame().reset_index().pivot(index='sample_rep', columns='leiden', values=0)\n",
    "display(HTML(df.to_html()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8add88ef-20cb-4626-8c66-d31c129728eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "with rc_context({'figure.figsize': (7.5, 3)}):\n",
    "    sc.pl.violin(adata, ['nCount_RNA', 'nFeature_RNA', 'pMt_RNA', 'pRb_RNA'], groupby='leiden', stripplot=False, inner='box', multi_panel=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a3c7a0c-9919-4024-969b-235d560ed238",
   "metadata": {},
   "source": [
    "# Dimensionality reduction and clustering (n_neighbors=30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59cd26f3-204d-4806-ad91-060d1899963f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dimensional reduction and clustering \n",
    "sc.pp.neighbors(adata, n_neighbors=30, metric='euclidean', n_pcs=dimred, use_rep='X_scanorama')\n",
    "sc.tl.leiden(adata, resolution=1)\n",
    "sc.tl.louvain(adata, resolution=1)\n",
    "sc.tl.umap(adata, min_dist=0.3, spread=1.0)\n",
    "\n",
    "# Plot \n",
    "sc.pl.umap(adata, color=['louvain', 'leiden', 'tissue', 'treatment', 'label_fine_haemosphere', 'sample_rep', 'cc_phase_class', 'pHb_RNA', 'pRb_RNA'], wspace=0.5, ncols=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55ce8235-5b21-487c-b363-9d5e7bfd655a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "p.3.8.12-FD20200109SPLENO",
   "language": "python",
   "name": "p.3.8.12-fd20200109spleno"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
