{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "50e4fbf7-815e-4b90-9012-a0afabdd59a5",
   "metadata": {},
   "source": [
    "# Figure 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cfaec12-a6e4-450e-a15e-e4a93255590b",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "800de35d-86e3-4b71-8dd3-911e631758f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "library_load <- suppressMessages(\n",
    "    \n",
    "    suppressWarnings(\n",
    "        \n",
    "        list(\n",
    "        \n",
    "            # Seurat \n",
    "            library(Seurat), \n",
    "\n",
    "            # TradeSeq\n",
    "            library(tradeSeq),\n",
    "\n",
    "            # CCI \n",
    "            library(CellChat), \n",
    "\n",
    "            # Data \n",
    "            library(tidyverse), \n",
    "            library(data.table), \n",
    "\n",
    "            # miloR\n",
    "            library(miloR), \n",
    "            library(ggbeeswarm), \n",
    "\n",
    "            # Plotting \n",
    "            library(ggplot2), \n",
    "            library(patchwork), \n",
    "            library(cowplot), \n",
    "            library(ComplexHeatmap), \n",
    "\n",
    "            # Pyhton compatibility\n",
    "            library(reticulate)\n",
    "\n",
    "        )\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15b0ec44-5e03-4e48-82cd-1042b31d1a5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Configure reticulate \n",
    "# use_condaenv(condaenv=\"p.3.10.16-FD20200109SPLENO\", conda=\"/nobackup/peer/fdeckert/miniconda3/bin/conda\", required=NULL)\n",
    "# py_config()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45028df8-0d7a-4a09-865f-5d769c33cf3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_seed <- 42\n",
    "set.seed(random_seed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ae344ae-fb5c-42da-af0e-e5559e7b3e09",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set working directory to project root\n",
    "setwd(\"/research/peer/fdeckert/FD20200109SPLENO\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b52cea3-0401-43b9-a99a-a97fc2a66c4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Source\n",
    "source(\"bin/so_pl.R\")\n",
    "source(\"bin/dea_pp.R\")\n",
    "source(\"bin/dea_pl.R\")\n",
    "source(\"bin/gsea_pp.R\")\n",
    "source(\"bin/gsea_pl.R\")\n",
    "source(\"bin/cci_pp.R\")\n",
    "source(\"bin/cci_pl.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ab8c554-ca21-4e65-8932-2a87d05f1f30",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting Theme\n",
    "source(\"plotting_global.R\")\n",
    "ggplot2::theme_set(theme_global_set(size_select=4)) # From project global source()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7d45dcf-0f7c-422e-815d-c982dd7d9e37",
   "metadata": {},
   "source": [
    "# Import scRNAseq data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04ba5c21-247b-4af4-ac1b-37c0bddd7b69",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- readRDS(\"data/scRNAseq/object/pp_1.rds\")\n",
    "so <- NormalizeData(so)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be7d13cf-4977-45e9-a028-89ea7e5d7350",
   "metadata": {},
   "source": [
    "# Differential expression analysis (DEA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b711403-c917-4872-9c8d-331b39b4cd30",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Response gene DEA \n",
    "contrasts_vec <- c(\"IFNAR_fl_CpG_D1\", \"IFNAR_fl_Baseline_D0\")\n",
    "dea_res_1 <- readRDS(paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))\n",
    "\n",
    "contrasts_vec <- c(\"IFNAR_fl_LysM_cre_CpG_D1\", \"IFNAR_fl_LysM_cre_Baseline_D0\")\n",
    "dea_res_2 <- readRDS(paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))\n",
    "\n",
    "contrasts_vec <- c(\"IFNAR_fl_CpG_D3\", \"IFNAR_fl_Baseline_D0\")\n",
    "dea_res_3 <- readRDS(paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))\n",
    "\n",
    "contrasts_vec <- c(\"IFNAR_fl_LysM_cre_CpG_D3\", \"IFNAR_fl_LysM_cre_Baseline_D0\")\n",
    "dea_res_4 <- readRDS(paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))\n",
    "\n",
    "# Genotype DEA\n",
    "contrasts_vec <- c(\"IFNAR_fl_LysM_cre_Baseline_D0\", \"IFNAR_fl_Baseline_D0\")\n",
    "dea_res_5 <- readRDS(paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))\n",
    "\n",
    "contrasts_vec <- c(\"IFNAR_fl_LysM_cre_CpG_D1\", \"IFNAR_fl_CpG_D1\")\n",
    "dea_res_6 <- readRDS(paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))\n",
    "\n",
    "contrasts_vec <- c(\"IFNAR_fl_LysM_cre_CpG_D3\", \"IFNAR_fl_CpG_D3\")\n",
    "dea_res_7 <- readRDS(paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8af20ea3-e859-4451-8822-8ed4bc6a06c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "adj_pval_thr <- 0.05\n",
    "log2_thr <- 0.50"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f9c804f-bc7b-49c4-87a1-c6e5fa94e653",
   "metadata": {},
   "source": [
    "# DEG count overview"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d0f4dd5-fd88-44f2-9388-42752bfe03aa",
   "metadata": {},
   "source": [
    "## Response count D1 and D3 vs Ctl "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e334cd1-b03d-49da-84f9-cd76cdb3033c",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- lapply(dea_res_1, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2_thr) %>% dplyr::pull(gene) %>% length}) %>% data.frame() %>% t() \n",
    "mat_2 <- lapply(dea_res_2, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2_thr) %>% dplyr::pull(gene) %>% length}) %>% data.frame() %>% t() \n",
    "mat_3 <- lapply(dea_res_3, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2_thr) %>% dplyr::pull(gene) %>% length}) %>% data.frame() %>% t() \n",
    "mat_4 <- lapply(dea_res_4, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2_thr) %>% dplyr::pull(gene) %>% length}) %>% data.frame() %>% t() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f059b523-aebe-4707-a779-0340e2a6fed1",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat <- list(mat_1, mat_2, mat_3, mat_4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7788e6c-8792-455f-9cb9-5530da268f45",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat <- lapply(mat, function(x) {\n",
    "    \n",
    "    x <- as.data.frame(x)\n",
    "    set_diff <- setdiff(make.names(levels(so$celltype_low)), rownames(x))\n",
    "    x[set_diff, ] <- 0\n",
    "    x[make.names(levels(so$celltype_low)), , drop=FALSE]\n",
    "\n",
    "}\n",
    "             )\n",
    "\n",
    "mat <- do.call(cbind, mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2a8afae-68d4-461f-a2ff-ca5464cf8341",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(mat) <- c(\"+/+\", \"cre/+\", \"+/+\", \"cre/+\")\n",
    "col_split <- c(\"D1\", \"D1\", \"D3\", \"D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea688900-a1d9-4611-a03c-f6cdccb65b5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/hm_deg_count_d1_d3_vs_ctl_cg_and_ko.pdf\", width=1.5, height=3)\n",
    "source(\"bin/dea_pl.R\")\n",
    "draw(dea_res_count_hm(as.matrix(mat), col_split=col_split, fontsize_select=2))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef419035-3b29-4615-8552-e10d0855634b",
   "metadata": {},
   "source": [
    "## Genotype comparison "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af472cc2-f9a8-4dae-8260-b655d0ebe33b",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- lapply(dea_res_5, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2_thr) %>% dplyr::pull(gene) %>% length}) %>% data.frame() %>% t() \n",
    "mat_2 <- lapply(dea_res_6, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2_thr) %>% dplyr::pull(gene) %>% length}) %>% data.frame() %>% t() \n",
    "mat_3 <- lapply(dea_res_7, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2_thr) %>% dplyr::pull(gene) %>% length}) %>% data.frame() %>% t() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bb300fd-dcd3-4ce8-80b8-50413c34b18c",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat <- list(mat_1, mat_2, mat_3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51d9b947-9c97-4038-a5bf-d50085fb9ceb",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat <- lapply(mat, function(x) {\n",
    "    \n",
    "    x <- as.data.frame(x)\n",
    "    set_diff <- setdiff(make.names(levels(so$celltype_low)), rownames(x))\n",
    "    x[set_diff, ] <- 0\n",
    "    x[make.names(levels(so$celltype_low)), , drop=FALSE]\n",
    "\n",
    "}\n",
    "             )\n",
    "\n",
    "mat <- do.call(cbind, mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b5c9ebc-f2e1-4b3e-8b73-58daf037702e",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(mat) <- c(\"+/+ vs cre/+\", \"+/+ vs cre/+\", \"+/+ vs cre/+\")\n",
    "col_split <- c(\"Ctl\", \"D1\", \"D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a1db0f8-b61a-4b5f-aa6c-79be5c683476",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/hm_deg_count_cg_vs_ko.pdf\", width=1.5, height=3)\n",
    "\n",
    "draw(dea_res_count_hm(as.matrix(mat), col_split=col_split, fontsize_select=2))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "caa082f8-151f-4c3d-b882-52dfe9e3b8a2",
   "metadata": {},
   "source": [
    "# Gene set enrichment analysis (GSEA)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8569bdd3-f557-4b93-b4ae-cc3056a655de",
   "metadata": {},
   "source": [
    "## genotype comparison - M7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab320482-f659-41b4-bf29-a7a83ab484e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "compute_gsea <- FALSE\n",
    "celltype_select <- c(\"MEP\", \"Proerythroblast\", \"Erythroblast\", \"cMo\", \"intMo\", \"ncMo\", \"RPM\", \"cDC1\", \"cDC2\", \"moDC\", \"cDC.mig.\", \"Basophil\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1359afa3-7e3c-478a-a60e-0be197bbb19a",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(compute_gsea) {\n",
    "\n",
    "    # Collapse gene set\n",
    "    gene_set <- msigdbr::msigdbr(species=\"mouse\", db_species=\"MM\", category=\"M7\", subcategory=NULL)\n",
    "    gene_set <- gene_set[grep(\"_UP\", gene_set$gs_name), ]\n",
    "    gene_set$gs_name <- sapply(strsplit(gene_set$gs_name, \"_\"), function(x) {paste(x[max(1, length(x)-3):(length(x)-2)], collapse=\"_\")})\n",
    "    gene_set$gs_name <- gsub(\"CELL_|CD4_|CD8_|MONOCYTE_|MACROPHAGE_|TREG_|LANGERHANS_|NEUTROPHIL_|CDC1_|CDC2_|MIGDC_|ETAC_|GD_|ILC_|PDC_\", \"\", gene_set$gs_name)\n",
    "    gene_set <- gene_set %>% dplyr::select(gene_symbol, gs_name) %>% dplyr::distinct()\n",
    "\n",
    "    gsea_res <- lapply(celltype_select, function(i) {\n",
    "    \n",
    "        gsea_res_1 <- gsea(dea_res_5[[i]], gene_set=gene_set) %>% dplyr::mutate(col_split=\"Ctl\", col_label=i)\n",
    "        gsea_res_2 <- gsea(dea_res_6[[i]], gene_set=gene_set) %>% dplyr::mutate(col_split=\"D1\", col_label=i)\n",
    "        gsea_res_3 <- gsea(dea_res_7[[i]], gene_set=gene_set) %>% dplyr::mutate(col_split=\"D3\", col_label=i)\n",
    "    \n",
    "        gsea_res <- rbind(gsea_res_1, gsea_res_2, gsea_res_3)\n",
    "    \n",
    "        return(gsea_res)\n",
    "        \n",
    "    }\n",
    "          )\n",
    "\n",
    "    saveRDS(gsea_res, \"result/gsea/scRNAseq/gsea_ctl_d1_d3_wt_vs_ko_m7.rds\")\n",
    "    \n",
    "} else {\n",
    "\n",
    "    gsea_res <- readRDS(\"result/gsea/scRNAseq/gsea_ctl_d1_d3_wt_vs_ko_m7.rds\")\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42d9360e-dc81-4ab3-8d8c-1a919d1664ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res <- do.call(rbind, gsea_res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "685414aa-068b-4b11-8f97-eaf9867f8b7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs_name_select <- gsea_res %>% dplyr::filter(padj<=0.05 & abs(NES)>=1) %>% dplyr::pull(pathway)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1dd31e51-eee3-462b-9486-4bdb61a8954e",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- gsea_res %>% dplyr::filter(pathway %in% gs_name_select) %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=padj,\n",
    "        values_fill=1\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6006c1c-5934-4bba-a3ab-254f2d5edce0",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_2 <- gsea_res %>% dplyr::filter(pathway %in% gs_name_select) %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=NES,\n",
    "        values_fill=0\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08115c33-eb78-4629-8e66-9abe5f18eb8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_split <- read.csv(\"data/reference/immune_dictionary/immune_dictionary.csv\") %>% tibble::column_to_rownames(\"gs_name\")\n",
    "col_label <- mat_1$col_label\n",
    "col_split <- mat_1$col_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3df90dc3-3d27-4b5c-a683-de6ad1aefe01",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- mat_1 %>% dplyr::select(-col_label, -col_split) %>% t()\n",
    "mat_2 <- mat_2 %>% dplyr::select(-col_label, -col_split) %>% t()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5971db3-87ca-4018-920f-397ea543608c",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_split <- row_split[rownames(mat_1), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3f20be2-86af-46b0-86da-7db847e27bc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "rownames(mat_1) <- row_split$norm\n",
    "rownames(mat_2) <- row_split$norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6f4092b-a24f-475e-819a-73c1583b7884",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "row_split <- row_split %>% dplyr::select(norm, family) %>% dplyr::rename(row_split_label=family) %>% dplyr::distinct()\n",
    "rownames(row_split) <- row_split$norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11f01b9c-6b82-45cb-a6ff-5ad736dd936e",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_label[col_label==\"cDC.mig.\"] <- \"cDC mig.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e581fe24-9fba-48ee-8747-206d8a1dda6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/hm_gsea_ctl_d1_d3_wt_vs_ko_m7.pdf\", width=8, height=12)\n",
    "\n",
    "gsea_hm(mat_1, mat_2, col_label=col_label, col_split=col_split, row_split=row_split, fontsize_select=2)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b21ed7b7-96a7-463f-94d0-5a8cede04a09",
   "metadata": {},
   "source": [
    "## genotype comparison - HM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c9289eb-8412-4d07-b34e-8ea15840dce8",
   "metadata": {},
   "outputs": [],
   "source": [
    "compute_gsea <- FALSE\n",
    "celltype_select <- c(\"MEP\", \"Proerythroblast\", \"Erythroblast\", \"cMo\", \"intMo\", \"ncMo\", \"RPM\", \"cDC1\", \"cDC2\", \"moDC\", \"cDC.mig.\", \"Basophil\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "885824cd-7f1d-4363-975d-14d058961671",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(compute_gsea) {\n",
    "\n",
    "    gsea_res <- lapply(celltype_select, function(i) {\n",
    "    \n",
    "        gsea_res_1 <- gsea(dea_res_5[[i]], category=\"MH\", subcategory=NULL, gene_set=NULL) %>% dplyr::mutate(col_split=\"Ctl\", col_label=i)\n",
    "        gsea_res_2 <- gsea(dea_res_6[[i]], category=\"MH\", subcategory=NULL, gene_set=NULL) %>% dplyr::mutate(col_split=\"D1\", col_label=i)\n",
    "        gsea_res_3 <- gsea(dea_res_7[[i]], category=\"MH\", subcategory=NULL, gene_set=NULL) %>% dplyr::mutate(col_split=\"D3\", col_label=i)\n",
    "    \n",
    "        gsea_res <- rbind(gsea_res_1, gsea_res_2, gsea_res_3)\n",
    "    \n",
    "        return(gsea_res)\n",
    "        \n",
    "    }\n",
    "          )\n",
    "\n",
    "    saveRDS(gsea_res, \"result/gsea/scRNAseq/gsea_ctl_d1_d3_wt_vs_ko_mh.rds\")\n",
    "    \n",
    "} else {\n",
    "\n",
    "    gsea_res <- readRDS(\"result/gsea/scRNAseq/gsea_ctl_d1_d3_wt_vs_ko_mh.rds\")\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba957b2c-78b2-4ba5-8a1a-e5ea52c90a21",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res <- do.call(rbind, gsea_res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63b1abc0-9358-418e-812d-de21ea45cb7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs_name_select <- gsea_res %>% dplyr::filter(padj<=0.05 & abs(NES)>=1) %>% dplyr::pull(pathway)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6139a685-8339-42e2-ae7d-7cbc061f55a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- gsea_res %>% dplyr::filter(pathway %in% gs_name_select) %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=padj,\n",
    "        values_fill=1\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1a51c94-5778-4d10-97ae-767a69147b47",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_2 <- gsea_res %>% dplyr::filter(pathway %in% gs_name_select) %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=NES,\n",
    "        values_fill=0\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55514ba2-b2b2-449c-8f67-509d0da3df7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_label <- mat_1$col_label\n",
    "col_split <- mat_1$col_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3f5482c-2d04-419f-afac-fa1776a3f428",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- mat_1 %>% dplyr::select(-col_label, -col_split) %>% t()\n",
    "mat_2 <- mat_2 %>% dplyr::select(-col_label, -col_split) %>% t()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57c3d582-4c80-42a8-a057-df8780f9a190",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_label[col_label==\"cDC.mig.\"] <- \"cDC mig.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abdc1726-c16c-4468-96d5-003e05e0aeb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/hm_gsea_ctl_d1_d3_wt_vs_ko_mh.pdf\", width=8, height=10)\n",
    "\n",
    "gsea_hm(mat_1, mat_2, col_label=col_label, col_split=col_split, row_split=NULL, pathway_suffix=\"HALLMARK\", fontsize_select=2)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25be8c31-f619-4e12-b5b2-b0ef62ed256d",
   "metadata": {},
   "source": [
    "## genotype comparison - GO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dd64847-8400-48f0-af90-99d8f0e02ef7",
   "metadata": {},
   "outputs": [],
   "source": [
    "compute_gsea <- FALSE\n",
    "celltype_select <- c(\"MEP\", \"Proerythroblast\", \"Erythroblast\", \"cMo\", \"intMo\", \"ncMo\", \"RPM\", \"cDC1\", \"cDC2\", \"moDC\", \"cDC.mig.\", \"Basophil\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e7734f6-65c0-4467-8e25-b7b7d4e8a7c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(compute_gsea) {\n",
    "\n",
    "    gsea_res <- lapply(celltype_select, function(i) {\n",
    "    \n",
    "        gsea_res_1 <- gsea_cp(dea_res_5[[i]]) %>% dplyr::mutate(col_split=\"Ctl\", col_label=i)\n",
    "        gsea_res_2 <- gsea_cp(dea_res_6[[i]]) %>% dplyr::mutate(col_split=\"D1\", col_label=i)\n",
    "        gsea_res_3 <- gsea_cp(dea_res_7[[i]]) %>% dplyr::mutate(col_split=\"D3\", col_label=i)\n",
    "    \n",
    "        gsea_res <- rbind(gsea_res_1, gsea_res_2, gsea_res_3)\n",
    "    \n",
    "        return(gsea_res)\n",
    "        \n",
    "    }\n",
    "          )\n",
    "\n",
    "    saveRDS(gsea_res, \"result/gsea/scRNAseq/gsea_ctl_d1_d3_wt_vs_ko_gobp.rds\")\n",
    "    \n",
    "} else {\n",
    "\n",
    "    gsea_res <- readRDS(\"result/gsea/scRNAseq/gsea_ctl_d1_d3_wt_vs_ko_gobp.rds\")\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7b5f832-67cc-443c-8b9d-cabf9347966c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res <- do.call(rbind, gsea_res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1777f5d7-18bb-42cf-a416-b6bf86a6343a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs_name_select <- gsea_res %>% dplyr::filter(p.adjust<=0.05 & abs(NES)>=1) %>% dplyr::pull(ID)\n",
    "gsea_res <- gsea_res %>% dplyr::filter(ID %in% gs_name_select)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39b94e20-83be-416e-92e6-cf3543f724f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res <- gsea_res[!gsea_res$Description %in% grep(\"T cell|B cell|CD4|CD8|lymphocyte|endothelial|epithelial\", gsea_res$Description, value=TRUE), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af296a68-40e3-41ae-8be2-1dbf6f56d93e",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(compute_gsea) {\n",
    "\n",
    "    gs_reduce <- gsea_cp_reduce(gsea_res)\n",
    "\n",
    "    saveRDS(gs_reduce, \"result/gsea/scRNAseq/gsea_reduce_ctl_d1_d3_wt_vs_ko_gobp.rds\")\n",
    "    \n",
    "} else {\n",
    "\n",
    "    gs_reduce <- readRDS(\"result/gsea/scRNAseq/gsea_reduce_ctl_d1_d3_wt_vs_ko_gobp.rds\")\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adddea3f-2873-4b63-a260-9d0c923dce87",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res <- gsea_res %>% dplyr::rename(pathway=Description, padj=p.adjust) %>% dplyr::filter(ID %in% gs_reduce$go)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9341e76-d801-4f73-83e2-1e48074ad4b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs_name_select <- gsea_res %>% dplyr::filter(padj<=0.05 & abs(NES)>=2) %>% dplyr::pull(ID)\n",
    "gsea_res <- gsea_res %>% dplyr::filter(ID %in% gs_name_select)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d12e8ed-a07f-4588-acc7-9cca036764e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- gsea_res %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=padj,\n",
    "        values_fill=1\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd5f1776-fdd8-428b-aed5-fe46b1419df6",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_2 <- gsea_res %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=NES,\n",
    "        values_fill=0\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47e01794-d392-4f49-88b3-520251c6b6cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_label <- mat_1$col_label\n",
    "col_split <- mat_1$col_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a369b354-e980-4542-ab09-6becb9c9d7d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- mat_1 %>% dplyr::select(-col_label, -col_split) %>% t()\n",
    "mat_2 <- mat_2 %>% dplyr::select(-col_label, -col_split) %>% t()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cb59cfb-091f-490f-bd12-bc886ee98807",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_split <- gs_reduce %>% dplyr::select(term, parentTerm) %>% dplyr::rename(row_split_label=parentTerm) %>% dplyr::distinct()\n",
    "rownames(row_split) <- gs_reduce$term"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f4cbe1d-cbc0-48fb-b76d-45650164ee86",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_label[col_label==\"cDC.mig.\"] <- \"cDC mig.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c2c3e03-1886-4da8-8fe0-fdc9ca96bfa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/hm_gsea_ctl_d1_d3_wt_vs_ko_gobp.pdf\", width=10, height=50)\n",
    "\n",
    "gsea_hm(mat_1, mat_2, col_label=col_label, col_split=col_split, row_split=row_split, pathway_suffix=\"WP\", fontsize_select=2)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f368b363-996d-44f1-a264-0e6c24685d49",
   "metadata": {},
   "source": [
    "## Heat map log2FC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "243c8a79-5666-4643-95b2-18a298373ba2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data base for annotation \n",
    "tf <- read.table(\"/research/peer/fdeckert/reference/animaltfdb/Mus_musculus_TF.txt\")[[2]]\n",
    "\n",
    "lr <- CellChat::CellChatDB.mouse[[1]] %>% dplyr::select(pathway_name, ligand, receptor.symbol, receptor.family, annotation) %>% dplyr::filter(annotation %in% c(\"Cell-Cell Contact\", \"Secreted Signaling\", \"ECM-Receptor\")) %>% \n",
    "    separate_rows(receptor.symbol, sep=\", \") %>% dplyr::distinct() %>%\n",
    "    separate_rows(ligand, sep=\", \") %>% dplyr::distinct() %>% \n",
    "    dplyr::rename(pathway=pathway_name, receptor=receptor.symbol, family=receptor.family)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc87ca95-73fe-4bde-b516-10df9eacbbd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype_select <- c(\"cMo\", \"intMo\", \"RPM\", \"cDC2\", \"Basophil\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45777e04-7133-4291-888c-e1d12130ba41",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get sig. DEA results \n",
    "mat_1 <- lapply(dea_res_1, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2_thr)})\n",
    "mat_2 <- lapply(dea_res_2, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2_thr)})\n",
    "mat_3 <- lapply(dea_res_3, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2_thr)})\n",
    "mat_4 <- lapply(dea_res_4, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2_thr)})\n",
    "mat_5 <- lapply(dea_res_5, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2_thr)})\n",
    "mat_6 <- lapply(dea_res_6, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2_thr)})\n",
    "mat_7 <- lapply(dea_res_7, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2_thr)})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbd2bf9b-ddae-4ef0-aece-bf35685e36b9",
   "metadata": {},
   "source": [
    "## Shared response genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5a08015-cdbc-4c9f-b2f4-1c637edfeb2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DE response genes D1 and D3\n",
    "de_response_1 <- Map(function(...) do.call(rbind, list(...)), mat_1, mat_2)\n",
    "de_response_pos_1 <- lapply(de_response_1, function(x) {x %>% group_by(gene) %>% filter(n()>=2 & all(sign(avg_log2FC>0))) %>% dplyr::pull(gene) %>% unique()})\n",
    "de_response_neg_1 <- lapply(de_response_1, function(x) {x %>% group_by(gene) %>% filter(n()>=2 & all(sign(avg_log2FC<0))) %>% dplyr::pull(gene) %>% unique()})\n",
    "                     \n",
    "de_response_2 <- Map(function(...) do.call(rbind, list(...)), mat_3, mat_4)\n",
    "de_response_pos_2 <- lapply(de_response_2, function(x) {x %>% group_by(gene) %>% filter(n()>=2 & all(sign(avg_log2FC>0))) %>% dplyr::pull(gene) %>% unique()})\n",
    "de_response_neg_2 <- lapply(de_response_2, function(x) {x %>% group_by(gene) %>% filter(n()>=2 & all(sign(avg_log2FC<0))) %>% dplyr::pull(gene) %>% unique()})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9fc0965-0a6f-443f-af63-409fde102c7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get DE genes between genotypes \n",
    "de_genotype <- Map(function(...) do.call(rbind, list(...)), mat_5, mat_6, mat_7)\n",
    "de_genotype <- lapply(de_genotype, function(x) {x$gene %>% unique()})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5666b180-e755-429d-bf7b-d3aa2495c329",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_samples <- c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_LysM_cre_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_LysM_cre_CpG_D1\", \"IFNAR_fl_CpG_D3\", \"IFNAR_fl_LysM_cre_CpG_D3\")\n",
    "col_split <- c(\"Ctl\", \"Ctl\", \"D1\", \"D1\", \"D3\", \"D3\")\n",
    "col_label <- c(\"+/+\", \"cre/+\", \"+/+\", \"cre/+\", \"+/+\", \"cre/+\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "524a10d8-ae1f-40b9-8a2c-99fab7253d6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_dea_fetch <- function(celltype_select) {\n",
    "\n",
    "    genes_pos <- c(de_response_pos_1[[celltype_select]], de_response_pos_2[[celltype_select]]) %>% unique()\n",
    "    genes_neg <- c(de_response_neg_2[[celltype_select]], de_response_neg_2[[celltype_select]]) %>% unique()\n",
    "\n",
    "    genes_pos <- genes_pos[!genes_pos %in% de_genotype[[celltype_select]]]\n",
    "    genes_neg <- genes_neg[!genes_neg %in% de_genotype[[celltype_select]]]\n",
    "\n",
    "    genes <- c(genes_pos, genes_neg)\n",
    "    row_split <- c(rep(\"up\", length(genes_pos)), rep(\"down\", length(genes_neg)))\n",
    "\n",
    "    # Save gene list with figures \n",
    "    write.csv(rbind(data.frame(genes=genes_pos, response=\"Up\"), data.frame(genes=genes_pos, response=\"Down\")), paste0(\"result/dea/scRNAseq/wilcox/fetch/hm_dea_d1_d3_vs_ctl_shared_\", tolower(celltype_select), \".csv\"), row.names=FALSE)\n",
    "\n",
    "    # Get pseudobulk counts \n",
    "    so_i <- so[, so$celltype_low==celltype_select & so$sample_group %in% col_samples]\n",
    "\n",
    "    # Get pseudobulk counts \n",
    "    mat <- GetAssayData(so_i, assay=\"RNA\", layer=\"counts\")\n",
    "    mat <- split(t(mat) %>% as.data.frame(), f=so_i$sample_group)\n",
    "\n",
    "    # Normalize and scale \n",
    "    mat <- lapply(mat, colSums) %>% do.call(rbind, .) %>% t()\n",
    "    mat <- t(t(mat) / colSums(mat)) *1e4\n",
    "    mat <- log1p(mat)\n",
    "    mat <- t(scale(t(as.matrix(mat)), center=TRUE, scale=TRUE))\n",
    "\n",
    "    # Select genes and sample order \n",
    "    mat <- mat[genes, col_samples]\n",
    "\n",
    "    row_genes <- c(\n",
    "    \n",
    "        \"Mki67\", \"Pcna\", \n",
    "        \"Spic\", \n",
    "        \"Ly6c2\", \n",
    "        \"Adgre\", \n",
    "        \"Cd163\", \n",
    "        \"Arg1\", \n",
    "        \"Gdf15\", \n",
    "        \"Spi1\", \n",
    "        \"Mertk\", \"Axl\", \"Timd4\", \"Marco\",  \n",
    "        \"Cebpa\", \"Cebpb\", \"Nr4a1\", \n",
    "        genes_pos[genes %in% tf], \n",
    "        genes_pos[genes %in% lr$ligand], \n",
    "        genes_pos[genes %in% lr$receptor], \n",
    "        grep(\"Ifit\", genes, value=TRUE), \n",
    "        grep(\"Irf\", genes, value=TRUE),\n",
    "        grep(\"Oas\", genes, value=TRUE), \n",
    "        grep(\"Isg\", genes, value=TRUE)\n",
    "    \n",
    "    )\n",
    "\n",
    "    dea_res_hm(mat, row_split=row_split, col_label=col_label, col_split=col_split, celltype=celltype_select, row_genes=row_genes, fontsize_select=2)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66df4161-2bf0-4485-b2ec-9a9ab199e849",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/hm_dea_d1_d3_vs_ctl_shared_cmo.pdf\", width=2, height=15)\n",
    "\n",
    "hm_dea_fetch(\"cMo\")\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5219fa8-f906-476c-888b-9bcffa4995d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/hm_dea_d1_d3_vs_ctl_shared_intmo.pdf\", width=2, height=10)\n",
    "\n",
    "hm_dea_fetch(\"intMo\")\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dd10bad-9535-4bbe-b710-d039b1fb42a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/hm_dea_d1_d3_vs_ctl_shared_rpm.pdf\", width=2, height=10)\n",
    "\n",
    "hm_dea_fetch(\"RPM\")\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02e68fc3-c9a4-4be9-b43c-80b0c8dd57f0",
   "metadata": {},
   "source": [
    "## DE pos response genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37c695f4-4442-4bff-ac4c-c3fd8c11b94c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DE response genes D1 and D3\n",
    "de_response_1 <- Map(function(...) do.call(rbind, list(...)), dea_res_1, dea_res_2)\n",
    "de_response_1 <- lapply(de_response_1, function(x) {x %>% dplyr::group_by(gene) %>% dplyr::filter(all(sign(avg_log2FC)>0)) %>% dplyr::pull(gene) %>% unique()})\n",
    "\n",
    "de_response_2 <- Map(function(...) do.call(rbind, list(...)), dea_res_3, dea_res_4)\n",
    "de_response_2 <- lapply(de_response_2, function(x) {x %>% dplyr::group_by(gene) %>% dplyr::filter(all(sign(avg_log2FC)>0)) %>% dplyr::pull(gene) %>% unique()})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96397274-d971-48bc-80e7-fb9760cf830d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get DE baseline genes \n",
    "de_genotype_ctl <- lapply(dea_res_5, function(x) {x %>% dplyr::pull(gene)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8549ae5-ce5f-4f07-acf1-228e7c323b46",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get DE genes between genotypes \n",
    "de_genotype <- Map(function(...) do.call(rbind, list(...)), dea_res_6, dea_res_7)\n",
    "de_genotype_pos <- lapply(de_genotype, function(x) {x %>% group_by(gene) %>% filter(all(sign(avg_log2FC>0))) %>% dplyr::pull(gene) %>% unique()})\n",
    "de_genotype_neg <- lapply(de_genotype, function(x) {x %>% group_by(gene) %>% filter(all(sign(avg_log2FC<0))) %>% dplyr::pull(gene) %>% unique()})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbc3e594-5d10-4de5-993b-7646d027c1e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_samples <- c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_LysM_cre_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_LysM_cre_CpG_D1\", \"IFNAR_fl_CpG_D3\", \"IFNAR_fl_LysM_cre_CpG_D3\")\n",
    "col_split <- c(\"Ctl\", \"Ctl\", \"D1\", \"D1\", \"D3\", \"D3\")\n",
    "col_label <- c(\"+/+\", \"cre/+\", \"+/+\", \"cre/+\", \"+/+\", \"cre/+\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54e57423-f7bc-442c-b027-5f295dd67c96",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_dea_fetch <- function(celltype_select) {\n",
    "\n",
    "    genes_pos <- de_genotype_pos[[celltype_select]]\n",
    "    genes_neg <- de_genotype_neg[[celltype_select]]\n",
    "\n",
    "    genes_pos <- genes_pos[!genes_pos %in% de_genotype_ctl[[celltype_select]]]\n",
    "    genes_neg <- genes_neg[!genes_neg %in% de_genotype_ctl[[celltype_select]]]\n",
    "\n",
    "    genes_pos <- genes_pos[genes_pos %in% c(de_response_1[[celltype_select]])]\n",
    "    genes_neg <- genes_neg[genes_neg %in% c(de_response_1[[celltype_select]])]\n",
    "\n",
    "    print(intersect(genes_pos, genes_neg))\n",
    "\n",
    "    genes <- c(genes_pos, genes_neg)\n",
    "    row_split <- c(rep(\"up\", length(genes_pos)), rep(\"down\", length(genes_neg)))\n",
    "\n",
    "    # Save gene list with figures \n",
    "    write.csv(rbind(data.frame(genes=genes_pos, response=\"Up (WT)\"), data.frame(genes=genes_pos, response=\"Down (WT)\")), paste0(\"result/dea/scRNAseq/wilcox/fetch/hm_dea_pos_response_wt_vs_ko_\", tolower(celltype_select), \".csv\"), row.names=FALSE)\n",
    "\n",
    "    # Get pseudobulk counts \n",
    "    so_i <- so[, so$celltype_low==celltype_select & so$sample_group %in% col_samples]\n",
    "\n",
    "    # Get pseudobulk counts \n",
    "    mat <- GetAssayData(so_i, assay=\"RNA\", layer=\"counts\")\n",
    "    mat <- split(t(mat) %>% as.data.frame(), f=so_i$sample_group)\n",
    "\n",
    "    # Normalize and scale \n",
    "    mat <- lapply(mat, colSums) %>% do.call(rbind, .) %>% t()\n",
    "    mat <- t(t(mat) / colSums(mat)) *1e4\n",
    "    mat <- log1p(mat)\n",
    "    mat <- t(scale(t(as.matrix(mat)), center=TRUE, scale=TRUE))\n",
    "\n",
    "    # Select genes and sample order \n",
    "    mat <- mat[genes, col_samples]\n",
    "\n",
    "    row_genes <- c(\n",
    "    \n",
    "        \"Mki67\", \"Pcna\", \n",
    "        \"Spic\", \n",
    "        \"Ly6c2\", \n",
    "        \"Adgre\", \n",
    "        \"Cd163\", \n",
    "        \"Arg1\", \n",
    "        \"Gdf15\", \n",
    "        \"Spi1\", \n",
    "        \"Mertk\", \"Axl\", \"Timd4\", \"Marco\",  \n",
    "        \"Cebpa\", \"Cebpb\", \"Nr4a1\", \n",
    "        genes_pos[genes %in% tf], \n",
    "        genes_pos[genes %in% lr$ligand], \n",
    "        genes_pos[genes %in% lr$receptor], \n",
    "        grep(\"Ifit\", genes, value=TRUE), \n",
    "        grep(\"Irf\", genes, value=TRUE),\n",
    "        grep(\"Oas\", genes, value=TRUE), \n",
    "        grep(\"Isg\", genes, value=TRUE)\n",
    "    \n",
    "    )\n",
    "\n",
    "    dea_res_hm(mat, row_split=row_split, col_label=col_label, col_split=col_split, celltype=celltype_select, row_genes=row_genes, fontsize_select=2)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c88cae8-f34b-4e3d-8143-106900a036d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/hm_dea_pos_response_wt_vs_ko_cmo.pdf\", width=2, height=10)\n",
    "\n",
    "hm_dea_fetch(\"cMo\")\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b570116e-4046-4360-a0ff-6505bd6fed5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/hm_dea_pos_response_wt_vs_ko_intmo.pdf\", width=2, height=10)\n",
    "\n",
    "hm_dea_fetch(\"intMo\")\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbde7adc-a2cf-4705-bf7b-b2ce060fc3ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/hm_dea_pos_response_wt_vs_ko_rpm.pdf\", width=2, height=10)\n",
    "\n",
    "hm_dea_fetch(\"RPM\")\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46e8f4ed-bede-42e4-9152-ff3c49afca60",
   "metadata": {},
   "source": [
    "## Violine gene expression plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4168949f-ccb8-43c3-a8dd-da784c206c3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_select <- c(\"Ly6c2\", \"Spic\", \"Cd163\", \"Gdf15\", \"Ccl2\", \"Slfn4\", \"Bcl2l1\", \"Hmox1\", \"Jun\")\n",
    "celltype_select <- c(\"cMo\", \"intMo\", \"RPM\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb9f59b6-a237-4468-892b-2d1d19272111",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat <- cbind(so@meta.data[, c(\"sample_group\", \"celltype_low\"), drop=FALSE], t(GetAssayData(so, assay=\"RNA\", slot=\"data\")[gene_select, ])) %>% \n",
    "    pivot_longer(cols=-c(sample_group, celltype_low), names_to=\"gene\", values_to=\"expression\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06a030b5-a3e0-4e6b-a25a-1367fe96696e",
   "metadata": {},
   "outputs": [],
   "source": [
    "map <- data.frame(\n",
    "    \n",
    "    sample_group=c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_LysM_cre_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_LysM_cre_CpG_D1\", \"IFNAR_fl_CpG_D3\", \"IFNAR_fl_LysM_cre_CpG_D3\"),\n",
    "    time_point=c(\"Ctl\", \"Ctl\", \"D1\", \"D1\", \"D3\", \"D3\"),\n",
    "    genotype=c(\"+/+\", \"cre/+\", \"+/+\", \"cre/+\", \"+/+\", \"cre/+\")\n",
    ")\n",
    "\n",
    "mat <- mat %>% dplyr::left_join(., map, by=join_by(sample_group)) %>% \n",
    "    dplyr::mutate(gene=factor(gene, levels=gene_select))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7ee6260-001a-485d-84db-7619e4a3d9cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "genotype <- c(\"#732E8F\", \"#D390BF\", \"#FA8775\")\n",
    "names(genotype) <- c(\"WT\", \"+/+\", \"cre/+\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d442739b-0a5a-4ce6-a930-1ba8def20c61",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=5, repr.plot.height=7.5)\n",
    "\n",
    "exp_p <- ggplot(mat %>% dplyr::filter(celltype_low %in% celltype_select), aes(x=time_point, y=expression, color=sample_group, fill=sample_group)) + \n",
    "\n",
    "        geom_jitter(position=position_jitterdodge(jitter.width=0.20, dodge.width=1.0), shape=21, stroke=0.1, size=0.75, show.legend=FALSE, color=\"black\") + \n",
    "        # geom_boxplot(position=position_dodge(width=1.0), width=0.20, color=\"black\", alpha=1, outlier.color=NA) + \n",
    "        \n",
    "        xlab(\"\") + ylab(\"log(CPTT)\") + \n",
    "        ggtitle(\"Spic\") + \n",
    "        facet_grid(gene~celltype_low, scales=\"free\") + \n",
    "        scale_color_manual(values=color$sample_group) + \n",
    "        scale_fill_manual(values=color$sample_group)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21cc23ae-0b66-47e4-86ec-efc515519adf",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/exp_myeloid.pdf\", width=3, height=8)\n",
    "\n",
    "set_height <- 1 * length(gene_select)\n",
    "set_width <- 3 * length(celltype_select)\n",
    "\n",
    "exp_p + ggh4x::force_panelsizes(cols=unit(set_width, \"mm\"), total_height=unit(set_height, \"cm\"))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e0d441e-f38d-43a0-8ff0-99c3b4fc6a0a",
   "metadata": {},
   "source": [
    "# Volcano plots  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11e942bf-4f62-43a3-b4af-af58db021e57",
   "metadata": {},
   "outputs": [],
   "source": [
    "label <- c(\"Icam1\", \"Itga4\", \"Itgb1\", \"Vcam1\", \"Gdf15\", \"Ccl2\", \"Tnf\", \"Il6\", \"Tgfb1\", \"Stat1\", \"Stat2\", \"Spic\", \"Irf8\", \"Siglec1\", \"Slc40a1\", \"Hmox1\", \"Blvrb\", \"Treml4\", \"Back1\")\n",
    "# label <- NULL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8f42ed7-cdf9-4104-84e3-bf1f27a2c398",
   "metadata": {},
   "outputs": [],
   "source": [
    "source(\"bin/dea_pl.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc4c81b2-ebd6-467b-bd09-dd1419e21664",
   "metadata": {},
   "outputs": [],
   "source": [
    "v_pl_1 <- v_pl(dea_res_5[[\"cMo\"]], title=\"cMo\", log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=0.5, color_pos=unlist(color$sample_group[\"IFNAR_fl_Baseline_D0\"]), color_neg=unlist(color$sample_group[\"IFNAR_fl_LysM_cre_Baseline_D0\"]), label=label, label_size=2, y_limit=4, aspect_ratio=0.8, top_label=5) + theme(legend.position=\"none\") + theme_global_set(size_select=4)\n",
    "v_pl_2 <- v_pl(dea_res_6[[\"cMo\"]], title=\"cMo\", log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=0.5, color_pos=unlist(color$sample_group[\"IFNAR_fl_Baseline_D0\"]), color_neg=unlist(color$sample_group[\"IFNAR_fl_LysM_cre_Baseline_D0\"]), label=label, label_size=2, y_limit=4, aspect_ratio=0.8, top_label=5) + theme(legend.position=\"none\") + theme_global_set(size_select=4)\n",
    "v_pl_3 <- v_pl(dea_res_7[[\"cMo\"]], title=\"cMo\", log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=0.5, color_pos=unlist(color$sample_group[\"IFNAR_fl_Baseline_D0\"]), color_neg=unlist(color$sample_group[\"IFNAR_fl_LysM_cre_Baseline_D0\"]), label=label, label_size=2, y_limit=4, aspect_ratio=0.8, top_label=5) + theme(legend.position=\"none\") + theme_global_set(size_select=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b91def5-6284-404c-bf23-039b79e543fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "v_pl_4 <- v_pl(dea_res_5[[\"intMo\"]], title=\"intMo\", log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=0.5, color_pos=unlist(color$sample_group[\"IFNAR_fl_CpG_D1\"]), color_neg=unlist(color$sample_group[\"IFNAR_fl_LysM_cre_CpG_D1\"]), label=label, label_size=2, y_limit=4, aspect_ratio=0.8, top_label=5) + theme(legend.position=\"none\") + theme_global_set(size_select=4)\n",
    "v_pl_5 <- v_pl(dea_res_6[[\"intMo\"]], title=\"intMo\", log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=0.5, color_pos=unlist(color$sample_group[\"IFNAR_fl_CpG_D1\"]), color_neg=unlist(color$sample_group[\"IFNAR_fl_LysM_cre_CpG_D1\"]), label=label, label_size=2, y_limit=4, aspect_ratio=0.8, top_label=5) + theme(legend.position=\"none\") + theme_global_set(size_select=4)\n",
    "v_pl_6 <- v_pl(dea_res_7[[\"intMo\"]], title=\"intMo\", log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=0.5, color_pos=unlist(color$sample_group[\"IFNAR_fl_CpG_D1\"]), color_neg=unlist(color$sample_group[\"IFNAR_fl_LysM_cre_CpG_D1\"]), label=label, label_size=2, y_limit=4, aspect_ratio=0.8, top_label=5) + theme(legend.position=\"none\") + theme_global_set(size_select=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "083742e3-6a82-47b7-9831-a9194085d935",
   "metadata": {},
   "outputs": [],
   "source": [
    "v_pl_7 <- v_pl(dea_res_5[[\"RPM\"]], title=\"RPM\", log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=0.5, color_pos=unlist(color$sample_group[\"IFNAR_fl_CpG_D3\"]), color_neg=unlist(color$sample_group[\"IFNAR_fl_LysM_cre_CpG_D3\"]), label=label, label_size=2, y_limit=4, aspect_ratio=0.8, top_label=5) + theme(legend.position=\"none\") + theme_global_set(size_select=4)\n",
    "v_pl_8 <- v_pl(dea_res_6[[\"RPM\"]], title=\"RPM\", log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=0.5, color_pos=unlist(color$sample_group[\"IFNAR_fl_CpG_D3\"]), color_neg=unlist(color$sample_group[\"IFNAR_fl_LysM_cre_CpG_D3\"]), label=label, label_size=2, y_limit=4, aspect_ratio=0.8, top_label=5) + theme(legend.position=\"none\") + theme_global_set(size_select=4)\n",
    "v_pl_9 <- v_pl(dea_res_7[[\"RPM\"]], title=\"RPM\", log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=0.5, color_pos=unlist(color$sample_group[\"IFNAR_fl_CpG_D3\"]), color_neg=unlist(color$sample_group[\"IFNAR_fl_LysM_cre_CpG_D3\"]), label=label, label_size=2, y_limit=4, aspect_ratio=0.8, top_label=5) + theme(legend.position=\"none\") + theme_global_set(size_select=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "523e4913-944c-4a8a-924f-93aa60bfb730",
   "metadata": {},
   "outputs": [],
   "source": [
    "v_pl_10 <- v_pl(dea_res_5[[\"MEP\"]], title=\"MEP\", log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=0.5, color_pos=unlist(color$sample_group[\"IFNAR_fl_CpG_D3\"]), color_neg=unlist(color$sample_group[\"IFNAR_fl_LysM_cre_CpG_D3\"]), label=label, label_size=2, y_limit=4, aspect_ratio=0.8, top_label=5) + theme(legend.position=\"none\") + theme_global_set(size_select=4)\n",
    "v_pl_11 <- v_pl(dea_res_6[[\"MEP\"]], title=\"MEP\", log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=0.5, color_pos=unlist(color$sample_group[\"IFNAR_fl_CpG_D3\"]), color_neg=unlist(color$sample_group[\"IFNAR_fl_LysM_cre_CpG_D3\"]), label=label, label_size=2, y_limit=4, aspect_ratio=0.8, top_label=5) + theme(legend.position=\"none\") + theme_global_set(size_select=4)\n",
    "v_pl_12 <- v_pl(dea_res_7[[\"MEP\"]], title=\"MEP\", log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=0.5, color_pos=unlist(color$sample_group[\"IFNAR_fl_CpG_D3\"]), color_neg=unlist(color$sample_group[\"IFNAR_fl_LysM_cre_CpG_D3\"]), label=label, label_size=2, y_limit=4, aspect_ratio=0.8, top_label=5) + theme(legend.position=\"none\") + theme_global_set(size_select=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "014339dd-10b1-4203-8785-c79803dbf9d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "v_pl_13 <- v_pl(dea_res_5[[\"Proerythroblast\"]], title=\"ProEB\", log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=0.5, color_pos=unlist(color$sample_group[\"IFNAR_fl_CpG_D3\"]), color_neg=unlist(color$sample_group[\"IFNAR_fl_LysM_cre_CpG_D3\"]), label=label, label_size=2, y_limit=4, aspect_ratio=0.8, top_label=5) + theme(legend.position=\"none\") + theme_global_set(size_select=4)\n",
    "v_pl_14 <- v_pl(dea_res_6[[\"Proerythroblast\"]], title=\"ProEB\", log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=0.5, color_pos=unlist(color$sample_group[\"IFNAR_fl_CpG_D3\"]), color_neg=unlist(color$sample_group[\"IFNAR_fl_LysM_cre_CpG_D3\"]), label=label, label_size=2, y_limit=4, aspect_ratio=0.8, top_label=5) + theme(legend.position=\"none\") + theme_global_set(size_select=4)\n",
    "v_pl_15 <- v_pl(dea_res_7[[\"Proerythroblast\"]], title=\"ProEB\", log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=0.5, color_pos=unlist(color$sample_group[\"IFNAR_fl_CpG_D3\"]), color_neg=unlist(color$sample_group[\"IFNAR_fl_LysM_cre_CpG_D3\"]), label=label, label_size=2, y_limit=4, aspect_ratio=0.8, top_label=5) + theme(legend.position=\"none\") + theme_global_set(size_select=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4aabe5e-fb32-4b23-9208-08278af383e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "v_pl_16 <- v_pl(dea_res_5[[\"Erythroblast\"]], title=\"EB\", log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=0.5, color_pos=unlist(color$sample_group[\"IFNAR_fl_CpG_D3\"]), color_neg=unlist(color$sample_group[\"IFNAR_fl_LysM_cre_CpG_D3\"]), label=label, label_size=2, y_limit=4, aspect_ratio=0.8, top_label=5) + theme(legend.position=\"none\") + theme_global_set(size_select=4)\n",
    "v_pl_17 <- v_pl(dea_res_6[[\"Erythroblast\"]], title=\"EB\", log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=0.5, color_pos=unlist(color$sample_group[\"IFNAR_fl_CpG_D3\"]), color_neg=unlist(color$sample_group[\"IFNAR_fl_LysM_cre_CpG_D3\"]), label=label, label_size=2, y_limit=4, aspect_ratio=0.8, top_label=5) + theme(legend.position=\"none\") + theme_global_set(size_select=4)\n",
    "v_pl_18 <- v_pl(dea_res_7[[\"Erythroblast\"]], title=\"EB\", log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=0.5, color_pos=unlist(color$sample_group[\"IFNAR_fl_CpG_D3\"]), color_neg=unlist(color$sample_group[\"IFNAR_fl_LysM_cre_CpG_D3\"]), label=label, label_size=2, y_limit=4, aspect_ratio=0.8, top_label=5) + theme(legend.position=\"none\") + theme_global_set(size_select=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e873235a-2be6-45db-9c85-846a837129bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/v_pl_cg_vs_ko.pdf\", width=3*2, height=6*2)\n",
    "\n",
    "gridExtra::grid.arrange(\n",
    "    \n",
    "    v_pl_1 %>% egg::set_panel_size(., width=unit(3, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    v_pl_2 %>% egg::set_panel_size(., width=unit(3, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    v_pl_3 %>% egg::set_panel_size(., width=unit(3, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    v_pl_4 %>% egg::set_panel_size(., width=unit(3, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    v_pl_5 %>% egg::set_panel_size(., width=unit(3, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    v_pl_6 %>% egg::set_panel_size(., width=unit(3, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    v_pl_7 %>% egg::set_panel_size(., width=unit(3, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    v_pl_8 %>% egg::set_panel_size(., width=unit(3, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    v_pl_9 %>% egg::set_panel_size(., width=unit(3, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    v_pl_10 %>% egg::set_panel_size(., width=unit(3, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    v_pl_11 %>% egg::set_panel_size(., width=unit(3, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    v_pl_12 %>% egg::set_panel_size(., width=unit(3, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    v_pl_13 %>% egg::set_panel_size(., width=unit(3, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    v_pl_14 %>% egg::set_panel_size(., width=unit(3, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    v_pl_15 %>% egg::set_panel_size(., width=unit(3, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    v_pl_16 %>% egg::set_panel_size(., width=unit(3, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    v_pl_17 %>% egg::set_panel_size(., width=unit(3, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    v_pl_18 %>% egg::set_panel_size(., width=unit(3, \"cm\"), height=unit(2.5, \"cm\")), ncol=3\n",
    "\n",
    ")\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff664abc-1793-481d-8b25-cefbb946187e",
   "metadata": {},
   "source": [
    "# CCI DEA "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec59b755-f507-4b45-aa40-e32141627ea4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load source \n",
    "source(\"bin/cci_pp.R\")\n",
    "source(\"bin/cci_pl.R\")\n",
    "\n",
    "# Parameter \n",
    "pval_prob_thr <- 0.05\n",
    "p_val_adj_dea <- 0.05\n",
    "pathway_name_thr <- c(\"ICAM\", \"VCAM\", \"SIRP\", \"KIT\", \"TGFb\", \"TNF\", \"CXCL\", \"CCL\")\n",
    "\n",
    "# CellChat results\n",
    "lr_res <- read.csv(\"result/cci/lr_res.csv\", row.names=1)\n",
    "\n",
    "# Add reversed Cell-cell interactions \n",
    "db <- CellChat::CellChatDB.mouse\n",
    "\n",
    "# Impute custom interactions\n",
    "options(stringsAsFactors=FALSE)\n",
    "interaction <- read.csv(\"data/reference/cellchat/interaction.csv\", row.names=1, na.strings=c(\"\"))\n",
    "interaction[is.na(interaction)] <- \"\"\n",
    "db$interaction <- rbind(db$interaction, interaction)\n",
    "\n",
    "# Get cell-cell contact interaction \n",
    "db <- CellChat::subsetDB(db, search=c(\"Cell-Cell Contact\"), key=\"annotation\")\n",
    "\n",
    "# Subset lr results for cell-cell contact forward \n",
    "lr_res_ccc_fwd <- lr_res %>% dplyr::filter(lr_res$interaction_name %in% db$interaction$interaction_name)\n",
    "\n",
    "# Reverse cell-cell contact \n",
    "lr_res_ccc_rev <- data.frame(\n",
    "\n",
    "    source=lr_res_ccc_fwd$target, \n",
    "    target=lr_res_ccc_fwd$source, \n",
    "    interaction_name=paste0(lr_res_ccc_fwd$receptor_symbol, \"_\", lr_res_ccc_fwd$ligand_symbol), \n",
    "    prob=lr_res_ccc_fwd$prob, \n",
    "    pval=lr_res_ccc_fwd$pval, \n",
    "    pathway_name=lr_res_ccc_fwd$pathway_name, \n",
    "    ligand_symbol=lr_res_ccc_fwd$receptor_symbol, \n",
    "    receptor_symbol=lr_res_ccc_fwd$ligand_symbol, \n",
    "    sample_group=lr_res_ccc_fwd$sample_group\n",
    "    \n",
    ")\n",
    "\n",
    "# Combine results \n",
    "lr_res_ccc <- rbind(lr_res_ccc_fwd, lr_res_ccc_rev)\n",
    "\n",
    "# Assemble LR results \n",
    "lr_res <- rbind(lr_res[!lr_res$interaction_name %in% lr_res_ccc$interaction_name, ], lr_res_ccc)\n",
    "\n",
    "# Subset by interaction pathway\n",
    "lr_res <- lr_res[lr_res$pathway_name %in% pathway_name_thr, ]\n",
    "lr_res$pathway_name <- factor(lr_res$pathway_name, levels=pathway_name_thr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67410385-21c6-4e7a-8671-1fdbdadf36b5",
   "metadata": {},
   "source": [
    "# DEA results "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d60c5142-d878-4dfe-957d-426e2a819a7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "dea_mat_5 <- dea_res_5[c(\"cMo\", \"intMo\", \"RPM\", \"cDC2\", \"Basophil\", \"MEP\", \"Proerythroblast\", \"Erythroblast\")]\n",
    "dea_mat_5 <- lapply(dea_mat_5, function(x) {list(\"Ctl\"=x)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1347d3d0-1bde-44f0-882f-3102ba6ba5ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "dea_mat_6<- dea_res_6[c(\"cMo\", \"intMo\", \"RPM\", \"cDC2\", \"Basophil\", \"MEP\", \"Proerythroblast\", \"Erythroblast\")]\n",
    "dea_mat_6 <- lapply(dea_mat_6, function(x) {list(\"D1\"=x)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f3261de-4089-4970-807f-50ace3e7ae72",
   "metadata": {},
   "outputs": [],
   "source": [
    "dea_mat_7 <- dea_res_7[c(\"cMo\", \"intMo\", \"RPM\", \"cDC2\", \"Basophil\", \"MEP\", \"Proerythroblast\", \"Erythroblast\")]\n",
    "dea_mat_7 <- lapply(dea_mat_7, function(x) {list(\"D3\"=x)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fa13ead-4344-4d8d-91bf-9fe99ff9bcd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "dea_mat <- Map(c, dea_mat_5, dea_mat_6, dea_mat_7)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c48a72b-db65-4fd1-a47a-81423d4d9799",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## PTPG B (+/+) vs D1 (+/+) and D3 (+/+)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f08fc2ce-da91-4161-80a9-ed3281425407",
   "metadata": {},
   "source": [
    "## cMo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "003fb3e9-c960-4a2d-8241-b743963df5d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"cMo\")\n",
    "target <- c(\"MEP\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "462e97b1-b38a-4f27-a5bd-7716913ac0a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80d21cd2-cd95-4e0b-95b0-c629a4b37417",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/cci_cmo_mep_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18e4ce5b-ea26-4e93-936a-0048d8f7e2d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"cMo\")\n",
    "target <- c(\"Proerythroblast\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8526ccc3-5d19-42bc-8f8a-9febc3749351",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d1a01fd-b574-4c76-8901-092caa18b404",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/cci_cmo_proeb_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d290092c-21c1-431a-ab30-1cf2562eff89",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"cMo\")\n",
    "target <- c(\"Erythroblast\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3fcc361-927a-4b1b-a265-80fc3cc8d5d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e2f70a2-0f05-46ea-a89b-f1974e85494e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/cci_cmo_eb_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f691ebc-9fc2-4997-bb21-bac8cbb90db6",
   "metadata": {},
   "source": [
    "## intMo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b65a947-44dc-4f2a-a16c-105d000e247a",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"intMo\")\n",
    "target <- c(\"MEP\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e643f889-0815-49f9-9034-eb6ba03f5bef",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aac4fd3a-632f-4bd9-8eef-2cade700a7d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/cci_intmo_mep_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f12cfed-1372-4a14-b341-6e776b765d6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"intMo\")\n",
    "target <- c(\"Proerythroblast\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16f44f38-859b-4337-a545-5c2e05da9178",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71a43bc6-400e-47b1-b23a-66a4e944d4ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/cci_intmo_proeb_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69480d57-2a72-41d9-8385-52117e0012a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"intMo\")\n",
    "target <- c(\"Erythroblast\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16af9b97-b869-4128-a6bb-194a597f5d5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba316e76-2191-48ab-a35a-1fe2a7d33a71",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/cci_intmo_eb_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "074532ad-a423-4100-b9b0-1a3d1904063b",
   "metadata": {},
   "source": [
    "## RPM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c89da626-687b-49c2-94b0-1cddcd97f8d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"RPM\")\n",
    "target <- c(\"MEP\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bba0449-c718-42f4-a8d4-eb16a204f4f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc02eac0-37bc-486e-9f4a-422f17f6604e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/cci_rpm_mep_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c90c12bd-e06c-4fa7-b3c8-e8842916db4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"RPM\")\n",
    "target <- c(\"Proerythroblast\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3225ebbc-095d-4d91-ae0d-e5d6b6d3f6a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a6c0dec-fcd0-4ef0-b368-a2892ae33c01",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/cci_rpm_proeb_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58ba8d68-224a-4ce3-9412-4623d6fbf1f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"RPM\")\n",
    "target <- c(\"Erythroblast\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72af504a-8e9e-4139-94af-1e93e722fc0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39388a7c-74b7-42d3-a138-248cd8600c7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/cci_rpm_eb_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62c60458-8f68-4833-8c7e-967ef4efe998",
   "metadata": {},
   "source": [
    "# Interaction plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca3919c0-95db-40cf-99bb-7fd54e463343",
   "metadata": {},
   "outputs": [],
   "source": [
    "sources <- c(\"RPM\", \"intMo\", \"cMo\")\n",
    "targets <- c(\"MEP\", \"Proerythroblast\", \"Erythroblast\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0037cc6c-331c-4ab5-a9d9-16b1cc3c9416",
   "metadata": {},
   "outputs": [],
   "source": [
    "cco_1 <- readRDS(\"data/scRNAseq/cci/cco_ifnar_fl_baseline_d0.rds\")\n",
    "cco_2 <- readRDS(\"data/scRNAseq/cci/cco_ifnar_fl_cpg_d1.rds\")\n",
    "cco_3 <- readRDS(\"data/scRNAseq/cci/cco_ifnar_fl_cpg_d3.rds\")\n",
    "cco_4 <- readRDS(\"data/scRNAseq/cci/cco_ifnar_fl_lysm_cre_baseline_d0.rds\")\n",
    "cco_5 <- readRDS(\"data/scRNAseq/cci/cco_ifnar_fl_lysm_cre_cpg_d1.rds\")\n",
    "cco_6 <- readRDS(\"data/scRNAseq/cci/cco_ifnar_fl_lysm_cre_cpg_d3.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3011b27-7789-4eca-8e3d-ab0ad506eafb",
   "metadata": {},
   "outputs": [],
   "source": [
    "cco_1 <- subsetCellChat(object=cco_1, idents.use=c(sources, targets))\n",
    "cco_2 <- subsetCellChat(object=cco_2, idents.use=c(sources, targets))\n",
    "cco_3 <- subsetCellChat(object=cco_3, idents.use=c(sources, targets))\n",
    "cco_4 <- subsetCellChat(object=cco_4, idents.use=c(sources, targets))\n",
    "cco_5 <- subsetCellChat(object=cco_5, idents.use=c(sources, targets))\n",
    "cco_6 <- subsetCellChat(object=cco_6, idents.use=c(sources, targets))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f3e2c0c-fef7-4539-880e-c30cca6baea8",
   "metadata": {},
   "outputs": [],
   "source": [
    "levels(cco_1@idents) <- c(targets, sources)\n",
    "levels(cco_2@idents) <- c(targets, sources)\n",
    "levels(cco_3@idents) <- c(targets, sources)\n",
    "levels(cco_4@idents) <- c(targets, sources)\n",
    "levels(cco_5@idents) <- c(targets, sources)\n",
    "levels(cco_6@idents) <- c(targets, sources)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b7bf337-daf2-4d4f-a30e-8507c6974e97",
   "metadata": {},
   "outputs": [],
   "source": [
    "color_use <- color$celltype_low[names(color$celltype_low) %in% c(sources, targets)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10ec20f9-774b-462e-ac37-991ccb3d2020",
   "metadata": {},
   "outputs": [],
   "source": [
    "cco_list <- list(cco_1, cco_2, cco_3, cco_4, cco_5, cco_6)\n",
    "\n",
    "q_prob_thr <- quantile(c(cco_1@net$prob, cco_2@net$prob, cco_3@net$prob), 0.90)\n",
    "for (i in seq_along(cco_list)) {\n",
    "    \n",
    "    prob_arr <- cco_list[[i]]@net$prob\n",
    "    prob_arr[prob_arr > q_prob_thr] <- q_prob_thr\n",
    "    cco_list[[i]]@net$prob <- prob_arr\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb06d56a-6d68-4be4-9277-a21e7987ef04",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/cp_cci_net_ctl_ko.pdf\", width=2.0, height=2.0)\n",
    "\n",
    "    replayPlot(netVisual_chord_gene(cco_list[[4]], slot.name=\"net\", sources.use=sources, targets.use=targets, signaling=pathway_name_thr, color.use=color_use, title.name=NULL, show.legend=FALSE, small.gap=1, big.gap=5, lab.cex=0.5))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4deee3b9-f1f4-4764-b9d8-ea6b10dd66d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/cp_cci_net_d1_ko.pdf\", width=2.0, height=2.0)\n",
    "\n",
    "    replayPlot(netVisual_chord_gene(cco_list[[5]], slot.name=\"net\", sources.use=sources, targets.use=targets, signaling=pathway_name_thr, color.use=color_use, title.name=NULL, show.legend=FALSE, small.gap=1, big.gap=5, lab.cex=0.5))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11ad2ca6-235a-462a-b090-a0804e3e3092",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/figure_6/cp_cci_net_d3_ko.pdf\", width=2.0, height=2.0)\n",
    "\n",
    "    replayPlot(netVisual_chord_gene(cco_list[[6]], slot.name=\"net\", sources.use=sources, targets.use=targets, signaling=pathway_name_thr, color.use=color_use, title.name=NULL, show.legend=FALSE, small.gap=1, big.gap=5, lab.cex=0.5))\n",
    "\n",
    "dev.off()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "r.4.4.1-FD20200109SPLENO",
   "language": "R",
   "name": "r.4.4.1-fd20200109spleno"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
