{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3e57bbb1-f1d1-4c7a-8248-430750ec09f6",
   "metadata": {},
   "source": [
    "# Integration\n",
    "\n",
    "**[Sanorama Hie et al., 2019](https://www.nature.com/articles/s41587-019-0113-3)**  \n",
    "[GitHub](https://github.com/brianhie/scanorama)  \n",
    "[Tutorial external API](https://scanpy-tutorials.readthedocs.io/en/latest/spatial/integration-scanorama.html)  \n",
    "[External external API tutorial](https://scanpy.readthedocs.io/en/stable/generated/scanpy.external.pp.scanorama_integrate.html)\n",
    "\n",
    "A [fix](https://github.com/theislab/single-cell-tutorial/issues/60) to run [scran pooling normalization](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0947-7) computeSumFactors in current python environment. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2b0fedc-315f-4a1d-9ccc-732ab77d0481",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import scanorama\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69fe0382-4510-449c-b271-0e16bb135384",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Working directory \n",
    "os.chdir('/research/peer/fdeckert/FD20200109SPLENO')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cd3e203-b6ce-4bbd-a703-aa6bfe43a6b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# rpy2 \n",
    "os.environ['R_HOME'] = '/home/fdeckert/bin/miniconda3/envs/p.3.8.12-FD20200109SPLENO/lib/R'\n",
    "\n",
    "# Communication \n",
    "import rpy2.rinterface_lib.callbacks\n",
    "import logging\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR) # Ignore R warnings\n",
    "\n",
    "# Convertion \n",
    "from rpy2.robjects import pandas2ri\n",
    "import anndata2ri\n",
    "\n",
    "# Activate conversion globaly \n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "456232b1-f037-479b-a282-7e61c41af4c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "library(scran)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b50e162-f6b9-4d41-baff-b6b1967ebe45",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting \n",
    "import rpy2.robjects as robjects\n",
    "color_load = robjects.r.source('plotting_global.R')\n",
    "color = dict()\n",
    "for i in range(len(color_load[0])):\n",
    "    color[color_load[0].names[i]] = {key : color_load[0][i].rx2(key)[0] for key in color_load[0][i].names}\n",
    "\n",
    "sc.set_figure_params(figsize=(5, 5))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ea7d29b-b7bd-4392-94f5-648f72786222",
   "metadata": {},
   "source": [
    "# Import AnnData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f819d96-770d-4d68-af96-9a9baa7d9c75",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad('data/object/qc.h5ad')\n",
    "adata = adata[adata.obs['doublet_class']=='Singlet'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8ca8d3f-2366-419f-8731-b6b48d913d62",
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_color(categories): \n",
    "    \n",
    "    categories = [x for x in categories if x in list(adata.obs.columns)]\n",
    "\n",
    "    for category in categories: \n",
    "        \n",
    "        adata.obs[category] = pd.Series(adata.obs[category], dtype='category')\n",
    "        \n",
    "        keys = list(color[category].keys())\n",
    "        keys = [x for x in keys if x in list(adata.obs[category])]\n",
    "\n",
    "        adata.obs[category] = adata.obs[category].cat.reorder_categories(keys)\n",
    "        adata.uns[category+'_colors'] = np.array([color[category].get(key) for key in keys], dtype=object)\n",
    "        \n",
    "# Set colors\n",
    "set_color(list(color.keys()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3075fda2-2ca4-4789-92eb-d60bf7ef2b81",
   "metadata": {},
   "source": [
    "# Subset adata by sample_group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f74ba75b-cafa-41cd-a257-ed5d0f0f1088",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub = dict()\n",
    "for sample_group in adata.obs['treatment'].unique():\n",
    "    adata_sub[sample_group] = adata[adata.obs['treatment']==sample_group].copy()\n",
    "adata_sub = list(adata_sub.values())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0121859c-48e3-4cd2-9217-d40c276901d9",
   "metadata": {},
   "source": [
    "# Compute scran size factor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64a77476-617c-4d3b-833c-93095b8aebee",
   "metadata": {},
   "outputs": [],
   "source": [
    "def scran_input(adata):\n",
    "    \n",
    "    adata_pp = adata.copy()\n",
    "    \n",
    "    # For computeSumFactors we need a good coverage for each gene\n",
    "    sc.pp.filter_genes(adata_pp, min_cells=20) \n",
    "    \n",
    "    # Compute groups \n",
    "    sc.pp.normalize_per_cell(adata_pp, counts_per_cell_after=1e6)\n",
    "    sc.pp.log1p(adata_pp)\n",
    "    sc.pp.pca(adata_pp, n_comps=15, svd_solver='arpack')\n",
    "    sc.pp.neighbors(adata_pp)\n",
    "    sc.tl.louvain(adata_pp, key_added='groups', resolution=1)\n",
    "    \n",
    "    input_groups = adata_pp.obs['groups']\n",
    "    \n",
    "    # Set count matrix\n",
    "    data_mat = adata.X.T\n",
    "    \n",
    "    data = {'input_groups': input_groups, 'data_mat': data_mat}\n",
    "    \n",
    "    return(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a81b9fc0-2d2e-4592-b040-14932998b09c",
   "metadata": {},
   "outputs": [],
   "source": [
    "scran_input = [scran_input(x) for x in adata_sub]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a861bc9a-2894-486a-8166-2cc180a75267",
   "metadata": {},
   "source": [
    "### NaCl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a845d51c-b348-4c00-9bd4-c802084c5ea5",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_mat = scran_input[0]['data_mat']\n",
    "input_groups = list(scran_input[0]['input_groups'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b45e775c-8977-4a5d-ba03-d089a3faae9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i data_mat -i input_groups -o size_factors\n",
    "size_factors = sizeFactors(computeSumFactors(SingleCellExperiment(list(counts=data_mat)), clusters=input_groups, min.mean=0.1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ebc7afe-6a9b-40d0-a003-cea4a4ad6c11",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub[0].obs['size_factors'] = size_factors"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6398c29-080a-486f-a556-6728e5e2ff92",
   "metadata": {},
   "source": [
    "### CpG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e038a80b-7602-4744-b1ea-ffd351aa3a25",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_mat = scran_input[1]['data_mat']\n",
    "input_groups = list(scran_input[1]['input_groups'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2249b069-8282-4cfe-8a5c-5a59f4cb68bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i data_mat -i input_groups -o size_factors\n",
    "size_factors = sizeFactors(computeSumFactors(SingleCellExperiment(list(counts=data_mat)), clusters=input_groups, min.mean=0.1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87ed4617-bf4f-4086-a52f-1d276ba97e78",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub[1].obs['size_factors'] = size_factors"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffd3d2b0-40f8-4702-a3ab-089a6be5ad1d",
   "metadata": {},
   "source": [
    "# Normalize "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb193f61-3aa3-4ff2-a391-17a1b248832d",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub[0].X /= adata_sub[0].obs['size_factors'].values[:,None]\n",
    "sc.pp.log1p(adata_sub[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab599bd2-34d2-4efd-abf5-4102a8225af9",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub[1].X /= adata_sub[1].obs['size_factors'].values[:,None]\n",
    "sc.pp.log1p(adata_sub[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa5e6f9a-14aa-4900-a05b-37172811599f",
   "metadata": {},
   "source": [
    "# Scale data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dc87b50-0091-4a4d-9649-73a37d1ad5a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.scale(adata_sub[0])\n",
    "sc.pp.scale(adata_sub[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3215d61a-7a70-4a36-9f99-5c25c575af5e",
   "metadata": {},
   "source": [
    "# Scanorama"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59fc40b1-4f37-4d0b-b1b8-7b4cbf562e62",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run Scanorama\n",
    "scanorama.integrate_scanpy(adata_sub, dimred=200, knn=50, verbose=True)\n",
    "\n",
    "# Concatenate scanorama output \n",
    "X_scanorama = [ad.obsm['X_scanorama'] for ad in adata_sub]\n",
    "X_scanorama = np.concatenate(X_scanorama)\n",
    "\n",
    "obs_names = [ad.obs_names for ad in adata_sub]\n",
    "obs_names = np.concatenate(obs_names)\n",
    "all(obs_names==adata.obs_names)\n",
    "\n",
    "# Add X_scanorama integration to adata \n",
    "adata.obsm[\"X_scanorama\"] = X_scanorama"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd432279-7a57-48fb-9d63-87eb761b5574",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Dimensional reduction and clustering \n",
    "sc.pp.neighbors(adata, n_neighbors=100, use_rep='X_scanorama')\n",
    "sc.tl.leiden(adata, resolution=1)\n",
    "sc.tl.louvain(adata, resolution=1)\n",
    "sc.tl.umap(adata)\n",
    "\n",
    "# Plot \n",
    "sc.pl.umap(adata, color=['louvain', 'leiden', 'tissue', 'treatment', 'label_fine_haemosphere', 'sample_rep', 'cc_phase_class', 'pHb_RNA', 'pRb_RNA'], wspace=0.5, ncols=3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c3138ed-8aa3-4bb0-84ae-3223109cee43",
   "metadata": {},
   "source": [
    "# Scanorama (8000 HVG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a65ee2c-be40-4c04-bcd1-c7ee53c92a86",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sct = sc.read_h5ad('data/object/so_sct.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e55a795-2a6d-48c3-ae8f-0a1d0520141f",
   "metadata": {},
   "outputs": [],
   "source": [
    "hvg_8000 = list(adata_sct.uns['hvg_int_8000'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efbc4050-fb2f-4874-a453-be9ea3c4656f",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub[0] = adata_sub[0][:, hvg_8000]\n",
    "adata_sub[1] = adata_sub[1][:, hvg_8000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f57797ed-1b97-468f-a9df-e82d5c513439",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run Scanorama\n",
    "scanorama.integrate_scanpy(adata_sub, dimred=200, knn=50, verbose=True)\n",
    "\n",
    "# Concatenate scanorama output \n",
    "X_scanorama = [ad.obsm['X_scanorama'] for ad in adata_sub]\n",
    "X_scanorama = np.concatenate(X_scanorama)\n",
    "\n",
    "obs_names = [ad.obs_names for ad in adata_sub]\n",
    "obs_names = np.concatenate(obs_names)\n",
    "all(obs_names==adata.obs_names)\n",
    "\n",
    "# Add X_scanorama integration to adata \n",
    "adata.obsm[\"X_scanorama\"] = X_scanorama"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b159b399-3830-4add-aba6-e0714813215c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dimensional reduction and clustering \n",
    "sc.pp.neighbors(adata, n_neighbors=100, use_rep='X_scanorama')\n",
    "sc.tl.leiden(adata, resolution=1)\n",
    "sc.tl.louvain(adata, resolution=1)\n",
    "sc.tl.umap(adata)\n",
    "\n",
    "# Plot \n",
    "sc.pl.umap(adata, color=['louvain', 'leiden', 'tissue', 'treatment', 'label_fine_haemosphere', 'sample_rep', 'cc_phase_class', 'pHb_RNA', 'pRb_RNA'], wspace=0.5, ncols=3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c3199bc-1bbe-4afb-b0b0-49a4406ea1ba",
   "metadata": {},
   "source": [
    "# Scanorama (4000 HVG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d1d38a6-6a4d-40a4-9fbb-fd6f8f4a6bb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sct = sc.read_h5ad('data/object/so_sct.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e00ef882-61c8-43db-90f5-140a511f0593",
   "metadata": {},
   "outputs": [],
   "source": [
    "hvg_4000 = list(adata_sct.uns['hvg_int_4000'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfb03636-4eab-414a-9ac0-0bbf97349acc",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub[0] = adata_sub[0][:, hvg_4000]\n",
    "adata_sub[1] = adata_sub[1][:, hvg_4000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1456017d-51d5-4b3e-8a19-759ac40d4065",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run Scanorama\n",
    "scanorama.integrate_scanpy(adata_sub, dimred=200, knn=50, verbose=True)\n",
    "\n",
    "# Concatenate scanorama output \n",
    "X_scanorama = [ad.obsm['X_scanorama'] for ad in adata_sub]\n",
    "X_scanorama = np.concatenate(X_scanorama)\n",
    "\n",
    "obs_names = [ad.obs_names for ad in adata_sub]\n",
    "obs_names = np.concatenate(obs_names)\n",
    "all(obs_names==adata.obs_names)\n",
    "\n",
    "# Add X_scanorama integration to adata \n",
    "adata.obsm[\"X_scanorama\"] = X_scanorama"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e6eb2ed-07b9-40d0-8cb9-b6b6f8e55d0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dimensional reduction and clustering \n",
    "sc.pp.neighbors(adata, n_neighbors=100, use_rep='X_scanorama')\n",
    "sc.tl.leiden(adata, resolution=1)\n",
    "sc.tl.louvain(adata, resolution=1)\n",
    "sc.tl.umap(adata)\n",
    "\n",
    "# Plot \n",
    "sc.pl.umap(adata, color=['louvain', 'leiden', 'tissue', 'treatment', 'label_fine_haemosphere', 'sample_rep', 'cc_phase_class', 'pHb_RNA', 'pRb_RNA'], wspace=0.5, ncols=3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0198581-15bb-4118-b0b6-c0ab67aaf4e5",
   "metadata": {},
   "source": [
    "# Scanorama (2000 HVG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74020670-2457-4131-b7d8-17569671f2aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sct = sc.read_h5ad('data/object/so_sct.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0924a45-d7c3-4ae3-8b2f-cc946ce432e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "hvg_2000 = list(adata_sct.uns['hvg_int_2000'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c62e5936-4435-4f4f-8201-436841a2426a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub[0] = adata_sub[0][:, hvg_2000]\n",
    "adata_sub[1] = adata_sub[1][:, hvg_2000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1ebb7ac-be64-4c07-abc7-6b5617837564",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run Scanorama\n",
    "scanorama.integrate_scanpy(adata_sub, dimred=200, knn=50, verbose=True)\n",
    "\n",
    "# Concatenate scanorama output \n",
    "X_scanorama = [ad.obsm['X_scanorama'] for ad in adata_sub]\n",
    "X_scanorama = np.concatenate(X_scanorama)\n",
    "\n",
    "obs_names = [ad.obs_names for ad in adata_sub]\n",
    "obs_names = np.concatenate(obs_names)\n",
    "all(obs_names==adata.obs_names)\n",
    "\n",
    "# Add X_scanorama integration to adata \n",
    "adata.obsm[\"X_scanorama\"] = X_scanorama"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9e93e44-79aa-4f78-beea-ad69070e7291",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dimensional reduction and clustering \n",
    "sc.pp.neighbors(adata, n_neighbors=100, use_rep='X_scanorama')\n",
    "sc.tl.leiden(adata, resolution=1)\n",
    "sc.tl.louvain(adata, resolution=1)\n",
    "sc.tl.umap(adata)\n",
    "\n",
    "# Plot \n",
    "sc.pl.umap(adata, color=['louvain', 'leiden', 'tissue', 'treatment', 'label_fine_haemosphere', 'sample_rep', 'cc_phase_class', 'pHb_RNA', 'pRb_RNA'], wspace=0.5, ncols=3)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "p.3.8.12-FD20200109SPLENO",
   "language": "python",
   "name": "p.3.8.12-fd20200109spleno"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
