{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e3a12995-25a2-438f-bdbd-c92c9cd21aba",
   "metadata": {},
   "source": [
    "# Haemosphere "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ab643ff-7cad-4991-822f-2409e7e5edef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tools\n",
    "library(edgeR)\n",
    "library(limma)\n",
    "\n",
    "library(SingleCellExperiment)\n",
    "library(SingleR)\n",
    "library(Seurat)\n",
    "\n",
    "# Source\n",
    "library(biomaRt)\n",
    "\n",
    "# Data \n",
    "library(dplyr)\n",
    "library(reshape2)\n",
    "\n",
    "# Plotting\n",
    "library(ggplot2)\n",
    "library(patchwork)\n",
    "library(ComplexHeatmap)\n",
    "library(ggplotify)\n",
    "library(gridExtra)\n",
    "library(grid)\n",
    "library(viridis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1201d88-0d5f-485f-94c5-84f4283c2727",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set working directory to project root\n",
    "setwd(\"/research/peer/fdeckert/FD20200109SPLENO\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f4145b5-c7f4-4079-a7cc-c1332ee4a6ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Source files\n",
    "source(\"plotting_global.R\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39ddcf5f-0c4b-4b74-a1eb-cf2611367f5e",
   "metadata": {},
   "source": [
    "# Parameter "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e24fd2d0-569c-446a-b6b8-79e5c9b2126e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Seurat files \n",
    "cnt_file <- \"data/haemosphere/Haemopedia-Mouse-RNASeq_raw.txt\"\n",
    "meta_file <- \"data/haemosphere/Haemopedia-Mouse-RNASeq_samples.txt\"\n",
    "\n",
    "so_file <- \"data/object/seurat.rds\"\n",
    "\n",
    "filter_min_cpm = 0.5\n",
    "filter_min_expressed_samples = 2 \n",
    "normalization_method = \"TMM\"\n",
    "adj_p_cutoff = 0.05\n",
    "min_rows = 50\n",
    "\n",
    "# Plotting Theme\n",
    "ggplot2::theme_set(theme_global_set()) # From project global source()\n",
    "\n",
    "# Settings \n",
    "options(digits = 4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aefee87a-629d-40c5-aa53-a1c99fef5955",
   "metadata": {},
   "source": [
    "# Helper functions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f55c4cf0-1e28-4aaa-bbfd-42e88bfa7789",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Takes a prcomp pca object and group labels as input to compute a PCA scattern plot of the first two PCA \n",
    "pca_scattern <- function(pca, group, group_color) {\n",
    "    \n",
    "    pca_data <- cbind(pca$x, data.frame(group = group))\n",
    "    pca_scattern <- ggplot(pca_data, aes(x = PC1, y = PC2, fill = group)) + \n",
    "      geom_point(color = \"black\", pch = 21, size = 3) + \n",
    "      ggtitle(\"log2(CPM) - z-score\") + \n",
    "      xlab(paste(\"PC1 (\", round(100 * (pca$sdev^2)[1] / sum(pca$sdev^2), digits = 1), \" %)\")) + \n",
    "      ylab(paste(\"PC2 (\", round(100 * (pca$sdev^2)[2] / sum(pca$sdev^2), digits = 1), \" %)\")) +\n",
    "      scale_fill_manual(values = group_color) +\n",
    "      guides(fill = guide_legend(title = \"Group\")) + \n",
    "      theme(aspect.ratio = 1, \n",
    "            panel.border = element_blank(), \n",
    "            axis.line    = element_line())\n",
    "    \n",
    "    return(pca_scattern)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cd0c096-191e-4e89-a48d-e0c9cd176e7e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "top_hm <- function(deg, cnt = NULL, meta = NULL, level = \"\", contrast = \"\", top = 50, min_contrast_log2cpm = 2) {\n",
    "\n",
    "    # Filter and sort cnt by meta samples \n",
    "    # Columns of cnt must be corespondend to meta sample id to compute mean cpm by group\n",
    "    cnt <- cnt[, colnames(cnt) %in% meta$sample_id]  \n",
    "    cnt <- cnt[, match(meta$sample_id, colnames(cnt))]\n",
    "    \n",
    "    # Compute cpm and log2(cpm)\n",
    "    cpm <- edgeR::cpm(cnt, normalized.lib.sizes = FALSE, log = FALSE, prior.count = 0)\n",
    "    log_cpm <- edgeR::cpm(cnt, normalized.lib.sizes = FALSE, log = TRUE, prior.count = 1)\n",
    "    \n",
    "    # Define grouping levels to compute the group mean log2 cpm\n",
    "    if(level == \"main_labels\") {group <- meta$main_labels}\n",
    "    if(level == \"fine_labels\") {group <- meta$fine_labels}\n",
    "    \n",
    "    # Mean group cpm and log2(cpm)\n",
    "    colnames(cpm) <- group    \n",
    "    cpm <- lapply(unique(group), function(i) data.frame(X = rowMeans(cpm[, grepl(i, colnames(cpm), fixed = TRUE)])))\n",
    "    cpm <- do.call(\"cbind\", cpm)\n",
    "    colnames(cpm) <- unique(group)\n",
    "                  \n",
    "    colnames(log_cpm) <- group\n",
    "    log_cpm <- lapply(unique(group), function(i) data.frame(X = rowMeans(log_cpm[, grepl(i, colnames(log_cpm), fixed = TRUE)])))\n",
    "    log_cpm <- do.call(\"cbind\", log_cpm)\n",
    "    colnames(log_cpm) <- unique(group)\n",
    "                  \n",
    "    # Compute log2fc between contrast and maximum group \n",
    "    max_log2fc <- data.frame(max_log2fc = log2(cpm[, contrast] / apply(cpm[, colnames(cpm) != contrast, drop = FALSE], 1, max)))\n",
    "    rownames(max_log2fc) <- rownames(cpm)\n",
    "    \n",
    "    # filter log_cpm and max_log2fc by deg top hits\n",
    "    log_cpm <- log_cpm[rownames(log_cpm) %in% rownames(deg), ]\n",
    "    max_log2fc <- max_log2fc[rownames(max_log2fc) %in% rownames(deg), , drop = FALSE]\n",
    "                      \n",
    "    # Sort log_cpm by max_log2fc \n",
    "    log_cpm <- log_cpm[arrange(max_log2fc, desc(max_log2fc)) %>% rownames(), ]\n",
    "                      \n",
    "    # Filter for min expression in contrast group \n",
    "    log_cpm <- log_cpm[log_cpm[, contrast] > min_contrast_log2cpm, ]\n",
    "\n",
    "    # Filter log_cpm top hits \n",
    "    if (nrow(log_cpm) > top) {log_cpm <- log_cpm[1:top, ]}\n",
    "                      \n",
    "    # ComplexHeatmap color ramp\n",
    "    color_range <- max(abs(log_cpm))\n",
    "    color_break <- seq(0, color_range, 0.01)\n",
    "    color_ramp <- viridis(length(color_break), option = \"magma\")\n",
    "\n",
    "                  \n",
    "    # Heat map \n",
    "    hm <- grid.grabExpr(draw(ComplexHeatmap::pheatmap(\n",
    "        mat           = as.matrix(log_cpm),\n",
    "        main          = contrast,\n",
    "        fontsize_row  =  10,\n",
    "        scale         = \"none\",\n",
    "        cluster_rows  = TRUE,\n",
    "        cluster_cols  = FALSE,\n",
    "        cellwidth      = 12, \n",
    "        cellheight     = 12, \n",
    "        clustering_distance_rows = \"euclidean\",\n",
    "        clustering_distance_cols = \"euclidean\",\n",
    "        clustering_method        = \"complete\",\n",
    "        show_row_dend = FALSE, \n",
    "        show_rownames = TRUE,\n",
    "        show_colnames = TRUE,\n",
    "        color         = color_ramp,\n",
    "        breaks        = color_break, \n",
    "        border_color  = NA)))\n",
    "\n",
    "    \n",
    "    return(list(hm, log_cpm))\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56cb93dc-3b07-4fe8-bd3c-81d0b038fa2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Contrast matrix for one vs average of all other explanatory variables \n",
    "contrast_one_vs_average <- function(design) {\n",
    "    \n",
    "    exp_variables <- colnames(design) \n",
    "    contrast <- list()\n",
    "    for(i in seq_along(exp_variables)) {\n",
    "        design_string <- paste(exp_variables[i], \"-\", \"(\", paste(exp_variables[-i], collapse = \"+\"), \")/\", length(exp_variables[-i]), sep = \"\")\n",
    "        contrast[[i]] <- makeContrasts(contrasts = design_string, levels = exp_variables)\n",
    "    }\n",
    "    \n",
    "    return(contrast)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4abb2265-0556-4d5b-a09a-8410b7534670",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute DEG matrix needs a contrast, design and count matrix object\n",
    "deg <- function(\n",
    "    \n",
    "    cnt, design, contrast, \n",
    "    filter_min_cpm = 0.5,\n",
    "    filter_min_expressed_samples = 2, \n",
    "    normalization_method = \"TMM\",\n",
    "    adj_p_cutoff = 0.05\n",
    "    \n",
    ") {\n",
    "    \n",
    "    # Filter and sort cnt by design matrix\n",
    "    cnt <- cnt[, colnames(cnt) %in% rownames(design)]\n",
    "    cnt <- cnt[, match(rownames(design), colnames(cnt))]\n",
    "    \n",
    "    # Create a DGEList object from cnt and normalize\n",
    "    cnt <- DGEList(cnt)\n",
    "    cnt <- calcNormFactors(cnt, method = normalization_method)\n",
    "    \n",
    "    # Keep only rows where filter_min_expressed_samples or more samples out of group1 and group2 must be greater than filter_min_cpm\n",
    "    cnt <- cnt[rowSums(edgeR::cpm(cnt, normalized.lib.sizes = TRUE, log = FALSE, prior.count = 0) > filter_min_cpm) >= filter_min_expressed_samples, ] # cpm default is normalized.lib.sizes = TRUE and log = FALSE\n",
    "    \n",
    "    # Limma fit to fit explanatory variable\n",
    "    fit <- lmFit(voom(cnt, design = design, normalize.method = \"none\"), design)\n",
    "    # For a linear model fit, compute moderated t-statistics, moderated F-statistic, and log-odds of differential expression by empirical Bayes moderation of the standard errors towards a common value.\n",
    "    eb <- eBayes(contrasts.fit(fit, contrast))\n",
    "    \n",
    "    # Fetch topTable\n",
    "    m = topTable(eb, adjust = \"fdr\", number = Inf)   # This should fetch every gene/probe sorted by adj p\n",
    "    m = m[m$adj.P.Val < adj_p_cutoff, ]\n",
    "\n",
    "    # select column subset and remove dots in adj.P.Val colname\n",
    "    m = m[c(\"logFC\", \"adj.P.Val\",\"AveExpr\")] # missing logFC\n",
    "    colnames(m)[2] = \"adjPValue\"\n",
    "\n",
    "    return(m)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ed8258a-8c90-4229-b9b0-85b7870032a3",
   "metadata": {},
   "source": [
    "# Import data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2976b647-8c6e-4ae6-8025-870dbf2c6a32",
   "metadata": {},
   "outputs": [],
   "source": [
    "cnt <- read.delim(cnt_file, row.names = 1)\n",
    "meta <- read.delim(meta_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9262f96b-8b57-4f62-8b8e-1a604dd1017c",
   "metadata": {},
   "source": [
    "# Select samples by meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0602a13f-32cb-430c-b935-cf02475d1a84",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select sample by cell_name \n",
    "cell_name <- c(\n",
    "    \n",
    "    \"Fob\", \"MZB\", \"GCB\", \"SplPlsC\", #B-cells SPLEEN \n",
    "    \"BasoBM\",  #Basophil BM \n",
    "    \"cDC2\", \"pDC\", #DC conventional and plasmacytoid SPLEEN\n",
    "    \"EoP\", \"Eo\", #Eosinophil progenitor and adult BM\n",
    "    \"MEP\", \"PreCFUE\", \"CFUE\", \"EryBlPB\", \"EryBlPO\", \"Retic\", #Erythrocyte BM\n",
    "    \"MonoBM\", \"Mac\", #Monocyte BM and macrophage PERITONEAL  \n",
    "    \"Mast\", #Mast cell PERITONEAL \n",
    "    \"MegTPO\", #Megakaryocyte cultured from BM\n",
    "    \"STHSC\", \"LSK\", \"MPP\", #Multi potential progenitor BM\n",
    "    \"NeutBM\", #Neutrophil BM\n",
    "    \"NK\", #Natural killer cell SPLEEN \n",
    "    \"CMP\", #Common myeloid Progenitor BM\n",
    "    \"GMP\", #Granulocyte macrophage progenitor BM\n",
    "    \"CLP\", #Common lymphoid progenitor BM\n",
    "    \"CD4T\", \"CD8T\" #T-cells total SPLEEN and LYMPHNODES\n",
    "    \n",
    ")\n",
    "\n",
    "meta <- meta[meta$cell_name %in% cell_name, ]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5f8c5e2-93f2-4edd-965e-dbe00334f24d",
   "metadata": {},
   "source": [
    "# Select sample and order count matrix by meta "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd3eaf8c-17bc-4add-93fe-e371bf3a4681",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter and order count matrix columns by meta data samples \n",
    "cnt <- cnt[, colnames(cnt) %in% meta$sample_id]\n",
    "cnt <- cnt[, match(meta$sample_id, colnames(cnt))]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5981b222-bb65-4ea2-8a47-05e7d65eb3a4",
   "metadata": {},
   "source": [
    "# Reset meta sample id and cnt column names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f5a257d-ead4-4efd-9bf9-a9532faae8fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta$sample_id <- make.names(paste0(meta$fine_labels, \".\", seq(1:nrow(meta))))\n",
    "colnames(cnt) <- meta$sample_id"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6f6a314-a303-4328-9708-45df66e09ee1",
   "metadata": {},
   "source": [
    "# Translate count matrix ENSEMBL to Symbol "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebd4d2c3-b65b-47b3-8c00-c3bcb2fad3f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get ensembl to symbol relationship from biomaRt \n",
    "mart <- useDataset(\"mmusculus_gene_ensembl\", useMart(\"ensembl\"))\n",
    "ens_to_sym <- getBM(filters = \"ensembl_gene_id\", attributes = c(\"ensembl_gene_id\", \"mgi_symbol\"), values = rownames(cnt), mart = mart)\n",
    "\n",
    "# Remove duplicates \n",
    "ens_to_sym <- ens_to_sym[!ens_to_sym$ensembl_gene_id %>% duplicated, ]\n",
    "ens_to_sym <- ens_to_sym[!ens_to_sym$mgi_symbol %>% duplicated, ]\n",
    "\n",
    "# Filter data by ensembl id in reference \n",
    "cnt <- cnt[rownames(cnt) %in% ens_to_sym$ensembl_gene_id, ]\n",
    "\n",
    "# order ens_to_sym ensembl id by rownames of data \n",
    "ens_to_sym <- ens_to_sym[match(ens_to_sym$ensembl_gene_id, rownames(cnt)), ]\n",
    "\n",
    "# Set mgi_symbol as data rownames\n",
    "rownames(cnt) <- ens_to_sym$mgi_symbol"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68600ec2-c367-4bf7-809a-0986c3c984cd",
   "metadata": {},
   "source": [
    "# Filter count matrix by genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44e17878-f498-4ecd-9124-7f55bf0c6091",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep only genes with counts greater 0\n",
    "cnt <- cnt[rowSums(cnt) > 0, ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c912779b-3d79-4ea8-a36b-5b19e5376223",
   "metadata": {},
   "source": [
    "# Set factor level "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d19079c7-ae4e-4d69-ae14-1133912feb88",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set factor level for labels \n",
    "meta$main_labels <- factor(meta$main_labels, levels = names(color$main_labels_haemosphere))\n",
    "meta$fine_labels <- factor(meta$fine_labels, levels = names(color$fine_labels_haemosphere))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94a77c2b-194d-4ef6-a9f8-50a758969668",
   "metadata": {},
   "source": [
    "# PCA on all data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "790392f4-6d37-433b-8237-bc34c2aa33d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 10, repr.plot.height = 5)\n",
    "\n",
    "# create a DGEList object from cnt and normalize\n",
    "cnt_pca <- cnt[rowSums(edgeR::cpm(cnt, normalized.lib.sizes = FALSE, log = FALSE) > 0.5) >= 2, ]\n",
    "pca <- prcomp(t(edgeR::cpm(cnt_pca, normalized.lib.sizes = FALSE, log = TRUE)), scale = TRUE, center = TRUE)\n",
    "pca_1 <- pca_scattern(pca, meta$main_labels, color$main_labels_haemosphere)\n",
    "pca_2 <- pca_scattern(pca, meta$fine_labels, color$fine_labels_haemosphere)\n",
    "\n",
    "sc_pca <- pca_1 + pca_2\n",
    "sc_pca\n",
    "ggsave(sc_pca, filename = \"result/plot/haemosphere/sc_pca.png\", width = 10, height = 5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e79e206-4665-4fa3-b985-0e646601fb45",
   "metadata": {},
   "source": [
    "# Design and contrast matrix\n",
    "\n",
    "Explanatory variable (factor): Cell lineage (design_cl) or cell type  (design_ct)  \n",
    "Indipendent meassurements (character): Cell type with batch information   \n",
    "Statistical model: Means model to get all possible combinations of levels  \n",
    "Design matrix: Coded as model.matrix(~ 0+explanatory_variable) which is here equivalent to model.matrix(~ explanatory_variable)   \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54a26cf3-ae61-40c2-9c1a-be81ec2f0d86",
   "metadata": {},
   "source": [
    "# Compare main label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27224aab-8c07-49e0-bf0c-2f94008034b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_ml <- meta\n",
    "meta_ml$main_labels <- make.names(meta_ml$main_labels)\n",
    "\n",
    "# Make design for means model parameterization\n",
    "design_ml = model.matrix(~0+as.factor(meta_ml$main_labels))\n",
    "colnames(design_ml) = unique(meta_ml$main_labels)\n",
    "rownames(design_ml) = meta_ml$sample_id\n",
    "\n",
    "# Create contrast \n",
    "contrast_ml <- contrast_one_vs_average(design_ml)\n",
    "\n",
    "# Compute deg for all contrast combination \n",
    "deg_ml <- list()\n",
    "for(i in seq_along(contrast_ml)) {deg_ml[[i]] <- deg(cnt = cnt, design = design_ml, contrast = contrast_ml[[i]])}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48451721-1a09-4c1f-a3a0-e910036fcbdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 30, repr.plot.height = 10)\n",
    "\n",
    "# Heat map of marker genes\n",
    "hm_ml <- lapply(seq_along(unique(meta_ml$main_labels)), function(i) top_hm(deg = deg_ml[[i]], cnt = cnt, meta = meta_ml, level = \"main_labels\", contrast = unique(meta_ml$main_labels)[i], top = 25, min_contrast_log2cpm = 2))\n",
    "                \n",
    "# Plot\n",
    "hm_ml_grid <- gridExtra::arrangeGrob(grobs = sapply(hm_ml, \"[\", 1), ncol = 7) \n",
    "ggsave(hm_ml_grid, filename = \"result/plot/haemosphere/hm_ml_grid.png\", width = 30, height = 10)\n",
    "hm_ml_grid %>% grid::grid.draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d07e1583-8e25-4ab9-bdf7-c5975ddb97c5",
   "metadata": {},
   "source": [
    "# Compare fine labels "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7308d63f-3cd1-4065-a8b3-c99cb6f2d46e",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_fl <- meta\n",
    "meta_fl$fine_labels <- make.names(meta_fl$fine_labels)\n",
    "\n",
    "# Make design for means model parameterization\n",
    "design_fl <- model.matrix(~0+as.factor(meta_fl$cell_name))\n",
    "colnames(design_fl) <- unique(meta_fl$cell_name)\n",
    "rownames(design_fl) <- meta_fl$sample_id\n",
    "\n",
    "# Create contrast \n",
    "contrast_fl <- contrast_one_vs_average(design_fl)\n",
    "\n",
    "# Compute deg for all contrast combination \n",
    "deg_fl <- list()\n",
    "for(i in seq_along(contrast_fl)) {deg_fl[[i]] <- deg(cnt = cnt, design = design_fl, contrast = contrast_fl[[i]])}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "555447f2-f7e9-433a-a040-abe9d24bcad7",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 30, repr.plot.height = 25)\n",
    "\n",
    "# Heat map of marker genes \n",
    "hm_fl <- lapply(seq_along(unique(meta_fl$fine_labels)), function(i) top_hm(deg = deg_fl[[i]], cnt = cnt, meta = meta_fl, level = \"fine_labels\", contrast = unique(meta_fl$fine_labels)[i], top = 25, min_contrast_log2cpm = 1))\n",
    "                \n",
    "# Plot\n",
    "hm_fl_grid <- gridExtra::arrangeGrob(grobs = sapply(hm_fl, \"[\", 1), ncol = 5) \n",
    "ggsave(hm_fl_grid, filename = \"result/plot/haemosphere/hm_fl_grid.png\", width = 30, height = 25)\n",
    "hm_fl_grid %>% grid::grid.draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42fd9a88-e00e-4ec2-ae92-6f211e49e652",
   "metadata": {},
   "source": [
    "# Compare erythoblast stages "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ac5e4a2-bd3d-46e7-99ca-d9d0232e36b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_ery <- meta[meta$main_labels == \"Ery\", ]\n",
    "meta_ery$fine_labels <- make.names(meta_ery$fine_labels)\n",
    "\n",
    "# Make design for means model parameterization\n",
    "design_ery <- model.matrix(~0+as.factor(meta_ery$cell_name))\n",
    "colnames(design_ery) <- unique(meta_ery$cell_name)\n",
    "rownames(design_ery) <- meta_ery$sample_id\n",
    "\n",
    "# Create contrast \n",
    "contrast_ery <- contrast_one_vs_average(design_ery)\n",
    "\n",
    "# Compute deg for all contrast combination \n",
    "deg_ery <- list()\n",
    "for(i in seq_along(contrast_ery)) {deg_ery[[i]] <- deg(cnt = cnt, design = design_ery, contrast = contrast_ery[[i]])}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93aa8e3f-39d0-4f94-b6cc-99ce8b8ee470",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 15, repr.plot.height = 6)\n",
    "\n",
    "# Heat map of marker genes \n",
    "hm_ery <- lapply(seq_along(unique(meta_ery$fine_labels)), function(i) top_hm(deg = deg_ery[[i]], cnt = cnt, meta = meta_ery, level = \"fine_labels\", contrast = unique(meta_ery$fine_labels)[i], top = 25, min_contrast_log2cpm = 1))\n",
    "                 \n",
    "# Plot\n",
    "hm_ery_grid <- gridExtra::arrangeGrob(grobs = sapply(hm_ery, \"[\", 1),, ncol = 6)      \n",
    "ggsave(hm_ery_grid, filename = \"result/plot/haemosphere/hm_ery_grid.png\", width = 15, height = 6)\n",
    "hm_ery_grid %>% grid::grid.draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2917c2ac-3c03-44b8-a814-ac1de39a4fec",
   "metadata": {},
   "source": [
    "# Compare multipotent progenitor cells "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f48530c4-d6f6-497a-8fee-a46384d64506",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_mpp <- meta[meta$main_labels == \"MPP\", ]\n",
    "meta_mpp$fine_labels <- make.names(meta_mpp$fine_labels)\n",
    "\n",
    "# Make design for means model parameterization\n",
    "design_mpp <- model.matrix(~0+as.factor(meta_mpp$cell_name))\n",
    "colnames(design_mpp) <- unique(meta_mpp$cell_name)\n",
    "rownames(design_mpp) <- meta_mpp$sample_id\n",
    "\n",
    "# Create contrast \n",
    "contrast_mpp <- contrast_one_vs_average(design_mpp)\n",
    "\n",
    "# Compute deg for all contrast combination \n",
    "deg_mpp <- list()\n",
    "for(i in seq_along(contrast_mpp)) {deg_mpp[[i]] <- deg(cnt = cnt, design = design_mpp, contrast = contrast_mpp[[i]])}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d5bfac5-9687-483f-854b-cd6eb7a94d37",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 10, repr.plot.height = 6)\n",
    "\n",
    "# Heat map of marker genes \n",
    "hm_mpp <- lapply(seq_along(unique(meta_mpp$fine_labels)), function(i) top_hm(deg = deg_mpp[[i]], cnt = cnt, meta = meta_mpp, level = \"fine_labels\", contrast = unique(meta_mpp$fine_labels)[i], top = 25, min_contrast_log2cpm = 1))\n",
    "\n",
    "# Plot\n",
    "hm_mpp_grid <- gridExtra::arrangeGrob(grobs = sapply(hm_mpp, \"[\", 1), ncol = 3)\n",
    "ggsave(hm_mpp_grid, filename = \"result/plot/haemosphere/hm_mpp_grid.png\", width = 10, height = 6)\n",
    "hm_mpp_grid %>% grid::grid.draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "feb00a72-97f1-4460-bc4c-1fec5ff36525",
   "metadata": {},
   "source": [
    "# Compare restricted progenitor cells "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "460fa29f-3d54-4fa7-b161-9d4a87e003f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_rpp <- meta[meta$main_labels == \"RPP\", ]\n",
    "meta_rpp$fine_labels <- make.names(meta_rpp$fine_labels)\n",
    "\n",
    "# Make design for means model parameterization\n",
    "design_rpp <- model.matrix(~0+as.factor(meta_rpp$cell_name))\n",
    "colnames(design_rpp) <- unique(meta_rpp$cell_name)\n",
    "rownames(design_rpp) <- meta_rpp$sample_id\n",
    "\n",
    "# Create contrast \n",
    "contrast_rpp <- contrast_one_vs_average(design_rpp)\n",
    "\n",
    "# Compute deg for all contrast combination \n",
    "deg_rpp <- list()\n",
    "for(i in seq_along(contrast_rpp)) {deg_rpp[[i]] <- deg(cnt = cnt, design = design_rpp, contrast = contrast_rpp[[i]])}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eec58673-2bd5-48d1-a6f6-2600d150cd28",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 5, repr.plot.height = 6)\n",
    "\n",
    "# Heat map of marker genes \n",
    "hm_rpp <- lapply(seq_along(unique(meta_rpp$main_labels)), function(i) top_hm(deg = deg_rpp[[i]], cnt = cnt, meta = meta_rpp, level = \"fine_labels\", contrast = unique(meta_rpp$main_labels)[i], top = 25, min_contrast_log2cpm = 1))\n",
    "                 \n",
    "# Plot\n",
    "hm_rpp_grid <- gridExtra::arrangeGrob(grobs = sapply(hm_rpp, \"[\", 1), ncol = 1)\n",
    "ggsave(hm_rpp_grid, filename = \"result/plot/haemosphere/hm_rpp_grid.png\", width = 5, height = 6)\n",
    "hm_rpp_grid %>% grid::grid.draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffdf1447-7520-481d-9360-52be9e4c60a6",
   "metadata": {},
   "source": [
    "# Create SingleR object "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d3cfa41-aade-438a-8fd5-b35c940bc97b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create SingleR object from Haemosphere RNA-seq data\n",
    "ref <- SummarizedExperiment(list(counts = cnt, logcounts = edgeR::cpm(cnt, log = TRUE, prior.count = 1)))\n",
    "ref$label.main <- as.character(meta$main_labels)\n",
    "ref$label.fine <- as.character(meta$fine_labels)\n",
    "\n",
    "saveRDS(ref, \"data/haemosphere/se_haemosphere.rds\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "r.4.1.0-FD20200109SPLENO",
   "language": "R",
   "name": "r.4.1.0-fd20200109spleno"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
