{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "50e4fbf7-815e-4b90-9012-a0afabdd59a5",
   "metadata": {},
   "source": [
    "# Figure panel 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cfaec12-a6e4-450e-a15e-e4a93255590b",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "800de35d-86e3-4b71-8dd3-911e631758f2",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "library_load <- suppressMessages(\n",
    "    \n",
    "    suppressWarnings(\n",
    "        \n",
    "        list(\n",
    "        \n",
    "            # Seurat \n",
    "            library(Seurat), \n",
    "\n",
    "            # Data \n",
    "            library(tidyverse), \n",
    "            library(data.table), \n",
    "\n",
    "\n",
    "            # Plotting \n",
    "            library(ggplot2), \n",
    "            library(patchwork), \n",
    "            library(ComplexHeatmap), \n",
    "            library(circlize), \n",
    "            library(viridis), \n",
    "            library(ggplotify), \n",
    "            library(ggrepel), \n",
    "            library(cowplot), \n",
    "\n",
    "            # Pyhton compatibility\n",
    "            library(reticulate)\n",
    "\n",
    "        )\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15b0ec44-5e03-4e48-82cd-1042b31d1a5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Configure reticulate \n",
    "# use_condaenv(condaenv=\"p.3.10.16-FD20200109SPLENO\", conda=\"/nobackup/peer/fdeckert/miniconda3/bin/conda\", required=NULL)\n",
    "# py_config()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45028df8-0d7a-4a09-865f-5d769c33cf3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_seed <- 42\n",
    "set.seed(random_seed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ae344ae-fb5c-42da-af0e-e5559e7b3e09",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set working directory to project root\n",
    "setwd(\"/research/peer/fdeckert/FD20200109SPLENO\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b52cea3-0401-43b9-a99a-a97fc2a66c4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Source\n",
    "source(\"bin/so_pl.R\")\n",
    "source(\"bin/tradeseq_pp.R\")\n",
    "source(\"bin/tradeseq_pl.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ab8c554-ca21-4e65-8932-2a87d05f1f30",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting Theme\n",
    "source(\"plotting_global.R\")\n",
    "ggplot2::theme_set(theme_global_set(size_select=4)) # From project global source()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "503a7fe3-9117-4ee3-9855-d8d6b205fe2a",
   "metadata": {},
   "source": [
    "# Load reference data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fdfb917-1882-4a96-8acd-6e467219c80e",
   "metadata": {},
   "outputs": [],
   "source": [
    "tf <- read.table(\"/research/peer/fdeckert/reference/animaltfdb/Mus_musculus_TF.txt\")[[2]]\n",
    "\n",
    "lr <- CellChat::CellChatDB.mouse[[1]] %>% dplyr::select(pathway_name, ligand, receptor.symbol, receptor.family, annotation) %>% dplyr::filter(annotation %in% c(\"Cell-Cell Contact\", \"Secreted Signaling\", \"ECM-Receptor\")) %>% \n",
    "    separate_rows(receptor.symbol, sep=\", \") %>% dplyr::distinct() %>%\n",
    "    separate_rows(ligand, sep=\", \") %>% dplyr::distinct() %>% \n",
    "    dplyr::rename(pathway=pathway_name, receptor=receptor.symbol, family=receptor.family)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40b50671-532a-4675-b47a-2818c4b5ef07",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a3be22c-7b99-4705-9fb3-1fc18dc1a4ef",
   "metadata": {},
   "source": [
    "## scRNAseq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cf85c48-d533-411e-a88f-9da7e7461076",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- readRDS(\"data/scRNAseq/object/pp_1.rds\")\n",
    "so <- NormalizeData(so)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1af207b4-6931-4e44-a4b2-e7fd648a30aa",
   "metadata": {},
   "source": [
    "## Erythroid lineage fitgam and make contrasts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97214ee5-e6c1-4151-9aaa-38d66d80ecd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "tradeseq_res <- readRDS(\"result/lineage/tradeseq_res_2.rds\")\n",
    "fitgam <- tradeseq_res[[\"fitgam\"]]\n",
    "ptpg <- tradeseq_res[[\"ptpg\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9caac726-52a9-4969-a302-920dcb6a52a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrast_1 <- c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\")\n",
    "ptpg_res_1 <- ptpg[ptpg[[\"contrast\"]]==paste0(contrast_1, collapse=\":\"), ]\n",
    "\n",
    "contrast_2 <- c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D3\")\n",
    "ptpg_res_2 <- ptpg[ptpg[[\"contrast\"]]==paste0(contrast_2, collapse=\":\"), ]\n",
    "\n",
    "contrast_3 <- c(\"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")\n",
    "ptpg_res_3 <- ptpg[ptpg[[\"contrast\"]]==paste0(contrast_3, collapse=\":\"), ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c703963f-7d61-4a2f-a407-34375a3ae53f",
   "metadata": {},
   "source": [
    "# UMAP erytime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dcef9a8-5b15-462f-8df9-e9f89735d2bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "dpt <- read.csv(\"result/pseudotime/ery_dpt.csv\", row.names=1)\n",
    "so <- AddMetaData(so, dpt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e832867-5f98-4fd7-91af-bb84045d76ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "fp_1 <- fplot(so, features=\"dpt_pseudotime\", pt_size=0.25, option=\"F\", size_select=4)\n",
    "fp_1[[1]]$data <- fp_1[[1]]$data[!is.na(fp_1[[1]]$data$feature), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1264d70-9cf4-41e0-b677-99f02035987e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_4/umap_erytime.pdf\", width=3.5, height=2.5)\n",
    "\n",
    "gridExtra::grid.arrange(\n",
    "    \n",
    "    fp_1 %>% egg::set_panel_size(., width=unit(4.5, \"cm\"), height=unit(4.5, \"cm\"))\n",
    "\n",
    ")\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc487a92-7bb2-4be5-8486-1e298bc754f6",
   "metadata": {},
   "source": [
    "# Erythroid progenitor marker \n",
    "\n",
    "All BFU-E potential in the cKit+ CD71low/Ter119low population resided in the CD150+ (Slamf1) fraction. Exclusion of mature erythroid precursor cells in spleen was facilitated by inclusion of CD24a as a negative selection marker. Of the surface markers tested (CD9, CD11a, CD34, CD48, CD63, and CD79b), only CD9 could further fractionate the BFU-E-containing CD150+ population in stressed spleen. To further discriminate putative multi-potent stress-progenitors from lineage restricted stress-BFU-E we also included Sca1, which in steady-state BM separates Sca1+ (Ly6a) hematopoietic stem cells (HSC) and Sca1â€“ myelo-erythroid progenitors."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "486204a0-62af-4d04-a55f-7cbe009e5cb3",
   "metadata": {},
   "source": [
    "## Module score "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5f24eb3-967a-453f-8a76-8887fe800821",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- readRDS(\"data/scRNAseq/object/pp.rds\")\n",
    "so <- NormalizeData(so)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79cd2846-db08-4ded-9259-3f168a38fc17",
   "metadata": {},
   "outputs": [],
   "source": [
    "fate_simplex_fwd <- read.csv(\"result/cellrank/fate_simplex_fwd_p.csv\", row.names=1) \n",
    "fate_simplex_fwd <- fate_simplex_fwd[rownames(fate_simplex_fwd) %in% colnames(so), ]\n",
    "colnames(fate_simplex_fwd) <- c(\"CR_1\", \"CR_2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e02243e6-66fb-4eca-8146-99f76d896f25",
   "metadata": {},
   "outputs": [],
   "source": [
    "so[[\"CR\"]] <- SeuratObject::CreateDimReducObject(embeddings=as.matrix(fate_simplex_fwd), key=\"CR_\", assay=\"RNA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc300815-8541-4cb8-bd88-61d4376839e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- so[, so$celltype_low %in% c(\"GMP\", \"NeuP\", \"BasoP\", \"MastP\", \"MegP\", \"MEP\", \"Proerythroblast\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61aa40f7-a08c-4acc-a8dc-52278ea788ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "features <- list(\n",
    "\n",
    "    msP_1=c(\"Hlf+\", \"Esam+\"), \n",
    "    msP_2=c(\"Kit+\", \"Flt3+\"), \n",
    "    msP_3=c(\"Cd24a+\", \"Tfrc+\"), \n",
    "    msP_4=c(\"Slamf1+\"),\n",
    "    msP_5=c(\"Fcgr2b+\", \"Fcgr3+\"), \n",
    "    msP_6=c(\"Ms4a2+\"),  \n",
    "    msP_7=c(\"Elane+\", \"Prtn3+\"), \n",
    "    msP_8=c(\"Gzmb+\"), \n",
    "    msP_9=c(\"Prss34+\"), \n",
    "    msP_10=c(\"Ms4a2+\", \"Gzmb+\", \"Prss34+\", \"Fcgr2b+\", \"Fcgr3+\", \"Cd34+\"), \n",
    "    msP_11=c(\"Cd9+\"), \n",
    "    msP_12=c(\"Ly6a+\"), \n",
    "    msP_13=c(\"Gata1+\", \"Klf1+\", \"Gata2-\")\n",
    "\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a28ba017-a2f4-497b-b7c2-48151420408c",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- UCell::AddModuleScore_UCell(so, features=features, assay=\"RNA\", slot=\"counts\", name=\"\", maxRank=2500)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f448864-541a-4a48-990d-0aedc602f7ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "pt_size <- 0.25\n",
    "order <- TRUE\n",
    "\n",
    "max_set <- 1\n",
    "min_cutoff <- 0\n",
    "max_cutoff <- 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a9a93ae-6ba9-471c-8fb3-631fabfc3413",
   "metadata": {},
   "outputs": [],
   "source": [
    "p_l <- list(\n",
    "\n",
    "    dplot(so, group_by=\"celltype_low\", group_color=color$celltype_low, reduction=\"CR\", pt_size=pt_size, legend_position=\"none\", size_select=4), \n",
    "    fplot(so, features=\"msP_1\", assay=\"RNA\", slot=\"data\", reduction=\"CR\", pt_size=pt_size, order=order, max_set=max_set, min_cutoff=min_cutoff, max_cutoff=max_cutoff, size_select=4), \n",
    "    fplot(so, features=\"msP_2\", assay=\"RNA\", slot=\"data\", reduction=\"CR\", pt_size=pt_size, order=order, max_set=max_set, min_cutoff=min_cutoff, max_cutoff=max_cutoff, size_select=4), \n",
    "    fplot(so, features=\"msP_3\", assay=\"RNA\", slot=\"data\", reduction=\"CR\", pt_size=pt_size, order=order, max_set=max_set, min_cutoff=min_cutoff, max_cutoff=max_cutoff, size_select=4), \n",
    "    fplot(so, features=\"msP_4\", assay=\"RNA\", slot=\"data\", reduction=\"CR\", pt_size=pt_size, order=order, max_set=max_set, min_cutoff=min_cutoff, max_cutoff=max_cutoff, size_select=4), \n",
    "    fplot(so, features=\"msP_5\", assay=\"RNA\", slot=\"data\", reduction=\"CR\", pt_size=pt_size, order=order, max_set=max_set, min_cutoff=min_cutoff, max_cutoff=max_cutoff, size_select=4), \n",
    "    fplot(so, features=\"msP_6\", assay=\"RNA\", slot=\"data\", reduction=\"CR\", pt_size=pt_size, order=order, max_set=max_set, min_cutoff=min_cutoff, max_cutoff=max_cutoff, size_select=4), \n",
    "    fplot(so, features=\"msP_7\", assay=\"RNA\", slot=\"data\", reduction=\"CR\", pt_size=pt_size, order=order, max_set=max_set, min_cutoff=min_cutoff, max_cutoff=max_cutoff, size_select=4), \n",
    "    fplot(so, features=\"msP_8\", assay=\"RNA\", slot=\"data\", reduction=\"CR\", pt_size=pt_size, order=order, max_set=max_set, min_cutoff=min_cutoff, max_cutoff=max_cutoff, size_select=4), \n",
    "    fplot(so, features=\"msP_9\", assay=\"RNA\", slot=\"data\", reduction=\"CR\", pt_size=pt_size, order=order, max_set=max_set, min_cutoff=min_cutoff, max_cutoff=max_cutoff, size_select=4), \n",
    "    fplot(so, features=\"msP_10\", assay=\"RNA\", slot=\"data\", reduction=\"CR\", pt_size=pt_size, order=order, max_set=max_set, min_cutoff=min_cutoff, max_cutoff=max_cutoff, size_select=4), \n",
    "    fplot(so, features=\"msP_11\", assay=\"RNA\", slot=\"data\", reduction=\"CR\", pt_size=pt_size, order=order, max_set=max_set, min_cutoff=min_cutoff, max_cutoff=max_cutoff, size_select=4), \n",
    "    fplot(so, features=\"msP_12\", assay=\"RNA\", slot=\"data\", reduction=\"CR\", pt_size=pt_size, order=order, max_set=max_set, min_cutoff=min_cutoff, max_cutoff=max_cutoff, size_select=4)\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f7d4f6d-f5fd-4352-af36-a986813872e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "p_l <- lapply(p_l, function(p) egg::set_panel_size(p, width=unit(3, \"cm\"), height=unit(3, \"cm\")))\n",
    "p_l <- do.call(gridExtra::arrangeGrob, c(p_l, ncol=10, nrow=ceiling(length(p_l)/10)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ad941f5-483f-42f5-8e95-6b953b1b2d71",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_4/cr_p_marker_genes.pdf\", width=10*2, height=2*ceiling(length(p_l)/10))\n",
    "\n",
    "grid::grid.draw(p_l)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c50631bb-4c53-4bb1-8f68-94b4593d0e19",
   "metadata": {},
   "source": [
    "# Progenitor distribution "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "436cacea-1dbc-465a-b3cc-a843e947b98e",
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- so@meta.data %>% dplyr::filter(celltype_low!=\"Proerythroblast\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68f316e2-8476-47fc-9458-9aa3fdcf265d",
   "metadata": {},
   "outputs": [],
   "source": [
    "features <- list(\n",
    "\n",
    "    msP_1=c(\"Hlf+\", \"Esam+\"), \n",
    "    msP_2=c(\"Kit+\", \"Flt3+\"), \n",
    "    msP_3=c(\"Cd24a+\", \"Tfrc+\"), \n",
    "    msP_4=c(\"Slamf1+\"),\n",
    "    msP_5=c(\"Fcgr2b+\", \"Fcgr3+\"), \n",
    "    msP_6=c(\"Ms4a2+\"),  \n",
    "    msP_7=c(\"Elane+\", \"Prtn3+\"), \n",
    "    msP_8=c(\"Gzmb+\"), \n",
    "    msP_9=c(\"Prss34+\"), \n",
    "    msP_10=c(\"Ms4a2+\", \"Gzmb+\", \"Prss34+\", \"Fcgr2b+\", \"Fcgr3+\", \"Cd34+\"), \n",
    "    msP_11=c(\"Cd9+\"), \n",
    "    msP_12=c(\"Ly6a+\"), \n",
    "    msP_13=c(\"Gata1+\", \"Klf1+\", \"Gata2-\")\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "120d310a-13ee-46d6-a949-4c6433ccd9f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GMP Hlf+ Esam+\n",
    "ratio <- data %>% dplyr::filter(celltype_low==\"GMP\") %>% dplyr::group_by(sample_group) %>% dplyr::summarise(ratio=sum(msP_1>0)/n()) %>% dplyr::pull(ratio)\n",
    "round(mean(100*ratio), digits=1)\n",
    "round(sd(100*ratio), digits=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0476ac8-a150-474a-9367-f0588a3082da",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MEP Hlf+ Esam+ \n",
    "ratio <- data %>% dplyr::filter(celltype_low==\"MEP\") %>% dplyr::group_by(sample_group) %>% dplyr::summarise(ratio=sum(msP_1>0)/n()) %>% dplyr::pull(ratio)\n",
    "round(mean(100*ratio), digits=1)\n",
    "round(sd(100*ratio), digits=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "771bce62-01f5-4e57-9990-04e0fa4dbf74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Kit+ Flt3+ \n",
    "ratio <- data %>% dplyr::filter(celltype_low!=\"Proerythroblast\") %>% dplyr::group_by(sample_group) %>% dplyr::summarise(ratio=sum(msP_2>0)/n()) %>% dplyr::pull(ratio)\n",
    "round(mean(100*ratio), digits=1)\n",
    "round(sd(100*ratio), digits=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2720996-a1d0-4d38-a1af-01ce0d947ccf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MEP Cd24+ Tfrc+ \n",
    "ratio <- data %>% dplyr::filter(celltype_low==\"MEP\") %>% dplyr::group_by(sample_group) %>% dplyr::summarise(ratio=sum(msP_3>0)/n()) %>% dplyr::pull(ratio)\n",
    "round(mean(100*ratio), digits=1)\n",
    "round(sd(100*ratio), digits=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e913061a-edd9-4278-b8f1-b8e68486d8d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Proerythroblast Cd24+ Tfrc+ \n",
    "ratio <- data %>% dplyr::filter(celltype_low==\"Proerythroblast\") %>% dplyr::group_by(sample_group) %>% dplyr::summarise(ratio=sum(msP_3>0)/n()) %>% dplyr::pull(ratio)\n",
    "round(mean(100*ratio), digits=1)\n",
    "round(sd(100*ratio), digits=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0668630f-e156-409c-b537-1c8a62f94634",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Granulocyte Cd24+ Tfrc+ \n",
    "ratio <- data %>% dplyr::filter(!(celltype_low %in% c(\"MEP\", \"Proerythroblast\"))) %>% dplyr::group_by(sample_group) %>% dplyr::summarise(ratio=sum(msP_3>0)/n()) %>% dplyr::pull(ratio)\n",
    "round(mean(100*ratio), digits=1)\n",
    "round(sd(100*ratio), digits=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2493b95c-c804-4a4e-b437-234bcca693fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MEP Cd24+ Tfrc+ \n",
    "ratio <- data %>% dplyr::filter(celltype_low==\"MEP\") %>% dplyr::group_by(sample_group) %>% dplyr::summarise(ratio=sum(msP_3>0)/n()) %>% dplyr::pull(ratio)\n",
    "round(mean(100*ratio), digits=1)\n",
    "round(sd(100*ratio), digits=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb7048a3-204a-444a-9643-2b3e3cd6d758",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MEP Cd24+ Tfrc+ with Slamf1\n",
    "ratio <- data %>% dplyr::filter(msP_3>0 & msP_4>0) %>% dplyr::group_by(sample_group) %>% dplyr::summarise(ratio=sum(celltype_low %in% c(\"MegP\", \"Proerythroblast\", \"MEP\"))/n()) %>% dplyr::pull(ratio)\n",
    "round(mean(100*ratio), digits=1)\n",
    "round(sd(100*ratio), digits=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10970992-ba49-4564-980b-76274b24a6f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MEP with BMP \n",
    "ratio <- data %>% dplyr::filter(celltype_low==\"MEP\") %>% dplyr::group_by(sample_group) %>% dplyr::summarise(ratio=sum(msP_10>0)/n()) %>% dplyr::pull(ratio)\n",
    "round(mean(100*ratio), digits=1)\n",
    "round(sd(100*ratio), digits=1)\n",
    "ratio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c472b3a6-ab9f-419a-b045-e92647082585",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Slamf1 with granulocyte potential \n",
    "ratio <- data %>% dplyr::filter(msP_4>0 & celltype_low==\"MEP\") %>% dplyr::group_by(sample_group) %>% dplyr::summarise(ratio=sum(msP_10>0)/n()) %>% dplyr::pull(ratio)\n",
    "round(mean(100*ratio), digits=1)\n",
    "round(sd(100*ratio), digits=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2ebf0fa-e61d-427a-9606-22b086881e7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Slamf1 without granulocyte potential \n",
    "ratio <- data %>% dplyr::filter(msP_10==0 & celltype_low==\"MEP\") %>% dplyr::group_by(sample_group) %>% dplyr::summarise(ratio=sum(msP_4>0)/n()) %>% dplyr::pull(ratio)\n",
    "round(mean(100*ratio), digits=1)\n",
    "round(sd(100*ratio), digits=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de41554c-0d0d-4149-b545-7dd978c5b2f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MEP without granulocyte potential at Ctl and D6 with Ly6a expression \n",
    "ratio <- data %>% dplyr::filter(msP_10==0 & celltype_low==\"MEP\" & sample_group %in% c(\"Bl6_NaCl_D6\", \"Bl6_CpG_D6\", \"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_LysM_cre_Baseline_D0\")) %>% dplyr::group_by(sample_group) %>% dplyr::summarise(ratio=sum(msP_12>0)/n()) %>% dplyr::pull(ratio)\n",
    "round(mean(100*ratio), digits=1)\n",
    "round(sd(100*ratio), digits=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "655c1dbf-bd88-4ce8-b80d-1167373dcb50",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MEP without granulocyte potential at D1 and D3 with Ly6a expression \n",
    "ratio <- data %>% dplyr::filter(msP_10==0 & celltype_low==\"MEP\" & !(sample_group %in% c(\"Bl6_NaCl_D6\", \"Bl6_CpG_D6\", \"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_LysM_cre_Baseline_D0\"))) %>% dplyr::group_by(sample_group) %>% dplyr::summarise(ratio=sum(msP_12>0)/n()) %>% dplyr::pull(ratio)\n",
    "round(mean(100*ratio), digits=1)\n",
    "round(sd(100*ratio), digits=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "761d6a31-02da-4cce-ad24-070ec949b3f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MEP without granulocyte potential at D1 and D3 with Ly6a expression \n",
    "ratio <- data %>% dplyr::filter(msP_10==0 & celltype_low==\"MEP\" & sample_group %in% c(\"IFNAR_fl_CpG_D3\")) %>% dplyr::group_by(sample_group) %>% dplyr::summarise(ratio=sum(msP_12>0)/n()) %>% dplyr::pull(ratio)\n",
    "round(mean(100*ratio), digits=1)\n",
    "round(sd(100*ratio), digits=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2f3d655-7609-4756-9334-cfb3505c9dc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MEP without granulocyte potential at D1 and D3 with Ly6a expression \n",
    "ratio <- data %>% dplyr::filter(msP_10==0 & celltype_low==\"MEP\" & sample_group %in% c(\"IFNAR_fl_LysM_cre_CpG_D3\")) %>% dplyr::group_by(sample_group) %>% dplyr::summarise(ratio=sum(msP_12>0)/n()) %>% dplyr::pull(ratio)\n",
    "round(mean(100*ratio), digits=1)\n",
    "round(sd(100*ratio), digits=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dca92d2c-56ad-44ca-9a70-b9de953f9cbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- so@meta.data %>% dplyr::filter(celltype_low %in% c(\"MEP\")) %>% \n",
    "    dplyr::mutate(group=ifelse(msP_10==0 & msP_4>0, \"MEP Slamf1+\", \"Negative\")) %>% \n",
    "    dplyr::mutate(group=ifelse(msP_10==0 & msP_4==0, \"MEP Slamf1-\", group)) %>% \n",
    "    dplyr::mutate(group=ifelse(msP_10>0 & msP_4>0, \"BMP Slamf1+\", group)) %>% \n",
    "    dplyr::mutate(group=ifelse(msP_10>0 & msP_4==0, \"BMP Slamf1-\", group)) %>% \n",
    "    dplyr::mutate(group=factor(group, levels=c(\"MEP Slamf1+\", \"MEP Slamf1-\", \"BMP Slamf1+\", \"BMP Slamf1-\"))) %>% \n",
    "    dplyr::mutate(dpi=ifelse(sample_group==\"Bl6_NaCl_D6\", \"D0\", as.character(dpi)))\n",
    "\n",
    "bp_1 <- ggplot(data, aes(x=genotype, fill=group)) + \n",
    "    geom_bar(position=\"fill\") + \n",
    "    facet_grid(~dpi, space=\"free\", scales=\"free\") + \n",
    "    scale_fill_manual(values=c(\"MEP Slamf1+\"=\"#ffcdb2\", \"MEP Slamf1-\"=\"#ffb4a2\", \"BMP Slamf1+\"=\"#83c5be\", \"BMP Slamf1-\"=\"#006d77\")) + \n",
    "    theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) + \n",
    "    guides(fill=guide_legend(ncol=1, override.aes=list(alpha=1, size=1.5), keywidth=0.25, keyheight=0.25, default.unit=\"cm\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6be95fe5-ce57-451f-9d98-d4e499f8d601",
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- so@meta.data %>% dplyr::filter(celltype_low %in% c(\"MEP\")) %>% dplyr::filter(msP_10==0) %>%\n",
    "    dplyr::mutate(group=ifelse(msP_10==0 & msP_11>0 & msP_12>0, \"MEP Cd9+ Ly6a+\", \"Negative\")) %>% \n",
    "    dplyr::mutate(group=ifelse(msP_10==0 & msP_11==0 & msP_12>0, \"MEP Cd9- Ly6a+\", group)) %>% \n",
    "    dplyr::mutate(group=ifelse(msP_10==0 & msP_11>0 & msP_12==0, \"MEP Cd9+ Ly6a-\", group)) %>% \n",
    "    dplyr::mutate(group=ifelse(msP_10==0 & msP_11==0 & msP_12==0, \"MEP Cd9- Ly6a-\", group)) %>% \n",
    "    dplyr::mutate(group=ifelse(msP_10>0, \"BMP\", group)) %>% \n",
    "    dplyr::mutate(group=factor(group, levels=c(\"MEP Cd9+ Ly6a+\", \"MEP Cd9- Ly6a+\", \"MEP Cd9+ Ly6a-\", \"MEP Cd9- Ly6a-\", \"BMP\"))) %>% \n",
    "    dplyr::mutate(dpi=ifelse(sample_group==\"Bl6_NaCl_D6\", \"D0\", as.character(dpi)))\n",
    "\n",
    "bp_2 <- ggplot(data, aes(x=genotype, fill=group)) + \n",
    "    geom_bar(position=\"fill\") + \n",
    "    facet_grid(~dpi, space=\"free\", scales=\"free\") + \n",
    "    scale_fill_manual(values=c(\"MEP Cd9+ Ly6a+\"=\"#ffcdb2\", \"MEP Cd9- Ly6a+\"=\"#ffb4a2\", \"MEP Cd9+ Ly6a-\"=\"#e5989b\", \"MEP Cd9- Ly6a-\"=\"#b5838d\")) + \n",
    "    theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) + \n",
    "    guides(fill=guide_legend(ncol=1, override.aes=list(alpha=1, size=1.5), keywidth=0.25, keyheight=0.25, default.unit=\"cm\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15684bab-500d-43a1-878f-7df6fc6b7972",
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- so@meta.data %>% \n",
    "    dplyr::mutate(group=ifelse(msP_3>0, \"Cd24a+ Tfrc+\", \"Negative\")) %>%\n",
    "    dplyr::mutate(group=factor(group, levels=c(\"Cd24a+ Tfrc+\", \"Negative\")))\n",
    "\n",
    "bp_3 <- ggplot(data, aes(x=celltype_low, fill=group)) + \n",
    "    geom_bar(position=\"fill\") + \n",
    "    scale_fill_manual(values=c(\"Cd24a+ Tfrc+\"=\"#ffcdb2\", \"Negative\"=\"#ffb4a2\")) + \n",
    "    theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) + \n",
    "    guides(fill=guide_legend(ncol=1, override.aes=list(alpha=1, size=1.5), keywidth=0.25, keyheight=0.25, default.unit=\"cm\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb728f66-5cfb-426b-b005-36028557749a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_4/bp_p_ratio.pdf\", width=3, height=6)\n",
    "\n",
    "gridExtra::grid.arrange(\n",
    "\n",
    "    bp_1 %>% egg::set_panel_size(., width=unit(c(3*0.3, 2*0.3, 2*0.3, 0.3), \"cm\"), height=unit(2.0, \"cm\")),\n",
    "    bp_2 %>% egg::set_panel_size(., width=unit(c(3*0.3, 2*0.3, 2*0.3, 0.3), \"cm\"), height=unit(2.0, \"cm\")),\n",
    "    bp_3 %>% egg::set_panel_size(., width=unit(c(7*0.3), \"cm\"), height=unit(2.0, \"cm\")), ncol=1\n",
    "        \n",
    ")\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "096b866c-b729-4e8a-a5cc-cc1bf7f078bd",
   "metadata": {},
   "source": [
    "# Cell cycle score "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff2ec75d-d355-4f55-aadd-2f37ac384d6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- readRDS(\"data/scRNAseq/object/pp.rds\")\n",
    "so <- NormalizeData(so)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "538bc82d-ae05-4431-8916-f8738b63796b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# so <- so[, so$celltype_low %in% c(\"MEP\", \"Proerythroblast\", \"Erythroblast\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f32126c4-9fe3-476e-aff9-56d5fdb90637",
   "metadata": {},
   "outputs": [],
   "source": [
    "cc_genes <- read.csv(\"data/reference/cellcycle/seurat.csv\", row.names=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be8fae4a-bea0-4a11-b5ef-53bad7e5a026",
   "metadata": {},
   "outputs": [],
   "source": [
    "features <- list(\n",
    "\n",
    "    msG2M=cc_genes[cc_genes$group==\"G2M\", ]$gene, \n",
    "    msS=cc_genes[cc_genes$group==\"S\", ]$gene, \n",
    "    msCC=c(\"Mki67\", \"Topa2\")\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23604831-9160-4105-8bfa-515dfb78115c",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- UCell::AddModuleScore_UCell(so, features=features, assay=\"RNA\", slot=\"data\", name=\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69b7798c-adfb-4871-9d1f-ce8dc37f587d",
   "metadata": {},
   "outputs": [],
   "source": [
    "so$cc_phase <- ifelse((so$msS>0 & so$msCC>0 & so$msS>so$msG2M), \"S\", ifelse((so$msG2M>0 & so$msCC>0 & so$msG2M>so$msS), \"G2M\", \"G1\")) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "680b4d15-4af3-43ac-abef-d264956e6100",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=35, repr.plot.height=5)\n",
    "\n",
    "ggplot(so@meta.data, aes(x=msCC, y=msS, color=msCC)) + geom_point() + theme(aspect.ratio=1) + theme_global_set(1) + facet_grid(~celltype_low)\n",
    "ggplot(so@meta.data, aes(x=msCC, y=msG2M, color=msCC)) + geom_point() + theme(aspect.ratio=1) + theme_global_set(1) + facet_grid(~celltype_low)\n",
    "ggplot(so@meta.data, aes(x=msG2M, y=msS, color=msCC)) + geom_point() + theme(aspect.ratio=1) + theme_global_set(1) + facet_grid(~celltype_low)\n",
    "ggplot(so@meta.data, aes(x=msG2M, y=msS, color=cc_phase)) + geom_point() + scale_color_manual(values=color$cc_phase) + theme(aspect.ratio=1) + theme_global_set(1) + facet_grid(~celltype_low) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7a3d234-adda-49eb-8730-9396b2af3617",
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- so@meta.data %>% dplyr::filter(celltype_low %in% c(\"MEP\")) %>% \n",
    "    dplyr::mutate(dpi=ifelse(sample_group==\"Bl6_NaCl_D6\", \"D0\", as.character(dpi))) %>% \n",
    "    dplyr::mutate(cc_phase=factor(cc_phase, levels=rev(c(\"G1\", \"S\", \"G2M\"))))\n",
    "\n",
    "bp_1 <- ggplot(data, aes(x=genotype, fill=cc_phase)) + \n",
    "    geom_bar(position=\"fill\") + \n",
    "    facet_grid(~dpi, space=\"free\", scales=\"free\") + \n",
    "    scale_fill_manual(values=color$cc_phase) + \n",
    "    guides(fill=guide_legend(ncol=1, override.aes=list(alpha=1, size=1.5), keywidth=0.25, keyheight=0.25, default.unit=\"cm\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da736d30-f279-4b4c-ae8f-e84fb7384600",
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- so@meta.data %>% dplyr::filter(celltype_low %in% c(\"Proerythroblast\")) %>% \n",
    "    dplyr::mutate(dpi=ifelse(sample_group==\"Bl6_NaCl_D6\", \"D0\", as.character(dpi))) %>% \n",
    "    dplyr::mutate(cc_phase=factor(cc_phase, levels=rev(c(\"G1\", \"S\", \"G2M\"))))\n",
    "\n",
    "bp_2 <- ggplot(data, aes(x=genotype, fill=cc_phase)) + \n",
    "    geom_bar(position=\"fill\") + \n",
    "    facet_grid(~dpi, space=\"free\", scales=\"free\") + \n",
    "    scale_fill_manual(values=color$cc_phase) + \n",
    "    guides(fill=guide_legend(ncol=1, override.aes=list(alpha=1, size=1.5), keywidth=0.25, keyheight=0.25, default.unit=\"cm\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebf0772e-ffc4-49a1-a83e-fa43217396c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- so@meta.data %>% dplyr::filter(celltype_low %in% c(\"Erythroblast\")) %>% \n",
    "    dplyr::mutate(dpi=ifelse(sample_group==\"Bl6_NaCl_D6\", \"D0\", as.character(dpi))) %>% \n",
    "    dplyr::mutate(cc_phase=factor(cc_phase, levels=rev(c(\"G1\", \"S\", \"G2M\"))))\n",
    "\n",
    "bp_3 <- ggplot(data, aes(x=genotype, fill=cc_phase)) + \n",
    "    geom_bar(position=\"fill\") + \n",
    "    facet_grid(~dpi, space=\"free\", scales=\"free\") + \n",
    "    scale_fill_manual(values=color$cc_phase) + \n",
    "    guides(fill=guide_legend(ncol=1, override.aes=list(alpha=1, size=1.5), keywidth=0.25, keyheight=0.25, default.unit=\"cm\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29b37f8a-2d6f-4882-91b1-54d7fa1af9d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_4/bp_cc_phase.pdf\", width=3, height=6)\n",
    "\n",
    "gridExtra::grid.arrange(\n",
    "\n",
    "    bp_1 %>% egg::set_panel_size(., width=unit(c(3*0.3, 2*0.3, 2*0.3, 0.3), \"cm\"), height=unit(2.0, \"cm\")),\n",
    "    bp_2 %>% egg::set_panel_size(., width=unit(c(3*0.3, 2*0.3, 2*0.3, 0.3), \"cm\"), height=unit(2.0, \"cm\")), \n",
    "    bp_3 %>% egg::set_panel_size(., width=unit(c(3*0.3, 2*0.3, 2*0.3, 0.3), \"cm\"), height=unit(2.0, \"cm\")), ncol=1\n",
    "        \n",
    ")\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8eb779ed-f6e4-4af5-96a8-d2bc4c5394fc",
   "metadata": {},
   "source": [
    "# Transcription factor heatmap of smoothed expression "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e035107b-adb2-4583-a58c-3da2f6f124a3",
   "metadata": {},
   "source": [
    "## Load pruned graph TF and target gens "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc236d73-800f-4dda-9127-a85b7af262f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "reg <- read.csv(\"result/scenic/scRNAseq/res_1/reg.csv\", row.names=1, skip=1)\n",
    "colnames(reg) <- c(\"TF\", \"target_genes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfd1642e-f6ea-41b0-9fc1-65fe3af9f52d",
   "metadata": {},
   "outputs": [],
   "source": [
    "tf <- c(\"Nfkb1\", \"Myb\", \"Fli1\", \"Gata2\", \"E2f3\", \"Bcl11a\", \"Myc\", \"Ikzf1\", \"Zfp148\", \"Irf8\", \"Irf1\", \"Irf7\", \"Stat1\", \"Stat2\", \"Stat3\", \"Bclaf1\", \"E2f4\", \"Gata1\", \"E2f7\", \"Tal1\", \"E2f1\", \"Mybl2\", \"Klf1\", \"E2f8\", \"Klf3\", \"E2f2\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "899f51af-9c22-4ad0-9682-840fb8e681ed",
   "metadata": {},
   "source": [
    "## PTCG Ctl (+/+) and D1 (+/+) and D3 (+/+)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8cd6f54-d905-407b-8bc1-7ff4b2ee27b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrast_vec <- c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")\n",
    "color_mat <- color$sample_group[c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "627ae948-ae6e-4eb1-8830-437775c5c1ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_diff_1 <- intersect(ptpg_res_1[ptpg_res_1$ptpg_class==\"Canonical\", ]$gene, ptpg_res_2[ptpg_res_2$ptpg_class==\"Canonical\", ]$gene) # Union of response genes D1 (+/+) and D3 (+/+)\n",
    "genes_diff_1 <- genes_diff_1[!genes_diff_1 %in% ptpg_res_3[ptpg_res_3$ptpg_class != \"Canonical\", ]$gene] # Remove genes that are perturbed between D1 (+/+) and D3 (+/+)\n",
    "length(genes_diff_1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbdbb11f-4e85-4479-8f2e-1d20b4bd60ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "pt_split_ptcg <- pt_split(fitgam, genes_diff_1, contrast_vec)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afd76efa-9b92-4b9c-8cba-ea3105ee6d06",
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_tg_mat_ptcg <- reg[reg$target_genes %in% names(pt_split_ptcg), ]\n",
    "tf_tg_mat_ptcg <- tf_tg_pp(tf_tg_mat_ptcg)\n",
    "tf_tg_mat_ptcg <- tf_tg_prune(tf_tg_mat_ptcg, pt_split_ptcg, tf_ratio=0, tf_sum=0, tf_fill=TRUE)\n",
    "tf_tg_mat_ptcg <- tf_tg_mat_ptcg[, tf]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9147c00-e481-4976-8969-90f390906611",
   "metadata": {},
   "outputs": [],
   "source": [
    "pt_tf_hm_1 <- pt_tf_hm(tf_tg_mat_ptcg, row_split=pt_split_ptcg, height=180, fontsize_select=2, cluster_rows=FALSE, cluster_columns=FALSE, use_raster=TRUE)\n",
    "pt_hm_1 <- pt_hm(fitgam, names(pt_split_ptcg), \"IFNAR_fl_Baseline_D0\", contrast_vec, 50, color_mat[1], row_split=pt_split_ptcg, width=0.75, height=180, fontsize_select=2, use_raster=TRUE)\n",
    "pt_hm_2 <- pt_hm(fitgam, names(pt_split_ptcg), \"IFNAR_fl_CpG_D1\", contrast_vec, 50, color_mat[2], row_split=pt_split_ptcg, width=0.75, height=180, fontsize_select=2, use_raster=TRUE)\n",
    "pt_hm_3 <- pt_hm(fitgam, names(pt_split_ptcg), \"IFNAR_fl_CpG_D3\", contrast_vec, 50, color_mat[3], row_split=pt_split_ptcg, width=0.75, height=180, fontsize_select=2, use_raster=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96d61cf1-4403-45ea-af36-c1458ef347f6",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "row_genes <- c(\n",
    "\n",
    "    \"Meis1\", \"Klf1\", \"Tal1\", \"Gata1\", \"Gata2\", # Transcription factors \n",
    "    \"Mki67\", \"Top2a\", # Proliferation \n",
    "    \"Pkm\", # Glycolysis\n",
    "    \"Pdk1\", # Glycolysis/aerobic metabolism\n",
    "    \"Hif1a\", \"Ldha\", # Glycolysis/anaerboic metabolism\n",
    "    \"Atp5a1\", \"Atp5b\", # Oxidative phosphorylation \n",
    "    \"Sod2\", # ROS detoxifiyer\n",
    "    \"G6pd\", \"Pgd\", # PP\n",
    "    \"Cpt1a\", \"Acadl\", # b-oxidatse\n",
    "    \"Mtor\", \"Rptor\", \"Rheb\",  # mTOR pathwy\n",
    "    \"Tfam\", \"Ssbp1\", \"Polg\", # Mitochondrial maintainance\n",
    "    \"Bnip3l\", # Mitochondrial removal \n",
    "    \"Dnmt1\", \"Dnmt3b\", \"Tet2\", \"Tet3\", # Epigenetics\n",
    "    \"Alas2\", \"Ncoa4\", # Heme metabolism \n",
    "    \"Icam4\", \"Epor\", \"Tfrc\" # Receptor\n",
    "\n",
    ")\n",
    "\n",
    "row_genes <- row_genes[row_genes %in% rownames(pt_tf_hm_1@matrix)]\n",
    "\n",
    "row_anno <- rowAnnotation(\n",
    "    \n",
    "    labels=anno_mark(\n",
    "        \n",
    "        at=which(rownames(pt_tf_hm_1@matrix) %in% row_genes),\n",
    "        labels=rownames(pt_tf_hm_1@matrix)[which(rownames(pt_tf_hm_1@matrix) %in% row_genes)],\n",
    "        side=\"left\", \n",
    "        labels_gp=gpar(fontsize=6), \n",
    "        link_width=unit(3, \"mm\")\n",
    "    \n",
    "    )\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df58f7ac-4142-4986-929b-2794225a20f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_4/hm_tf_ptcg_ctl_d1_d3.pdf\", width=5, height=10)\n",
    "\n",
    "draw(row_anno + pt_tf_hm_1 + pt_hm_1 + pt_hm_2 + pt_hm_3)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8825815-2dcc-40a2-93c3-513e348d43a1",
   "metadata": {},
   "source": [
    "## PTPG D1 (+/+) or D3 (+/+) UP vs Ctl (+/+)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10048c6d-bc14-4f86-8847-8acc18bffe07",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrast_vec <- c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")\n",
    "color_mat <- color$sample_group[c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e34b49e-58a7-478f-a9a3-0dde30e6b452",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "genes_diff_1 <- intersect(ptpg_res_1[ptpg_res_1$ptpg_class==\"IFNAR_fl_CpG_D1\", ]$gene, ptpg_res_2[ptpg_res_2$ptpg_class==\"IFNAR_fl_CpG_D3\", ]$gene) # Union of response genes UP D1 (+/+) and D3 (+/+) vs Ctl (+/+)\n",
    "length(genes_diff_1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f1cdfb3-bdf3-4256-ae78-1583edbaaf56",
   "metadata": {},
   "outputs": [],
   "source": [
    "pt_split_ptpg <- pt_split(fitgam, genes_diff_1, contrast_vec)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45ca11e6-949d-47f9-b522-d5a1f28cfcfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_tg_mat_ptcg <- reg[reg$target_genes %in% names(pt_split_ptpg), ]\n",
    "tf_tg_mat_ptcg <- tf_tg_pp(tf_tg_mat_ptcg)\n",
    "tf_tg_mat_ptcg <- tf_tg_prune(tf_tg_mat_ptcg, pt_split_ptpg, tf_ratio=0, tf_sum=0, tf_fill=TRUE)\n",
    "tf_tg_mat_ptcg[tf[!tf %in% colnames(tf_tg_mat_ptcg)]] <- 0\n",
    "tf_tg_mat_ptcg <- tf_tg_mat_ptcg[, tf]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c637b700-a1ea-4aad-b6cc-b77a72c019c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "pt_tf_hm_1 <- pt_tf_hm(tf_tg_mat_ptcg, row_split=pt_split_ptpg, height=50, fontsize_select=2, cluster_rows=FALSE, cluster_columns=FALSE, use_raster=FALSE)\n",
    "pt_hm_1 <- pt_hm(fitgam, names(pt_split_ptpg), \"IFNAR_fl_Baseline_D0\", contrast_vec, 50, color_mat[1], row_split=pt_split_ptpg, width=0.75, height=50, fontsize_select=2, use_raster=TRUE)\n",
    "pt_hm_2 <- pt_hm(fitgam, names(pt_split_ptpg), \"IFNAR_fl_CpG_D1\", contrast_vec, 50, color_mat[2], row_split=pt_split_ptpg, width=0.75, height=50, fontsize_select=2, use_raster=TRUE)\n",
    "pt_hm_3 <- pt_hm(fitgam, names(pt_split_ptpg), \"IFNAR_fl_CpG_D3\", contrast_vec, 50, color_mat[3], row_split=pt_split_ptpg, width=0.75, height=50, fontsize_select=2, use_raster=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03c4fad2-2427-4575-85fe-698536e6496e",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "row_genes <- c(\n",
    "    \n",
    "    \"Meis1\", \"Klf1\", \"Tal1\", \"Gata1\", \"Gata2\", # Transcription factors \n",
    "    \"Mki67\", \"Top2a\", # Proliferation \n",
    "    \"Pkm\", # Glycolysis\n",
    "    \"Pdk1\", # Glycolysis/aerobic metabolism\n",
    "    \"Hif1a\", \"Ldha\", # Glycolysis/anaerboic metabolism\n",
    "    \"Atp5a1\", \"Atp5b\", # Oxidative phosphorylation \n",
    "    \"Sod2\", # ROS detoxifiyer\n",
    "    \"G6pd\", \"Pgd\", # PP\n",
    "    \"Cpt1a\", \"Acadl\", # b-oxidatse\n",
    "    \"Mtor\", \"Rptor\", \"Rheb\",  # mTOR pathwy\n",
    "    \"Tfam\", \"Ssbp1\", \"Polg\", # Mitochondrial maintainance\n",
    "    \"Bnip3l\", # Mitochondrial removal \n",
    "    \"Dnmt1\", \"Dnmt3b\", \"Tet2\", \"Tet3\", # Epigenetics\n",
    "    \"Alas2\", \"Ncoa4\", # Heme metabolism \n",
    "    \"Icam4\", \"Epor\", \"Tfrc\", # Receptor\n",
    "    \"Kit\", \"Cd24a\", \"Tfrc\", \"Slamf1\", \"Cd9\", \"Ly6a\", \n",
    "    grep(\"Isg|Ifi|Irf|Socs|Smad|Nfkb|Stat|Icam|Smad|Bmp\", genes_diff_1, value=TRUE), \n",
    "    grep(\"Tlr[0-9]+|Nlrp|Nlrc|Nod|Ddx58|Ifih1|Clec[0-9]+|Fc[grme]|Ifnar|Ifngr|Ifnlr|Il[0-9]+r[a-z]?|Ccr[0-9]+|Cxcr[0-9]+|Cx3cr1|Xcr1|Cd3|Cd79|Klra|Klrb|Klrc|Klrd|Klre|Klrf|Klrg|Kit|Cd[0-9]+[a-zA-Z]?|Tgfb[123]|Tgfbr[123]|Smad[1-9]|Bmp[0-9]+|Bmpr1a|Bmpr1b|Bmpr2|Acvr1|Acvr2a|Acvr2b|Gdf[0-9]+|Kitl|Kit|Nr3c1|Ptch[12]|Smo|Cdon|Boc|Gas1\", genes_diff_1, value=TRUE)\n",
    "\n",
    ")\n",
    "\n",
    "row_genes <- row_genes[row_genes %in% rownames(pt_tf_hm_1@matrix)]\n",
    "\n",
    "row_anno <- rowAnnotation(\n",
    "    \n",
    "    labels=anno_mark(\n",
    "        at=which(rownames(pt_tf_hm_1@matrix) %in% row_genes),\n",
    "        labels=rownames(pt_tf_hm_1@matrix)[which(rownames(pt_tf_hm_1@matrix) %in% row_genes)],\n",
    "        side=\"left\", \n",
    "        labels_gp=gpar(fontsize=6), \n",
    "        link_width=unit(3, \"mm\")\n",
    "    \n",
    "    )\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b33fba92-8db2-4c0c-bb3d-6f63e3f48752",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_4/hm_tf_ptpg_d1_or_d3_up_vs_ctl.pdf\", width=6, height=10)\n",
    "\n",
    "draw(row_anno + pt_tf_hm_1 + pt_hm_1 + pt_hm_2 + pt_hm_3)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a670e9d0-814e-45a6-961e-073df67acdb8",
   "metadata": {},
   "source": [
    "## PTPG D1 (+/+) or D3 (+/+) DOWN vs Ctl (+/+)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0341f0f3-8366-453b-9387-4a5d29a78caf",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrast_vec <- c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")\n",
    "color_mat <- color$sample_group[c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd51b58e-3af6-4f5f-be45-252034634722",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "genes_diff_1 <- union(ptpg_res_1[ptpg_res_1$ptpg_class==\"IFNAR_fl_Baseline_D0\", ]$gene, ptpg_res_2[ptpg_res_2$ptpg_class==\"IFNAR_fl_Baseline_D0\", ]$gene) # Union of response genes DOWN D1 (+/+) and D3 (+/+) vs Ctl (+/+)\n",
    "length(genes_diff_1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05aad538-0a74-460a-be6b-224fe8212ff9",
   "metadata": {},
   "outputs": [],
   "source": [
    "pt_split_ptpg <- pt_split(fitgam, genes_diff_1, contrast_vec)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d79b75c5-bfd1-4959-b3f3-ed25f18b97f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_tg_mat_ptcg <- reg[reg$target_genes %in% names(pt_split_ptpg), ]\n",
    "tf_tg_mat_ptcg <- tf_tg_pp(tf_tg_mat_ptcg)\n",
    "tf_tg_mat_ptcg <- tf_tg_prune(tf_tg_mat_ptcg, pt_split_ptpg, tf_ratio=0, tf_sum=0, tf_fill=TRUE)\n",
    "tf_tg_mat_ptcg[tf[!tf %in% colnames(tf_tg_mat_ptcg)]] <- 0\n",
    "tf_tg_mat_ptcg <- tf_tg_mat_ptcg[, tf]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72a083da-60c8-488b-805a-e7ebae82570f",
   "metadata": {},
   "outputs": [],
   "source": [
    "pt_tf_hm_1 <- pt_tf_hm(tf_tg_mat_ptcg, row_split=pt_split_ptpg, height=50, fontsize_select=2, cluster_rows=FALSE, cluster_columns=FALSE, use_raster=FALSE)\n",
    "pt_hm_1 <- pt_hm(fitgam, names(pt_split_ptpg), \"IFNAR_fl_Baseline_D0\", contrast_vec, 50, color_mat[1], row_split=pt_split_ptpg, width=0.75, height=50, fontsize_select=2, use_raster=TRUE)\n",
    "pt_hm_2 <- pt_hm(fitgam, names(pt_split_ptpg), \"IFNAR_fl_CpG_D1\", contrast_vec, 50, color_mat[2], row_split=pt_split_ptpg, width=0.75, height=50, fontsize_select=2, use_raster=TRUE)\n",
    "pt_hm_3 <- pt_hm(fitgam, names(pt_split_ptpg), \"IFNAR_fl_CpG_D3\", contrast_vec, 50, color_mat[3], row_split=pt_split_ptpg, width=0.75, height=50, fontsize_select=2, use_raster=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b00509f3-215d-4eeb-a52e-593a52fcceb2",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "row_genes <- rownames(pt_tf_hm_1@matrix)\n",
    "\n",
    "row_anno <- rowAnnotation(\n",
    "    \n",
    "    labels=anno_mark(\n",
    "        at=which(rownames(pt_tf_hm_1@matrix) %in% row_genes),\n",
    "        labels=rownames(pt_tf_hm_1@matrix)[which(rownames(pt_tf_hm_1@matrix) %in% row_genes)],\n",
    "        side=\"left\", \n",
    "        labels_gp=gpar(fontsize=6), \n",
    "        link_width=unit(3, \"mm\")\n",
    "    \n",
    "    )\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23c7543d-faf1-41c0-8db0-b9af1eae44ba",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_4/hm_tf_ptpg_d1_or_d3_down_vs_ctl.pdf\", width=6, height=10)\n",
    "\n",
    "draw(row_anno + pt_tf_hm_1 + pt_hm_1 + pt_hm_2 + pt_hm_3)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71e657fc-ccd7-4d9e-86d4-45b0e83f61c4",
   "metadata": {},
   "source": [
    "## PTPG D1 (+/+) UP vs D3 (+/+) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b21ca043-f81b-4476-9083-f5133db7c00e",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrast_vec <- c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")\n",
    "color_mat <- color$sample_group[c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3212d57-6bf4-47f2-aa9e-1727bad8053c",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "genes_diff_1 <- ptpg_res_3[ptpg_res_3$ptpg_class==\"IFNAR_fl_CpG_D1\", ]$gene # Genes up on D1 (+/+) vs D3 (+/+)\n",
    "length(genes_diff_1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e22f78ef-2c81-4c99-a875-61e146616278",
   "metadata": {},
   "outputs": [],
   "source": [
    "pt_split_ptpg <- pt_split(fitgam, genes_diff_1, contrast_vec)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47e6a39c-b3e3-49d9-8cbe-bd50ff78ed33",
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_tg_mat_ptcg <- reg[reg$target_genes %in% names(pt_split_ptpg), ]\n",
    "tf_tg_mat_ptcg <- tf_tg_pp(tf_tg_mat_ptcg)\n",
    "tf_tg_mat_ptcg <- tf_tg_prune(tf_tg_mat_ptcg, pt_split_ptpg, tf_ratio=0, tf_sum=0, tf_fill=TRUE)\n",
    "tf_tg_mat_ptcg <- tf_tg_mat_ptcg[, colnames(tf_tg_mat_ptcg) %in% tf]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9dde111-b8f1-4229-8b0f-54186ff177e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_fill <- setdiff(tf, colnames(tf_tg_mat_ptcg))\n",
    "tf_tg_mat_ptcg[, tf_fill] <- 0\n",
    "tf_tg_mat_ptcg <- tf_tg_mat_ptcg[, tf, drop=FALSE]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17b57e90-bfb8-476e-be69-74c5e94d90f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "pt_tf_hm_1 <- pt_tf_hm(tf_tg_mat_ptcg, row_split=pt_split_ptpg, height=50, fontsize_select=2, cluster_rows=FALSE, cluster_columns=FALSE, use_raster=FALSE)\n",
    "pt_hm_1 <- pt_hm(fitgam, names(pt_split_ptpg), \"IFNAR_fl_Baseline_D0\", contrast_vec, 50, color_mat[1], row_split=pt_split_ptpg, width=0.75, height=50, fontsize_select=2, use_raster=TRUE)\n",
    "pt_hm_2 <- pt_hm(fitgam, names(pt_split_ptpg), \"IFNAR_fl_CpG_D1\", contrast_vec, 50, color_mat[2], row_split=pt_split_ptpg, width=0.75, height=50, fontsize_select=2, use_raster=TRUE)\n",
    "pt_hm_3 <- pt_hm(fitgam, names(pt_split_ptpg), \"IFNAR_fl_CpG_D3\", contrast_vec, 50, color_mat[3], row_split=pt_split_ptpg, width=0.75, height=50, fontsize_select=2, use_raster=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43c85f26-5e46-4603-8573-4380221d6dd1",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "row_genes <- rownames(pt_tf_hm_1@matrix)\n",
    "\n",
    "row_anno <- rowAnnotation(\n",
    "    \n",
    "    labels=anno_mark(\n",
    "        at=which(rownames(pt_tf_hm_1@matrix) %in% row_genes),\n",
    "        labels=rownames(pt_tf_hm_1@matrix)[which(rownames(pt_tf_hm_1@matrix) %in% row_genes)],\n",
    "        side=\"left\", \n",
    "        labels_gp=gpar(fontsize=6), \n",
    "        link_width=unit(3, \"mm\")\n",
    "    \n",
    "    )\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c7f014c-4f7e-4973-a92a-1d64c6720e8b",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_4/hm_tf_ptpg_d1_up_vs_d3.pdf\", width=6, height=10)\n",
    "\n",
    "draw(row_anno + pt_tf_hm_1 + pt_hm_1 + pt_hm_2 + pt_hm_3)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a6a2686-4b27-455d-a021-00cbc98c7d5e",
   "metadata": {},
   "source": [
    "## PTPG D1 (+/+) DOWN vs D3 (+/+) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a09c4ed5-66dc-4aac-bfb3-639c30d46c36",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrast_vec <- c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")\n",
    "color_mat <- color$sample_group[c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed76bd1a-96a3-4066-b35d-91bdbc4ea830",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "genes_diff_1 <- ptpg_res_3[ptpg_res_3$ptpg_class==\"IFNAR_fl_CpG_D3\", ]$gene # Genes up on D1 (+/+) vs D3 (+/+)\n",
    "length(genes_diff_1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55db1857-fa21-44c8-b2e3-0f75227dde40",
   "metadata": {},
   "outputs": [],
   "source": [
    "pt_split_ptpg <- pt_split(fitgam, genes_diff_1, contrast_vec)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f42dd734-cf54-428a-af40-31cfb0fd2b61",
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_tg_mat_ptcg <- reg[reg$target_genes %in% names(pt_split_ptpg), ]\n",
    "tf_tg_mat_ptcg <- tf_tg_pp(tf_tg_mat_ptcg)\n",
    "tf_tg_mat_ptcg <- tf_tg_prune(tf_tg_mat_ptcg, pt_split_ptpg, tf_ratio=0, tf_sum=0, tf_fill=TRUE)\n",
    "tf_tg_mat_ptcg <- tf_tg_mat_ptcg[, colnames(tf_tg_mat_ptcg) %in% tf]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de907ca8-28a5-47b0-ba4e-20b0fab475c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_fill <- setdiff(tf, colnames(tf_tg_mat_ptcg))\n",
    "tf_tg_mat_ptcg[, tf_fill] <- 0\n",
    "tf_tg_mat_ptcg <- tf_tg_mat_ptcg[, tf, drop=FALSE]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6058bb74-3ca0-46a9-882d-f4ca540af2b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pt_tf_hm_1 <- pt_tf_hm(tf_tg_mat_ptcg, row_split=pt_split_ptpg, height=50, fontsize_select=2, cluster_rows=FALSE, cluster_columns=FALSE, use_raster=FALSE)\n",
    "pt_hm_1 <- pt_hm(fitgam, names(pt_split_ptpg), \"IFNAR_fl_Baseline_D0\", contrast_vec, 50, color_mat[1], row_split=pt_split_ptpg, width=0.75, height=50, fontsize_select=2, use_raster=TRUE)\n",
    "pt_hm_2 <- pt_hm(fitgam, names(pt_split_ptpg), \"IFNAR_fl_CpG_D1\", contrast_vec, 50, color_mat[2], row_split=pt_split_ptpg, width=0.75, height=50, fontsize_select=2, use_raster=TRUE)\n",
    "pt_hm_3 <- pt_hm(fitgam, names(pt_split_ptpg), \"IFNAR_fl_CpG_D3\", contrast_vec, 50, color_mat[3], row_split=pt_split_ptpg, width=0.75, height=50, fontsize_select=2, use_raster=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87984417-2555-41a8-8555-b907fb6aa3f3",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "row_genes <- rownames(pt_tf_hm_1@matrix)\n",
    "\n",
    "row_anno <- rowAnnotation(\n",
    "    \n",
    "    labels=anno_mark(\n",
    "        at=which(rownames(pt_tf_hm_1@matrix) %in% row_genes),\n",
    "        labels=rownames(pt_tf_hm_1@matrix)[which(rownames(pt_tf_hm_1@matrix) %in% row_genes)],\n",
    "        side=\"left\", \n",
    "        labels_gp=gpar(fontsize=6), \n",
    "        link_width=unit(3, \"mm\")\n",
    "    \n",
    "    )\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8792287c-0b72-4d34-957a-5da90ce933bf",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_4/hm_tf_ptpg_d1_down_vs_d3.pdf\", width=6, height=10)\n",
    "\n",
    "draw(row_anno + pt_tf_hm_1 + pt_hm_1 + pt_hm_2 + pt_hm_3)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a16ce23-529e-4392-b980-80b33514ca7a",
   "metadata": {},
   "source": [
    "# Line plots smoothed expression "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8e2c61a-1a8c-43e2-8552-a2ff1b2b212e",
   "metadata": {},
   "source": [
    "# Stress erythroid marker "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df7c05d7-2839-40e7-832a-7429a0bfee46",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrast_vec <- c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")\n",
    "color_mat <- color$sample_group[c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13c929e9-44db-439d-9eec-e446ea00f46f",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes <- c(\"Kit\", \"Cd24a\", \"Tfrc\", \"Slamf1\", \"Cd9\", \"Ly6a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8cdbd7e-1d47-4b2a-896b-86856195bee7",
   "metadata": {},
   "outputs": [],
   "source": [
    "lp_1 <- lapply(genes, function(i) {plot_smooth(fitgam, gene=i, point=FALSE, condition_color=color_mat, line_size=unit(1.5/2.141959, \"pt\"), condition_line_type=TRUE) + annotate(\"text\", x=0.5, y=0, label=i, hjust=0.5, vjust=-1, size=2) + theme(legend.position=\"none\")})\n",
    "lp_1 <- lapply(lp_1, function(p) egg::set_panel_size(p, width=unit(2.0, \"cm\"), height=unit(0.75, \"cm\")))\n",
    "lp_1 <- do.call(gridExtra::arrangeGrob, c(lp_1, ncol=10, nrow=ceiling(length(lp_1)/10)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cb74d1c-0723-48d6-864f-d358c0c31ff2",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_4/lp_stress_erythroid_marker.pdf\", width=10*1.25, height=ceiling(length(lp_1)/10))\n",
    "\n",
    "grid::grid.draw(lp_1)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0df20ee5-1901-42d0-950c-1165fef934be",
   "metadata": {},
   "source": [
    "## PTCG Ctl (+/+) and D1 (+/+) and D3 (+/+) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1edfd2d-fd14-417e-bae8-92685d754fb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrast_vec <- c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")\n",
    "color_mat <- color$sample_group[c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "259c2757-4c2b-4bd5-8ea8-35e20f96cc33",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes <- c(\"Meis1\", \"Klf1\", \"Tal1\", \"Gata1\", \"Epor\", \"Tfrc\", \"Alas2\")\n",
    "genes %in% intersect(ptpg_res_1[ptpg_res_1$ptpg_class==\"Canonical\", ]$gene, ptpg_res_2[ptpg_res_2$ptpg_class==\"Canonical\", ]$gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "486ec15b-1093-4764-b6fc-c098544e4e0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "lp_1 <- lapply(genes, function(i) {plot_smooth(fitgam, gene=i, point=FALSE, condition_color=color_mat, line_size=unit(1.5/2.141959, \"pt\"), condition_line_type=FALSE) + annotate(\"text\", x=0.5, y=0, label=i, hjust=0.5, vjust=-1, size=2) + theme(legend.position=\"none\")})\n",
    "lp_1 <- lapply(lp_1, function(p) egg::set_panel_size(p, width=unit(2.0, \"cm\"), height=unit(0.75, \"cm\")))\n",
    "lp_1 <- do.call(gridExtra::arrangeGrob, c(lp_1, ncol=10, nrow=ceiling(length(lp_1)/10)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e151243e-47d3-44c6-bc00-7780e1afef1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_4/lp_ptcg_ctl_d1_d3.pdf\", width=10*1.25, height=ceiling(length(lp_1)/10))\n",
    "\n",
    "grid::grid.draw(lp_1)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22b1717e-368d-4345-a36d-43edc0882024",
   "metadata": {},
   "source": [
    "## PTPG D1 (+/+) or D3 (+/+) UP vs Ctl (+/+)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be565ac3-d2dd-41ec-9e85-c0f48351df45",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrast_vec <- c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")\n",
    "color_mat <- color$sample_group[c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9566503b-60a0-48b0-adeb-8b37f0a1b5bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes <- union(ptpg_res_1[ptpg_res_1$ptpg_class==\"IFNAR_fl_CpG_D1\" , ]$gene, ptpg_res_2[ptpg_res_2$ptpg_class==\"IFNAR_fl_CpG_D3\" , ]$gene)\n",
    "genes <- pt_split(fitgam, genes, contrast_vec) %>% names()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "819072cf-65a4-4642-b313-7756c38c8f2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "lp_1 <- lapply(genes, function(i) {plot_smooth(fitgam, gene=i, point=FALSE, condition_color=color_mat, line_size=unit(1.5/2.141959, \"pt\"), condition_line_type=TRUE) + annotate(\"text\", x=0.5, y=0, label=i, hjust=0.5, vjust=-1, size=2) + theme(legend.position=\"none\")})\n",
    "lp_1 <- lapply(lp_1, function(p) egg::set_panel_size(p, width=unit(2.0, \"cm\"), height=unit(0.75, \"cm\")))\n",
    "lp_1 <- do.call(gridExtra::arrangeGrob, c(lp_1, ncol=10, nrow=ceiling(length(lp_1)/10)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8091e666-c14f-415c-972f-8610215cd043",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_4/lp_ptpg_d1_or_d3_up_vs_ctl.pdf\", width=10*1.25, height=ceiling(length(lp_1)/10))\n",
    "\n",
    "grid::grid.draw(lp_1)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a788da2-8fae-4024-8e16-b05ff7ac16a6",
   "metadata": {},
   "source": [
    "## PTPG D1 (+/+) or D3 (+/+) DOWN vs Ctl (+/+)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08b7bf33-2a4f-49b1-a75a-674b1ba4335b",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrast_vec <- c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")\n",
    "color_mat <- color$sample_group[c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00008cea-0887-4d78-8f8c-1b4aad242c5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes <- union(ptpg_res_1[ptpg_res_1$ptpg_class==\"IFNAR_fl_Baseline_D0\" , ]$gene, ptpg_res_2[ptpg_res_2$ptpg_class==\"IFNAR_fl_Baseline_D0\" , ]$gene)\n",
    "genes <- pt_split(fitgam, genes, contrast_vec) %>% names()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "220c5171-9946-4c3d-916f-f6d245bee971",
   "metadata": {},
   "outputs": [],
   "source": [
    "lp_1 <- lapply(genes, function(i) {plot_smooth(fitgam, gene=i, point=FALSE, condition_color=color_mat, line_size=unit(1.5/2.141959, \"pt\"), condition_line_type=TRUE) + annotate(\"text\", x=0.5, y=0, label=i, hjust=0.5, vjust=-1, size=2) + theme(legend.position=\"none\")})\n",
    "lp_1 <- lapply(lp_1, function(p) egg::set_panel_size(p, width=unit(2.0, \"cm\"), height=unit(0.75, \"cm\")))\n",
    "lp_1 <- do.call(gridExtra::arrangeGrob, c(lp_1, ncol=10, nrow=ceiling(length(lp_1)/10)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "760b409b-9561-4973-b6f1-97d6d4a7361c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_4/lp_ptpg_d1_or_d3_down_vs_ctl.pdf\", width=10*1.25, height=ceiling(length(lp_1)/10))\n",
    "\n",
    "grid::grid.draw(lp_1)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95d930b2-bbee-4056-99bd-9a8184c70a9c",
   "metadata": {},
   "source": [
    "## PTPG D1 (+/+) or D3 (+/+) vs Ctl (+/+)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8583f67-98e3-4bb3-9c12-aac39f1ebfd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrast_vec <- c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")\n",
    "color_mat <- color$sample_group[c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd206cb6-521b-46d2-86ff-4c0762c833a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes <- union(ptpg_res_1[ptpg_res_1$ptpg_class==\"Perturbed\" , ]$gene, ptpg_res_2[ptpg_res_2$ptpg_class==\"Perturbed\" , ]$gene)\n",
    "genes <- pt_split(fitgam, genes, contrast_vec) %>% names()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef4bdb4f-e0dc-4363-8e04-840021b471b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# lp_1 <- lapply(genes, function(i) {plot_smooth(fitgam, gene=i, point=FALSE, condition_color=color_mat, line_size=unit(1.5/2.141959, \"pt\"), condition_line_type=TRUE) + annotate(\"text\", x=0.5, y=0, label=i, hjust=0.5, vjust=-1, size=2) + theme(legend.position=\"none\")})\n",
    "# lp_1 <- lapply(lp_1, function(p) egg::set_panel_size(p, width=unit(2.0, \"cm\"), height=unit(0.75, \"cm\")))\n",
    "# lp_1 <- do.call(gridExtra::arrangeGrob, c(lp_1, ncol=10, nrow=ceiling(length(lp_1)/10)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0f00b6d-0948-4ca4-aa5b-807ff43e6e67",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pdf(\"result/figures/1_scRNAseq/panel_4/lp_ptpg_d1_or_d3_vs_ctl.pdf\", width=10*1.25, height=ceiling(length(lp_1)/10))\n",
    "\n",
    "# grid::grid.draw(lp_1)\n",
    "\n",
    "# dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bed53560-e4f6-49db-90f3-b2b3d03ddcf8",
   "metadata": {},
   "source": [
    "## PTPG D1 (+/+) UP vs D3 (+/+) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d5eaecc-bcc7-42d9-9dd6-2fa69069c31f",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrast_vec <- c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")\n",
    "color_mat <- color$sample_group[c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb1f13ba-cb13-41ec-8812-70f20e3a001e",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes <- ptpg_res_3[ptpg_res_3$ptpg_class==\"IFNAR_fl_CpG_D1\" , ]$gene\n",
    "genes <- pt_split(fitgam, genes, contrast_vec) %>% names()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5313eee8-15f0-4d0f-88b2-aed85b404c2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "lp_1 <- lapply(genes, function(i) {plot_smooth(fitgam, gene=i, point=FALSE, condition_color=color_mat, line_size=unit(1.5/2.141959, \"pt\"), condition_line_type=TRUE) + annotate(\"text\", x=0.5, y=0, label=i, hjust=0.5, vjust=-1, size=2) + theme(legend.position=\"none\")})\n",
    "lp_1 <- lapply(lp_1, function(p) egg::set_panel_size(p, width=unit(2.0, \"cm\"), height=unit(0.75, \"cm\")))\n",
    "lp_1 <- do.call(gridExtra::arrangeGrob, c(lp_1, ncol=10, nrow=ceiling(length(lp_1)/10)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d35bbf9d-ab32-479b-8ed5-1bf0aebc1f39",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_4/lp_ptpg_d1_up_vs_d3.pdf\", width=10*1.25, height=ceiling(length(lp_1)/10))\n",
    "\n",
    "grid::grid.draw(lp_1)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16157e6e-144b-4b1f-8da9-74eb9540fd0d",
   "metadata": {},
   "source": [
    "## PTPG D1 (+/+) DOWN vs D3 (+/+) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a111b645-ad11-4258-b39c-710ce7d99fe7",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrast_vec <- c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")\n",
    "color_mat <- color$sample_group[c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1ef32d7-b730-49b9-80d6-7c2a6ba4a9d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes <- ptpg_res_3[ptpg_res_3$ptpg_class==\"IFNAR_fl_CpG_D3\" , ]$gene\n",
    "genes <- pt_split(fitgam, genes, contrast_vec) %>% names()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45ad58df-23f3-426a-9252-f56d9c207672",
   "metadata": {},
   "outputs": [],
   "source": [
    "lp_1 <- lapply(genes, function(i) {plot_smooth(fitgam, gene=i, point=FALSE, condition_color=color_mat, line_size=unit(1.5/2.141959, \"pt\"), condition_line_type=TRUE) + annotate(\"text\", x=0.5, y=0, label=i, hjust=0.5, vjust=-1, size=2) + theme(legend.position=\"none\")})\n",
    "lp_1 <- lapply(lp_1, function(p) egg::set_panel_size(p, width=unit(2.0, \"cm\"), height=unit(0.75, \"cm\")))\n",
    "lp_1 <- do.call(gridExtra::arrangeGrob, c(lp_1, ncol=10, nrow=ceiling(length(lp_1)/10)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63e961a5-1a38-4d8c-bfd8-09a75a2d1040",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_4/lp_ptpg_d1_down_vs_d3.pdf\", width=10*1.25, height=ceiling(length(lp_1)/10))\n",
    "\n",
    "grid::grid.draw(lp_1)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce398d9f-6d72-45e4-a66a-7dc1f7c653ec",
   "metadata": {},
   "source": [
    "## PTPG D1 (+/+) vs D3 (+/+)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86bd0c93-3bd6-49f2-a16d-29b4d862f6f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrast_vec <- c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")\n",
    "color_mat <- color$sample_group[c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32bd1b69-a281-42b1-a5a6-220b046039c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes <- ptpg_res_3[ptpg_res_3$ptpg_class==\"Perturbed\" , ]$gene\n",
    "genes <- pt_split(fitgam, genes, contrast_vec) %>% names()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d48e185-4523-4b2b-9b7c-45ab1a77fdb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# lp_1 <- lapply(genes, function(i) {plot_smooth(fitgam, gene=i, point=FALSE, condition_color=color_mat, line_size=unit(1.5/2.141959, \"pt\"), condition_line_type=TRUE) + annotate(\"text\", x=0.5, y=0, label=i, hjust=0.5, vjust=-1, size=2) + theme(legend.position=\"none\")})\n",
    "# lp_1 <- lapply(lp_1, function(p) egg::set_panel_size(p, width=unit(2.0, \"cm\"), height=unit(0.75, \"cm\")))\n",
    "# lp_1 <- do.call(gridExtra::arrangeGrob, c(lp_1, ncol=10, nrow=ceiling(length(lp_1)/10)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fdfd773-8d63-4f19-bb98-6e0222135c5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pdf(\"result/figures/1_scRNAseq/panel_4/lp_ptpg_d1_vs_d3.pdf\", width=10*1.25, height=ceiling(length(lp_1)/10))\n",
    "\n",
    "# grid::grid.draw(lp_1)\n",
    "\n",
    "# dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62d50ffc-29bf-4ac9-be63-98e111011a26",
   "metadata": {},
   "source": [
    "# Pie chart PT gene counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75d4e66c-a22f-4a0f-88c7-a045bdf79d41",
   "metadata": {},
   "outputs": [],
   "source": [
    "ptpg_class_color <- c(\"Perturbed\"=\"#7f7f7f\", \"Canonical\"=\"#AEAEAE\")\n",
    "data <- table(ptpg_res_1 %>% dplyr::mutate(ptpg_class=ifelse(ptpg_class==\"Canonical\", \"Canonical\", \"Perturbed\")) %>% dplyr::pull(ptpg_class)) %>% data.frame() %>% dplyr::rename(group=Var1, count=Freq) %>% dplyr::mutate(group=factor(group, levels=names(ptpg_class_color)))\n",
    "data\n",
    "p_pl_1 <- ggplot(data, aes(x=\"\", y=count, fill=group)) +\n",
    "    geom_col(width=1) +\n",
    "    coord_polar(theta=\"y\") + \n",
    "    scale_fill_manual(values=ptpg_class_color) + \n",
    "    ggtitle(\"Ctl (+/+) vs D1 (+/+)\") + \n",
    "    guides(fill=guide_legend(ncol=1, override.aes=list(alpha=1, size=1.5), keywidth=0.25, keyheight=0.25, default.unit=\"cm\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "858d488f-859e-4586-9fee-c52812f4eb6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ptpg_class_color <- c(\"Perturbed\"=\"#7f7f7f\", \"Canonical\"=\"#AEAEAE\")\n",
    "data <- table(ptpg_res_2 %>% dplyr::mutate(ptpg_class=ifelse(ptpg_class==\"Canonical\", \"Canonical\", \"Perturbed\")) %>% dplyr::pull(ptpg_class)) %>% data.frame() %>% dplyr::rename(group=Var1, count=Freq) %>% dplyr::mutate(group=factor(group, levels=names(ptpg_class_color)))\n",
    "data\n",
    "p_pl_2 <- ggplot(data, aes(x=\"\", y=count, fill=group)) +\n",
    "    geom_col(width=1) +\n",
    "    coord_polar(theta=\"y\") + \n",
    "    scale_fill_manual(values=ptpg_class_color) + \n",
    "    ggtitle(\"Ctl (+/+) vs D1 (+/+)\") + \n",
    "    guides(fill=guide_legend(ncol=1, override.aes=list(alpha=1, size=1.5), keywidth=0.25, keyheight=0.25, default.unit=\"cm\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53634b73-3358-4c18-8a3d-2ba131b15d45",
   "metadata": {},
   "outputs": [],
   "source": [
    "ptpg_class_color <- c(\"Perturbed\"=\"#7f7f7f\", \"Canonical\"=\"#AEAEAE\")\n",
    "data <- table(ptpg_res_3 %>% dplyr::mutate(ptpg_class=ifelse(ptpg_class==\"Canonical\", \"Canonical\", \"Perturbed\")) %>% dplyr::pull(ptpg_class)) %>% data.frame() %>% dplyr::rename(group=Var1, count=Freq) %>% dplyr::mutate(group=factor(group, levels=names(ptpg_class_color)))\n",
    "data\n",
    "p_pl_3 <- ggplot(data, aes(x=\"\", y=count, fill=group)) +\n",
    "    geom_col(width=1) +\n",
    "    coord_polar(theta=\"y\") + \n",
    "    scale_fill_manual(values=ptpg_class_color) + \n",
    "    ggtitle(\"Ctl (+/+) vs D1 (+/+)\") + \n",
    "    guides(fill=guide_legend(ncol=1, override.aes=list(alpha=1, size=1.5), keywidth=0.25, keyheight=0.25, default.unit=\"cm\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ded92634-af5c-4ca8-9c83-0d020d581f24",
   "metadata": {},
   "outputs": [],
   "source": [
    "ptpg_class_color <- color$sample_group[c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D1\")]\n",
    "data <- table(ptpg_res_1 %>% dplyr::filter(ptpg_class %in% names(ptpg_class_color)) %>% dplyr::pull(ptpg_class)) %>% data.frame() %>% dplyr::rename(group=Var1, count=Freq) %>% dplyr::mutate(group=factor(group, levels=names(ptpg_class_color)))\n",
    "data\n",
    "p_pl_4 <- ggplot(data, aes(x=\"\", y=count, fill=group)) +\n",
    "    geom_col(width=1) +\n",
    "    coord_polar(theta=\"y\") + \n",
    "    scale_fill_manual(values=ptpg_class_color) + \n",
    "    ggtitle(\"Ctl (+/+) vs D1 (+/+)\") + \n",
    "    guides(fill=guide_legend(ncol=1, override.aes=list(alpha=1, size=1.5), keywidth=0.25, keyheight=0.25, default.unit=\"cm\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "090c818a-a1d3-4b37-bbf9-91a797f544cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "ptpg_class_color <- color$sample_group[c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_CpG_D3\")]\n",
    "ptpg_class_color[2] <- scales::alpha(ptpg_class_color[2], 0.5)\n",
    "data <- table(ptpg_res_2 %>% dplyr::filter(ptpg_class %in% names(ptpg_class_color)) %>% dplyr::pull(ptpg_class)) %>% data.frame() %>% dplyr::rename(group=Var1, count=Freq) %>% dplyr::mutate(group=factor(group, levels=names(ptpg_class_color)))\n",
    "data\n",
    "p_pl_5 <- ggplot(data, aes(x=\"\", y=count, fill=group)) +\n",
    "    geom_col(width=1) +\n",
    "    coord_polar(theta=\"y\") + \n",
    "    scale_fill_manual(values=ptpg_class_color) + \n",
    "    ggtitle(\"Ctl (+/+) vs D1 (+/+)\") + \n",
    "    guides(fill=guide_legend(ncol=1, override.aes=list(alpha=1, size=1.5), keywidth=0.25, keyheight=0.25, default.unit=\"cm\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92db9a1d-2970-406a-a36b-9c4f9f9851c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "ptpg_class_color <- color$sample_group[c(\"IFNAR_fl_CpG_D1\", \"IFNAR_fl_CpG_D3\")]\n",
    "ptpg_class_color[2] <- scales::alpha(ptpg_class_color[2], 0.5)\n",
    "data <- table(ptpg_res_3 %>% dplyr::filter(ptpg_class %in% names(ptpg_class_color)) %>% dplyr::pull(ptpg_class)) %>% data.frame() %>% dplyr::rename(group=Var1, count=Freq) %>% dplyr::mutate(group=factor(group, levels=names(ptpg_class_color)))\n",
    "data\n",
    "p_pl_6 <- ggplot(data, aes(x=\"\", y=count, fill=group)) +\n",
    "    geom_col(width=1) +\n",
    "    coord_polar(theta=\"y\") + \n",
    "    scale_fill_manual(values=ptpg_class_color) + \n",
    "    ggtitle(\"Ctl (+/+) vs D1 (+/+)\") + \n",
    "    guides(fill=guide_legend(ncol=1, override.aes=list(alpha=1, size=1.5), keywidth=0.25, keyheight=0.25, default.unit=\"cm\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "491e4aa0-c17c-47ce-9bb7-156d240e44b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_4/p_pl_ptg_counts.pdf\", width=5, height=7.5)\n",
    "\n",
    "gridExtra::grid.arrange(\n",
    "    \n",
    "    p_pl_1 %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    p_pl_2 %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    p_pl_3 %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    p_pl_4 %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    p_pl_5 %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    p_pl_6 %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    ncol=2\n",
    "    \n",
    ")\n",
    "\n",
    "dev.off()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R (r.4.4.1)",
   "language": "R",
   "name": "r.4.4.1"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
