{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2922ed7e-69ca-4218-b5e6-774ec9b9fd9f",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Limma - Differential expression analysis (DEA)\n",
    "\n",
    "https://support.bioconductor.org/p/p133524/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9c0771c-10e9-403b-8946-ffe67fcaf8c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "library_load <- suppressMessages(\n",
    "    \n",
    "    suppressWarnings(\n",
    "    \n",
    "            list(\n",
    "                \n",
    "                # Seurat \n",
    "                library(Seurat),\n",
    "\n",
    "                # RNA seq\n",
    "                library(edgeR), \n",
    "                \n",
    "                # GSEA \n",
    "                library(msigdbr), \n",
    "                library(fgsea), \n",
    "\n",
    "                # Data \n",
    "                library(tidyverse), \n",
    "                library(openxlsx), \n",
    "\n",
    "                # Plotting \n",
    "                library(ComplexHeatmap), \n",
    "                library(circlize), \n",
    "                library(viridis), \n",
    "                library(ggplotify), \n",
    "                library(patchwork), \n",
    "                library(RColorBrewer), \n",
    "\n",
    "                # Python \n",
    "                library(reticulate)\n",
    "            \n",
    "            )\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "265bdd64-1810-4aae-b202-857082cd822f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Configure reticulate \n",
    "use_condaenv(condaenv='p.3.9.19-FD20200109SPLENO', conda=\"/nobackup/peer/fdeckert/miniconda3/bin/conda\", required=NULL)\n",
    "py_config()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "825af51a-df61-427b-b6d8-25faf713d620",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)\n",
    "options(dplyr.summarise.inform=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db5a2237-eb11-4713-bc16-2b499d255902",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_seed <- 42\n",
    "set.seed(random_seed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "355e209b-fd26-436e-8c3e-364c3b6ed71d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set working directory to project root\n",
    "setwd(\"/research/peer/fdeckert/FD20200109SPLENO\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "719f92fe-4d71-4584-beea-7b76342a78a3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Source files\n",
    "source(\"plotting_global.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3775cb16-8a5d-497f-95e7-705c21933256",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Plotting Theme\n",
    "ggplot2::theme_set(theme_global_set(size_select=1)) # From project global source()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2279528f-9b9d-47fe-9180-b6f318017f61",
   "metadata": {},
   "source": [
    "# Parameter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbfd3ca6-5875-4d61-8e3b-3eceaad88184",
   "metadata": {},
   "outputs": [],
   "source": [
    "cache_test <- FALSE\n",
    "\n",
    "cnt_min=1\n",
    "cell_min=3\n",
    "\n",
    "adj_pval_thr <- 0.05\n",
    "log2_thr <- 0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3220ee3b-879f-4abd-bf0a-5c8db3e2eafd",
   "metadata": {},
   "source": [
    "# Helper function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "454cd27c-c502-49ce-8955-b747b09604a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_select <- function(so, cnt_min=3, cell_min=1) {\n",
    "    \n",
    "    cnt <- GetAssayData(so, assay=\"RNA\", layer=\"counts\")\n",
    "    cnt <- cnt[rowSums(cnt>=cnt_min)>=cell_min, ]\n",
    "\n",
    "    so <- CreateSeuratObject(counts=cnt, meta.data=so@meta.data)\n",
    "    \n",
    "    return(so)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4986747-78b8-4b01-92cd-585428f8e6cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "limma_fit <- function(so) {\n",
    "\n",
    "    # Get Counts \n",
    "    cnt <- GetAssayData(so, assay=\"RNA\", slot=\"counts\")\n",
    "        \n",
    "    # Prepare count data for split \n",
    "    cnt <- as.matrix(cnt)\n",
    "\n",
    "    # Get grouping variables from cnt matrix\n",
    "    infection <- so[[\"infection\", drop=TRUE]] %>% as.character()\n",
    "    sample_group <- so[[\"sample_group\", drop=TRUE]] %>% as.character()\n",
    "    \n",
    "    # Design matrix \n",
    "    design <- model.matrix(~0+infection)\n",
    "    colnames(design) <- gsub(\"infection\", \"\", colnames(design))\n",
    "    \n",
    "    # Run voomLmFit\n",
    "    fit <- edgeR::voomLmFit(\n",
    "        \n",
    "        counts=cnt, \n",
    "        design=design, \n",
    "        block=NULL, \n",
    "        normalize.method=\"none\", \n",
    "        lib.size=NULL, \n",
    "        sample.weights=FALSE, \n",
    "        plot=FALSE\n",
    "\n",
    "    )\n",
    "    \n",
    "    return(fit)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d0c1762-2694-4c28-9d4f-c298736d766c",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrasts_fit <- function(fit, contrasts_vec) {\n",
    "    \n",
    "\n",
    "    if (all(contrasts_vec %in% colnames(fit[[1]]))) {\n",
    "        \n",
    "        # Contrast fit\n",
    "        contrasts_vec <- paste0(contrasts_vec[1], \"-\", contrasts_vec[2])\n",
    "        \n",
    "        contrasts <- limma::makeContrasts(contrasts=contrasts_vec, levels=colnames(fit[[1]]))\n",
    "        contrasts_fit <- limma::contrasts.fit(fit, contrasts=contrasts)\n",
    "\n",
    "        # eBayes fit \n",
    "        efit <- limma::eBayes(contrasts_fit)\n",
    "\n",
    "        # Get result table\n",
    "        res <- limma::topTable(efit, sort.by=\"P\", n=Inf, p.value=1, lfc=0, coef=1)\n",
    "        res$gene <- rownames(res)\n",
    "        \n",
    "    } else {\n",
    "        \n",
    "        res <- NULL\n",
    "        \n",
    "    }\n",
    "\n",
    "    \n",
    "    return(res)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45ec7f6c-104b-4bf6-bd18-814751b50bd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "vp <- function(dea, log2_thr=0, adj_pval_thr=0.1, top_label=10, title=NULL, color_pos=c(\"pos\"=\"#0000ffff\"), color_neg=c(\"neg\"=\"#fd8008ff\")) {\n",
    "    \n",
    "    # Annotate entries significance by log2_thr and adj_pval_thr\n",
    "    dea$adj.P.Val <- ifelse(dea$adj.P.Val == 0, min(dea$adj.P.Val), dea$adj.P.Val)\n",
    "    dea$sig <- ifelse((abs(dea$logFC) >= log2_thr) & (dea$adj.P.Val <= adj_pval_thr), \"s\", \"ns\")\n",
    "    \n",
    "    \n",
    "    # Set color based on significance and direction of dea e.g. positive and negative \n",
    "    dea$color <- ifelse(dea$sig == \"s\" & dea$logFC > 0, names(color_pos), \"n.s.\")\n",
    "    dea$color <- ifelse(dea$sig == \"s\" & dea$logFC < 0, names(color_neg), dea$color)\n",
    "    \n",
    "    color <- c(color_pos, \"gray\", \"black\", color_neg)\n",
    "    names(color) <- c(names(color_pos), \"n.s.\", \"black\", names(color_neg))\n",
    "    \n",
    "    # Create labels based log2FC and adj.P.Val\n",
    "    dea_pos <- dea[dea$logFC > 0 & dea$sig == \"s\", ]\n",
    "    dea_neg <- dea[dea$logFC < 0 & dea$sig == \"s\", ]\n",
    "\n",
    "    pos_labels_log2FC <- dea_pos[rev(order(dea_pos$logFC)), ][1:top_label, ] %>% rownames()\n",
    "    neg_labels_log2FC <- dea_neg[order(dea_neg$logFC), ][1:top_label, ] %>% rownames()\n",
    "    \n",
    "    pos_labels_adj.P.Val <- dea_pos[order(dea_pos$adj.P.Val), ][1:top_label, ] %>% rownames()\n",
    "    neg_labels_adj.P.Val <- dea_neg[order(dea_neg$adj.P.Val), ][1:top_label, ] %>% rownames()\n",
    "    \n",
    "    pos_labels <- c(pos_labels_log2FC, pos_labels_adj.P.Val)\n",
    "    neg_labels <- c(neg_labels_log2FC, neg_labels_adj.P.Val)\n",
    "    \n",
    "    # Set labels \n",
    "    dea$label <- ifelse(rownames(dea) %in% c(pos_labels, neg_labels), rownames(dea), NA)\n",
    "\n",
    "    # Plot\n",
    "    vp <- ggplot(dea, aes(x=AveExpr, y=logFC, fill=dea$color, label=label), alpha=1) + \n",
    "    \n",
    "        geom_point(size=4, shape=21, color=\"white\") + \n",
    "        geom_hline(aes(yintercept=0), linetype=\"dotted\", colour=\"black\") +\n",
    "        ggrepel::geom_text_repel(segment.color=\"black\", force=20, force_pull=1, max.overlaps=getOption(\"ggrepel.max.overlaps\", default=100), size=5, alpha=1, guide=\"none\", segment.size=0.1, color='black') + \n",
    "        ylim(-max(abs(dea$logFC)), max(abs(dea$logFC))) +  \n",
    "        ggtitle(title) + xlab(\"average expression [ratio]\") + ylab(\"average log2FC\") + \n",
    "        scale_fill_manual(values=color, name=\"DEA\") + \n",
    "    \n",
    "        guides(\n",
    "            \n",
    "            color=guide_legend(order=1, title=\"Group\", size=2, keywidth=0.75, keyheight=0.75), \n",
    "            alpha=\"none\"\n",
    "            \n",
    "        ) + \n",
    "    \n",
    "    theme(\n",
    "        \n",
    "        legend.position=\"right\", \n",
    "        aspect.ratio=1\n",
    "        \n",
    "    )\n",
    "    \n",
    "    return(vp)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "231b18c0-8a8f-4cee-a1ab-aebde4d8a11a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea <- function(res_dea, gene_set) {\n",
    "    \n",
    "    # Set gene name \n",
    "    res_dea$gene_name <- rownames(res_dea)\n",
    "    \n",
    "    # Make ranks \n",
    "    res_dea$adj.P.Val <- ifelse(res_dea$adj.P.Val==0, min(res_dea$adj.P.Val\t[res_dea$adj.P.Val>0]), res_dea$adj.P.Val)\n",
    "    res_dea$sign_log_adj_p_values <- -log10(res_dea$adj.P.Val) * sign(res_dea$logFC)\n",
    "    \n",
    "    ranks <- res_dea$sign_log_adj_p_values\n",
    "    names(ranks) <- res_dea$gene_name\n",
    "    ranks <- ranks[order(ranks)]\n",
    "    ranks <- rev(ranks)\n",
    "    \n",
    "    # Retain only pathways that overlap with dea lsit\n",
    "    gene_set_filter <- lapply(gene_set, function(x) {sum(names(ranks) %in% x)>=1})\n",
    "    gene_set <- gene_set[unlist(gene_set_filter)]\n",
    "\n",
    "    gsea_res <- fgsea(\n",
    "        \n",
    "        pathways=gene_set,\n",
    "        stats=ranks,\n",
    "        nperm=100000, \n",
    "        minSize=5,\n",
    "        maxSize=Inf\n",
    "        \n",
    "    )\n",
    "    \n",
    "    return(gsea_res)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c85fc1c-4240-4294-a496-10a441931d0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_pl <- function(gsea, pval_thr=0.1, title=NULL, size_range=5, pathway_suffix=NULL, top=20) {\n",
    "    \n",
    "    # Set GSEA data frame \n",
    "    gsea <- as.data.frame(gsea)\n",
    "    gsea <- na.omit(gsea) \n",
    "    \n",
    "    # Set color names \n",
    "    color <- color$infection\n",
    "    \n",
    "    # Fix pathway names\n",
    "    if(!is.null(pathway_suffix)) {gsea$pathway <- gsub(pathway_suffix, \"\", gsea$pathway)}\n",
    "    gsea$pathway <- gsub(\"_\", \" \", gsea$pathway)\n",
    "    \n",
    "    # Filter hits \n",
    "    gsea_up <- gsea[sign(gsea$NES)==+1, ]\n",
    "    gsea_down <- gsea[sign(gsea$NES)==-1, ]\n",
    "    \n",
    "    gsea_up <- gsea_up[order(gsea_up$pval), ][1:top, ]\n",
    "    gsea_down <- gsea_down[order(gsea_down$pval), ][1:top, ]\n",
    "    \n",
    "    gsea <- rbind(gsea_up, gsea_down)\n",
    "    gsea <- na.omit(gsea)\n",
    "    gsea <- distinct(gsea)\n",
    "    \n",
    "    # Add color \n",
    "    gsea$color <- ifelse(sign(gsea$ES)==-1, names(color)[1], names(color)[2])\n",
    "    gsea$color <- ifelse(gsea$pval<=pval_thr, gsea$color, NA)\n",
    "\n",
    "    gsea$sign_log_pval_values <- -log10(gsea$pval) * sign(gsea$ES)\n",
    "    \n",
    "    # Short pathway names \n",
    "    gsea <- mutate(gsea, pathway = str_sub(pathway, 1, 40))\n",
    "    \n",
    "    # Order hits \n",
    "    gsea <- gsea[order(gsea$sign_log_pval_values), ]\n",
    "    gsea$pathway <- factor(gsea$pathway, levels=gsea$pathway)\n",
    "\n",
    "    x_max <- max(abs(gsea$sign_log_pval_values)) + 0.1\n",
    "    if(x_max<abs(log10(pval_thr))) {x_max <- abs(log10(pval_thr)) + 0.1}\n",
    "    \n",
    "    # Plot \n",
    "    plot <- ggplot(gsea, aes(x=sign_log_pval_values, y=pathway, color=color)) + \n",
    "        \n",
    "        geom_vline(xintercept=-log10(pval_thr), linetype=\"dashed\") + \n",
    "        geom_vline(xintercept=log10(pval_thr), linetype=\"dashed\") +\n",
    "    \n",
    "        geom_point(aes(size=abs(NES))) +\n",
    "\n",
    "        ggtitle(title) +\n",
    "        xlab(\"Signed -log10 adj. p-value\") + ylab(\"\") + \n",
    "        xlim(-x_max, x_max) + \n",
    "        scale_color_manual(values=color, na.value=\"black\", drop=FALSE) +\n",
    "        scale_size(range=c(0, size_range)) + \n",
    "        guides(\n",
    "            \n",
    "            color=guide_legend(order=1, title=\"Group\", size=5, keywidth=0.75, keyheight=0.75), \n",
    "            size=guide_legend(order=2, title=\"Abs. (NES)\", keywidth=0.75, keyheight=0.75)\n",
    "            \n",
    "        ) +\n",
    "    \n",
    "        theme(\n",
    "            \n",
    "            legend.position=\"right\", \n",
    "            legend.justification=\"top\", \n",
    "            axis.text.y=element_text(hjust=1, vjust=0.5)\n",
    "            \n",
    "        )\n",
    "    \n",
    "    return(plot)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7abff79-2e3f-473c-84de-010bbb929a08",
   "metadata": {},
   "source": [
    "# Run DEA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e714eab-4552-4129-9dc6-17b6dafa4819",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load Seurat object \n",
    "so <- readRDS(\"data/BSA_0355_SM01_10x_SPLENO/ANALYSIS/pp.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f78fcac-c1ca-4c34-800a-093d7558a2e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Order results \n",
    "cell_type_order <- levels(so$leiden_cell_type_main)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6094441a-98d1-4c73-bfd6-2309d74e86b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare data \n",
    "so <- SplitObject(so, split.by=\"leiden_cell_type_main\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d1cc0e3-5c27-44d3-8f2c-8fd18189ae0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Order cell types \n",
    "so <- so[cell_type_order]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19809c58-2922-4148-81dd-cdd247edae42",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset features\n",
    "so <- lapply(so, feature_select, cnt_min=cnt_min, cell_min=cell_min)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63bc332d-1fd2-420a-bb05-470e48e1cd65",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DEA voom fit \n",
    "if(!cache_test) {\n",
    "    \n",
    "    fit <- lapply(so, limma_fit)\n",
    "    saveRDS(fit, \"result/dea/scRNAseq/limma/baseline_vs_cpg.rds\")\n",
    "    \n",
    "} else {\n",
    "    \n",
    "    fit <- readRDS(\"result/dea/scRNAseq/limma/baseline_vs_cpg.rds\")\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1095e5f8-305b-4d9d-8bf5-85bba0365bbd",
   "metadata": {},
   "source": [
    "## Baseline vs CpG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfc3501b-0cbb-4269-adb8-40516cbdb83f",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrasts_vec <- c(\"CpG\", \"Baseline\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c6ef73d-154e-4250-b3e9-f5efd8c9d3c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_dea_1 <- lapply(fit, contrasts_fit, contrasts_vec=contrasts_vec)\n",
    "res_dea_1 <- res_dea_1[!lapply(res_dea_1, is.null) %>% unlist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbe5537a-202a-49b6-9e12-65ad77532bbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "openxlsx::write.xlsx(res_dea_1, paste0(\"result/dea/scRNAseq/limma/baseline_vs_cpg.xlsx\"))\n",
    "saveRDS(res_dea_1, paste0(\"result/dea/scRNAseq/limma/baseline_vs_cpg.rds\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92919e2c-2f29-4e6b-a0bd-a3d739f0dbb2",
   "metadata": {},
   "source": [
    "## Volcano "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8861146-1021-4c72-96f0-a621d04b7cc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=4*6, repr.plot.height=6)\n",
    "\n",
    "vp_1 <- lapply(names(res_dea_1), function(i) vp(res_dea_1[[i]], title=i, log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, color_pos=unlist(color$infection[\"CpG\"]), color_neg=unlist(color$infection[\"Baseline\"])) + theme_global_set(size_select=1))\n",
    "               \n",
    "ggpubr::ggarrange(plotlist=vp_1, ncol=4, common.legend=TRUE, legend=\"bottom\") %>% print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc30258f-5dce-4b08-ab13-6fc7ecc917d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(paste0(\"result/dea/scRNAseq/limma/baseline_vs_cpg.pdf\"), onefile=TRUE, width=6, height=6)\n",
    "\n",
    "for (i in seq(length(vp_1))) {plot(vp_1[[i]])}\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fb4c420-dbb4-48d4-a47b-d8a3278c942f",
   "metadata": {},
   "source": [
    "## Gene set enrichment analysis (Hallmark)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ec604af-4e0b-4f9d-ba99-bb47e1a2ae67",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs_mm <- msigdbr(species=\"Mus musculus\", category=\"H\", subcategory=NULL)\n",
    "gs_mm <- split(x=gs_mm$gene_symbol, f=gs_mm$gs_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86e8c0ec-bf97-48ce-825d-66a7bc6658f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_gsea_1 <- lapply(res_dea_1, function(res_dea) gsea(res_dea=res_dea, gene_set=gs_mm))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "394c8ca6-51ce-459f-b45d-2395260dfe5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=4*7, repr.plot.height=12)\n",
    "\n",
    "gsea_pl_1 <- lapply(names(res_gsea_1), function(i) gsea_pl(res_gsea_1[[i]], pval_thr=0.1, title=i, size_range=5, pathway_suffix=\"HALLMARK\", top=25))\n",
    "                    \n",
    "ggpubr::ggarrange(plotlist=gsea_pl_1, ncol=4, common.legend=TRUE, legend=\"bottom\") %>% print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d2502b5-9764-4b0b-a98b-a80a5d492459",
   "metadata": {},
   "source": [
    "## Gene set enrichment analysis (Hallmark)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e8e9c72-131e-4023-967a-d89f77bb7832",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs_mm <- msigdbr(species=\"Mus musculus\", category=\"C2\", subcategory=\"WIKIPATHWAYS\")\n",
    "gs_mm <- split(x=gs_mm$gene_symbol, f=gs_mm$gs_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c51ad7ae-53a0-44bb-a586-2eac6200c248",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_gsea_1 <- lapply(res_dea_1, function(res_dea) gsea(res_dea=res_dea, gene_set=gs_mm))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60dfd74b-5103-4c6b-bd82-abcddf35b249",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=4*7, repr.plot.height=12)\n",
    "\n",
    "gsea_pl_1 <- lapply(names(res_gsea_1), function(i) gsea_pl(res_gsea_1[[i]], pval_thr=0.1, title=i, size_range=5, pathway_suffix=\"WP\", top=25))\n",
    "                    \n",
    "ggpubr::ggarrange(plotlist=gsea_pl_1, ncol=4, common.legend=TRUE, legend=\"bottom\") %>% print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9d6100a-9c12-4433-9580-101393f96de8",
   "metadata": {},
   "source": [
    "# Session info "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81d55d8b-7de5-4183-8773-cd100506ab69",
   "metadata": {},
   "outputs": [],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "r.4.4.1-FD20200109SPLENO",
   "language": "R",
   "name": "r.4.4.1-fd20200109spleno"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
