{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2922ed7e-69ca-4218-b5e6-774ec9b9fd9f",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Limma - Differential expression analysis (DEA)\n",
    "\n",
    "https://support.bioconductor.org/p/p133524/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b676ad89-2831-4e63-8daa-473c262fc167",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9c0771c-10e9-403b-8946-ffe67fcaf8c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "library_load <- suppressMessages(\n",
    "    \n",
    "    suppressWarnings(\n",
    "    \n",
    "            list(\n",
    "                \n",
    "                # Seurat \n",
    "                library(Seurat), \n",
    "                \n",
    "                # RNA seq\n",
    "                library(edgeR), \n",
    "\n",
    "                # Data \n",
    "                library(tidyverse), \n",
    "                library(openxlsx), \n",
    "\n",
    "                # Plotting \n",
    "                library(ComplexHeatmap), \n",
    "                library(circlize), \n",
    "                library(viridis), \n",
    "                library(ggplotify), \n",
    "                library(patchwork), \n",
    "                library(RColorBrewer), \n",
    "\n",
    "                # Python \n",
    "                library(reticulate)\n",
    "            \n",
    "            )\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "265bdd64-1810-4aae-b202-857082cd822f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Configure reticulate \n",
    "use_condaenv(condaenv='p.3.10.16-FD20200109SPLENO', conda=\"/nobackup/peer/fdeckert/miniconda3/bin/conda\", required=NULL)\n",
    "py_config()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "825af51a-df61-427b-b6d8-25faf713d620",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)\n",
    "options(dplyr.summarise.inform=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db5a2237-eb11-4713-bc16-2b499d255902",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_seed <- 42\n",
    "set.seed(random_seed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "355e209b-fd26-436e-8c3e-364c3b6ed71d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set working directory to project root\n",
    "setwd(\"/research/peer/fdeckert/FD20200109SPLENO\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "719f92fe-4d71-4584-beea-7b76342a78a3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Source files\n",
    "source(\"plotting_global.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3775cb16-8a5d-497f-95e7-705c21933256",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Plotting Theme\n",
    "ggplot2::theme_set(theme_global_set(size_select=1)) # From project global source()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2279528f-9b9d-47fe-9180-b6f318017f61",
   "metadata": {},
   "source": [
    "# Parameter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbfd3ca6-5875-4d61-8e3b-3eceaad88184",
   "metadata": {},
   "outputs": [],
   "source": [
    "pct_min=5\n",
    "\n",
    "cnt_min=3\n",
    "cell_min=3\n",
    "\n",
    "adj_pval_thr <- 0.01\n",
    "log2_thr <- 0.50\n",
    "\n",
    "adj_pval_gsea_thr <- 0.1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77bd9b19-1cef-4e8e-a884-6a748dd48d8c",
   "metadata": {},
   "source": [
    "# Import data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e714eab-4552-4129-9dc6-17b6dafa4819",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load Seurat object \n",
    "so <- readRDS(\"data/scRNAseq/object/pp.rds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3220ee3b-879f-4abd-bf0a-5c8db3e2eafd",
   "metadata": {},
   "source": [
    "# Helper function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "454cd27c-c502-49ce-8955-b747b09604a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_select <- function(so, pct_min=0, cnt_min=3, cell_min=1) {\n",
    "\n",
    "    # Get count matrix \n",
    "    cnt <- GetAssayData(so, assay=\"RNA\", layer=\"counts\")\n",
    "\n",
    "    # Filter by percentage \n",
    "    cnt <- cnt[(100*rowSums(cnt>0)/ncol(cnt))>=pct_min, ]\n",
    "\n",
    "    # Filer by expression \n",
    "    cnt <- cnt[rowSums(cnt>=cnt_min)>=cell_min, ]\n",
    "\n",
    "    # return genes \n",
    "    genes <- rownames(cnt)\n",
    "        \n",
    "    return(genes)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4986747-78b8-4b01-92cd-585428f8e6cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "limma_fit <- function(so, pct_min=0, cnt_min=0, cell_min=0) {\n",
    "\n",
    "    message(paste0(\"Limma fit \", unique(so$celltype_low), \"(\", ncol(so), \")\"))\n",
    "\n",
    "    genes <- lapply(unique(so$sample_group), function(i) {\n",
    "\n",
    "        genes_i <- feature_select(so[, so$sample_group==i], pct_min=pct_min, cnt_min=cnt_min, cell_min=cell_min)\n",
    "        \n",
    "    }\n",
    "          )\n",
    "\n",
    "    genes <- do.call(\"c\", genes) %>% unique()\n",
    "\n",
    "    message(paste0(\"Using \", length(genes), \" genes after filtering\"))\n",
    "    so <- so[genes, ]\n",
    "    \n",
    "    # Get Counts \n",
    "    cnt <- GetAssayData(so, assay=\"RNA\", slot=\"counts\")\n",
    "        \n",
    "    # Prepare count data for split \n",
    "    cnt <- as.matrix(cnt)\n",
    "\n",
    "    # Get grouping variables from cnt matrix\n",
    "    group <- as.character(so[[\"group\", drop=TRUE]])\n",
    "    \n",
    "    # Design matrix \n",
    "    design <- model.matrix(~0+group)\n",
    "    colnames(design) <- gsub(\"group\", \"\", colnames(design))\n",
    "    \n",
    "    # Run voomLmFit\n",
    "    fit <- edgeR::voomLmFit(\n",
    "        \n",
    "        counts=cnt, \n",
    "        design=design, \n",
    "        normalize.method=\"none\", \n",
    "        lib.size=NULL, \n",
    "        sample.weights=FALSE, \n",
    "        plot=FALSE\n",
    "\n",
    "    )\n",
    "    \n",
    "    return(fit)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d0c1762-2694-4c28-9d4f-c298736d766c",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrasts_fit <- function(fit, contrasts_vec) {\n",
    "    \n",
    "\n",
    "    if (all(contrasts_vec %in% colnames(fit[[1]]))) {\n",
    "\n",
    "        # Store cell numbers \n",
    "        n_cells_1 <- sum(fit$design[, contrasts_vec[1]])\n",
    "        n_cells_2 <- sum(fit$design[, contrasts_vec[2]])\n",
    "        \n",
    "        # Contrast fit\n",
    "        contrasts_vec <- paste0(contrasts_vec[1], \"-\", contrasts_vec[2])\n",
    "        \n",
    "        contrasts <- limma::makeContrasts(contrasts=contrasts_vec, levels=colnames(fit[[1]]))\n",
    "        contrasts_fit <- limma::contrasts.fit(fit, contrasts=contrasts)\n",
    "\n",
    "        # eBayes fit \n",
    "        efit <- limma::eBayes(contrasts_fit)\n",
    "\n",
    "        # Get result table\n",
    "        res <- limma::topTable(efit, sort.by=\"P\", n=Inf, p.value=1, lfc=0, coef=1)\n",
    "        res$gene <- rownames(res)\n",
    "\n",
    "        # Add cell number \n",
    "        res$n_cells_1 <- n_cells_1\n",
    "        res$n_cells_2 <- n_cells_2\n",
    "        \n",
    "    } else {\n",
    "        \n",
    "        res <- NULL\n",
    "        \n",
    "    }\n",
    "\n",
    "    return(res)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45ec7f6c-104b-4bf6-bd18-814751b50bd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "vp <- function(dea, log2_thr=0, adj_pval_thr=0.1, top_label=10, title=NULL, color_pos=c(\"pos\"=\"#0000ffff\"), color_neg=c(\"neg\"=\"#fd8008ff\")) {\n",
    "    \n",
    "    # Annotate entries significance by log2_thr and adj_pval_thr\n",
    "    dea$adj.P.Val <- ifelse(dea$adj.P.Val == 0, min(dea$adj.P.Val), dea$adj.P.Val)\n",
    "    dea$sig <- ifelse((abs(dea$logFC) >= log2_thr) & (dea$adj.P.Val <= adj_pval_thr), \"s\", \"ns\")\n",
    "    \n",
    "    # Set color based on significance and direction of dea e.g. positive and negative \n",
    "    dea$color <- ifelse(dea$sig == \"s\" & dea$logFC > 0, names(color_pos), \"n.s.\")\n",
    "    dea$color <- ifelse(dea$sig == \"s\" & dea$logFC < 0, names(color_neg), dea$color)\n",
    "    \n",
    "    color <- c(color_pos, \"gray\", \"black\", color_neg)\n",
    "    names(color) <- c(names(color_pos), \"n.s.\", \"black\", names(color_neg))\n",
    "    \n",
    "    # Create labels based log2FC and adj.P.Val\n",
    "    dea_pos <- dea[dea$logFC > 0 & dea$sig == \"s\", ]\n",
    "    dea_neg <- dea[dea$logFC < 0 & dea$sig == \"s\", ]\n",
    "\n",
    "    pos_labels_log2FC <- dea_pos[rev(order(dea_pos$logFC)), ][1:top_label, ] %>% rownames()\n",
    "    neg_labels_log2FC <- dea_neg[order(dea_neg$logFC), ][1:top_label, ] %>% rownames()\n",
    "    \n",
    "    pos_labels_adj.P.Val <- dea_pos[order(dea_pos$adj.P.Val), ][1:top_label, ] %>% rownames()\n",
    "    neg_labels_adj.P.Val <- dea_neg[order(dea_neg$adj.P.Val), ][1:top_label, ] %>% rownames()\n",
    "    \n",
    "    pos_labels <- c(pos_labels_log2FC, pos_labels_adj.P.Val)\n",
    "    neg_labels <- c(neg_labels_log2FC, neg_labels_adj.P.Val)\n",
    "    \n",
    "    # Set labels \n",
    "    dea$label <- ifelse(rownames(dea) %in% c(pos_labels, neg_labels), rownames(dea), NA)\n",
    "\n",
    "    # Plot\n",
    "    vp <- ggplot(dea, aes(x=AveExpr, y=logFC, fill=dea$color, label=label), alpha=1) + \n",
    "    \n",
    "        geom_point(size=4, shape=21, color=\"white\") + \n",
    "        geom_hline(aes(yintercept=-log2_thr), linetype=\"dotted\", colour=\"black\") +\n",
    "        geom_hline(aes(yintercept=+log2_thr), linetype=\"dotted\", colour=\"black\") +\n",
    "        ggrepel::geom_text_repel(segment.color=\"black\", force=20, force_pull=1, max.overlaps=getOption(\"ggrepel.max.overlaps\", default=100), size=5, alpha=1, guide=\"none\", segment.size=0.1, color='black') + \n",
    "        # ylim(-max(abs(dea$logFC)), max(abs(dea$logFC))) +  \n",
    "        scale_y_continuous(limits=c(-2, +2)) + \n",
    "        ggtitle(title) + xlab(\"average expression\") + ylab(\"average log2FC\") + \n",
    "        scale_fill_manual(values=c(color_pos, color_neg), name=\"DEA\") + \n",
    "    \n",
    "        guides(\n",
    "            \n",
    "            color=guide_legend(order=1, title=\"Group\", size=2, keywidth=0.75, keyheight=0.75), \n",
    "            alpha=\"none\"\n",
    "            \n",
    "        ) + \n",
    "    \n",
    "    theme(\n",
    "        \n",
    "        legend.position=\"right\", \n",
    "        aspect.ratio=1\n",
    "        \n",
    "    ) +\n",
    "    \n",
    "    annotate(\"text\", x=Inf, y=Inf, label=paste(\"n =\", dea$n_cells_1[1]), hjust=1.1, vjust=1.5, size=6) +\n",
    "    annotate(\"text\", x=Inf, y=-Inf, label=paste(\"n =\", dea$n_cells_2[1]), hjust=1.1, vjust=-0.5, size=6)\n",
    "    \n",
    "    return(vp)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "020f79df-48ff-44fe-ac9b-717db76ee90e",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea <- function(result, category=\"H\", subcategory=NULL) {\n",
    "    \n",
    "    # Set mgi symbols\n",
    "    result$mgi_symbol <- rownames(result)\n",
    "    \n",
    "    # Get gene set\n",
    "    gene_set <- msigdbr::msigdbr(species=\"Mus musculus\", category=category, subcategory=subcategory)\n",
    "    gene_set <- split(gene_set, x=gene_set$gene_symbol, f=gene_set$gs_name)\n",
    "    \n",
    "    # Set gene names \n",
    "    result <- na.omit(result)\n",
    "    \n",
    "    # Make ranks \n",
    "    result$P.Value <- ifelse(result$P.Value==0, min(result$P.Value[result$P.Value>0]), result$P.Value)\n",
    "    result$sign_log_adj_p_values <- -log10(result$P.Value) * sign(result$logFC)\n",
    "    \n",
    "    ranks <- result$sign_log_adj_p_values\n",
    "    names(ranks) <- result$mgi_symbol\n",
    "    ranks <- ranks[order(ranks)]\n",
    "    ranks <- rev(ranks)\n",
    "    \n",
    "    # Retain only pathways that overlap with result lsit\n",
    "    gene_set_filter <- lapply(gene_set, function(x) {sum(result$mgi_symbol %in% x)>=1})\n",
    "    gene_set <- gene_set[unlist(gene_set_filter)]\n",
    "\n",
    "    gsea <- fgsea::fgsea(\n",
    "        \n",
    "        pathways=gene_set,\n",
    "        stats=ranks,\n",
    "        nperm=100000, \n",
    "        minSize=5,\n",
    "        maxSize=500\n",
    "        \n",
    "    )\n",
    "    \n",
    "    return(gsea)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adca4566-5e5f-4d75-9980-9156a7f2420c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_plot <- function(gsea, adj_pval_thr=0.1, title=NULL, color=c(RColorBrewer::brewer.pal(8, \"Set1\")[1], RColorBrewer::brewer.pal(8, \"Set1\")[2]), color_names=c(\"Pos\", \"Neg\"), size_range=5, pathway_suffix=NULL, top=20) {\n",
    "    \n",
    "    # Set GSEA data frame \n",
    "    gsea <- as.data.frame(gsea)\n",
    "    gsea <- na.omit(gsea) \n",
    "    \n",
    "    # Set color names \n",
    "    color <- setNames(color, color_names)\n",
    "    \n",
    "    # Fix pathway names\n",
    "    if(!is.null(pathway_suffix)) {gsea$pathway <- gsub(pathway_suffix, \"\", gsea$pathway)}\n",
    "    gsea$pathway <- gsub(\"_\", \" \", gsea$pathway)\n",
    "    \n",
    "    # Filter hits \n",
    "    gsea_up <- gsea[sign(gsea$NES)==+1, ]\n",
    "    gsea_down <- gsea[sign(gsea$NES)==-1, ]\n",
    "    \n",
    "    gsea_up <- gsea_up[order(gsea_up$pval), ][1:top, ]\n",
    "    gsea_down <- gsea_down[order(gsea_down$pval), ][1:top, ]\n",
    "    \n",
    "    gsea <- rbind(gsea_up, gsea_down)\n",
    "    gsea <- na.omit(gsea)\n",
    "    gsea <- distinct(gsea)\n",
    "    \n",
    "    # Add color \n",
    "    gsea$color <- ifelse(sign(gsea$ES)==1, names(color)[1], names(color)[2])\n",
    "    gsea$color <- ifelse(gsea$padj<=adj_pval_thr, gsea$color, NA)\n",
    "\n",
    "    gsea$sign_log_pval_values <- -log10(gsea$padj) * sign(gsea$ES)\n",
    "    \n",
    "    # Order hits \n",
    "    gsea <- gsea[rev(order(gsea$sign_log_pval_values)), ]\n",
    "    gsea$pathway <- factor(gsea$pathway, levels=rev(gsea$pathway))\n",
    "\n",
    "    x_max <- max(abs(gsea$sign_log_pval_values))\n",
    "    if(x_max<abs(log10(adj_pval_thr))) {x_max <- abs(log10(adj_pval_thr))}\n",
    "    x_max <- ceiling(ceiling(x_max))\n",
    "    \n",
    "    # Plot \n",
    "    plot <- ggplot(gsea, aes(x=sign_log_pval_values, y=pathway, color=color)) + \n",
    "        \n",
    "        geom_vline(xintercept=-log10(adj_pval_thr), linetype=\"dashed\") + \n",
    "        geom_vline(xintercept=log10(adj_pval_thr), linetype=\"dashed\") +\n",
    "    \n",
    "        geom_point(aes(size=abs(NES))) +\n",
    "\n",
    "        ggtitle(title) +\n",
    "        xlab(\"Signed -log10 adj. p-value\") + ylab(\"\") + \n",
    "        scale_x_continuous(breaks=c(-x_max, 0, x_max), limits=c(-x_max, x_max)) +\n",
    "        scale_color_manual(values=color, na.value=\"black\", drop=FALSE) +\n",
    "        scale_size(range=c(0, size_range)) + \n",
    "        guides(\n",
    "            \n",
    "            color=guide_legend(order=1, title=\"Agent\", keywidth=0.75, keyheight=0.75, override.aes=list(size=4)), \n",
    "            size=guide_legend(order=2, title=\"Abs. (NES)\", keywidth=0.75, keyheight=0.75)\n",
    "            \n",
    "        ) +\n",
    "    \n",
    "        theme(\n",
    "            \n",
    "            legend.position=\"bottom\", \n",
    "            legend.justification=\"top\", \n",
    "            axis.text.y=element_text(size=14, hjust=1, vjust=0.5, face=\"plain\", margin=margin(t=0, r=2, b=0, l=0), color=\"black\")\n",
    "            \n",
    "        ) \n",
    "    \n",
    "    return(plot)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7abff79-2e3f-473c-84de-010bbb929a08",
   "metadata": {},
   "source": [
    "# Run DEA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "692880b5-0ec9-4076-9662-2233fbecca3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set group variable\n",
    "so$group <- so$sample_group"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c25d1d88-f9b4-4073-8fd2-117be8b0dfa6",
   "metadata": {},
   "source": [
    "# Fit BSF model "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d84c5fce-1547-42aa-9152-06481cbc7b32",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_0 <- subset(so, subset=facility==\"BSF\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6094441a-98d1-4c73-bfd6-2309d74e86b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare data \n",
    "so_0 <- SplitObject(so_0, split.by=\"celltype_low\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b02887ee-50bf-4fe4-b94f-d6e84d3f6688",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Order \n",
    "celltype_order <- names(color$celltype_low)[names(color$celltype_low) %in% names(so_0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d1cc0e3-5c27-44d3-8f2c-8fd18189ae0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Order cell types \n",
    "so_0 <- so_0[celltype_order]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63bc332d-1fd2-420a-bb05-470e48e1cd65",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DEA voom fit \n",
    "if(!cache_fit) {\n",
    "\n",
    "    fit <- lapply(so_0, function(x) limma_fit(x, pct_min=pct_min, cnt_min=cnt_min, cell_min=cell_min))\n",
    "    saveRDS(fit, \"result/dea/scRNAseq/limma/fit_0.rds\")\n",
    "    \n",
    "} else {\n",
    "    \n",
    "    fit <- readRDS(\"result/dea/scRNAseq/limma/fit_0.rds\")\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1095e5f8-305b-4d9d-8bf5-85bba0365bbd",
   "metadata": {},
   "source": [
    "# Bl6 NaCl D6 vs Bl6 CpG D6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfc3501b-0cbb-4269-adb8-40516cbdb83f",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrasts_vec <- c(\"Bl6_CpG_D6\", \"Bl6_NaCl_D6\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c6ef73d-154e-4250-b3e9-f5efd8c9d3c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_1 <- lapply(fit, contrasts_fit, contrasts_vec=contrasts_vec)\n",
    "res_1 <- res_1[!lapply(res_1, is.null) %>% unlist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8861146-1021-4c72-96f0-a621d04b7cc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=6*6, repr.plot.height=6)\n",
    "\n",
    "vp_1 <- lapply(names(res_1), function(i) vp(res_1[[i]], title=i, log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, color_pos=unlist(color$infection[\"CpG\"]), color_neg=unlist(color$infection[\"NaCl\"])) + theme_global_set(size_select=1))\n",
    "ggpubr::ggarrange(plotlist=vp_1, ncol=6, common.legend=TRUE, legend=\"bottom\") %>% print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc30258f-5dce-4b08-ab13-6fc7ecc917d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(paste0(\"result/dea/scRNAseq/limma/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".pdf\"), onefile=TRUE, width=6, height=6)\n",
    "for (i in seq(length(vp_1))) {plot(vp_1[[i]])}\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbe5537a-202a-49b6-9e12-65ad77532bbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(res_1) <- make.names(names(res_1))\n",
    "openxlsx::write.xlsx(res_1, paste0(\"result/dea/scRNAseq/limma/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".xlsx\"))\n",
    "saveRDS(res_1, paste0(\"result/dea/scRNAseq/limma/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2af83ace-3377-4baa-8f80-cbcea6e935db",
   "metadata": {},
   "source": [
    "# Fit VBC model "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87b8e072-aea9-4559-860e-25197e553ddf",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_1 <- subset(so, subset=facility==\"VBC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "249532d3-9ba8-405b-95dc-27e551b2f25a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare data \n",
    "so_1 <- SplitObject(so_1, split.by=\"celltype_low\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f33c744-5f1d-4782-8422-1fed1c68ffcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Order \n",
    "celltype_order <- names(color$celltype_low)[names(color$celltype_low) %in% names(so_1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4fe5445-d8dc-4d41-8fa1-80ed2fbd6153",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Order cell types \n",
    "so_1 <- so_1[celltype_order]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd2ea8d5-3c00-44ed-9c47-7514c393d743",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DEA voom fit \n",
    "if(!cache_fit) {\n",
    "\n",
    "    fit <- lapply(so_1, function(x) limma_fit(x, pct_min=pct_min, cnt_min=cnt_min, cell_min=cell_min))\n",
    "    saveRDS(fit, \"result/dea/scRNAseq/limma/fit_1.rds\")\n",
    "    \n",
    "} else {\n",
    "    \n",
    "    fit <- readRDS(\"result/dea/scRNAseq/limma/fit_1.rds\")\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2f028a7-ad39-411b-be83-7be479c71b69",
   "metadata": {},
   "source": [
    "# IFNAR fl/fl LysM cre baseline vs IFNAR fl/fl baseline "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c3e98d3-3edc-46a8-af9e-628c6910e655",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrasts_vec <- c(\"IFNAR_fl_LysM_cre_Baseline_D0\", \"IFNAR_fl_Baseline_D0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f5ba83c-27fd-44c3-a87c-6550623571e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_1 <- lapply(fit, contrasts_fit, contrasts_vec=contrasts_vec)\n",
    "res_1 <- res_1[!lapply(res_1, is.null) %>% unlist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "156cbaa2-11e4-4b9a-a041-1f267fdadbd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e000cff-c28e-4824-a70e-ff4c32d6b3e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=6*6, repr.plot.height=6)\n",
    "\n",
    "vp_1 <- lapply(names(res_1), function(i) vp(res_1[[i]], title=i, log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, color_pos=unlist(color$genotype[\"IFNAR_fl_LysM_cre\"]), color_neg=unlist(color$genotype[\"IFNAR_fl\"])) + theme_global_set(size_select=1))\n",
    "ggpubr::ggarrange(plotlist=vp_1, ncol=4, common.legend=TRUE, legend=\"bottom\") %>% print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37b5731e-0054-4ebc-99ff-08481846c0fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "vp_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ac4a0fd-c4fb-4059-ac65-d5409d099dfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(paste0(\"result/dea/scRNAseq/limma/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".pdf\"), onefile=TRUE, width=6, height=6)\n",
    "for (i in seq(length(vp_1))) {plot(vp_1[[i]])}\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "362d9248-3ba3-4d2e-b1e2-e185a4194153",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(res_1) <- make.names(names(res_1))\n",
    "openxlsx::write.xlsx(res_1, paste0(\"result/dea/scRNAseq/limma/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".xlsx\"))\n",
    "saveRDS(res_1, paste0(\"result/dea/scRNAseq/limma/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c07ea381-570e-4abf-9693-f89660f37e74",
   "metadata": {},
   "source": [
    "# IFNAR fl/fl LysM cre D1 vs IFNAR fl/fl D1 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08f0006b-3130-4f97-be62-7fab315cd5ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrasts_vec <- c(\"IFNAR_fl_LysM_cre_CpG_D1\", \"IFNAR_fl_CpG_D1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a5b08f6-68fc-4146-ad1d-a367cce91229",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_1 <- lapply(fit, contrasts_fit, contrasts_vec=contrasts_vec)\n",
    "res_1 <- res_1[!lapply(res_1, is.null) %>% unlist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de93abeb-5a60-41c1-9848-ddf1a0ecf0cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=6*6, repr.plot.height=6)\n",
    "\n",
    "vp_1 <- lapply(names(res_1), function(i) vp(res_1[[i]], title=i, log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, color_pos=unlist(color$genotype[\"IFNAR_fl_LysM_cre\"]), color_neg=unlist(color$genotype[\"IFNAR_fl\"])) + theme_global_set(size_select=1))\n",
    "ggpubr::ggarrange(plotlist=vp_1, ncol=4, common.legend=TRUE, legend=\"bottom\") %>% print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd924fba-73c5-406f-9d22-087da6780e7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(paste0(\"result/dea/scRNAseq/limma/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".pdf\"), onefile=TRUE, width=6, height=6)\n",
    "for (i in seq(length(vp_1))) {plot(vp_1[[i]])}\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c645338b-1635-4e60-b4d9-b2638d1ecd6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(res_1) <- make.names(names(res_1))\n",
    "openxlsx::write.xlsx(res_1, paste0(\"result/dea/scRNAseq/limma/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".xlsx\"))\n",
    "saveRDS(res_1, paste0(\"result/dea/scRNAseq/limma/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e4d3fbf-3253-4e44-a73e-19c3b5c0d37a",
   "metadata": {},
   "source": [
    "# IFNAR fl/fl LysM cre D3 vs IFNAR fl/fl D3 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bbb133d-c1b4-424b-90d8-3939d55d20d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrasts_vec <- c(\"IFNAR_fl_LysM_cre_CpG_D3\", \"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33b5a579-9a5b-4f30-bc83-e1a2a80ac4b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_1 <- lapply(fit, contrasts_fit, contrasts_vec=contrasts_vec)\n",
    "res_1 <- res_1[!lapply(res_1, is.null) %>% unlist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6add059-07a5-4034-be19-9c31a45b3b58",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=6*6, repr.plot.height=6)\n",
    "\n",
    "vp_1 <- lapply(names(res_1), function(i) vp(res_1[[i]], title=i, log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, color_pos=unlist(color$genotype[\"IFNAR_fl_LysM_cre\"]), color_neg=unlist(color$genotype[\"IFNAR_fl\"])) + theme_global_set(size_select=1))\n",
    "ggpubr::ggarrange(plotlist=vp_1, ncol=4, common.legend=TRUE, legend=\"bottom\") %>% print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f998abd1-9e8a-4a81-a1dd-dcb503ee005e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(paste0(\"result/dea/scRNAseq/limma/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".pdf\"), onefile=TRUE, width=6, height=6)\n",
    "for (i in seq(length(vp_1))) {plot(vp_1[[i]])}\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e4f0432-b3aa-4f7a-8daa-fd10de496a81",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(res_1) <- make.names(names(res_1))\n",
    "openxlsx::write.xlsx(res_1, paste0(\"result/dea/scRNAseq/limma/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".xlsx\"))\n",
    "saveRDS(res_1, paste0(\"result/dea/scRNAseq/limma/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9d6100a-9c12-4433-9580-101393f96de8",
   "metadata": {},
   "source": [
    "# Session info "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81d55d8b-7de5-4183-8773-cd100506ab69",
   "metadata": {},
   "outputs": [],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "r.4.4.1-FD20200109SPLENO",
   "language": "R",
   "name": "r.4.4.1-fd20200109spleno"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
