{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "19528665-4058-447f-b1ba-fc7416f49216",
   "metadata": {},
   "source": [
    "# Figure panel 1 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9e813b01-e528-49c6-80a4-193b83bfec4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "40017520-c2ab-41fa-8ab0-2b39416c9a43",
   "metadata": {},
   "outputs": [],
   "source": [
    "library_load <- suppressMessages(\n",
    "    \n",
    "    suppressWarnings(\n",
    "        \n",
    "        list(\n",
    "        \n",
    "            # Seurat \n",
    "            library(Seurat), \n",
    "\n",
    "            # Data \n",
    "            library(tidyverse), \n",
    "            library(data.table), \n",
    "\n",
    "            # miloR\n",
    "            library(miloR), \n",
    "            library(ggbeeswarm), \n",
    "\n",
    "            # Plotting \n",
    "            library(ggplot2), \n",
    "            library(patchwork), \n",
    "            library(cowplot), \n",
    "            library(ComplexHeatmap), \n",
    "\n",
    "            # Pyhton compatibility\n",
    "            library(reticulate)\n",
    "\n",
    "        )\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d766baa7-9020-46b5-a0b6-2e12ed509509",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Configure reticulate \n",
    "# use_condaenv(condaenv=\"p.3.10.16-FD20200109SPLENO\", conda=\"/nobackup/peer/fdeckert/miniconda3/bin/conda\", required=NULL)\n",
    "# py_config()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "24b2e988-79fc-4654-9702-4b7873d83abc",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_seed <- 42\n",
    "set.seed(random_seed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bfcb3262-d1b1-4421-b8a8-f29a4a15f30a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set working directory to project root\n",
    "setwd(\"/research/peer/fdeckert/FD20200109SPLENO\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "67ffec3f-08dc-4142-bdfc-c336d3290cd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting Theme\n",
    "source(\"plotting_global.R\")\n",
    "ggplot2::theme_set(theme_global_set(size_select=4)) # From project global source()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9a22187f-3bcf-43bb-842b-c3eb4eed0384",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Helper functions\n",
    "source(\"bin/so_pl.R\")\n",
    "\n",
    "source(\"bin/dea_pp.R\")\n",
    "source(\"bin/dea_pl.R\")\n",
    "\n",
    "source(\"bin/gsea_pp.R\")\n",
    "source(\"bin/gsea_pl.R\")\n",
    "\n",
    "source(\"bin/dp_feature.R\")\n",
    "source(\"bin/grn_pl.R\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "478b78a4-7a0a-4085-adb0-88038bef97a9",
   "metadata": {},
   "source": [
    "# Import data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "905cc467-30bb-494b-a9bc-b367c27cea1a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Normalizing layer: counts\n",
      "\n"
     ]
    }
   ],
   "source": [
    "so <- readRDS(\"data/scRNAseq/object/pp_0.rds\")\n",
    "so <- NormalizeData(so)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a6659edf-0dac-4c1d-8065-5a8e0743613d",
   "metadata": {},
   "outputs": [],
   "source": [
    "so$sample_group <- factor(so$sample_group, levels=c(\"Bl6_NaCl_D6\", \"Bl6_CpG_D6\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e6bbc96-17af-4d6c-b713-c8213c32d0ea",
   "metadata": {},
   "source": [
    "# Compute percentage primitive and definitive hemoglobin "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8a304a8b-3bed-4aa2-8436-20bd9835364d",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- PercentageFeatureSet(\n",
    "    so,\n",
    "    features = c(\"Hba-x\", \"Hbb-y\", \"Hbb-bh1\"),\n",
    "    col.name=\"pct_counts_primitive_HB\", \n",
    "    assay = \"RNA\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "170602cc-7019-4b3a-9cdc-46dc74e2f112",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- PercentageFeatureSet(\n",
    "    so,\n",
    "    features = c(\"Hba-a1\", \"Hba-a2\", \"Hbb-bt\", \"Hbb-bs\"),\n",
    "    col.name=\"pct_counts_definitive_HB\", \n",
    "    assay = \"RNA\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1d7812a-cf4d-4b15-9ac5-77448a8f2c18",
   "metadata": {},
   "source": [
    "# UMAP leiden celltype low"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "877997cd-7572-4d41-a57e-146814f28507",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mScale for \u001b[32mfill\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mfill\u001b[39m, which will replace the existing scale.\n"
     ]
    }
   ],
   "source": [
    "dp_1 <- dplot(so, group_by=\"celltype_low\", alpha=1.0, pt_size=0.25, shuffle=TRUE, size_select=4, label=TRUE, label_box=TRUE, label_size=6) + \n",
    "\n",
    "    scale_color_manual(values=color$celltype_low) + \n",
    "    scale_fill_manual(values=color$celltype_low) + \n",
    "    guides(color=guide_legend(ncol=1, override.aes=list(alpha=1, size=1.5), keywidth=0, keyheight=0.25, default.unit=\"cm\")) + \n",
    "    theme(\n",
    "    \n",
    "        legend.text.align=0\n",
    "    \n",
    "    ) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e4b340b0-9fc7-434b-9648-15d6a6865227",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_1/umap_celltype_low.pdf\", width=3.5, height=2.5)\n",
    "\n",
    "dp_1\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ad0b6e0-1d89-43c0-a711-a2d7fb946950",
   "metadata": {},
   "source": [
    "# UMAP FACS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0d2d850a-7370-442b-afed-a90787a087e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "so$facs <- factor(so$facs, levels=c(\"Progenitor\", \"Myeloid\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7fa005cf-212b-476c-9f64-e95050e31140",
   "metadata": {},
   "outputs": [],
   "source": [
    "dp_1 <- dplot(so, group_by=\"facs\", alpha=1.0, pt_size=0.25, shuffle=FALSE, size_select=4, label=FALSE, label_box=FALSE, label_size=6) + \n",
    "\n",
    "    scale_color_manual(values=color$facs) + \n",
    "    scale_fill_manual(values=color$facs) + \n",
    "    guides(color=guide_legend(ncol=1, override.aes=list(alpha=1, size=1.5), keywidth=0, keyheight=0.25, default.unit=\"cm\")) + \n",
    "    theme(\n",
    "    \n",
    "        legend.text.align=0\n",
    "    \n",
    "    ) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "1b052d21-4458-40c0-87ed-16926154cce4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_1/umap_facs.pdf\", width=3.5, height=2.5)\n",
    "\n",
    "gridExtra::grid.arrange(\n",
    "    \n",
    "    dp_1 %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\"))\n",
    "\n",
    ")\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27116752-a42c-4c6b-94e0-3160a32935a1",
   "metadata": {},
   "source": [
    "# UMAP Ctr vs CpG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1d06f8f2-65ab-4cc5-b26d-3562d2ed21ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_tmp <- so"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d59748a6-5c60-4ada-8086-38f3dc77aeac",
   "metadata": {},
   "outputs": [],
   "source": [
    "color_tmp <- color$sample_group\n",
    "color_tmp[\"Bl6_CpG_D6\"] <- \"#7F7F7F\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "9f8e9e00-6635-42e3-9b03-19c2b18c4a50",
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- cbind(so_tmp[[\"umap\"]]@cell.embeddings, so_tmp@meta.data)\n",
    "data <- data[data$sample_group==\"Bl6_NaCl_D6\", ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "9f7847fb-234c-4ae4-a349-f13e66221ab8",
   "metadata": {},
   "outputs": [],
   "source": [
    "dp_1 <- dplot(so_tmp, group_by=\"sample_group\", alpha=1, pt_size=0.25, shuffle=FALSE, size_select=4, order=c(\"Bl6_NaCl_D6\", \"Bl6_CpG_D6\")) + \n",
    "    scale_color_manual(values=color_tmp) + \n",
    "    guides(color=guide_legend(ncol=1, override.aes=list(alpha=1, size=1.5), keywidth=0, keyheight=0.25, default.unit=\"cm\")) + \n",
    "    stat_density_2d(data=data, aes(x=UMAP_1, y=UMAP_2), color=\"black\", size=0.2, adjust=0.5, n=100, contour=TRUE) + \n",
    "    theme(legend.text.align=0) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b6cbdad9-6952-46fd-bb4f-8d04552693fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_1/umap_wt_nacl_d6.pdf\", width=3.5, height=2.5)\n",
    "\n",
    "dp_1\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "219c3b2d-0c0f-49a8-acc1-a1f3d6e47b38",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_tmp <- so"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "dacd7c53-9bd5-4041-bdfb-db64815f4741",
   "metadata": {},
   "outputs": [],
   "source": [
    "color_tmp <- color$sample_group\n",
    "color_tmp[\"Bl6_NaCl_D6\"] <- \"#7F7F7F\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f9cccea6-7e71-43ea-af48-20d922b3d997",
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- cbind(so_tmp[[\"umap\"]]@cell.embeddings, so_tmp@meta.data)\n",
    "data <- data[data$sample_group==\"Bl6_CpG_D6\", ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8f92c3c8-fcdc-43b7-99ff-f0b4b087b67c",
   "metadata": {},
   "outputs": [],
   "source": [
    "dp_1 <- dplot(so_tmp, group_by=\"sample_group\", alpha=1, pt_size=0.25, shuffle=FALSE, size_select=4, order=c(\"Bl6_CpG_D6\", \"Bl6_NaCl_D6\")) + \n",
    "    scale_color_manual(values=color_tmp) + \n",
    "    guides(color=guide_legend(ncol=1, override.aes=list(alpha=1, size=1.5), keywidth=0, keyheight=0.25, default.unit=\"cm\")) + \n",
    "    stat_density_2d(data=data, aes(x=UMAP_1, y=UMAP_2), color=\"black\", size=0.2, adjust=0.5, n=100, contour=TRUE) + \n",
    "    theme(legend.text.align=0) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "81dccefb-e443-4344-8639-3cc79db8797f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_1/umap_wt_cpg_d6.pdf\", width=3.5, height=2.5)\n",
    "\n",
    "dp_1\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfbeca61-ece9-48c7-89ea-09a768815112",
   "metadata": {},
   "source": [
    "# Barplot FACS ratio per celltype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "a945f2b9-619a-4f95-9690-c33d9a84ebec",
   "metadata": {},
   "outputs": [],
   "source": [
    "bp_1 <- ggplot(so@meta.data %>% dplyr::mutate(celltype_low=factor(celltype_low, levels=rev(levels(celltype_low)))), aes(x=celltype_low, fill=facs)) + \n",
    "    geom_bar(position=\"fill\", width=0.8, color=\"black\", size=0.1) + \n",
    "    scale_fill_manual(values=color$facs) + \n",
    "    ggtitle(\"\") + xlab(\"\") + ylab(\"Celltype [ratio]\") + \n",
    "    coord_flip() +\n",
    "    theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) + \n",
    "    guides(fill=guide_legend(ncol=1, override.aes=list(alpha=1, size=1.5), keywidth=0.25, keyheight=0.25, default.unit=\"cm\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "49119f92-35b9-4360-b4bf-d69f31b504d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_1/bp_facs_celltype_low.pdf\", width=2.5, height=2.5)\n",
    "\n",
    "bp_1\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7d05579-c782-422a-92fd-8cd13999587c",
   "metadata": {},
   "source": [
    "# Barplot celltype ratio  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "52b41989-7ef0-4e64-8aea-845164a8e0a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "so$celltype_high <- ifelse(so$celltype_low %in% c(\"MEP\", \"Proerythroblast\", \"Erythroblast\"), \"E\", ifelse(so$celltype_low %in% c(\"MegP\", \"GMP\", \"NeuP\", \"BasoP\", \"MastP\"), \"P\", \"M\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "eee1d6a5-719c-41df-a640-bdd73439d088",
   "metadata": {},
   "outputs": [],
   "source": [
    "so$sample_group_rep <- factor(so$sample_group_rep, levels=c(\"Bl6_NaCl_D6_Rep1\", \"Bl6_NaCl_D6_Rep2\", \"Bl6_CpG_D6_Rep1\",  \"Bl6_CpG_D6_Rep2\" ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "f53273ca-def3-4330-8e92-0cd858be4097",
   "metadata": {},
   "outputs": [],
   "source": [
    "bp_1 <- ggplot(so@meta.data[so$celltype_high==\"E\", ], aes(x=sample_group_rep, fill=celltype_low)) + \n",
    "    geom_bar(position=\"fill\", width=0.8, color=\"black\", size=0.1) + \n",
    "    scale_fill_manual(values=color$celltype_low) + \n",
    "    ggtitle(\"\") + xlab(\"\") + ylab(\"Celltype [ratio]\") + \n",
    "    facet_wrap(~celltype_high, scales=\"free\", ncol=1) + \n",
    "    scale_x_discrete(labels=c(\"Rep1\", \"Rep2\", \"Rep1\", \"Rep2\")) + \n",
    "    theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) + \n",
    "    guides(fill=guide_legend(ncol=1, override.aes=list(alpha=1, size=1.5), keywidth=0.25, keyheight=0.25, default.unit=\"cm\"))\n",
    "\n",
    "bp_2 <- ggplot(so@meta.data[so$celltype_high==\"P\", ], aes(x=sample_group_rep, fill=celltype_low)) + \n",
    "    geom_bar(position=\"fill\", width=0.8, color=\"black\", size=0.1) + \n",
    "    scale_fill_manual(values=color$celltype_low) + \n",
    "    ggtitle(\"\") + xlab(\"\") + ylab(\"Celltype [ratio]\") + \n",
    "    facet_wrap(~celltype_high, scales=\"free\", ncol=1) + \n",
    "    scale_x_discrete(labels=c(\"Rep1\", \"Rep2\", \"Rep1\", \"Rep2\")) +\n",
    "    theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) + \n",
    "    guides(fill=guide_legend(ncol=1, override.aes=list(alpha=1, size=1.5), keywidth=0.25, keyheight=0.25, default.unit=\"cm\"))\n",
    "\n",
    "bp_3 <- ggplot(so@meta.data[so$celltype_high==\"M\", ], aes(x=sample_group_rep, fill=celltype_low)) + \n",
    "    geom_bar(position=\"fill\", width=0.8, color=\"black\", size=0.1) + \n",
    "    scale_fill_manual(values=color$celltype_low) + \n",
    "    ggtitle(\"\") + xlab(\"\") + ylab(\"Celltype [ratio]\") + \n",
    "    facet_wrap(~celltype_high, scales=\"free\", ncol=1) + \n",
    "    scale_x_discrete(labels=c(\"Rep1\", \"Rep2\", \"Rep1\", \"Rep2\")) +\n",
    "    theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) + \n",
    "    guides(fill=guide_legend(ncol=1, override.aes=list(alpha=1, size=1.5), keywidth=0.25, keyheight=0.25, default.unit=\"cm\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "91f654be-88e5-4f17-b279-0d4b58529be2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_1/bp_celltype_low.pdf\", width=2, height=4)\n",
    "\n",
    "gridExtra::grid.arrange(\n",
    "\n",
    "    bp_1 %>% egg::set_panel_size(., width=unit(4*0.3, \"cm\"), height=unit(2.0, \"cm\")),\n",
    "    bp_2 %>% egg::set_panel_size(., width=unit(4*0.3, \"cm\"), height=unit(2.0, \"cm\")),\n",
    "    bp_3 %>% egg::set_panel_size(., width=unit(4*0.3, \"cm\"), height=unit(2.0, \"cm\")), ncol=1\n",
    "        \n",
    ")\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "768b200d-54da-4fb8-b861-15367f853bf7",
   "metadata": {},
   "source": [
    "# Celltype marker genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "795724a6-7679-4c3c-b5a3-05ed2d56aab2",
   "metadata": {},
   "outputs": [],
   "source": [
    "marker_genes=data.frame(\n",
    "\n",
    "    celltype=c(\n",
    "        \n",
    "        \"Prog.\", \"Prog.\", \"Prog.\", \"Prog.\", \"Prog.\",\n",
    "\n",
    "        \"NeuP\", \n",
    "               \n",
    "        \"Baso.\", \"Baso.\", \"Baso.\", \"Baso.\", \n",
    "               \n",
    "        \"Mast\", \"Mast\", \n",
    "\n",
    "        \"MegP\", \"MegP\", \n",
    "               \n",
    "        \"Erythroid\", \"Erythroid\", \"Erythroid\", \"Erythroid\", \"Erythroid\", \"Erythroid\", \"Erythroid\", \"Erythroid\", \"Erythroid\", \n",
    "               \n",
    "        \"Mono.\", \"Mono.\", \"Mono.\", \"Mono.\", \n",
    "               \n",
    "        \"RPM\", \"RPM\", \"RPM\", \n",
    "               \n",
    "        \"cDC\", \"cDC\", \"cDC\", \"cDC\", \"cDC\", \"cDC\", \"cDC\"\n",
    "               \n",
    "    ), \n",
    "    \n",
    "    genes=c(\n",
    "        \n",
    "        \"Hlf\", \"Cd34\", \"Kit\", \"Meis1\", \"Spi1\",  \n",
    "\n",
    "        \"Elane\", \n",
    "            \n",
    "        \"Cebpa\", \"Mcpt8\", \"Clec12a\", \"Itga2\", \n",
    "            \n",
    "        \"Mcpt4\", \"Cma1\", \n",
    "\n",
    "        \"Pf4\", \"Itga2b\",\n",
    "            \n",
    "        \"Gata2\", \"Gata1\", \"Klf1\", \"Icam4\", \"Epor\", \"Sox6\", \"Tfrc\", \"Alas2\", \"Slc4a1\",\n",
    "            \n",
    "        \"Ly6c2\", \"Ccr2\", \"Cx3cr1\", \"Itgam\",\n",
    "            \n",
    "        \"Adgre1\", \"Spic\", \"Cd163\",  \n",
    "            \n",
    "        \"Sirpa\", \"Flt3\", \"Xcr1\", \"Cd8a\", \"Cd4\", \"Cd209a\", \"Ccr7\"\n",
    "    \n",
    "    )\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "f2fb5226-5faf-4c65-893c-12766c69860d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'celltype_low'. You can override using the\n",
      "`.groups` argument.\n",
      "\u001b[1m\u001b[22mScale for \u001b[32mfill\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32mfill\u001b[39m, which will replace the existing scale.\n"
     ]
    }
   ],
   "source": [
    "dp_1 <- dp_feature(so, marker_genes$genes, group_by=\"celltype_low\", group_by_order=rev(levels(so$celltype_low)), split=marker_genes$celltype, split_order=unique(marker_genes$celltype), range_max=2.5) + \n",
    "    scale_y_discrete(labels=rev(levels(so$celltype_low))) + theme_global_set(4) + \n",
    "\n",
    "        theme(\n",
    "            legend.position=\"bottom\", \n",
    "            panel.spacing=unit(0.1, \"lines\"), \n",
    "            axis.text.x=element_text(angle=90, vjust=0.5, hjust=1, face=\"italic\"),\n",
    "            axis.text.y=element_text(angle=0, vjust=0.5, hjust=1), \n",
    "            strip.text.x=element_text(angle=90, vjust=0.5, hjust=0), \n",
    "            strip.text=element_text(margin=margin(0.1, 0.1, 0.1, 0.1, \"lines\")), \n",
    "            strip.background=element_rect(fill=\"transparent\", color=NA), \n",
    "            legend.key.size=unit(0.25, \"cm\"), \n",
    "            legend.key.height=unit(0.25, \"cm\"), \n",
    "            legend.key.width=unit(0.25, \"cm\")\n",
    "            \n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "9cb1cfb6-cc6b-4aaa-8013-ae8442c3e44d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_1/dp_celltype_low_marker_genes.pdf\", width=4.0, height=3.2)\n",
    "\n",
    "dp_1\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bf5bcd4-c72d-4485-815d-cf2f8259a6f2",
   "metadata": {},
   "source": [
    "# Celltype abundance test with milo (NaCl D6 vs CpG D6 WT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "dc7a8d20-65de-4f73-a2e9-1aad8bb9ba36",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import Milo object\n",
    "milo <- readRDS(paste0(\"data/scRNAseq/milo/milo_k30.rds\"))\n",
    "results <- readRDS(\"data/scRNAseq/milo/resl.rds\")\n",
    "result <- results[[2]][[\"res_da\"]][[1]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "1ac3f800-7de0-4658-b5a1-a40dee931414",
   "metadata": {},
   "outputs": [],
   "source": [
    "alpha <- 0.05\n",
    "group_by <- \"celltype_low\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "ceb6ba6f-d5b5-4149-a400-dd92f96a19e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Beeswarm \n",
    "data <- suppressMessages(plotDAbeeswarm(result, group.by=group_by, alpha=alpha))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "fc1c6d8a-cd46-4539-a61d-47f81bfc7c83",
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- data$data %>%\n",
    "    mutate(is_signif=ifelse(SpatialFDR < alpha, 1, 0)) %>%\n",
    "    mutate(logFC_color=ifelse(is_signif==1, logFC, NA)) %>%\n",
    "    arrange(group_by) %>%\n",
    "    mutate(Nhood=factor(Nhood, levels=unique(Nhood))) %>%\n",
    "    mutate(pos_x=pos_x, pos_y=pos_y) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "e4154f92-698a-477a-b8de-73f0032b8540",
   "metadata": {},
   "outputs": [],
   "source": [
    "y_breaks <- c(-floor(max(abs(data$logFC))), -floor(max(abs(data$logFC)))/2, 0, floor(max(abs(data$logFC)))/2, floor(max(abs(data$logFC))))\n",
    "n_groups <- length(unique(data$group_by))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "c4f166ea-6184-466f-98b6-3abb63ba631b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22mScale for \u001b[32my\u001b[39m is already present.\n",
      "Adding another scale for \u001b[32my\u001b[39m, which will replace the existing scale.\n"
     ]
    }
   ],
   "source": [
    "options(repr.plot.width=4, repr.plot.height=10)\n",
    "\n",
    "dabp <- ggplot(data, aes(pos_x, pos_y, color=logFC_color)) +\n",
    "    scale_color_gradient2(low=color$sample_group[\"Bl6_NaCl_D6\"], high=color$sample_group[\"Bl6_CpG_D6\"]) + \n",
    "    guides(color=\"none\") +\n",
    "    xlab(\"\") + ylab(\"Log Fold Change\") + ggtitle(\"Differential abundance\") + \n",
    "    scale_x_discrete(labels=setNames(levels(data$group_by), seq(1 ,n_groups))) + \n",
    "    scale_y_continuous(breaks=y_breaks, labels=y_breaks) +\n",
    "    geom_hline(yintercept=0, size=0.1) + \n",
    "    geom_point(size=0.25) + \n",
    "    stat_summary(aes(x=group_by, y=logFC), fun.y=median, geom=\"point\", size=0.5, color=\"black\") +\n",
    "    coord_flip() +\n",
    "    ylim(-max(abs(data$logFC)), max(abs(data$logFC))) + \n",
    "    theme(axis.text.y=element_text(vjust=0.5, hjust=1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "f5387f6e-93e1-498b-9f05-4ec6dd75fe17",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_1/dp_erythroid_milo.pdf\", width=1.5, height=2.5)\n",
    "\n",
    "dabp + theme_global_set(size_select=4)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "595ee323-5f12-45b0-b4d7-0cb69bc29219",
   "metadata": {},
   "source": [
    "# DEA "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cceecae4-f218-492d-8a6e-cbf033b97807",
   "metadata": {},
   "source": [
    "# DEA overview "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "7e12fcb9-dd6c-4f51-a519-0bae975a5ffc",
   "metadata": {},
   "outputs": [],
   "source": [
    "adj_pval_thr <- 0.05\n",
    "log2FC_thr  <- 0.25"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "7f45d5e3-a487-4042-b8ab-3ef1c7e1366e",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrasts_vec <- c(\"Bl6_CpG_D6\", \"Bl6_NaCl_D6\")\n",
    "dea_res_1 <- readRDS(paste0(\"result/dea/scRNAseq/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "01c8a747-9b90-4840-b957-93635ae50371",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- lapply(dea_res_1, function(x) {x %>% dplyr::filter(p_val_adj <= adj_pval_thr & abs(avg_log2FC)>=log2FC_thr ) %>% dplyr::pull(gene) %>% length}) %>% data.frame() %>% t()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "1b9e3659-718d-4101-98bd-e5eb91fa2682",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_1/hm_dea_count.pdf\", width=1.5, height=3)\n",
    "\n",
    "draw(dea_count_hm(mat_1, fontsize_select=2))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91f37f59-a718-4a22-b8d0-f2012166bd3b",
   "metadata": {},
   "source": [
    "# Gene set enrichment analysis (GSEA)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ce27e0e-68bf-4058-af4a-479add646b23",
   "metadata": {},
   "source": [
    "## Gene Set M7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "04981535-3a57-4dbc-8527-44b34e8c2e3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "compute_gsea <- FALSE\n",
    "celltype_select <- names(dea_res_1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "13fac257-ee16-4050-a090-14619f18d836",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(compute_gsea) {\n",
    "\n",
    "    # Collapse gene set\n",
    "    gene_set <- msigdbr::msigdbr(species=\"mouse\", db_species=\"MM\", category=\"M7\", subcategory=NULL)\n",
    "    gene_set <- gene_set[grep(\"_UP\", gene_set$gs_name), ]\n",
    "    gene_set$gs_name <- sapply(strsplit(gene_set$gs_name, \"_\"), function(x) {paste(x[max(1, length(x)-3):(length(x)-2)], collapse=\"_\")})\n",
    "    gene_set$gs_name <- gsub(\"CELL_|CD4_|CD8_|MONOCYTE_|MACROPHAGE_|TREG_|LANGERHANS_|NEUTROPHIL_|CDC1_|CDC2_|MIGDC_|ETAC_|GD_|ILC_|PDC_\", \"\", gene_set$gs_name)\n",
    "    gene_set <- gene_set %>% dplyr::select(gene_symbol, gs_name) %>% dplyr::distinct()\n",
    "\n",
    "    gsea_res <- lapply(celltype_select, function(i) {\n",
    "    \n",
    "        gsea_res <- gsea(dea_res_1[[i]], gene_set=gene_set) %>% dplyr::mutate(col_split=\"Ctl vs D6\", col_label=i)\n",
    "    \n",
    "        return(gsea_res)\n",
    "        \n",
    "    }\n",
    "          )\n",
    "\n",
    "    saveRDS(gsea_res, \"result/gsea/scRNAseq/gsea_d6_vs_ctl.rds\")\n",
    "    \n",
    "} else {\n",
    "\n",
    "    gsea_res <- readRDS(\"result/gsea/scRNAseq/gsea_d6_vs_ctl.rds\")\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "a6fd5626-7d17-4661-9b30-0430310f34c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res <- do.call(rbind, gsea_res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "b0cd21ac-9d3c-4a5f-b555-3cbaaf66cb8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs_name_select <- gsea_res %>% dplyr::filter(padj<=0.05 & abs(NES)>=1) %>% dplyr::pull(pathway)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "4b9e7085-4e5e-4ba2-9ee3-722caadac477",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- gsea_res %>% dplyr::filter(pathway %in% gs_name_select) %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=padj,\n",
    "        values_fill=1\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "1ab24b38-874d-4d3d-aa3d-6f0b32d6597a",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_2 <- gsea_res %>% dplyr::filter(pathway %in% gs_name_select) %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=NES,\n",
    "        values_fill=0\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "e7f65d85-b27c-4825-a8ee-808643b45cf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_split <- read.csv(\"data/reference/immune_dictionary/immune_dictionary.csv\") %>% tibble::column_to_rownames(\"gs_name\")\n",
    "col_label <- mat_1$col_label\n",
    "col_split <- mat_1$col_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "c71f8ab2-97f8-49b4-b4a2-96a7b0a5acbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- mat_1 %>% dplyr::select(-col_label, -col_split) %>% t()\n",
    "mat_2 <- mat_2 %>% dplyr::select(-col_label, -col_split) %>% t()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "35c8aead-6942-42ae-9973-8545e2a0a5a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_split <- row_split[rownames(mat_1), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "e26ac88e-1b3d-498b-9023-5dec70d97aef",
   "metadata": {},
   "outputs": [],
   "source": [
    "rownames(mat_1) <- row_split$norm\n",
    "rownames(mat_2) <- row_split$norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "4f70a30a-8af1-4965-9860-e647aa87c8bc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "row_split <- row_split %>% dplyr::select(norm, family) %>% dplyr::rename(row_split_label=family) %>% dplyr::distinct()\n",
    "rownames(row_split) <- row_split$norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "0302a198-e973-4c87-9423-0364cdc3d4ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_label[col_label==\"cDC.mig.\"] <- \"cDC mig.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "dfc0bb34-2395-41f4-9f7b-4e32e7c7f913",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_1/hm_gsea_d6_vs_ctl_m7.pdf\", width=8, height=12)\n",
    "\n",
    "gsea_hm(mat_1, mat_2, col_label=col_label, col_split=col_split, row_split=row_split, color_neg=color$sample_group[1], color_pos=color$sample_group[2], fontsize_select=2)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90bad978-c7d3-4455-b5c7-622af3578707",
   "metadata": {},
   "source": [
    "## genotype comparison - HM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "7acfe897-3642-4663-ae6c-aabfdab8f24d",
   "metadata": {},
   "outputs": [],
   "source": [
    "compute_gsea <- FALSE\n",
    "celltype_select <- names(dea_res_1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "eb06ee9a-c49d-4c61-a28e-a475fe94785f",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(compute_gsea) {\n",
    "\n",
    "    gsea_res <- lapply(celltype_select, function(i) {\n",
    "    \n",
    "        gsea_res <- gsea(dea_res_1[[i]], category=\"MH\", subcategory=NULL, gene_set=NULL) %>% dplyr::mutate(col_split=\"Ctl vs D6\", col_label=i)\n",
    "    \n",
    "        return(gsea_res)\n",
    "        \n",
    "    }\n",
    "          )\n",
    "\n",
    "    saveRDS(gsea_res, \"result/gsea/scRNAseq/gsea_d6_vs_ctl_mh.rds\")\n",
    "    \n",
    "} else {\n",
    "\n",
    "    gsea_res <- readRDS(\"result/gsea/scRNAseq/gsea_d6_vs_ctl_mh.rds\")\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "4cd4f962-5e29-4fae-b761-cd6dd54d39fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res <- do.call(rbind, gsea_res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "7cdea7ff-33f7-4337-a967-7a7850651484",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs_name_select <- gsea_res %>% dplyr::filter(padj<=0.05 & abs(NES)>=1) %>% dplyr::pull(pathway)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "62915c37-e10c-4c34-936e-d2219d9b4b82",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs_name_select <- c(\n",
    "    \n",
    "    \"HALLMARK_E2F_TARGETS\",\n",
    "    \"HALLMARK_MYC_TARGETS_V1\",\n",
    "    \"HALLMARK_MYC_TARGETS_V2\",\n",
    "    \"HALLMARK_HEME_METABOLISM\",\n",
    "    \n",
    "    \"HALLMARK_INTERFERON_ALPHA_RESPONSE\",\n",
    "    \"HALLMARK_TNFA_SIGNALING_VIA_NFKB\",\n",
    "    \n",
    "    \"HALLMARK_OXIDATIVE_PHOSPHORYLATION\"\n",
    "\n",
    ") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "25759190-5456-428a-965b-5b99e91a3476",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- gsea_res %>% dplyr::filter(pathway %in% gs_name_select) %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=padj,\n",
    "        values_fill=1\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "69eef7d3-df6d-430d-912f-bd96e754f498",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_2 <- gsea_res %>% dplyr::filter(pathway %in% gs_name_select) %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=NES,\n",
    "        values_fill=0\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "a8162dba-70d4-43a6-95aa-197c852b9e5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- mat_1[, c(\"col_label\", \"col_split\", gs_name_select)]\n",
    "mat_2 <- mat_2[, c(\"col_label\", \"col_split\", gs_name_select)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "15c95499-4751-484a-8f67-515ed72ab631",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_label <- mat_1$col_label\n",
    "col_split <- mat_1$col_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "242bd45f-e785-41ad-828e-1617ce5c6ce4",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- mat_1 %>% dplyr::select(-col_label, -col_split) %>% t()\n",
    "mat_2 <- mat_2 %>% dplyr::select(-col_label, -col_split) %>% t()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "8d2673fb-a7bd-4973-acce-d705b030ac9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_label[col_label==\"cDC.mig.\"] <- \"cDC mig.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "cfe536dd-55f2-4596-abb9-5f9a8041c2e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_1/hm_gsea_d6_vs_ctl_mh.pdf\", width=8, height=10)\n",
    "\n",
    "gsea_hm(mat_1, mat_2, col_label=col_label, col_split=col_split, row_split=NULL, pathway_suffix=\"HALLMARK\", color_neg=color$sample_group[1], color_pos=color$sample_group[2], fontsize_select=2)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "075a2f6c-09b1-4771-ba2a-e10c48b32566",
   "metadata": {},
   "source": [
    "## genotype comparison - Reactome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "16fe2bde-e7a3-4b87-84e9-1a6846f4018c",
   "metadata": {},
   "outputs": [],
   "source": [
    "compute_gsea <- FALSE\n",
    "celltype_select <- names(dea_res_1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "f451e215-fecc-4863-a5ea-cb7fed06f0ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(compute_gsea) {\n",
    "\n",
    "    gsea_res <- lapply(celltype_select, function(i) {\n",
    "    \n",
    "        gsea_res <- gsea(dea_res_1[[i]], category=\"M2\", subcategory=\"CP:REACTOME\", gene_set=NULL) %>% dplyr::mutate(col_split=\"Ctl vs D6\", col_label=i)\n",
    "    \n",
    "        return(gsea_res)\n",
    "        \n",
    "    }\n",
    "          )\n",
    "\n",
    "    saveRDS(gsea_res, \"result/gsea/scRNAseq/gsea_d6_vs_ctl_m2_cp_reactome.rds\")\n",
    "    \n",
    "} else {\n",
    "\n",
    "    gsea_res <- readRDS(\"result/gsea/scRNAseq/gsea_d6_vs_ctl_m2_cp_reactome.rds\")\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "18853fef-0d89-4059-ac1a-6423880579ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res <- do.call(rbind, gsea_res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "5ef7e8b3-83b9-4cd6-ad1d-86031042ffa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs_name_select <- gsea_res %>% dplyr::filter(padj<=0.05 & abs(NES)>=1) %>% dplyr::pull(pathway)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "cac83f0d-f063-4d5f-a00d-1d0e2ca391f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- gsea_res %>% dplyr::filter(pathway %in% gs_name_select) %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=padj,\n",
    "        values_fill=1\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "6021bdfa-c0fe-406c-b676-a6f627513223",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_2 <- gsea_res %>% dplyr::filter(pathway %in% gs_name_select) %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=NES,\n",
    "        values_fill=0\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "aa542890-b676-4c00-a11f-6abba460e07c",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_label <- mat_1$col_label\n",
    "col_split <- mat_1$col_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "aad59cc9-f76d-4c90-aa18-c6fb19e04607",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- mat_1 %>% dplyr::select(-col_label, -col_split) %>% t()\n",
    "mat_2 <- mat_2 %>% dplyr::select(-col_label, -col_split) %>% t()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "837d76ae-97dd-4867-8434-9bdf0ac2fa66",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_label[col_label==\"cDC.mig.\"] <- \"cDC mig.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "39f7023b-52ee-47a4-bed8-3591d899b9f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_1/hm_gsea_d6_vs_ctl_m2_cp_reactome.pdf\", width=8, height=20)\n",
    "\n",
    "gsea_hm(mat_1, mat_2, col_label=col_label, col_split=col_split, row_split=NULL, pathway_suffix=\"REACTOME\", color_neg=color$sample_group[1], color_pos=color$sample_group[2], fontsize_select=2)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6de203d0-c383-47f3-83df-cd5d4be5554a",
   "metadata": {},
   "source": [
    "## genotype comparison - GO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "5a782c5e-6700-41ee-bda0-b855fb53ce20",
   "metadata": {},
   "outputs": [],
   "source": [
    "compute_gsea <- FALSE\n",
    "celltype_select <- names(dea_res_1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "a9d66ff3-da62-4f9d-8e33-b7b9b0370c8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(compute_gsea) {\n",
    "\n",
    "    gsea_res <- lapply(celltype_select, function(i) {\n",
    "    \n",
    "        gsea_res <- gsea_cp(dea_res_1[[i]]) %>% dplyr::mutate(col_split=\"Ctl vs D6\", col_label=i)\n",
    "    \n",
    "        return(gsea_res)\n",
    "        \n",
    "    }\n",
    "          )\n",
    "\n",
    "    saveRDS(gsea_res, \"result/gsea/scRNAseq/gsea_d6_vs_ctl_gobp.rds\")\n",
    "    \n",
    "} else {\n",
    "\n",
    "    gsea_res <- readRDS(\"result/gsea/scRNAseq/gsea_d6_vs_ctl_gobp.rds\")\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "b240cba9-ccc3-4fa8-8df2-9cda44705071",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res <- do.call(rbind, gsea_res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "0c428014-d6c9-4fb8-8afb-a28408ae5671",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs_name_select <- gsea_res %>% dplyr::filter(p.adjust<=0.05 & abs(NES)>=1) %>% dplyr::pull(ID)\n",
    "gsea_res <- gsea_res %>% dplyr::filter(ID %in% gs_name_select)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "b83f087e-7775-4400-ad1f-bb9a3880507c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res <- gsea_res[!gsea_res$Description %in% grep(\"T cell|B cell|CD4|CD8|lymphocyte|endothelial|epithelial\", gsea_res$Description, value=TRUE), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "4711c623-6d49-4617-a786-07f162299176",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(compute_gsea) {\n",
    "\n",
    "    gs_reduce <- gsea_cp_reduce(gsea_res)\n",
    "\n",
    "    saveRDS(gs_reduce, \"result/gsea/scRNAseq/gsea_reduce_d6_vs_ctl_gobp.rds\")\n",
    "    \n",
    "} else {\n",
    "\n",
    "    gs_reduce <- readRDS(\"result/gsea/scRNAseq/gsea_reduce_d6_vs_ctl_gobp.rds\")\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "ce8a2ab7-7084-4aab-8be3-996e572bd06f",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res <- gsea_res %>% dplyr::rename(pathway=Description, padj=p.adjust) %>% dplyr::filter(ID %in% gs_reduce$go)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "5734e574-3a05-4176-8237-c64af097c98d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs_name_select <- gsea_res %>% dplyr::filter(padj<=0.1 & abs(NES)>=1) %>% dplyr::pull(ID)\n",
    "gsea_res <- gsea_res %>% dplyr::filter(ID %in% gs_name_select)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "3b96908d-baee-40b1-a0f3-14914fac1cce",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- gsea_res %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=padj,\n",
    "        values_fill=1\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "b6bea6c4-9ea6-4d1b-beb4-df6dea6ef5f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_2 <- gsea_res %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=NES,\n",
    "        values_fill=0\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "c349829c-b541-4479-80f8-a22fa1280730",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_label <- mat_1$col_label\n",
    "col_split <- mat_1$col_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "b5fd08ea-3a2e-4f53-8e35-96da191c5a09",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- mat_1 %>% dplyr::select(-col_label, -col_split) %>% t()\n",
    "mat_2 <- mat_2 %>% dplyr::select(-col_label, -col_split) %>% t()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "30f4bc3b-9dfe-4bbb-a296-afbf0905c719",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_split <- gs_reduce %>% dplyr::select(term, parentTerm) %>% dplyr::rename(row_split_label=parentTerm) %>% dplyr::distinct()\n",
    "rownames(row_split) <- gs_reduce$term"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "679ae0ad-c530-4fb8-a9fd-d783c2dc4fc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_label[col_label==\"cDC.mig.\"] <- \"cDC mig.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "fb938b73-c34c-4335-8f19-7f698afb46d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_1/hm_gsea_d6_vs_ctl_gobp.pdf\", width=10, height=100)\n",
    "\n",
    "gsea_hm(mat_1, mat_2, col_label=col_label, col_split=col_split, row_split=row_split, color_neg=color$sample_group[1], color_pos=color$sample_group[2], pathway_suffix=\"WP\", fontsize_select=2)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37745697-0fbc-4416-980d-8363f55ce442",
   "metadata": {},
   "source": [
    "## Volcano plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "5576a004-511d-46a9-ba0d-3c060041ab4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrasts_vec <- c(\"Bl6_CpG_D6\", \"Bl6_NaCl_D6\")\n",
    "dea_res_1 <- readRDS(paste0(\"result/dea/scRNAseq/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "id": "16e19805-6842-49b1-a2b9-0a4315f04e5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "labels <- list(\n",
    "\n",
    "    \"RPM\"=c(\n",
    "        \n",
    "        \"Gdf15\", \"Spic\", \"Zeb2\", \"Vcam1\", \"Mertk\", \"Pparg\", # RPM marker \n",
    "        \"Cd44\", \"Lrp1\", \"Lgf1\", # Residency\n",
    "        \"Slc40a1\", \"Hmox1\", \"Abca1\", # Iron metabolism \n",
    "        \"Cybb\", \"Ncf1\", \"Ctss\", \"Fcgr2\", \"Ctsl\", \"Ctsd\", \"Lgals3\", \"Marco\", \"Msr1\", \"Fcgr1\", # Phagocytosis \n",
    "        \"H2-Ab1\", \"H2-Eb1\", \"Cd74\", \n",
    "        \"Nfkbia\", \"Relb\", # NFKB\n",
    "        \"Batf\", \"Batf3\", \n",
    "        \"Ifit3\", \"Ddx60\", \"Ifitm1\", \"Ifitm2\", \"Ifitm6\", \"Bst2\", \"Psmb10\", \"Parp9\", \"Ccl5\", # Interferon signaling\n",
    "        \"Cxcl9\", \"Bcl2a1a\", \"Bcl2a1b\", \"Bcl2a1d\" # Survival \n",
    "        \n",
    "        \n",
    "    \n",
    "    ), \n",
    "\n",
    "    \"intMo\"=c(\n",
    "        \n",
    "        \"Gdf15\", \"Spic\", \"Zeb2\", \"Vcam1\", \"Mertk\", \"Pparg\", # RPM marker \n",
    "        \"Cd44\", \"Lrp1\", \"Lgf1\", # Residency\n",
    "        \"Slc40a1\", \"Hmox1\", \"Abca1\", # Iron metabolism \n",
    "        \"Cybb\", \"Ncf1\", \"Ctss\", \"Fcgr2\", \"Ctsl\", \"Ctsd\", \"Lgals3\", \"Marco\", \"Msr1\", \"Fcgr1\", # Phagocytosis \n",
    "        \"H2-Ab1\", \"H2-Eb1\", \"Cd74\", \n",
    "        \"Nfkbia\", \"Relb\", # NFKB\n",
    "        \"Batf\", \"Batf3\", \n",
    "        \"Ifit3\", \"Ddx60\", \"Ifitm1\", \"Ifitm2\", \"Ifitm6\", \"Bst2\", \"Psmb10\", \"Parp9\", \"Ccl5\", # Interferon signaling\n",
    "        \"Cxcl9\", \"Bcl2a1a\", \"Bcl2a1b\", \"Bcl2a1d\" # Survival \n",
    "    \n",
    "    ), \n",
    "\n",
    "    \"cMo\"=c(\n",
    "        \n",
    "        \"Gdf15\", \"Spic\", \"Zeb2\", \"Vcam1\", \"Mertk\", \"Pparg\", # RPM marker \n",
    "        \"Cd44\", \"Lrp1\", \"Lgf1\", # Residency\n",
    "        \"Slc40a1\", \"Hmox1\", \"Abca1\", # Iron metabolism \n",
    "        \"Cybb\", \"Ncf1\", \"Ctss\", \"Fcgr2\", \"Ctsl\", \"Ctsd\", \"Lgals3\", \"Marco\", \"Msr1\", \"Fcgr1\", # Phagocytosis \n",
    "        \"H2-Ab1\", \"H2-Eb1\", \"Cd74\", \n",
    "        \"Nfkbia\", \"Relb\", # NFKB\n",
    "        \"Batf\", \"Batf3\", \n",
    "        \"Ifit3\", \"Ddx60\", \"Ifitm1\", \"Ifitm2\", \"Ifitm6\", \"Bst2\", \"Psmb10\", \"Parp9\", \"Ccl5\", # Interferon signaling\n",
    "        \"Cxcl9\", \"Bcl2a1a\", \"Bcl2a1b\", \"Bcl2a1d\" # Survival \n",
    "    \n",
    "    )\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "1a3a3c36-02c4-452e-b59e-136eb78fbac9",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=6*6, repr.plot.height=6)\n",
    "\n",
    "vp_1 <- lapply(names(dea_res_1), function(i) {\n",
    "\n",
    "    if(i %in% names(labels)) {\n",
    "\n",
    "        label <- labels[[i]]\n",
    "        \n",
    "    } else {\n",
    "\n",
    "        label <- NULL\n",
    "        \n",
    "    }\n",
    "\n",
    "    v_pl(dea_res_1[[i]], title=i, log2FC_thr=log2FC_thr , adj_pval_thr=adj_pval_thr, color_pos=color$sample_group[contrasts_vec[1]], color_neg=color$sample_group[contrasts_vec[2]], label=label, top_label=20, label_merge=FALSE, point_size=0.5, label_size=2, aspect_ratio=0.8, y_limit=3) + theme(legend.position=\"none\") + theme_global_set(size_select=4)\n",
    "\n",
    "}\n",
    "              )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "90142b65-7b5c-47c6-9a1a-b20d2546dd99",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_1/v_plot_d6_vs_ctl.pdf\", width=5*2, height=4*2)\n",
    "\n",
    "plots <- lapply(vp_1, function(p) egg::set_panel_size(p, width=unit(3, \"cm\"), height=unit(3, \"cm\")))\n",
    "do.call(gridExtra::grid.arrange, c(plots, ncol=5, nrow=4, newpage=FALSE))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62641897-b5b7-4069-80c7-00379d508ff5",
   "metadata": {},
   "source": [
    "## log2FC DEA heatmap of selected genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "66b41ea0-c3f0-409a-8154-112aaf6899a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "log2fc_thr <- 0.25\n",
    "p_val_adj_thr <- 0.05\n",
    "celltype_select <- c(\"cMo\", \"ncMo\", \"intMo\", \"RPM\", \"cDC1\", \"cDC2\", \"moDC\", \"cDC mig.\", \"Basophil\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "99da191d-36a0-41a1-a92c-66fc23f21831",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_dea <- lapply(dea_res_1, function(x) {x %>% dplyr::filter(abs(avg_log2FC)>=log2fc_thr & p_val_adj<=p_val_adj_thr) %>% dplyr::pull(gene)})\n",
    "genes_dea <- do.call(\"c\", genes_dea) %>% unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "7b520818-cafe-427e-81ab-9684f7e6fa17",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_genes <- list(\n",
    "    \n",
    "    \"MHC I\"=c(\"H2-K1\", \"H2-D1\", \"B2m\"),\n",
    "    # \"MHC II\"=c(\"H2-Ab1\", \"H2-Eb1\", \"Cd74\"),\n",
    "    \"Antigen/Proteasome\"=c(\"Psmb8\", \"Psmb9\", \"Psme1\", \"Psme2\"),\n",
    "    # \"Phagosome/Autophagy\"=c(\"Ctss\", \"Ctsb\", \"Lamp1\", \"Lamp2\", \"Atp6v0d2\", \"Map1lc3b\", \"Becn1\", \"Atg5\"),\n",
    "    \"IFN signaling\"=c(\"Ly6a\", \"Ifitm1\", \"Ifitm2\", \"Ifitm3\", \"Gbp2\", \"Gbp7\", \"Gbp8\"),\n",
    "    # \"NFkB/MAPK\"=c(\"Tlr9\", \"Myd88\", \"Irak4\", \"Irak1\", \"Traf6\", \"Map3k7\", \"Mapk14\", \"Mapk8\", \"Nfkb1\", \"Nfkb2\", \"Rela\", \"Relb\", \"Nfkbia\"),\n",
    "    # \"AP1\"=c(\"Fos\", \"Jun\", \"Junb\", \"Jund\", \"Fosb\", \"Fosl1\", \"Atf3\", \"Batf3\"),\n",
    "    # \"Inflammasome\"=c(\"Pycard\", \"Nlrp3\", \"Nlrp1b\", \"Il1b\", \"Il18\"),\n",
    "    # \"Survival\"=c(\"Cxcl9\", \"Havcr2\", \"Bcl2a1a\", \"Bcl2a1b\", \"Bcl2a1d\"), \n",
    "    # \"ROS/NOX2 complex\"=c(\"Cybb\", \"Cyba\", \"Ncf1\", \"Ncf2\", \"Ncf4\", \"Rac2\"),\n",
    "    # \"RPM iron/heme program\"=c(\"Spic\", \"Slc40a1\", \"Hmox1\", \"Fth1\", \"Ftl1\", \"Hamp\"), \n",
    "    # \"Glycolysis/Lactate\"=c(\"Slc2a1\", \"Hk2\", \"Pfkp\", \"Pfkfb3\", \"Aldoa\", \"Gapdh\", \"Pgk1\", \"Pkm\", \"Ldha\", \"Slc16a1\"),\n",
    "    \"OXPHOS\"=c(\"Ndufb8\", \"Uqcrb\", \"Cox5a\", \"Cox8a\", \"Atp5b\", \"Atp5c1\")\n",
    "    # \"Pentose Phosphate\"=c(\"G6pdx\", \"Pgls\", \"Rpe\", \"Rpia\", \"Tkt\", \"Taldo1\"),\n",
    "    # \"Hexosamine Pathway\"=c(\"Gfpt1\", \"Gfpt2\", \"Gnpda1\", \"Ogt\", \"Oga\")\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "9f9c444b-eb36-4e5d-b3d5-bb1dec602164",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [],
      "text/latex": [],
      "text/markdown": [],
      "text/plain": [
       "named character(0)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "row_split <- setNames(unlist(row_genes, use.names = FALSE), rep(names(row_genes), sapply(row_genes, length)))\n",
    "row_split[!row_split %in% rownames(so)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "4eb23a39-fc9d-40dd-abcf-a273323b15c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_split <- row_split[row_split %in% genes_dea]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "0c7405ef-0693-43a2-9bf2-102e091791f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- lapply(names(dea_res_1), function(i) {dea_res_1[[i]] %>% dplyr::mutate(col_name=i) %>% dplyr::select(avg_log2FC, col_name) %>% tibble::rownames_to_column(\"gene\")})\n",
    "mat_2 <- lapply(names(dea_res_1), function(i) {dea_res_1[[i]] %>% dplyr::mutate(col_name=i) %>% dplyr::select(p_val_adj, col_name) %>% tibble::rownames_to_column(\"gene\")})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "c766ef1e-3f49-460e-a2db-fd4517a717bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- do.call(\"rbind\", mat_1)  %>% reshape2::dcast(., gene~col_name, value.var=\"avg_log2FC\", fill=0) %>% tibble::column_to_rownames(\"gene\")\n",
    "mat_2 <- do.call(\"rbind\", mat_2) %>% reshape2::dcast(., gene~col_name, value.var=\"p_val_adj\", fill=1) %>% tibble::column_to_rownames(\"gene\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "ff4c8ac2-a923-4b33-8295-c0bb6eb91078",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(mat_1)[colnames(mat_1)==\"cDC.mig.\"] <- \"cDC mig.\"\n",
    "colnames(mat_2)[colnames(mat_2)==\"cDC.mig.\"] <- \"cDC mig.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "0ac4e76a-0a9f-4001-9def-d7d6afa3cdc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- mat_1[row_split, celltype_select]\n",
    "mat_2 <- mat_2[row_split, celltype_select]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "7dde1409-9bd2-4cb8-a990-78e5c51d69d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1[is.na(mat_1)] <- 0\n",
    "mat_2[is.na(mat_2)] <- 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "f3ff8030-4b90-430d-8f69-d26465a85cbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_bool_thr <- rowSums(mat_2<=0.05)>0\n",
    "\n",
    "mat_1 <- mat_1[sig_bool_thr, ]\n",
    "mat_2 <- mat_2[sig_bool_thr, ]\n",
    "row_split <- row_split[sig_bool_thr]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "98aa4cb4-9c83-4d5e-b109-f1874878fbeb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_1/hm_dea_res_myeloid.pdf\", width=3, height=10)\n",
    "source(\"bin/dea_pl.R\")\n",
    "dea_res_hm(mat_1, mat_2, row_split=factor(names(row_split), levels=unique(names(row_split))), scale_width=1, color_min=color$sample_group[contrasts_vec[2]], color_max=color$sample_group[contrasts_vec[1]], cluster_rows=FALSE, cluster_columns=FALSE, fontsize_select=2)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ac3ae92-6903-4020-b103-baea35de071e",
   "metadata": {},
   "source": [
    "# Expression UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3b21a70-d502-435f-8d50-ad2fc56975f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- NormalizeData(so, normalization.method=\"LogNormalize\", scale.factor=1e4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a8462c0-b855-4f4e-885c-7bb6d13706c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_id_thr <- so@reductions$umap@cell.embeddings %>% as.data.frame() %>% dplyr::filter(UMAP_1==min(UMAP_1) | UMAP_1==max(UMAP_1) | UMAP_2==min(UMAP_2) | UMAP_2==max(UMAP_2)) %>% rownames()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22c87b7c-143e-475c-9af1-ff6229dc47b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_tmp <- so[, (so$celltype_low %in% c(\"GMP\", \"NeuP\", \"BasoP\", \"Basophil\", \"MastP\", \"MegP\", \"MEP\", \"Proerythroblast\", \"Erythroblast\")) | so$cell_id %in% cell_id_thr]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf952d91-7fe2-4a01-85de-76df91d449c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_1/umap_marker_genes.pdf\", width=10, height=10)\n",
    "\n",
    "    gridExtra::grid.arrange(\n",
    "\n",
    "        fplot(so_tmp, features=\"pct_counts_definitive_HB\", assay=\"RNA\", slot=\"data\", pt_size=0.25, size_select=4, max_cutoff=30) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "        fplot(so_tmp, features=\"pct_counts_primitive_HB\", assay=\"RNA\", slot=\"data\", pt_size=0.25, size_select=4, max_cutoff=30) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "        fplot(so_tmp, features=\"Cd34\", assay=\"RNA\", slot=\"data\", pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "        fplot(so_tmp, features=\"Meis1\", assay=\"RNA\", slot=\"data\", pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "        fplot(so_tmp, features=\"Gata1\", assay=\"RNA\", slot=\"data\", pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")),\n",
    "        fplot(so_tmp, features=\"Gata2\", assay=\"RNA\", slot=\"data\", pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")),\n",
    "        fplot(so_tmp, features=\"Klf1\", assay=\"RNA\", slot=\"data\", pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")),\n",
    "        fplot(so_tmp, features=\"Pf4\", assay=\"RNA\", slot=\"data\", pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")),\n",
    "        fplot(so_tmp, features=\"Sox6\", assay=\"RNA\", slot=\"data\", pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")),\n",
    "        fplot(so_tmp, features=\"Elane\", assay=\"RNA\", slot=\"data\", pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")),\n",
    "        fplot(so_tmp, features=\"Prss34\", assay=\"RNA\", slot=\"data\", pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")),\n",
    "        fplot(so_tmp, features=\"Cma1\", assay=\"RNA\", slot=\"data\", pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "        fplot(so_tmp, features=\"Spi1\", assay=\"RNA\", slot=\"data\", pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "            \n",
    "        fplot(so_tmp, restrict=c(\"sample_group\", \"Bl6_NaCl_D6\"), features=\"Gzmb\", assay=\"RNA\", slot=\"data\", pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "        fplot(so_tmp, restrict=c(\"sample_group\", \"Bl6_CpG_D6\"), features=\"Gzmb\", assay=\"RNA\", slot=\"data\", pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")),\n",
    "            \n",
    "        fplot(so_tmp, restrict=c(\"sample_group\", \"Bl6_NaCl_D6\"), features=\"Hes1\", assay=\"RNA\", slot=\"data\", pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "        fplot(so_tmp, restrict=c(\"sample_group\", \"Bl6_CpG_D6\"), features=\"Hes1\", assay=\"RNA\", slot=\"data\", pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")),\n",
    "            \n",
    "        fplot(so_tmp, restrict=c(\"sample_group\", \"Bl6_NaCl_D6\"), features=\"Egr1\", assay=\"RNA\", slot=\"data\", pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "        fplot(so_tmp, restrict=c(\"sample_group\", \"Bl6_CpG_D6\"), features=\"Egr1\", assay=\"RNA\", slot=\"data\", pt_size=0.25, size_select=4, max_set=5) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "        \n",
    "        ncol=4\n",
    "    \n",
    "    )\n",
    "\n",
    "dev.off()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "r.4.4.1-FD20200109SPLENO",
   "language": "R",
   "name": "r.4.4.1-fd20200109spleno"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
