{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "781951e7-14df-437b-b205-0b07dbd3b2d4",
   "metadata": {},
   "source": [
    "# BSF and VBC CellRanger import "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65a530a6-e368-4009-a10f-a276ccb5893c",
   "metadata": {},
   "outputs": [],
   "source": [
    "library_load <- suppressMessages(\n",
    "    \n",
    "    suppressWarnings(\n",
    "    \n",
    "        list(\n",
    "            \n",
    "            # Seurat \n",
    "            library(Seurat),\n",
    "            \n",
    "            # HTO demultiplex\n",
    "            library(demuxmix), \n",
    "            library(DropletUtils), \n",
    "\n",
    "            # Module score \n",
    "            library(UCell), \n",
    "\n",
    "            # SingleR\n",
    "            library(SingleR), \n",
    "            library(SingleCellExperiment), \n",
    "            \n",
    "            # Data \n",
    "            library(tidyverse),\n",
    "            \n",
    "            # Plot \n",
    "            library(ggplot2), \n",
    "            \n",
    "            # Pyhton compatibility\n",
    "            library(reticulate)\n",
    "            \n",
    "        )\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d733565b-5b12-431e-9a3c-6854305e0edc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Configure reticulate \n",
    "use_condaenv(condaenv='p.3.10.16-FD20200109SPLENO', conda=\"/nobackup/peer/fdeckert/miniconda3/bin/conda\", required=NULL)\n",
    "py_config()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15f4dc19-f82a-4af4-8204-2e597dd39288",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc45c9ca-7211-45a0-b6d3-98a47d103200",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_seed <- 42\n",
    "set.seed(random_seed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fdbfd25-0727-48e7-b0c3-9966d410a7a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set working directory to project root\n",
    "setwd(\"/research/peer/fdeckert/FD20200109SPLENO\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3196fa1d-e0d5-473f-9ed3-43cd37d0d1da",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Source files\n",
    "source(\"plotting_global.R\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f4f2b51-73d3-4593-996e-618557ec639c",
   "metadata": {},
   "source": [
    "# Parameter settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb37668f-b3a1-46a9-a9e1-fb5cb45722a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Seurat parameters \n",
    "min_cells    <- 0 # Filter out features which are only present in less than min_cells cells - default 3\n",
    "min_features <- 0 # Filter out cells which have fewer than min_features features - default 200\n",
    "\n",
    "# Directories\n",
    "run_path_bsf <- \"/nobackup/peer/fdeckert/FD20200109SPLENO/data/BSA_0355_SM01_10x_SPLENO/\"\n",
    "run_path_vbc <- \"/nobackup/peer/fdeckert/FD20200109SPLENO/data/VBC_R18518_22WYJ2LT3/\"\n",
    "\n",
    "# Files\n",
    "h5ad_file <- \"data/scRNAseq/object/raw.h5ad\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "059427c9-7dbd-44ae-8202-d9fe3d38608d",
   "metadata": {},
   "source": [
    "# BSF import "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8953045a-7a75-4b0f-938c-8fef9d9763b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_bsf <- data.frame(\n",
    "\n",
    "    run_path=run_path_bsf, \n",
    "    sample_dir=c(\"M6_control_Mac\", \"M6_control_Prog\", \"M8_control_Mac\", \"M8_control_Prog\", \"M1_CpG_Mac\", \"M1_CpG_Prog\", \"M2_CpG_Mac\", \"M2_CpG_Prog\"), \n",
    "    facility=\"BSF\", \n",
    "    lib_assay=\"3_prime\", \n",
    "    facs=rep(c(\"Myeloid\", \"Progenitor\"), 4), \n",
    "    infection=rep(c(\"NaCl\", \"CpG\"), each=4),\n",
    "    dpi=rep(\"D6\", each=8), \n",
    "    sample_rep=rep(c(\"Rep1\", \"Rep1\", \"Rep2\", \"Rep2\"), 2), \n",
    "    genotype=rep(c(\"Bl6\"), 8), \n",
    "    sex=rep(c(\"female\"), 8), \n",
    "    age=rep(c(\"11 wks\"), 8)\n",
    "    \n",
    ") %>% \n",
    "    dplyr::mutate(sample_group=paste0(genotype, \"_\", infection, \"_\", dpi)) %>% \n",
    "    dplyr::mutate(sample_group_rep=paste0(genotype, \"_\", infection, \"_\", dpi, \"_\", sample_rep)) %>% \n",
    "    dplyr::mutate(sample_name=paste0(genotype, \"_\", infection, \"_\", dpi, \"_\", facs, \"_\", sample_rep))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea2319d2-ce37-4005-ac84-16a7cc69af0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_bsf <- lapply(split(meta_bsf, f=meta_bsf$sample_dir), function(x) {\n",
    "    \n",
    "    sample_path <- paste0(x$run_path, x$sample_dir, \"/outs/filtered_feature_bc_matrix\")\n",
    "    message(paste(\"import:\", sample_path))\n",
    "\n",
    "    # Read 10x data in as sparse matrix (sm)\n",
    "    gex <- Read10X(data.dir=sample_path)\n",
    "\n",
    "    # Transform sparse matrix to seurat object (so)\n",
    "    so <- CreateSeuratObject(counts=gex, min.cells=min_cells, min.features=min_features)\n",
    "    \n",
    "    # Meta \n",
    "    meta <- dplyr::cross_join(so@meta.data, x)\n",
    "    rownames(meta) <- colnames(so)\n",
    "    \n",
    "    so@meta.data <- meta\n",
    "    \n",
    "    # Set Ident\n",
    "    so$orig.ident <- so$sample_name\n",
    "    \n",
    "    return(so)\n",
    "    \n",
    "}\n",
    "            )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f73e3269-b624-4639-9959-4d4d0feb494f",
   "metadata": {},
   "source": [
    "# VBC import and "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95c6c75a-b006-401b-be74-01774705a6a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "cnt_vbc <- Read10X(paste0(run_path_vbc, \"22WYJ2LT3_R18518/outs/filtered_feature_bc_matrix/\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67bd1c80-b5ba-4915-9e86-aebbd21f3b5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gex_vbc <- cnt_vbc[[1]]\n",
    "hto_vbc <- cnt_vbc[[2]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69d32c63-bf6d-4720-a0a1-6f1c2c175180",
   "metadata": {},
   "source": [
    "## Sample HTO demultiplexing "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88abd64f-a222-4b8f-ae56-b229fe9d13d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Demuxmix\n",
    "dm <- demuxmix(hto=as.matrix(hto_vbc), model=\"naive\")\n",
    "dm <- dmmClassify(dm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "914ce631-5b7e-45d0-845f-a2e267c74e56",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(dm) <- c(\"dm_hto\", \"dm_prob\", \"dm_class\")\n",
    "dm <- tibble::rownames_to_column(dm, \"cell_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34bfacb7-2e5b-4721-8408-49f15be85b20",
   "metadata": {},
   "outputs": [],
   "source": [
    "table(dm[dm$dm_class==\"singlet\", ]$dm_hto)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b83260d-2c42-42c4-8b97-da8a23abe908",
   "metadata": {},
   "outputs": [],
   "source": [
    "# HashedDrops\n",
    "hd <- hashedDrops(hto_vbc, confident.min=0.5) %>% as.data.frame()\n",
    "hd$Best <- paste0(\"Sample\", hd$Best)\n",
    "hd$Second <- paste0(\"Sample\", hd$Second)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84b74121-4757-48f2-90f6-756619166c6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "hd$hd_class <- ifelse(hd$Doublet, \"multiplet\", ifelse(!hd$Confident, \"uncertain\", \"singlet\"))\n",
    "hd <- hd[, c(\"Best\", \"LogFC\", \"hd_class\")]\n",
    "colnames(hd) <- c(\"hd_hto\", \"hd_logfc\", \"hd_class\")\n",
    "hd <- tibble::rownames_to_column(hd, \"cell_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b63061b2-9b9f-42e6-80d4-578a69d89aed",
   "metadata": {},
   "outputs": [],
   "source": [
    "table(hd[hd$hd_class==\"singlet\", ]$hd_hto)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc2904ab-7110-4bd7-92fa-1447de718f51",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Combine results and use hd label for joining \n",
    "hto_demux <- dplyr::left_join(dm, hd, by=join_by(cell_id)) %>% dplyr::rename(\"hto_label\"=\"hd_hto\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e01df3f-3b19-4a49-934b-b0394aab2c69",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_vbc <- data.frame(\n",
    "\n",
    "    run_path=run_path_vbc, \n",
    "    sample_dir=\"22WYJ2LT3_R18518\", \n",
    "    facility=\"VBC\", \n",
    "    lib_assay=\"5_prime\", \n",
    "    facs=rep(c(\"Mix\"), 6), \n",
    "    infection=rep(c(\"CpG\", \"CpG\", \"Baseline\"), each=2),\n",
    "    dpi=c(\"D3\", \"D3\", \"D1\", \"D1\", \"D0\", \"D0\"), \n",
    "    sample_rep=rep(\"Rep1\", 6), \n",
    "    genotype=rep(c(\"IFNAR_fl\", \"IFNAR_fl_LysM_cre\"), 3), \n",
    "    sex=rep(c(\"female\"), 6), \n",
    "    age=rep(c(\"3 wks\"), 6), \n",
    "    hto_label=paste0(\"Sample\", 1:6)\n",
    "\n",
    ") %>% \n",
    "    dplyr::mutate(sample_group=paste0(genotype, \"_\", infection, \"_\", dpi)) %>% \n",
    "    dplyr::mutate(sample_group_rep=paste0(genotype, \"_\", infection, \"_\", dpi, \"_\", sample_rep)) %>% \n",
    "    dplyr::mutate(sample_name=paste0(genotype, \"_\", infection, \"_\", dpi, \"_\", facs, \"_\", sample_rep))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1437382-6c59-45a7-a3d5-ba83fe47a337",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_vbc <- dplyr::left_join(meta_vbc, hto_demux, by=join_by(hto_label)) %>% tibble::column_to_rownames(\"cell_id\") %>% dplyr::filter(hd_class==\"singlet\") # Filtered by hd singlet classification "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fd29035-79fc-4719-90dd-3b59bb8349f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "gex_vbc <- gex_vbc[, rownames(meta_vbc)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95a5fa3a-669d-4b9c-8547-51ad0c63f1d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_vbc <- Seurat::CreateSeuratObject(counts=gex_vbc, assay=\"RNA\", meta.data=meta_vbc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4a82488-1bd3-446b-b2c4-4fbe8cfd5d32",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_vbc <- Seurat::SplitObject(so_vbc, split.by=\"sample_name\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed26bd78-1ec6-4b6c-8ec6-8e374bd070bb",
   "metadata": {},
   "source": [
    "# Merge data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e5ac6e6-3a93-41d1-89b3-8892fdbcb84e",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- c(so_bsf, so_vbc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9977eca-33eb-4ddc-a0a8-6a905ae5875d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge Seurat objects \n",
    "so <- merge(x=so[[1]], y=so[2:length(so)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20ec0a66-e2c6-4073-84ac-1dcaa42161ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge layers \n",
    "so <- JoinLayers(object=so, layers=\"counts\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca382399-731f-4841-8fa2-f989379c2c2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set cell_id \n",
    "so$cell_id <- colnames(so)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53024a72-ac0e-4a03-a10d-61a5a72a2c15",
   "metadata": {},
   "outputs": [],
   "source": [
    "so$nCount_RNA <- NULL\n",
    "so$nFeature_RNA <- NULL"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41aaa628-124e-43fc-8515-588050211106",
   "metadata": {},
   "source": [
    "# Automated cell type annotation "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fcf5cd1-8007-46dd-bce7-78bc2ec26550",
   "metadata": {},
   "source": [
    "## ImmGen "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f31a838-3752-4260-863c-c0505b89f0dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load reference data \n",
    "sce_ref <- celldex::ImmGenData()\n",
    "    \n",
    "# Seurat object to SingleCellExperiment\n",
    "sce_qry <- SingleCellExperiment::SingleCellExperiment(assays=list(counts=Seurat::GetAssayData(so, assay=\"RNA\", layer=\"counts\")))\n",
    "sce_qry <- scuttle::logNormCounts(sce_qry)\n",
    "\n",
    "# Predict labels\n",
    "label_main_immgen <- SingleR::SingleR(test=sce_qry, ref=sce_ref, labels=sce_ref$label.main, assay.type.test=\"logcounts\", assay.type.ref=\"logcounts\", de.method=\"classic\", fine.tune=FALSE) %>%\n",
    "    as.data.frame() %>% dplyr::select(pruned.labels, delta.next) %>% dplyr::rename(label_main_immgen=pruned.labels, delta_score_main_immgen=delta.next)\n",
    "\n",
    "label_fine_immgen <- SingleR::SingleR(test=sce_qry, ref=sce_ref, labels=sce_ref$label.fine, assay.type.test=\"logcounts\", assay.type.ref=\"logcounts\", de.method=\"classic\", fine.tune=FALSE) %>%\n",
    "    as.data.frame() %>% dplyr::select(pruned.labels, delta.next) %>% dplyr::rename(label_fine_immgen=pruned.labels, delta_score_fine_immgen=delta.next)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6c0d5c2-0745-42cd-aec3-5b630a565298",
   "metadata": {},
   "source": [
    "## Haemopedia Mouse RNAseq from Haemosphere"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7180042-a9dc-4235-8fe1-131feb89de2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load reference data \n",
    "meta_ref <- read.delim(\"/research/peer/fdeckert/reference/bulkRNAseq/haemosphere/data/Haemopedia-Mouse-RNASeq_samples.txt\") %>% tibble::column_to_rownames(\"sampleId\")\n",
    "mat_ref <- read.delim(\"/research/peer/fdeckert/reference/bulkRNAseq/haemosphere/data/Haemopedia-Mouse-RNASeq_raw.txt\", row.names=1)[, rownames(meta_ref)]\n",
    "\n",
    "# Annotate ENSEML id with gene symbol\n",
    "gtf <- rtracklayer::import(\"/research/peer/fdeckert/reference/genome/GRCm38/Mus_musculus.GRCm38.102.gtf\") %>% as.data.frame() \n",
    "\n",
    "convert <- gtf[, c(\"gene_id\", \"gene_name\")] %>% dplyr::distinct() %>% dplyr::group_by(gene_name) %>% dplyr::mutate(n=n()) %>% dplyr::filter(n==1) %>% tibble::column_to_rownames(\"gene_id\") %>% dplyr::select(-n)\n",
    "gene_id <- intersect(rownames(mat_ref), rownames(convert)) %>% unique()\n",
    "convert <- convert[which(rownames(convert) %in% gene_id), , drop=FALSE]\n",
    "mat_ref <- mat_ref[which(rownames(mat_ref) %in% gene_id), , drop=FALSE]\n",
    "rownames(mat_ref) <- convert[rownames(mat_ref), ]\n",
    "\n",
    "# Load reference data\n",
    "sce_ref <- SingleCellExperiment::SingleCellExperiment(list(counts=mat_ref), colData=meta_ref)\n",
    "sce_ref <- scuttle::logNormCounts(sce_ref)\n",
    "\n",
    "# AnnData object to SingleCellExperiment\n",
    "sce_qry <- SingleCellExperiment::SingleCellExperiment(assays=list(counts=Seurat::GetAssayData(so, assay=\"RNA\", layer=\"counts\")))\n",
    "sce_qry <- scuttle::logNormCounts(sce_qry)\n",
    "\n",
    "# Predict labels\n",
    "label_main_haemopedia <- SingleR::SingleR(test=sce_qry, ref=sce_ref, labels=sce_ref$celltype, assay.type.test=\"logcounts\", assay.type.ref=\"logcounts\", de.method=\"classic\", fine.tune=FALSE) %>%\n",
    "    as.data.frame() %>% dplyr::select(pruned.labels, delta.next) %>% dplyr::rename(label_main_haemopedia=pruned.labels, delta_score_main_haemopedia=delta.next)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02558988-81a5-4fdf-9535-53073b0de0b6",
   "metadata": {},
   "source": [
    "## Add cell type annotation to meta data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2a9e100-15b4-4772-85a4-c3f3fde8a5e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- AddMetaData(so, cbind(label_main_immgen, label_fine_immgen, label_main_haemopedia))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d17c93e-1f5e-4642-bb5f-ee91b1fe356a",
   "metadata": {},
   "source": [
    "# Cell cycle scoring "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c367fdee-f939-4ef7-8245-fc4a64fe669d",
   "metadata": {},
   "source": [
    "## Human to mouse gene symbol convert"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b81393a3-d13c-4bc9-a902-69e8f4d54b70",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Seurat cell cycle genes \n",
    "cc_genes_s <- cc.genes.updated.2019$s.genes\n",
    "cc_genes_g2m <- cc.genes.updated.2019$g2m.genes\n",
    "\n",
    "# Get mouse orthologs from human gene simbols\n",
    "httr::set_config(httr::config(ssl_verifypeer=FALSE))\n",
    "\n",
    "hgnc_mart <- biomaRt::useMart(\"ensembl\", dataset=\"hsapiens_gene_ensembl\", host=\"https://dec2021.archive.ensembl.org/\")\n",
    "mm_mart <- biomaRt::useMart(\"ensembl\", dataset=\"mmusculus_gene_ensembl\", host=\"https://dec2021.archive.ensembl.org/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eab46b9e-7b8e-4149-85b3-d87fba8a86e8",
   "metadata": {},
   "source": [
    "## Seurat gene set based on Tirosh et al, 2015"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "437b4395-771c-4488-8be7-6e76d81eeef6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Conver genes\n",
    "cc_genes_s <- biomaRt::getLDS(attributes=c(\"hgnc_symbol\"), filters=\"hgnc_symbol\", values=cc_genes_s, mart=hgnc_mart, attributesL=c(\"mgi_symbol\"), martL=mm_mart, uniqueRows=TRUE)[, 2]\n",
    "cc_genes_g2m <- biomaRt::getLDS(attributes=c(\"hgnc_symbol\"), filters=\"hgnc_symbol\", values=cc_genes_g2m, mart=hgnc_mart, attributesL=c(\"mgi_symbol\"), martL=mm_mart, uniqueRows=TRUE)[, 2]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b77a7efd-9411-4516-8465-1c4584bfe7ab",
   "metadata": {},
   "source": [
    "## Compute Module score with UCell "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96d3a33f-0d45-4421-a14c-e7f020366158",
   "metadata": {},
   "outputs": [],
   "source": [
    "so$msG2M <- NULL\n",
    "so$msS <- NULL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "baf5e098-b60e-4699-9f78-801544d557a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cell cycle scoring anc class\n",
    "so <- UCell::AddModuleScore_UCell(so, features=list(msS_RNA=cc_genes_s, msG2M_RNA=cc_genes_g2m), assay=\"RNA\", slot=\"counts\", name=\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60efdcb8-644d-4a5c-a633-ddea2b48bda2",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(so@meta.data) <- gsub(\"msS_RNA\", \"S_score\", colnames(so@meta.data))\n",
    "colnames(so@meta.data) <- gsub(\"msG2M_RNA\", \"G2M_score\", colnames(so@meta.data))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2e7ed5c-a15e-4f65-ad77-a31274019427",
   "metadata": {},
   "source": [
    "# Save output file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2dad1c0-2b3f-4b5b-8013-640201ee8f4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store data as h5ad \n",
    "adata <- import(\"anndata\", as=\"adata\", convert=FALSE)\n",
    "pd <- import(\"pandas\", as=\"pd\", convert=FALSE)\n",
    "np <- import(\"numpy\", as=\"np\", convert=FALSE)\n",
    "    \n",
    "# Transform dgCMatrix to sparse sc_sparse matrix\n",
    "X <- GetAssayData(so, assay=\"RNA\", layer=\"counts\")    \n",
    "X <- adata$AnnData(X=X)$X$T\n",
    "\n",
    "# Combine \n",
    "adata <- adata$AnnData(X=X, obs=so@meta.data)\n",
    "adata$var_names <- rownames(GetAssayData(so, assay=\"RNA\", layer=\"counts\"))\n",
    "\n",
    "adata$raw <- adata\n",
    "\n",
    "# Save\n",
    "adata$write_h5ad(h5ad_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc9924e5-ef07-44b0-b931-3d741a6238c0",
   "metadata": {},
   "source": [
    "# Session info "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec9bbf97-1b78-4609-b46c-e7b7e4652fd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "r.4.4.1-FD20200109SPLENO",
   "language": "R",
   "name": "r.4.4.1-fd20200109spleno"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
