{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b413faa8-5b6b-4220-82a4-eab372ee6abf",
   "metadata": {},
   "source": [
    "# SingleR annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f5cdc9f2-418a-49b6-b649-187356ba46ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Attaching SeuratObject\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘SummarizedExperiment’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Seurat’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘celldex’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:SingleR’:\n",
      "\n",
      "    BlueprintEncodeData, DatabaseImmuneCellExpressionData,\n",
      "    HumanPrimaryCellAtlasData, ImmGenData, MonacoImmuneData,\n",
      "    MouseRNAseqData, NovershternHematopoieticData\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(Seurat)\n",
    "library(SingleCellExperiment)\n",
    "library(SingleR)\n",
    "library(celldex)\n",
    "\n",
    "\n",
    "library(dplyr)\n",
    "library(ggplot2)\n",
    "library(RColorBrewer)\n",
    "library(patchwork)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2d3404c6-2fa3-452d-93bb-5c556ae77ca7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set working directory to project root\n",
    "setwd(gsub(\"/script/seurat\", \"\", getwd()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "09d05d70-aed6-4a07-be32-e17a5a36b9ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Source files\n",
    "source(\"plotting_global.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1c72eaa1-5a55-4dcc-aee0-47a0815f895d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parameter \n",
    "so_file <- \"data/seurat_object/so_qc.rds\"\n",
    "loom_qc_file <- \"data/loom_object/loom_qc\"\n",
    "\n",
    "# Plotting Theme\n",
    "theme_set(theme_global_set()) # From project global source()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15f7527f-21de-449b-b078-a47d127051ce",
   "metadata": {},
   "source": [
    "# Import Seurat object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "46d86fbc-b655-4c4f-b694-25daa172f903",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- readRDS(so_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "908a69b3-5876-4971-9cb7-f87e30f41f81",
   "metadata": {},
   "source": [
    "# Annotate cells with SingleR\n",
    "SingleR identifies marker genes from the reference and uses them to compute assignment scores (based on the Spearman correlation across markers) for each cell in the test dataset against each label in the reference. The label with the highest score is the assigned to the test cell, possibly with further fine-tuning to resolve closely related labels.  \n",
    "\n",
    "first.labels: Labels before fine-tuning\n",
    "labels: Labels after fine-tuning\n",
    "pruning: labels after pruning "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6230bd08-0d72-422e-8c4e-c1a810a0abed",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load celldex Immgen reference\n",
    "ref <- celldex::ImmGenData(ensembl = FALSE)\n",
    "\n",
    "# Seurat to SingleCellExperiment \n",
    "sce <- SingleCellExperiment(list(counts = so@assays$RNA@counts))\n",
    "\n",
    "# Predict labels\n",
    "label_main <- SingleR::SingleR(test = sce, ref = ref, labels = ref$label.main, assay.type.test = \"counts\", de.method = \"classic\")\n",
    "label_fine <- SingleR::SingleR(test = sce, ref = ref, labels = ref$label.fine, assay.type.test = \"counts\", de.method = \"classic\")\n",
    "\n",
    "# Add labels to Seurat object\n",
    "label_main_meta <- as.data.frame(label_main) %>% dplyr::select(pruned.labels, tuning.scores.first) %>% dplyr::rename(main_labels = pruned.labels, main_delta_score = tuning.scores.first)\n",
    "so <- AddMetaData(so, label_main_meta)\n",
    "so$main_labels <- factor(so$main_labels, levels = names(so_color$main_labels))\n",
    "\n",
    "label_fine_meta <- as.data.frame(label_fine) %>% dplyr::select(pruned.labels, tuning.scores.first) %>% dplyr::rename(fine_labels = pruned.labels, fine_delta_score = tuning.scores.first)\n",
    "so <- AddMetaData(so, label_fine_meta)\n",
    "# so$fine_labels <- factor(so$main_labels, levels = names(so_color$main_labels)) # Dont have a sort yet"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "933c5b0c-4837-45e9-85f6-5752d455579a",
   "metadata": {},
   "source": [
    "# Dim Plot "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d8a0577-60a3-4a9c-a628-ab68b3dd5a7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "reduction <- \"rna_umap_nno\"\n",
    "dplot_1 <- DimPlot(so, reduction = reduction, group.by = \"rna_snn_res.0.8\", label = TRUE) & \n",
    "    theme(aspect.ratio = 1, legend.position = \"none\")\n",
    "dplot_2 <- DimPlot(so, reduction = reduction, group.by = \"main_labels\", label = FALSE) & \n",
    "    theme(aspect.ratio = 1, legend.position = \"bottom\") & \n",
    "    scale_color_manual(values = so_color$main_labels, na.value = \"dark gray\") & \n",
    "    guides(color = guide_legend(ncol = 3, override.aes = list(size = 3)))\n",
    "dplot_3 <- DimPlot(so, reduction = reduction, group.by = \"tissue\", label = FALSE) & \n",
    "    theme(aspect.ratio = 1, legend.position = \"bottom\") & \n",
    "    scale_color_manual(values = so_color$tissue, na.value = \"dark gray\") & \n",
    "    guides(color = guide_legend(ncol = 3, override.aes = list(size = 3)))\n",
    "dplot_4 <- DimPlot(so, reduction = reduction, group.by = \"treatment\", label = FALSE) & \n",
    "    theme(aspect.ratio = 1, legend.position = \"bottom\") & \n",
    "    scale_color_manual(values = so_color$treatment, na.value = \"dark gray\") & \n",
    "    guides(color = guide_legend(ncol = 3, override.aes = list(size = 3)))\n",
    "\n",
    "options(repr.plot.width = 10, repr.plot.height = 10)\n",
    "dplot <- dplot_1 + dplot_2 + dplot_3 + dplot_4 + plot_layout(ncol = 2)\n",
    "dplot\n",
    "ggsave(dplot, filename = \"result/plot/seurat/dimplot_rna_1.png\", width = 9, height = 9)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "497af529-5855-4647-a8e7-c869650eeb25",
   "metadata": {},
   "source": [
    "# Save output "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f3bad26-d135-4299-8cc3-80fd361c1836",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(so, so_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e57b1ae-8c82-4c10-b980-a4c56c6b02de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write loom object \n",
    "library(SeuratDisk)\n",
    "as.loom(so, filename = loom_qc_file, verbose = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a94edee-5a05-4ab5-b195-892f1fc76dae",
   "metadata": {},
   "source": [
    "# Session info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "caaadd17-79e6-424a-945f-56651d786208",
   "metadata": {},
   "outputs": [],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "r.4.1.0-FD20200109SPLENO",
   "language": "R",
   "name": "r.4.1.0-fd20200109spleno"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
