{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cd04ed1a-d1ae-441f-9d5e-88c0aef441ca",
   "metadata": {},
   "source": [
    "# SingleR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4e9ed315-3005-46a2-b6bc-bed3a2094875",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "library_load <- suppressMessages(\n",
    "    \n",
    "    suppressWarnings(\n",
    "    \n",
    "        list(\n",
    "\n",
    "            # Seurat \n",
    "            library(Seurat), \n",
    "\n",
    "            # SingleR\n",
    "            library(SingleR), \n",
    "            library(SingleCellExperiment), \n",
    "            \n",
    "            # Reference\n",
    "            library(celldex), \n",
    "\n",
    "            # Data \n",
    "            library(tidyverse), \n",
    "\n",
    "            # Pyhton compatibility\n",
    "            library(reticulate)\n",
    "\n",
    "        )\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "54ccbe8d-8ffc-47af-ae05-7096223241bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "python:         /nobackup/peer/fdeckert/miniconda3/envs/p.3.9.19-FD20200109SPLENO/bin/python\n",
       "libpython:      /nobackup/peer/fdeckert/miniconda3/envs/p.3.9.19-FD20200109SPLENO/lib/libpython3.9.so\n",
       "pythonhome:     /nobackup/peer/fdeckert/miniconda3/envs/p.3.9.19-FD20200109SPLENO:/nobackup/peer/fdeckert/miniconda3/envs/p.3.9.19-FD20200109SPLENO\n",
       "version:        3.9.19 | packaged by conda-forge | (main, Mar 20 2024, 12:50:21)  [GCC 12.3.0]\n",
       "numpy:          /nobackup/peer/fdeckert/miniconda3/envs/p.3.9.19-FD20200109SPLENO/lib/python3.9/site-packages/numpy\n",
       "numpy_version:  1.26.4\n",
       "\n",
       "NOTE: Python version was forced by use_python() function"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Configure reticulate \n",
    "use_condaenv(condaenv='p.3.9.19-FD20200109SPLENO', conda=\"/nobackup/peer/fdeckert/miniconda3/bin/conda\", required=NULL)\n",
    "py_config()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "182e13b6-823c-435b-93c3-ab7817c80e43",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1] ‘1.14.0’"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "packageVersion(\"celldex\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4716eb83-18c8-4136-adff-31fe7b75b817",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "194d3c7e-d8da-4872-83d0-3765a85c45fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_seed <- 42\n",
    "set.seed(random_seed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b86906ba-f0a1-4edb-b39a-5b4b35f5d901",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set working directory to project root\n",
    "setwd(\"/research/peer/fdeckert/FD20200109SPLENO\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "265bff8d-daa3-46c3-8ddd-2a6b1d3a3a22",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Source files\n",
    "source(\"plotting_global.R\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dfe8c56-8c9d-4dc0-a5d4-d6d075b84a27",
   "metadata": {},
   "source": [
    "# Parameter settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f1e02234-e351-4946-b71d-40c09cceb047",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Files \n",
    "raw_rds_file <- \"data/BSA_0355_SM01_10x_SPLENO/ANALYSIS/raw.rds\"\n",
    "\n",
    "# Plotting Theme\n",
    "ggplot2::theme_set(theme_global_set(size_select=1)) # From project global source()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "354a4c0c-2393-4815-a7a8-dd8beb9e3127",
   "metadata": {},
   "source": [
    "# Import Seurat object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b406cedc-cfeb-4376-a330-9305d17c3e62",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- readRDS(raw_rds_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4d3739ac-67c6-4b1f-b272-1f95fd431bdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- subset(so, subset=nCount_RNA>=1000)\n",
    "so <- subset(so, subset=pMt_RNA<=15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "e39cdf35-85ce-4f1a-a806-8c9dba3cb7e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "cnt <- GetAssayData(so, assay=\"RNA\", slot=\"counts\")\n",
    "cnt <- cnt[rowSums(cnt>=3)>=1, ]\n",
    "so <- CreateSeuratObject(counts=cnt, meta.data=so@meta.data, project=\"FD20200109SPLENO\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "2edccc30-0d56-4351-b135-1d70b78834a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- NormalizeData(so, normalization.method=\"LogNormalize\", verbose=FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31ba27ad-c348-4589-bc82-67df6e26a09b",
   "metadata": {},
   "source": [
    "# ImmGen "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "20fac1cd-4802-4195-8e9e-77befbbec1cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load reference data \n",
    "ref <- celldex::ImmGenData()\n",
    "    \n",
    "# Seurat object to SingleCellExperiment\n",
    "sce <- SingleCellExperiment(list(logcounts=GetAssayData(so, assay=\"RNA\", layer=\"data\")))\n",
    "\n",
    "# Predict labels\n",
    "label_main_immgen <- SingleR::SingleR(test=sce, ref=ref, labels=ref$label.main, assay.type.test=\"logcounts\", assay.type.ref=\"logcounts\", de.method=\"classic\", fine.tune=TRUE) %>%\n",
    "    as.data.frame() %>% dplyr::select(pruned.labels, delta.next) %>% dplyr::rename(label_main_immgen=pruned.labels, delta_score_main_immgen=delta.next)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c7531d3-8403-444e-b6c9-d0cca292a610",
   "metadata": {},
   "source": [
    "# Haemopedia Mouse RNAseq from Haemosphere"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be60aff4-1c87-4204-984b-fc25ad992fd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load reference data \n",
    "meta <- read.delim(\"/research/peer/fdeckert/reference/bulkRNAseq/haemosphere/data/Haemopedia-Mouse-RNASeq_samples.txt\") %>% tibble::column_to_rownames(\"sampleId\")\n",
    "cnt <- read.delim(\"/research/peer/fdeckert/reference/bulkRNAseq/haemosphere/data/Haemopedia-Mouse-RNASeq_raw.txt\", row.names=1)[, rownames(meta)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a73b60bd-b078-476a-a947-6fdfe863c026",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Annotate ENSEML id with gene symbol\n",
    "gtf <- rtracklayer::import(\"/research/peer/fdeckert/reference/genome/GRCm38/Mus_musculus.GRCm38.102.gtf\") %>% as.data.frame() \n",
    "\n",
    "convert <- gtf[, c(\"gene_id\", \"gene_name\")] %>% dplyr::distinct() %>% dplyr::group_by(gene_name) %>% dplyr::mutate(n=n()) %>% dplyr::filter(n==1) %>% tibble::column_to_rownames(\"gene_id\") %>% dplyr::select(-n)\n",
    "gene_id <- intersect(rownames(cnt), rownames(convert)) %>% unique()\n",
    "convert <- convert[which(rownames(convert) %in% gene_id), , drop=FALSE]\n",
    "cnt <- cnt[which(rownames(cnt) %in% gene_id), , drop=FALSE]\n",
    "rownames(cnt) <- convert[rownames(cnt), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d164544-eb90-409c-9b9e-ce1097a8c24c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load reference data\n",
    "ref <- SingleCellExperiment(list(counts=cnt), colData=meta)\n",
    "\n",
    "# Seurat object to SingleCellExperiment\n",
    "sce <- SingleCellExperiment(list(counts=GetAssayData(so, assay=\"RNA\", layer=\"counts\")))\n",
    "\n",
    "# Predict labels\n",
    "label_main_haemopedia <- SingleR::SingleR(test=sce, ref=ref, labels=ref$celltype, assay.type.test=\"counts\", assay.type.ref=\"counts\", de.method=\"classic\", fine.tune=TRUE) %>%\n",
    "    as.data.frame() %>% dplyr::select(pruned.labels, delta.next) %>% dplyr::rename(label_main_haemopedia=pruned.labels, delta_score_main_haemopedia=delta.next)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb99b15f-3d74-46f6-8b15-4be587b69413",
   "metadata": {},
   "source": [
    "# Save results "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "475f3b92-ea2a-4f7d-b035-5568a09b8e72",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(cbind(label_main_immgen, label_main_haemopedia), \"result/singler/singler.csv\", row.names=TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3573893c-81fa-4f9d-aef4-764b80bc18c6",
   "metadata": {},
   "source": [
    "# Session info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8650431-fd79-4f68-bc8f-57c8d4decdef",
   "metadata": {},
   "outputs": [],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "r.4.4.1-FD20200109SPLENO",
   "language": "R",
   "name": "r.4.4.1-fd20200109spleno"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
