{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a5ee1f18-3e91-4382-8557-c685d809e82f",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Compute Kernel for CellRank \n",
    "\n",
    "**RNA velocity** [Bergen et al., 2021](https://www.embopress.org/doi/full/10.15252/msb.202110282)  \n",
    "Beyond the scope of computational modeling, the statistical power of the methods depends on the curvature in the phase portrait since a lack of curvature challenges current models to distinguish whether an up- or down-regulation is occurring. The overall curvature of deviation from the steady-state line in the phase portrait is mostly impacted by the ratios of splicing to degradation rates (Box 1), **indicating that statistical inference is limited to genes where splicing is faster or comparable to degradation, while a small ratio would yield straight lines rather than an interpretable curvature.**  \n",
    "\n",
    "**MURK genes** [GitHub](https://github.com/mebarile/Gata1_Erythroid_kinetics/blob/main/Figure_6/search_and_rank_murk_human.ipynb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c66dc3dc-7672-47df-8b08-2c7b2d60a105",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import scanpy.external as sce\n",
    "import scvelo as scv\n",
    "\n",
    "import scipy\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import matplotlib.pyplot as mpl\n",
    "\n",
    "import os\n",
    "import sys\n",
    "\n",
    "import warnings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75fdf0cf-0f11-4d08-a678-61ab38221488",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Root directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbb90621-4619-4a99-bb3d-83f438ae844d",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/research/peer/fdeckert/FD20200109SPLENO')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a82ac18c-db15-4cd5-90fe-9e9ae4869be4",
   "metadata": {},
   "source": [
    "## Custom modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6eb5cad7-bc8d-4573-ad6e-a855e070bc71",
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.append('bin/')\n",
    "from adata_qc import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "004c3dcd-cf1a-4815-b59f-c7a9a7f27c9a",
   "metadata": {},
   "source": [
    "## Settup rpy2 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adec0f57-8ae0-49e8-8812-95643646eee3",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.environ['R_HOME'] = '/nobackup/peer/fdeckert/miniconda3/envs/r.4.4.1-FD20200109SPLENO/lib/R'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4aeec546-beb6-45e0-b84d-b75d730b8ebb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import rpy2\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "313e080e-e566-4d0c-bfb4-8e79a3d43217",
   "metadata": {},
   "source": [
    "## Figures "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b066e56-2bf8-4d2c-a5a0-355ccf7ccc9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(figsize=(5, 5), dpi_save=1200, fontsize=12, frameon=False, facecolor='white')\n",
    "mpl.rcParams['figure.facecolor'] = 'white'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25178657-5e7f-4ddf-87dd-007f4aed89fd",
   "metadata": {},
   "source": [
    "# Import data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44885fc5-ccf4-4255-9cfb-13128af39e96",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad('data/scRNAseq/object/pp.h5ad')\n",
    "adata_v = sc.read_h5ad('data/scRNAseq/object/velocyto.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "157796d7-d20f-4390-99c4-6b6797b3271f",
   "metadata": {},
   "source": [
    "# Transfer velocyto matrices "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6b3e39e-0378-4d27-82a5-d980c6172a73",
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_names = adata.obs_names\n",
    "var_names = adata.var_names.intersection(adata_v.var_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b0554a2-8f54-49b2-9837-7f67e905e0e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[obs_names, var_names]\n",
    "adata_v = adata_v[obs_names, var_names]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ac32793-627f-44e6-8276-0700a8a75be7",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X = adata_v.X\n",
    "adata.layers = adata_v.layers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55c0ec67-915b-40cb-9943-e2debc125850",
   "metadata": {},
   "outputs": [],
   "source": [
    "# UMAP \n",
    "sc.pl.umap(sc.pp.normalize_total(adata.raw.to_adata(), copy=True), color=['celltype_low', 'Cd34', 'Ly6a', 'Procr', 'Kit', 'Spi1', 'Klf1', 'Gata1', 'Gata2', 'Nfe2', 'Fli1', 'Pf4', 'Mpl', 'Gfi1', 'Cebpa', 'Irf8', 'Elane', 'Mcpt8', 'Prss34', 'Cd74', 'Csf1r', 'Car2', 'Dntt', 'Flt3', 'Lyz2'], legend_loc='on data', ncols=5, size=50, use_raw=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fab1fb8d-7371-480e-93ad-f0e805fca8d0",
   "metadata": {},
   "source": [
    "# Store full matrix "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b224f123-ba67-44b6-9f28-1a6fafc12c77",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_tmp = adata.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bce4b56-70ee-4a82-ae3b-9cdd5b1c5e1c",
   "metadata": {},
   "source": [
    "# Helper function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "891d27eb-c810-4200-8cca-eec40906f1e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "###########\n",
    "### PCA ###\n",
    "###########\n",
    "def pca_workflow(adata, n_comps): \n",
    "    \n",
    "    adata = adata.raw.to_adata()\n",
    "    \n",
    "    sc.pp.highly_variable_genes(adata, flavor='seurat_v3', n_top_genes=2000, subset=False)\n",
    "    \n",
    "    sc.pp.normalize_total(adata)\n",
    "    sc.pp.log1p(adata)\n",
    "    \n",
    "    sc.pp.pca(adata, n_comps=n_comps, use_highly_variable=True)\n",
    "    \n",
    "    X_pca = adata.obsm['X_pca']\n",
    "    \n",
    "    return(X_pca)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc29d3c7-dcc2-4192-9820-9b5e34f07465",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################\n",
    "### Velocity kernel ###\n",
    "#######################\n",
    "def vk_workflow(adata, suffix, compute=False):\n",
    "    \n",
    "    if compute: \n",
    "        \n",
    "        # Set genes to include for moment computation \n",
    "        adata.var['m_genes'] = adata.var.index.isin(scv.pp.filter_and_normalize(adata, min_shared_counts=20, subset_highly_variable=False, n_top_genes=None, log=False, copy=True).var_names)\n",
    "        \n",
    "        # Compute PCA \n",
    "        adata.obsm['X_pca'] = pca_workflow(adata.copy(), n_comps=30)\n",
    "        \n",
    "        # Compute neighbors on PCA \n",
    "        sc.pp.neighbors(adata, n_neighbors=30, n_pcs=30, use_rep='X_pca')\n",
    "        \n",
    "        # Normalize X, spliced, and unspliced\n",
    "        sc.pp.normalize_total(adata, layers=['spliced', 'unspliced'], target_sum=1e6)\n",
    "        sc.pp.log1p(adata)\n",
    "\n",
    "        # Subset by var names filter\n",
    "        adata = adata[:, adata.var['m_genes']].copy()\n",
    "\n",
    "        # First-/second-order moments are computed for each cell across its nearest neighbors\n",
    "        scv.pp.moments(adata, n_pcs=None, n_neighbors=None)\n",
    "        scv.tl.recover_dynamics(adata)\n",
    "\n",
    "        # Estimates velocities in a gene-specific manner\n",
    "        scv.tl.velocity(adata, mode='dynamical')\n",
    "            \n",
    "        # Get velocity and top likelihood fit genes \n",
    "        v_df = adata.var\n",
    "        v_df = v_df[v_df.velocity_genes]\n",
    "        v_df = v_df[v_df.fit_likelihood>=0.1]\n",
    "            \n",
    "        v_genes = v_df.index\n",
    "            \n",
    "        # Velocity graph (re-compute)\n",
    "        scv.tl.velocity_graph(adata, gene_subset=v_genes)\n",
    "\n",
    "        # Velocity pseudotime \n",
    "        scv.tl.velocity_pseudotime(adata)\n",
    "            \n",
    "        # Terminal and root cells \n",
    "        scv.tl.recover_latent_time(adata)\n",
    "            \n",
    "        # Fit differential kinetics on velocity genes\n",
    "        # scv.tl.differential_kinetic_test(adata, var_names=adata.var_names[adata.var.velocity_genes], groupby='celltype_low')\n",
    "\n",
    "        # CellRank kernel \n",
    "        from cellrank.kernels import VelocityKernel\n",
    "        vk = VelocityKernel(adata)\n",
    "        \n",
    "        vk.write('data/scRNAseq/object/cellrank/kernel/vk'+suffix+'.pickle', write_adata=True)\n",
    "        \n",
    "    else: \n",
    "        \n",
    "        from cellrank.kernels import VelocityKernel\n",
    "        vk = VelocityKernel.read(fname='data/scRNAseq/object/cellrank/kernel/vk'+suffix+'.pickle')\n",
    "        \n",
    "    return vk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3831079-426e-45b0-a028-f0dc0e40a505",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################\n",
    "### Palantir kernel ###\n",
    "#######################\n",
    "def pk_workflow(adata, suffix, s_genes, t_genes, compute=False):\n",
    "    \n",
    "    if compute:\n",
    "        \n",
    "        # Use the full matrix instead of velocity matrix \n",
    "        adata = adata.raw.to_adata()\n",
    "        adata.raw = adata.copy()\n",
    "        \n",
    "        # Compute PCA \n",
    "        adata.obsm['X_pca'] = pca_workflow(adata.copy(), n_comps=300)\n",
    "        \n",
    "        # Compute neighbors on PCA \n",
    "        sc.pp.neighbors(adata, n_neighbors=30, n_pcs=300, use_rep='X_pca')\n",
    "        \n",
    "        # Palantir diffusion maps based on PCA\n",
    "        sce.tl.palantir(adata, n_components=10, knn=30)\n",
    "\n",
    "        # Normalize for module score computation \n",
    "        sc.pp.normalize_total(adata)\n",
    "        sc.pp.log1p(adata)\n",
    "\n",
    "        # Compute starting cell\n",
    "        for k, v in s_genes.items():\n",
    "            sc.tl.score_genes(adata, v, score_name='t_score_'+ k.lower(), use_raw=False)\n",
    "            t_cell=adata.obs['t_score_'+ k.lower()].idxmax()\n",
    "            s_genes[k] = t_cell\n",
    "        s_id = pd.Series(list(s_genes.keys()), list(s_genes.values()))\n",
    "\n",
    "        # Compute terminal state cells\n",
    "        for k, v in t_genes.items():\n",
    "            sc.tl.score_genes(adata, v, score_name='t_score_'+ k.lower(), use_raw=False)\n",
    "            t_cell=adata.obs['t_score_'+ k.lower()].idxmax()\n",
    "            t_genes[k] = t_cell\n",
    "        t_id = pd.Series(list(t_genes.keys()), list(t_genes.values()))\n",
    "\n",
    "        adata.obs = adata.obs.join(pd.DataFrame({'pt_cells':pd.concat([s_id, t_id]).tolist()}, index=pd.concat([s_id, t_id]).index), how='left')\n",
    "        adata.obs['pt_cells'] = adata.obs['pt_cells'].astype('category')\n",
    "\n",
    "        # Set back raw data  \n",
    "        adata = adata.raw.to_adata()\n",
    "        adata.raw = adata.copy()\n",
    "        \n",
    "        # Palantir results \n",
    "        pr_res = sce.tl.palantir_results(\n",
    "            adata,\n",
    "            early_cell=s_id.index[0], \n",
    "            terminal_states=t_id, \n",
    "            ms_data='X_palantir_multiscale',\n",
    "            num_waypoints=500, \n",
    "            use_early_cell_as_start=False\n",
    "        )\n",
    "        \n",
    "        # Get pseudotime \n",
    "        adata.obs['ppt_pseudotime'] = pr_res.pseudotime\n",
    "        \n",
    "        # Use original data\n",
    "        adata = adata.raw.to_adata()\n",
    "        adata.raw = adata.copy()\n",
    "        \n",
    "\n",
    "        # CellRank kernel \n",
    "        from cellrank.kernels import PseudotimeKernel\n",
    "        pk = PseudotimeKernel(adata, time_key='ppt_pseudotime')\n",
    "\n",
    "        pk.write('data/scRNAseq/object/cellrank/kernel/pk'+suffix+'.pickle', write_adata=True)\n",
    "    \n",
    "    else: \n",
    "        \n",
    "        from cellrank.kernels import PseudotimeKernel\n",
    "        pk = PseudotimeKernel.read(fname='data/scRNAseq/object/cellrank/kernel/pk'+suffix+'.pickle')\n",
    "    \n",
    "    return pk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d77f904-fbdc-460f-ade1-1a1b9825b4f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "###########################\n",
    "### Connectivity kernel ###\n",
    "###########################\n",
    "def ck_workflow(adata, suffix, compute=False):\n",
    "    \n",
    "    if compute: \n",
    "        \n",
    "        # Neighbor graph on latent space \n",
    "        sc.pp.neighbors(adata, n_neighbors=30, n_pcs=None, use_rep='latent')\n",
    "        \n",
    "        # CellRank kernel \n",
    "        from cellrank.kernels import ConnectivityKernel\n",
    "        ck = ConnectivityKernel(adata)\n",
    "        \n",
    "        ck.write('data/scRNAseq/object/cellrank/kernel/ck'+suffix+'.pickle', write_adata=True)\n",
    "        \n",
    "    else: \n",
    "        \n",
    "        from cellrank.kernels import ConnectivityKernel\n",
    "        ck = ConnectivityKernel.read(fname='data/scRNAseq/object/cellrank/kernel/ck'+suffix+'.pickle')\n",
    "        \n",
    "    return ck"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb5ac464-09ec-455c-b7de-5de10eab1fcd",
   "metadata": {},
   "source": [
    "# Kernel Erythroid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29e30844-ab44-4f1a-8086-2463ac7c4d67",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype_ery = [\n",
    "\n",
    "    'MegP', \n",
    "    'MEP',\n",
    "    'Proerythroblast', \n",
    "    'Erythroblast'\n",
    "    \n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96a22bd3-e1f0-4770-96eb-5a5bf999d07d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset by Erythroid cluster \n",
    "adata = adata_tmp[adata_tmp.obs.celltype_low.isin(celltype_ery)].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1d4aa07-31c7-4ede-91b6-4150ac1fe197",
   "metadata": {},
   "source": [
    "## Kernel Erythroid (WT Baseline) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc74ff2d-53ab-4859-b5e7-79413c310d33",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata_i = adata.copy()\n",
    "adata_i = adata_i[(adata_i.obs.celltype_low.isin(celltype_ery)).values & (adata_i.obs.sample_group=='Bl6_NaCl_D6').values].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "140be474-853e-431d-9385-ca62047e0334",
   "metadata": {},
   "outputs": [],
   "source": [
    "vk_i = vk_workflow(adata=adata_i.copy(), suffix='_ery_wt_baseline_d6', compute=False)\n",
    "pk_i = pk_workflow(adata=adata_i.copy(), suffix='_ery_wt_baseline_d6', compute=False, s_genes=dict(Start=['Cd34', 'Kit', 'Gata2']), t_genes=dict(Ery=['Hbb-bt'], Meg=['Pf4']))\n",
    "ck_i = ck_workflow(adata=adata_i.copy(), suffix='_ery_wt_baseline_d6', compute=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f12f10f3-ee39-43d7-9763-4fdac54bf1b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "warnings.simplefilter(\"ignore\")\n",
    "sc.set_figure_params(figsize=(2.5, 5), dpi_save=1200, fontsize=10, color_map='magma', frameon=False)\n",
    "scv.set_figure_params(figsize=(2.5, 5), dpi_save=1200, fontsize=10, color_map='magma', frameon=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73324a47-cce0-479a-afe5-39972da3fe69",
   "metadata": {},
   "outputs": [],
   "source": [
    "scv.pl.velocity_embedding_stream(vk_i.adata, basis='X_umap', color=['celltype_low', 'velocity_pseudotime', 'latent_time', 'root_cells', 'end_points', 'S_score_scale', 'G2M_score_scale'], legend_loc='on data', ncols=7, size=50, density=1.5, recompute=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b82e1173-b302-4e25-b03c-05c941b26795",
   "metadata": {},
   "outputs": [],
   "source": [
    "scv.pl.umap(pk_i.adata, color=['celltype_low', 'ppt_pseudotime', 'pt_cells', 't_score_start', 't_score_meg', 't_score_ery'], legend_loc='on data', ncols=6, size=50)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e444a372-4e01-47de-bcdc-1914804b4c1d",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Kernel Erythroid (WT CpG D6) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0da4b32-152c-47b8-a90f-7aab1c1261a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_i = adata.copy()\n",
    "adata_i = adata_i[(adata_i.obs.celltype_low.isin(celltype_ery)).values & (adata_i.obs.sample_group=='Bl6_CpG_D6').values].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f579848b-7abe-4b79-92b7-11c8ada3154a",
   "metadata": {},
   "outputs": [],
   "source": [
    "vk_i = vk_workflow(adata=adata_i.copy(), suffix='_ery_wt_cpg_d6', compute=False)\n",
    "pk_i = pk_workflow(adata=adata_i.copy(), suffix='_ery_wt_cpg_d6', compute=False, s_genes=dict(Start=['Cd34', 'Kit', 'Gata2']), t_genes=dict(Ery=['Hbb-bt'], Meg=['Pf4']))\n",
    "ck_i = ck_workflow(adata=adata_i.copy(), suffix='_ery_wt_cpg_d6', compute=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75971a8d-4121-4eb2-bbd0-d0f79e0887da",
   "metadata": {},
   "outputs": [],
   "source": [
    "warnings.simplefilter(\"ignore\")\n",
    "sc.set_figure_params(figsize=(2.5, 5), dpi_save=1200, fontsize=10, color_map='magma', frameon=False)\n",
    "scv.set_figure_params(figsize=(2.5, 5), dpi_save=1200, fontsize=10, color_map='magma', frameon=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45b6c8d0-21a8-4692-a47a-e4898e940299",
   "metadata": {},
   "outputs": [],
   "source": [
    "scv.pl.velocity_embedding_stream(vk_i.adata, basis='X_umap', color=['celltype_low', 'velocity_pseudotime', 'latent_time', 'root_cells', 'end_points', 'S_score_scale', 'G2M_score_scale'], legend_loc='on data', ncols=7, size=100, density=1.5, recompute=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5ae3c5f-5ca8-4314-9e7d-c74b40c66ace",
   "metadata": {},
   "outputs": [],
   "source": [
    "scv.pl.umap(pk_i.adata, color=['celltype_low', 'ppt_pseudotime', 'pt_cells', 't_score_start', 't_score_meg', 't_score_ery'], legend_loc='on data', ncols=10, size=100)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7b04c82-c465-46c8-b1ef-de40f819f728",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Kernel Erythroid (IFNAR fl/fl) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58be102a-bab9-495e-a704-52e037c26565",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_i = adata.copy()\n",
    "adata_i = adata_i[(adata_i.obs.celltype_low.isin(celltype_ery)).values & (adata_i.obs.genotype=='IFNAR_fl').values].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7d940fd-c1c2-4859-b7a0-ce2afcf94207",
   "metadata": {},
   "outputs": [],
   "source": [
    "vk_i = vk_workflow(adata=adata_i.copy(), suffix='_ery_ifnar_fl', compute=False)\n",
    "pk_i = pk_workflow(adata=adata_i.copy(), suffix='_ery_ifnar_fl', compute=False, s_genes=dict(Start=['Cd34', 'Kit', 'Gata2']), t_genes=dict(Ery=['Hbb-bt'], Meg=['Pf4']))\n",
    "ck_i = ck_workflow(adata=adata_i.copy(), suffix='_ery_ifnar_fl', compute=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7521081b-88d9-4cd4-9ab5-89ff3859dff1",
   "metadata": {},
   "outputs": [],
   "source": [
    "warnings.simplefilter(\"ignore\")\n",
    "sc.set_figure_params(figsize=(2.5, 5), dpi_save=1200, fontsize=10, color_map='magma', frameon=False)\n",
    "scv.set_figure_params(figsize=(2.5, 5), dpi_save=1200, fontsize=10, color_map='magma', frameon=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bb18f48-c9bf-4260-8d03-a4aa6b5e8f1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "scv.pl.velocity_embedding_stream(vk_i.adata, basis='X_umap', color=['celltype_low', 'velocity_pseudotime', 'latent_time', 'root_cells', 'end_points', 'S_score_scale', 'G2M_score_scale'], legend_loc='on data', ncols=7, size=100, density=1.5, recompute=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1351f05-84f9-4c6c-9e6e-da58b5645c31",
   "metadata": {},
   "outputs": [],
   "source": [
    "scv.pl.umap(pk_i.adata, color=['celltype_low', 'ppt_pseudotime', 'pt_cells', 't_score_start', 't_score_meg', 't_score_ery'], legend_loc='on data', ncols=10, size=100)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0cedcd0-d060-41b1-9c7a-d21c0aa9d06a",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Kernel Erythroid (IFNAR fl/fl LyzM Cre) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e68f5809-4a9a-4b58-a4aa-addd1cdb07fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_i = adata.copy()\n",
    "adata_i = adata_i[(adata_i.obs.celltype_low.isin(celltype_ery)).values & (adata_i.obs.genotype=='IFNAR_fl_LysM_cre').values].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "923a64b3-7809-42d3-b05e-85682e0f51b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "vk_i = vk_workflow(adata=adata_i.copy(), suffix='_ery_ifnar_fl_lysm_cre', compute=False)\n",
    "pk_i = pk_workflow(adata=adata_i.copy(), suffix='_ery_ifnar_fl_lysm_cre', compute=False, s_genes=dict(Start=['Cd34', 'Kit', 'Gata2']), t_genes=dict(Ery=['Hbb-bt'], Meg=['Pf4']))\n",
    "ck_i = ck_workflow(adata=adata_i.copy(), suffix='_ery_ifnar_fl_lysm_cre', compute=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7143769-17f7-4c69-9fa9-bad953881b55",
   "metadata": {},
   "outputs": [],
   "source": [
    "warnings.simplefilter(\"ignore\")\n",
    "sc.set_figure_params(figsize=(2.5, 5), dpi_save=1200, fontsize=10, color_map='magma', frameon=False)\n",
    "scv.set_figure_params(figsize=(2.5, 5), dpi_save=1200, fontsize=10, color_map='magma', frameon=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f74ebdd-209c-4e63-820b-0c038454e2d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "scv.pl.velocity_embedding_stream(vk_i.adata, basis='X_umap', color=['celltype_low', 'velocity_pseudotime', 'latent_time', 'root_cells', 'end_points', 'S_score_scale', 'G2M_score_scale'], legend_loc='on data', ncols=7, size=100, density=1.5, recompute=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52e4521a-56bc-4271-bc18-057331f54190",
   "metadata": {},
   "outputs": [],
   "source": [
    "scv.pl.umap(pk_i.adata, color=['celltype_low', 'ppt_pseudotime', 'pt_cells', 't_score_start', 't_score_meg', 't_score_ery'], legend_loc='on data', ncols=10, size=100)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cdd6f6a5-6574-4dd0-aedf-93000cd48bf8",
   "metadata": {},
   "source": [
    "# Kernel Myeloid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c0d993a-f15c-41e2-97fc-6ec643bb93bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype_m = [\n",
    "\n",
    "    'cMo', \n",
    "    'intMo',\n",
    "    'ncMo', \n",
    "    'RPM'\n",
    "    \n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92e5830a-0191-4e20-9650-4792626d5852",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset by Erythroid cluster \n",
    "adata = adata_tmp[adata_tmp.obs.celltype_low.isin(celltype_m)].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27015653-cad8-4c11-99d6-9d2d61a7c3a2",
   "metadata": {},
   "source": [
    "## Kernel Myeloid (WT Baseline) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0569f5c6-3ac7-4cfa-bc89-6bc39bff7900",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "adata_i = adata.copy()\n",
    "adata_i = adata_i[(adata_i.obs.celltype_low.isin(celltype_m)).values & (adata_i.obs.sample_group=='Bl6_NaCl_D6').values].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9980967b-2cf0-41d8-a4a5-89b612affebb",
   "metadata": {},
   "outputs": [],
   "source": [
    "vk_i = vk_workflow(adata=adata_i.copy(), suffix='_m_wt_baseline_d6', compute=True)\n",
    "pk_i = pk_workflow(adata=adata_i.copy(), suffix='_m_wt_baseline_d6', compute=True, s_genes=dict(Start=['Ly6c2']), t_genes=dict(ncMo=['Cx3cr1', 'Nr4a1'], RPM=['Spic'], moDC=['Flt3']))\n",
    "ck_i = ck_workflow(adata=adata_i.copy(), suffix='_m_wt_baseline_d6', compute=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "570d89da-eea9-4f6d-ba74-d790ddb000c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "warnings.simplefilter(\"ignore\")\n",
    "sc.set_figure_params(figsize=(2.5, 5), dpi_save=1200, fontsize=10, color_map='magma', frameon=False)\n",
    "scv.set_figure_params(figsize=(2.5, 5), dpi_save=1200, fontsize=10, color_map='magma', frameon=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "226f346f-cca9-4027-9148-2ae83b8beb22",
   "metadata": {},
   "outputs": [],
   "source": [
    "scv.pl.velocity_embedding_stream(vk_i.adata, basis='X_umap', color=['celltype_low', 'velocity_pseudotime', 'latent_time', 'root_cells', 'end_points', 'S_score_scale', 'G2M_score_scale'], legend_loc='on data', ncols=7, size=50, density=1.5, recompute=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb18a2e1-148a-47ce-af36-cb9f4ed7092e",
   "metadata": {},
   "outputs": [],
   "source": [
    "scv.pl.umap(pk_i.adata, color=['celltype_low', 'ppt_pseudotime', 'pt_cells', 't_score_start', 't_score_ncmo', 't_score_rpm', 't_score_modc'], legend_loc='on data', ncols=7, size=50)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ed5b35e-e8db-4b91-b657-31f459d9ffce",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Kernel Myeloid (WT CpG D6) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7bc1e74-4381-4f12-bbb9-05034195ae08",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_i = adata.copy()\n",
    "adata_i = adata_i[(adata_i.obs.celltype_low.isin(celltype_m)).values & (adata_i.obs.sample_group=='Bl6_CpG_D6').values].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "260b1a37-c799-4d37-b1f1-0a4f3f2347d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "vk_i = vk_workflow(adata=adata_i.copy(), suffix='_m_wt_cpg_d6', compute=True)\n",
    "pk_i = pk_workflow(adata=adata_i.copy(), suffix='_m_wt_cpg_d6', compute=True, s_genes=dict(Start=['Ly6c2']), t_genes=dict(ncMo=['Cx3cr1', 'Nr4a1'], RPM=['Spic'], moDC=['Flt3']))\n",
    "ck_i = ck_workflow(adata=adata_i.copy(), suffix='_m_wt_cpg_d6', compute=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98a17ac6-d868-403b-9727-3a03dd517e5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "warnings.simplefilter(\"ignore\")\n",
    "sc.set_figure_params(figsize=(2.5, 5), dpi_save=1200, fontsize=10, color_map='magma', frameon=False)\n",
    "scv.set_figure_params(figsize=(2.5, 5), dpi_save=1200, fontsize=10, color_map='magma', frameon=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0ee17c5-1b71-4eaf-826e-cd4b817a3322",
   "metadata": {},
   "outputs": [],
   "source": [
    "scv.pl.velocity_embedding_stream(vk_i.adata, basis='X_umap', color=['celltype_low', 'velocity_pseudotime', 'latent_time', 'root_cells', 'end_points', 'S_score_scale', 'G2M_score_scale'], legend_loc='on data', ncols=7, size=100, density=1.5, recompute=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b5c41a3-ddcf-4bdd-8897-31a7174d1e6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "scv.pl.umap(pk_i.adata, color=['celltype_low', 'ppt_pseudotime', 'pt_cells', 't_score_start', 't_score_ncmo', 't_score_rpm', 't_score_modc'], legend_loc='on data', ncols=7, size=50)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb73efa8-aebb-485a-81d9-06a540927d00",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Kernel Myeloid (IFNAR fl/fl) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba3e2c68-9896-4079-8730-4d0f4d40af59",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_i = adata.copy()\n",
    "adata_i = adata_i[(adata_i.obs.celltype_low.isin(celltype_m)).values & (adata_i.obs.genotype=='IFNAR_fl').values].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b463f3d9-3fd0-4d5d-b737-f932892a0283",
   "metadata": {},
   "outputs": [],
   "source": [
    "vk_i = vk_workflow(adata=adata_i.copy(), suffix='_m_ifnar_fl', compute=True)\n",
    "pk_i = pk_workflow(adata=adata_i.copy(), suffix='_m_ifnar_fl', compute=True, s_genes=dict(Start=['Ly6c2']), t_genes=dict(ncMo=['Cx3cr1', 'Nr4a1'], RPM=['Spic'], moDC=['Flt3']))\n",
    "ck_i = ck_workflow(adata=adata_i.copy(), suffix='_m_ifnar_fl', compute=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6aabb3c3-0f2c-42e8-9eb5-7d70bfffab2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "warnings.simplefilter(\"ignore\")\n",
    "sc.set_figure_params(figsize=(2.5, 5), dpi_save=1200, fontsize=10, color_map='magma', frameon=False)\n",
    "scv.set_figure_params(figsize=(2.5, 5), dpi_save=1200, fontsize=10, color_map='magma', frameon=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73873aec-e589-474b-8e8a-ebfc75a240de",
   "metadata": {},
   "outputs": [],
   "source": [
    "scv.pl.velocity_embedding_stream(vk_i.adata, basis='X_umap', color=['celltype_low', 'velocity_pseudotime', 'latent_time', 'root_cells', 'end_points', 'S_score_scale', 'G2M_score_scale'], legend_loc='on data', ncols=7, size=100, density=1.5, recompute=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a06cb940-c772-4408-862d-e48a200a0658",
   "metadata": {},
   "outputs": [],
   "source": [
    "scv.pl.umap(pk_i.adata, color=['celltype_low', 'ppt_pseudotime', 'pt_cells', 't_score_start', 't_score_ncmo', 't_score_rpm', 't_score_modc'], legend_loc='on data', ncols=7, size=50)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f10a0139-e422-4364-8431-32c31cc9d754",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Kernel Myeloid (IFNAR fl/fl LyzM Cre) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0adaf305-ae82-4a14-a4fd-10e067134d83",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_i = adata.copy()\n",
    "adata_i = adata_i[(adata_i.obs.celltype_low.isin(celltype_m)).values & (adata_i.obs.genotype=='IFNAR_fl_LysM_cre').values].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "277d8849-22e0-4aa4-915f-ced09f07e25e",
   "metadata": {},
   "outputs": [],
   "source": [
    "vk_i = vk_workflow(adata=adata_i.copy(), suffix='_m_ifnar_fl_lysm_cre', compute=True)\n",
    "pk_i = pk_workflow(adata=adata_i.copy(), suffix='_m_ifnar_fl_lysm_cre', compute=True, s_genes=dict(Start=['Ly6c2']), t_genes=dict(ncMo=['Cx3cr1', 'Nr4a1'], RPM=['Spic'], moDC=['Flt3']))\n",
    "ck_i = ck_workflow(adata=adata_i.copy(), suffix='_m_ifnar_fl_lysm_cre', compute=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2a857d1-1b44-41b0-8656-2a3b6ec0352a",
   "metadata": {},
   "outputs": [],
   "source": [
    "warnings.simplefilter(\"ignore\")\n",
    "sc.set_figure_params(figsize=(2.5, 5), dpi_save=1200, fontsize=10, color_map='magma', frameon=False)\n",
    "scv.set_figure_params(figsize=(2.5, 5), dpi_save=1200, fontsize=10, color_map='magma', frameon=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a0a18b0-b012-493f-bbd8-9ba772822a3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "scv.pl.velocity_embedding_stream(vk_i.adata, basis='X_umap', color=['celltype_low', 'velocity_pseudotime', 'latent_time', 'root_cells', 'end_points', 'S_score_scale', 'G2M_score_scale'], legend_loc='on data', ncols=7, size=100, density=1.5, recompute=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfe066c1-c33f-4aec-8517-a488f2587f70",
   "metadata": {},
   "outputs": [],
   "source": [
    "scv.pl.umap(pk_i.adata, color=['celltype_low', 'ppt_pseudotime', 'pt_cells', 't_score_start', 't_score_ncmo', 't_score_rpm', 't_score_modc'], legend_loc='on data', ncols=7, size=50)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "p.3.10.16-FD20200109SPLENO",
   "language": "python",
   "name": "p.3.10.16-fd20200109spleno"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
