{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8dab95dd-2e8e-472e-b1ef-e9566cd08bfb",
   "metadata": {},
   "source": [
    "# Differential exprssion analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75fd7b70-81de-41ed-b00b-54dd26951307",
   "metadata": {},
   "outputs": [],
   "source": [
    "library_load <- suppressMessages(\n",
    "    \n",
    "    suppressWarnings(\n",
    "        \n",
    "        list(\n",
    "        \n",
    "            # DEA \n",
    "            library(DESeq2), \n",
    "            library(IHW), # Independent Hypothesis Weighting\n",
    "            library(edgeR), # For CPM \n",
    "            \n",
    "            # Bioconductor tools\n",
    "            library(rtracklayer), # readGFF, export\n",
    "            library(Rsubread), # featureCounts\n",
    "            \n",
    "            # Data\n",
    "            library(dplyr), \n",
    "            library(rtracklayer), \n",
    "            \n",
    "            # Plotting \n",
    "            library(ggplot2), \n",
    "            library(patchwork), \n",
    "            library(ComplexHeatmap), \n",
    "            library(grid) # For gpar\n",
    "            \n",
    "        )\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7ee4d36-c8a2-4cf8-a1a0-bfa6233f4b12",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c965fc68-e06b-404d-b9b7-6da6d498a9f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set working directory to project root\n",
    "setwd(\"C:/Users/Florian Deckert/Projects/FD20200109SPLENO/\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d5d13bf-d70f-405b-b7b8-a96b73cbd542",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting Theme\n",
    "source(\"plotting_global.R\")\n",
    "ggplot2::theme_set(theme_global_set(size_select=1)) # From project global source()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f90310d-559d-44a3-9939-3a0593b92d0b",
   "metadata": {},
   "source": [
    "# Parameter "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7023a517-7859-49c1-b586-e1a1a3409e1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "cache <- TRUE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9179c5f4-a7d2-4d61-a846-9a3cf4a06789",
   "metadata": {},
   "outputs": [],
   "source": [
    "gft_file <- \"/research/peer/fdeckert/reference/genome/GRCm39/gencode.vM33.annotation.gtf.gz\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e12b975-c74d-4509-979f-a38bd9c70253",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Create count matrix from BSF STAR output "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "409c196b-1bf0-4a6c-86e2-669ffe662842",
   "metadata": {},
   "outputs": [],
   "source": [
    "star_bam <- list.files(\"/nobackup/peer/fdeckert/FD20200109SPLENO/data/BSA_0885_FACS_FD/\", pattern=\".bam$\", full.names=TRUE, recursive=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16e424b2-94f8-4b50-ab7d-43a6073c2c33",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(!cache) {\n",
    "    \n",
    "    feature_count <- featureCounts(\n",
    "    \n",
    "        star_bam,\n",
    "\n",
    "        # Annotation\n",
    "        annot.ext=gft_file, \n",
    "        isGTFAnnotationFile=TRUE, \n",
    "\n",
    "        # Paired-end mode \n",
    "        isPairedEnd=FALSE, \n",
    "\n",
    "        # Feature type\n",
    "        GTF.featureType=\"gene\",\n",
    "\n",
    "        # Strand specificity\n",
    "        strandSpecific=1, \n",
    "\n",
    "        # read filtering\n",
    "        ignoreDup=TRUE\n",
    "        \n",
    "    \n",
    "    )\n",
    "    \n",
    "    saveRDS(feature_count, \"data/bulkRNAseq/object/feature_count.rds\")\n",
    "\n",
    "} else {\n",
    "    \n",
    "    feature_count <- readRDS(\"data/bulkRNAseq/object/feature_count.rds\")\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "622921ec-256d-477a-a30b-c28206eef5f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# feature_count <- read.table(\"data/bulkRNAseq/object/cnt_bsf.tsv\", head=TRUE)\n",
    "# feature_count <- rbind(\n",
    "    \n",
    "#     feature_count %>% dplyr::select(-source, -gene_id, -gene_type, -level, -mgi_id, -havana_gene, -location) %>% dplyr::group_by(gene_name) %>% dplyr::filter(n()==1), \n",
    "#     feature_count %>% dplyr::select(-source, -gene_id, -gene_type, -level, -mgi_id, -havana_gene, -location) %>% dplyr::group_by(gene_name) %>% dplyr::filter(n()>1) %>% summarise(across(where(is.numeric), sum, na.rm=TRUE), .groups=\"drop\")\n",
    "\n",
    "# ) %>% na.omit() %>% tibble::column_to_rownames(\"gene_name\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9e3fcc8-d8d3-431b-a115-e26fefb15540",
   "metadata": {},
   "source": [
    "# Import BSF DESeq2 object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1408dad1-7a89-44ba-9473-bd829bf6ea1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare meta data \n",
    "meta <- read.csv(\"/nobackup/peer/fdeckert/FD20200109SPLENO/data/BSA_0885_FACS_FD/BSA_0884_FACS.csv\", row.names=2)\n",
    "meta <- meta[grep(\"FD\", rownames(meta), value=TRUE), ]\n",
    "meta$sample_name <- rownames(meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc3c7562-8ecf-4998-bb6b-5dabeb435ded",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove monocyte low input sample \n",
    "meta <- meta[meta$sample_name!=\"FD_80\", ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f839bf54-f599-44cf-92e7-358150a10881",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare count matrix \n",
    "# cnt <- feature_count[[\"counts\"]]\n",
    "cnt <- feature_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78de3792-f746-40cb-964b-1a5e41f8a3cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store full bam file names \n",
    "bam_files <- colnames(cnt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9417a27a-c1d1-4b9a-899c-0346046c106b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace BAM file names with sample names \n",
    "colnames(cnt) <- gsub(\"_0|_00\", \"_\", colnames(cnt))\n",
    "colnames(cnt) <- gsub(\"star_sample_|_S.*\", \"\", colnames(cnt), fixed=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db5cacf9-f950-4c93-b4b8-89fd24b9e69b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add BAM path to meta \n",
    "meta <- dplyr::left_join(meta, data.frame(sample_name=colnames(cnt), bam_path=paste0(\"/nobackup/peer/fdeckert/FD20200109SPLENO/data/BSA_0885_FACS_FD/mm39_ERCC92/\", gsub(\".bam\", \"\", basename(bam_files)), \"/\", bam_files)), by = join_by(sample_name))\n",
    "rownames(meta) <- meta$sample_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d037d2f-310b-42fa-823a-4f14c7a6fec6",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta <- meta[intersect(rownames(meta), colnames(cnt)), ]\n",
    "cnt <- cnt[, intersect(rownames(meta), colnames(cnt))]\n",
    "all(rownames(meta)==colnames(cnt))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a73d078-ae78-419b-acad-48da73e08b79",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta <- meta[, c(\"sample_name\", \"Cell.Type\", \"Disease\", \"Disease\", \"Genotype\", \"Material.Type\", \"bam_path\")]\n",
    "colnames(meta) <- c(\"sample_name\", \"celltype\", \"disease\", \"dpi\", \"genotype\", \"lib_protocol\", \"bam_path\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41e2c2ce-ba30-4fc2-ae90-df0bfd7ff6ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta <- dplyr::mutate(\n",
    "    \n",
    "    meta, celltype=recode(celltype, \"Monocyte\"=\"cMo\", \"preRPM\"=\"PreRPM\", \"RPM\"=\"RPM\", \"EB\"=\"EB\"), \n",
    "    disease=recode(disease, \"Baseline\"=\"Baseline\", \"CpG_D1\"=\"CpG\", \"CpG_D3\"=\"CpG\", \"CpG_D7\"=\"CpG\"), \n",
    "    dpi=recode(dpi, \"Baseline\"=\"D0\", \"CpG_D1\"=\"D1\", \"CpG_D3\"=\"D3\", \"CpG_D7\"=\"D6\"), \n",
    "    genotype=recode(genotype, \"wt/wt\"=\"IFNAR_fl\", \"cre/+\"=\"IFNAR_fl_LysM_cre\"), \n",
    "    lib_protocol=recode(lib_protocol, \"RNA low input\"=\"low_input\", \"RNA normal input >10ng\"=\"normal_input\")\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbb74cbb-2dda-4cff-87d5-2e52325c1228",
   "metadata": {},
   "source": [
    "## Set factor levels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51e8af53-f389-468a-a167-a353cd2bc7ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set factor levels \n",
    "meta$celltype <- factor(meta$celltype, levels=c(\"cMo\", \"PreRPM\", \"RPM\", \"EB\"))\n",
    "meta$disease <- factor(meta$disease, levels=c(\"Baseline\", \"CpG\"))\n",
    "meta$dpi <- factor(meta$dpi, levels=c(\"D0\", \"D1\", \"D3\", \"D6\"))\n",
    "meta$genotype <- factor(meta$genotype, levels=c(\"IFNAR_fl\", \"IFNAR_fl_LysM_cre\"))\n",
    "meta$lib_protocol <- factor(meta$lib_protocol, levels=c(\"low_input\", \"normal_input\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f989559-fbfe-4286-a9c7-43673eddf295",
   "metadata": {},
   "source": [
    "## Check count matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6bb477d-59bb-4a7c-a483-0ee5fcfff259",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter genes with 0 counts \n",
    "cnt <- cnt[rowSums(cnt)>0, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3aa6e39-d135-4f4f-a5a1-4cc2e6ed4f66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter out spike in \n",
    "cnt <- cnt[grepl(\"ENSMUSG\", rownames(cnt)), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e629d93a-78bf-4b37-a3af-f80a8203517b",
   "metadata": {},
   "outputs": [],
   "source": [
    "cnt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dff66949-4681-4b18-8195-b00e70c72111",
   "metadata": {},
   "source": [
    "## Order data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fa3120a-8c16-4bc0-a2a7-c89ada08c5e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reset order of samples in cnt and meta\n",
    "cnt <- cnt[, intersect(colnames(cnt), rownames(meta))]\n",
    "meta <- meta[intersect(colnames(cnt), rownames(meta)), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e2e7cb7-0b13-456e-b368-905c524cddce",
   "metadata": {},
   "outputs": [],
   "source": [
    "all(rownames(meta)==colnames(cnt))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97400128-0682-407d-bc6d-f7aa7ea405f6",
   "metadata": {},
   "source": [
    "## Compute library size "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9a2cfa0-4a15-4a8f-accc-89cc6d0909a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get data frame with library size per sample \n",
    "library_size <- data.frame(sample_name=colnames(cnt), library_size=colSums(cnt))\n",
    "\n",
    "# Use the sample_name column to combine meta data with library size. \n",
    "meta <- dplyr::left_join(meta, library_size, by=\"sample_name\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b283934d-920b-4def-b1b5-cfafff9b00c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dplyr removes rownames so re-set\n",
    "rownames(meta) <- meta$sample_name"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f33ed0c1-9eb3-4aba-89e4-d71fad43df60",
   "metadata": {},
   "source": [
    "# Convert from ENSEML ID to MGI Symbol "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e06078c-3ae9-4706-b7e0-60ba70aa05f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import_convert <- function(gtf_path=gft_file) {\n",
    "    \n",
    "    # Load reference gtf \n",
    "    gtf <- rtracklayer::import(gtf_path)\n",
    "    gtf <- as.data.frame(gtf)\n",
    "    # Subset data \n",
    "    convert <- gtf[, c(\"gene_id\", \"gene_name\")] \n",
    "    # Rename columns \n",
    "    colnames(convert) <- c(\"ensembl_gene_id\", \"mgi_symbol\")\n",
    "    # Select unique entries \n",
    "    convert <- unique(convert)\n",
    "    convert <- na.omit(convert)\n",
    "\n",
    "    message(paste(\"Total number of ENSEMBL ID matched with MGI SYMBOL:\", nrow(convert)))\n",
    "    \n",
    "    # Filter convert table for unique entries\n",
    "    ensembl_gene_id_multiple <- table(convert$ensembl_gene_id)\n",
    "    ensembl_gene_id_multiple <- ensembl_gene_id_multiple[ensembl_gene_id_multiple>1]\n",
    "    convert <- convert[!convert$ensembl_gene_id %in% names(ensembl_gene_id_multiple), ]\n",
    "    message(paste(\"Unique ENSEMBL ID:\", nrow(convert)))\n",
    "\n",
    "    mgi_symbol_multiple <- table(convert$mgi_symbol)\n",
    "    mgi_symbol_multiple <- mgi_symbol_multiple[mgi_symbol_multiple>1]\n",
    "    convert <- convert[!convert$mgi_symbol %in% names(mgi_symbol_multiple), ]\n",
    "    message(paste(\"Unique MGI SYMBOL:\", nrow(convert)))\n",
    "    \n",
    "    return(convert)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc66c9e0-6291-4450-b0b4-924303055c75",
   "metadata": {},
   "outputs": [],
   "source": [
    "convert <- import_convert()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55ef5b4f-b4e6-4909-a2e6-104917c86b35",
   "metadata": {},
   "outputs": [],
   "source": [
    "ensembl_to_mgi <- function(data, convert) {\n",
    "     \n",
    "    # Filter data and convert by shared ensemble_gene_id \n",
    "    rownames(convert) <- convert$ensembl_gene_id\n",
    "    data <- data[rownames(data) %in% rownames(convert), ]\n",
    "    convert <- convert[rownames(convert) %in% rownames(data), ]\n",
    "    convert <- convert[rownames(data), ]\n",
    "\n",
    "    # Convert ensemble_gene_id to mgi_symbol\n",
    "    rownames(data) <- convert$mgi_symbol\n",
    "    \n",
    "    # Add full annotation \n",
    "    data$mgi_symbol <- convert$mgi_symbol\n",
    "    data$ensembl_gene_id <- rownames(convert)\n",
    "    \n",
    "    return(data)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc8bcc79-f067-4e0a-9369-b418de4bcb02",
   "metadata": {},
   "source": [
    "# Subset counts to enseble IDs present in the converter "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "caecdb41-d20f-44f1-8c8d-167e12fb5119",
   "metadata": {},
   "outputs": [],
   "source": [
    "message(paste(\"ENSEMBL ID from count data not fround in ensemble converter:\", table(rownames(cnt) %in% rownames(convert))[1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ceecfd2f-e6b6-41a4-87a9-1db59ada3790",
   "metadata": {},
   "outputs": [],
   "source": [
    "cnt <- cnt[rownames(cnt) %in% convert[, 1], ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15e9ca7a-3ddd-46c9-b4b1-5716f708ceb2",
   "metadata": {},
   "source": [
    "# Plot library size "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a99d017-a6f6-44f0-a877-282962074c15",
   "metadata": {},
   "outputs": [],
   "source": [
    "# First order the samples by celltype and dpi. Then set the sample_name as factor leves. That way in the folloing plot the samples will be nicely ordered. \n",
    "meta <- dplyr::arrange(meta, celltype)\n",
    "meta$sample_name <- factor(meta$sample_name, levels=meta$sample_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdc8c2ca-4888-4b83-bc96-4823c230eb65",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=25, repr.plot.height=5)\n",
    "\n",
    "# Library sizes counts\n",
    "ggplot(meta, aes(x=sample_name, y=library_size, fill=celltype)) + \n",
    "    geom_bar(stat=\"identity\") +\n",
    "    ggtitle(\"Library size\") + xlab(\"Sample\") + ylab(\"Library size [counts]\") + \n",
    "    # scale_fill_manual(values=color$cell_type_main) +\n",
    "    facet_grid(~dpi, scales=\"free_x\", space=\"free\") + \n",
    "    theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3213094-e2fa-42cb-8933-60b6da0f1837",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=25, repr.plot.height=5)\n",
    "\n",
    "# Library sizes log10(counts)\n",
    "ggplot(meta, aes(x=sample_name, y=log10(library_size), fill=celltype)) + \n",
    "    geom_bar(stat=\"identity\") +\n",
    "    ggtitle(\"Library size\") + xlab(\"Sample\") + ylab(\"Library size [log10(counts)]\") + \n",
    "    # scale_fill_manual(values=color$cell_type_main) +\n",
    "    facet_grid(~dpi, scales=\"free_x\", space=\"free\") + \n",
    "    theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51f5d69e-8c71-42ee-ad45-1ca4cb44412c",
   "metadata": {},
   "source": [
    "## Gene count distribution "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb81ce53-ecb3-4fbf-b634-91c33160411a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For plotting we need to convert the counts from wide to long format (http://www.cookbook-r.com/Manipulating_data/Converting_data_between_wide_and_long_format/)\n",
    "data <- reshape2::melt(cnt, varnames=c(\"ENSEMBL_ID\", \"sample_name\"), value.name=\"counts\")\n",
    "\n",
    "# Combine with meta data \n",
    "data <- dplyr::left_join(data, meta, by=\"sample_name\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "174b61ed-0c84-4523-b62c-82dc4ce6b97b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Again, set factor levels for plotting  \n",
    "data <- dplyr::arrange(data, celltype)\n",
    "data$sample_name <- factor(data$sample_name, levels=unique(data$sample_name)) # The sample names are now repeated and we need to use unique for the levels. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d8749b3-d6c3-4b44-b249-c9af39260ef9",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=25, repr.plot.height=5)\n",
    "\n",
    "ggplot(data, aes(x=sample_name, y=log2(counts), fill=celltype)) + \n",
    "    geom_boxplot() +\n",
    "    ggtitle(\"Gene count distribution\") + xlab(\"Sample\") + ylab(\"Gene count [log2(counts)]\") + \n",
    "    # scale_fill_manual(values=color$cell_type_main) + \n",
    "    facet_grid(~dpi, scales=\"free_x\", space=\"free\") + \n",
    "    theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df022d79-efef-4406-9704-cac583289bf9",
   "metadata": {},
   "source": [
    "# Ensure order of data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7931de88-0bb2-4df4-825e-da189ae2b71b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reset order of samples in cnt and meta\n",
    "cnt <- cnt[, intersect(colnames(cnt), rownames(meta))]\n",
    "meta <- meta[intersect(colnames(cnt), rownames(meta)), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c65a469-9769-4d2d-afdb-f1ffd24a7323",
   "metadata": {},
   "outputs": [],
   "source": [
    "all(rownames(meta)==colnames(cnt))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e92ab03f-5f70-47f8-b255-2b9381e8b1a3",
   "metadata": {},
   "source": [
    "# Principal component analysis "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f39f5029-48ee-4fbf-94d6-2972a5d6de4e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pca_plot <- function(pca_data) {\n",
    "    \n",
    "    pca_plot_1 <- ggplot(pca_data, aes(PC1, PC2, color=celltype, fill=celltype, shape=genotype)) + \n",
    "        geom_point(size=5, stroke=1) + \n",
    "        ggtitle(\"PCA (Celltype)\") + \n",
    "        xlab(paste0(\"PC1: \", round(100 * attr(pca_data, \"percentVar\"))[1],\"% variance\")) +\n",
    "        ylab(paste0(\"PC2: \", round(100 * attr(pca_data, \"percentVar\"))[2],\"% variance\")) + \n",
    "        # scale_color_manual(\"Cell type\", values=color$cell_type_main) +\n",
    "        # scale_fill_manual(\"Cell type\", values=color$cell_type_main) + \n",
    "        scale_shape_manual(\"Genotype\", values=setNames(c(21, 24), c(\"IFNAR_fl\", \"IFNAR_fl_LysM_cre\"))) + \n",
    "        theme(aspect.ratio=1, legend.position=\"right\") + guides(fill=guide_legend(byrow=TRUE)) + guides(color=guide_legend(byrow=TRUE)) + guides(shape=guide_legend(byrow=TRUE))\n",
    "\n",
    "    pca_plot_2 <- ggplot(pca_data, aes(PC1, PC2, color=dpi, fill=dpi, shape=genotype)) + \n",
    "        geom_point(size=5, stroke=1) + \n",
    "        ggtitle(\"PCA (DPI)\") + \n",
    "        xlab(paste0(\"PC1: \", round(100 * attr(pca_data, \"percentVar\"))[1],\"% variance\")) +\n",
    "        ylab(paste0(\"PC2: \", round(100 * attr(pca_data, \"percentVar\"))[2],\"% variance\")) + \n",
    "        scale_color_manual(\"DPI\", values=color$dpi) +\n",
    "        scale_fill_manual(\"DPI\", values=color$dpi) + \n",
    "        scale_shape_manual(\"Genotype\", values=setNames(c(21, 24), c(\"IFNAR_fl\", \"IFNAR_fl_LysM_cre\"))) + \n",
    "        theme(aspect.ratio=1, legend.position=\"right\") + guides(fill=guide_legend(byrow=TRUE)) + guides(color=guide_legend(byrow=TRUE)) + guides(shape=guide_legend(byrow=TRUE))\n",
    "\n",
    "    pca_plot_3 <- ggplot(pca_data, aes(PC1, PC2, color=lib_protocol, fill=lib_protocol, shape=genotype)) + \n",
    "        geom_point(size=5, stroke=1) + \n",
    "        ggtitle(\"PCA (Library protocol)\") + \n",
    "        xlab(paste0(\"PC1: \", round(100 * attr(pca_data, \"percentVar\"))[1],\"% variance\")) +\n",
    "        ylab(paste0(\"PC2: \", round(100 * attr(pca_data, \"percentVar\"))[2],\"% variance\")) + \n",
    "        scale_color_manual(\"Protocol\", values=c(\"low_input\"=\"#f54260\", \"normal_input\"=\"#4287f5\")) +\n",
    "        scale_fill_manual(\"Protocol\", values=c(\"low_input\"=\"#f54260\", \"normal_input\"=\"#4287f5\")) + \n",
    "        scale_shape_manual(\"Genotype\", values=setNames(c(21, 24), c(\"IFNAR_fl\", \"IFNAR_fl_LysM_cre\"))) + \n",
    "        theme(aspect.ratio=1, legend.position=\"right\") + guides(fill=guide_legend(byrow=TRUE)) + guides(color=guide_legend(byrow=TRUE)) + guides(shape=guide_legend(byrow=TRUE))\n",
    "\n",
    "    pca_plot_4 <- ggplot(pca_data, aes(PC1, PC2, color=library_size, fill=library_size, shape=genotype)) + \n",
    "        geom_point(size=5, stroke=1) + \n",
    "        ggtitle(\"PCA (Library size)\") + \n",
    "        xlab(paste0(\"PC1: \", round(100 * attr(pca_data, \"percentVar\"))[1],\"% variance\")) +\n",
    "        ylab(paste0(\"PC2: \", round(100 * attr(pca_data, \"percentVar\"))[2],\"% variance\")) + \n",
    "        scale_color_continuous(\"Library size\") +\n",
    "        scale_fill_continuous(\"Library size\") + \n",
    "        scale_shape_manual(\"Genotype\", values=setNames(c(21, 24), c(\"IFNAR_fl\", \"IFNAR_fl_LysM_cre\"))) + \n",
    "        theme(aspect.ratio=1, legend.position=\"right\") + guides(shape=guide_legend(byrow=TRUE))\n",
    "    \n",
    "    # Plot\n",
    "    pca_plot_1 + pca_plot_2 + pca_plot_3 + pca_plot_4 + patchwork::plot_layout(ncol=4)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce942bf4-b17d-4de4-9c29-18e6a8f4a7af",
   "metadata": {},
   "source": [
    "## All samples "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b501d92-0445-43cf-b00b-65f996dbd265",
   "metadata": {},
   "outputs": [],
   "source": [
    "dss <- DESeqDataSetFromMatrix(countData=cnt, colData=meta, design=~0+sample_name)\n",
    "vst <- DESeq2::vst(dss, fitType=\"local\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0118e4a7-c5bd-4da4-b794-ef78ef1926ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=4*7.5, repr.plot.height=7.5)\n",
    "\n",
    "pca_data <- DESeq2::plotPCA(vst, intgroup=colnames(meta), returnData=TRUE, ntop=500)\n",
    "pca_plot(pca_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "311eb34c-0d1f-4f92-9caf-d0e57b52d6e8",
   "metadata": {},
   "source": [
    "## Myeloid subset "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02961628-759c-443a-9547-6f31dbb9325c",
   "metadata": {},
   "outputs": [],
   "source": [
    "dss <- DESeqDataSetFromMatrix(countData=cnt[, meta$celltype!=\"EB\"], colData=meta[meta$celltype!=\"EB\", ], design=~0+sample_name)\n",
    "vst <- DESeq2::vst(dss, fitType=\"local\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08bc60b9-8212-44d1-939a-5518fa86fd1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=4*7.5, repr.plot.height=7.5)\n",
    "\n",
    "pca_data <- DESeq2::plotPCA(vst, intgroup=colnames(meta), returnData=TRUE, ntop=500)\n",
    "pca_plot(pca_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d925217c-2ed3-46db-8eee-0e27dd833ae4",
   "metadata": {},
   "source": [
    "## Erythroblast subset "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0b3271a-16cd-48a0-9984-0213d565c1df",
   "metadata": {},
   "outputs": [],
   "source": [
    "dss <- DESeqDataSetFromMatrix(countData=cnt[, meta$celltype==\"EB\"], colData=meta[meta$celltype==\"EB\", ], design=~0+sample_name)\n",
    "vst <- DESeq2::vst(dss, fitType=\"local\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "092f5630-4221-4f60-913a-a9f64b30077a",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=4*7.5, repr.plot.height=7.5)\n",
    "\n",
    "pca_data <- DESeq2::plotPCA(vst, intgroup=colnames(meta), returnData=TRUE, ntop=500)\n",
    "pca_plot(pca_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc9e0ecb-9577-404f-a33d-3fe305352859",
   "metadata": {},
   "source": [
    "# Save objects "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e659875-50e7-47e2-8953-485bfe9114b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(meta, \"data/bulkRNAseq/meta.rds\")\n",
    "saveRDS(ensembl_to_mgi(as.data.frame(cnt), convert=convert), \"data/bulkRNAseq/cnt.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b9f2904-fb76-4637-82db-904af2e28781",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(meta, \"data/bulkRNAseq/meta.csv\")\n",
    "write.csv(ensembl_to_mgi(as.data.frame(cnt), convert=convert), \"data/bulkRNAseq/cnt.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R (r.4.4.1)",
   "language": "R",
   "name": "r.4.4.1"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
