{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "018f633e-2d9a-4eda-b60a-a9abdb415c56",
   "metadata": {},
   "source": [
    "# Cluster differential expressed genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "a0d53ec4-a116-4cfa-ab36-2318815f27d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "library_load <- suppressMessages(\n",
    "    \n",
    "    list(\n",
    "        \n",
    "        # Seurat \n",
    "        library(Seurat), \n",
    "        \n",
    "        # GSEA\n",
    "        library(GSEABase), \n",
    "        library(fgsea), \n",
    "        library(msigdbr), \n",
    "        \n",
    "        # Data \n",
    "        library(tidyverse), \n",
    "        \n",
    "        # Plotting \n",
    "        library(ggplot2), \n",
    "        library(patchwork), \n",
    "        library(ComplexHeatmap)\n",
    "        \n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "9a8a6f68-25fc-43d8-b80c-f374258973f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set working directory to project root\n",
    "setwd(\"/research/peer/fdeckert/FD20200109SPLENO\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "cecebef3-1e5b-41c2-801e-07db6ceeeb02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Source files\n",
    "source(\"plotting_global.R\")\n",
    "source(\"bin/SeuratDEG.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "c7cd4c07-fe1d-4207-8da7-10d34ee6ef64",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting Theme\n",
    "ggplot2::theme_set(theme_global_set()) # From project global source()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "3ab6ce44-788c-43b3-911e-362f97d24d5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a761082-0825-45da-b534-5eeccaf0d7a0",
   "metadata": {},
   "source": [
    "# Parameter settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "80994038-883d-4134-b959-576f0d491660",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_file <- \"data/object/seurat_int.rds\"\n",
    "\n",
    "# Clustering \n",
    "cluster <- \"integrated_snn_res.0.8\"\n",
    "\n",
    "# DEG\n",
    "test_use <- \"wilcox\"\n",
    "\n",
    "deg_compute <- FALSE\n",
    "deg_file <- \"result/DEG/treatment_int_deg_wilcox_logfc_0_minpct_0.rds\"\n",
    "\n",
    "# GSEA\n",
    "gsea_compute <- FALSE\n",
    "gsea_file <- \"result/GSEA/treatment_int_gsea_wilcox_logfc_0_minpct_0.rds\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9007923-f242-4f84-a953-61f019150e31",
   "metadata": {},
   "source": [
    "# Import Objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "a626b763-c375-493b-837a-a0c902d308dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- readRDS(so_file)\n",
    "DefaultAssay(so) <- \"RNA\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0d0778e-3394-4d5b-9347-a97ee194d006",
   "metadata": {},
   "source": [
    "### For SCVI the data slot needs to be reset to count normalized data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "cb797a10-4745-4b51-b6dd-0c8e83f613bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(cluster==\"SCVI_snn_res.0.8\") {so <- NormalizeData(so)}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccb16e37-f354-4cff-8ead-12d288a2b51d",
   "metadata": {},
   "source": [
    "### Remove assays except RNA "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "a451852f-d535-4d2b-91d0-b99f06aac80d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Just to be absolut certain that the downstream analysis is done on the RNA assay all other assays are removed\n",
    "assays <- names(so@assays)[names(so@assays)!=\"RNA\"]\n",
    "if(length(assays) > 0) {for(assay in assays) {so[[assay]] <- NULL}}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5115e06-d166-4dab-844c-cbfb8454ba0b",
   "metadata": {},
   "source": [
    "# Type I interferone response (MSigDB; GO:0034340)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd191659-27d8-45de-ba05-93589f34c338",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs <- msigdbr(species=\"Mus musculus\", category=\"C5\")\n",
    "gs_inf <- gs[gs$gs_exact_source %in% \"GO:0034340\", ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1763e762-7534-4775-ac13-7cbda2265938",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- AddModuleScore(so, features=list(gs_inf$gene_symbol), assays=\"RNA\", slot=\"data\", ctrl=100, nbin=25, name=\"msGO0034340_RNA\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f5a631e-283f-4d9d-8804-0d8e591fea5b",
   "metadata": {},
   "source": [
    "## Plot module score "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "398a5c57-33bc-43c7-a741-e6bc33b64ad3",
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- so@meta.data[, c(\"treatment\", \"msGO0034340_RNA1\", cluster)] %>%\n",
    "    dplyr::group_by_at(c(\"treatment\", cluster)) %>%\n",
    "    dplyr::mutate(msGO0034340_RNA1=ifelse(msGO0034340_RNA1<0, 0, msGO0034340_RNA1)) %>%\n",
    "    dplyr::summarise(mean_ms=mean(msGO0034340_RNA1)) %>%\n",
    "    pivot_wider(names_from=treatment, values_from=mean_ms) %>%\n",
    "    dplyr::mutate(cluster=paste(\"Cluster \", .[, 1, drop=TRUE])) %>%\n",
    "    dplyr::mutate(fc=log2(CpG/NaCl)) %>%\n",
    "    dplyr::mutate(fc=ifelse(is.infinite(fc), 0, fc)) %>% \n",
    "    dplyr::select(-CpG, -NaCl, -as.symbol(cluster)) %>%\n",
    "    column_to_rownames(var=\"cluster\") %>% t()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5df92bec-c2ba-4fba-9e4e-c63d0772574f",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10, repr.plot.height=3)\n",
    "ComplexHeatmap::pheatmap(\n",
    "    main=\"Module score for interferon genes GO:00343402\", \n",
    "    name=\"log2FC(MS)\", \n",
    "    mat=data,  \n",
    "    cellwidth=20, \n",
    "    cellheight=20, \n",
    "    scale=\"none\",\n",
    "    cluster_rows=FALSE,\n",
    "    cluster_cols=FALSE,\n",
    "    show_rownames=FALSE,\n",
    "    show_colnames=TRUE, \n",
    "    color=colorRampPalette(c(color$treatment[\"NaCl\"], \"white\", color$treatment[\"CpG\"]))(50), \n",
    "    breaks=seq(-max(abs(na.omit(data[1, ]))), max(abs(na.omit(data[1, ]))), 0.001), \n",
    "    border_color=NA\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1fb2711-944a-443b-92a9-b6bf4e7a0866",
   "metadata": {},
   "source": [
    "# Differently gene expression (DGE) analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab8db824-8e7e-4c32-bfff-f9eac084c730",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set Ident and split by test group \n",
    "Idents(so) <- \"treatment\"\n",
    "so <- SplitObject(so, split.by=cluster)\n",
    "\n",
    "# Sort objects by cluster id \n",
    "so <- so[as.character(1:length(so))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d878829-1e97-4bef-a9d7-4b85a5c282ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- lapply(so, function(so) {\n",
    "    \n",
    "    cnt <- GetAssayData(so, assay=\"RNA\", slot=\"counts\")\n",
    "    cnt <- cnt[rowSums(cnt) > 0, ]\n",
    "    \n",
    "    so <- subset(so, features=rownames(cnt))\n",
    "    \n",
    "    return(so)\n",
    "    \n",
    "}\n",
    "      )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "7898411b-7e5b-4e6c-821c-001dd351dde8",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(deg_compute) {\n",
    "    \n",
    "    deg <- list()\n",
    "    for(i in names(so)) {\n",
    "        \n",
    "        deg[[i]] <- FindMarkers(so[[i]], ident.1=\"CpG\", ident.2=\"NaCl\", logfc.threshold=0, min.pct=0, assay=\"RNA\", test.use=test_use, verbose=TRUE)\n",
    "        \n",
    "    }\n",
    "    \n",
    "    saveRDS(deg, deg_file)\n",
    "    \n",
    "} else {\n",
    "    \n",
    "    deg <- readRDS(deg_file)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01b35917-10f7-4aa4-9d7e-a10d75f5794d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Change names to cluster \n",
    "names(deg) <- paste0(\"Cluster \", 1:length(deg))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08402719-3110-459d-aff4-2bba3e046d3b",
   "metadata": {},
   "source": [
    "## Volcaon plots "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b3aafba-09c4-4efe-9e25-9a5b0477c94a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add cluster information to deg for plotting\n",
    "for(i in 1:length(deg)) {deg[[i]] <- mutate(deg[[i]], cluster=i)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aef274fe-0db1-4745-a4ac-aea1484547af",
   "metadata": {},
   "outputs": [],
   "source": [
    "source(\"bin/SeuratQC.R\")\n",
    "vp <- lapply(deg, deg_volcano_plot, log2_thold=0.25, adjpvalue_thold=0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2261d099-ccbd-4211-8d16-2467ff820088",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=20, repr.plot.height=30)\n",
    "wrap_plots(vp, ncol=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "089b73ac-4b33-40d0-ba4c-6bf96433053d",
   "metadata": {},
   "source": [
    "# Gene set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fae5ae3c-9500-40dd-b04d-f68a612a6868",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs=msigdbr(species=\"Mus musculus\", category=\"H\")\n",
    "gs_list=split(x=gs$gene_symbol, f=gs$gs_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e70d7b0-2a4e-42b3-a03e-03fbbf7fcf29",
   "metadata": {},
   "source": [
    "# Ranked gene list by Signed -log10 adjusted p-values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e909674-f7fe-45b8-8d7f-eb8f90f09a37",
   "metadata": {},
   "outputs": [],
   "source": [
    "ranks <- list()\n",
    "for(i in 1:length(deg)) {\n",
    "    \n",
    "    deg_i <- deg[[i]]\n",
    "    deg_i$p_val_adj <- ifelse(deg_i$p_val_adj == 0, .Machine$double.xmin, deg_i$p_val_adj)\n",
    "    deg_i$sign_log_adj_p_values <- -log10(deg_i$p_val_adj) * sign(deg_i$avg_log2FC)\n",
    "    \n",
    "    ranks_i <- deg_i$sign_log_adj_p_values\n",
    "    names(ranks_i) <- rownames(deg_i)\n",
    "    \n",
    "    ranks_i <- ranks_i[order(-ranks_i)]\n",
    "    \n",
    "    ranks[[i]] <- ranks_i\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e19a5c9-ef65-448f-aa8a-a0bf2b6ced4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(ranks) <- paste0(\"Cluster \", 1:length(ranks))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54a4a8f9-df37-403e-8458-3092ab190545",
   "metadata": {},
   "source": [
    "## Plot ranked gene lists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a79ba794-b3a8-4f0f-97a4-9d9089c048ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- list()\n",
    "for(i in names(ranks)) {\n",
    "    \n",
    "    data[[i]] <- data.frame(\n",
    "        \n",
    "        score=ranks[[i]], \n",
    "        rank=1:length(ranks[[i]]), \n",
    "        cluster=i, \n",
    "        treatment=ifelse(ranks[[i]]>0, \"CpG\", ifelse(ranks[[i]]<0, \"NaCl\", NA))\n",
    "        \n",
    "    )\n",
    "    \n",
    "}\n",
    "\n",
    "# Combine list and set factor level for cluster\n",
    "data <- do.call(\"rbind\", data)\n",
    "data$cluster <- factor(data$cluster, levels=names(ranks))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a298828c-d929-49ba-b671-3bb0f1995ae9",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=20, repr.plot.height=9)\n",
    "ggplot(data, aes(x=rank, y=score, fill=treatment)) + \n",
    "    geom_bar(stat=\"identity\") + \n",
    "    scale_fill_manual(values=color$treatment, na.value=\"gray\") + \n",
    "    facet_wrap(~cluster, ncol=4) + \n",
    "    theme(legend.position=\"none\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc05fa8e-04c9-4cb6-9481-cd562d884f9f",
   "metadata": {},
   "source": [
    "# GSEA "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df75026b-b6b6-4f64-a8b5-372baf4e8f66",
   "metadata": {},
   "outputs": [],
   "source": [
    "ranks <- ranks[sapply(ranks, function(x) {sum(10^-abs(x) <= 0.05) >= 10})]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c034eeb5-6d54-45aa-8650-98f5bafcde8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(gsea_compute) {\n",
    "    \n",
    "    gsea <- list()\n",
    "    for(i in 1:length(ranks)) {\n",
    "        \n",
    "        gsea[[i]] <- fgsea(\n",
    "            \n",
    "            pathways=gs_list,\n",
    "            stats=ranks[[i]],\n",
    "            minSize=15,\n",
    "            maxSize=500\n",
    "            \n",
    "        )\n",
    "    }\n",
    "\n",
    "    saveRDS(gsea, gsea_file)\n",
    "    \n",
    "} else {\n",
    "    \n",
    "    gsea <- readRDS(gsea_file)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af75c88a-8ebb-4c6e-8183-98bd68264f33",
   "metadata": {},
   "source": [
    "## Filter add cluster information to gsea"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4985a276-f5c5-4668-a862-d8fcb16d665a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set gsea list names to cluster \n",
    "names(gsea) <- names(ranks) \n",
    "gsea <- gsea[!sapply(gsea, is.null)]\n",
    "for(i in names(gsea)) {gsea[[i]]$cluster <- i} "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "863bff8d-a406-46e0-a4a3-ac8106dca0f7",
   "metadata": {},
   "source": [
    "# GSEA plotting "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "087c257e-834b-41a6-8032-52b080be359a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_plot <- lapply(gsea, function(gsea, padj_thr=0.25, top=20) {\n",
    "    \n",
    "    # Correct pathway names \n",
    "    names <- str_split(gsea$pathway, pattern=\"HALLMARK_\")\n",
    "    names <- sapply(names, \"[[\", 2)\n",
    "    names <- str_split(names, pattern=\"_\")\n",
    "    names <- do.call(\"c\", lapply(names, paste, collapse=\" \"))\n",
    "    gsea$names <- names\n",
    "\n",
    "    # Add color \n",
    "    gsea$treatment <- ifelse(sign(gsea$ES)==1, \"CpG\", \"NaCl\")\n",
    "    gsea$treatment <- ifelse(gsea$padj<=padj_thr, gsea$treatment, NA)\n",
    "    \n",
    "    # Set 0 pvalues to min value \n",
    "    gsea$padj <- ifelse(gsea$padj == 0, .Machine$double.xmin, gsea$padj)\n",
    "\n",
    "    # Signed -log10 adjusted p-values \n",
    "    gsea$sign_log_adj_p_values <- -log10(gsea$padj) * sign(gsea$ES)\n",
    "    \n",
    "    # Set axis limits \n",
    "    if(max(abs(gsea$sign_log_adj_p_values)) >= 2) {\n",
    "        \n",
    "        x_max <- max(abs(gsea$sign_log_adj_p_values)) + 0.5\n",
    "                     \n",
    "    } else {\n",
    "        \n",
    "        x_max <- 2.5\n",
    "        \n",
    "    }\n",
    "    \n",
    "    # Take top 20 hits by adjusted pvalue\n",
    "    gsea <- gsea[order(padj)]\n",
    "    gsea <- gsea[1:top, ]\n",
    "    \n",
    "    # Order by adjusted p value\n",
    "    gsea <- gsea[order(sign_log_adj_p_values)]\n",
    "\n",
    "    # Set names to factor\n",
    "    gsea$names <- factor(gsea$names, levels=gsea$names)\n",
    "    \n",
    "    # Plot \n",
    "    plot <- ggplot(gsea, aes(x=sign_log_adj_p_values, y=names, color=treatment)) + \n",
    "        geom_point(aes(size=abs(NES))) +\n",
    "        geom_vline(xintercept=-log10(padj_thr)) + \n",
    "        geom_vline(xintercept=log10(padj_thr)) +\n",
    "        ggtitle(gsea$cluster[1]) +\n",
    "        xlab(\"Signed -log10 adj. p-value\") + ylab(\"\") + \n",
    "        xlim(-x_max, x_max) + \n",
    "        scale_color_manual(values=color$treatment, na.value=\"gray\") +\n",
    "        guides(\n",
    "            color=guide_legend(order=1), \n",
    "            size=guide_legend(order=2, title=\"Abs. (NES)\")\n",
    "        ) + \n",
    "        theme(\n",
    "            legend.position=\"right\", \n",
    "            legend.justification=\"top\"\n",
    "        )\n",
    "    \n",
    "    return(plot)\n",
    "    \n",
    "}\n",
    "                   )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "262eebe9-8515-4992-a5b2-30e2e0d60c18",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=25, repr.plot.height=floor(length(gsea_plot)/4) * 7.5)\n",
    "wrap_plots(gsea_plot, ncol=4)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "r.4.1.0-FD20200109SPLENO",
   "language": "R",
   "name": "r.4.1.0-fd20200109spleno"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
