{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b61c73b7-e11c-411d-b1f2-dc34e9123df5",
   "metadata": {},
   "source": [
    "# pySCENIC gene regulatory network analysis of pre-processed data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "055763ce-b6f6-49e6-b3d8-665d87f2b938",
   "metadata": {},
   "source": [
    "[GitHub](https://github.com/aertslab/pySCENIC?tab=readme-ov-file#id16)  \n",
    "[Tutorial](https://pyscenic.readthedocs.io/en/latest/tutorial.html)  \n",
    "[Worflow publication](https://www.nature.com/articles/s41596-020-0336-2)  \n",
    "[Data base](https://resources.aertslab.org/cistarget/databases/)  \n",
    "\n",
    "[About group comparisions](https://github.com/aertslab/pySCENIC/issues/292)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddfdfc22-3caf-4aeb-80b2-8db493766e2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "181ddac2-90c8-49e4-9233-90a9e27e9dcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyscenic.rss import regulon_specificity_scores\n",
    "from pyscenic.plotting import plot_rss\n",
    "\n",
    "from arboreto.algo import grnboost2\n",
    "\n",
    "import scanpy as sc\n",
    "import loompy as lp\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import glob\n",
    "\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcb7fa6b-4a3d-4354-a141-35015f2de8a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/research/peer/fdeckert/FD20200109SPLENO')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40773c94-88c6-4379-82a2-5ac1fa88fed1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# rpy2 \n",
    "os.environ['R_HOME'] = '/nobackup/peer/fdeckert/miniconda3/envs/r.4.4.1-FD20200109SPLENO/lib/R'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "226f9d89-a61d-403c-9f74-c9bebb5cd891",
   "metadata": {},
   "source": [
    "## Plotting "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "643336a9-3071-4ff5-ad3a-d30549264bcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.settings.vector_friendly = False\n",
    "\n",
    "sc.set_figure_params(figsize=(5, 5), dpi_save=1200, fontsize=12, frameon=False, facecolor=\"white\")\n",
    "sc.settings.figdir = 'result/figures/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8cf95cf-06ab-4e9a-a7e9-28c6fa12b35b",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set()\n",
    "sns.set(style='whitegrid', font_scale=0.8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0c79280-543e-4282-aa64-5e7b5438b7c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams.update(\n",
    "    \n",
    "    {\n",
    "    \n",
    "        'font.size': 12,          # Default font size\n",
    "        'axes.titlesize': 12,     # Title font size\n",
    "        'axes.labelsize': 12,     # X and Y label font size\n",
    "        'xtick.labelsize': 10,    # X tick label font size\n",
    "        'ytick.labelsize': 10,    # Y tick label font size\n",
    "        'legend.fontsize': 12     # Legend font size\n",
    "\n",
    "    }\n",
    "        )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fe9dc11-7cb5-43c3-922d-1e167e0334b7",
   "metadata": {},
   "source": [
    "# Create AnnData from bulk RNAseq experiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b3c76e3-8286-4ae8-9a85-f83bf3cd7a55",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.AnnData(\n",
    "    \n",
    "    pd.read_csv('data/bulkRNAseq/object/cnt.csv', index_col=0).drop(['mgi_symbol', 'ensembl_gene_id'], axis=1).transpose(), \n",
    "    obs=pd.read_csv('data/bulkRNAseq/object/meta.csv', index_col=0)\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4a0a57a-712c-4739-a8c2-d5f0874c4332",
   "metadata": {},
   "source": [
    "# Subset "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94a312f7-914b-4c06-b2d3-24f657d30f66",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[:, (adata.X>=3).sum(axis=0)>=3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4d9b4bf-bd0c-4b91-b418-f0646ec4e1fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21278c55-a3af-4719-b775-c3bf09b0c254",
   "metadata": {},
   "source": [
    "# Export loom file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35a3a2d9-cf75-49eb-ae20-026bc48c4684",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_attrs = {\n",
    "    \n",
    "    \"Gene\": np.array(adata.var_names),\n",
    "}\n",
    "\n",
    "col_attrs = {\n",
    "    \n",
    "    \"CellID\": np.array(adata.obs_names),\n",
    "    \"nGene\": np.array(np.sum(adata.X.transpose()>0, axis=0)).flatten(),\n",
    "    \"nUMI\": np.array(np.sum(adata.X.transpose(), axis=0)).flatten(),\n",
    "    \n",
    "}\n",
    "\n",
    "lp.create(\"data/bulkRNAseq/scenic/adata.loom\", adata.X.transpose(), row_attrs, col_attrs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "757348fa-e8a9-4f9b-a8c7-9d86629347f7",
   "metadata": {},
   "source": [
    "# pySCENIC workflow "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d62562cd-3ac0-412b-a8cf-b7841429b97d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Scenic data base files \n",
    "tf_file = \"/nobackup/peer/fdeckert/scenic/mm10/mm_mgi_tfs.txt\"\n",
    "db_file = ' '.join(glob.glob(\"/nobackup/peer/fdeckert/scenic/mm10/*feather\"))\n",
    "motif_file = \"/nobackup/peer/fdeckert/scenic/mm10/motifs-v10nr_clust-nr.mgi-m0.001-o0.0.tbl\"\n",
    "\n",
    "# Loom \n",
    "input_loom = \"/research/peer/fdeckert/FD20200109SPLENO/data/bulkRNAseq/scenic/adata.loom\"\n",
    "\n",
    "# Result  \n",
    "result_dir = \"/research/peer/fdeckert/FD20200109SPLENO/result/scenic/bulkRNAseq\"\n",
    "result_loom = \"pyscenic.loom\"\n",
    "\n",
    "# Available workers \n",
    "num_workers = os.cpu_count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1545d22f-44a0-41a1-8755-003ab9e55c13",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(result_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfc7d51f-05f9-444a-a2f2-09ec32dd0ad1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !/nobackup/peer/fdeckert/miniconda3/envs/pyscenic.0.12.1/bin/pyscenic grn {input_loom} {tf_file} \\\n",
    "#     --method grnboost2 \\\n",
    "#     --output adj.csv \\\n",
    "#     --num_workers {num_workers}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f0bd0e7-32ca-47f1-a45a-d40687dd1f01",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !/nobackup/peer/fdeckert/miniconda3/envs/pyscenic.0.12.1/bin/pyscenic ctx adj.csv {db_file} \\\n",
    "#     --annotations_fname {motif_file} \\\n",
    "#     --expression_mtx_fname {input_loom} \\\n",
    "#     --output reg.csv \\\n",
    "#     --mask_dropouts \\\n",
    "#     --num_workers {num_workers}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3ed91c3-ede8-47a0-807c-27ad476762f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !/nobackup/peer/fdeckert/miniconda3/envs/pyscenic.0.12.1/bin/pyscenic aucell \\\n",
    "#     {input_loom} \\\n",
    "#     reg.csv \\\n",
    "#     --output {result_loom} \\\n",
    "#     --num_workers {num_workers}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55c9e75c-9ee2-4543-beb5-021d74a0eaa2",
   "metadata": {},
   "outputs": [],
   "source": [
    "lf = lp.connect(result_loom, mode='r+', validate=False)\n",
    "auc_mtx = pd.DataFrame(lf.ca.RegulonsAUC, index=lf.ca.CellID)\n",
    "adata.obsm['X_regulon'] = auc_mtx\n",
    "lf.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5652aa8-3e82-4c71-8579-4eeec2424514",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.neighbors(adata, use_rep='X_regulon')\n",
    "sc.tl.umap(adata)\n",
    "sc.pl.umap(adata, color=['celltype'], size=100)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyscenic.0.12.1",
   "language": "python",
   "name": "pyscenic.0.12.1"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.21"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
