{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "91386b4f-9250-4761-9153-03fc206ea91e",
   "metadata": {},
   "source": [
    "# pySCENIC gene regulatory network analysis of pre-processed data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f67a3a41-85a4-46bc-854d-019889ed324a",
   "metadata": {},
   "source": [
    "[GitHub](https://github.com/aertslab/pySCENIC?tab=readme-ov-file#id16)  \n",
    "[Tutorial](https://pyscenic.readthedocs.io/en/latest/tutorial.html)  \n",
    "[Worflow publication](https://www.nature.com/articles/s41596-020-0336-2)  \n",
    "[Data base](https://resources.aertslab.org/cistarget/databases/)  \n",
    "\n",
    "[About group comparisions](https://github.com/aertslab/pySCENIC/issues/292)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81e651e8-75f1-42ac-abac-c7bc2dea9dbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10ad739c-34af-46ba-9240-cfea00550cb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from arboreto.algo import grnboost2\n",
    "from pyscenic.utils import modules_from_adjacencies\n",
    "from pyscenic.prune import prune2df, df2regulons\n",
    "from pyscenic.aucell import aucell\n",
    "\n",
    "from arboreto.utils import load_tf_names\n",
    "from ctxcore.rnkdb import FeatherRankingDatabase as RankingDatabase\n",
    "\n",
    "import scanpy as sc\n",
    "import loompy as lp\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import pickle\n",
    "\n",
    "import glob\n",
    "\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bb01017-f0bf-4445-a42e-170e16a426a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/research/peer/fdeckert/FD20200109SPLENO')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79e51737-fc2d-496d-9d34-217a2d6552d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# rpy2 \n",
    "os.environ['R_HOME'] = '/nobackup/peer/fdeckert/miniconda3/envs/r.4.4.1-FD20200109SPLENO/lib/R'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78f64b93-303f-45bd-9a7c-acbf365551d3",
   "metadata": {},
   "source": [
    "## Plotting "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac5a8128-893a-476a-bfa6-1ca74741593c",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.settings.vector_friendly = False\n",
    "\n",
    "sc.set_figure_params(figsize=(5, 5), dpi_save=1200, fontsize=12, frameon=False, facecolor='white')\n",
    "sc.settings.figdir = 'result/figures/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8653e311-df5a-4e36-8cd1-34d1b5e629e3",
   "metadata": {},
   "source": [
    "# pySCENIC workflow "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e977dbc3-6e0f-4c78-8ac2-844d9f9ff617",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Scenic data base files \n",
    "tf_file = '/research/peer/fdeckert/reference/animaltfdb/Mus_musculus_TF.txt'\n",
    "db_file = glob.glob('/nobackup/peer/fdeckert/scenic/mm10/*feather')\n",
    "motif_file = '/nobackup/peer/fdeckert/scenic/mm10/motifs-v10nr_clust-nr.mgi-m0.001-o0.0.tbl'\n",
    "\n",
    "adj_file = 'adj.csv'\n",
    "module_file = 'module.pkl'\n",
    "reg_file = 'reg.pkl'\n",
    "auc_file = 'auc_mtx.csv'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9644954e-e81d-484a-8e4d-53f78fedef08",
   "metadata": {},
   "source": [
    "# Store animaltfdb in pySCENIC reference "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5daa03e-34da-4a5e-82a3-a52b544f3d94",
   "metadata": {},
   "outputs": [],
   "source": [
    "animaltfdb = pd.read_csv(tf_file, sep=\"\\t\", header=0)\n",
    "animaltfdb[\"Symbol\"].to_csv('/nobackup/peer/fdeckert/scenic/mm10/animaltfdb.txt', index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21ece1c5-ee6b-4935-b0d9-1230f68070c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_file = '/nobackup/peer/fdeckert/scenic/mm10/animaltfdb.txt'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bf93267-5eec-4dba-bebd-916a363c0c3a",
   "metadata": {},
   "source": [
    "# Load reference data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "696199a9-8555-4b8f-835b-4ac58da1ee79",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load TF list\n",
    "tf_names = load_tf_names(tf_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f04039eb-eea8-4d7d-aa24-e19ee71ed266",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load motif feather\n",
    "def name(fname): return os.path.splitext(os.path.basename(fname))[0]\n",
    "db = [RankingDatabase(fname=fname, name=name(fname)) for fname in db_file]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e34d7b18-6a88-4a72-b87b-bb5ba8d861c0",
   "metadata": {},
   "source": [
    "# pySCENIC analysis for erythroid lineage from BSF samples "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fde609a-5376-41e2-85da-fecc9b8ffb88",
   "metadata": {},
   "source": [
    "## Subset AnnData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b8e3b85-d126-46fe-8d80-2621a3ae2266",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad('/research/peer/fdeckert/FD20200109SPLENO/data/scRNAseq/object/pp.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20384f82-4065-4083-a3d1-7a277f90ca7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[(adata.obs.celltype_low.isin(['MegP', 'MEP', 'Proerythroblast', 'Erythroblast'])) & (adata.obs.facility=='BSF')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e0e7bd2-1b6e-49c7-9702-848e2984e4f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_1 = adata[:, (adata.X>=5).sum(axis=0)>=1].var_names.to_list()\n",
    "genes_2 = adata[:, (adata.X>=1).sum(axis=0)>=20].var_names.to_list()\n",
    "genes = list(set(genes_1 + genes_2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fcc5f1c-2315-4b0f-a1e7-68ea0c5dbf3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[:, genes].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b557ad8-0ffd-41b5-a02d-1ed0c828cec9",
   "metadata": {},
   "source": [
    "## Set out dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40c3a3ca-5da3-43d5-a874-1a2b068c723e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Result directory   \n",
    "result_dir = '/research/peer/fdeckert/FD20200109SPLENO/result/scenic/scRNAseq/res_0'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5b8967d-e166-4347-9005-444d1a182584",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(result_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89f5cb4e-ddc7-42e1-adc4-efbf79dc15e9",
   "metadata": {},
   "source": [
    "## Expression matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ae3e778-1917-4188-8ae2-8d079f06a8f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat = pd.DataFrame(adata.X.A if hasattr(adata.X, 'A') else adata.X, index=adata.obs_names, columns=adata.var_names)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21cf3d14-e36c-423a-80de-e970ebb04b06",
   "metadata": {},
   "source": [
    "## Compute graph adjacency "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77bbff74-a4c2-4209-adbf-08780f589055",
   "metadata": {},
   "outputs": [],
   "source": [
    "adj_cache = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "191f0a15-42ce-485d-871a-a8191f282def",
   "metadata": {},
   "outputs": [],
   "source": [
    "if adj_cache: \n",
    "\n",
    "    # Load pre-compute adjacency\n",
    "    adj = pd.read_csv(adj_file)\n",
    "\n",
    "else: \n",
    "    \n",
    "    # Run GRNBoost2 \n",
    "    adj = grnboost2(mat, tf_names=tf_names, seed=42, verbose=True)\n",
    "\n",
    "    # Store results\n",
    "    adj.to_csv(adj_file, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adfa3a8a-56c3-45c8-966d-34130e6e6b13",
   "metadata": {},
   "source": [
    "## Create modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24f6dd58-1f41-43bb-81bb-feaed9bcf015",
   "metadata": {},
   "outputs": [],
   "source": [
    "module_cache = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e644cd35-893a-420d-8f8c-16afe06afdaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "if module_cache: \n",
    "\n",
    "    # Load pre-computed adjacency \n",
    "    with open(module_file, 'rb') as f: module = pickle.load(f)\n",
    "\n",
    "else: \n",
    "\n",
    "    module = list(modules_from_adjacencies(adj, mat, rho_mask_dropouts=False))\n",
    "    with open(module_file, 'wb') as f: pickle.dump(module, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68cc6f5f-33cc-41b7-a6d6-93f9238f2a8a",
   "metadata": {},
   "source": [
    "## Prune graph by biding motifs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "197f98f6-8a0e-4ecb-b2bf-f7c2622718cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "reg_cache = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f47538a-2a01-4b84-b9cc-6ee1aa0cee48",
   "metadata": {},
   "outputs": [],
   "source": [
    "if reg_cache: \n",
    "\n",
    "    # Load pre-computed regulons \n",
    "    with open(reg_file, 'rb') as f: reg = pickle.load(f)\n",
    "\n",
    "else: \n",
    "\n",
    "    reg = prune2df(db, module, motif_annotations_fname=motif_file, filter_for_annotation=True, num_workers=32)\n",
    "\n",
    "    # Save as pkl for further processing \n",
    "    with open(reg_file, 'wb') as f: pickle.dump(reg, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0426d57e-5220-4cc6-8ad2-9231e9acd4a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "reg_df = reg[('Enrichment', 'TargetGenes')].reset_index()\n",
    "reg_df = reg_df.explode(('Enrichment', 'TargetGenes'))\n",
    "reg_df['TargetGene'] = reg_df[('Enrichment', 'TargetGenes')].apply(lambda x: x[0])\n",
    "reg_df = reg_df[['TF', 'TargetGene']]\n",
    "reg_df = reg_df.drop_duplicates().reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2d6b049-362c-4e38-abb3-75f3d76ba3e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "reg_df.to_csv('reg.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c62bab4-daf6-4980-a72d-2539e3b2285f",
   "metadata": {},
   "source": [
    "## AUC matrix "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2762d299-b81a-439c-9943-5fc4923f2751",
   "metadata": {},
   "outputs": [],
   "source": [
    "reg = df2regulons(reg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f9aa61a-0dfa-4fa1-b255-5643b5bd79a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "auc_mtx = aucell(mat, reg, num_workers=4)\n",
    "auc_mtx.to_csv(auc_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e5745a5-e040-42b4-a881-df511c7e47fb",
   "metadata": {},
   "source": [
    "## AUC embedding UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8be05704-7d3e-476d-a0b2-e585d8c71c8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obsm['X_regulon'] = auc_mtx\n",
    "sc.pp.neighbors(adata, use_rep='X_regulon')\n",
    "sc.tl.umap(adata)\n",
    "sc.pl.umap(adata, color=['celltype_low', 'sample_group'], size=20)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86e9d60f-0852-4c5f-802b-f4261e8aa317",
   "metadata": {},
   "source": [
    "# pySCENIC analysis for erythroid lineage from VBC samples "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c56f0667-fa14-4024-9065-00a396c89543",
   "metadata": {},
   "source": [
    "## Subset AnnData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc36209b-b969-4b51-84da-892f1e32ef53",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad('/research/peer/fdeckert/FD20200109SPLENO/data/scRNAseq/object/pp.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c12ecde9-7502-4436-8511-3647104725ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[(adata.obs.celltype_low.isin(['MegP', 'MEP', 'Proerythroblast', 'Erythroblast'])) & (adata.obs.facility=='VBC')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44b9bb02-c07a-4be9-a21d-e36fca848365",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_1 = adata[:, (adata.X>=5).sum(axis=0)>=1].var_names.to_list()\n",
    "genes_2 = adata[:, (adata.X>=1).sum(axis=0)>=20].var_names.to_list()\n",
    "genes = list(set(genes_1 + genes_2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f314dae-6f42-45ed-963d-ca804e824200",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[:, genes].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1e34d24-6734-4edf-b566-9c575980a9c2",
   "metadata": {},
   "source": [
    "## Set out dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2aabf229-a107-40a0-8791-869c6be96c28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Result directory   \n",
    "result_dir = '/research/peer/fdeckert/FD20200109SPLENO/result/scenic/scRNAseq/res_1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18fabc11-cbe5-4f60-b1e1-4452ab4e0159",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(result_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5529db52-a77c-4c5e-84d5-74aacc75b948",
   "metadata": {},
   "source": [
    "## Expression matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "723323a4-1a01-4e49-965c-7141bc803c4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat = pd.DataFrame(adata.X.A if hasattr(adata.X, 'A') else adata.X, index=adata.obs_names, columns=adata.var_names)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca8c539a-5342-426a-9d65-37ac1e709cd1",
   "metadata": {},
   "source": [
    "## Compute graph adjacency "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e8dc5c0-a5db-40ae-a4eb-0df96167068d",
   "metadata": {},
   "outputs": [],
   "source": [
    "adj_cache = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9ba836f-0d49-4d49-83dd-d23db97c464c",
   "metadata": {},
   "outputs": [],
   "source": [
    "if adj_cache: \n",
    "\n",
    "    # Load pre-compute adjacency\n",
    "    adj = pd.read_csv(adj_file)\n",
    "\n",
    "else: \n",
    "    \n",
    "    # Run GRNBoost2 \n",
    "    adj = grnboost2(mat, tf_names=tf_names, seed=42, verbose=True)\n",
    "\n",
    "    # Store results\n",
    "    adj.to_csv(adj_file, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19df3e88-5dd8-4421-b2ed-4358f3c27bd5",
   "metadata": {},
   "source": [
    "## Create modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0153d95-a1dc-4b7a-af14-9fbbfcf974cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "module_cache = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58f1f031-5f4b-48d1-8c38-f11a36a70996",
   "metadata": {},
   "outputs": [],
   "source": [
    "if module_cache: \n",
    "\n",
    "    # Load pre-computed adjacency \n",
    "    with open(module_file, 'rb') as f: module = pickle.load(f)\n",
    "\n",
    "else: \n",
    "\n",
    "    module = list(modules_from_adjacencies(adj, mat, rho_mask_dropouts=False))\n",
    "    with open(module_file, 'wb') as f: pickle.dump(module, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1b112c7-c92a-4517-99cc-a91d98eedd2b",
   "metadata": {},
   "source": [
    "## Prune graph by biding motifs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fd4b23d-b47d-489a-aae5-5bc9038b0888",
   "metadata": {},
   "outputs": [],
   "source": [
    "reg_cache = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8af8958a-e1ef-42ca-bf12-5e1a232b903a",
   "metadata": {},
   "outputs": [],
   "source": [
    "if reg_cache: \n",
    "\n",
    "    # Load pre-computed regulons \n",
    "    with open(reg_file, 'rb') as f: reg = pickle.load(f)\n",
    "\n",
    "else: \n",
    "\n",
    "    reg = prune2df(db, module, motif_annotations_fname=motif_file, filter_for_annotation=True, num_workers=32)\n",
    "\n",
    "    # Save as pkl for further processing \n",
    "    with open(reg_file, 'wb') as f: pickle.dump(reg, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4446ad5b-a030-4c0e-beba-27d8db5d883c",
   "metadata": {},
   "outputs": [],
   "source": [
    "reg_df = reg[('Enrichment', 'TargetGenes')].reset_index()\n",
    "reg_df = reg_df.explode(('Enrichment', 'TargetGenes'))\n",
    "reg_df['TargetGene'] = reg_df[('Enrichment', 'TargetGenes')].apply(lambda x: x[0])\n",
    "reg_df = reg_df[['TF', 'TargetGene']]\n",
    "reg_df = reg_df.drop_duplicates().reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fdfbd7a-4cd2-4c13-92e8-82fcf51ed3ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "reg_df.to_csv('reg.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5244741e-48e8-48d7-a6ff-0a1dc36343aa",
   "metadata": {},
   "source": [
    "## AUC matrix "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e5bc2aa-59ee-444c-bdf8-7d7a39f6c8fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "reg = df2regulons(reg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66c20972-ca1f-43f3-bb34-86b636a94b8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "auc_mtx = aucell(mat, reg, num_workers=4)\n",
    "auc_mtx.to_csv(auc_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2c9ce4f-d6a3-4064-a065-015567ec3071",
   "metadata": {},
   "source": [
    "## AUC embedding UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c95b6d1f-b72c-4a13-aba2-c5ed9dde5b6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obsm['X_regulon'] = auc_mtx\n",
    "sc.pp.neighbors(adata, use_rep='X_regulon')\n",
    "sc.tl.umap(adata)\n",
    "sc.pl.umap(adata, color=['celltype_low', 'sample_group'], size=20)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd4916b3-0c8a-42f8-8156-d49bc726a9fe",
   "metadata": {},
   "source": [
    "# pySCENIC analysis for myeloid lineage from BSF samples "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e8e1c8f-5f66-426a-b4fe-ef3c7d5c4491",
   "metadata": {},
   "source": [
    "## Subset AnnData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "377b2dfe-ecdb-4e8e-aa6f-b286f14967c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad('/research/peer/fdeckert/FD20200109SPLENO/data/scRNAseq/object/pp.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "19fa66b4-de7c-4f86-8341-0e877e5fe86f",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[(adata.obs.celltype_low.isin(['cMo', 'intMo', 'ncMo', 'RPM'])) & (adata.obs.facility=='BSF')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "ba839b17-e30a-413e-b90d-8f04ef3b6294",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_1 = adata[:, (adata.X>=5).sum(axis=0)>=1].var_names.to_list()\n",
    "genes_2 = adata[:, (adata.X>=1).sum(axis=0)>=20].var_names.to_list()\n",
    "genes = list(set(genes_1 + genes_2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "95a08a83-1319-49a6-82d3-c48952920bf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[:, genes].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67427a56-578f-4c88-a7c7-93b26a4df563",
   "metadata": {},
   "source": [
    "## Set out dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "6b92ac7d-3615-4de1-a35c-5626072a39b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Result directory   \n",
    "result_dir = '/research/peer/fdeckert/FD20200109SPLENO/result/scenic/scRNAseq/res_2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "ef7fe894-ce0c-43fe-a1f7-374dc46d5a17",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(result_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e31e592-8a99-49f1-b2d9-10e3036ae36c",
   "metadata": {},
   "source": [
    "## Expression matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "88385d82-45ba-44c8-b510-66e05c6aba04",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat = pd.DataFrame(adata.X.A if hasattr(adata.X, 'A') else adata.X, index=adata.obs_names, columns=adata.var_names)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "149e156c-2a4e-4260-81ed-f1b533a81b9a",
   "metadata": {},
   "source": [
    "## Compute graph adjacency "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "9f1c3115-3466-4eb6-b337-6d3f67162a47",
   "metadata": {},
   "outputs": [],
   "source": [
    "adj_cache = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "c21d238b-0802-492f-9860-8530f6dee995",
   "metadata": {},
   "outputs": [],
   "source": [
    "if adj_cache: \n",
    "\n",
    "    # Load pre-compute adjacency\n",
    "    adj = pd.read_csv(adj_file)\n",
    "\n",
    "else: \n",
    "    \n",
    "    # Run GRNBoost2 \n",
    "    adj = grnboost2(mat, tf_names=tf_names, seed=42, verbose=True)\n",
    "\n",
    "    # Store results\n",
    "    adj.to_csv(adj_file, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f5d1742-26d1-438e-bb92-9783d936603a",
   "metadata": {},
   "source": [
    "## Create modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "957eee10-d1a1-4431-8d95-1fd39b43d6f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "module_cache = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "77295476-e3aa-4e3c-a3fe-09e8bd953e8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "if module_cache: \n",
    "\n",
    "    # Load pre-computed adjacency Spic\n",
    "    \n",
    "    with open(module_file, 'rb') as f: module = pickle.load(f)\n",
    "\n",
    "else: \n",
    "\n",
    "    module = list(modules_from_adjacencies(adj, mat, rho_mask_dropouts=False))\n",
    "    with open(module_file, 'wb') as f: pickle.dump(module, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecc65147-f02c-471b-a630-c58053928d74",
   "metadata": {},
   "source": [
    "## Prune graph by biding motifs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "e70f682f-b2c6-41ed-9ac3-cca5eea2495e",
   "metadata": {},
   "outputs": [],
   "source": [
    "reg_cache = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "8182a47a-fde2-4039-bc85-4fb4c4fcb8d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "if reg_cache: \n",
    "\n",
    "    # Load pre-computed regulons \n",
    "    with open(reg_file, 'rb') as f: reg = pickle.load(f)\n",
    "\n",
    "else: \n",
    "\n",
    "    reg = prune2df(db, module, motif_annotations_fname=motif_file, filter_for_annotation=True, num_workers=32)\n",
    "\n",
    "    # Save as pkl for further processing \n",
    "    with open(reg_file, 'wb') as f: pickle.dump(reg, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "d948f8ab-1f14-4ed5-922f-56974d8e60b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "reg_df = reg[('Enrichment', 'TargetGenes')].reset_index()\n",
    "reg_df = reg_df.explode(('Enrichment', 'TargetGenes'))\n",
    "reg_df['TargetGene'] = reg_df[('Enrichment', 'TargetGenes')].apply(lambda x: x[0])\n",
    "reg_df = reg_df[['TF', 'TargetGene']]\n",
    "reg_df = reg_df.drop_duplicates().reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e5d41e8-c912-436f-aa8a-de3b95b09944",
   "metadata": {},
   "outputs": [],
   "source": [
    "reg_df.to_csv('reg.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "058e7cd3-9924-4f88-b574-cfa52dfef83d",
   "metadata": {},
   "source": [
    "## AUC matrix "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e213024-80b9-4b75-9351-6e2e53263288",
   "metadata": {},
   "outputs": [],
   "source": [
    "reg = df2regulons(reg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b53e3587-5d57-45c2-98d0-3cf0cf98a6cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "auc_mtx = aucell(mat, reg, num_workers=4)\n",
    "auc_mtx.to_csv(auc_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1eb4a3be-6e3e-42ed-8c09-b38d1e735940",
   "metadata": {},
   "source": [
    "## AUC embedding UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ee65858-4be9-4838-90d3-70aba2f25180",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obsm['X_regulon'] = auc_mtx\n",
    "sc.pp.neighbors(adata, use_rep='X_regulon')\n",
    "sc.tl.umap(adata)\n",
    "sc.pl.umap(adata, color=['celltype_low', 'sample_group'], size=20)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fea60725-0157-427e-9f2c-1de6eafdbdcb",
   "metadata": {},
   "source": [
    "# pySCENIC analysis for myeloid lineage from VBC samples "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a7007c4-54ae-4770-80cd-7267aa9b56f4",
   "metadata": {},
   "source": [
    "## Subset AnnData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "276aa781-961b-4fb3-b7e7-cee0150d79a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad('/research/peer/fdeckert/FD20200109SPLENO/data/scRNAseq/object/pp.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f761b2b2-5636-4c07-bb11-f41a178bc9af",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[(adata.obs.celltype_low.isin(['cMo', 'intMo', 'ncMo', 'RPM'])) & (adata.obs.facility=='VBC')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4780adf-b21d-4126-86f4-9932a2436cf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_1 = adata[:, (adata.X>=5).sum(axis=0)>=1].var_names.to_list()\n",
    "genes_2 = adata[:, (adata.X>=1).sum(axis=0)>=20].var_names.to_list()\n",
    "genes = list(set(genes_1 + genes_2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e96a2a29-976e-4336-b8d5-bc55eb87ac99",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[:, genes].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdcc5d64-4e8b-42d2-b42a-d2da6a91558a",
   "metadata": {},
   "source": [
    "## Set out dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e69b9c33-15ac-4d82-be25-b53851e72dab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Result directory   \n",
    "result_dir = '/research/peer/fdeckert/FD20200109SPLENO/result/scenic/scRNAseq/res_3'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2a2b344-3d08-493a-9053-85e582d22259",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(result_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcce3aec-7130-4eea-be57-083a1bb04432",
   "metadata": {},
   "source": [
    "## Expression matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c30032c7-7a10-4998-b5bf-4dcd7f9da64c",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat = pd.DataFrame(adata.X.A if hasattr(adata.X, 'A') else adata.X, index=adata.obs_names, columns=adata.var_names)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c6cb5c8-1e16-4f14-80d1-80dfcb9e91e8",
   "metadata": {},
   "source": [
    "## Compute graph adjacency "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fea6e49a-d968-4e5d-b3ce-af1cad7d936b",
   "metadata": {},
   "outputs": [],
   "source": [
    "adj_cache = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eaffc8e9-35cd-4ca2-a6e4-687c245b7369",
   "metadata": {},
   "outputs": [],
   "source": [
    "if adj_cache: \n",
    "\n",
    "    # Load pre-compute adjacency\n",
    "    adj = pd.read_csv(adj_file)\n",
    "\n",
    "else: \n",
    "    \n",
    "    # Run GRNBoost2 \n",
    "    adj = grnboost2(mat, tf_names=tf_names, seed=42, verbose=True)\n",
    "\n",
    "    # Store results\n",
    "    adj.to_csv(adj_file, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04cd0e74-2db8-4d35-afb9-96f8908af506",
   "metadata": {},
   "source": [
    "## Create modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f05d2a20-9b55-4e8a-9eaf-dec334073214",
   "metadata": {},
   "outputs": [],
   "source": [
    "module_cache = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9877e0cd-3901-4ad0-b0d7-c8085c546465",
   "metadata": {},
   "outputs": [],
   "source": [
    "if module_cache: \n",
    "\n",
    "    # Load pre-computed adjacency \n",
    "    with open(module_file, 'rb') as f: module = pickle.load(f)\n",
    "\n",
    "else: \n",
    "\n",
    "    module = list(modules_from_adjacencies(adj, mat, rho_mask_dropouts=False))\n",
    "    with open(module_file, 'wb') as f: pickle.dump(module, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9a13a53-fada-4ba3-94ba-53d3c9495767",
   "metadata": {},
   "source": [
    "## Prune graph by biding motifs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43340168-57b3-4918-977c-dccb6d42ee84",
   "metadata": {},
   "outputs": [],
   "source": [
    "reg_cache = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d258c2a8-3bb3-4b1e-8a22-a036334b9676",
   "metadata": {},
   "outputs": [],
   "source": [
    "if reg_cache: \n",
    "\n",
    "    # Load pre-computed regulons \n",
    "    with open(reg_file, 'rb') as f: reg = pickle.load(f)\n",
    "\n",
    "else: \n",
    "\n",
    "    reg = prune2df(db, module, motif_annotations_fname=motif_file, filter_for_annotation=True, num_workers=32)\n",
    "\n",
    "    # Save as pkl for further processing \n",
    "    with open(reg_file, 'wb') as f: pickle.dump(reg, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "972dd39c-6dfc-451d-a620-bdfba8c3bec8",
   "metadata": {},
   "outputs": [],
   "source": [
    "reg_df = reg[('Enrichment', 'TargetGenes')].reset_index()\n",
    "reg_df = reg_df.explode(('Enrichment', 'TargetGenes'))\n",
    "reg_df['TargetGene'] = reg_df[('Enrichment', 'TargetGenes')].apply(lambda x: x[0])\n",
    "reg_df = reg_df[['TF', 'TargetGene']]\n",
    "reg_df = reg_df.drop_duplicates().reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9496def0-f7ff-4d1b-a7d2-66e4dbea840a",
   "metadata": {},
   "outputs": [],
   "source": [
    "reg_df.to_csv('reg.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ecd45d9-4ff1-4767-8288-48c960cf2236",
   "metadata": {},
   "source": [
    "## AUC matrix "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1581d0bb-7410-4c24-9694-62c98fe439d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "reg = df2regulons(reg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe84ccc9-c322-4b1f-9a1a-a374496f14a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "auc_mtx = aucell(mat, reg, num_workers=4)\n",
    "auc_mtx.to_csv(auc_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5196e3c3-6f53-4d2f-9c8f-e1a3e17847e3",
   "metadata": {},
   "source": [
    "## AUC embedding UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4330e665-d8f1-402c-9460-8037aa8f5279",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obsm['X_regulon'] = auc_mtx\n",
    "sc.pp.neighbors(adata, use_rep='X_regulon')\n",
    "sc.tl.umap(adata)\n",
    "sc.pl.umap(adata, color=['celltype_low', 'sample_group'], size=20)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyscenic.0.12.1",
   "language": "python",
   "name": "pyscenic.0.12.1"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
