{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "22a20c70-716d-48cf-9ca0-0e8a438ca1e2",
   "metadata": {},
   "source": [
    "# Lamian lineage marker genes and DEA along pseudotime \n",
    "\n",
    "**Lamian**  \n",
    "[Preprint](https://www.biorxiv.org/content/10.1101/2021.07.10.451910v1.full)  \n",
    "[GitHub](https://github.com/Winnie09/Lamian)  \n",
    "[Tutorial](https://winnie09.github.io/Wenpin_Hou/pages/Lamian.html)  \n",
    "\n",
    "**Condiment**  \n",
    "[Preprint](https://www.biorxiv.org/content/10.1101/2021.03.09.433671v1.full)  \n",
    "[GitHub](https://github.com/HectorRDB/condiments)  \n",
    "[Tutorial](https://hectorrdb.github.io/condimentsPaper/)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e79e7ae2-5f48-468f-97ee-44e0c9dc3ec0",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9268cdcf-e507-4b8b-a6a7-29254c35c64e",
   "metadata": {},
   "outputs": [],
   "source": [
    "library_load <- suppressMessages(\n",
    "    \n",
    "    list(\n",
    "        \n",
    "        # Seurat \n",
    "        library(Seurat), \n",
    "        \n",
    "        # Condiment\n",
    "        library(condiments), \n",
    "        library(tradeSeq), \n",
    "        library(SingleCellExperiment), \n",
    "        # library(clusterExperiment), \n",
    "        \n",
    "        # GO \n",
    "        library(msigdbr), \n",
    "        library(fgsea), \n",
    "        \n",
    "        # Data \n",
    "        library(tidyverse), \n",
    "        library(openxlsx), \n",
    "        \n",
    "        # Plotting \n",
    "        library(ComplexHeatmap), \n",
    "        library(circlize), \n",
    "        library(viridis), \n",
    "        library(ggplotify), \n",
    "        library(ComplexUpset), \n",
    "        \n",
    "        # Prallel \n",
    "        library(BiocParallel), \n",
    "        \n",
    "        # Python \n",
    "        library(reticulate)\n",
    "        \n",
    "    )\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74f5bb66-148c-41ee-aaba-608daa0f65d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_seed <- 42\n",
    "set.seed(random_seed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56442cde-5f05-4bb9-b7cb-d6ac316dd8ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(future.globals.maxSize=5000*1024^2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97b081ac-fd83-4f27-8d14-8999d61f535c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ht_opt$message=FALSE # ComplexHeatmap "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfa1fa83-841b-4566-8b14-5f5ec0d6e383",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set working directory to project root\n",
    "setwd(\"/research/peer/fdeckert/FD20200109SPLENO\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02e1e788-04af-4d9c-90f7-8179069579a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Source files\n",
    "source(\"plotting_global.R\")\n",
    "source(\"bin/tradeseq.R\")\n",
    "source(\"bin/seurat_qc.R\")\n",
    "source(\"bin/seurat_dea.R\")\n",
    "source(\"bin/cell_type.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84c33c57-2326-4fcc-bb9a-f8dbd60d1f54",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting Theme\n",
    "ggplot2::theme_set(theme_global_set(size_select=1)) # From project global source()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12a4a4a4-1cf9-4cad-a018-91b37f1b07ae",
   "metadata": {},
   "source": [
    "# Parameter settings and data import "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a5dc025-b813-4856-a38f-b1909e05428d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parallel computing \n",
    "workers <- future::availableCores()\n",
    "print(workers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e13e1ed-4e77-4806-b795-ce951e646809",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Seurat object \n",
    "so <- readRDS(\"data/object/pp.rds\")\n",
    "\n",
    "# CellRank absorption probabilities \n",
    "absorption_probabilities_prog_nacl <- read.csv(\"result/cellrank/absorption_probabilities_prog_nacl.csv\", row.names=1, check.names=FALSE)\n",
    "absorption_probabilities_prog_cpg <- read.csv(\"result/cellrank/absorption_probabilities_prog_cpg.csv\", row.names=1,check.names=FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19834e4e-fa0a-40da-9ab2-dc72f3f6aa80",
   "metadata": {},
   "source": [
    "# Prepare data  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b9b0eaf-e6ab-4385-858c-ab03b9942f2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type_prog=c(\n",
    "    \n",
    "    \"MastP\",\n",
    "    \"MegP\", \n",
    "    \"MEP (1)\", \n",
    "    \"MEP (2)\", \n",
    "    \"MEP (3)\",\n",
    "    \"MEP (4)\", \n",
    "    \"ProEB (1)\",\n",
    "    \"ProEB (2)\",\n",
    "    \"ProEB (3)\",\n",
    "    \"ProEB (4)\",\n",
    "    \"EB (1)\",\n",
    "    \"EB (2)\",\n",
    "    \"EB (3)\", \n",
    "    \"EB (4)\",\n",
    "    \"EB (5)\"\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7945e0c9-cbd3-48f4-b2de-4363fa6ce9b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_prog <- subset(so, subset=cell_type_fine %in% cell_type_prog)\n",
    "so_prog$treatment <- factor(so_prog$treatment, levels=names(color$treatment))\n",
    "so_prog$cell_type_fine <- factor(so_prog$cell_type_fine, levels=cell_type_fine_order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30bbcc2b-e855-44ab-8b2a-056e7bce8629",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_prog <- AddMetaData(so_prog, rbind(absorption_probabilities_prog_nacl, absorption_probabilities_prog_cpg))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18156d59-f131-4421-bf36-9664bc41a5fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_prog_nacl <- subset(so_prog, subset=treatment==\"NaCl\")\n",
    "so_prog_cpg <- subset(so_prog, subset=treatment==\"CpG\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "910deffc-393d-42cc-9b4d-ca05270a90c8",
   "metadata": {},
   "source": [
    "# Absorption probability box plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65eabe5f-88d6-43f6-953f-49d75f3df814",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=15, repr.plot.height=5)\n",
    "\n",
    "data <- select(so_prog@meta.data, treatment, cell_type_fine, MastP, MegP, EB..5.) %>% \n",
    "    reshape2::melt(id_vars=c(\"treatment\", \"cell_type_fine\")) %>% as.data.frame() %>% \n",
    "    dplyr::mutate(cell_type_fine=factor(cell_type_fine, levels=cell_type_prog)) %>% \n",
    "    dplyr::mutate(treatment=factor(treatment, levels=c(\"NaCl\", \"CpG\")))\n",
    "\n",
    "plot <- ggplot(data, aes(x=cell_type_fine, y=value, color=cell_type_fine, fill=cell_type_fine)) + \n",
    "    geom_boxplot(outlier.size=0.5) + \n",
    "    facet_grid(treatment~variable) + \n",
    "    scale_color_manual(values=color[[\"cell_type_fine_detail\"]][names(color[[\"cell_type_fine_detail\"]]) %in% so_prog$cell_type_fine]) + \n",
    "    scale_fill_manual(values=color[[\"cell_type_fine_detail\"]][names(color[[\"cell_type_fine_detail\"]]) %in% so_prog$cell_type_fine]) + \n",
    "    theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0.5))\n",
    "\n",
    "plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1eef99c1-290c-41d5-b0bd-34830999c893",
   "metadata": {},
   "source": [
    "# DEA helper function "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69c160c7-10d2-42e3-bb65-fc49fe5dc14f",
   "metadata": {},
   "outputs": [],
   "source": [
    "voom_lm_fit <- function(grouping_var_i, so, grouping_var, pseudobatch_var, sample_weights) {\n",
    "    \n",
    "    # Get relevant genes for grouping_var \n",
    "    cnt <- GetAssayData(so, assay=\"RNA\", slot=\"counts\")\n",
    "    cnt <- cnt[, so[[grouping_var, drop=TRUE]]==grouping_var_i]\n",
    "    cnt <- cnt[rowSums(cnt>3)>3, ]\n",
    "    genes_i <- rownames(cnt)\n",
    "    \n",
    "    # Get Counts \n",
    "    cnt <- GetAssayData(so, assay=\"RNA\", slot=\"counts\")\n",
    "    \n",
    "    # Subset counts \n",
    "    cnt <- cnt[genes_i, ]\n",
    "    \n",
    "    # Prepare count data for split \n",
    "    cnt <- t(as.matrix(cnt))\n",
    "    cnt <- as.data.frame(cnt)\n",
    "        \n",
    "    # Make pseudobulks by suming single cell counts\n",
    "    cnt <- split(cnt, f=paste0(so[[grouping_var, drop=TRUE]], \"|\", so[[pseudobatch_var, drop=TRUE]]))\n",
    "    cnt <- lapply(names(cnt), function(i) {x <- data.frame(counts=colSums(cnt[[i]])); colnames(x) <- i; return(x)})\n",
    "    cnt <- do.call(cbind, cnt)\n",
    "        \n",
    "    # Get grouping variables from cnt matrix\n",
    "    grouping_var_vec <- sapply(strsplit(colnames(cnt), \"\\\\|\"), `[[`, 1) # Used for design matrix\n",
    "    pseudobatch_var_vec <- sapply(strsplit(colnames(cnt), \"\\\\|\"), `[[`, 2) # Used for design matrix\n",
    "    \n",
    "    # Design matrix \n",
    "    grouping_var_vec <- as.factor(grouping_var_vec)\n",
    "    grouping_var_vec <- relevel(grouping_var_vec, ref=grouping_var_i)\n",
    "    pseudobatch_var_vec <- as.character(pseudobatch_var_vec)\n",
    "    design <- model.matrix(~0+grouping_var_vec)\n",
    "    \n",
    "    suppressMessages(\n",
    "        \n",
    "        fit <- edgeR::voomLmFit(\n",
    "\n",
    "            counts=cnt, \n",
    "            design=design, \n",
    "            sample.weights=sample_weights, \n",
    "            var.design=design\n",
    "\n",
    "        )\n",
    "        \n",
    "    )\n",
    "    \n",
    "    return(fit)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39670e15-bc5d-4a2b-bbed-5b3467966c6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ebayes <- function(fit) {\n",
    "    \n",
    "    # Get grouping var and design from fit\n",
    "    design <- fit$design\n",
    "    \n",
    "    # Define result list\n",
    "    result <- list()\n",
    "    \n",
    "    # Grouping var vec \n",
    "    grouping_var_vec <- grep(\"grouping_var_vec\", colnames(design), value=TRUE)\n",
    "    grouping_var_ref <- grouping_var_vec[1]\n",
    "    grouping_var_qry <- grouping_var_vec[-1]\n",
    "    \n",
    "    # Contrast fit\n",
    "    contrasts_vec <- paste0(grouping_var_ref, \"-(\", paste0(grouping_var_qry, collapse=\"+\"), \")/\", length(grouping_var_qry))\n",
    "    contrasts <- limma::makeContrasts(contrasts=contrasts_vec, levels=colnames(design))\n",
    "    contrasts_fit <- limma::contrasts.fit(fit, contrasts=contrasts)\n",
    "\n",
    "    # eBayes fit \n",
    "    efit <- limma::eBayes(contrasts_fit)\n",
    "\n",
    "    # Get result table\n",
    "    result_df <- limma::topTable(efit, sort.by=\"P\", n=Inf, p.value=1, lfc=0, coef=1)\n",
    "\n",
    "    # Convert results to Seurat format \n",
    "    colnames(result_df)[1] <- \"avg_log2FC\"\n",
    "    colnames(result_df)[4] <- \"p_value\"\n",
    "    colnames(result_df)[5] <- \"p_val_adj\"\n",
    "\n",
    "    return(result_df)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cf7a458-73fd-4d03-bbdc-ba73795d6cfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "dea_vp <- function(dea, log2fc_thr=1, p_adj_thr=0.05, top_label=10, title=NULL, color_neg=RColorBrewer::brewer.pal(8, \"Set1\")[1], color_pos=RColorBrewer::brewer.pal(8, \"Set1\")[2]) {\n",
    "\n",
    "    # Set rownames to genes\n",
    "    if(\"gene\" %in% colnames(dea)) {rownames(dea) <- dea$gene}\n",
    "    \n",
    "    # Annotate entries significance by log2fc_thr and p_adj_thr\n",
    "    dea$p_val_adj <- ifelse(dea$p_val_adj == 0, .Machine$double.xmin, dea$p_val_adj)\n",
    "    dea$sig <- ifelse(abs(dea$avg_log2FC) >= log2fc_thr & -log10(dea$p_val_adj) >= -log10(p_adj_thr), \"s\", \"ns\")\n",
    "    \n",
    "    # Set color based on significance and direction of dea e.g. positive and negative \n",
    "    dea$color <- ifelse(dea$sig == \"s\" & dea$avg_log2FC > 0, \"s_pos\", \"ns\")\n",
    "    dea$color <- ifelse(dea$sig == \"s\" & dea$avg_log2FC < 0, \"s_neg\", dea$color)\n",
    "    \n",
    "    color <- c(color_neg, \"gray\", \"black\", color_pos)\n",
    "    names(color) <- c(\"s_pos\", \"ns\", \"black\", \"s_neg\")\n",
    "    \n",
    "    # Create labels based log2FC and p_val_adj\n",
    "    dea_pos <- dea[dea$avg_log2FC > 0 & dea$sig == \"s\", ]\n",
    "    dea_neg <- dea[dea$avg_log2FC < 0 & dea$sig == \"s\", ]\n",
    "\n",
    "    pos_labels_log2FC <- dea_pos[rev(order(dea_pos$avg_log2FC)), ][1:top_label, ] %>% rownames()\n",
    "    neg_labels_log2FC <- dea_neg[order(dea_neg$avg_log2FC), ][1:top_label, ] %>% rownames()\n",
    "    \n",
    "    pos_labels_p_val_adj <- dea_pos[order(dea_pos$p_val_adj), ][1:top_label, ] %>% rownames()\n",
    "    neg_labels_p_val_adj <- dea_neg[order(dea_neg$p_val_adj), ][1:top_label, ] %>% rownames()\n",
    "    \n",
    "    pos_labels <- c(pos_labels_log2FC, pos_labels_p_val_adj)\n",
    "    neg_labels <- c(neg_labels_log2FC, neg_labels_p_val_adj)\n",
    "    \n",
    "    # Set labels \n",
    "    dea$label <- ifelse(rownames(dea) %in% c(pos_labels, neg_labels), rownames(dea), NA)\n",
    "\n",
    "    # Plot\n",
    "    vp <- ggplot(dea, aes(x=avg_log2FC, y=-log10(p_val_adj), fill=dea$color, label=label), alpha=1) + \n",
    "    \n",
    "        geom_point(size=4, shape=21, color=\"white\") + \n",
    "        geom_vline(aes(xintercept=log2fc_thr), linetype=\"dotted\", colour=\"black\") +\n",
    "        geom_vline(aes(xintercept=-log2fc_thr), linetype=\"dotted\", colour=\"black\") +\n",
    "        geom_hline(aes(yintercept=-log10(p_adj_thr)), linetype=\"dotted\", colour=\"black\") +\n",
    "        ggrepel::geom_text_repel(segment.color=\"black\", force=20, force_pull=1, max.overlaps=getOption(\"ggrepel.max.overlaps\", default=100), size=5, alpha=1, guide=\"none\", segment.size=0.1, color=\"black\") + \n",
    "        xlim(-max(abs(dea$avg_log2FC)), max(abs(dea$avg_log2FC))) +  \n",
    "        ylim(0, max(-log10(dea$p_val_adj))+1) + \n",
    "        ggtitle(title) + xlab(\"average log2FC\") + ylab(\"-log10(adj. p-value)\") + \n",
    "        scale_fill_manual(values=color) + \n",
    "    \n",
    "        guides(\n",
    "            \n",
    "            color=guide_legend(order=1, title=\"Group\", size=2, keywidth=0.75, keyheight=0.75), \n",
    "            alpha=\"none\"\n",
    "            \n",
    "        ) + \n",
    "    \n",
    "        theme(\n",
    "\n",
    "            legend.position=\"none\", \n",
    "            aspect.ratio=1\n",
    "\n",
    "        )\n",
    "    \n",
    "    return(vp)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64a93ac6-3a09-4e59-a29c-37498a894e28",
   "metadata": {},
   "source": [
    "# Endpoint DEA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b70704fd-35f8-40f7-93c3-41e766f4cb56",
   "metadata": {},
   "outputs": [],
   "source": [
    "grouping_var=\"cell_type_fine\"\n",
    "pseudobatch_var=\"sample_name\"\n",
    "\n",
    "p_adj_thr=0.05\n",
    "sample_weights=FALSE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "224d8e68-6028-4d0a-9fd0-f3769920a5a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_prog_ep <- subset(so_prog, subset=cell_type_fine %in% c(\"MastP\", \"MegP\", \"EB (5)\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3544b20e-fc42-4efd-9707-dee60ebcbabf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store original names for plotting \n",
    "so_prog_ep$grouping_var_set <- so_prog_ep[[grouping_var, drop=TRUE]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec710df8-0fce-4793-b0d5-b0e7351a8577",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set names \n",
    "so_prog_ep[[grouping_var]] <- make.names(so_prog_ep[[grouping_var, drop=TRUE]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9b9f1d1-e6c4-4d3b-bc6e-81028637e17b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DEA voom fit \n",
    "fit <- lapply(unique(so_prog_ep[[grouping_var, drop=TRUE]]), voom_lm_fit, so=so_prog_ep, grouping_var=grouping_var, pseudobatch_var=pseudobatch_var, sample_weights=sample_weights)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84649389-6127-4407-ae6e-21400852f6cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get results \n",
    "results <- lapply(fit, ebayes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb79bba5-5b79-47e0-b41c-32a4e0ec1ef2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set original grouping names \n",
    "rename_df <- distinct(so_prog_ep@meta.data[, c(grouping_var, \"grouping_var_set\")])\n",
    "rename_df[1] <- paste0(\"grouping_var_vec\", rename_df[[1]])\n",
    "rename_df <- rename_df[order(rename_df$grouping_var_set), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68919fb4-e518-4f41-a5a5-ccd461c30e0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(results) <- paste0(\"grouping_var_vec\", unique(so_prog_ep[[grouping_var, drop=TRUE]]))\n",
    "results <- results[rename_df[[1]]]\n",
    "names(results) <- rename_df[[2]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40abbe4c-5995-408b-aaec-2b93336053f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat <- GetAssayData(so_prog_ep, assay=\"RNA\", slot=\"counts\")\n",
    "mat_order <- so_prog_ep$grouping_var_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d938ef6-e8d2-4388-bdf1-685111554b2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "results <- lapply(names(results), function(i, mat, mat_order) {\n",
    "    \n",
    "    x <- results[[i]]\n",
    "    mat <- mat[rownames(x), ]\n",
    "    mat <- mat>0\n",
    "    x$pct_1<-rowSums(mat[, mat_order==i])/sum(mat_order==i)\n",
    "    x$pct_2<-rowSums(mat[, mat_order!=i])/sum(mat_order!=i)\n",
    "\n",
    "    return(x)\n",
    "\n",
    "}, mat=mat, mat_order=mat_order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "483d6227-ef42-4057-98bc-a7be1249226e",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(results) <- rename_df[[2]]\n",
    "results <- results[c(\"MastP\", \"MegP\", \"EB (5)\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "372f4757-6921-4abc-a33c-e5d48348ae05",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=8*3, repr.plot.height=8*1)\n",
    "\n",
    "vp <- lapply(names(results), function(i) {dea_vp(results[[i]], title=i, log2fc_thr=0, p_adj_thr=p_adj_thr)})\n",
    "wrap_plots(vp, ncol=3, nrow=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9747b929-8562-4199-995a-ee0798801c85",
   "metadata": {},
   "source": [
    "# Start point DEA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "767f2062-d306-465f-9a1e-88cc0f534e5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "grouping_var=\"cell_type_fine\"\n",
    "pseudobatch_var=\"sample_name\"\n",
    "\n",
    "p_adj_thr=0.05\n",
    "sample_weights=FALSE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76b2243a-7031-42d6-ab96-a5b4b4f9ccea",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_prog_ep <- subset(so_prog, subset=cell_type_fine %in% c(\"MEP (1)\", \"MEP (2)\", \"MEP (3)\", \"MEP (4)\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2f4afc1-6e57-4703-b6a1-5967ca5ee21d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store original names for plotting \n",
    "so_prog_ep$grouping_var_set <- so_prog_ep[[grouping_var, drop=TRUE]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "979226f5-4da6-4ca7-bc34-51021ca44efa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set names \n",
    "so_prog_ep[[grouping_var]] <- make.names(so_prog_ep[[grouping_var, drop=TRUE]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e40c605-691a-414b-a239-2669b6a7e820",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DEA voom fit \n",
    "fit <- lapply(unique(so_prog_ep[[grouping_var, drop=TRUE]]), voom_lm_fit, so=so_prog_ep, grouping_var=grouping_var, pseudobatch_var=pseudobatch_var, sample_weights=sample_weights)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d70337d-e9ed-4e39-a832-8e5ae594b66f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get results \n",
    "results <- lapply(fit, ebayes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07783926-754c-4704-a61e-615545d0f9ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set original grouping names \n",
    "rename_df <- distinct(so_prog_ep@meta.data[, c(grouping_var, \"grouping_var_set\")])\n",
    "rename_df[1] <- paste0(\"grouping_var_vec\", rename_df[[1]])\n",
    "rename_df <- rename_df[order(rename_df$grouping_var_set), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "716fee45-ade6-4a11-b08b-e52d0bbd440c",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(results) <- paste0(\"grouping_var_vec\", unique(so_prog_ep[[grouping_var, drop=TRUE]]))\n",
    "results <- results[rename_df[[1]]]\n",
    "names(results) <- rename_df[[2]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1520777e-b9c1-4c24-9f81-ed001f7ebf6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat <- GetAssayData(so_prog_ep, assay=\"RNA\", slot=\"counts\")\n",
    "mat_order <- so_prog_ep$grouping_var_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15dc6f7e-4fb2-423d-b801-b3d079921b86",
   "metadata": {},
   "outputs": [],
   "source": [
    "results <- lapply(names(results), function(i, mat, mat_order) {\n",
    "    \n",
    "    x <- results[[i]]\n",
    "    mat <- mat[rownames(x), ]\n",
    "    mat <- mat>0\n",
    "    x$pct_1<-rowSums(mat[, mat_order==i])/sum(mat_order==i)\n",
    "    x$pct_2<-rowSums(mat[, mat_order!=i])/sum(mat_order!=i)\n",
    "\n",
    "    return(x)\n",
    "\n",
    "}, mat=mat, mat_order=mat_order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8465ec69-682d-457f-b1e3-da52c5aee68b",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(results) <- rename_df[[2]]\n",
    "results <- results[c(\"MEP (1)\", \"MEP (2)\", \"MEP (3)\", \"MEP (4)\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64e67b94-c76a-4a53-aedc-7a1f57296791",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=8*4, repr.plot.height=8*1)\n",
    "\n",
    "vp <- lapply(names(results), function(i) {dea_vp(results[[i]], title=i, log2fc_thr=0, p_adj_thr=p_adj_thr)})\n",
    "wrap_plots(vp, ncol=4, nrow=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "703ffe0b-ef26-44d9-ae49-3bdf8732fc2f",
   "metadata": {},
   "source": [
    "# Condiment + Diffusion pseudotime "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d71d0c04-af33-43f8-b76b-e156210f45f6",
   "metadata": {},
   "source": [
    "## Prepare data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6367610-7c34-4103-a5f9-eeba067effe7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=25, repr.plot.height=5)\n",
    "\n",
    "so_eb <- suppressMessages(import_cellrank(so_prog, suffix=\"_eb\", absorption_probability_col=\"EB..5.\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5f665d2-21a1-43bc-aee0-a0e0725dfb10",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10, repr.plot.height=2.5)\n",
    "\n",
    "pdp_1 <- pseudotime_density_plot(so_eb, pseudotime=\"dpt_pseudotime\") \n",
    "pdp_2 <- pseudotime_density_plot(so_eb, pseudotime=\"dpt_pseudotime_treatment\")\n",
    "\n",
    "pdp_1 + pdp_2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15929900-e3f4-4dbb-b827-c67ae82d5fe6",
   "metadata": {},
   "source": [
    "## Scale pseudotime for each treatment group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f0ee885-29e2-4415-aa18-01c2a6775b6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "dpt_pseudotime_nacl <- subset(so_eb, subset=treatment==\"NaCl\")@meta.data[\"dpt_pseudotime\"]\n",
    "dpt_pseudotime_nacl$dpt_pseudotime <- scales::rescale(dpt_pseudotime_nacl$dpt_pseudotime)\n",
    "\n",
    "dpt_pseudotime_cpg <- subset(so_eb, subset=treatment==\"CpG\")@meta.data[\"dpt_pseudotime\"]\n",
    "dpt_pseudotime_cpg$dpt_pseudotime <- scales::rescale(dpt_pseudotime_cpg$dpt_pseudotime)\n",
    "\n",
    "dpt_pseudotime <- rbind(dpt_pseudotime_nacl, dpt_pseudotime_cpg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e88d68e9-a426-4eec-9019-aec479b424e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary(subset(so_eb, subset=treatment==\"NaCl\")$dpt_pseudotime)\n",
    "summary(subset(so_eb, subset=treatment==\"CpG\")$dpt_pseudotime)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc568cfd-cb73-43c2-8588-6f323c56ae78",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_eb$dpt_pseudotime <- NULL\n",
    "so_eb <- AddMetaData(so_eb, dpt_pseudotime)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5cb619d-6859-468f-b3ff-a97925c84abb",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary(subset(so_eb, subset=treatment==\"NaCl\")$dpt_pseudotime)\n",
    "summary(subset(so_eb, subset=treatment==\"CpG\")$dpt_pseudotime)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a43185c-c7b8-4369-bb03-782c4e643944",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10, repr.plot.height=2.5)\n",
    "\n",
    "pdp_1 <- pseudotime_density_plot(so_eb, pseudotime=\"dpt_pseudotime\") \n",
    "pdp_2 <- pseudotime_density_plot(so_eb, pseudotime=\"dpt_pseudotime_treatment\")\n",
    "\n",
    "pdp_1 + pdp_2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "812f6642-e6c9-46a2-a857-c3573f395711",
   "metadata": {},
   "source": [
    "# TradeSeq workflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec0e08a6-4e98-4e4e-9d64-e04175253e4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "cnt_nacl <- GetAssayData(subset(so_eb, subset=treatment==\"NaCl\"), assay=\"RNA\", slot=\"counts\")\n",
    "genes_nacl <- rownames(cnt_nacl[rowSums(cnt_nacl>=3)>5, ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36f2e896-e112-4079-8f27-2d8150545c82",
   "metadata": {},
   "outputs": [],
   "source": [
    "cnt_cpg <- GetAssayData(subset(so_eb, subset=treatment==\"CpG\"), assay=\"RNA\", slot=\"counts\")\n",
    "genes_cpg <- rownames(cnt_cpg[rowSums(cnt_cpg>=3)>5, ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6729e37-d75d-4d1a-9704-640af15473fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes <- unique(c(genes_nacl, genes_cpg))\n",
    "genes <- genes[!genes==\"Igkc\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ca2fc17-ff2c-4560-8cd6-7c90f2c4f6ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "tradeseq_eb <- tradeseq_workflow(so_eb, so_eb$dpt_pseudotime, conditions=factor(so_eb$treatment, levels=c(\"NaCl\", \"CpG\")), genes=genes, suffix=\"_eb\", log2_thr=0, nknots=7, workers=workers, compute=FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5d1ee93-fedd-4dc4-a395-2ff4b31a69b7",
   "metadata": {},
   "source": [
    "## Evaluate K"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8abf60e3-67ed-4825-833c-0f25a58cbffe",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=15, repr.plot.height=5)\n",
    "\n",
    "plot_evalutateK_results(tradeseq_eb[[\"evaluate_k\"]], k=NULL, aicDiff=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "626282bd-181a-426c-a644-972b527775d5",
   "metadata": {},
   "source": [
    "## Pseudotime associated genes (PTAG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b99c3d68-1fa7-462c-b082-048292d688bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "association_eb <- tradeseq_eb[[\"association\"]]\n",
    "\n",
    "colnames(association_eb) <- c(\"wald_stat\", \"df\", \"pval\", \"wald_stat_nacl\", \"df_nacl\", \"pval_nacl\", \"wald_stat_cpg\", \"df_cpg\", \"pval_cpg\", \"mean_logfc\")\n",
    "association_eb <- association_eb[, c(\"wald_stat_nacl\", \"df_nacl\", \"pval_nacl\", \"wald_stat_cpg\", \"df_cpg\", \"pval_cpg\")]\n",
    "\n",
    "association_eb$pval_nacl <- ifelse(association_eb$pval_nacl==0, min(na.omit(association_eb$pval_nacl[association_eb$pval_nacl>0])), association_eb$pval_nacl)\n",
    "association_eb$pval_cpg <- ifelse(association_eb$pval_cpg==0, min(na.omit(association_eb$pval_cpg[association_eb$pval_cpg>0])), association_eb$pval_cpg)\n",
    "\n",
    "association_eb$padj_nacl <- p.adjust(association_eb$pval_nacl, \"fdr\")\n",
    "association_eb$padj_cpg <- p.adjust(association_eb$pval_cpg, \"fdr\")\n",
    "\n",
    "association_eb$padj_nacl <- ifelse(association_eb$padj_nacl==0, min(na.omit(association_eb$padj_nacl[association_eb$padj_nacl>0])), association_eb$padj_nacl)\n",
    "association_eb$padj_cpg <- ifelse(association_eb$padj_cpg==0, min(na.omit(association_eb$padj_cpg[association_eb$padj_cpg>0])), association_eb$padj_cpg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d66d5a1-0ad8-49f1-9ed5-e8630284a029",
   "metadata": {},
   "outputs": [],
   "source": [
    "association_eb <- association_eb[!(is.na(association_eb$wald_stat_nacl) & is.na(association_eb$wald_stat_cpg)), ]\n",
    "association_eb <- association_eb[(association_eb$wald_stat_nacl>0 | association_eb$wald_stat_cpg>0), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53f7ba4a-79aa-4b55-9454-55542c4898ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "association_eb_nacl <- association_eb[association_eb$wald_stat_nacl > 0 & is.na(association_eb$wald_stat_nacl)==FALSE, ]\n",
    "association_eb_cpg <- association_eb[association_eb$wald_stat_cpg > 0 & is.na(association_eb$wald_stat_cpg)==FALSE, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "374bafde-ffca-4e24-9efe-cf175945ddeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "association_eb_nacl$p_value <- 1-stats::pchisq(association_eb_nacl$wald_stat_nacl, df=association_eb_nacl$df_nacl, lower.tail=TRUE, log.p=FALSE)\n",
    "association_eb_nacl$fdr <- p.adjust(association_eb_nacl$p_value, method=\"fdr\")\n",
    "\n",
    "association_eb_cpg$p_value <- 1-stats::pchisq(association_eb_cpg$wald_stat_cpg, df=association_eb_cpg$df_cpg, lower.tail=TRUE, log.p=FALSE)\n",
    "association_eb_cpg$fdr <- p.adjust(association_eb_cpg$p_value, method=\"fdr\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2728f42-ee2f-4f83-ad85-d94a58d81034",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_ptag_nacl <- rownames(association_eb_nacl[association_eb_nacl$fdr <= 0.001, ])\n",
    "genes_ptag_cpg <- rownames(association_eb_cpg[association_eb_cpg$fdr <= 0.001, ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61cee46e-d6ed-4811-9b15-0c090e4053eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "association_eb$ptag_nacl <- ifelse(rownames(association_eb) %in% genes_ptag_nacl, TRUE, FALSE)\n",
    "association_eb$ptag_cpg <- ifelse(rownames(association_eb) %in% genes_ptag_cpg, TRUE, FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa82b7ac-190e-46fe-912b-0e78b34138d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10, repr.plot.height=5)\n",
    "\n",
    "mean_nacl <- mean(log(association_eb_nacl$wald_stat_nacl))\n",
    "sd_nacl <- sd(log(association_eb_nacl$wald_stat_nacl))\n",
    "\n",
    "p_1 <- ggplot(association_eb[association_eb$wald_stat_nacl > 0, ], aes(x=log(wald_stat_nacl))) + \n",
    "    geom_histogram(bins=25) + \n",
    "    geom_vline(xintercept=mean_nacl, color=\"red\", size=1) + \n",
    "    geom_vline(xintercept=mean_nacl-sd_nacl, color=\"red\", size=1, linetype=\"longdash\") + \n",
    "    geom_vline(xintercept=mean_nacl+sd_nacl, color=\"red\", size=1, linetype=\"longdash\") + \n",
    "    ggtitle(\"Wald stat (NaCl)\") + \n",
    "    theme(aspect.ratio=1)\n",
    "\n",
    "mean_cpg <- mean(log(association_eb_cpg$wald_stat_cpg))\n",
    "sd_cpg <- sd(log(association_eb_cpg$wald_stat_cpg))\n",
    "\n",
    "p_2 <- ggplot(association_eb[association_eb$wald_stat_cpg > 0, ], aes(x=log(wald_stat_cpg))) + \n",
    "    geom_histogram(bins=25) + \n",
    "    geom_vline(xintercept=mean_cpg, color=\"red\", size=1) + \n",
    "    geom_vline(xintercept=mean_cpg-sd_cpg, color=\"red\", size=1, linetype=\"longdash\") + \n",
    "    geom_vline(xintercept=mean_cpg+sd_cpg, color=\"red\", size=1, linetype=\"longdash\") + \n",
    "    ggtitle(\"Wald stat (CpG)\") + \n",
    "    theme(aspect.ratio=1)\n",
    "\n",
    "p_1 + p_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "186e75d4-af04-44c6-a393-439fa26f8981",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10, repr.plot.height=5)\n",
    "\n",
    "p_1 <- ggplot(association_eb, aes(x=-log10(padj_nacl), y=log(wald_stat_nacl), color=ptag_nacl)) + \n",
    "    geom_point(size=2) + \n",
    "    ggtitle(\"Wald stat vs padj (NaCl)\") + \n",
    "    scale_color_manual(values=c(\"gray\", \"black\")) + \n",
    "    theme(aspect.ratio=1, legend.position=\"none\")\n",
    "\n",
    "p_2 <- ggplot(association_eb, aes(x=-log10(padj_cpg), y=log(wald_stat_cpg), color=ptag_cpg)) + \n",
    "    geom_point(size=2) + \n",
    "    ggtitle(\"Wald stat vs padj (CpG)\") + \n",
    "    scale_color_manual(values=c(\"gray\", \"black\")) + \n",
    "    theme(aspect.ratio=1, legend.position=\"none\")\n",
    "\n",
    "p_1 + p_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "207bfacc-aa50-4d9b-b00f-20046a409c9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "association_eb <- association_eb[(association_eb$ptag_nacl | association_eb$ptag_cpg), ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ade978f-a7a1-4de3-b34f-a5d21e64f4b5",
   "metadata": {},
   "source": [
    "## Pseudotime condition genes (PTCG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "143141e4-0811-4896-b92b-ef809e3b205c",
   "metadata": {},
   "outputs": [],
   "source": [
    "condition_eb <- conditionTest(tradeseq_eb[[\"fitgam\"]], l2fc=log2(1))\n",
    "condition_eb <- condition_eb[rownames(condition_eb) %in% rownames(association_eb), ]\n",
    "\n",
    "colnames(condition_eb) <- c(\"wald_stat_ptcg\", \"df_ptcg\", \"pval_ptcg\")\n",
    "\n",
    "condition_eb$pval_ptcg <- ifelse(condition_eb$pval_ptcg==0, min(na.omit(condition_eb$pval_ptcg[condition_eb$pval_ptcg>0])), condition_eb$pval_ptcg)\n",
    "condition_eb$padj_ptcg <- p.adjust(condition_eb$pval_ptcg, \"fdr\")\n",
    "condition_eb$padj_ptcg <- ifelse(condition_eb$padj_ptcg==0, min(na.omit(condition_eb$padj_ptcg[condition_eb$padj_ptcg>0])), condition_eb$padj_ptcg)\n",
    "\n",
    "condition_eb$ptcg <- ifelse(condition_eb$padj_ptcg<=0.001, TRUE, FALSE)\n",
    "condition_eb$ptcg <- ifelse(is.na(condition_eb$ptcg), FALSE, condition_eb$ptcg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7dd296a-5280-4f16-ae55-db60551cd20b",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=5, repr.plot.height=5)\n",
    "\n",
    "p_1 <- ggplot(condition_eb, aes(x=-log10(padj_ptcg), y=log10(wald_stat_ptcg), color=ptcg)) + \n",
    "    geom_point(size=2) + \n",
    "    ggtitle(\"Condition genes\") + \n",
    "    scale_color_manual(values=c(\"gray\", \"black\")) + \n",
    "    theme(\n",
    "        aspect.ratio=1, \n",
    "        legend.position=\"none\"\n",
    "    )\n",
    "\n",
    "p_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c1c403d-0425-4411-8dc3-3ae3573ad8e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "association_eb$ptcg <- ifelse(rownames(association_eb) %in% rownames(condition_eb[condition_eb$ptcg, ]), TRUE, FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e2272ab-4c0b-4652-9a41-02adacabc3cf",
   "metadata": {},
   "source": [
    "## Combine and save PTAG/PTCG results "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed2d1997-2807-4ffa-aed3-56a094d70d0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "association_eb <- cbind(condition_eb[, c(\"wald_stat_ptcg\", \"df_ptcg\", \"pval_ptcg\", \"padj_ptcg\")], association_eb)\n",
    "saveRDS(association_eb, \"result/tradeseq/ptg_eb.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "932689d9-5665-4c06-a941-d839838c608c",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=5, repr.plot.height=5)\n",
    "\n",
    "upset(data=association_eb, intersect=c(\"ptag_nacl\", \"ptag_cpg\", \"ptcg\"), width_ratio=0.1, sort_sets=FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7aaa8a6c-677b-46be-9bc8-28ebecd0715d",
   "metadata": {},
   "source": [
    "## Smoothed gene expression "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2ea2bf2-3791-4147-bf0b-9cd7ce5394b0",
   "metadata": {},
   "source": [
    "### Marker genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d742abb-3804-4f20-85b1-760bd8295a11",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes <- c(\"Kit\", \"Cd47\", \"Gata1\", \"Gata2\", \"Epor\", \"Tfrc\", \"Klf1\", \"Tal1\", \"Stat1\", \"Stat5a\", \"Rb1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d7c540d-9516-426b-8abb-973b15b705c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10*3, repr.plot.height=ceiling(length(genes)/10)*3)\n",
    "\n",
    "p <- lapply(genes, function(i) plot_smooth(tradeseq_eb[[\"fitgam\"]], i, point=FALSE) + ggtitle(i) + theme(legend.position=\"none\"))\n",
    "wrap_plots(p, ncol=10, nrow=ceiling(length(genes)/10))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb2f9a92-7e32-4544-8c17-074b2a6523c8",
   "metadata": {},
   "source": [
    "### Transcription factor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6ee8049-ceab-42b8-a847-dfd07b2913ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes <- read.delim(\"data/annotation/animaltfdb3/Mus_musculus_TF.txt\")$Symbol %>% unique()\n",
    "genes <- genes[genes %in% rownames(association_eb[(association_eb$ptag_nacl==TRUE | association_eb$ptag_cpg==TRUE), ])]\n",
    "genes <- genes[genes %in% rownames(tradeseq_eb[[\"fitgam\"]])]\n",
    "length(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25932aa3-5834-4d6f-9bcd-8d9deeb7374c",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10*2.5, repr.plot.height=ceiling(length(genes)/10)*2.5)\n",
    "\n",
    "p <- lapply(genes, function(i) plot_smooth(tradeseq_eb[[\"fitgam\"]], i) + ggtitle(i) + theme(legend.position=\"none\"))\n",
    "wrap_plots(p, ncol=10, nrow=ceiling(length(genes)/10))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adb3bc50-7255-4f46-bab8-c268fb456679",
   "metadata": {},
   "source": [
    "### Ligand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "904389d5-7986-4884-9fc9-57e353346455",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes <- CellChat::CellChatDB.mouse$interaction[[\"ligand\"]]%>% unique()\n",
    "genes <- genes[genes %in% rownames(association_eb[(association_eb$ptag_nacl==TRUE | association_eb$ptag_cpg==TRUE), ])]\n",
    "genes <- genes[genes %in% rownames(tradeseq_eb[[\"fitgam\"]])]\n",
    "length(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b0a0b5f-5ac9-48b0-a407-03ebb17f1336",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10*2.5, repr.plot.height=ceiling(length(genes)/10)*2.5)\n",
    "\n",
    "p <- lapply(genes, function(i) plot_smooth(tradeseq_eb[[\"fitgam\"]], i) + ggtitle(i) + theme(legend.position=\"none\"))\n",
    "wrap_plots(p, ncol=10, nrow=ceiling(length(genes)/10))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5855c0a9-433b-42a8-b5c5-a947b037649a",
   "metadata": {},
   "source": [
    "### Receptor "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98896894-e7cd-4e17-978d-e351a7f32e57",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes <- CellChat::CellChatDB.mouse$interaction[[\"receptor\"]] %>% unique()\n",
    "genes <- genes[genes %in% rownames(association_eb[(association_eb$ptag_nacl==TRUE | association_eb$ptag_cpg==TRUE), ])]\n",
    "genes <- genes[genes %in% rownames(tradeseq_eb[[\"fitgam\"]])]\n",
    "length(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78a600b9-efd7-493c-b10b-6cf417e50494",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10*2.5, repr.plot.height=ceiling(length(genes)/10)*2.5)\n",
    "\n",
    "p <- lapply(genes, function(i) plot_smooth(tradeseq_eb[[\"fitgam\"]], i) + ggtitle(i) + theme(legend.position=\"none\"))\n",
    "wrap_plots(p, ncol=10, nrow=ceiling(length(genes)/10))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "r.4.1.0-FD20200109SPLENO",
   "language": "R",
   "name": "r.4.1.0-fd20200109spleno"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
