{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bb817fcf-c6d9-49ee-9dda-70412056f1bc",
   "metadata": {},
   "source": [
    "# Differential expression analysis (DEA)\n",
    "\n",
    "## CpG vs NaCl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98503a66-391e-4cc0-a25a-07280874bda1",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)\n",
    "options(dplyr.summarise.inform=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1b912db-f423-4ca4-bd89-d211f9231c95",
   "metadata": {},
   "outputs": [],
   "source": [
    "library_load <- suppressMessages(\n",
    "    \n",
    "    list(\n",
    "        \n",
    "        # Seurat \n",
    "        library(Seurat), \n",
    "    \n",
    "        \n",
    "        # GSEA \n",
    "        library(msigdbr), \n",
    "        library(fgsea), \n",
    "        \n",
    "        # Data \n",
    "        library(tidyverse), \n",
    "        library(openxlsx), \n",
    "        \n",
    "        # Plotting \n",
    "        library(ComplexHeatmap), \n",
    "        library(circlize), \n",
    "        library(viridis), \n",
    "        library(ggplotify), \n",
    "        library(patchwork), \n",
    "        library(ggrepel), \n",
    "        \n",
    "        # Python \n",
    "        library(reticulate)\n",
    "        \n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efe5b0e6-b9ee-419a-b7aa-fc41b5f46919",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Configure reticulate \n",
    "use_condaenv(condaenv='p.3.8.12-FD20200109SPLENO', conda=\"/nobackup/peer/fdeckert/miniconda3/bin/conda\", required=NULL)\n",
    "py_config()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd46bf4f-5b76-499b-ae8c-a3038368b1bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_seed <- 42\n",
    "set.seed(random_seed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b02fdf07-34ca-4e76-abdc-5dd1696cb0a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set working directory to project root\n",
    "setwd(\"/research/peer/fdeckert/FD20200109SPLENO\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbb10b2b-2010-48f6-bb8e-96c2b1e239c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Source files\n",
    "source(\"plotting_global.R\")\n",
    "source(\"bin/seurat_qc.R\")\n",
    "source(\"bin/pbDEA.R\")\n",
    "source(\"bin/gsea.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ec40480-8638-44a6-b9f2-7f8d55359e8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting Theme\n",
    "ggplot2::theme_set(theme_global_set(size_select=1)) # From project global source()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba7147ea-8464-44e2-8049-b439337881d5",
   "metadata": {},
   "source": [
    "# Parameter settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a90f53a7-9679-4d20-9bb5-e5066d84127e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Seurat object file\n",
    "so_rds_file <- \"data/object/pp.rds\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b20f495e-1320-4ae9-9fdb-1c4bae1f5d54",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DEA settings \n",
    "grouping_var=\"treatment\"\n",
    "\n",
    "random_effect_var=\"sample_group\"\n",
    "random_effect_var_cell_min=3\n",
    "random_effect_var_min=2\n",
    "\n",
    "pseudoreplicate_n=5\n",
    "pseudoreplicate_var=\"pseudoreplicate\"\n",
    "\n",
    "cnt_min=3\n",
    "cell_min=1\n",
    "feature_group_min=2\n",
    "\n",
    "p_adj_thr=0.1\n",
    "sample_weights=FALSE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8bf2e4f-387d-4e72-8e3f-f58e44e90e34",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Result basename \n",
    "basename <- \"result/dea/scRNAseq/nacl_vs_cpg\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "978c68b8-490b-48fa-88ce-7126cc71a604",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use cache \n",
    "cache <- FALSE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20d0f224-963b-4f11-812d-8a597972d419",
   "metadata": {},
   "source": [
    "# Import Seurat object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d10e7ab8-ef30-4f62-9419-a80d605687d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- readRDS(so_rds_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf81cfce-4c9f-4811-8db2-afbf8e45fcf7",
   "metadata": {},
   "source": [
    "# Check if all cell types have both groups present for DEA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e4640b2-ea7f-429d-83c2-ed72e838efdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "so$cell_type_fine <- factor(so$cell_type_fine, levels=names(color$cell_type_fine))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bdf8adc-3a09-4815-8232-aab2b443943b",
   "metadata": {},
   "outputs": [],
   "source": [
    "table <- group_by(so@meta.data, cell_type_fine, sample_group) %>% \n",
    "    dplyr::summarise(n=n()) %>% \n",
    "    reshape2::dcast(., formula=sample_group~cell_type_fine) %>% \n",
    "    column_to_rownames(var=\"sample_group\") %>% \n",
    "    replace(is.na(.), 0)\n",
    "\n",
    "knitr::kable(table, \"html\") %>% as.character() %>% IRdisplay::display_html()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e12307d5-e732-4453-857a-317f717221e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=6, repr.plot.height=6)\n",
    "\n",
    "table <- group_by(so@meta.data, cell_type_fine, treatment) %>% \n",
    "    dplyr::summarise(n=n()) %>% \n",
    "    reshape2::dcast(., formula=treatment~cell_type_fine) %>% \n",
    "    column_to_rownames(var=\"treatment\") %>% \n",
    "    replace(is.na(.), 0)\n",
    "\n",
    "ratio <- sweep(table, MARGIN=1, FUN=\"/\", STATS=rowSums(table)) %>% t %>% as.data.frame %>% tibble::rownames_to_column(., \"cell_type_fine\") %>% \n",
    "    dplyr::mutate(label=ifelse(abs(NaCl-CpG)>=0.01, cell_type_fine, NA))\n",
    "\n",
    "ratio_p <- ggplot(ratio, aes(x=NaCl, y=CpG, color=cell_type_fine, label=label)) + \n",
    "    \n",
    "    geom_point(size=3) + \n",
    "    geom_abline() + \n",
    "    xlim(0, max(ratio[, 2:3])) + ylim(0, max(ratio[, 2:3])) + \n",
    "    geom_text_repel(segment.color=\"black\", force=20, force_pull=1, max.overlaps=getOption(\"ggrepel.max.overlaps\", default=100), size=5, alpha=1, guide=\"none\", segment.size=0.1, color='black') + \n",
    "    scale_color_manual(values=color$cell_type_fine) + \n",
    "    theme(\n",
    "        aspect.ratio=1, \n",
    "        legend.position=\"none\"\n",
    "    \n",
    "    )\n",
    "\n",
    "ratio_p"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "972fc566-9bcf-410b-aabb-0a0544a61707",
   "metadata": {},
   "source": [
    "# Pseudobatch DEA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ebd20a2-f8b6-4deb-9093-be0d8d8234df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set pseudobulk id\n",
    "so[[pseudoreplicate_var]] <- so@meta.data %>% \n",
    "    dplyr::group_by_(\"cell_type_fine\", grouping_var, random_effect_var) %>% \n",
    "    dplyr::mutate(pseudoreplicate_id=sample(rep(1:pseudoreplicate_n, length.out=n()))) %>% \n",
    "    dplyr::mutate(pseudoreplicate_id=paste0(treatment, \"_\", sample_rep, \"_\", pseudoreplicate_id)) %>% \n",
    "    dplyr::pull(pseudoreplicate_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e984035-caed-4b29-a940-cb626063f220",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare data \n",
    "so <- SplitObject(so, split.by=\"cell_type_fine\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4951ded-e36e-4d85-aaa5-3ee52d7b331f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cell type order \n",
    "cell_type_order <- names(color$cell_type_fine)\n",
    "so <- so[cell_type_order]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04015dc3-ba4c-4a69-a3b9-d7aaa5a7f2f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(!cache) {\n",
    "    \n",
    "    # Subset samples \n",
    "    so <- lapply(so, group_select, grouping_var=grouping_var, random_effect_var=random_effect_var, random_effect_var_cell_min=random_effect_var_cell_min, random_effect_var_min=random_effect_var_min)\n",
    "    \n",
    "    # Subset features\n",
    "    so <- lapply(so, feature_select, cnt_min=cnt_min, cell_min=cell_min, random_effect_var=random_effect_var, feature_group_min=feature_group_min)\n",
    "    \n",
    "    # Voom lmFit\n",
    "    results <- lapply(so, voomlmfit_group, grouping_var=grouping_var, random_effect_var=random_effect_var, pseudoreplicate_var=pseudoreplicate_var, contrast_vec=c(\"NaCl\", \"CpG\"), sample_weights=sample_weights) \n",
    "    \n",
    "    # Write xlsx\n",
    "    write_results(results, basename)\n",
    "    \n",
    "} else {\n",
    "    \n",
    "    results <- readRDS(paste0(basename, \"_dea.rds\"))\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6121f68-02f4-4f98-a5fd-81e5c60eb604",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=4*8, repr.plot.height=1*8)\n",
    "\n",
    "vp_treatment <- lapply(seq_along(results), function(i) {dea_vp(results[[i]], title=names(results)[i], log2fc_thr=0, p_adj_thr=p_adj_thr, color_neg=color$treatment[1], color_pos=color$treatment[2]) + theme_global_set(size_select=1)})\n",
    "ggpubr::ggarrange(plotlist=vp_treatment, ncol=4, common.legend=FALSE, legend=\"bottom\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43c8aa89-eb14-4221-ba57-6363fbdf7b5e",
   "metadata": {},
   "source": [
    "# GSEA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b63e990e-ae9c-4b80-a36a-d4bbc822d1b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs_mm <- msigdbr(species=\"Mus musculus\", category=\"C2\", subcategory=\"WIKIPATHWAYS\")\n",
    "gs_mm <- split(x=gs_mm$gene_symbol, f=gs_mm$gs_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "451c519a-a99c-457e-82a6-5a83a1ecace1",
   "metadata": {},
   "outputs": [],
   "source": [
    "results <- lapply(results, function(x) {x$gene <- rownames(x); return(x)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbf4cd71-d675-4ebd-9048-1d477cf183d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(!cache) {\n",
    "    \n",
    "    gsea <- lapply(1:length(results), function(i) {gsea_from_dea(dea=results[[i]], gene_set=gs_mm, gene_name=\"gene\")})\n",
    "    saveRDS(gsea, paste0(basename, \"_gsea.rds\"))\n",
    "    \n",
    "    names(gsea) <- gsub(\"\\\\.\", \"\", make.names(names(results)))\n",
    "    write.xlsx(gsea, paste0(basename, \"_gsea.xlsx\"), colNames=TRUE)\n",
    "    \n",
    "} else {\n",
    "    \n",
    "    gsea <- readRDS(paste0(basename, \"_gsea.rds\"))\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abae022d-9afd-4bbc-9dc5-b140b91bdb78",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=2*20, repr.plot.height=1*10)\n",
    "\n",
    "gsea_p <- lapply(1:length(results), function(i) {gsea_plot(gsea[[i]], pval_thr=p_adj_thr, pathway_suffix=\"WP\", title=names(results)[i]) + theme_global_set(size_select=1)})\n",
    "ggpubr::ggarrange(plotlist=gsea_p, ncol=2, common.legend=FALSE, legend=\"bottom\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0aedf3eb-b1cc-4b3d-baea-3a4b258c9406",
   "metadata": {},
   "source": [
    "# Heatmap - Top DEA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb05329e-c319-4740-9ee9-144dc6341a80",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "hm_dea_genes <- function(dea_file, cell_type_order, cell_type_color, gene_names, log2fc_sign, padj_thr_group=1, n_slice_group=0, padj_thr=0.1, n_slice=10, color_max=\"red\") {\n",
    "    \n",
    "    # DEA results \n",
    "    dea <- readRDS(dea_file)\n",
    "    dea <- do.call(rbind, dea)\n",
    "    dea <- dea %>% dplyr::filter(ident %in% cell_type_order) \n",
    "    \n",
    "    # Re-set cell types in case some are missing from DEA\n",
    "    cell_type_order <- cell_type_order[cell_type_order %in% dea$ident]\n",
    "    \n",
    "    # Check groups - Don't display group with less DEG than n_sclice for padj_thr_group\n",
    "    cell_type_group <- dea %>% \n",
    "        dplyr::filter(sign(avg_log2FC)==log2fc_sign & p_val_adj <= padj_thr_group) %>% \n",
    "        dplyr::group_by(ident) %>% \n",
    "        dplyr::summarise(n=n()) %>%\n",
    "        dplyr::filter(n>=n_slice_group) %>% \n",
    "        dplyr::pull(ident) \n",
    "\n",
    "    # Get cell type and top DEA genes\n",
    "    dea_genes <- dea %>%\n",
    "        dplyr::filter(ident %in% cell_type_group & sign(avg_log2FC)==log2fc_sign) %>% \n",
    "        dplyr::group_by(ident) %>% \n",
    "        dplyr::arrange(ident, desc(abs(avg_log2FC)), p_val_adj) %>% \n",
    "        slice(1:n_slice) %>%\n",
    "        dplyr::select(ident, gene, p_val_adj) \n",
    "\n",
    "    # Get log2FC matrix \n",
    "    mat_1 <- dea %>% \n",
    "        dplyr::filter(gene %in% dea_genes$gene) %>% \n",
    "        dplyr::select(avg_log2FC, ident, gene) %>% \n",
    "        reshape2::dcast(., gene ~ ident, value.var=\"avg_log2FC\") %>% \n",
    "        replace(is.na(.), 0)\n",
    "    \n",
    "    # Get padj matrix \n",
    "    mat_2 <- dea %>% \n",
    "        dplyr::filter(gene %in% dea_genes$gene) %>% \n",
    "        dplyr::select(p_val_adj, ident, gene) %>% \n",
    "        reshape2::dcast(., gene ~ ident, value.var=\"p_val_adj\") %>% \n",
    "        replace(is.na(.), 1)\n",
    "    \n",
    "    # Combine dea_genes with mat and preserve gene names\n",
    "    mat_1 <- left_join(dea_genes, mat_1, by=\"gene\") %>% \n",
    "        dplyr::mutate(ident=factor(ident, levels=cell_type_order)) %>% \n",
    "        dplyr::arrange(ident) %>% \n",
    "        as.data.frame() \n",
    "    \n",
    "    mat_2 <- left_join(dea_genes, mat_2, by=\"gene\") %>% \n",
    "        dplyr::mutate(ident=factor(ident, levels=cell_type_order)) %>% \n",
    "        dplyr::arrange(ident) %>% \n",
    "        as.data.frame()\n",
    "    \n",
    "    # Store column split and column name \n",
    "    row_split_1 <- mat_1$ident\n",
    "    column_names_1 <- mat_1$gene\n",
    "    \n",
    "    column_split_2 <- mat_2$ident\n",
    "    column_names_2 <- mat_2$gene\n",
    "    \n",
    "    # Remove cell type and gene names from matirx \n",
    "    mat_1 <- mat_1 %>% dplyr::select(-ident, -gene, -p_val_adj) %>% as.matrix \n",
    "    mat_2 <- mat_2 %>% dplyr::select(-ident, -gene, -p_val_adj) %>% as.matrix \n",
    "    \n",
    "    # Transform matrix and set column names\n",
    "    mat_1 <- t(mat_1)\n",
    "    mat_2 <- t(mat_2)\n",
    "    \n",
    "    colnames(mat_1) <- column_names_1\n",
    "    colnames(mat_2) <- column_names_2\n",
    "    \n",
    "    # Color function - expression z-score\n",
    "    if(log2fc_sign>0) {\n",
    "        \n",
    "        color_ramp_mat <- c(\"#ffffff\", max_color=color_max)\n",
    "        breaks_mat <- seq(0, 2, length.out=length(color_ramp_mat))\n",
    "    \n",
    "    } else {\n",
    "        \n",
    "        color_ramp_mat <- c(color_max, max_color=\"#ffffff\")\n",
    "        breaks_mat <- seq(-2,  0, length.out=length(color_ramp_mat))\n",
    "    \n",
    "    }\n",
    "\n",
    "    color_function_mat <- circlize::colorRamp2(breaks_mat, color_ramp_mat) \n",
    "    \n",
    "    # Top annotation - cell type\n",
    "    bottom_annotation <- HeatmapAnnotation(\n",
    "\n",
    "        df=data.frame(\n",
    "\n",
    "            top_annotation_dpi=cell_type_order\n",
    "\n",
    "\n",
    "        ), \n",
    "\n",
    "        annotation_legend_param=list(\n",
    "\n",
    "            top_annotation_dpi=list(title=\"Cell type\", title_gp=gpar(fontsize=18, fontface=\"plain\"), labels_gp=gpar(fontsize=16), grid_width=unit(6, \"mm\"), grid_height=unit(6, \"mm\"))\n",
    "\n",
    "        ), \n",
    "\n",
    "        col=list(\n",
    "\n",
    "            top_annotation_dpi=cell_type_color\n",
    "\n",
    "        ), \n",
    "\n",
    "        simple_anno_size=unit(6, \"mm\"), \n",
    "        show_annotation_name=FALSE, \n",
    "        show_legend=FALSE\n",
    "\n",
    "    )\n",
    "    \n",
    "    # Order columns \n",
    "    mat_1 <- mat_1[rev(cell_type_order), ]\n",
    "    mat_2 <- mat_2[rev(cell_type_order), ]\n",
    "    \n",
    "    mat_1 <- t(mat_1)\n",
    "    mat_2 <- t(mat_2)\n",
    "    \n",
    "    # Heat map \n",
    "    ComplexHeatmap::Heatmap(\n",
    "\n",
    "        matrix=mat_1, \n",
    "        # name=\"log2(FC)\",  \n",
    "\n",
    "        col=color_function_mat, \n",
    "        na_col=\"white\", \n",
    "\n",
    "        width=unit(5.5*ncol(mat_1), \"mm\"), \n",
    "        height=unit(5.5*nrow(mat_1), \"mm\"), \n",
    "\n",
    "        # column_title=NULL, \n",
    "        column_title_gp=gpar(fontsize=18, fontface=\"bold\"), \n",
    "\n",
    "        row_names_gp=gpar(fontsize=16, fontface=\"plain\"), \n",
    "        column_names_gp=gpar(fontsize=18, fontface=\"plain\"), \n",
    "\n",
    "        cluster_rows=FALSE, \n",
    "        show_row_dend=FALSE,    \n",
    "        row_split=row_split_1,\n",
    "        row_title_gp=gpar(fontsize=16, fontface=\"plain\"), \n",
    "        row_gap=unit(2, \"mm\"),\n",
    "        show_row_names=TRUE,\n",
    "        row_names_side=\"right\",\n",
    "\n",
    "        cluster_columns=FALSE,\n",
    "        cluster_column_slices=FALSE, \n",
    "        show_column_dend=FALSE, \n",
    "        column_split=NULL,\n",
    "        column_gap=unit(2, \"mm\"), \n",
    "        show_column_names=TRUE, \n",
    "        \n",
    "        bottom_annotation=bottom_annotation, \n",
    "        top_annotation=NULL, \n",
    "        right_annotation=NULL, \n",
    "\n",
    "        rect_gp=gpar(col=NA, lwd=0, alpha=1), \n",
    "\n",
    "        heatmap_legend_param=list(title=\"log2FC\", title_gp=gpar(fontsize=18, fontface=\"plain\"), labels_gp=gpar(fontsize=16), legend_height=unit(5*6, \"mm\"), grid_width=unit(6, \"mm\")), \n",
    "\n",
    "        border=TRUE, \n",
    "        border_gp=gpar(col=\"black\", lwd=unit(0.5, \"mm\")), \n",
    "\n",
    "        use_raster=FALSE, \n",
    "        \n",
    "        cell_fun = function(j, i, x, y, w, h, fill) {\n",
    "            \n",
    "        if(mat_2[i, j] < 0.1) {\n",
    "            \n",
    "            grid.text(\"*\", x, y)\n",
    "        \n",
    "            }\n",
    "        \n",
    "        }\n",
    "\n",
    "    ) \n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4462130-0bff-4627-8ef8-b59c2f590642",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_slice <- 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59ae44f7-5851-45ee-a89e-f4357a30ecf9",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=15, repr.plot.height=length(results)*n_slice/4.5)\n",
    "\n",
    "hm_1 <- hm_dea_genes(\n",
    "    \n",
    "    dea_file=paste0(basename, \"_dea.rds\"), \n",
    "    cell_type_order=names(color$cell_type_fine), \n",
    "    cell_type_color=unlist(color$cell_type_fine),\n",
    "    gene_names=NULL, \n",
    "    log2fc_sign=+1, padj_thr_group=1, n_slice_group=1, padj_thr=padj_thr, n_slice=n_slice, color_max=color$treatment[2]\n",
    "    \n",
    ")\n",
    "\n",
    "ggsave(paste0(basename, \"_dea_hm_pos.pdf\"), as.ggplot(hm_1), width=15, height=length(results)*n_slice/4.5, dpi=1200, limitsize=FALSE)\n",
    "hm_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef0349df-1230-413a-a667-341aa409ba4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=15, repr.plot.height=length(results)*n_slice/4.5)\n",
    "\n",
    "hm_1 <- hm_dea_genes(\n",
    "    \n",
    "    dea_file=paste0(basename, \"_dea.rds\"), \n",
    "    cell_type_order=names(color$cell_type_fine), \n",
    "    cell_type_color=unlist(color$cell_type_fine),\n",
    "    gene_names=NULL, \n",
    "    log2fc_sign=-1, padj_thr_group=1, n_slice_group=1, padj_thr=padj_thr, n_slice=n_slice, color_max=color$treatment[1]\n",
    "    \n",
    ")\n",
    "\n",
    "ggsave(paste0(basename, \"_dea_hm_neg.pdf\"), as.ggplot(hm_1), width=15, height=length(results)*n_slice/4.5, dpi=1200, limitsize=FALSE)\n",
    "hm_1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dce9a93d-2b23-4af0-aa55-5b52fe71e543",
   "metadata": {},
   "source": [
    "# Heatmap - TF "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72fb350d-f5bb-4367-9c65-8593c8611444",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "hm_dea_genes_bio_class <- function(dea_file, bio_class, column_title=NULL, cell_type_order, cell_type_color, padj_thr=0.1, n_slice=10, color_neg=\"blue\", color_pos=\"red\") {\n",
    "    \n",
    "    # DEA results \n",
    "    dea <- readRDS(dea_file)\n",
    "    dea <- do.call(rbind, dea)\n",
    "    dea <- dea %>% dplyr::filter(ident %in% cell_type_order) \n",
    "    \n",
    "    # Filter for biotype \n",
    "    dea <- dea[dea[[bio_class]], ]\n",
    "    \n",
    "    # Re-set cell types in case some are missing from DEA\n",
    "    cell_type_order <- cell_type_order[cell_type_order %in% dea$ident]\n",
    "    \n",
    "    # Get cell type and top DEA genes\n",
    "    dea_genes <- dea %>%\n",
    "        dplyr::filter(p_val_adj <= padj_thr) %>%\n",
    "        dplyr::select(ident, gene, p_val_adj)\n",
    "\n",
    "    # Get log2FC matrix \n",
    "    mat_1 <- dea %>% \n",
    "        dplyr::filter(gene %in% dea_genes$gene) %>% \n",
    "        dplyr::select(avg_log2FC, ident, gene) %>% \n",
    "        reshape2::dcast(., gene ~ ident, value.var=\"avg_log2FC\") %>% \n",
    "        replace(is.na(.), 0) %>% \n",
    "        tibble::column_to_rownames(\"gene\") %>% \n",
    "        unique()\n",
    "    \n",
    "    # Get padj matrix \n",
    "    mat_2 <- dea %>% \n",
    "        dplyr::filter(gene %in% dea_genes$gene) %>% \n",
    "        dplyr::select(p_val_adj, ident, gene) %>% \n",
    "        reshape2::dcast(., gene ~ ident, value.var=\"p_val_adj\") %>% \n",
    "        replace(is.na(.), 1) %>% \n",
    "        tibble::column_to_rownames(\"gene\") %>% \n",
    "        unique()\n",
    "    \n",
    "    # Color function - log2FC\n",
    "    color_ramp_mat <- c(color_neg, \"#ffffff\", color_pos)\n",
    "    breaks_mat <- seq(-2,  2, length.out=length(color_ramp_mat))\n",
    "\n",
    "    color_function_mat <- circlize::colorRamp2(breaks_mat, color_ramp_mat) \n",
    "    \n",
    "    # Top annotation - cell type\n",
    "    bottom_annotation <- HeatmapAnnotation(\n",
    "\n",
    "        df=data.frame(\n",
    "\n",
    "            top_annotation_dpi=cell_type_order\n",
    "\n",
    "\n",
    "        ), \n",
    "\n",
    "        annotation_legend_param=list(\n",
    "\n",
    "            top_annotation_dpi=list(title=\"Cell type\", title_gp=gpar(fontsize=18, fontface=\"plain\"), labels_gp=gpar(fontsize=16), grid_width=unit(6, \"mm\"), grid_height=unit(6, \"mm\"))\n",
    "\n",
    "        ), \n",
    "\n",
    "        col=list(\n",
    "\n",
    "            top_annotation_dpi=cell_type_color\n",
    "\n",
    "        ), \n",
    "\n",
    "        simple_anno_size=unit(6, \"mm\"), \n",
    "        show_annotation_name=FALSE, \n",
    "        show_legend=FALSE\n",
    "\n",
    "    )    \n",
    "    \n",
    "    # Heat map \n",
    "    ComplexHeatmap::Heatmap(\n",
    "\n",
    "        matrix=mat_1, \n",
    "        # name=\"log2(FC)\",  \n",
    "\n",
    "        col=color_function_mat, \n",
    "        na_col=\"white\", \n",
    "\n",
    "        width=unit(5.5*ncol(mat_1), \"mm\"), \n",
    "        height=unit(5.5*nrow(mat_1), \"mm\"), \n",
    "\n",
    "        column_title=column_title, \n",
    "        column_title_gp=gpar(fontsize=18, fontface=\"bold\"), \n",
    "\n",
    "        row_names_gp=gpar(fontsize=16, fontface=\"plain\"), \n",
    "        column_names_gp=gpar(fontsize=18, fontface=\"plain\"), \n",
    "\n",
    "        cluster_rows=TRUE, \n",
    "        show_row_dend=FALSE,    \n",
    "        # row_split=row_split_1,\n",
    "        row_title_gp=gpar(fontsize=16, fontface=\"plain\"), \n",
    "        row_gap=unit(2, \"mm\"),\n",
    "        show_row_names=TRUE,\n",
    "        row_names_side=\"right\",\n",
    "\n",
    "        cluster_columns=FALSE,\n",
    "        cluster_column_slices=FALSE, \n",
    "        show_column_dend=FALSE, \n",
    "        column_split=NULL,\n",
    "        column_gap=unit(2, \"mm\"), \n",
    "        show_column_names=TRUE, \n",
    "        \n",
    "        bottom_annotation=bottom_annotation, \n",
    "        top_annotation=NULL, \n",
    "        right_annotation=NULL, \n",
    "\n",
    "        rect_gp=gpar(col=NA, lwd=0, alpha=1), \n",
    "\n",
    "        heatmap_legend_param=list(title=\"log2FC\", title_gp=gpar(fontsize=18, fontface=\"plain\"), labels_gp=gpar(fontsize=16), legend_height=unit(5*6, \"mm\"), grid_width=unit(6, \"mm\")), \n",
    "\n",
    "        border=TRUE, \n",
    "        border_gp=gpar(col=\"black\", lwd=unit(0.5, \"mm\")), \n",
    "\n",
    "        use_raster=FALSE, \n",
    "        \n",
    "        cell_fun = function(j, i, x, y, w, h, fill) {\n",
    "            \n",
    "        if(mat_2[i, j] < 0.1) {\n",
    "            \n",
    "            grid.text(\"*\", x, y)\n",
    "        \n",
    "            }\n",
    "        \n",
    "        }\n",
    "\n",
    "    )\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e7bbf17-73b0-4aa7-8a7c-d1cbd4ace6dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_1 <- hm_dea_genes_bio_class(\n",
    "    \n",
    "    dea_file=paste0(basename, \"_dea.rds\"), \n",
    "    bio_class=\"transcription_factor\", \n",
    "    column_title=\"TF\", \n",
    "    cell_type_order=names(color$cell_type_fine), \n",
    "    cell_type_color=unlist(color$cell_type_fine),\n",
    "    padj_thr=p_adj_thr, color_neg=color$treatment[[1]], color_pos=color$treatment[[2]]\n",
    "    \n",
    ")\n",
    "\n",
    "hm_2 <- hm_dea_genes_bio_class(\n",
    "    \n",
    "    dea_file=paste0(basename, \"_dea.rds\"), \n",
    "    bio_class=\"ligand\", \n",
    "    column_title=\"Ligand\", \n",
    "    cell_type_order=names(color$cell_type_fine), \n",
    "    cell_type_color=unlist(color$cell_type_fine),\n",
    "    padj_thr=p_adj_thr, color_neg=color$treatment[[1]], color_pos=color$treatment[[2]]\n",
    "    \n",
    ")\n",
    "\n",
    "hm_3 <- hm_dea_genes_bio_class(\n",
    "    \n",
    "    dea_file=paste0(basename, \"_dea.rds\"), \n",
    "    bio_class=\"receptor\", \n",
    "    column_title=\"Receptor\", \n",
    "    cell_type_order=names(color$cell_type_fine), \n",
    "    cell_type_color=unlist(color$cell_type_fine),\n",
    "    padj_thr=p_adj_thr, color_neg=color$treatment[[1]], color_pos=color$treatment[[2]]\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93bd462b-08c0-4a11-9884-4478b217174e",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10, repr.plot.height=ceiling(1+dim(hm_1)[1]/4))\n",
    "\n",
    "ggsave(paste0(basename, \"_dea_hm_tf.pdf\"), as.ggplot(hm_1), width=15, height=ceiling(1+dim(hm_1)[1]/4), dpi=1200, limitsize=FALSE)\n",
    "hm_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb9be230-49f1-46cc-80d7-c6571875b4a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10, repr.plot.height=ceiling(1+dim(hm_2)[1]/4))\n",
    "\n",
    "ggsave(paste0(basename, \"_dea_hm_ligand.pdf\"), as.ggplot(hm_2), width=15, height=ceiling(1+dim(hm_2)[1]/4), dpi=1200, limitsize=FALSE)\n",
    "hm_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "939f1e6a-0649-414a-b907-5b4f21b89a1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10, repr.plot.height=ceiling(1+dim(hm_3)[1]/4))\n",
    "\n",
    "ggsave(paste0(basename, \"_dea_hm_receptor.pdf\"), as.ggplot(hm_3), width=15, height=ceiling(1+dim(hm_3)[1]/4), dpi=1200, limitsize=FALSE)\n",
    "hm_3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b09cb12-7626-444f-8970-0c2e2b5dd0b0",
   "metadata": {},
   "source": [
    "# Compare DEA between cell types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b051bd7e-dab3-4bf5-822e-bfb91ceb4ff8",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_dea_cor <- function(file, log2FC_sign, score_type, color_min, color_max, padj_thr_gs=0.1, n_thr_gs=5, cache=TRUE) {\n",
    "\n",
    "    # DEA Ranking and GSEA function \n",
    "    gsea_run <- function(dea, gene_set, score_type) {\n",
    "    \n",
    "        ranks <- set_names(-log10(dea$p_val_adj) * sign(dea$avg_log2FC), dea$gene)\n",
    "        ranks <- ranks[order(ranks, decreasing=TRUE)]\n",
    "\n",
    "        gsea <- fgsea(\n",
    "\n",
    "            pathways=gene_set,\n",
    "            stats=ranks,\n",
    "            nperm=100000, \n",
    "            minSize=0, \n",
    "            maxSize=Inf, \n",
    "            scoreType=score_type\n",
    "\n",
    "        )\n",
    "\n",
    "        return(gsea)\n",
    "    \n",
    "    }\n",
    "\n",
    "    # Load DEA results from file \n",
    "    dea <- readRDS(paste0(basename, \"_dea.rds\"))\n",
    "    dea_order <- names(dea)\n",
    "    dea <- do.call(\"rbind\", dea)\n",
    "\n",
    "    # Extract gene sets\n",
    "    gs_dea <- dplyr::group_by(dea, ident) %>% filter(p_val_adj <= padj_thr_gs & sign(avg_log2FC)==log2FC_sign) %>% dplyr::filter(n() > n_thr_gs)\n",
    "    gs_dea <- data.frame(gs_name=gs_dea$ident, gene_symbol=gs_dea$gene)\n",
    "    gs_dea <- split(x=gs_dea$gene_symbol, f=gs_dea$gs_name)\n",
    "\n",
    "    # Run GSEA \n",
    "    dea <- split(dea, f=dea$ident)\n",
    "\n",
    "    if(!cache) {\n",
    "        \n",
    "        gsea <- lapply(dea, function(x) gsea_run(x, gene_set=gs_dea, score_type=score_type))\n",
    "        saveRDS(gsea, paste0(basename, \"_gsea_dea_\", score_type, \".rds\"))\n",
    "        \n",
    "    } else {\n",
    "        \n",
    "        gsea <- readRDS(paste0(basename, \"_gsea_dea_\", score_type, \".rds\"))\n",
    "        \n",
    "    }\n",
    "      \n",
    "    # Get GSEA matrix with NES\n",
    "    mat <- lapply(gsea, function(x, y) {\n",
    "    \n",
    "        x$NES <- ifelse(x$padj <= 0.05, x$NES, 0)\n",
    "        x <- x[, c(\"pathway\", \"NES\")]\n",
    "\n",
    "        order <- y$pathway\n",
    "        y <- y[!y$pathway %in% x$pathway, , drop=FALSE]\n",
    "\n",
    "        if(nrow(y) > 0) {x <- rbind(x, y)} else {x <- x}\n",
    "\n",
    "        x <- tibble::column_to_rownames(x, \"pathway\") %>% as.data.frame()\n",
    "        x <- x[order, ]\n",
    "\n",
    "        return(x)\n",
    "\n",
    "    }, y=data.frame(pathway=names(gs_dea), NES=NA)\n",
    "\n",
    "                 )\n",
    "                   \n",
    "    mat <- do.call(\"cbind\", mat)\n",
    "    colnames(mat) <- names(gsea)\n",
    "    rownames(mat) <- names(gs_dea)\n",
    "    mat[is.na(mat)] <- 0\n",
    "    \n",
    "    # Set geneset to columns \n",
    "    mat <- t(mat)\n",
    "    dea_order <- dea_order[dea_order %in% colnames(mat)]\n",
    "    mat <- mat[dea_order, dea_order]\n",
    "                   \n",
    "    # Get absolut NES values \n",
    "    mat <- abs(mat)\n",
    "\n",
    "    # Color function \n",
    "    color_ramp_mat <- c(color_min, color_max)\n",
    "    breaks_mat <- seq(min(mat, na.rm=TRUE), quantile(mat, 0.99, na.rm=TRUE), length.out=2)\n",
    "    color_function_mat <- circlize::colorRamp2(breaks_mat, color_ramp_mat) \n",
    "                   \n",
    "    hm <- ComplexHeatmap::Heatmap(\n",
    "    \n",
    "        matrix=mat, \n",
    "\n",
    "        col=color_function_mat, \n",
    "        na_col=\"white\", \n",
    "\n",
    "        width=unit(5.5*ncol(mat), \"mm\"), \n",
    "        height=unit(5.5*nrow(mat), \"mm\"), \n",
    "\n",
    "        row_title=\"DEA\", \n",
    "        row_title_gp=gpar(fontsize=18, fontface=\"bold\"), \n",
    "\n",
    "        column_title=\"DEA gene module\", \n",
    "        column_title_gp=gpar(fontsize=18, fontface=\"bold\"), \n",
    "\n",
    "        row_names_gp=gpar(fontsize=16, fontface=\"plain\"), \n",
    "        column_names_gp=gpar(fontsize=18, fontface=\"plain\"), \n",
    "\n",
    "        # cluster_rows=function(mat) hclust(dist(mat), method=\"complete\"), \n",
    "        cluster_rows=FALSE, \n",
    "        show_row_dend=TRUE,    \n",
    "        row_gap=unit(2, \"mm\"),\n",
    "        show_row_names=TRUE,\n",
    "        row_names_side=\"right\",\n",
    "\n",
    "        cluster_columns=FALSE,\n",
    "        cluster_column_slices=FALSE, \n",
    "        show_column_dend=TRUE, \n",
    "        column_split=NULL,\n",
    "        column_gap=unit(2, \"mm\"), \n",
    "        show_column_names=TRUE, \n",
    "\n",
    "        top_annotation=NULL, \n",
    "        right_annotation=NULL, \n",
    "\n",
    "        rect_gp=gpar(col=NA, lwd=0, alpha=1), \n",
    "\n",
    "        heatmap_legend_param=list(title=\"NES\", title_gp=gpar(fontsize=18, fontface=\"plain\"), labels_gp=gpar(fontsize=16), legend_height=unit(5*6, \"mm\"), grid_width=unit(6, \"mm\")), \n",
    "\n",
    "        border=TRUE, \n",
    "        border_gp=gpar(col=\"black\", lwd=unit(0.5, \"mm\")), \n",
    "\n",
    "        use_raster=FALSE\n",
    "\n",
    "    )\n",
    "                   \n",
    "    return(hm)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bbb7d19-5e87-45dc-b080-72e92a72097f",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_1 <- hm_dea_cor(\n",
    "    \n",
    "    file=basename, \n",
    "    log2FC_sign=+1,\n",
    "    score_type=\"pos\",\n",
    "    color_min=\"#ffffff\",\n",
    "    color_max=color$treatment[2], \n",
    "    cache=cache\n",
    "\n",
    ")\n",
    "\n",
    "hm_2 <- hm_dea_cor(\n",
    "    \n",
    "    file=basename, \n",
    "    log2FC_sign=-1,\n",
    "    score_type=\"neg\",\n",
    "    color_min=\"#ffffff\",\n",
    "    color_max=color$treatment[1], \n",
    "    cache=cache\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b34faff-92fa-4a25-b4a8-bfc2e48ad276",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=7.5, repr.plot.height=ceiling(2+dim(hm_1)[1]/4))\n",
    "\n",
    "ggsave(paste0(basename, \"_dea_gsea_hm_pos.pdf\"), as.ggplot(hm_1), width=15, height=ceiling(2+dim(hm_1)[1]/4), dpi=1200, limitsize=FALSE)\n",
    "hm_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8130da08-73b2-41b1-a8a1-dd723ae3257a",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=7.5, repr.plot.height=ceiling(2+dim(hm_2)[1]/4))\n",
    "\n",
    "ggsave(paste0(basename, \"_dea_gsea_hm_neg.pdf\"), as.ggplot(hm_2), width=15, height=ceiling(2+dim(hm_2)[1]/4), dpi=1200, limitsize=FALSE)\n",
    "hm_2"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "r.4.1.0-FD20200109SPLENO",
   "language": "R",
   "name": "r.4.1.0-fd20200109spleno"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
