{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2bb26613-8e42-44a6-b7ac-d4c2ba9fac1d",
   "metadata": {},
   "source": [
    "# RNA-velocity MURK gene identification \n",
    "\n",
    "[Barile et al. 2021](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02414-y) showed that the transcription kinetics along erythroid development are not constant and hence violating the assumption of scVelo. We too expect multiple rate kinetics (MURK) genes between treatment conditions and adapt the [procedure](https://github.com/mebarile/Gata1_Erythroid_kinetics) to identify them. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "6bfc7854-eac5-4ee9-a9cb-571b7d24ebd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scvelo as scv\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy\n",
    "import scanpy as sc\n",
    "\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "id": "75495ed6-b12a-4752-b35a-4885b72512fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.settings.vector_friendly = False\n",
    "scv.set_figure_params( dpi=300, dpi_save = 300, frameon=False, figsize = (7,4), format='png',fontsize=25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "id": "0fbe1d69-43c1-4022-a8a2-3c2a3e8f53a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/research/peer/fdeckert/FD20200109SPLENO')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f7f1d73-ee70-496f-997b-80246a882d92",
   "metadata": {},
   "source": [
    "# Import data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "id": "26ace9af-5ffb-488e-b8bb-d683818557d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad('data/object/velocyto.h5ad')\n",
    "obs = pd.read_csv('data/object/int/meta/meta.csv', index_col=0)\n",
    "obsm = pd.read_csv('data/object/int/reductions/X_umap/reduction.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c40765d-687b-4cc7-aadb-e8a2f677cbf2",
   "metadata": {},
   "source": [
    "# Filter velocity matrix by cell types from meta "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "85b60c14-806d-4282-9a6f-159119ab6fc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter obs by Ery annotation \n",
    "population_names = ['Ery (1)', 'Ery (2)', 'Ery (3)', 'Ery (4)', 'Ery (5)', 'Ery (6)']\n",
    "obs = obs[obs['leiden_annotation'].isin(population_names)]\n",
    "\n",
    "# Filter obsm by cell index\n",
    "obsm = obsm[obsm.index.isin(obs.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "id": "a3fa8f2d-678b-4c94-bfc7-ca757a481f6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter velocity adata by obs \n",
    "adata = adata[adata.obs.index.isin(obs.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "id": "7072f342-f159-4d55-a2b9-06ad78f636ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Order index to match velocity adata \n",
    "obs = obs.reindex(adata.obs.index)\n",
    "obsm = obsm.reindex(adata.obs.index)\n",
    "\n",
    "adata.obs = obs\n",
    "adata.obsm['X_umap'] = obsm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e475beab-143d-425f-b75f-a4ee405e6358",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_temp = adata.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdb8b89d-0883-43fc-a419-474511f03220",
   "metadata": {},
   "source": [
    "# Highly variable genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ca8faf6-1ede-43c0-a40b-c5559abd6107",
   "metadata": {},
   "outputs": [],
   "source": [
    "def hvg_select(subset, adata):\n",
    "    \n",
    "    # HVG on all data \n",
    "    adata = adata[adata.obs['treatment']==subset]\n",
    "    sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "    sc.pp.log1p(adata)\n",
    "    sc.pp.highly_variable_genes(adata, min_mean=0.05, min_disp=0.1)\n",
    "    hvg_1 = adata.var_names[adata.var.highly_variable].tolist()\n",
    "\n",
    "    # HVG on scvelo filtered data\n",
    "    adata = adata_temp.copy()\n",
    "    adata = adata[adata.obs['treatment']==subset]\n",
    "    scv.pp.filter_and_normalize(adata, min_shared_counts=20, n_top_genes=2000)\n",
    "\n",
    "    hvg_2 = adata.var_names.tolist()\n",
    "\n",
    "    # Combine HVG\n",
    "    hvg = list(set(hvg_1) | set(hvg_2))\n",
    "    \n",
    "    # Reset adata\n",
    "    adata = adata_temp.copy()\n",
    "    \n",
    "    return(hvg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c11c3a04-a7dc-48cc-9455-d4a6fb8f9607",
   "metadata": {},
   "outputs": [],
   "source": [
    "hvg_nacl = hvg_select(\"NaCl\", adata)\n",
    "hvg_cpg = hvg_select(\"CpG\", adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "141f9bb9-8af6-478b-97b6-a64f3daf6420",
   "metadata": {},
   "outputs": [],
   "source": [
    "hvg = list(set(hvg_nacl) & set(hvg_cpg))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70754a15-5416-456f-b15d-c93439001e85",
   "metadata": {},
   "source": [
    "# Moments imputation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdb8cc88-8ce2-4e17-bba0-9a634bcfc06d",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata_temp.copy()\n",
    "adata = adata[:,hvg]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72ac733b-98e5-44ed-8f65-fda401151bdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "scv.pp.filter_and_normalize(adata)\n",
    "scv.pp.moments(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "341e4509-05e0-43d6-9b47-d1612ee11712",
   "metadata": {},
   "outputs": [],
   "source": [
    "# initialize matrix for spliced (mat_s) and unspliced (unspliced) counts\n",
    "mat_s = pd.DataFrame(adata.layers['Ms'].copy(), index=adata.obs_names, columns=adata.var_names)\n",
    "mat_u = pd.DataFrame(adata.layers['Mu'].copy(), index=adata.obs_names, columns=adata.var_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19c99904-2c5a-440a-b4f7-791fc01d0dfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# initialize dataframes for slopes (df_sl) and slope error (df_ds)\n",
    "df_sl = pd.DataFrame(index=adata.var_names)\n",
    "df_ds = pd.DataFrame(index=adata.var_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "491dd1bd-85ba-4105-ba04-a01376a83292",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute slopes and errors\n",
    "for pop in population_names:\n",
    "\n",
    "    vec = adata.obs['leiden_annotation'] == pop\n",
    "    \n",
    "    x = mat_s.loc[vec]\n",
    "    y = mat_u.loc[vec]\n",
    "    \n",
    "    mx = np.mean(x)\n",
    "    my = np.mean(y)\n",
    "    x_mx = x - mx\n",
    "    y_my = y - my\n",
    "    xy = np.sum(x_mx * y_my, axis=0)\n",
    "    X2 = np.sum(x_mx**2, axis=0)\n",
    "    sl = xy / X2\n",
    "    inter = my - sl * mx\n",
    "    n = np.sum(vec)\n",
    "                \n",
    "    df_sl[pop] = pd.DataFrame(sl)     \n",
    "    \n",
    "    pred = sl * x + inter\n",
    "    sse = (pred - y)**2\n",
    "    SSE = np.sum(sse, axis=0)\n",
    "                \n",
    "    ds = np.sqrt(SSE/(n-2)/X2)\n",
    "    df_ds[pop] = pd.DataFrame(ds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14ae8bad-0178-4900-b1be-63e32149401a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# initialize dataframes for slopes (df_exp), expression error (df_de) and 95% quantile (df_quant)\n",
    "df_exp = pd.DataFrame(index=adata.var_names)\n",
    "df_de = pd.DataFrame(index=adata.var_names)\n",
    "df_quan = pd.DataFrame(index=adata.var_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "571d29c9-666e-4970-a4c8-7ab8ef0df1e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute average expression and its error\n",
    "for pop in population_names:\n",
    "\n",
    "    vec = adata.obs['leiden_annotation'] == pop\n",
    "    n = np.sum(vec)\n",
    "    expr = np.mean(mat_s.loc[vec], axis = 0)\n",
    "    dex = np.std(mat_s.loc[vec], axis = 0)/np.sqrt(n)\n",
    "    \n",
    "    df_exp[pop] = pd.DataFrame(expr)\n",
    "    df_de[pop] = pd.DataFrame(dex)\n",
    "\n",
    "for pop in population_names:\n",
    "    vec = adata.obs['leiden_annotation'] == pop\n",
    "    n = np.sum(vec)\n",
    "    \n",
    "    df_quan[pop] = scipy.stats.t.ppf(0.95, n - 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e7d1402-d44a-4abd-abe8-0cc0e6fe133d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute minimum and maximum estimates of the slopes\n",
    "df_conf = df_ds * df_quan\n",
    "\n",
    "df_min = df_sl - df_conf\n",
    "df_max = df_sl + df_conf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d46e419-ffd9-4fe3-87ee-467ccbcdfdf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "population_names = ['Ery (1)', 'Ery (2)', 'Ery (3)', 'Ery (4)', 'Ery (5)', 'Ery (6)']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "091b9a2c-d30b-437b-bd0b-57dc7f4bdbb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# decide if a change is slope is significative (Erythroid 3 expressed more than the others, its minimum slope grater than the previous maximum slope and positive)\n",
    "vec_mean = (((df_exp['Ery (6)'] >  df_exp['Ery (5)'])\n",
    "            &(df_exp['Ery (6)'] >  df_exp['Ery (4)']))\n",
    "            &(df_exp['Ery (6)'] >  df_exp['Ery (3)'])\n",
    "            &(df_exp['Ery (6)'] >  df_exp['Ery (2)'])\n",
    "            &(df_exp['Ery (6)'] >  df_exp['Ery (1)']))\n",
    "\n",
    "vec_sl = (df_sl['Ery (6)'] > df_sl['Ery (5)']) \n",
    "\n",
    "vec_pos =  df_sl['Ery (6)']>0\n",
    "\n",
    "vec_test = df_min['Ery (6)'] > df_max['Ery (5)']\n",
    "\n",
    "vec_tot = ((np.array(vec_sl) & np.array(vec_mean) & np.array(vec_pos) & np.array(vec_test)) | \n",
    " (vec_mean & np.array(df_sl['Ery (6)']<0)) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29fb3d62-e6cd-475a-9be7-00b00c332b71",
   "metadata": {},
   "outputs": [],
   "source": [
    "# How many MURK genes\n",
    "np.sum(vec_tot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d329e3f-495e-4ac3-85ee-0e232b31a093",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(murk_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e2c7c27-7506-4f97-acce-f377506ea8f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MURK genes for GO analysis Figure 3D\n",
    "murk_genes = adata.var_names[vec_tot]\n",
    "pd.DataFrame(murk_genes).to_csv('result/scvelo/resultmurk_genes_mouse.csv', index=None,header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a866487-ba15-425b-81c5-bf540fc9bce5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# recalculate the slopes scaling for the avarage of gene expression in order to range the genes for \"MURKiness\"\n",
    "\n",
    "df_sl2 = pd.DataFrame(index=adata.var_names)\n",
    "\n",
    "for pop in population_names:\n",
    "    print(pop)\n",
    "    vec = adata.obs['leiden_annotation'] == pop\n",
    "    \n",
    "    x = mat_s.loc[vec]/np.max(mat_s.loc[vec])\n",
    "    y = mat_u.loc[vec]/np.max(mat_u.loc[vec])\n",
    "    \n",
    "    mx = np.mean(x)\n",
    "    my = np.mean(y)\n",
    "    x_mx = x - mx\n",
    "    y_my = y - my\n",
    "    xy = np.abs(np.sum(x_mx * y_my, axis = 0))\n",
    "    X2 = np.sum(x_mx**2,axis = 0)\n",
    "    sl = xy / X2\n",
    "\n",
    "    df_sl2[pop] = pd.DataFrame(sl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f757800a-b221-4805-9f25-14aebc9ed038",
   "metadata": {},
   "outputs": [],
   "source": [
    "test = pd.read_csv('murk_genes_mouse.csv', header=0, index_col=0).index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b77f2b61-6046-4f5f-8813-a8bc3c45c504",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f7c3e5d-36dc-44b9-9fbc-3985aa9999a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(murk_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c55de7c2-6f19-4e9b-a6bb-e5c6f2c88bc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "murk_genes[murk_genes.isin(test)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0f79ae1-a0bb-478a-822d-0ff0dfc635af",
   "metadata": {},
   "outputs": [],
   "source": [
    "murk_genes"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "p.3.6.13-FD20200109SPLENO",
   "language": "python",
   "name": "p.3.6.13-fd20200109spleno"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
