{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cd04ed1a-d1ae-441f-9d5e-88c0aef441ca",
   "metadata": {},
   "source": [
    "# QC and filter seurat object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e9ed315-3005-46a2-b6bc-bed3a2094875",
   "metadata": {},
   "outputs": [],
   "source": [
    "library_load <- suppressMessages(\n",
    "    \n",
    "    list(\n",
    "        \n",
    "        # Seurat \n",
    "        library(Seurat), \n",
    "        \n",
    "        # Scran doubletFinder implementation \n",
    "        library(scDblFinder), \n",
    "        \n",
    "        # FastMNN implementation \n",
    "        library(SeuratWrappers), \n",
    "        \n",
    "        # Data \n",
    "        library(tidyverse), \n",
    "        \n",
    "        # Reticulate\n",
    "        library(reticulate), \n",
    "        \n",
    "        # Biomart \n",
    "        library(biomaRt), \n",
    "        \n",
    "        # Plotting \n",
    "        library(ggplot2), \n",
    "        library(knitr),\n",
    "        library(kableExtra),\n",
    "        library(IRdisplay)\n",
    "        \n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b923e9df-dfa7-4bc7-bb59-ae87cecc7faa",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_seed <- 42\n",
    "set.seed(random_seed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6ab1590-bda7-46f7-b50a-5fdde52a7846",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac325532-cdf7-4143-8775-463c6852b68e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Set working directory to project root\n",
    "setwd(\"/research/peer/fdeckert/FD20200109SPLENO\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44a53f47-22fb-424c-868c-8bc2fe748580",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Source files\n",
    "source(\"plotting_global.R\")\n",
    "source(\"bin/SeuratQC.R\")\n",
    "\n",
    "# SingleRPlot\n",
    "source(\"bin/SingleRQC.R\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dfe8c56-8c9d-4dc0-a5d4-d6d075b84a27",
   "metadata": {},
   "source": [
    "# Parameter settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1e02234-e351-4946-b71d-40c09cceb047",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Files \n",
    "so_raw_file <- \"data/object/raw.rds\"\n",
    "\n",
    "so_qc_rds <- \"data/object/qc.rds\"\n",
    "so_qc_h5ad <- \"data/object/qc.h5ad\"\n",
    "\n",
    "# Plotting Theme\n",
    "ggplot2::theme_set(theme_global_set()) # From project global source()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "354a4c0c-2393-4815-a7a8-dd8beb9e3127",
   "metadata": {},
   "source": [
    "# Import Seurat object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b406cedc-cfeb-4376-a330-9305d17c3e62",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_raw <- readRDS(so_raw_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b069077-d33d-458c-8ecd-6bcdad56379a",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_raw$tissue <- factor(so_raw$tissue, levels=c(\"Myeloid\", \"Progenitor\"))\n",
    "so_raw$treatment <- factor(so_raw$treatment, levels=c(\"NaCl\", \"CpG\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72d58dc5-d02c-44be-91b1-10b983728b17",
   "metadata": {},
   "source": [
    "# Rank plot\n",
    "\n",
    "Empty droplets were determined with CellRanger V3.0.2 Lun et al., 2019 EmptyDrop heuristic. RNAse activity of granulocytes might be wrongly identified as empty cells by CellRanger.\n",
    "\n",
    "**Typical Sample** A steep drop-off is indicative of good separation between the cell-associated barcodes and the barcodes associated with empty GEMs. A ideal barcode rank plot has a distincitve shape, which is referred to as a \"cliff and knee\".  \n",
    "**Heterogeneous Sample** Heterogeneous populations of cells in a sample result in two \"cliff and knee\" distributions. However, there should still be clear separation between the bacodes.   \n",
    "**Compromised Sample** Round curve and lack of steep cliff may indicate low sample quality or loss of single-cell behavior. This can be due to a wetting failure, premature cell lysis, or low cell viability.   \n",
    "**Compromised Sample** Defined cliff and knee, but the total number of barcodes detected may be lower than expected. This can be caused by a sample clog or inaccurate cell count.   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48f4c35b-90b9-4a08-b8d7-99c86c7a8ddf",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10, repr.plot.height=5)\n",
    "rank_plot_qc_1 <- rank_plot_qc(so_raw, color_color=color$cellranger_class, formular=tissue~sample_rep+treatment)\n",
    "rank_plot_qc_1\n",
    "ggsave(rank_plot_qc_1, filename=\"result/qc/plot/rank_plot_qc.png\", width=10, height=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "295859a3-b6c3-44de-88da-311555aff740",
   "metadata": {},
   "source": [
    "# Filter by CellRanger class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04d17536-94d8-4229-bdab-36ea8bfc78b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_qc <- subset(so_raw, subset=cellranger_class == \"Cell\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df2b16b5-e83c-41ce-b42e-c1ebb95d2a76",
   "metadata": {},
   "source": [
    "# Compute cell cycle score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ce0c5d0-8ba7-49a2-a3ad-fdca0666b0ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get Seurat cell cycle genes \n",
    "cc_genes_seurat_s <- cc.genes.updated.2019$s.genes\n",
    "cc_genes_seurat_g2m <- cc.genes.updated.2019$g2m.genes\n",
    "\n",
    "# Get mouse orthologs from human gene simbols\n",
    "httr::set_config(httr::config(ssl_verifypeer=FALSE))\n",
    "\n",
    "hgnc_mart <- useMart(\"ensembl\", dataset=\"hsapiens_gene_ensembl\", host=\"https://dec2021.archive.ensembl.org/\")\n",
    "mm_mart <- useMart(\"ensembl\", dataset=\"mmusculus_gene_ensembl\", host=\"https://dec2021.archive.ensembl.org/\")\n",
    "\n",
    "cc_genes_seurat_s <- getLDS(attributes=c(\"hgnc_symbol\"), filters=\"hgnc_symbol\", values=cc_genes_seurat_s, mart=hgnc_mart, attributesL=c(\"mgi_symbol\"), martL=mm_mart, uniqueRows=TRUE)\n",
    "cc_genes_seurat_s <- cc_genes_seurat_s[, 2]\n",
    "    \n",
    "cc_genes_seurat_g2m <- getLDS(attributes=c(\"hgnc_symbol\"), filters=\"hgnc_symbol\", values=cc_genes_seurat_g2m, mart=hgnc_mart, attributesL=c(\"mgi_symbol\"), martL=mm_mart, uniqueRows=TRUE)\n",
    "cc_genes_seurat_g2m <- cc_genes_seurat_g2m[, 2]\n",
    "\n",
    "# Compute cell cycle score \n",
    "so_qc <- CellCycleScoring(so_qc, s.features=cc_genes_seurat_s, g2m.features=cc_genes_seurat_g2m, set.ident=FALSE)\n",
    "\n",
    "colnames(so_qc@meta.data) <- gsub(\"Phase\", \"cc_phase_class\", colnames(so_qc@meta.data))\n",
    "colnames(so_qc@meta.data) <- gsub(\"S.Score\", \"msS_RNA\", colnames(so_qc@meta.data))\n",
    "colnames(so_qc@meta.data) <- gsub(\"G2M.Score\", \"msG2M_RNA\", colnames(so_qc@meta.data)) \n",
    "    \n",
    "so_qc$msCC_diff_RNA <- so_qc$msS_RNA - so_qc$msG2M_RNA\n",
    "so_qc$cc_phase_class <- factor(so_qc$cc_phase_class, levels=names(color$cc_phase_class))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74c15b7e-0341-4220-926e-81dc97b70671",
   "metadata": {},
   "source": [
    "# SingleR annotation\n",
    "SingleR identifies marker genes from the reference and uses them to compute assignment score (based on the Spearman correlation across markers) for each cell in the test dataset against each label in the reference. The label with the highest doublet_stat is the assigned to the test cell, possibly with further fine-tuning to resolve closely related labels.  \n",
    "\n",
    "first.labels: Labels before fine-tuning  \n",
    "labels: Labels after fine-tuning  \n",
    "pruning: labels after pruning  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "209ba001-e97c-480a-b99f-69fef37e3e06",
   "metadata": {},
   "source": [
    "## ImmGen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6b80c48-2f3b-4e3d-aed7-70b1dda6f2b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(FALSE) {\n",
    "    \n",
    "    # Load reference data \n",
    "    ref <- ImmGenData()\n",
    "    \n",
    "    # Seurat object to SingleCellExperiment\n",
    "    sce <- SingleCellExperiment(list(counts=so_qc@assays$RNA@counts))\n",
    "\n",
    "    # # Predict labels\n",
    "    label_main <- SingleR::SingleR(test=sce, ref=ref, labels=ref$label.main, assay.type.test=\"counts\", de.method=\"classic\")\n",
    "    label_fine <- SingleR::SingleR(test=sce, ref=ref, labels=ref$label.fine, assay.type.test=\"counts\", de.method=\"classic\")\n",
    "\n",
    "    saveRDS(label_main, \"result/SingleR/label_main_immgen.rds\")\n",
    "    saveRDS(label_fine, \"result/SingleR/label_fine_immgen.rds\")\n",
    "\n",
    "} else {\n",
    "    \n",
    "    label_main <- readRDS(\"result/SingleR/label_main_immgen.rds\")\n",
    "    label_fine <- readRDS(\"result/SingleR/label_fine_immgen.rds\")\n",
    "    \n",
    "}\n",
    "\n",
    "# Add labels to Seurat object\n",
    "label_main_meta <- as.data.frame(label_main) %>% dplyr::select(pruned.labels, tuning.scores.first) %>% dplyr::rename(label_main_immgen=pruned.labels, delta_score_main_immgen=tuning.scores.first)\n",
    "label_fine_meta <- as.data.frame(label_fine) %>% dplyr::select(pruned.labels, tuning.scores.first) %>% dplyr::rename(label_fine_immgen=pruned.labels, delta_score_fine_immgen=tuning.scores.first)\n",
    "\n",
    "so_qc <- AddMetaData(so_qc, cbind(label_main_meta, label_fine_meta))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05f262fa-e98d-4890-a042-19a3b3bdcc82",
   "metadata": {},
   "source": [
    "## Haemosphere"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16e3bd14-4c8f-4045-bf8e-3e3a876e473b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "if(FALSE) {\n",
    "    \n",
    "    # Load reference data \n",
    "    ref <- readRDS(\"data/haemosphere/se_haemosphere.rds\")\n",
    "    \n",
    "    # Seurat object to SingleCellExperiment\n",
    "    sce <- SingleCellExperiment(list(counts=so_qc@assays$RNA@counts))\n",
    "\n",
    "    # # Predict labels\n",
    "    label_main <- SingleR::SingleR(test=sce, ref=ref, labels=ref$label.main, assay.type.test=\"counts\", de.method=\"classic\")\n",
    "    label_fine <- SingleR::SingleR(test=sce, ref=ref, labels=ref$label.fine, assay.type.test=\"counts\", de.method=\"classic\")\n",
    "\n",
    "    saveRDS(label_main, \"result/SingleR/label_main_haemosphere.rds\")\n",
    "    saveRDS(label_fine, \"result/SingleR/label_fine_haemosphere.rds\")\n",
    "\n",
    "} else {\n",
    "    \n",
    "    label_main <- readRDS(\"result/SingleR/label_main_haemosphere.rds\")\n",
    "    label_fine <- readRDS(\"result/SingleR/label_fine_haemosphere.rds\")\n",
    "    \n",
    "}\n",
    "\n",
    "# Add labels to Seurat object\n",
    "label_main_meta <- as.data.frame(label_main) %>% dplyr::select(pruned.labels, tuning.scores.first) %>% dplyr::rename(label_main_haemosphere=pruned.labels, delta_score_main_haemosphere=tuning.scores.first)\n",
    "label_fine_meta <- as.data.frame(label_fine) %>% dplyr::select(pruned.labels, tuning.scores.first) %>% dplyr::rename(label_fine_haemosphere=pruned.labels, delta_score_fine_haemosphere=tuning.scores.first)\n",
    "\n",
    "so_qc <- AddMetaData(so_qc, cbind(label_main_meta, label_fine_meta))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f0d4f96-bb55-40f8-ad3c-e4ecc362aee6",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10, repr.plot.height=10)\n",
    "ggplot(so_qc@meta.data, aes(y=label_fine_haemosphere, fill=label_fine_haemosphere)) + \n",
    "    geom_bar() + \n",
    "    xlab(\"log10(cell count)\") + ylab(\"\") + ggtitle(\"Fine Haemosphere label\") +\n",
    "    scale_x_log10(breaks=c(0, 10, 100, 1000)) +\n",
    "    facet_grid(tissue~treatment+sample_rep) + \n",
    "    scale_fill_manual(values=color$label_fine_haemosphere)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f83eb77-6035-4ea0-b964-3c09fde197ad",
   "metadata": {},
   "source": [
    "# Set Seurat object QC class\n",
    "\n",
    "We recognize a substantial number or cells annotated as erythrocytes in the Myeloid sort. Since we do not expect those cells based on the FACS panel they are probably doublet cells or phagocytes which took up a erythrocyte. As shown in the scatter plots below those cells show either a very high or low feature count respectively. While it might be of interest keeping the phagocytes the low number of features detected – probably to the exaggerated  Hb mRNA count – makes downstream analysis difficult. We therefore exclude cells from the Myeloid sort which were annotated as Ery. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc2f6e07-3648-40bc-a3af-ccdc6c71903e",
   "metadata": {},
   "outputs": [],
   "source": [
    "qc_class_set <- function(so) {\n",
    "    \n",
    "    # QC filter for all cells\n",
    "    so$pMt_RNA_min <- -Inf\n",
    "    so$pMt_RNA_max <- 5\n",
    "    \n",
    "    so$nCount_RNA_min <- 1500\n",
    "    so$nCount_RNA_max <- max(so$nCount_RNA)\n",
    "\n",
    "    so$qc_class <- ifelse(\n",
    "        so$cellranger_class == \"Cell\" & \n",
    "        so$nCount_RNA <= so$nCount_RNA_max & \n",
    "        so$nCount_RNA > so$nCount_RNA_min &\n",
    "        so$pMt_RNA <= so$pMt_RNA_max &\n",
    "        so$pMt_RNA > so$pMt_RNA_min,\n",
    "        \"pass\", \"fail\"\n",
    "      )\n",
    "    \n",
    "    # Remove Ery from Myeloid sort before computing nFeature_RNA\n",
    "    if(so$tissue[1]==\"Myeloid\") {\n",
    "        \n",
    "        so$qc_class <- ifelse(\n",
    "            so$label_main_haemosphere==\"Ery\" | so$label_fine_haemosphere %in% c(\"preCFUE\", \"CFUE\", \"pbEry\", \"poEry\", \"Retic\") | so$pHb_RNA >= 5, \n",
    "            \"fail\", so$qc_class\n",
    "        )\n",
    "        \n",
    "    }\n",
    "    \n",
    "    so_tmp <- subset(so, subset=qc_class==\"pass\")\n",
    "    \n",
    "    so$nFeature_RNA_max <- quantile(so_tmp$nFeature_RNA, 0.99)\n",
    "    so$nFeature_RNA_min <- quantile(so_tmp$nFeature_RNA, 0.01)\n",
    "    \n",
    "    so$qc_class <- ifelse(\n",
    "        so$qc_class == \"pass\" &\n",
    "        so$nFeature_RNA <= so$nFeature_RNA_max & \n",
    "        so$nFeature_RNA > so$nFeature_RNA_min, \n",
    "        \"pass\", \"fail\"\n",
    "      )\n",
    "    \n",
    "    return(so)\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea49e921-9d90-4e2a-bc60-c1b06818cdc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_qc <- Seurat::SplitObject(so_qc, split.by=\"sample_name\")\n",
    "so_qc <- lapply(so_qc, qc_class_set)\n",
    "so_qc <- merge(so_qc[[1]], so_qc[2:length(so_qc)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf88f35f-3f76-4181-aa3b-0daaf470c32c",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_qc$tissue <- factor(so_qc$tissue, levels=c(\"Myeloid\", \"Progenitor\"))\n",
    "so_qc$treatment <- factor(so_qc$treatment, levels=c(\"NaCl\", \"CpG\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "207a9009-9445-44c8-9908-7e1b60e63251",
   "metadata": {},
   "source": [
    "# QC plots "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb31706c-1b1b-4f26-a593-d8907bd38532",
   "metadata": {},
   "source": [
    "## Density plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fa41dec-0eea-44f5-8377-15f8e551705f",
   "metadata": {},
   "outputs": [],
   "source": [
    "density_plot_qc_1 <- density_plot_qc(so=so_qc, title=\"Density plot UMI count\", x=nCount_RNA, xlab=\"log10(UMI count)\", min=nCount_RNA_min, max=nCount_RNA_max, fill_color=color$tissue)\n",
    "density_plot_qc_2 <- density_plot_qc(so=so_qc, title=\"Density plot Feature count\", x=nFeature_RNA, xlab=\"log10(Feature count)\", min=nFeature_RNA_min, max=nFeature_RNA_max, fill_color=color$tissue)\n",
    "density_plot_qc_3 <- density_plot_qc(so=so_qc, title=\"Density plot Mt %\", x=pMt_RNA, xlab=\"Mt [%]\", min=0, max=pMt_RNA_max, log10=FALSE, fill_color=color$tissue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e516954c-2bea-40c3-be9b-df383277c160",
   "metadata": {},
   "outputs": [],
   "source": [
    "ggsave(density_plot_qc_1, filename=\"result/qc/plot/density_plot_qc_umi.png\", width=10, height=5)\n",
    "ggsave(density_plot_qc_2, filename=\"result/qc/plot/density_plot_qc_feature.png\", width=10, height=5)\n",
    "ggsave(density_plot_qc_3, filename=\"result/qc/plot/density_plot_qc_mt.png\", width=10, height=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01d7439d-47c3-43c7-8473-d0240ef3777f",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=30, repr.plot.height=5)\n",
    "density_plot_qc_1 + density_plot_qc_2 + density_plot_qc_3 + plot_layout(nrow=1) & theme(legend.position=\"bottom\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc9d7850-4f6d-49a7-8992-20242f1b2d4a",
   "metadata": {},
   "source": [
    "## Scattern plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a73ac64-2e0a-4709-b4f1-d24dc94332af",
   "metadata": {},
   "outputs": [],
   "source": [
    "scattern_plot_qc_1 <- scattern_plot_qc(so=so_qc, title=\"Mitochondrial gene percentage\", color=pMt_RNA)\n",
    "scattern_plot_qc_2 <- scattern_plot_qc(so=so_qc, title=\"Hemoglobin gene percentage\", color=pHb_RNA)\n",
    "scattern_plot_qc_3 <- scattern_plot_qc(so=so_qc, title=\"Ribsonmal gene percentage\", color=pRb_RNA)\n",
    "scattern_plot_qc_4 <- scattern_plot_qc(so=so_qc, title=\"Fine labels\", color=label_fine_haemosphere) + scale_color_manual(values=color$label_fine_haemosphere)\n",
    "scattern_plot_qc_5 <- scattern_plot_qc(so=so_qc, title=\"QC class\", color=qc_class) + scale_color_manual(values=c(\"fail\"=\"red\", \"pass\"=\"black\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d53bcb99-0012-49db-afa4-4f0b46ff3974",
   "metadata": {},
   "outputs": [],
   "source": [
    "ggsave(scattern_plot_qc_1, filename=\"result/qc/plot/sc_mt.png\", width=10, height=5)\n",
    "ggsave(scattern_plot_qc_2, filename=\"result/qc/plot/sc_hg.png\", width=10, height=5)\n",
    "ggsave(scattern_plot_qc_3, filename=\"result/qc/plot/sc_rb.png\", width=10, height=5)\n",
    "ggsave(scattern_plot_qc_4, filename=\"result/qc/plot/sc_fl.png\", width=10, height=5)\n",
    "ggsave(scattern_plot_qc_5, filename=\"result/qc/plot/sc_qc.png\", width=10, height=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bb5d0f5-3bcf-40c1-9fc5-d19ad45a489b",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=15, repr.plot.height=15)\n",
    "scattern_plot_qc_1 + scattern_plot_qc_2 + scattern_plot_qc_3 + scattern_plot_qc_4 + scattern_plot_qc_5 + plot_layout(nrow=3) & theme(legend.position=\"bottom\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6d63748-676a-4648-838d-9b4d72e35587",
   "metadata": {},
   "source": [
    "## Box plots "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4866bc8d-ff44-4a9a-acea-e7e682fd3917",
   "metadata": {},
   "outputs": [],
   "source": [
    "box_plot_qc_1 <- box_plot_qc(so=so_qc, y=nCount_RNA, fill=tissue, ylab=\"UMI [count]\", ymin=0)\n",
    "box_plot_qc_2 <- box_plot_qc(so=so_qc, y=nFeature_RNA, fill=tissue, ylab=\"Feature [count]\", ymin=0)\n",
    "box_plot_qc_3 <- box_plot_qc(so=so_qc, y=pMt_RNA, fill=tissue, ylab=\"Mt [%]\", ymin=0, ymax=100)\n",
    "box_plot_qc_4 <- box_plot_qc(so=so_qc, y=pHb_RNA, fill=tissue, ylab=\"Hb [%]\", ymin=0, ymax=100)\n",
    "box_plot_qc_5 <- box_plot_qc(so=so_qc, y=pRb_RNA, fill=tissue, ylab=\"Rbl [%]\", ymin=0, ymax=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1364bd2-ba90-4df9-94ed-3d76fa5eaa12",
   "metadata": {},
   "outputs": [],
   "source": [
    "ggsave(box_plot_qc_1[[1]] + box_plot_qc_1[[2]] + plot_layout(guides=\"collect\", ncol=1) & theme(legend.position=\"bottom\"), filename=\"result/qc/plot/box_plot_qc_1.png\", width=7.5, height=5)\n",
    "ggsave(box_plot_qc_2[[1]] + box_plot_qc_2[[2]] + plot_layout(guides=\"collect\", ncol=1) & theme(legend.position=\"bottom\"), filename=\"result/qc/plot/box_plot_qc_2.png\", width=7.5, height=5)\n",
    "ggsave(box_plot_qc_3[[1]] + box_plot_qc_3[[2]] + plot_layout(guides=\"collect\", ncol=1) & theme(legend.position=\"bottom\"), filename=\"result/qc/plot/box_plot_qc_3.png\", width=7.5, height=5)\n",
    "ggsave(box_plot_qc_4[[1]] + box_plot_qc_4[[2]] + plot_layout(guides=\"collect\", ncol=1) & theme(legend.position=\"bottom\"), filename=\"result/qc/plot/box_plot_qc_4.png\", width=7.5, height=5)\n",
    "ggsave(box_plot_qc_5[[1]] + box_plot_qc_5[[2]] + plot_layout(guides=\"collect\", ncol=1) & theme(legend.position=\"bottom\"), filename=\"result/qc/plot/box_plot_qc_5.png\", width=7.5, height=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08ad030a-f42b-461c-8d29-4ad0a3cb8fa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=15, repr.plot.height=20)\n",
    "box_plot_qc_1[[1]] + box_plot_qc_1[[2]] + box_plot_qc_2[[1]]  + box_plot_qc_2[[2]] + \n",
    "box_plot_qc_3[[1]] + box_plot_qc_3[[2]] + box_plot_qc_4[[1]]  + box_plot_qc_4[[2]] + \n",
    "box_plot_qc_5[[1]] + box_plot_qc_5[[2]] + plot_spacer() + plot_spacer() + plot_layout(guides=\"collect\", ncol=2) & theme(legend.position=\"none\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d68a8bed-f077-4610-b6ae-f2a63c65d3dd",
   "metadata": {},
   "source": [
    "# Filter by QC class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bf84b82-a006-4cff-83c7-768c28a73434",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_qc <- subset(so_qc, subset = qc_class==\"pass\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70c5a7b3-bce2-4d9a-b5cc-29cca3458045",
   "metadata": {},
   "source": [
    "# UMAP and clustering of all samples on all genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3fa54e5-d48c-40fd-9db1-ed6f74fbd06a",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_qc <- NormalizeData(so_qc, assay=\"RNA\")\n",
    "so_qc <- ScaleData(so_qc, assay=\"RNA\", verbose=FALSE)\n",
    "so_qc <- FindVariableFeatures(so_qc, nfeatures=3000, assay=\"RNA\", verbose=FALSE)\n",
    "so_qc <- RunPCA(so_qc, npcs=50, assay=\"RNA\", features=NULL, verbose=FALSE)\n",
    "so_qc <- FindNeighbors(so_qc, dims=1:10, k.param=20, verbose=FALSE)\n",
    "so_qc <- FindClusters(so_qc, verbose=FALSE, resolution=1, algorithm=1, group.singletons=TRUE)\n",
    "so_qc <- RunUMAP(so_qc, dims=1:50, n.neighbors=30, verbose=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04927611-29ac-47e4-96b9-e0bdfa68d159",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=22.5, repr.plot.height=15)\n",
    "\n",
    "dplot_1 <- dplot(so_qc, reduction=\"umap\", group_by=\"seurat_clusters\", label=TRUE) + ggtitle(\"Louvain cluster\")\n",
    "dplot_2 <- dplot(so_qc, reduction=\"umap\", group_by=\"tissue\") + scale_color_manual(values=color$tissue) + ggtitle(\"Tissue\")\n",
    "dplot_3 <- dplot(so_qc, reduction=\"umap\", group_by=\"treatment\") + scale_color_manual(values=color$treatment) + ggtitle(\"Treatment\")\n",
    "dplot_4 <- dplot(so_qc, reduction=\"umap\", group_by=\"cc_phase_class\") + scale_color_manual(values=color$cc_phase_class) + ggtitle(\"Cell cycle\")\n",
    "dplot_5 <- dplot(so_qc, reduction=\"umap\", group_by=\"label_main_haemosphere\") + scale_color_manual(values=color$label_main_haemosphere) + ggtitle(\"Haemosphere label (main)\")\n",
    "dplot_6 <- dplot(so_qc, reduction=\"umap\", group_by=\"label_fine_haemosphere\") + scale_color_manual(values=color$label_fine_haemosphere) + ggtitle(\"Haemosphere label (fine)\")\n",
    "dplot_7 <- dplot(so_qc, reduction=\"umap\", group_by=\"sample_rep\") + ggtitle(\"Replicate\")\n",
    "fplot_1 <- fplot(so_qc, reduction=\"umap\", features=\"Trac\") + scale_colour_viridis_c(option=\"inferno\") + ggtitle(\"Trac\")\n",
    "fplot_2 <- fplot(so_qc, reduction=\"umap\", features=\"Igkc\") + scale_colour_viridis_c(option=\"inferno\") + ggtitle(\"Igkc\")\n",
    "\n",
    "dplot_comb <- dplot_1 + dplot_2 + dplot_3 + dplot_4 + dplot_5 + dplot_6 + dplot_7 + fplot_1 + fplot_2 + plot_layout(ncol=3)\n",
    "dplot_comb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80299ea1-b3c4-4626-ac03-e794a4c8c8d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_qc@meta.data %>% dplyr::group_by(seurat_clusters, treatment, sample_rep) %>% \n",
    "    dplyr::summarise(n=n()) %>% data.frame() %>% \n",
    "    tidyr::spread(seurat_clusters, n) %>% \n",
    "    kableExtra::kable(\"html\") %>% as.character() %>% IRdisplay::display_html()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca96a203-1779-443c-ab20-059b341a0a0a",
   "metadata": {},
   "source": [
    "# Doublet detection \n",
    "\n",
    "I use the strategy from [Sala et al., 2019](https://www.nature.com/articles/s41586-019-0933-9) which was also used in [Barile et al., 2021](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02414-y) with the source documented [here](https://github.com/MarioniLab/EmbryoTimecourse2018/blob/master/analysis_scripts/atlas/5_doublets/doublets.Rmd). We only do the refined method with combined samples. The functions used in the script are documented [here](https://github.com/MarioniLab/EmbryoTimecourse2018/blob/master/analysis_scripts/atlas/core_functions.R). We do doublet detection on sample groups (Myeloid and Progenitor cells of a replicated) and integrated those groups per batch and then by treatment with FastMNN"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0772771-50f9-4939-b0c8-392ebf09c7bc",
   "metadata": {},
   "source": [
    "## Doublet score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4c89386-b6f2-4429-97d3-f00e7ca27b18",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split data by sample group \n",
    "so_qc <- SplitObject(so_qc, split.by=\"sample_group\")\n",
    "\n",
    "# Compute doublet density per sample group \n",
    "so_qc <- lapply(so_qc, function(so) {\n",
    "    \n",
    "    DefaultAssay(so) <- \"RNA\"\n",
    "    \n",
    "    # Remove empty genes from split \n",
    "    cnt <- GetAssayData(so, assay=\"RNA\", slot=\"counts\")\n",
    "    cnt <- cnt[rowSums(cnt) > 0, ]\n",
    "    \n",
    "    # Highly variable features \n",
    "    so <- ScaleData(so, features=rownames(cnt), assay=\"RNA\", verbose=FALSE)\n",
    "    so <- FindVariableFeatures(so, assay=\"RNA\", nfeatures=2000)\n",
    "    \n",
    "    # Compute doublet density \n",
    "    so$doublet_score <- scDblFinder::computeDoubletDensity(cnt, subset.row=VariableFeatures(so), size.factors.norm=so$nCount_RNA, dims=50)\n",
    "    so$doublet_score_log2 <- log2(so$doublet_score)\n",
    "    \n",
    "    return(so)\n",
    "    \n",
    "}\n",
    "               )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5088c71-7c4b-4288-b734-2e3f34a98619",
   "metadata": {},
   "source": [
    "## Sample group doublets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3776ffaf-180c-463d-b4b6-499d2680a128",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute sub-cluster per sample group \n",
    "so_qc <- lapply(so_qc, function(so) {\n",
    "    \n",
    "    # Set default assay \n",
    "    DefaultAssay(so) <- \"RNA\"\n",
    "    \n",
    "    # Highly variable genes across all data sets \n",
    "    cnt <- GetAssayData(so, assay=\"RNA\", slot=\"data\")\n",
    "    cnt <- cnt[rowSums(cnt)>0, ]\n",
    "    \n",
    "    # Cluster function     \n",
    "    so <- ScaleData(so, features=rownames(cnt), assay=\"RNA\", verbose=FALSE)\n",
    "    so <- RunPCA(so, npcs=50, features=rownames(cnt), assay=\"RNA\", verbose=FALSE)\n",
    "    so <- FindNeighbors(so, dims=1:50, k.param=15, reduction=\"pca\", verbose=FALSE)\n",
    "    so <- FindClusters(so, verbose=FALSE, resolution=1, algorithm=1)\n",
    "\n",
    "    # Compute sub-cluster \n",
    "    cluster_list <- list()\n",
    "    for(cluster_id in unique(so$seurat_clusters)) {\n",
    "        \n",
    "        sink(\"NUL\")\n",
    "        so <- FindSubCluster(so, cluster_id, graph.name=\"RNA_snn\", subcluster.name=\"doublet_cluster\", resolution=1, algorithm=1)\n",
    "        sink()\n",
    "\n",
    "        cluster <- so@meta.data[so$seurat_clusters==cluster_id, ] \n",
    "        cluster <- cluster[, c(\"cell_id\", \"doublet_cluster\")]\n",
    "\n",
    "        cluster_list[[cluster_id]] <- cluster\n",
    "\n",
    "    }\n",
    "\n",
    "    cluster <- do.call(rbind, cluster_list)\n",
    "    rownames(cluster) <- cluster$cell_id\n",
    "    cluster <- cluster[, \"doublet_cluster\", drop=FALSE]\n",
    "\n",
    "    so <- AddMetaData(so, cluster)\n",
    "    \n",
    "    return(so)\n",
    "    \n",
    "}\n",
    "               )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bb0baed-de11-4703-b803-3096918c9a4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "doublet_stat <- lapply(so_qc, function(so) {\n",
    "    \n",
    "    meta <- so@meta.data\n",
    "    \n",
    "    # Function to computes MAD for values of x greater the median\n",
    "    mad_upper <- function(x){\n",
    "\n",
    "        x <- x-median(x)\n",
    "        x_mad <- mad(x[x>0], center=0)\n",
    "        return(x_mad)\n",
    "\n",
    "    }\n",
    "\n",
    "    # Comput median doublet density per cluster\n",
    "    doublet_stat <- dplyr::group_by(meta, doublet_cluster, sample_group) %>% \n",
    "        dplyr::summarise(doublet_cluster_density=median(doublet_score))\n",
    "\n",
    "    # Compute median-centered, MAD-estimated variance normal distribution of cluster medians. \n",
    "    doublet_stat$p=pnorm(\n",
    "        \n",
    "        doublet_stat$doublet_cluster_density, \n",
    "        mean=median(doublet_stat$doublet_cluster_density), \n",
    "        sd=mad_upper(doublet_stat$doublet_cluster_density), \n",
    "        lower.tail=FALSE\n",
    "        \n",
    "    )\n",
    "    \n",
    "    return(doublet_stat)\n",
    "    \n",
    "}\n",
    "               )\n",
    "\n",
    "doublet_stat <- do.call(rbind, doublet_stat)\n",
    "doublet_stat$fdr=p.adjust(doublet_stat$p, method=\"fdr\")\n",
    "    \n",
    "# Call doublet cluster in meta data \n",
    "so_qc <- lapply(names(so_qc), function(sample_group){\n",
    "    \n",
    "    so <- so_qc[[sample_group]]\n",
    "    \n",
    "    so$doublet_class <- ifelse(so$doublet_cluster %in% doublet_stat[doublet_stat$fdr < 0.1 & doublet_stat$sample_group==sample_group, ]$doublet_cluster, \"Doublet\", \"Singlet\")\n",
    "    \n",
    "    return(so)\n",
    "    \n",
    "}\n",
    "              )\n",
    "\n",
    "names(so_qc) <- lapply(so_qc, function(so) {so$sample_group[1]})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0246672e-7863-4c6f-8b79-3489375447a3",
   "metadata": {},
   "source": [
    "## Treatment group doublets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03491209-cc20-42ca-8fa1-4f9871a663e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# FastMNN on batch  \n",
    "so_qc_1 <- RunFastMNN(so_qc[c(\"NaCl_Rep2\", \"NaCl_Rep1\")], assay=\"RNA\", reconstructed.assay=\"MNN\")\n",
    "so_qc_2 <- RunFastMNN(so_qc[c(\"CpG_Rep2\", \"CpG_Rep1\")], assay=\"RNA\", reconstructed.assay=\"MNN\")\n",
    "\n",
    "# FastMNN on treatment groups \n",
    "so_qc <- RunFastMNN(list(so_qc_1, so_qc_2), assay=\"MNN\", reconstructed.assay=\"MNN\")\n",
    "    \n",
    "DefaultAssay(so_qc) <- \"MNN\"\n",
    "\n",
    "so_qc <- FindNeighbors(so_qc, dims=1:50, k.param=15, reduction=\"mnn\", graph.name=c(\"MNN_nn\", \"MNN_snn\"), verbose=FALSE)\n",
    "so_qc <- FindClusters(so_qc, verbose=FALSE, resolution=1, algorithm=1, graph.name=\"MNN_snn\") \n",
    "so_qc <- RunUMAP(so_qc, dims=1:50, n.neighbors=30, reduction=\"mnn\", verbose=FALSE)\n",
    "\n",
    "so_qc$doublet_cluster_int <- so_qc$seurat_clusters\n",
    "\n",
    "DefaultAssay(so_qc) <- \"RNA\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "556382fc-9d40-45dd-a8d2-723b9ef3b03e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to computes MAD for values of x greater the median\n",
    "mad_upper <- function(x){\n",
    "    \n",
    "    x <- x-median(x)\n",
    "    x_mad <- mad(x[x>0], center=0)\n",
    "    return(x_mad)\n",
    "\n",
    "}\n",
    "    \n",
    "doublet_stat <- dplyr::group_by(so_qc@meta.data, doublet_cluster_int) %>% \n",
    "    dplyr::mutate(cluster_count=n()) %>% dplyr::ungroup() %>% \n",
    "    dplyr::group_by(doublet_cluster_int, doublet_class, cluster_count) %>% \n",
    "    dplyr::summarise(doublet_ratio=n()) %>% dplyr::ungroup() %>% \n",
    "    dplyr::mutate(doublet_ratio=doublet_ratio/cluster_count) %>% \n",
    "    dplyr::filter(doublet_class==\"Doublet\")\n",
    "\n",
    "doublet_stat$p=pnorm(doublet_stat$doublet_ratio, mean=median(doublet_stat$doublet_ratio), sd=mad_upper(doublet_stat$doublet_ratio), lower.tail=FALSE)\n",
    "doublet_stat$fdr=p.adjust(doublet_stat$p, method=\"fdr\")\n",
    "    \n",
    "so_qc$doublet_class_int <- ifelse(so_qc$doublet_cluster_int %in% doublet_stat[doublet_stat$fdr < 0.1, ]$doublet_cluster_int, \"Doublet\", \"Singlet\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c91b80c5-c691-4dee-b66e-148f59eff1d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=22.5, repr.plot.height=20)\n",
    "\n",
    "dplot_1 <- dplot(so_qc, reduction=\"umap\", group_by=\"treatment\") + scale_color_manual(values=color$treatment) + ggtitle(\"Treatment\")\n",
    "dplot_2 <- dplot(so_qc, reduction=\"umap\", group_by=\"tissue\") + scale_color_manual(values=color$tissue) + ggtitle(\"Tissue\")\n",
    "dplot_3 <- dplot(so_qc, reduction=\"umap\", group_by=\"cc_phase_class\") + scale_color_manual(values=color$cc_phase_class) + ggtitle(\"Cell cycle\")\n",
    "dplot_4 <- dplot(so_qc, reduction=\"umap\", group_by=\"label_fine_haemosphere\") + scale_color_manual(values=color$label_fine_haemosphere) + ggtitle(\"Haemosphere label\")\n",
    "fplot_1 <- fplot(so_qc, reduction=\"umap\", features=\"doublet_score_log2\") + scale_colour_viridis_c(option=\"inferno\") + ggtitle(\"Log2(Doublet score)\")\n",
    "dplot_5 <- dplot(so_qc, reduction=\"umap\", group_by=\"doublet_class\") + scale_color_manual(values=c(\"Doublet\"=\"black\", \"Singlet\"=\"light grey\")) + ggtitle(\"Doublet class sample\")\n",
    "dplot_6 <- dplot(so_qc, reduction=\"umap\", group_by=\"doublet_class_int\") + scale_color_manual(values=c(\"Doublet\"=\"black\", \"Singlet\"=\"light grey\")) + ggtitle(\"Doublet class integrated\")\n",
    "fplot_2 <- fplot(so_qc, reduction=\"umap\", features=\"pMt_RNA\") + scale_colour_viridis_c(option=\"inferno\") + ggtitle(\"Percentage Mt\")\n",
    "fplot_3 <- fplot(so_qc, reduction=\"umap\", features=\"pHb_RNA\") + scale_colour_viridis_c(option=\"inferno\") + ggtitle(\"Percentage Hb\")\n",
    "fplot_4 <- fplot(so_qc, reduction=\"umap\", features=\"pRb_RNA\") + scale_colour_viridis_c(option=\"inferno\") + ggtitle(\"Percentage Rb\")\n",
    "fplot_5 <- fplot(so_qc, reduction=\"umap\", features=\"nCount_RNA\") + scale_colour_viridis_c(option=\"inferno\") + ggtitle(\"UMI count\")\n",
    "fplot_6 <- fplot(so_qc, reduction=\"umap\", features=\"nFeature_RNA\") + scale_colour_viridis_c(option=\"inferno\") + ggtitle(\"Feature count\")\n",
    "\n",
    "dplot_comb <- dplot_1 + dplot_2 + dplot_3 + dplot_4 + fplot_1 + dplot_5 + dplot_6 + fplot_2 + fplot_3 + fplot_4 + fplot_5 + fplot_6 + plot_layout(ncol=3)\n",
    "dplot_comb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c5d0e98-3515-4063-a687-1ca173acadf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=22.5, repr.plot.height=5)\n",
    "\n",
    "dplot_1 <- dplot(so_qc, reduction=\"umap\", group_by=\"seurat_clusters\", label=TRUE) + ggtitle(\"Louvain cluster\")\n",
    "fplot_1 <- fplot(so_qc, reduction=\"umap\", features=\"Trac\") + scale_colour_viridis_c(option=\"inferno\") + ggtitle(\"Trac\")\n",
    "fplot_2 <- fplot(so_qc, reduction=\"umap\", features=\"Igkc\") + scale_colour_viridis_c(option=\"inferno\") + ggtitle(\"Igkc\")\n",
    "\n",
    "dplot_comb <- dplot_1 + fplot_1 + fplot_2 + plot_layout(ncol=3)\n",
    "dplot_comb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f3fe672-bc91-4489-8238-c9a5d1553bfd",
   "metadata": {},
   "source": [
    "# Filter out doublets, T- and B- lymphocytes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a685e5f-552e-4a99-89b1-04474cece19c",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_qc <- subset(so_qc, subset=doublet_class==\"Singlet\")\n",
    "so_qc <- subset(so_qc, subset=doublet_class_int==\"Singlet\")\n",
    "so_qc <- subset(so_qc, subset=label_fine_haemosphere!=\"CLP\") # Filter out T lymphocytes\n",
    "so_qc <- subset(so_qc, subset=label_fine_haemosphere!=\"T-cell\") # Filter out T lymphocytes\n",
    "so_qc <- subset(so_qc, subset=label_fine_haemosphere!=\"NK\") # Filter out T lymphocytes\n",
    "so_qc <- subset(so_qc, subset=label_main_haemosphere!=\"NK\") # Filter out T lymphocytes\n",
    "so_qc <- subset(so_qc, subset=seurat_clusters!=18) # Filter out T lymphocytes\n",
    "so_qc <- subset(so_qc, subset=seurat_clusters!=19) # Filter out B lymphocytes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5af2668b-2a1a-4dd2-9d65-68facc961e52",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=22.5, repr.plot.height=20)\n",
    "\n",
    "dplot_1 <- dplot(so_qc, reduction=\"umap\", group_by=\"treatment\") + scale_color_manual(values=color$treatment) + ggtitle(\"Treatment\")\n",
    "dplot_2 <- dplot(so_qc, reduction=\"umap\", group_by=\"tissue\") + scale_color_manual(values=color$tissue) + ggtitle(\"Tissue\")\n",
    "dplot_3 <- dplot(so_qc, reduction=\"umap\", group_by=\"cc_phase_class\") + scale_color_manual(values=color$cc_phase_class) + ggtitle(\"Cell cycle\")\n",
    "dplot_4 <- dplot(so_qc, reduction=\"umap\", group_by=\"label_fine_haemosphere\") + scale_color_manual(values=color$label_fine_haemosphere) + ggtitle(\"Haemosphere label\")\n",
    "fplot_1 <- fplot(so_qc, reduction=\"umap\", features=\"doublet_score_log2\") + scale_colour_viridis_c(option=\"inferno\") + ggtitle(\"Log2(Doublet score)\")\n",
    "dplot_5 <- dplot(so_qc, reduction=\"umap\", group_by=\"doublet_class\") + scale_color_manual(values=c(\"Doublet\"=\"black\", \"Singlet\"=\"light grey\")) + ggtitle(\"Doublet class sample\")\n",
    "dplot_6 <- dplot(so_qc, reduction=\"umap\", group_by=\"doublet_class_int\") + scale_color_manual(values=c(\"Doublet\"=\"black\", \"Singlet\"=\"light grey\")) + ggtitle(\"Doublet class integrated\")\n",
    "fplot_2 <- fplot(so_qc, reduction=\"umap\", features=\"pMt_RNA\") + scale_colour_viridis_c(option=\"inferno\") + ggtitle(\"Percentage Mt\")\n",
    "fplot_3 <- fplot(so_qc, reduction=\"umap\", features=\"pHb_RNA\") + scale_colour_viridis_c(option=\"inferno\") + ggtitle(\"Percentage Hb\")\n",
    "fplot_4 <- fplot(so_qc, reduction=\"umap\", features=\"pRb_RNA\") + scale_colour_viridis_c(option=\"inferno\") + ggtitle(\"Percentage Rb\")\n",
    "fplot_5 <- fplot(so_qc, reduction=\"umap\", features=\"nCount_RNA\") + scale_colour_viridis_c(option=\"inferno\") + ggtitle(\"UMI count\")\n",
    "fplot_6 <- fplot(so_qc, reduction=\"umap\", features=\"nFeature_RNA\") + scale_colour_viridis_c(option=\"inferno\") + ggtitle(\"Feature count\")\n",
    "\n",
    "dplot_comb <- dplot_1 + dplot_2 + dplot_3 + dplot_4 + fplot_1 + dplot_5 + dplot_6 + fplot_2 + fplot_3 + fplot_4 + fplot_5 + fplot_6 + plot_layout(ncol=3)\n",
    "dplot_comb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81540eea-60d6-44de-94bd-74d7ff5bd645",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=22.5, repr.plot.height=5)\n",
    "\n",
    "dplot_1 <- dplot(so_qc, reduction=\"umap\", group_by=\"seurat_clusters\", label=TRUE) + ggtitle(\"Louvain cluster\")\n",
    "fplot_1 <- fplot(so_qc, reduction=\"umap\", features=\"Trac\") + scale_colour_viridis_c(option=\"inferno\") + ggtitle(\"Trac\")\n",
    "fplot_2 <- fplot(so_qc, reduction=\"umap\", features=\"Igkc\") + scale_colour_viridis_c(option=\"inferno\") + ggtitle(\"Igkc\")\n",
    "\n",
    "dplot_comb <- dplot_1 + fplot_1 + fplot_2 + plot_layout(ncol=3)\n",
    "dplot_comb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e835b18-c54c-4d07-888c-cb18bab44277",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique(so_qc$label_fine_haemosphere)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d485173-78e9-4db8-8bf1-e79c9e7b42f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique(so_qc$label_main_haemosphere)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "034ef09c-6d91-4ba1-b886-af4a9b85bc8d",
   "metadata": {},
   "source": [
    "# Save result objso_qc_rds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95cb0e1a-a21e-4272-a3d8-47834d85e573",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store data as RDS \n",
    "saveRDS(so_qc, so_qc_rds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc9ee337-5daf-482e-bb7c-7f3ab20940b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store data as h5ad \n",
    "adata <- import(\"anndata\", as=\"adata\", convert=FALSE)\n",
    "pd <- import(\"pandas\", as=\"pd\", convert=FALSE)\n",
    "np <- import(\"numpy\", as=\"np\", convert=FALSE)\n",
    "\n",
    "    \n",
    "# Transform dgCMatrix to \n",
    "X <- GetAssayData(so_qc, assay=\"RNA\", slot=\"counts\") %>% as.matrix() %>% t()\n",
    "X <- np$array(X, dtype=np$int32)\n",
    "    \n",
    "adata <- adata$AnnData(X=X, obs=so_qc@meta.data)\n",
    "adata$var_names <- rownames(GetAssayData(so_qc, assay=\"RNA\", slot=\"counts\"))\n",
    "\n",
    "adata$raw <- adata\n",
    "adata$write_h5ad(so_qc_h5ad)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3573893c-81fa-4f9d-aef4-764b80bc18c6",
   "metadata": {},
   "source": [
    "# Session info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8650431-fd79-4f68-bc8f-57c8d4decdef",
   "metadata": {},
   "outputs": [],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "r.4.1.0-FD20200109SPLENO",
   "language": "R",
   "name": "r.4.1.0-fd20200109spleno"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
