{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8dab95dd-2e8e-472e-b1ef-e9566cd08bfb",
   "metadata": {},
   "source": [
    "# Panel 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75fd7b70-81de-41ed-b00b-54dd26951307",
   "metadata": {},
   "outputs": [],
   "source": [
    "library_load <- suppressMessages(\n",
    "    \n",
    "    suppressWarnings(\n",
    "        \n",
    "        list(\n",
    "        \n",
    "            # Seurat \n",
    "            library(Seurat), \n",
    "\n",
    "            # For PCA\n",
    "            library(DESeq2), \n",
    "            \n",
    "            # Data\n",
    "            library(dplyr), \n",
    "\n",
    "            # STRING\n",
    "            library(STRINGdb), \n",
    "            library(igraph), \n",
    "            library(tidygraph), \n",
    "            library(ggraph), \n",
    "            \n",
    "            # Plotting \n",
    "            library(ggplot2), \n",
    "            library(patchwork), \n",
    "            library(ComplexHeatmap), \n",
    "            library(grid) # For gpar\n",
    "            \n",
    "        )\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7ee4d36-c8a2-4cf8-a1a0-bfa6233f4b12",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c965fc68-e06b-404d-b9b7-6da6d498a9f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set working directory to project root\n",
    "setwd(\"C:/Users/Florian Deckert/Projects/FD20200109SPLENO/\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "929a6504-28cd-45e6-ad0d-cf4f6ca2c6f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Source\n",
    "source(\"bin/so_pl.R\")\n",
    "source(\"bin/dea_pl.R\")\n",
    "source(\"bin/gsea_pp.R\")\n",
    "source(\"bin/gsea_pl.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d5d13bf-d70f-405b-b7b8-a96b73cbd542",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting Theme\n",
    "source(\"plotting_global.R\")\n",
    "ggplot2::theme_set(theme_global_set(size_select=1)) # From project global source()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6690aad8-9f43-4e72-8fee-c36424e70168",
   "metadata": {},
   "source": [
    "# Import and parameter "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b3435a9-39ba-4711-b42c-31686f6e473e",
   "metadata": {},
   "outputs": [],
   "source": [
    "singler <- read.csv(\"result/singler/bulkRNAseq/singler_res.csv\", row.names=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef27618a-2747-4373-93a8-42f2ea2a872c",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- readRDS(\"data/scRNAseq/object/pp_0.rds\")\n",
    "so <- AddMetaData(so, singler)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc0789c6-4eaf-4451-b289-97a2870cd4fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- NormalizeData(so)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1df8b657-62e9-4069-bca8-2b5c77cf0ae0",
   "metadata": {},
   "source": [
    "# SingleR UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7cee7d3-91b8-4ce3-b444-ee6cb1025b4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "fp_1 <- list(\n",
    "\n",
    "    fplot(so, features=\"cmo_score\", pt_size=0.25, min_cutoff=0, order=TRUE, restrict=c(\"sample_group\", \"Bl6_NaCl_D6\"), size_select=4), \n",
    "    fplot(so, features=\"cmo_score\", pt_size=0.25, min_cutoff=0, order=TRUE, restrict=c(\"sample_group\", \"Bl6_CpG_D6\"), size_select=4), \n",
    "    fplot(so, features=\"rpm_score\", pt_size=0.25, min_cutoff=0, order=TRUE, restrict=c(\"sample_group\", \"Bl6_NaCl_D6\"), size_select=4), \n",
    "    fplot(so, features=\"rpm_score\", pt_size=0.25, min_cutoff=0, order=TRUE, restrict=c(\"sample_group\", \"Bl6_CpG_D6\"), size_select=4), \n",
    "    fplot(so, features=\"Spic\", pt_size=0.25, min_cutoff=0, order=TRUE, slot=\"data\", assay=\"RNA\", restrict=c(\"sample_group\", \"Bl6_NaCl_D6\"), size_select=4), \n",
    "    fplot(so, features=\"Spic\", pt_size=0.25, min_cutoff=0, order=TRUE, slot=\"data\", assay=\"RNA\", restrict=c(\"sample_group\", \"Bl6_CpG_D6\"), size_select=4), \n",
    "    fplot(so, features=\"Gdf15\", pt_size=0.25, min_cutoff=0, order=TRUE, slot=\"data\", assay=\"RNA\", restrict=c(\"sample_group\", \"Bl6_NaCl_D6\"), size_select=4), \n",
    "    fplot(so, features=\"Gdf15\", pt_size=0.25, min_cutoff=0, order=TRUE, slot=\"data\", assay=\"RNA\", restrict=c(\"sample_group\", \"Bl6_CpG_D6\"), size_select=4)\n",
    "\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1a84498-163f-4dca-8570-11438db87b83",
   "metadata": {},
   "outputs": [],
   "source": [
    "fp_1 <- lapply(fp_1, function(p) egg::set_panel_size(p, width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")))\n",
    "fp_1 <- do.call(gridExtra::arrangeGrob, c(fp_1, ncol=2, nrow=length(fp_1)/2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "032de9cc-bd32-451e-b153-41d2b7eb7710",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/fp_rpm_score.pdf\", width=2*2.0, height=1.5*length(fp_1)/2)\n",
    "\n",
    "grid::grid.draw(fp_1)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a105560-61e6-488b-bf05-a2ef56cc54f2",
   "metadata": {},
   "source": [
    "# DEA result count heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed17047e-5bff-4860-90bb-24acc30a13fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "adj_pval_thr <- 0.05\n",
    "log2FC_thr <- 0.50"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fa1788d-ff7e-4d08-ab74-de71abfb3af0",
   "metadata": {},
   "source": [
    "## Fetch results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48a7ef59-2c1e-40bf-8304-e91a9480306f",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_dea_1 <- readRDS(\"result/dea/bulkRNAseq/res_dea_cmo.rds\")\n",
    "res_dea_2 <- readRDS(\"result/dea/bulkRNAseq/res_dea_rpm.rds\")\n",
    "res_dea_3 <- readRDS(\"result/dea/bulkRNAseq/res_dea_ery.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32c5fe3e-2c06-406a-824c-ec57c370c2c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "dpi_lrt <- dplyr::left_join(\n",
    "\n",
    "    do.call(rbind, \n",
    "    \n",
    "        list(\n",
    "            \n",
    "            res_dea_1[[\"res\"]][[\"dpi_wt_lrt\"]] %>% dplyr::mutate(celltype=\"cMo\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            \n",
    "            res_dea_2[[\"res\"]][[\"dpi_wt_lrt\"]] %>% dplyr::mutate(celltype=\"RPM\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            \n",
    "            res_dea_3[[\"res\"]][[\"dpi_wt_lrt\"]] %>% dplyr::mutate(celltype=\"Ery\") %>% tibble::rownames_to_column(\"gene\")\n",
    "        \n",
    "            )\n",
    "    \n",
    "        ) %>% dplyr::select(gene, log2FoldChange, ihw, celltype) %>% dplyr::rename(dpi_wt_lrt_fc=log2FoldChange, dpi_wt_lrt_ihw=ihw), \n",
    "\n",
    "    do.call(rbind, \n",
    "    \n",
    "        list(\n",
    "            \n",
    "            res_dea_1[[\"res\"]][[\"dpi_ko_lrt\"]] %>% dplyr::mutate(celltype=\"cMo\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            \n",
    "            res_dea_2[[\"res\"]][[\"dpi_ko_lrt\"]] %>% dplyr::mutate(celltype=\"RPM\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            \n",
    "            res_dea_3[[\"res\"]][[\"dpi_ko_lrt\"]] %>% dplyr::mutate(celltype=\"Ery\") %>% tibble::rownames_to_column(\"gene\")\n",
    "        \n",
    "            )\n",
    "    \n",
    "        ) %>% dplyr::select(gene, log2FoldChange, ihw, celltype) %>% dplyr::rename(dpi_ko_lrt_fc=log2FoldChange, dpi_ko_lrt_ihw=ihw), \n",
    "\n",
    "\n",
    "    by=join_by(gene, celltype)\n",
    "\n",
    ") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83b57e71-7071-4bec-b70b-73ba1ad981d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "dpi_wald <- dplyr::left_join(\n",
    "\n",
    "    do.call(rbind, \n",
    "    \n",
    "        list(\n",
    "            \n",
    "            res_dea_1[[\"res\"]][[\"dpi_wt_wald\"]][[1]] %>% dplyr::mutate(dpi=\"D1\", celltype=\"cMo\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            res_dea_1[[\"res\"]][[\"dpi_wt_wald\"]][[2]] %>% dplyr::mutate(dpi=\"D3\", celltype=\"cMo\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            res_dea_1[[\"res\"]][[\"dpi_wt_wald\"]][[3]] %>% dplyr::mutate(dpi=\"D6\", celltype=\"cMo\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            \n",
    "            res_dea_2[[\"res\"]][[\"dpi_wt_wald\"]][[1]] %>% dplyr::mutate(dpi=\"D1\", celltype=\"RPM\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            res_dea_2[[\"res\"]][[\"dpi_wt_wald\"]][[2]] %>% dplyr::mutate(dpi=\"D3\", celltype=\"RPM\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            res_dea_2[[\"res\"]][[\"dpi_wt_wald\"]][[3]] %>% dplyr::mutate(dpi=\"D6\", celltype=\"RPM\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            \n",
    "            res_dea_3[[\"res\"]][[\"dpi_wt_wald\"]][[1]] %>% dplyr::mutate(dpi=\"D1\", celltype=\"Ery\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            res_dea_3[[\"res\"]][[\"dpi_wt_wald\"]][[2]] %>% dplyr::mutate(dpi=\"D3\", celltype=\"Ery\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            res_dea_3[[\"res\"]][[\"dpi_wt_wald\"]][[3]] %>% dplyr::mutate(dpi=\"D6\", celltype=\"Ery\") %>% tibble::rownames_to_column(\"gene\")\n",
    "        \n",
    "            )\n",
    "    \n",
    "        ) %>% dplyr::select(gene, log2FoldChange, ihw, dpi, celltype) %>% dplyr::rename(dpi_wt_wald_fc=log2FoldChange, dpi_wt_wald_ihw=ihw), \n",
    "\n",
    "    do.call(rbind, \n",
    "    \n",
    "        list(\n",
    "            \n",
    "            res_dea_1[[\"res\"]][[\"dpi_ko_wald\"]][[1]] %>% dplyr::mutate(dpi=\"D1\", celltype=\"cMo\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            res_dea_1[[\"res\"]][[\"dpi_ko_wald\"]][[2]] %>% dplyr::mutate(dpi=\"D3\", celltype=\"cMo\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            res_dea_1[[\"res\"]][[\"dpi_ko_wald\"]][[3]] %>% dplyr::mutate(dpi=\"D6\", celltype=\"cMo\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "\n",
    "            res_dea_2[[\"res\"]][[\"dpi_ko_wald\"]][[1]] %>% dplyr::mutate(dpi=\"D1\", celltype=\"RPM\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            res_dea_2[[\"res\"]][[\"dpi_ko_wald\"]][[2]] %>% dplyr::mutate(dpi=\"D3\", celltype=\"RPM\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            res_dea_2[[\"res\"]][[\"dpi_ko_wald\"]][[3]] %>% dplyr::mutate(dpi=\"D6\", celltype=\"RPM\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            \n",
    "            res_dea_3[[\"res\"]][[\"dpi_ko_wald\"]][[1]] %>% dplyr::mutate(dpi=\"D1\", celltype=\"Ery\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            res_dea_3[[\"res\"]][[\"dpi_ko_wald\"]][[2]] %>% dplyr::mutate(dpi=\"D3\", celltype=\"Ery\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            res_dea_3[[\"res\"]][[\"dpi_ko_wald\"]][[3]] %>% dplyr::mutate(dpi=\"D6\", celltype=\"Ery\") %>% tibble::rownames_to_column(\"gene\")\n",
    "        \n",
    "            )\n",
    "    \n",
    "        ) %>% dplyr::select(gene, log2FoldChange, ihw, dpi, celltype) %>% dplyr::rename(dpi_ko_wald_fc=log2FoldChange, dpi_ko_wald_ihw=ihw), \n",
    "\n",
    "\n",
    "    by=join_by(gene, dpi, celltype)\n",
    "\n",
    ") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f4240ee-c47d-493f-b2c4-ea05f195555d",
   "metadata": {},
   "outputs": [],
   "source": [
    "dpi_genotype_wald <- do.call(rbind, \n",
    "    \n",
    "        list(\n",
    "            \n",
    "            res_dea_1[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[2]] %>% dplyr::mutate(dpi=\"D1\", celltype=\"cMo\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            res_dea_1[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[3]] %>% dplyr::mutate(dpi=\"D3\", celltype=\"cMo\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            res_dea_1[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[4]] %>% dplyr::mutate(dpi=\"D6\", celltype=\"cMo\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            \n",
    "            res_dea_2[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[2]] %>% dplyr::mutate(dpi=\"D1\", celltype=\"RPM\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            res_dea_2[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[3]] %>% dplyr::mutate(dpi=\"D3\", celltype=\"RPM\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            res_dea_2[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[4]] %>% dplyr::mutate(dpi=\"D6\", celltype=\"RPM\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            \n",
    "            res_dea_3[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[2]] %>% dplyr::mutate(dpi=\"D1\", celltype=\"Ery\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            res_dea_3[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[3]] %>% dplyr::mutate(dpi=\"D3\", celltype=\"Ery\") %>% tibble::rownames_to_column(\"gene\"), \n",
    "            res_dea_3[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[4]] %>% dplyr::mutate(dpi=\"D6\", celltype=\"Ery\") %>% tibble::rownames_to_column(\"gene\")\n",
    "        \n",
    "            )\n",
    "    \n",
    "        ) %>% dplyr::select(gene, log2FoldChange, ihw, dpi, celltype) %>% dplyr::rename(genotype_wt_vs_ko_wald_fc=log2FoldChange, genotype_wt_vs_ko_wald_ihw=ihw)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bda20e1-1392-4e55-a9f0-2f614e19c240",
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- dplyr::left_join(dpi_genotype_wald, dpi_wald, by=join_by(gene, dpi, celltype)) %>% dplyr::left_join(., dpi_lrt, by=join_by(gene, celltype))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c998107b-94f0-4f28-888e-620150d6ae3a",
   "metadata": {},
   "source": [
    "## Response genes (jitter plot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8734966-d0e9-4b90-8ed7-410b7cf88ecc",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- rbind(\n",
    "\n",
    "    data %>% dplyr::select(dpi, gene, celltype, dpi_wt_wald_fc, dpi_wt_wald_ihw, dpi_wt_lrt_fc, dpi_wt_lrt_ihw) %>% dplyr::rename(dpi_wald_fc=dpi_wt_wald_fc, dpi_wald_ihw=dpi_wt_wald_ihw, dpi_lrt_fc=dpi_wt_lrt_fc, dpi_lrt_ihw=dpi_wt_lrt_ihw) %>% dplyr::mutate(genotype=\"+/+\"), \n",
    "    data %>% dplyr::select(dpi, gene, celltype, dpi_ko_wald_fc, dpi_ko_wald_ihw, dpi_ko_lrt_fc, dpi_ko_lrt_ihw) %>% dplyr::rename(dpi_wald_fc=dpi_ko_wald_fc, dpi_wald_ihw=dpi_ko_wald_ihw, dpi_lrt_fc=dpi_ko_lrt_fc, dpi_lrt_ihw=dpi_ko_lrt_ihw) %>% dplyr::mutate(genotype=\"cre/+\")\n",
    "    \n",
    ") %>% \n",
    "    dplyr::mutate(sig=ifelse(genotype==\"+/+\" & dpi_wald_ihw<=adj_pval_thr & abs(dpi_wald_fc)>=log2FC_thr & dpi_lrt_ihw<=adj_pval_thr, \"+/+\", ifelse(genotype==\"cre/+\" & dpi_wald_ihw<=adj_pval_thr & abs(dpi_wald_fc)>=log2FC_thr & dpi_lrt_ihw<=adj_pval_thr, \"cre/+\", \"n.s.\"))) %>% \n",
    "    dplyr::mutate(celltype=factor(celltype, levels=c(\"cMo\", \"RPM\", \"Ery\"))) %>% na.omit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b44b1312-0e30-4dcd-aafb-842bc1fc9e44",
   "metadata": {},
   "outputs": [],
   "source": [
    "# UP-only comparison\n",
    "up_counts <- mat_1 %>% dplyr::mutate(sign=ifelse(dpi_wald_fc>0, \"(UP)\", \"(DOWN)\")) %>% dplyr::filter(sig!=\"n.s.\") %>% dplyr::select(gene, dpi, celltype, sig, sign) %>%\n",
    "  filter(sig %in% c(\"+/+\", \"cre/+\"), sign == \"(UP)\") %>%\n",
    "  distinct(gene, dpi, celltype, sig) %>%\n",
    "  mutate(present = 1L) %>%\n",
    "  pivot_wider(names_from = sig, values_from = present, values_fill = 0L) %>%\n",
    "  mutate(category = case_when(\n",
    "    `+/+` == 1L & `cre/+` == 1L ~ \"UP_shared\",\n",
    "    `+/+` == 1L & `cre/+` == 0L ~ \"UP_+/+_only\",\n",
    "    `+/+` == 0L & `cre/+` == 1L ~ \"UP_cre/+_only\"\n",
    "  )) %>%\n",
    "  count(dpi, celltype, category)\n",
    "\n",
    "# DOWN-only comparison\n",
    "down_counts <- mat_1 %>% dplyr::mutate(sign=ifelse(dpi_wald_fc>0, \"(UP)\", \"(DOWN)\")) %>% dplyr::filter(sig!=\"n.s.\") %>% dplyr::select(gene, dpi, celltype, sig, sign) %>%\n",
    "  filter(sig %in% c(\"+/+\", \"cre/+\"), sign == \"(DOWN)\") %>%\n",
    "  distinct(gene, dpi, celltype, sig) %>%\n",
    "  mutate(present = 1L) %>%\n",
    "  pivot_wider(names_from = sig, values_from = present, values_fill = 0L) %>%\n",
    "  mutate(category = case_when(\n",
    "    `+/+` == 1L & `cre/+` == 1L ~ \"DOWN_shared\",\n",
    "    `+/+` == 1L & `cre/+` == 0L ~ \"DOWN_+/+_only\",\n",
    "    `+/+` == 0L & `cre/+` == 1L ~ \"DOWN_cre/+_only\"\n",
    "  )) %>%\n",
    "  count(dpi, celltype, category)\n",
    "\n",
    "counts_1 <- bind_rows(up_counts, down_counts) %>% arrange(dpi, celltype, category)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2653e53f-bde2-468b-b537-0e8f1a51c44b",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- mat_1 %>% dplyr::mutate(sign_log10_padj=-sign(dpi_wald_fc)*log10(dpi_wald_ihw)) %>% dplyr::mutate(sign_log10_padj=ifelse(abs(sign_log10_padj)>60, sign(dpi_wald_fc)*60, sign_log10_padj)) %>% \n",
    "    dplyr::mutate(dpi_wald_fc=ifelse(abs(dpi_wald_fc)>15, sign(dpi_wald_fc)*15, dpi_wald_fc))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a32d02fe-59b1-47be-a91c-1f8a6d0570f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "jp_1 <- ggplot(mat_1 %>% dplyr::filter(celltype==\"cMo\"), aes(y=sign_log10_padj, x=genotype, color=sig)) + \n",
    "    geom_jitter(data=mat_1 %>% dplyr::filter(celltype==\"cMo\" & sig==\"n.s.\"), size=0.1) + \n",
    "    geom_jitter(data=mat_1 %>% dplyr::filter(celltype==\"cMo\" & sig!=\"n.s.\"), size=0.1) + \n",
    "    geom_hline(yintercept=0) +\n",
    "    ylim(-60, +60) + \n",
    "    scale_color_manual(values=c(\"+/+\"=\"#cd34b5\", \"cre/+\"=\"#FFAC1E\", \"n.s.\"=\"#7f7f7f\")) + \n",
    "    ylab(\"\") + xlab(\"\") + \n",
    "    facet_grid(~dpi, scales=\"free\") + theme(aspect.ratio=2) + theme_global_set(4)\n",
    "\n",
    "jp_2 <- ggplot(mat_1 %>% dplyr::filter(celltype==\"RPM\"), aes(y=sign_log10_padj, x=genotype, color=sig)) + \n",
    "    geom_jitter(data=mat_1 %>% dplyr::filter(celltype==\"RPM\" & sig==\"n.s.\"), size=0.1) + \n",
    "    geom_jitter(data=mat_1 %>% dplyr::filter(celltype==\"RPM\" & sig!=\"n.s.\"), size=0.1) + \n",
    "    geom_hline(yintercept=0) +\n",
    "    ylim(-60, +60) + \n",
    "    scale_color_manual(values=c(\"+/+\"=\"#cd34b5\", \"cre/+\"=\"#FFAC1E\", \"n.s.\"=\"#7f7f7f\")) + \n",
    "    ylab(\"\") + xlab(\"\") + \n",
    "    facet_grid(~dpi, scales=\"free\") + theme(aspect.ratio=2) + theme_global_set(4)\n",
    "\n",
    "jp_3 <- ggplot(mat_1 %>% dplyr::filter(celltype==\"Ery\"), aes(y=sign_log10_padj, x=genotype, color=sig)) + \n",
    "    geom_jitter(data=mat_1 %>% dplyr::filter(celltype==\"Ery\" & sig==\"n.s.\"), size=0.1) + \n",
    "    geom_jitter(data=mat_1 %>% dplyr::filter(celltype==\"Ery\" & sig!=\"n.s.\"), size=0.1) +  \n",
    "    geom_hline(yintercept=0) +\n",
    "    ylim(-60, +60) + \n",
    "    scale_color_manual(values=c(\"+/+\"=\"#cd34b5\", \"cre/+\"=\"#FFAC1E\", \"n.s.\"=\"#7f7f7f\")) + \n",
    "    ylab(\"\") + xlab(\"\") + \n",
    "    facet_grid(~dpi, scales=\"free\") + theme(aspect.ratio=2) + theme_global_set(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "009d5ba6-d9aa-4bdf-8b1c-6cdb9c01404b",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/jp_dea_count_dpi.pdf\", width=3, height=4)\n",
    "\n",
    "    gridExtra::grid.arrange(\n",
    "            \n",
    "        jp_1 %>% egg::set_panel_size(., width=unit(1, \"cm\"), height=unit(1.5, \"cm\")), \n",
    "        jp_2 %>% egg::set_panel_size(., width=unit(1, \"cm\"), height=unit(1.5, \"cm\")), \n",
    "        jp_3 %>% egg::set_panel_size(., width=unit(1, \"cm\"), height=unit(1.5, \"cm\")), \n",
    "        nrow=3\n",
    "        \n",
    "    )\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aabf60bd-ba07-449a-b40d-f6af685ffbc5",
   "metadata": {},
   "source": [
    "## Genotype perturbation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86ca7cde-c5b0-4f12-babc-55576586433d",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- data %>% dplyr::select(gene, genotype_wt_vs_ko_wald_fc, genotype_wt_vs_ko_wald_ihw, dpi, celltype) %>% \n",
    "    dplyr::mutate(sig=ifelse(genotype_wt_vs_ko_wald_ihw<=adj_pval_thr & genotype_wt_vs_ko_wald_fc<=-log2FC_thr, \"+/+\", ifelse(genotype_wt_vs_ko_wald_ihw<=adj_pval_thr & genotype_wt_vs_ko_wald_fc>=+log2FC_thr, \"cre/+\", \"n.s.\"))) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a564672-e40d-4268-bda9-814ef6010149",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- mat_1 %>% dplyr::mutate(sign_log10_padj=-sign(genotype_wt_vs_ko_wald_fc)*log10(genotype_wt_vs_ko_wald_ihw)) %>% dplyr::mutate(sign_log10_padj=ifelse(abs(sign_log10_padj)>10, sign(genotype_wt_vs_ko_wald_fc)*10, sign_log10_padj)) %>% \n",
    "    dplyr::mutate(genotype_wt_vs_ko_wald_fc=ifelse(abs(genotype_wt_vs_ko_wald_fc)>10, sign(genotype_wt_vs_ko_wald_fc)*10, genotype_wt_vs_ko_wald_fc))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44628086-54a7-43e4-8155-468426a51c02",
   "metadata": {},
   "outputs": [],
   "source": [
    "jp_1 <- ggplot(mat_1 %>% dplyr::filter(celltype==\"cMo\"), aes(y=sign_log10_padj, x=dpi, color=sig)) + \n",
    "    geom_jitter(data=mat_1 %>% dplyr::filter(celltype==\"cMo\" & sig==\"n.s.\"), size=0.1) + \n",
    "    geom_jitter(data=mat_1 %>% dplyr::filter(celltype==\"cMo\" & sig!=\"n.s.\"), size=0.1) + \n",
    "    geom_hline(yintercept=0) +\n",
    "    ylim(-10, +10) + \n",
    "    scale_color_manual(values=c(\"+/+\"=\"#cd34b5\", \"cre/+\"=\"#FFAC1E\", \"n.s.\"=\"#7f7f7f\")) + \n",
    "    ylab(\"\") + xlab(\"\") + \n",
    "    facet_grid(~dpi, scales=\"free\") + theme(aspect.ratio=2) + theme_global_set(4)\n",
    "\n",
    "jp_2 <- ggplot(mat_1 %>% dplyr::filter(celltype==\"RPM\"), aes(y=sign_log10_padj, x=dpi, color=sig)) + \n",
    "    geom_jitter(data=mat_1 %>% dplyr::filter(celltype==\"RPM\" & sig==\"n.s.\"), size=0.1) + \n",
    "    geom_jitter(data=mat_1 %>% dplyr::filter(celltype==\"RPM\" & sig!=\"n.s.\"), size=0.1) + \n",
    "    geom_hline(yintercept=0) +\n",
    "    ylim(-10, +10) + \n",
    "    scale_color_manual(values=c(\"+/+\"=\"#cd34b5\", \"cre/+\"=\"#FFAC1E\", \"n.s.\"=\"#7f7f7f\")) + \n",
    "    ylab(\"\") + xlab(\"\") + \n",
    "    facet_grid(~dpi, scales=\"free\") + theme(aspect.ratio=2) + theme_global_set(4)\n",
    "\n",
    "jp_3 <- ggplot(mat_1 %>% dplyr::filter(celltype==\"Ery\"), aes(y=sign_log10_padj, x=dpi, color=sig)) + \n",
    "    geom_jitter(data=mat_1 %>% dplyr::filter(celltype==\"Ery\" & sig==\"n.s.\"), size=0.1) + \n",
    "    geom_jitter(data=mat_1 %>% dplyr::filter(celltype==\"Ery\" & sig!=\"n.s.\"), size=0.1) + \n",
    "    geom_hline(yintercept=0) +\n",
    "    ylim(-10, +10) + \n",
    "    scale_color_manual(values=c(\"+/+\"=\"#cd34b5\", \"cre/+\"=\"#FFAC1E\", \"n.s.\"=\"#7f7f7f\")) + \n",
    "    ylab(\"\") + xlab(\"\") + \n",
    "    facet_grid(~dpi, scales=\"free\") + theme(aspect.ratio=2) + theme_global_set(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "405308e4-3f73-4bd8-af32-2dd0f1264991",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_2 <- mat_1 %>% dplyr::filter(sig!=\"n.s.\") %>% dplyr::group_by(dpi, celltype, sig) %>% dplyr::summarise(n=n()) %>% dplyr::arrange(celltype, dpi, sig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97852370-a95a-4e42-9cca-beaf9124cde1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/jp_dea_count_genotype.pdf\", width=3, height=4)\n",
    "\n",
    "    gridExtra::grid.arrange(\n",
    "            \n",
    "        jp_1 %>% egg::set_panel_size(., width=unit(0.5, \"cm\"), height=unit(1.5, \"cm\")), \n",
    "        jp_2 %>% egg::set_panel_size(., width=unit(0.5, \"cm\"), height=unit(1.5, \"cm\")), \n",
    "        jp_3 %>% egg::set_panel_size(., width=unit(0.5, \"cm\"), height=unit(1.5, \"cm\")), \n",
    "        nrow=3\n",
    "        \n",
    "    )\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb5fc0a4-9a52-483d-82aa-6af27fb41773",
   "metadata": {},
   "source": [
    "# Gene set enrichment analysis (GSEA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c22adf09-f8e1-4184-9f65-49123f62d7f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "adj_pval_thr <- 0.05\n",
    "log2FC_thr <- 0.50"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb2fd498-3202-4cff-93d8-82d1dadaad05",
   "metadata": {},
   "source": [
    "## Prepare results from genotype wald test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5b3b640-fa9c-4a1f-b2b1-590b26c16059",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_dea_1 <- readRDS(\"result/dea/bulkRNAseq/res_dea_cmo.rds\")\n",
    "res_dea_2 <- readRDS(\"result/dea/bulkRNAseq/res_dea_rpm.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "084d7a9b-c062-424c-ab34-31efc1ef760e",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_dea <- list(\"cMo\"=res_dea_1[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]], \"RPM\"=res_dea_2[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4a6485f-14a6-45da-8ca0-16dc34f51304",
   "metadata": {},
   "source": [
    "## Genotype comparison - HM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24d7c671-a5ac-4138-9bfc-ffeb900b678a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res_1 <- lapply(seq_along(res_dea[[1]]), function(i) gsea(res_dea[[1]][[i]] %>% dplyr::rename(avg_log2FC=log2FoldChange, p_val_adj=ihw), category=\"MH\", subcategory=NULL) %>% dplyr::mutate(col_split=\"cMo\", col_label=paste0(\"group_\", i)))\n",
    "gsea_res_2 <- lapply(seq_along(res_dea[[2]]), function(i) gsea(res_dea[[2]][[i]] %>% dplyr::rename(avg_log2FC=log2FoldChange, p_val_adj=ihw), category=\"MH\", subcategory=NULL) %>% dplyr::mutate(col_split=\"RPM\", col_label=paste0(\"group_\", i)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "864f7770-3c81-4433-8312-4143d86e056b",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res_1 <- do.call(\"rbind\", gsea_res_1)\n",
    "gsea_res_2 <- do.call(\"rbind\", gsea_res_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82ee3541-8136-4e92-9a49-58062ec4d213",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res <- rbind(gsea_res_1, gsea_res_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9535590-44cb-4df7-bdba-75af1de3bb8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs_name_select <- gsea_res %>% dplyr::filter(padj<=0.05) %>% dplyr::pull(pathway) %>% unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3600b30-75ec-427e-9f41-fad546db85e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- gsea_res %>% dplyr::filter(pathway %in% gs_name_select) %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=padj,\n",
    "        values_fill=1\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c60e5c0-425a-4ea1-85c9-7ad02adac45b",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_2 <- gsea_res %>% dplyr::filter(pathway %in% gs_name_select) %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=NES,\n",
    "        values_fill=0\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db14b5e9-16ca-436e-b72e-d1522f0c3e22",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- mat_1[, c(\"col_label\", \"col_split\", gs_name_select)]\n",
    "mat_2 <- mat_2[, c(\"col_label\", \"col_split\", gs_name_select)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49d7a1a4-b2a2-4bfd-9c59-09d124917a27",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_label <- mat_1$col_label\n",
    "col_split <- mat_1$col_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f5e626c-902c-43a0-a56d-f59a6d7a84f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- mat_1 %>% dplyr::select(-col_label, -col_split) %>% t()\n",
    "mat_2 <- mat_2 %>% dplyr::select(-col_label, -col_split) %>% t()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93c1afaa-77cf-404c-9e0b-cf0fea5cb109",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/hm_gsea_genotype_dpi_mh.pdf\", width=8, height=50)\n",
    "\n",
    "gsea_hm(mat_1, mat_2, col_label=col_label, col_split=col_split, row_split=NULL, pathway_suffix=NULL, color_neg=color$sample_group[1], color_pos=color$sample_group[2], fontsize_select=2)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98777de4-21ea-4fb0-9e4f-e04e7a24f1fe",
   "metadata": {},
   "source": [
    "## Genotype comparison - GOBP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfc24469-f1c2-4800-95ca-2c8f2893bce5",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res_1 <- lapply(seq_along(res_dea[[1]]), function(i) gsea(res_dea[[1]][[i]] %>% dplyr::rename(avg_log2FC=log2FoldChange, p_val_adj=ihw), category=\"M5\", subcategory=\"GO:BP\") %>% dplyr::mutate(col_split=\"cMo\", col_label=paste0(\"group_\", i)))\n",
    "gsea_res_2 <- lapply(seq_along(res_dea[[2]]), function(i) gsea(res_dea[[2]][[i]] %>% dplyr::rename(avg_log2FC=log2FoldChange, p_val_adj=ihw), category=\"M5\", subcategory=\"GO:BP\") %>% dplyr::mutate(col_split=\"RPM\", col_label=paste0(\"group_\", i)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c260852-48e9-4119-8991-763abde2741c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res_1 <- do.call(\"rbind\", gsea_res_1)\n",
    "gsea_res_2 <- do.call(\"rbind\", gsea_res_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "085b7a40-325d-427f-b843-9aadc6eaab93",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res <- rbind(gsea_res_1, gsea_res_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0741ff7e-00de-4090-904e-9d92b8ad6ed7",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs_name_select <- gsea_res %>% dplyr::filter(padj<=0.05) %>% dplyr::pull(pathway) %>% unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "723ecf70-ba2b-4548-beda-ed2a2c0cd1c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- gsea_res %>% dplyr::filter(pathway %in% gs_name_select) %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=padj,\n",
    "        values_fill=1\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "974f53bd-1f61-4097-aed7-1524a8a4fccd",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_2 <- gsea_res %>% dplyr::filter(pathway %in% gs_name_select) %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=NES,\n",
    "        values_fill=0\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "732bdb50-c64c-4f99-897b-e4fabf6390b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- mat_1[, c(\"col_label\", \"col_split\", gs_name_select)]\n",
    "mat_2 <- mat_2[, c(\"col_label\", \"col_split\", gs_name_select)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64b22ab4-cd77-4764-9251-fb3c273202e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_label <- mat_1$col_label\n",
    "col_split <- mat_1$col_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2d8eb54-b006-4777-8819-f91dcd0c8cfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- mat_1 %>% dplyr::select(-col_label, -col_split) %>% t()\n",
    "mat_2 <- mat_2 %>% dplyr::select(-col_label, -col_split) %>% t()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "783c2c1e-9809-4f00-b544-4dfcd0ae0a1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/hm_gsea_genotype_dpi_gobp.pdf\", width=8, height=50)\n",
    "\n",
    "gsea_hm(mat_1, mat_2, col_label=col_label, col_split=col_split, row_split=NULL, pathway_suffix=\"GOBP\", color_neg=color$sample_group[1], color_pos=color$sample_group[2], fontsize_select=2)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f50c094-a064-42e8-a7a2-b5c7a0312863",
   "metadata": {},
   "source": [
    "# DEA heatmap "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58b354ff-326a-4cdc-a374-8d04978d80b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "adj_pval_thr <- 0.05\n",
    "log2FC_thr <- 0.50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "082888be-3f98-43ac-bfde-45bfa1250703",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_dea_fetch <- function(mat_1, dea_res, col_samples, col_split, fontsize_select, celltype, celltype_select, scale_mode=\"all\") {\n",
    "\n",
    "    mat_1 <- Seurat::CreateSeuratObject(counts=mat_1)\n",
    "    mat_1 <- NormalizeData(mat_1, normalization.method=\"RC\", scale.factor=1e6)\n",
    "    mat_1 <- GetAssayData(mat_1, assay=\"RNA\", layer=\"data\") %>% as.data.frame() \n",
    "\n",
    "    if(scale_mode==\"all\") {\n",
    "\n",
    "        mat_1 <- t(apply(mat_1, 1, scales::rescale))\n",
    "        mat_1 <- mat_1[row_split, col_samples]\n",
    "\n",
    "        colnames(mat_1) <- col_label\n",
    "\n",
    "        row_split <- setNames(unlist(row_genes, use.names=FALSE), rep(names(row_genes), sapply(row_genes, length)))\n",
    "        row_split <- row_split[row_split %in% rownames(mat_1)]\n",
    "        \n",
    "        mat_1 <- mat_1[row_split, celltype==celltype_select]\n",
    "        col_split <- col_split[celltype==celltype_select]\n",
    "        \n",
    "    } else if (scale_mode==\"sample\") {\n",
    "\n",
    "        mat_1 <- mat_1[row_split, col_samples]\n",
    "        mat_1 <- mat_1[row_split, col_samples]\n",
    "\n",
    "        colnames(mat_1) <- col_label\n",
    "\n",
    "        row_split <- setNames(unlist(row_genes, use.names=FALSE), rep(names(row_genes), sapply(row_genes, length)))\n",
    "        row_split <- row_split[row_split %in% rownames(mat_1)]\n",
    "        \n",
    "        mat_1 <- mat_1[row_split, celltype==celltype_select]\n",
    "        col_split <- col_split[celltype==celltype_select]\n",
    "        \n",
    "        mat_1 <- t(apply(mat_1, 1, scales::rescale))\n",
    "        \n",
    "    }\n",
    "    \n",
    "    # DEA log2FC \n",
    "    mat_2 <- purrr::reduce(\n",
    "        \n",
    "        list(\n",
    "            \n",
    "            res_dea[[celltype_select]][[1]] %>% dplyr::mutate(Ctl=log2FoldChange) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, Ctl), \n",
    "            res_dea[[celltype_select]][[2]] %>% dplyr::mutate(D1=log2FoldChange) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, D1), \n",
    "            res_dea[[celltype_select]][[3]] %>% dplyr::mutate(D3=log2FoldChange) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, D3), \n",
    "            res_dea[[celltype_select]][[4]] %>% dplyr::mutate(D6=log2FoldChange) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, D6)\n",
    "        \n",
    "        ), full_join, by=\"gene\"\n",
    "    \n",
    "    ) %>% tibble::column_to_rownames(\"gene\")\n",
    "\n",
    "    mat_2[setdiff(row_split, rownames(mat_2)), ] <- 0\n",
    "    mat_2 <- mat_2[row_split, ]\n",
    "    mat_2[is.na(mat_2)] <- 0\n",
    "\n",
    "    # DEA adj p-values\n",
    "    mat_3 <- purrr::reduce(\n",
    "        \n",
    "        list(\n",
    "            \n",
    "            res_dea[[celltype_select]][[1]] %>% dplyr::mutate(Ctl=padj) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, Ctl), \n",
    "            res_dea[[celltype_select]][[2]] %>% dplyr::mutate(D1=padj) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, D1), \n",
    "            res_dea[[celltype_select]][[3]] %>% dplyr::mutate(D3=padj) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, D3), \n",
    "            res_dea[[celltype_select]][[4]] %>% dplyr::mutate(D6=padj) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, D6)\n",
    "        \n",
    "        ), full_join, by=\"gene\"\n",
    "    \n",
    "    ) %>% tibble::column_to_rownames(\"gene\")\n",
    "\n",
    "    mat_3[setdiff(row_split, rownames(mat_3)), ] <- 1\n",
    "    mat_3 <- mat_3[row_split, ]\n",
    "    mat_3[is.na(mat_3)] <- 1\n",
    "    mat_3[(mat_3>adj_pval_thr & abs(mat_2)<log2FC_thr)] <- 1\n",
    "\n",
    "    # Set log2FC to 0 if not significant \n",
    "    mat_2[abs(mat_2) < log2FC_thr] <- 0\n",
    "\n",
    "    # Row split\n",
    "    row_split=names(row_split)\n",
    "    row_split=factor(row_split, levels=names(row_genes))\n",
    "    \n",
    "    hm_1 <- dea_exp_hm(mat_1, col_split=col_split, row_split=row_split, fontsize_select=2, breaks_limit=1, scale_width=0.5, legend_title=\"Exp.\", cluster_rows=FALSE)\n",
    "    hm_2 <- dea_res_hm(mat_2, mat_3, p_val_adj_thr=adj_pval_thr, row_split=row_split, scale_width=1, color_min=\"#cd34b5\", color_max=\"#FFAC1E\", cluster_rows=FALSE, fontsize_select=2)\n",
    "\n",
    "    hm <- hm_2 + hm_1\n",
    "\n",
    "    return(hm)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19d0a61c-470d-4c21-b611-efcfac0be901",
   "metadata": {},
   "source": [
    "# Prepare meta data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56fad3ad-74ed-4e35-a5d5-8dc3a6ada04c",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta <- readRDS(\"data/bulkRNAseq/object/meta.rds\") %>% dplyr::filter(celltype %in% c(\"cMo\", \"RPM\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "996c8765-b837-4ecd-bd7f-dff628f75465",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta <- meta %>% dplyr::arrange(celltype, dpi, genotype)\n",
    "\n",
    "col_samples <- as.character(meta$sample_name)\n",
    "col_label <- as.character(meta$genotype)\n",
    "col_split <- as.character(meta$dpi)\n",
    "celltype <- as.character(meta$celltype)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43321fd0-d20d-4ab5-b67c-e70e90b1e2da",
   "metadata": {},
   "source": [
    "## Prepare results from genotype wald test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ca0e799-fce5-4809-a936-da81b86f63a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_dea_1 <- readRDS(\"result/dea/bulkRNAseq/res_dea_cmo.rds\")\n",
    "res_dea_2 <- readRDS(\"result/dea/bulkRNAseq/res_dea_rpm.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c774fab-50b8-4885-8fca-8e80bb57faa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] <- res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::mutate(response_class_wt=ifelse(response_gene_wt, response_class_wt, NA), response_class_ko=ifelse(response_gene_ko, response_class_ko, NA))\n",
    "res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] <- res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::mutate(response_class_wt=ifelse(response_gene_wt, response_class_wt, NA), response_class_ko=ifelse(response_gene_ko, response_class_ko, NA))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2904a8a2-ccc5-4541-9b86-d99a53250086",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] <- res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter((response_gene_wt | response_gene_ko) & perturbation_gene) %>% dplyr::filter(!(is.na(response_class_wt) & is.na(response_class_ko)))\n",
    "res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] <- res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter((response_gene_wt | response_gene_ko) & perturbation_gene) %>% dplyr::filter(!(is.na(response_class_wt) & is.na(response_class_ko)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f620af1d-f8a6-413b-ae48-6cbba521b5d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_dea <- list(\"cMo\"=res_dea_1[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]], \"RPM\"=res_dea_2[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dada08c7-e237-4ed7-9db8-7486b60f2aa8",
   "metadata": {},
   "source": [
    "## Prepare expression mat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca71f644-1d5b-49d0-985a-db04a3aae74a",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta <- readRDS(\"data/bulkRNAseq/object/meta.rds\") %>% dplyr::filter(celltype %in% c(\"cMo\", \"RPM\")) %>% droplevels()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9b3f3a4-6ae5-4c1e-8355-4022cfec4a93",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- readRDS(\"data/bulkRNAseq/object/cnt.rds\") %>% dplyr::select(-mgi_symbol, -ensembl_gene_id)\n",
    "mat_1 <- mat_1[, rownames(meta)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62b2c9e5-4257-4129-acf0-e6cacb6c9b2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "all(rownames(meta)==colnames(mat_1))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "496cf99e-ad96-482a-95c0-4f9046615b98",
   "metadata": {},
   "source": [
    "## cMo cre/+ UP vs +/+"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb4bd382-03ac-4b72-8916-6724cb8cb530",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_select <- lapply(res_dea_1[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]], function(x) x %>% dplyr::filter(ihw<=adj_pval_thr & log2FoldChange>=log2FC_thr) %>% rownames())\n",
    "gene_select <- do.call(\"c\", gene_select) %>% unique()\n",
    "gene_select <- gene_select[!gene_select %in% grep(\"Gm|Rps|Rpl\", gene_select, value=TRUE)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fadc3113-a073-46e0-9f6b-ee0af62e17eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_genes <- list(\n",
    "\n",
    "    \"Repressed\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Repressed\" & response_class_ko==\"Repressed\") %>% rownames(), \n",
    "    \"Activated\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Activated\" & response_class_ko==\"Activated\") %>% rownames(), \n",
    "    \"Activated (+/+)\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Activated\" & is.na(response_class_ko)) %>% rownames(), \n",
    "    \"Activated (cre/+)\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(is.na(response_class_wt) & response_class_ko==\"Activated\") %>% rownames(), \n",
    "    \"Repressed (+/+)\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Repressed\" & is.na(response_class_ko)) %>% rownames(), \n",
    "    \"Repressed (cre/+)\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(is.na(response_class_wt) & response_class_ko==\"Repressed\") %>% rownames(),\n",
    "    \"Opposed\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt!=response_class_ko) %>% rownames()\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80472b28-ec90-44cf-aeb3-b52b4c6ee101",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_genes <- lapply(row_genes, function(x) x[x %in% gene_select])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bedc430-5f60-40fa-930f-361650f6aae5",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_split <- setNames(unlist(row_genes, use.names=FALSE), rep(names(row_genes), sapply(row_genes, length)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ecc30ca-c191-4c38-9272-54c199a78012",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_1 <- hm_dea_fetch(mat_1, dea_res, col_samples, col_split, fontsize_select=2, celltype, celltype_select=\"cMo\", scale_mode=\"sample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41a57f3c-3056-4146-85e5-d133a651d539",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/hm_dea_ko_up_vs_wt_cmo.pdf\", width=10, height=70)\n",
    "\n",
    "hm_1\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3432d272-f3c9-4c4e-af38-e9074ed23e57",
   "metadata": {},
   "source": [
    "## cMo cre/+ DOWN vs +/+"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2574e706-00af-4b9b-9d76-5fb933b368a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_select <- lapply(res_dea_1[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]], function(x) x %>% dplyr::filter(ihw<=adj_pval_thr & log2FoldChange<=-log2FC_thr) %>% rownames())\n",
    "gene_select <- do.call(\"c\", gene_select) %>% unique()\n",
    "gene_select <- gene_select[!gene_select %in% grep(\"Gm|Rps|Rpl\", gene_select, value=TRUE)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9c4e563-44af-46fa-8acf-0b72f4785f61",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_genes <- list(\n",
    "\n",
    "    \"Repressed\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Repressed\" & response_class_ko==\"Repressed\") %>% rownames(), \n",
    "    \"Activated\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Activated\" & response_class_ko==\"Activated\") %>% rownames(), \n",
    "    \"Activated (+/+)\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Activated\" & is.na(response_class_ko)) %>% rownames(), \n",
    "    \"Activated (cre/+)\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(is.na(response_class_wt) & response_class_ko==\"Activated\") %>% rownames(), \n",
    "    \"Repressed (+/+)\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Repressed\" & is.na(response_class_ko)) %>% rownames(), \n",
    "    \"Repressed (cre/+)\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(is.na(response_class_wt) & response_class_ko==\"Repressed\") %>% rownames(),\n",
    "    \"Opposed\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt!=response_class_ko) %>% rownames()\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac9f2f6d-97b6-4fb9-9a51-e391e8f383f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_genes <- lapply(row_genes, function(x) x[x %in% gene_select])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cfb8d8e-2ee2-4ff1-902d-66839e4ea94e",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_split <- setNames(unlist(row_genes, use.names=FALSE), rep(names(row_genes), sapply(row_genes, length)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cbb7804-cb08-4558-b7b8-6a352aeab896",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_1 <- hm_dea_fetch(mat_1, dea_res, col_samples, col_split, fontsize_select=2, celltype, celltype_select=\"cMo\", scale_mode=\"sample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42547209-9685-411d-a7fc-f9d971a16f85",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/hm_dea_ko_down_vs_wt_cmo.pdf\", width=10, height=80)\n",
    "\n",
    "hm_1\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9253b434-f1e1-4780-9331-f188a65075e2",
   "metadata": {},
   "source": [
    "## RPM cre/+ UP vs +/+"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "432b7646-24a2-4ea7-86a9-1374dacb614d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_select <- lapply(res_dea_2[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]], function(x) x %>% dplyr::filter(ihw<=adj_pval_thr & log2FoldChange>=log2FC_thr) %>% rownames())\n",
    "gene_select <- do.call(\"c\", gene_select) %>% unique()\n",
    "gene_select <- gene_select[!gene_select %in% grep(\"Gm|Rps|Rpl\", gene_select, value=TRUE)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8b5cbd7-84f7-443e-8553-1a3c1d8591b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_genes <- list(\n",
    "\n",
    "    \"Repressed\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Repressed\" & response_class_ko==\"Repressed\") %>% rownames(), \n",
    "    \"Activated\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Activated\" & response_class_ko==\"Activated\") %>% rownames(), \n",
    "    \"Activated (+/+)\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Activated\" & is.na(response_class_ko)) %>% rownames(), \n",
    "    \"Activated (cre/+)\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(is.na(response_class_wt) & response_class_ko==\"Activated\") %>% rownames(), \n",
    "    \"Repressed (+/+)\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Repressed\" & is.na(response_class_ko)) %>% rownames(), \n",
    "    \"Repressed (cre/+)\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(is.na(response_class_wt) & response_class_ko==\"Repressed\") %>% rownames(),\n",
    "    \"Opposed\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt!=response_class_ko) %>% rownames()\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e655dacc-6783-4f97-8e01-1fbe59cfac20",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_genes <- lapply(row_genes, function(x) x[x %in% gene_select])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0b75fd7-0b5a-4288-9c8c-7f1ac3d379cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_split <- setNames(unlist(row_genes, use.names=FALSE), rep(names(row_genes), sapply(row_genes, length)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab05c0df-8e6e-4a42-a884-bee5355c9634",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_1 <- hm_dea_fetch(mat_1, dea_res, col_samples, col_split, fontsize_select=2, celltype, celltype_select=\"RPM\", scale_mode=\"sample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0b6754a-d5f7-48cb-948d-f295d633acc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/hm_dea_ko_up_vs_wt_rpm.pdf\", width=10, height=30)\n",
    "\n",
    "hm_1\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93a8414c-e9e8-4f70-a0b5-066bdc5904d9",
   "metadata": {},
   "source": [
    "## RPM cre/+ DOWN vs +/+"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a5ff6d0-23a5-459c-a1e9-9c2f853c6434",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_select <- lapply(res_dea_2[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]], function(x) x %>% dplyr::filter(ihw<=adj_pval_thr & log2FoldChange<=-log2FC_thr) %>% rownames())\n",
    "gene_select <- do.call(\"c\", gene_select) %>% unique()\n",
    "gene_select <- gene_select[!gene_select %in% grep(\"Gm|Rps|Rpl\", gene_select, value=TRUE)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7448988b-ed1b-4e78-893d-56edbfbc872a",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_genes <- list(\n",
    "\n",
    "    \"Repressed\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Repressed\" & response_class_ko==\"Repressed\") %>% rownames(), \n",
    "    \"Activated\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Activated\" & response_class_ko==\"Activated\") %>% rownames(), \n",
    "    \"Activated (+/+)\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Activated\" & is.na(response_class_ko)) %>% rownames(), \n",
    "    \"Activated (cre/+)\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(is.na(response_class_wt) & response_class_ko==\"Activated\") %>% rownames(), \n",
    "    \"Repressed (+/+)\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Repressed\" & is.na(response_class_ko)) %>% rownames(), \n",
    "    \"Repressed (cre/+)\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(is.na(response_class_wt) & response_class_ko==\"Repressed\") %>% rownames(),\n",
    "    \"Opposed\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt!=response_class_ko) %>% rownames()\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ac46025-38dd-4916-a3c6-16e6c9096116",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_genes <- lapply(row_genes, function(x) x[x %in% gene_select])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff2b842d-adea-4723-9f48-a92a2a126130",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_split <- setNames(unlist(row_genes, use.names=FALSE), rep(names(row_genes), sapply(row_genes, length)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8965cc0c-10c9-4cde-b067-6b97675f8760",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_1 <- hm_dea_fetch(mat_1, dea_res, col_samples, col_split, fontsize_select=2, celltype, celltype_select=\"RPM\", scale_mode=\"sample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7ba094b-e251-49f7-b90a-edd84166471c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/hm_dea_ko_down_vs_wt_rpm.pdf\", width=10, height=25)\n",
    "\n",
    "hm_1\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9bb614d-9663-4dfb-8e0f-e7f9c79dbdf8",
   "metadata": {},
   "source": [
    "## Selected genes classical monocytes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbd9da69-4c7e-406b-b079-6db9f31013f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_genes <- list(\n",
    "\n",
    "    \"G1\"=c(\"Ddx60\", \"Ifih1\", \"Ifit1\", \"Ifit2\", \"Mx1\", \"Oasl1\", \"Oasl2\", \"Ccl2\", \"Cxcl10\"), # Interferon stimulation \n",
    "    \"G2\"=c(\"Hmox1\", \"MMaf\", \"Mafb\", \"Vcam1\", \"Fcer1g\", \"Hck\", \"Prkcd\", \"C1qb\", \"C5ar1\", \"Msr1\", \"Axl\"), # Macrophage development\n",
    "    \"G3\"=c(\"Slc40a1\", \"Slc48a1\", \"Ftl1\", \"Fth1\", \"Hmox2\", \"Nfe2l2\", \"Acsl4\", \"Sod2\", \"Prdx1\"), # Erythrophagocytosis \n",
    "    \"G4\"=c(\"Il1b\", \"Casp1\", \"Casp8\", \"Cflar\"), # Inflammasome\n",
    "    # \"G5\"=c(\"Tnfaip3\", \"Tnip1\", \"Il18bp\", \"Il10\"), # Antinfommatory \n",
    "    \"G6\"=c(\"Atp5b\", \"Uqcrh\", \"Cox7a2l\", \"Sod1\", \"Gpx4\", \"Pon2\"), # OXPHOS \n",
    "    \"G7\"=c(\"Psmb3\", \"Psmb5\", \"Psmb6\", \"Psmb8\") # Proteasome \n",
    "    # \"G8\"=c(\"Ctse\", \"Ctsh\", \"Sting1\") # Lysosome \n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ebddaf2-4af4-4798-81bb-985a6a95ab98",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_split <- setNames(unlist(row_genes, use.names=FALSE), rep(names(row_genes), sapply(row_genes, length)))\n",
    "row_split[!row_split %in% rownames(mat_1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "665c532f-1af4-4fe7-a3a2-a7b4a2b619b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_1 <- hm_dea_fetch(mat_1, dea_res, col_samples, col_split, fontsize_select=2, celltype, celltype_select=\"cMo\", scale_mode=\"sample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85a88ad3-a481-4df3-8439-0bad4c6a376d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/hm_dea_cmo_selected.pdf\", width=10, height=10)\n",
    "\n",
    "hm_1\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "003a85a6-1b6e-400b-940b-c8be6a7507e8",
   "metadata": {},
   "source": [
    "## Selected genes classical RPMs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "655823c9-a636-4ba9-99f2-1391624c8594",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_genes <- list(\n",
    "\n",
    "    \"G1\"=c(\"Stat2\", \"Irf7\", \"Ifit1\", \"Ifit2\", \"Ifitm3\", \"Ifi205\", \"Ifi206\", \"Ifi208\", \"Ifi211\", \"Usp18\"), # IFN-response \n",
    "    \"G2\"=c(\"Eif2ak2\", \"Atf4\", \"Egr1\", \"Gdf15\"), # Integrated stress response. Eif2ak2 is PKR, a kinase activator of \n",
    "    \"G3\"=c(\"Sod2\", \"Cflar\", \"Kitl\", \"Hif1a\", \"Ldha\"), \n",
    "    \"G4\"=c(\"Top2a\", \"Ube2c\", \"Cdk1\", \"Mcm5\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b417979b-82c6-486c-8c43-40c74a979210",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_split <- setNames(unlist(row_genes, use.names=FALSE), rep(names(row_genes), sapply(row_genes, length)))\n",
    "row_split[!row_split %in% rownames(mat_1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45afbd75-ad9f-4945-9cc9-9bf8ea4d6c9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_1 <- hm_dea_fetch(mat_1, dea_res, col_samples, col_split, fontsize_select=2, celltype, celltype_select=\"RPM\", scale_mode=\"sample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac3a66ec-98ad-4415-aec0-130e777a98eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/hm_dea_rpm_selected.pdf\", width=10, height=10)\n",
    "\n",
    "hm_1\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffe5afa4-52c9-446d-bea7-0ff54c6f3761",
   "metadata": {},
   "source": [
    "# PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ec278e6-2cb8-4327-b1b7-ad848f201f2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta <- readRDS(\"data/bulkRNAseq/object/meta.rds\") \n",
    "cnt <- readRDS(\"data/bulkRNAseq/object/cnt.rds\") %>% dplyr::select(-mgi_symbol, -ensembl_gene_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "820c6767-3509-4ade-8192-420f497ef753",
   "metadata": {},
   "source": [
    "## cMo "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19adb2da-83e9-431d-9fce-55235a532085",
   "metadata": {},
   "outputs": [],
   "source": [
    "cnt_1 <- cnt[, meta$celltype==\"cMo\"]\n",
    "meta_1 <- meta[meta$celltype==\"cMo\", ]\n",
    "all(rownames(meta_1)==colnames(cnt_1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71520cf9-5709-432b-801a-e666858452e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_1 <- c(\n",
    "    \n",
    "    res_dea_1[[\"resc\"]][[\"rc_wt\"]] %>% dplyr::filter(response_gene) %>% rownames(), \n",
    "    res_dea_1[[\"resc\"]][[\"rc_ko\"]] %>% dplyr::filter(response_gene) %>% rownames()\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da30d552-1531-4cfe-98c2-96652d5cb044",
   "metadata": {},
   "outputs": [],
   "source": [
    "cnt_1 <- cnt_1[rownames(cnt_1) %in% genes_1, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77395d73-e057-4543-a061-6401ca53555a",
   "metadata": {},
   "outputs": [],
   "source": [
    "dss <- DESeqDataSetFromMatrix(countData=cnt_1, colData=meta_1, design=~dpi)\n",
    "vst <- DESeq2::vst(dss)\n",
    "vst <- assay(vst)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d5130aa-0eed-4018-bbc3-cfb87be20e34",
   "metadata": {},
   "outputs": [],
   "source": [
    "rv  <- rowVars(vst)\n",
    "genes <- names(rv[order(rv, decreasing=TRUE)][1:100])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3241ab9-1192-4ecf-bf88-5cb1643bb973",
   "metadata": {},
   "outputs": [],
   "source": [
    "pca <- prcomp(t(vst[genes, , drop=FALSE]), center=TRUE, scale.=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "971525a6-939e-44df-a8f0-7c1c4b64ae2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- cbind(pca$x[, c(\"PC1\", \"PC2\")], meta_1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adf304f3-c803-4eca-9e99-0ca183912715",
   "metadata": {},
   "outputs": [],
   "source": [
    "centroids <- data %>% group_by(dpi) %>% summarise(PC1=mean(PC1), PC2=mean(PC2), .groups=\"drop\")\n",
    "segments <- data %>% left_join(centroids, by = \"dpi\", suffix = c(\"\", \"_centroid\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a7f5b6e-b2fc-47b7-b1d1-c3c8675f1b87",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pca_p <- ggplot(data, aes(x=PC1, y=PC2)) + \n",
    "    geom_point(data=centroids, size=1, shape=4, stroke=1.5) +\n",
    "    geom_segment(data=segments,aes(xend=PC1_centroid, yend=PC2_centroid, group=dpi), alpha=0.4) +\n",
    "    geom_point(aes(color=dpi, fill=dpi, shape=genotype), size=1, stroke=1) + \n",
    "    ggtitle(\"PCA (cMo)\") + \n",
    "    xlab(paste0(\"PC1: \", round(100*(pca$sdev^2)/sum(pca$sdev^2))[1],\"% variance\")) +\n",
    "    ylab(paste0(\"PC2: \", round(100*(pca$sdev^2)/sum(pca$sdev^2))[2],\"% variance\")) +\n",
    "    scale_color_manual(values=color$dpi) + \n",
    "    theme(aspect.ratio=1, legend.position=\"right\") + guides(fill=guide_legend(byrow=TRUE)) + guides(color=guide_legend(byrow=TRUE)) + guides(shape=guide_legend(byrow=TRUE)) + \n",
    "    theme_global_set(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c2044e6-434e-4f02-9ae7-1d9d6b13eb0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/pca_cmo.pdf\", width=5, height=2)\n",
    "\n",
    "pca_p\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62ab49f8-8bdb-4718-ade2-e180a694f0e9",
   "metadata": {},
   "source": [
    "## RPM "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78285722-83ca-4ba7-b47e-b19510565fcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "cnt_1 <- cnt[, meta$celltype==\"RPM\"]\n",
    "meta_1 <- meta[meta$celltype==\"RPM\", ]\n",
    "all(rownames(meta_1)==colnames(cnt_1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ec481be-aba4-426a-b33d-f124dca70c01",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_1 <- c(\n",
    "    \n",
    "    res_dea_2[[\"resc\"]][[\"rc_wt\"]] %>% dplyr::filter(response_gene) %>% rownames(), \n",
    "    res_dea_2[[\"resc\"]][[\"rc_ko\"]] %>% dplyr::filter(response_gene) %>% rownames()\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd690fcc-cbf4-4d71-8491-5e28de9d686f",
   "metadata": {},
   "outputs": [],
   "source": [
    "cnt_1 <- cnt_1[rownames(cnt_1) %in% genes_1, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71fb8c59-7ba7-4fe9-9a98-f6381ef4f83e",
   "metadata": {},
   "outputs": [],
   "source": [
    "dss <- DESeqDataSetFromMatrix(countData=cnt_1, colData=meta_1, design=~dpi)\n",
    "vst <- DESeq2::vst(dss)\n",
    "vst <- assay(vst)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fdcf254-7ef1-4e58-b5e2-c8493176209f",
   "metadata": {},
   "outputs": [],
   "source": [
    "rv  <- rowVars(vst)\n",
    "genes <- names(rv[order(rv, decreasing=TRUE)][1:100])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9a0f4bd-adad-4c7d-b06c-840509320a3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pca <- prcomp(t(vst[genes, , drop=FALSE]), center=TRUE, scale.=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6d587e7-2ec2-404b-ab8d-5cf65bde70b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- cbind(pca$x[, c(\"PC1\", \"PC2\")], meta_1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd5abf9e-994d-4605-8ed2-4d4f7db9da86",
   "metadata": {},
   "outputs": [],
   "source": [
    "centroids <- data %>% group_by(dpi) %>% summarise(PC1=mean(PC1), PC2=mean(PC2), .groups=\"drop\")\n",
    "segments <- data %>% left_join(centroids, by = \"dpi\", suffix = c(\"\", \"_centroid\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f62b826-2e84-47a9-9967-26df1b07bd4f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pca_p <- ggplot(data, aes(x=PC1, y=PC2)) + \n",
    "    geom_point(data=centroids, size=1, shape=4, stroke=1.5) +\n",
    "    geom_segment(data=segments,aes(xend=PC1_centroid, yend=PC2_centroid, group=dpi), alpha=0.4) +\n",
    "    geom_point(aes(color=dpi, fill=dpi, shape=genotype), size=1, stroke=1) + \n",
    "    ggtitle(\"PCA (cMo)\") + \n",
    "    xlab(paste0(\"PC1: \", round(100*(pca$sdev^2)/sum(pca$sdev^2))[1],\"% variance\")) +\n",
    "    ylab(paste0(\"PC2: \", round(100*(pca$sdev^2)/sum(pca$sdev^2))[2],\"% variance\")) +\n",
    "    scale_color_manual(values=color$dpi) + \n",
    "    theme(aspect.ratio=1, legend.position=\"right\") + guides(fill=guide_legend(byrow=TRUE)) + guides(color=guide_legend(byrow=TRUE)) + guides(shape=guide_legend(byrow=TRUE)) + \n",
    "    theme_global_set(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4ab1525-39e6-423b-a20a-bdd51a2f7cb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/pca_rpm.pdf\", width=5, height=2)\n",
    "\n",
    "pca_p\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91e9d08d-9e21-4193-b99d-0b5c821a19b0",
   "metadata": {},
   "source": [
    "# STRING "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8179409e-e3c2-4e45-9043-63a16c3fb296",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create STRING DB for mouse \n",
    "string_db <- STRINGdb::STRINGdb$new(version=\"12.0\", species=10090, score_threshold=400, network_type=\"full\", link_data=\"combined_only\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "655709a8-a610-4052-8a42-b390794d05de",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_dea_1 <- readRDS(\"result/dea/bulkRNAseq/res_dea_cmo.rds\")\n",
    "res_dea_2 <- readRDS(\"result/dea/bulkRNAseq/res_dea_rpm.rds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "895caecd-e46d-48b0-8fca-490392ce3c47",
   "metadata": {},
   "source": [
    "## cMo D1 UP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1683f75-657a-44d5-943f-c94205160782",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mapp DEA results to STRING DB\n",
    "string_db_dea <- res_dea_1[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[2]] %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, log2FoldChange, ihw) %>% dplyr::rename(logFC=log2FoldChange, padj=ihw) %>% na.omit() %>% dplyr::filter(padj<=adj_pval_thr & logFC>=+log2FC_thr) %>% string_db$map(., \"gene\", removeUnmappedRows=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4131d88c-0f30-4b4e-86b2-4682b8ae304a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Labe PPI vertices \n",
    "ppi_v <- string_db_dea %>% dplyr::distinct() %>% dplyr::rename(name=STRING_id) %>% \n",
    "    dplyr::group_by(name) %>% dplyr::mutate(n=n()) %>% dplyr::filter(n==1) %>% dplyr::select(name, gene, logFC, padj) %>% \n",
    "    dplyr::mutate(node_group=ifelse(padj<=0.05 & logFC<=-0.50, \"neg\", ifelse(padj<=0.05 & logFC>=0.50, \"pos\", \"n.s.\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60070b86-a42f-4d79-a9e5-9f8cddf182a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create PPI edges from DEA PPI\n",
    "ppi_e <- string_db$get_interactions(string_db_dea$STRING_id) %>% dplyr::filter(from %in% ppi_v$name & to %in% ppi_v$name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "857f8003-b488-4d6e-b316-95d1f9522f72",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute igraph object\n",
    "g <- igraph::graph_from_data_frame(\n",
    "    \n",
    "    d=ppi_e,\n",
    "    vertices=ppi_v, \n",
    "    directed=FALSE\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccca0b99-a73c-4de1-8e65-31e539e476cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Connected components \n",
    "min_component_size <- 4  # keep components with >= 4 vertices\n",
    "\n",
    "g_components <- igraph::components(g)\n",
    "g <- igraph::delete_vertices(g, which(g_components$csize[g_components$membership] < min_component_size))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f3e308f-24b0-4c3a-a1cd-399b56b5b830",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute Louvain cluster on combined score \n",
    "E(g)$weight <- E(g)$combined_score\n",
    "\n",
    "louvain <- igraph::cluster_louvain(g, weights=E(g)$weight)\n",
    "V(g)$louvain <- membership(louvain)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c254d89e-1e42-405b-9fe1-5b34bf5f868f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute GO enrichment for Louvain \n",
    "enrichment_louvain <- tibble::tibble(name=V(g)$name, louvain=V(g)$louvain) %>% group_by(louvain) %>%\n",
    "    \n",
    "    group_modify(~{\n",
    "        \n",
    "        enrichment <- string_db$get_enrichment(.x$name, category=\"Process\", methodMT=\"fdr\", iea=TRUE) %>% arrange(fdr)\n",
    "        if (is.null(enrichment) | nrow(enrichment)==0) {\n",
    "            \n",
    "            return(tibble::tibble())\n",
    "        \n",
    "        } else {\n",
    "\n",
    "            return(enrichment %>% arrange(fdr))\n",
    "            \n",
    "        }\n",
    "        \n",
    "    }\n",
    "                ) %>% ungroup()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77461335-c9be-485a-9349-498278db2378",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fetch GO enrichment \n",
    "enrichment_label <- enrichment_louvain %>% group_by(louvain) %>% arrange(fdr) %>% slice(1) %>% ungroup() %>% transmute(louvain, label=description)\n",
    "V(g)$enrichment_label <- enrichment_label$label[match(V(g)$louvain, enrichment_label$louvain)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71610b40-ea93-4ba9-9bcf-67395946c92d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove vertices without enrichment label \n",
    "g <- igraph::delete_vertices(g, which(is.na(V(g)$enrichment_label)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cba5311-b706-44ae-9207-2562cfa1e9e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot graph \n",
    "g_1 <- tidygraph::as_tbl_graph(g) %>% ggraph::ggraph(., layout=\"lgl\") +\n",
    "    geom_edge_link(alpha=0.2, width=0.1, color=\"#000000\") +\n",
    "    geom_node_point(aes(fill=factor(enrichment_label), size=abs(logFC)), color=\"black\", shape=21, stroke=0.1) +\n",
    "    # geom_node_text(aes(label=ifelse(gene %in% c(\"Slc40a1\", \"Slc48a1\", \"Nr1h3\", \"Hmox1\", \"Stat2\", \"Irf7\", \"Eif2ak2\", \"Gdf15\"), gene, NA)), repel=FALSE, size=2.1, color=\"black\") + \n",
    "    # scale_fill_manual(values=c(\"#6d6875\", \"#4f6d7a\", \"#4a6fa5\", \"#161b33\")) +\n",
    "    scale_size_continuous(range=c(0.5, 3)) + \n",
    "    theme(aspect.ratio=1) + theme_global_set(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c73779e-4d24-44aa-b4be-4e8f69519a1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/string_cmo_d1_ko_up.pdf\", width=10, height=10)\n",
    "\n",
    "gridExtra::grid.arrange(egg::set_panel_size(g_1, width=unit(5, \"cm\"), height=unit(5, \"cm\")))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "191ee485-6afa-4c80-a5ce-4f73c05067fc",
   "metadata": {},
   "source": [
    "## cMo D1 DOWN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2856a8d3-6202-40f6-b2d8-5727c6098cd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mapp DEA results to STRING DB\n",
    "string_db_dea <- res_dea_1[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[2]] %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, log2FoldChange, ihw) %>% dplyr::rename(logFC=log2FoldChange, padj=ihw) %>% na.omit() %>% dplyr::filter(padj<=adj_pval_thr & logFC<=-log2FC_thr) %>% string_db$map(., \"gene\", removeUnmappedRows=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c30715b6-d7aa-4ca8-975e-52de4aeec9d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Labe PPI vertices \n",
    "ppi_v <- string_db_dea %>% dplyr::distinct() %>% dplyr::rename(name=STRING_id) %>% \n",
    "    dplyr::group_by(name) %>% dplyr::mutate(n=n()) %>% dplyr::filter(n==1) %>% dplyr::select(name, gene, logFC, padj) %>% \n",
    "    dplyr::mutate(node_group=ifelse(padj<=0.05 & logFC<=-0.50, \"neg\", ifelse(padj<=0.05 & logFC>=0.50, \"pos\", \"n.s.\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ad6cae6-fa41-4e48-8a82-3c4d5a077d36",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create PPI edges from DEA PPI\n",
    "ppi_e <- string_db$get_interactions(string_db_dea$STRING_id) %>% dplyr::filter(from %in% ppi_v$name & to %in% ppi_v$name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a068eb8-af00-4b16-99cd-b7f5c36230bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute igraph object\n",
    "g <- igraph::graph_from_data_frame(\n",
    "    \n",
    "    d=ppi_e,\n",
    "    vertices=ppi_v, \n",
    "    directed=FALSE\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a67a9acd-41d4-40cf-9348-ec0c4fdaed22",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Connected components \n",
    "min_component_size <- 4  # keep components with >= 4 vertices\n",
    "\n",
    "g_components <- igraph::components(g)\n",
    "g <- igraph::delete_vertices(g, which(g_components$csize[g_components$membership] < min_component_size))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a27dff6b-1728-4826-a44f-9ee5de545809",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute Louvain cluster on combined score \n",
    "E(g)$weight <- E(g)$combined_score\n",
    "\n",
    "louvain <- igraph::cluster_louvain(g, weights=E(g)$weight)\n",
    "V(g)$louvain <- membership(louvain)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f050b62-e29c-4f1d-a339-c65553861c58",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute GO enrichment for Louvain \n",
    "enrichment_louvain <- tibble::tibble(name=V(g)$name, louvain=V(g)$louvain) %>% group_by(louvain) %>%\n",
    "    \n",
    "    group_modify(~{\n",
    "        \n",
    "        enrichment <- string_db$get_enrichment(.x$name, category=\"Process\", methodMT=\"fdr\", iea=TRUE) %>% arrange(fdr)\n",
    "        if (is.null(enrichment) | nrow(enrichment)==0) {\n",
    "            \n",
    "            return(tibble::tibble())\n",
    "        \n",
    "        } else {\n",
    "\n",
    "            return(enrichment %>% arrange(fdr))\n",
    "            \n",
    "        }\n",
    "        \n",
    "    }\n",
    "                ) %>% ungroup()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc9c2b3e-6046-4bcf-b802-8f007742124d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fetch GO enrichment \n",
    "enrichment_label <- enrichment_louvain %>% group_by(louvain) %>% arrange(fdr) %>% slice(1) %>% ungroup() %>% transmute(louvain, label=description)\n",
    "V(g)$enrichment_label <- enrichment_label$label[match(V(g)$louvain, enrichment_label$louvain)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e01619a-6649-47b3-af2b-eea6f2632fdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove vertices without enrichment label \n",
    "g <- igraph::delete_vertices(g, which(is.na(V(g)$enrichment_label)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "352c2747-620c-44c0-a1aa-10f2607281fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot graph \n",
    "g_1 <- tidygraph::as_tbl_graph(g) %>% ggraph::ggraph(., layout=\"lgl\") +\n",
    "    geom_edge_link(alpha=0.2, width=0.1, color=\"#000000\") +\n",
    "    geom_node_point(aes(fill=factor(enrichment_label), size=abs(logFC)), color=\"black\", shape=21, stroke=0.1) +\n",
    "    # geom_node_text(aes(label=ifelse(gene %in% c(\"Slc40a1\", \"Slc48a1\", \"Nr1h3\", \"Hmox1\", \"Stat2\", \"Irf7\", \"Eif2ak2\", \"Gdf15\"), gene, NA)), repel=FALSE, size=2.1, color=\"black\") + \n",
    "    # scale_fill_manual(values=c(\"#6d6875\", \"#4f6d7a\", \"#4a6fa5\", \"#161b33\")) +\n",
    "    scale_size_continuous(range=c(0.5, 3)) + \n",
    "    theme(aspect.ratio=1) + theme_global_set(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a9827f7-33b1-4f20-a0cc-00447d1b29ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/string_cmo_d1_ko_down.pdf\", width=10, height=10)\n",
    "\n",
    "gridExtra::grid.arrange(egg::set_panel_size(g_1, width=unit(5, \"cm\"), height=unit(5, \"cm\")))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af97fa88-c2bc-4e13-a2e7-17c672d2c99a",
   "metadata": {},
   "source": [
    "## RPM D1 UP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1469ee3-e430-4785-b18e-5d00f2ddec23",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mapp DEA results to STRING DB\n",
    "string_db_dea <- res_dea_2[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[2]] %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, log2FoldChange, ihw) %>% dplyr::rename(logFC=log2FoldChange, padj=ihw) %>% na.omit() %>% dplyr::filter(padj<=adj_pval_thr & logFC>=+log2FC_thr) %>% string_db$map(., \"gene\", removeUnmappedRows=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fee0ad68-f3ac-41f0-a77a-ab170cb8b71b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Labe PPI vertices \n",
    "ppi_v <- string_db_dea %>% dplyr::distinct() %>% dplyr::rename(name=STRING_id) %>% \n",
    "    dplyr::group_by(name) %>% dplyr::mutate(n=n()) %>% dplyr::filter(n==1) %>% dplyr::select(name, gene, logFC, padj) %>% \n",
    "    dplyr::mutate(node_group=ifelse(padj<=0.05 & logFC<=-0.50, \"neg\", ifelse(padj<=0.05 & logFC>=0.50, \"pos\", \"n.s.\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e222ea7-0f5d-4850-a89e-be9eda7ef021",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create PPI edges from DEA PPI\n",
    "ppi_e <- string_db$get_interactions(string_db_dea$STRING_id) %>% dplyr::filter(from %in% ppi_v$name & to %in% ppi_v$name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "528e4f65-4ee6-4630-8696-1d71c753e16a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute igraph object\n",
    "g <- igraph::graph_from_data_frame(\n",
    "    \n",
    "    d=ppi_e,\n",
    "    vertices=ppi_v, \n",
    "    directed=FALSE\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74c4a734-7f32-4e55-8914-70d6bb11714d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Connected components \n",
    "min_component_size <- 4  # keep components with >= 4 vertices\n",
    "\n",
    "g_components <- igraph::components(g)\n",
    "g <- igraph::delete_vertices(g, which(g_components$csize[g_components$membership] < min_component_size))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b49394cf-e0ed-4768-8536-575244184e61",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute Louvain cluster on combined score \n",
    "E(g)$weight <- E(g)$combined_score\n",
    "\n",
    "louvain <- igraph::cluster_louvain(g, weights=E(g)$weight)\n",
    "V(g)$louvain <- membership(louvain)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1816e9e-59c6-4380-9e25-c0551726e4f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute GO enrichment for Louvain \n",
    "enrichment_louvain <- tibble::tibble(name=V(g)$name, louvain=V(g)$louvain) %>% group_by(louvain) %>%\n",
    "    \n",
    "    group_modify(~{\n",
    "        \n",
    "        enrichment <- string_db$get_enrichment(.x$name, category=\"Process\", methodMT=\"fdr\", iea=TRUE) %>% arrange(fdr)\n",
    "        if (is.null(enrichment) | nrow(enrichment)==0) {\n",
    "            \n",
    "            return(tibble::tibble())\n",
    "        \n",
    "        } else {\n",
    "\n",
    "            return(enrichment %>% arrange(fdr))\n",
    "            \n",
    "        }\n",
    "        \n",
    "    }\n",
    "                ) %>% ungroup()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac039034-85b3-49e2-8d27-56d631e8aca1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fetch GO enrichment \n",
    "enrichment_label <- enrichment_louvain %>% group_by(louvain) %>% arrange(fdr) %>% slice(1) %>% ungroup() %>% transmute(louvain, label=description)\n",
    "V(g)$enrichment_label <- enrichment_label$label[match(V(g)$louvain, enrichment_label$louvain)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc2fdda3-e1f6-43a3-b23f-431513a049ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove vertices without enrichment label \n",
    "g <- igraph::delete_vertices(g, which(is.na(V(g)$enrichment_label)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5814bcdd-8f1a-470a-b683-3598334a11c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot graph \n",
    "g_1 <- tidygraph::as_tbl_graph(g) %>% ggraph::ggraph(., layout=\"lgl\") +\n",
    "    geom_edge_link(alpha=0.2, width=0.1, color=\"#000000\") +\n",
    "    geom_node_point(aes(fill=factor(enrichment_label), size=abs(logFC)), color=\"black\", shape=21, stroke=0.1) +\n",
    "    # geom_node_text(aes(label=ifelse(gene %in% c(\"Slc40a1\", \"Slc48a1\", \"Nr1h3\", \"Hmox1\", \"Stat2\", \"Irf7\", \"Eif2ak2\", \"Gdf15\"), gene, NA)), repel=FALSE, size=2.1, color=\"black\") + \n",
    "    # scale_fill_manual(values=c(\"#6d6875\", \"#4f6d7a\", \"#4a6fa5\", \"#161b33\")) +\n",
    "    scale_size_continuous(range=c(0.5, 3)) + \n",
    "    theme(aspect.ratio=1) + theme_global_set(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f797d848-f44e-48ce-90a7-fb2afa12060c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/string_rpm_d1_ko_up.pdf\", width=10, height=10)\n",
    "\n",
    "gridExtra::grid.arrange(egg::set_panel_size(g_1, width=unit(5, \"cm\"), height=unit(5, \"cm\")))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "963585e7-9c7c-4924-94a5-13ff25800028",
   "metadata": {},
   "source": [
    "## RPM D1 UP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23cda913-66e3-45f0-bee1-6ee9b37b453e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mapp DEA results to STRING DB\n",
    "string_db_dea <- res_dea_2[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[2]] %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, log2FoldChange, ihw) %>% dplyr::rename(logFC=log2FoldChange, padj=ihw) %>% na.omit() %>% dplyr::filter(padj<=adj_pval_thr & logFC<=-log2FC_thr) %>% string_db$map(., \"gene\", removeUnmappedRows=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6205ab8-ee38-4dd8-b10a-dac18281fd34",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Labe PPI vertices \n",
    "ppi_v <- string_db_dea %>% dplyr::distinct() %>% dplyr::rename(name=STRING_id) %>% \n",
    "    dplyr::group_by(name) %>% dplyr::mutate(n=n()) %>% dplyr::filter(n==1) %>% dplyr::select(name, gene, logFC, padj) %>% \n",
    "    dplyr::mutate(node_group=ifelse(padj<=0.05 & logFC<=-0.50, \"neg\", ifelse(padj<=0.05 & logFC>=0.50, \"pos\", \"n.s.\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b158d4a-efb2-4df8-82ca-e2a525e67ed2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create PPI edges from DEA PPI\n",
    "ppi_e <- string_db$get_interactions(string_db_dea$STRING_id) %>% dplyr::filter(from %in% ppi_v$name & to %in% ppi_v$name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93b31c44-e4f6-44fe-b2ea-d1ddc8f37abd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute igraph object\n",
    "g <- igraph::graph_from_data_frame(\n",
    "    \n",
    "    d=ppi_e,\n",
    "    vertices=ppi_v, \n",
    "    directed=FALSE\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e69ebb93-e380-4c17-a388-aab24e7dab7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Connected components \n",
    "min_component_size <- 4  # keep components with >= 4 vertices\n",
    "\n",
    "g_components <- igraph::components(g)\n",
    "g <- igraph::delete_vertices(g, which(g_components$csize[g_components$membership] < min_component_size))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acbd8a0b-8e85-4304-ac68-1bb9376f4ba6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute Louvain cluster on combined score \n",
    "E(g)$weight <- E(g)$combined_score\n",
    "\n",
    "louvain <- igraph::cluster_louvain(g, weights=E(g)$weight)\n",
    "V(g)$louvain <- membership(louvain)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f583383b-bc59-4037-ada8-4b8f9ea82ca3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute GO enrichment for Louvain \n",
    "enrichment_louvain <- tibble::tibble(name=V(g)$name, louvain=V(g)$louvain) %>% group_by(louvain) %>%\n",
    "    \n",
    "    group_modify(~{\n",
    "        \n",
    "        enrichment <- string_db$get_enrichment(.x$name, category=\"Process\", methodMT=\"fdr\", iea=TRUE) %>% arrange(fdr)\n",
    "        if (is.null(enrichment) | nrow(enrichment)==0) {\n",
    "            \n",
    "            return(tibble::tibble())\n",
    "        \n",
    "        } else {\n",
    "\n",
    "            return(enrichment %>% arrange(fdr))\n",
    "            \n",
    "        }\n",
    "        \n",
    "    }\n",
    "                ) %>% ungroup()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc453160-b5f0-47d0-a569-45a368507ce9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fetch GO enrichment \n",
    "enrichment_label <- enrichment_louvain %>% group_by(louvain) %>% arrange(fdr) %>% slice(1) %>% ungroup() %>% transmute(louvain, label=description)\n",
    "V(g)$enrichment_label <- enrichment_label$label[match(V(g)$louvain, enrichment_label$louvain)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5693fadf-8324-4246-8c01-6d09bb8510aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove vertices without enrichment label \n",
    "g <- igraph::delete_vertices(g, which(is.na(V(g)$enrichment_label)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcbc00b4-5293-408c-b013-ef7c71e13682",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot graph \n",
    "g_1 <- tidygraph::as_tbl_graph(g) %>% ggraph::ggraph(., layout=\"lgl\") +\n",
    "    geom_edge_link(alpha=0.2, width=0.1, color=\"#000000\") +\n",
    "    geom_node_point(aes(fill=factor(enrichment_label), size=abs(logFC)), color=\"black\", shape=21, stroke=0.1) +\n",
    "    # geom_node_text(aes(label=ifelse(gene %in% c(\"Slc40a1\", \"Slc48a1\", \"Nr1h3\", \"Hmox1\", \"Stat2\", \"Irf7\", \"Eif2ak2\", \"Gdf15\"), gene, NA)), repel=FALSE, size=2.1, color=\"black\") + \n",
    "    # scale_fill_manual(values=c(\"#6d6875\", \"#4f6d7a\", \"#4a6fa5\", \"#161b33\")) +\n",
    "    scale_size_continuous(range=c(0.5, 3)) + \n",
    "    theme(aspect.ratio=1) + theme_global_set(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a274c850-a9e8-4caa-bb4d-0af1926f72ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/string_rpm_d1_ko_down.pdf\", width=10, height=10)\n",
    "\n",
    "gridExtra::grid.arrange(egg::set_panel_size(g_1, width=unit(5, \"cm\"), height=unit(5, \"cm\")))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e3a6a2b-eabc-4236-98bc-811bac2d5698",
   "metadata": {},
   "source": [
    "# Session Info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfa8ea83-1199-4f04-94e6-646773902ff7",
   "metadata": {},
   "outputs": [],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dd11ad7-2143-4080-97cf-4e086e3dc3f9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R [conda env:r.4.4.1]",
   "language": "R",
   "name": "conda-env-r.4.4.1-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
