{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8dab95dd-2e8e-472e-b1ef-e9566cd08bfb",
   "metadata": {},
   "source": [
    "# Panel 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75fd7b70-81de-41ed-b00b-54dd26951307",
   "metadata": {},
   "outputs": [],
   "source": [
    "library_load <- suppressMessages(\n",
    "    \n",
    "    suppressWarnings(\n",
    "        \n",
    "        list(\n",
    "        \n",
    "            # Seurat \n",
    "            library(Seurat), \n",
    "            \n",
    "            # Data\n",
    "            library(dplyr), \n",
    "            \n",
    "            # Plotting \n",
    "            library(ggplot2), \n",
    "            library(patchwork), \n",
    "            library(ComplexHeatmap), \n",
    "            library(grid) # For gpar\n",
    "            \n",
    "        )\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7ee4d36-c8a2-4cf8-a1a0-bfa6233f4b12",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c965fc68-e06b-404d-b9b7-6da6d498a9f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set working directory to project root\n",
    "setwd(\"/research/peer/fdeckert/FD20200109SPLENO\") # You need to adjust this path. I will help. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "929a6504-28cd-45e6-ad0d-cf4f6ca2c6f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Source\n",
    "source(\"bin/so_pl.R\")\n",
    "source(\"bin/dea_pl.R\")\n",
    "source(\"bin/gsea_pp.R\")\n",
    "source(\"bin/gsea_pl.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d5d13bf-d70f-405b-b7b8-a96b73cbd542",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting Theme\n",
    "source(\"plotting_global.R\")\n",
    "ggplot2::theme_set(theme_global_set(size_select=1)) # From project global source()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6690aad8-9f43-4e72-8fee-c36424e70168",
   "metadata": {},
   "source": [
    "# Import and parameter "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b3435a9-39ba-4711-b42c-31686f6e473e",
   "metadata": {},
   "outputs": [],
   "source": [
    "singler <- read.csv(\"result/singler/bulkRNAseq/singler_res.csv\", row.names=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef27618a-2747-4373-93a8-42f2ea2a872c",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- readRDS(\"data/scRNAseq/object/pp_0.rds\")\n",
    "so <- AddMetaData(so, singler)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc0789c6-4eaf-4451-b289-97a2870cd4fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- NormalizeData(so)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1df8b657-62e9-4069-bca8-2b5c77cf0ae0",
   "metadata": {},
   "source": [
    "# SingleR UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7cee7d3-91b8-4ce3-b444-ee6cb1025b4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "fp_1 <- list(\n",
    "\n",
    "    fplot(so, features=\"cmo_score\", pt_size=0.25, min_cutoff=0, order=TRUE, restrict=c(\"sample_group\", \"Bl6_NaCl_D6\"), size_select=4), \n",
    "    fplot(so, features=\"cmo_score\", pt_size=0.25, min_cutoff=0, order=TRUE, restrict=c(\"sample_group\", \"Bl6_CpG_D6\"), size_select=4), \n",
    "    fplot(so, features=\"rpm_score\", pt_size=0.25, min_cutoff=0, order=TRUE, restrict=c(\"sample_group\", \"Bl6_NaCl_D6\"), size_select=4), \n",
    "    fplot(so, features=\"rpm_score\", pt_size=0.25, min_cutoff=0, order=TRUE, restrict=c(\"sample_group\", \"Bl6_CpG_D6\"), size_select=4), \n",
    "    fplot(so, features=\"Spic\", pt_size=0.25, min_cutoff=0, order=TRUE, slot=\"data\", assay=\"RNA\", restrict=c(\"sample_group\", \"Bl6_NaCl_D6\"), size_select=4), \n",
    "    fplot(so, features=\"Spic\", pt_size=0.25, min_cutoff=0, order=TRUE, slot=\"data\", assay=\"RNA\", restrict=c(\"sample_group\", \"Bl6_CpG_D6\"), size_select=4), \n",
    "    fplot(so, features=\"Gdf15\", pt_size=0.25, min_cutoff=0, order=TRUE, slot=\"data\", assay=\"RNA\", restrict=c(\"sample_group\", \"Bl6_NaCl_D6\"), size_select=4), \n",
    "    fplot(so, features=\"Gdf15\", pt_size=0.25, min_cutoff=0, order=TRUE, slot=\"data\", assay=\"RNA\", restrict=c(\"sample_group\", \"Bl6_CpG_D6\"), size_select=4)\n",
    "\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1a84498-163f-4dca-8570-11438db87b83",
   "metadata": {},
   "outputs": [],
   "source": [
    "fp_1 <- lapply(fp_1, function(p) egg::set_panel_size(p, width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")))\n",
    "fp_1 <- do.call(gridExtra::arrangeGrob, c(fp_1, ncol=2, nrow=length(fp_1)/2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "032de9cc-bd32-451e-b153-41d2b7eb7710",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/fp_rpm_score.pdf\", width=2*2.0, height=1.5*length(fp_1)/2)\n",
    "\n",
    "grid::grid.draw(fp_1)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a105560-61e6-488b-bf05-a2ef56cc54f2",
   "metadata": {},
   "source": [
    "# DEA result count heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48a7ef59-2c1e-40bf-8304-e91a9480306f",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_dea_1 <- readRDS(\"result/dea/bulkRNAseq/res_dea_cmo.rds\")\n",
    "res_dea_2 <- readRDS(\"result/dea/bulkRNAseq/res_dea_rpm.rds\")\n",
    "res_dea_3 <- readRDS(\"result/dea/bulkRNAseq/res_dea_ery.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "447b4c5a-6bbf-46af-9fa9-e8612f9211bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "genotype_wald_1 <- purrr::reduce(\n",
    "    \n",
    "    list(\n",
    "        \n",
    "        res_dea_1[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[1]] %>% dplyr::mutate(Ctl=sum((ihw<=0.05 & abs(log2FoldChange)>=0.25))) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, Ctl), \n",
    "        res_dea_1[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[2]] %>% dplyr::mutate(D1=sum((ihw<=0.05 & abs(log2FoldChange)>=0.25))) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, D1), \n",
    "        res_dea_1[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[3]] %>% dplyr::mutate(D3=sum((ihw<=0.05 & abs(log2FoldChange)>=0.25))) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, D3), \n",
    "        res_dea_1[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[4]] %>% dplyr::mutate(D6=sum((ihw<=0.05 & abs(log2FoldChange)>=0.25))) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, D6)\n",
    "    \n",
    "    ), full_join, by=\"gene\"\n",
    "\n",
    ") %>% dplyr::select(-gene) %>% dplyr::distinct()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df6746c9-8144-45df-8266-29ae56ed73df",
   "metadata": {},
   "outputs": [],
   "source": [
    "genotype_wald_2 <- purrr::reduce(\n",
    "    \n",
    "    list(\n",
    "        \n",
    "        res_dea_2[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[1]] %>% dplyr::mutate(Ctl=sum((ihw<=0.05 & abs(log2FoldChange)>=0.25))) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, Ctl), \n",
    "        res_dea_2[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[2]] %>% dplyr::mutate(D1=sum((ihw<=0.05 & abs(log2FoldChange)>=0.25))) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, D1), \n",
    "        res_dea_2[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[3]] %>% dplyr::mutate(D3=sum((ihw<=0.05 & abs(log2FoldChange)>=0.25))) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, D3), \n",
    "        res_dea_2[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[4]] %>% dplyr::mutate(D6=sum((ihw<=0.05 & abs(log2FoldChange)>=0.25))) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, D6)\n",
    "    \n",
    "    ), full_join, by=\"gene\"\n",
    "\n",
    ") %>% dplyr::select(-gene) %>% dplyr::distinct()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a84bc3c-552f-4711-8d80-2a36856637f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "genotype_wald_3 <- purrr::reduce(\n",
    "    \n",
    "    list(\n",
    "        \n",
    "        res_dea_3[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[1]] %>% dplyr::mutate(Ctl=sum((ihw<=0.05 & abs(log2FoldChange)>=0.25))) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, Ctl), \n",
    "        res_dea_3[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[2]] %>% dplyr::mutate(D1=sum((ihw<=0.05 & abs(log2FoldChange)>=0.25))) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, D1), \n",
    "        res_dea_3[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[3]] %>% dplyr::mutate(D3=sum((ihw<=0.05 & abs(log2FoldChange)>=0.25))) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, D3), \n",
    "        res_dea_3[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]][[4]] %>% dplyr::mutate(D6=sum((ihw<=0.05 & abs(log2FoldChange)>=0.25))) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, D6)\n",
    "    \n",
    "    ), full_join, by=\"gene\"\n",
    "\n",
    ") %>% dplyr::select(-gene) %>% dplyr::distinct()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ccccc7d-71ec-484f-9517-2114b54bd549",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- rbind(genotype_wald_1, genotype_wald_2, genotype_wald_3)\n",
    "rownames(mat_1) <- c(\"cMo\", \"RPM\", \"EB\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c760c29-f96c-42d9-aeda-335892122cad",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/hm_dea_count.pdf\", width=1.5, height=3)\n",
    "\n",
    "draw(dea_count_hm(mat_1, fontsize_select=2))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb5fc0a4-9a52-483d-82aa-6af27fb41773",
   "metadata": {},
   "source": [
    "# Gene set enrichment analysis (GSEA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c22adf09-f8e1-4184-9f65-49123f62d7f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "adj_pval_thr <- 0.05\n",
    "log2FC_thr <- 0.25"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb2fd498-3202-4cff-93d8-82d1dadaad05",
   "metadata": {},
   "source": [
    "## Prepare results from genotype wald test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5b3b640-fa9c-4a1f-b2b1-590b26c16059",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_dea_1 <- readRDS(\"result/dea/bulkRNAseq/res_dea_cmo.rds\")\n",
    "res_dea_2 <- readRDS(\"result/dea/bulkRNAseq/res_dea_rpm.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "084d7a9b-c062-424c-ab34-31efc1ef760e",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_dea <- list(\"cMo\"=res_dea_1[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]], \"RPM\"=res_dea_2[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4a6485f-14a6-45da-8ca0-16dc34f51304",
   "metadata": {},
   "source": [
    "## genotype comparison - HM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24d7c671-a5ac-4138-9bfc-ffeb900b678a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res_1 <- lapply(seq_along(res_dea[[1]]), function(i) gsea(res_dea[[1]][[i]] %>% dplyr::rename(avg_log2FC=log2FoldChange, p_val_adj=ihw), category=\"MH\", subcategory=NULL) %>% dplyr::mutate(col_split=\"cMo\", col_label=paste0(\"group_\", i)))\n",
    "gsea_res_2 <- lapply(seq_along(res_dea[[2]]), function(i) gsea(res_dea[[2]][[i]] %>% dplyr::rename(avg_log2FC=log2FoldChange, p_val_adj=ihw), category=\"MH\", subcategory=NULL) %>% dplyr::mutate(col_split=\"RPM\", col_label=paste0(\"group_\", i)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "864f7770-3c81-4433-8312-4143d86e056b",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res_1 <- do.call(\"rbind\", gsea_res_1)\n",
    "gsea_res_2 <- do.call(\"rbind\", gsea_res_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82ee3541-8136-4e92-9a49-58062ec4d213",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res <- rbind(gsea_res_1, gsea_res_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9535590-44cb-4df7-bdba-75af1de3bb8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs_name_select <- gsea_res %>% dplyr::filter(padj<=0.05 & abs(NES)>=0) %>% dplyr::pull(pathway) %>% unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3600b30-75ec-427e-9f41-fad546db85e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- gsea_res %>% dplyr::filter(pathway %in% gs_name_select) %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=padj,\n",
    "        values_fill=1\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c60e5c0-425a-4ea1-85c9-7ad02adac45b",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_2 <- gsea_res %>% dplyr::filter(pathway %in% gs_name_select) %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=NES,\n",
    "        values_fill=0\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db14b5e9-16ca-436e-b72e-d1522f0c3e22",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- mat_1[, c(\"col_label\", \"col_split\", gs_name_select)]\n",
    "mat_2 <- mat_2[, c(\"col_label\", \"col_split\", gs_name_select)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49d7a1a4-b2a2-4bfd-9c59-09d124917a27",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_label <- mat_1$col_label\n",
    "col_split <- mat_1$col_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f5e626c-902c-43a0-a56d-f59a6d7a84f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- mat_1 %>% dplyr::select(-col_label, -col_split) %>% t()\n",
    "mat_2 <- mat_2 %>% dplyr::select(-col_label, -col_split) %>% t()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93c1afaa-77cf-404c-9e0b-cf0fea5cb109",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/hm_gsea_genotype_dpi_mh.pdf\", width=8, height=10)\n",
    "\n",
    "gsea_hm(mat_1, mat_2, col_label=col_label, col_split=col_split, row_split=NULL, pathway_suffix=\"HALLMARK\", color_neg=color$sample_group[1], color_pos=color$sample_group[2], fontsize_select=2)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f50c094-a064-42e8-a7a2-b5c7a0312863",
   "metadata": {},
   "source": [
    "# DEA heatmap "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58b354ff-326a-4cdc-a374-8d04978d80b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "adj_pval_thr <- 0.05\n",
    "log2FC_thr <- 0.25"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "082888be-3f98-43ac-bfde-45bfa1250703",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_dea_fetch <- function(mat_1, dea_res, col_samples, col_split, fontsize_select, celltype, celltype_select, scale_mode=\"all\") {\n",
    "\n",
    "    mat_1 <- Seurat::CreateSeuratObject(counts=mat_1)\n",
    "    mat_1 <- NormalizeData(mat_1, normalization.method=\"RC\", scale.factor=1e6)\n",
    "    mat_1 <- GetAssayData(mat_1, assay=\"RNA\", layer=\"data\") %>% as.data.frame() \n",
    "\n",
    "    if(scale_mode==\"all\") {\n",
    "\n",
    "        mat_1 <- t(apply(mat_1, 1, scales::rescale))\n",
    "        mat_1 <- mat_1[row_split, col_samples]\n",
    "\n",
    "        colnames(mat_1) <- col_label\n",
    "\n",
    "        row_split <- setNames(unlist(row_genes, use.names=FALSE), rep(names(row_genes), sapply(row_genes, length)))\n",
    "        row_split <- row_split[row_split %in% rownames(mat_1)]\n",
    "        \n",
    "        mat_1 <- mat_1[row_split, celltype==celltype_select]\n",
    "        col_split <- col_split[celltype==celltype_select]\n",
    "        \n",
    "    } else if (scale_mode==\"sample\") {\n",
    "\n",
    "        mat_1 <- mat_1[row_split, col_samples]\n",
    "        mat_1 <- mat_1[row_split, col_samples]\n",
    "\n",
    "        colnames(mat_1) <- col_label\n",
    "\n",
    "        row_split <- setNames(unlist(row_genes, use.names=FALSE), rep(names(row_genes), sapply(row_genes, length)))\n",
    "        row_split <- row_split[row_split %in% rownames(mat_1)]\n",
    "        \n",
    "        mat_1 <- mat_1[row_split, celltype==celltype_select]\n",
    "        col_split <- col_split[celltype==celltype_select]\n",
    "        \n",
    "        mat_1 <- t(apply(mat_1, 1, scales::rescale))\n",
    "        \n",
    "    }\n",
    "    \n",
    "    # DEA log2FC \n",
    "    mat_2 <- purrr::reduce(\n",
    "        \n",
    "        list(\n",
    "            \n",
    "            res_dea[[celltype_select]][[1]] %>% dplyr::mutate(Ctl=log2FoldChange) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, Ctl), \n",
    "            res_dea[[celltype_select]][[2]] %>% dplyr::mutate(D1=log2FoldChange) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, D1), \n",
    "            res_dea[[celltype_select]][[3]] %>% dplyr::mutate(D3=log2FoldChange) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, D3), \n",
    "            res_dea[[celltype_select]][[4]] %>% dplyr::mutate(D6=log2FoldChange) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, D6)\n",
    "        \n",
    "        ), full_join, by=\"gene\"\n",
    "    \n",
    "    ) %>% tibble::column_to_rownames(\"gene\")\n",
    "\n",
    "    mat_2[setdiff(row_split, rownames(mat_2)), ] <- 0\n",
    "    mat_2 <- mat_2[row_split, ]\n",
    "    mat_2[is.na(mat_2)] <- 0\n",
    "\n",
    "    # DEA adj p-values\n",
    "    mat_3 <- purrr::reduce(\n",
    "        \n",
    "        list(\n",
    "            \n",
    "            res_dea[[celltype_select]][[1]] %>% dplyr::mutate(Ctl=padj) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, Ctl), \n",
    "            res_dea[[celltype_select]][[2]] %>% dplyr::mutate(D1=padj) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, D1), \n",
    "            res_dea[[celltype_select]][[3]] %>% dplyr::mutate(D3=padj) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, D3), \n",
    "            res_dea[[celltype_select]][[4]] %>% dplyr::mutate(D6=padj) %>% tibble::rownames_to_column(\"gene\") %>% dplyr::select(gene, D6)\n",
    "        \n",
    "        ), full_join, by=\"gene\"\n",
    "    \n",
    "    ) %>% tibble::column_to_rownames(\"gene\")\n",
    "\n",
    "    mat_3[setdiff(row_split, rownames(mat_3)), ] <- 1\n",
    "    mat_3 <- mat_3[row_split, ]\n",
    "    mat_3[is.na(mat_3)] <- 1\n",
    "    mat_3[(mat_3>adj_pval_thr & abs(mat_2)<log2FC_thr)] <- 1\n",
    "\n",
    "    # Set log2FC to 0 if not significant \n",
    "    mat_2[abs(mat_2) < log2FC_thr] <- 0\n",
    "\n",
    "    # Row split\n",
    "    row_split=names(row_split)\n",
    "    row_split=factor(row_split, levels=names(row_genes))\n",
    "    \n",
    "    hm_1 <- dea_exp_hm(mat_1, col_split=col_split, row_split=row_split, fontsize_select=2, breaks_limit=2, scale_width=0.5, legend_title=\"Exp.\", cluster_rows=FALSE)\n",
    "    hm_2 <- dea_res_hm(mat_2, mat_3, p_val_adj_thr=adj_pval_thr, row_split=row_split, scale_width=1, color_min=\"#cd34b5\", color_max=\"#FFAC1E\", cluster_rows=FALSE, fontsize_select=2)\n",
    "\n",
    "    hm <- hm_2 + hm_1\n",
    "\n",
    "    return(hm)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19d0a61c-470d-4c21-b611-efcfac0be901",
   "metadata": {},
   "source": [
    "# Prepare meta data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56fad3ad-74ed-4e35-a5d5-8dc3a6ada04c",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta <- readRDS(\"data/bulkRNAseq/object/meta.rds\") %>% dplyr::filter(celltype %in% c(\"cMo\", \"RPM\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "996c8765-b837-4ecd-bd7f-dff628f75465",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta <- meta %>% dplyr::arrange(celltype, dpi, genotype)\n",
    "\n",
    "col_samples <- as.character(meta$sample_name)\n",
    "col_label <- as.character(meta$genotype)\n",
    "col_split <- as.character(meta$dpi)\n",
    "celltype <- as.character(meta$celltype)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43321fd0-d20d-4ab5-b67c-e70e90b1e2da",
   "metadata": {},
   "source": [
    "## Prepare results from genotype wald test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ca0e799-fce5-4809-a936-da81b86f63a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_dea_1 <- readRDS(\"result/dea/bulkRNAseq/res_dea_cmo.rds\")\n",
    "res_dea_2 <- readRDS(\"result/dea/bulkRNAseq/res_dea_rpm.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c774fab-50b8-4885-8fca-8e80bb57faa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] <- res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::mutate(response_class_wt=ifelse(response_gene_wt, response_class_wt, NA), response_class_ko=ifelse(response_gene_ko, response_class_ko, NA))\n",
    "res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] <- res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::mutate(response_class_wt=ifelse(response_gene_wt, response_class_wt, NA), response_class_ko=ifelse(response_gene_ko, response_class_ko, NA))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2904a8a2-ccc5-4541-9b86-d99a53250086",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] <- res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter((response_gene_wt | response_gene_ko) & perturbation_gene) %>% dplyr::filter(!(is.na(response_class_wt) & is.na(response_class_ko)))\n",
    "res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] <- res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter((response_gene_wt | response_gene_ko) & perturbation_gene) %>% dplyr::filter(!(is.na(response_class_wt) & is.na(response_class_ko)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f620af1d-f8a6-413b-ae48-6cbba521b5d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_dea <- list(\"cMo\"=res_dea_1[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]], \"RPM\"=res_dea_2[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dada08c7-e237-4ed7-9db8-7486b60f2aa8",
   "metadata": {},
   "source": [
    "## Prepare z-score mat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca71f644-1d5b-49d0-985a-db04a3aae74a",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta <- readRDS(\"data/bulkRNAseq/object/meta.rds\") %>% dplyr::filter(celltype %in% c(\"cMo\", \"RPM\")) %>% droplevels()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9b3f3a4-6ae5-4c1e-8355-4022cfec4a93",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- readRDS(\"data/bulkRNAseq/object/cnt.rds\") %>% dplyr::select(-mgi_symbol, -ensembl_gene_id)\n",
    "mat_1 <- mat_1[, rownames(meta)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62b2c9e5-4257-4129-acf0-e6cacb6c9b2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "all(rownames(meta)==colnames(mat_1))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "496cf99e-ad96-482a-95c0-4f9046615b98",
   "metadata": {},
   "source": [
    "## cMo cre/+ UP vs +/+"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb4bd382-03ac-4b72-8916-6724cb8cb530",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_select <- lapply(res_dea_1[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]], function(x) x %>% dplyr::filter(ihw<=adj_pval_thr & log2FoldChange>=log2FC_thr) %>% rownames())\n",
    "gene_select <- do.call(\"c\", gene_select) %>% unique()\n",
    "gene_select <- gene_select[!gene_select %in% grep(\"Gm|Rps|Rpl\", gene_select, value=TRUE)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fadc3113-a073-46e0-9f6b-ee0af62e17eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_genes <- list(\n",
    "\n",
    "    \"Repressed\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Repressed\" & response_class_ko==\"Repressed\") %>% rownames(), \n",
    "    \"Activated\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Activated\" & response_class_ko==\"Activated\") %>% rownames(), \n",
    "    \"Activated (+/+)\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Activated\" & is.na(response_class_ko)) %>% rownames(), \n",
    "    \"Activated (cre/+)\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(is.na(response_class_wt) & response_class_ko==\"Activated\") %>% rownames(), \n",
    "    \"Repressed (+/+)\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Repressed\" & is.na(response_class_ko)) %>% rownames(), \n",
    "    \"Repressed (cre/+)\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(is.na(response_class_wt) & response_class_ko==\"Repressed\") %>% rownames(),\n",
    "    \"Opposed\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt!=response_class_ko) %>% rownames()\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80472b28-ec90-44cf-aeb3-b52b4c6ee101",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_genes <- lapply(row_genes, function(x) x[x %in% gene_select])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bedc430-5f60-40fa-930f-361650f6aae5",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_split <- setNames(unlist(row_genes, use.names=FALSE), rep(names(row_genes), sapply(row_genes, length)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ecc30ca-c191-4c38-9272-54c199a78012",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_1 <- hm_dea_fetch(mat_1, dea_res, col_samples, col_split, fontsize_select=2, celltype, celltype_select=\"cMo\", scale_mode=\"sample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41a57f3c-3056-4146-85e5-d133a651d539",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/hm_dea_ko_up_vs_wt_cmo.pdf\", width=10, height=70)\n",
    "\n",
    "hm_1\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3432d272-f3c9-4c4e-af38-e9074ed23e57",
   "metadata": {},
   "source": [
    "## cMo cre/+ DOWN vs +/+"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2574e706-00af-4b9b-9d76-5fb933b368a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_select <- lapply(res_dea_1[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]], function(x) x %>% dplyr::filter(ihw<=adj_pval_thr & log2FoldChange<=-log2FC_thr) %>% rownames())\n",
    "gene_select <- do.call(\"c\", gene_select) %>% unique()\n",
    "gene_select <- gene_select[!gene_select %in% grep(\"Gm|Rps|Rpl\", gene_select, value=TRUE)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9c4e563-44af-46fa-8acf-0b72f4785f61",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_genes <- list(\n",
    "\n",
    "    \"Repressed\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Repressed\" & response_class_ko==\"Repressed\") %>% rownames(), \n",
    "    \"Activated\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Activated\" & response_class_ko==\"Activated\") %>% rownames(), \n",
    "    \"Activated (+/+)\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Activated\" & is.na(response_class_ko)) %>% rownames(), \n",
    "    \"Activated (cre/+)\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(is.na(response_class_wt) & response_class_ko==\"Activated\") %>% rownames(), \n",
    "    \"Repressed (+/+)\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Repressed\" & is.na(response_class_ko)) %>% rownames(), \n",
    "    \"Repressed (cre/+)\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(is.na(response_class_wt) & response_class_ko==\"Repressed\") %>% rownames(),\n",
    "    \"Opposed\"=res_dea_1[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt!=response_class_ko) %>% rownames()\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac9f2f6d-97b6-4fb9-9a51-e391e8f383f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_genes <- lapply(row_genes, function(x) x[x %in% gene_select])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cfb8d8e-2ee2-4ff1-902d-66839e4ea94e",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_split <- setNames(unlist(row_genes, use.names=FALSE), rep(names(row_genes), sapply(row_genes, length)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cbb7804-cb08-4558-b7b8-6a352aeab896",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_1 <- hm_dea_fetch(mat_1, dea_res, col_samples, col_split, fontsize_select=2, celltype, celltype_select=\"cMo\", scale_mode=\"sample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42547209-9685-411d-a7fc-f9d971a16f85",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/hm_dea_ko_down_vs_wt_cmo.pdf\", width=10, height=80)\n",
    "\n",
    "hm_1\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9253b434-f1e1-4780-9331-f188a65075e2",
   "metadata": {},
   "source": [
    "## RPM cre/+ UP vs +/+"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "432b7646-24a2-4ea7-86a9-1374dacb614d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_select <- lapply(res_dea_2[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]], function(x) x %>% dplyr::filter(ihw<=adj_pval_thr & log2FoldChange>=log2FC_thr) %>% rownames())\n",
    "gene_select <- do.call(\"c\", gene_select) %>% unique()\n",
    "gene_select <- gene_select[!gene_select %in% grep(\"Gm|Rps|Rpl\", gene_select, value=TRUE)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8b5cbd7-84f7-443e-8553-1a3c1d8591b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_genes <- list(\n",
    "\n",
    "    \"Repressed\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Repressed\" & response_class_ko==\"Repressed\") %>% rownames(), \n",
    "    \"Activated\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Activated\" & response_class_ko==\"Activated\") %>% rownames(), \n",
    "    \"Activated (+/+)\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Activated\" & is.na(response_class_ko)) %>% rownames(), \n",
    "    \"Activated (cre/+)\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(is.na(response_class_wt) & response_class_ko==\"Activated\") %>% rownames(), \n",
    "    \"Repressed (+/+)\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Repressed\" & is.na(response_class_ko)) %>% rownames(), \n",
    "    \"Repressed (cre/+)\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(is.na(response_class_wt) & response_class_ko==\"Repressed\") %>% rownames(),\n",
    "    \"Opposed\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt!=response_class_ko) %>% rownames()\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e655dacc-6783-4f97-8e01-1fbe59cfac20",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_genes <- lapply(row_genes, function(x) x[x %in% gene_select])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0b75fd7-0b5a-4288-9c8c-7f1ac3d379cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_split <- setNames(unlist(row_genes, use.names=FALSE), rep(names(row_genes), sapply(row_genes, length)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab05c0df-8e6e-4a42-a884-bee5355c9634",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_1 <- hm_dea_fetch(mat_1, dea_res, col_samples, col_split, fontsize_select=2, celltype, celltype_select=\"RPM\", scale_mode=\"sample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0b6754a-d5f7-48cb-948d-f295d633acc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/hm_dea_ko_up_vs_wt_rpm.pdf\", width=10, height=30)\n",
    "\n",
    "hm_1\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93a8414c-e9e8-4f70-a0b5-066bdc5904d9",
   "metadata": {},
   "source": [
    "## RPM cre/+ DOWN vs +/+"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a5ff6d0-23a5-459c-a1e9-9c2f853c6434",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_select <- lapply(res_dea_2[[\"res\"]][[\"dpi_wt_vs_ko_wald\"]], function(x) x %>% dplyr::filter(ihw<=adj_pval_thr & log2FoldChange<=-log2FC_thr) %>% rownames())\n",
    "gene_select <- do.call(\"c\", gene_select) %>% unique()\n",
    "gene_select <- gene_select[!gene_select %in% grep(\"Gm|Rps|Rpl\", gene_select, value=TRUE)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7448988b-ed1b-4e78-893d-56edbfbc872a",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_genes <- list(\n",
    "\n",
    "    \"Repressed\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Repressed\" & response_class_ko==\"Repressed\") %>% rownames(), \n",
    "    \"Activated\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Activated\" & response_class_ko==\"Activated\") %>% rownames(), \n",
    "    \"Activated (+/+)\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Activated\" & is.na(response_class_ko)) %>% rownames(), \n",
    "    \"Activated (cre/+)\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(is.na(response_class_wt) & response_class_ko==\"Activated\") %>% rownames(), \n",
    "    \"Repressed (+/+)\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt==\"Repressed\" & is.na(response_class_ko)) %>% rownames(), \n",
    "    \"Repressed (cre/+)\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(is.na(response_class_wt) & response_class_ko==\"Repressed\") %>% rownames(),\n",
    "    \"Opposed\"=res_dea_2[[\"resc\"]][[\"pc_wt_vs_ko\"]] %>% dplyr::filter(response_class_wt!=response_class_ko) %>% rownames()\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ac46025-38dd-4916-a3c6-16e6c9096116",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_genes <- lapply(row_genes, function(x) x[x %in% gene_select])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff2b842d-adea-4723-9f48-a92a2a126130",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_split <- setNames(unlist(row_genes, use.names=FALSE), rep(names(row_genes), sapply(row_genes, length)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8965cc0c-10c9-4cde-b067-6b97675f8760",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_1 <- hm_dea_fetch(mat_1, dea_res, col_samples, col_split, fontsize_select=2, celltype, celltype_select=\"RPM\", scale_mode=\"sample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7ba094b-e251-49f7-b90a-edd84166471c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/hm_dea_ko_down_vs_wt_rpm.pdf\", width=10, height=25)\n",
    "\n",
    "hm_1\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9bb614d-9663-4dfb-8e0f-e7f9c79dbdf8",
   "metadata": {},
   "source": [
    "# Selected genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 638,
   "id": "bbd9da69-4c7e-406b-b079-6db9f31013f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_genes <- list(\n",
    "\n",
    "    \"G1\"=c(\"Ccr2\", \"Ly6c2\", \"Spic\", \"Cd68\", \"Cd163\"), \n",
    "    \"G2\"=c(\"Vcam1\", \"Itga9\"), \n",
    "    \"G3\"=c(\"Ccl24\", \"Kitl\", \"Gdf15\", \"Il1b\", \"Il18bp\", \"Il10\", \"Ccl2\", \"Ccl12\", \"Ccl4\", \"Cxcl10\"), \n",
    "    \"G4\"=c(\"Slc40a1\", \"Slc48a1\", \"Ftl1\", \"Fth1\", \"Hmox1\", \"Hmox2\", \"Nfe2l2\", \"Acsl4\", \"Sod2\", \"Prdx1\"), \n",
    "    \"G5\"=c(\"Stat1\", \"Stat2\", \"Irf9\", \"Irf7\", \"Ifih1\", \"Isg15\", \"Isg20\", \"Ifi44\", \"Ifit1\", \"Ifit2\", \"Ifit3\", \"Ifitm1\", \"Ifitm3\", \"Usp18\", \"Mx1\", \"Oasl1\", \"Oasl2\", \"Ddx60\", \"Cd274\")\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 639,
   "id": "3ebddaf2-4af4-4798-81bb-985a6a95ab98",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [],
      "text/latex": [],
      "text/markdown": [],
      "text/plain": [
       "named character(0)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "row_split <- setNames(unlist(row_genes, use.names=FALSE), rep(names(row_genes), sapply(row_genes, length)))\n",
    "row_split[!row_split %in% rownames(mat_1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 640,
   "id": "665c532f-1af4-4fe7-a3a2-a7b4a2b619b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Normalizing layer: counts\n",
      "\n",
      "Normalizing layer: counts\n",
      "\n"
     ]
    }
   ],
   "source": [
    "hm_1 <- hm_dea_fetch(mat_1, dea_res, col_samples, col_split, fontsize_select=2, celltype, celltype_select=\"cMo\", scale_mode=\"all\")\n",
    "hm_2 <- hm_dea_fetch(mat_1, dea_res, col_samples, col_split, fontsize_select=2, celltype, celltype_select=\"RPM\", scale_mode=\"all\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 641,
   "id": "85a88ad3-a481-4df3-8439-0bad4c6a376d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"result/figures/2_bulkRNAseq/panel_1/hm_dea_cmo_rpm_selected.pdf\", width=10, height=10)\n",
    "\n",
    "hm_1 + hm_2\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e3a6a2b-eabc-4236-98bc-811bac2d5698",
   "metadata": {},
   "source": [
    "# Session Info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfa8ea83-1199-4f04-94e6-646773902ff7",
   "metadata": {},
   "outputs": [],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "r.4.4.1-FD20200109SPLENO",
   "language": "R",
   "name": "r.4.4.1-fd20200109spleno"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
