{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bb817fcf-c6d9-49ee-9dda-70412056f1bc",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Wilcox - Differential expression analysis (DEA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5b2b347-8534-452f-9e7f-f59af940c6a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61a8e980-2d54-4ddd-9cd7-ab47446e5911",
   "metadata": {},
   "outputs": [],
   "source": [
    "library_load <- suppressMessages(\n",
    "    \n",
    "    list(\n",
    "        \n",
    "        # Seurat \n",
    "        library(Seurat), \n",
    "        library(SeuratWrappers), \n",
    "    \n",
    "        # IHW \n",
    "        library(IHW), \n",
    "        \n",
    "        # Data \n",
    "        library(tidyverse), \n",
    "        library(openxlsx), \n",
    "        \n",
    "        # Plotting \n",
    "        library(ComplexHeatmap), \n",
    "        library(circlize), \n",
    "        library(viridis), \n",
    "        library(ggplotify), \n",
    "        library(patchwork), \n",
    "        \n",
    "        # Python \n",
    "        library(reticulate)\n",
    "        \n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70ad0e86-1004-4429-a175-3b063a78978f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Configure reticulate \n",
    "use_condaenv(condaenv='p.3.10.16-FD20200109SPLENO', conda=\"/nobackup/peer/fdeckert/miniconda3/bin/conda\", required=NULL)\n",
    "py_config()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1d5d15f-0039-427f-b6b1-95fe6758b054",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)\n",
    "options(dplyr.summarise.inform=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9722f4e5-454a-4030-b347-b5b2838f12cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_seed <- 42\n",
    "set.seed(random_seed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84e77268-2afe-4d87-aadb-87398dad3ffb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set working directory to project root\n",
    "setwd(\"/research/peer/fdeckert/FD20200109SPLENO\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9395fdb9-947c-4c8f-89eb-d408a104ae54",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Source files\n",
    "source(\"plotting_global.R\")\n",
    "source(\"bin/dea_pl.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba441961-9293-498b-868f-3eea05c0dd3c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Plotting Theme\n",
    "ggplot2::theme_set(theme_global_set(size_select=1)) # From project global source()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f65da44f-545f-4825-8cd2-195f696135f4",
   "metadata": {},
   "source": [
    "# Parameter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "265add99-0ef6-4768-8c39-ffee91f87459",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "pct_min=5\n",
    "\n",
    "cnt_min=3\n",
    "cell_min=3\n",
    "\n",
    "adj_pval_thr <- 0.01\n",
    "log2_thr <- 0.50\n",
    "\n",
    "adj_pval_gsea_thr <- 0.1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f141ef2-7c2b-425f-9ff1-f8d020a4ad94",
   "metadata": {},
   "source": [
    "# Import data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d15553e8-5cd8-4f35-b1a4-da2ad33dbc4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load Seurat object \n",
    "so <- readRDS(\"data/scRNAseq/object/pp.rds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74f9cae7-c1a6-44a8-b49c-6a7890e42880",
   "metadata": {},
   "source": [
    "# Helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f098058-1868-4965-aa2d-fd398de8dd80",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_select <- function(so, pct_min=0, cnt_min=3, cell_min=1) {\n",
    "\n",
    "    # Get count matrix \n",
    "    cnt <- GetAssayData(so, assay=\"RNA\", layer=\"counts\")\n",
    "\n",
    "    # Filter by percentage \n",
    "    cnt <- cnt[(100*rowSums(cnt>0)/ncol(cnt))>=pct_min, ]\n",
    "\n",
    "    # Filer by expression \n",
    "    cnt <- cnt[rowSums(cnt>=cnt_min)>=cell_min, ]\n",
    "\n",
    "    # return genes \n",
    "    genes <- rownames(cnt)\n",
    "        \n",
    "    return(genes)\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a7163f4-3c28-46a1-9b50-b38553cf492d",
   "metadata": {},
   "outputs": [],
   "source": [
    "wilcox <- function(so, ident, ident_1=NULL, ident_2=NULL, only_pos=FALSE, avg_log2FC_threshold=0, pct_min=0, cnt_min=0, cell_min=0, test_use=\"wilcox\", verbose=TRUE) {\n",
    "\n",
    "    # Select genes per group \n",
    "    genes_ident_1 <- feature_select(so[, so$group==ident_1], pct_min=pct_min, cnt_min=cnt_min, cell_min=cell_min)\n",
    "    genes_ident_2 <- feature_select(so[, so$group==ident_2], pct_min=pct_min, cnt_min=cnt_min, cell_min=cell_min)\n",
    "\n",
    "    genes <- union(genes_ident_1, genes_ident_2)\n",
    "\n",
    "    genes <- genes[!genes %in% c(\"Igha\", \"Igkc\", \"Tcrg-C1\")]\n",
    "\n",
    "    if(verbose) message(paste0(so$celltype_low[1], \" genes selected: \", length(genes)))\n",
    "\n",
    "    so <- so[genes, ]\n",
    "    \n",
    "    # Drop empty levels \n",
    "    so@meta.data <- droplevels(so@meta.data)\n",
    "    \n",
    "    # Check number of cells \n",
    "    n_cells_1 <- sum(so@meta.data[[ident]]==ident_1)\n",
    "    n_cells_2 <- sum(so@meta.data[[ident]]==ident_2)\n",
    "\n",
    "    check_1 <- n_cells_1 >= 3\n",
    "    check_2 <- n_cells_2 >= 3\n",
    "    \n",
    "    if(check_1 & check_2) {\n",
    "        \n",
    "        so <- SetIdent(so, value=ident)\n",
    "        so <- NormalizeData(so)\n",
    "        res <- RunPresto(so, ident.1=ident_1, ident.2=ident_2, logfc.threshold=avg_log2FC_threshold, min.pct=0, only.pos=only_pos, test.use=test_use)\n",
    "\n",
    "        # Adjusted p-value with IHW \n",
    "        res$mean_exp <- rowMeans(GetAssayData(so, assay=\"RNA\", slot=\"data\")[rownames(so), ])\n",
    "        res$p_val_adj <- IHW::adj_pvalues(IHW::ihw(res$p_val ~ res$mean_exp, alpha=0.05))\n",
    "\n",
    "        # Annotate results \n",
    "        res$gene <- rownames(res)\n",
    "\n",
    "        # N cells per group \n",
    "        res$n_cells_1 <- n_cells_1\n",
    "        res$n_cells_2 <- n_cells_2\n",
    "\n",
    "        return(res)\n",
    "        \n",
    "    } else {\n",
    "        \n",
    "        return(NULL)\n",
    "        \n",
    "    }\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13176fa8-2712-407b-a71e-f7eae8aa6831",
   "metadata": {},
   "source": [
    "# Run DEA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56770717-efba-44b6-83af-2fbb85dec9af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set group variable\n",
    "so$group <- so$sample_group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9be2c6d-2ef0-4e94-b3e2-a2be4bb287a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype_order <- levels(so$celltype_low)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c66dfae-1c02-401d-9044-7a1edab54e7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare data \n",
    "so <- SplitObject(so, split.by=\"celltype_low\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7daa9d4-af00-422c-b812-59834d5baf76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Order cell types \n",
    "so <- so[celltype_order]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77c258ff-d8a7-4811-889b-86f733d86d23",
   "metadata": {},
   "source": [
    "# Bl6 NaCl D6 vs Bl6 CpG D6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7c21f33-5630-47dc-ad3b-e9a905082f71",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrasts_vec <- c(\"Bl6_CpG_D6\", \"Bl6_NaCl_D6\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "529c3c2e-7d90-4b2d-98ca-f4666f4cd84d",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_1 <- lapply(so, function(x) subset(x, subset=group %in% contrasts_vec))\n",
    "so_1 <- so_1[!lapply(so_1, is.null) %>% unlist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9e49aba-b749-4a28-9f86-9eb98a570ec8",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_1 <- lapply(so_1, function(so) {wilcox(\n",
    "    \n",
    "    so=so, \n",
    "    ident=\"group\", \n",
    "    ident_1=contrasts_vec[1], \n",
    "    ident_2=contrasts_vec[2], \n",
    "    only_pos=FALSE, \n",
    "    avg_log2FC_threshold=0, \n",
    "    pct_min=pct_min, \n",
    "    cnt_min=cnt_min, \n",
    "    cell_min=cell_min,  \n",
    "    test_use=\"wilcox\"\n",
    "    \n",
    ")\n",
    "                                  }\n",
    "               )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49da2c41-3ab8-4120-8c55-8c420c415744",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_1 <- res_1[!lapply(res_1, is.null) %>% unlist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37ee52ac-5860-40ce-a872-b66a0d8d23c2",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=6*6, repr.plot.height=6)\n",
    "\n",
    "vp_1 <- lapply(names(res_1), function(i) v_pl(res_1[[i]], title=i, log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=2, color_pos=unlist(color$infection[\"CpG\"]), color_neg=unlist(color$infection[\"NaCl\"]), top_label=20) + theme_global_set(size_select=1))\n",
    "ggpubr::ggarrange(plotlist=vp_1, ncol=6, common.legend=TRUE, legend=\"bottom\") %>% print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbaa8486-6139-4741-b86d-32928251ed7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".pdf\"), onefile=TRUE, width=12, height=6)\n",
    "for (i in seq(length(vp_1))) {plot(vp_1[[i]])}\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae26c32e-e9e6-4129-873d-3b18b7492491",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(res_1) <- make.names(names(res_1))\n",
    "openxlsx::write.xlsx(res_1, paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".xlsx\"))\n",
    "saveRDS(res_1, paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "124ae512-eadc-4a76-8684-7343964f8c03",
   "metadata": {},
   "source": [
    "# IFNAR fl/fl baseline vs IFNAR fl/fl CpG D1 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3860b76-259c-42c7-97e4-eb0dda8b0ede",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrasts_vec <- c(\"IFNAR_fl_CpG_D1\", \"IFNAR_fl_Baseline_D0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44d11b63-d3e0-429f-9807-178feb23d938",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_1 <- lapply(so, function(x) tryCatch({subset(x, subset=group %in% contrasts_vec)}, error=function(e) {NULL}))\n",
    "so_1 <- so_1[!lapply(so_1, is.null) %>% unlist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a46b5c1-f5e7-4ae4-bce1-3630b3eab5be",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_1 <- lapply(so_1, function(so) {wilcox(\n",
    "    \n",
    "    so=so, \n",
    "    ident=\"group\", \n",
    "    ident_1=contrasts_vec[1], \n",
    "    ident_2=contrasts_vec[2], \n",
    "    only_pos=FALSE, \n",
    "    avg_log2FC_threshold=0, \n",
    "    pct_min=pct_min, \n",
    "    cnt_min=cnt_min, \n",
    "    cell_min=cell_min,  \n",
    "    test_use=\"wilcox\"\n",
    "    \n",
    ")\n",
    "                                  }\n",
    "               )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "880b1bbc-3586-4713-a501-383b8cfe188e",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_1 <- res_1[!lapply(res_1, is.null) %>% unlist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59f2e6f5-5744-4599-b093-55f5ad780f62",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=6*6, repr.plot.height=6)\n",
    "\n",
    "vp_1 <- lapply(names(res_1), function(i) v_pl(res_1[[i]], title=i, log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=2, color_pos=unlist(color$sample_group[contrasts_vec[1]]), color_neg=unlist(color$sample_group[contrasts_vec[2]]), top_label=20) + theme_global_set(size_select=1))\n",
    "ggpubr::ggarrange(plotlist=vp_1, ncol=4, common.legend=TRUE, legend=\"bottom\") %>% print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "634b8398-a1cd-42d5-8c2d-56dd84ac8034",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".pdf\"), onefile=TRUE, width=12, height=6)\n",
    "for (i in seq(length(vp_1))) {plot(vp_1[[i]])}\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58ae046f-9c0f-4c8d-ad6c-be0e1604f8c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(res_1) <- make.names(names(res_1))\n",
    "openxlsx::write.xlsx(res_1, paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".xlsx\"))\n",
    "saveRDS(res_1, paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d52aa1b5-bf6d-4b19-9f8f-93d4f0516ef1",
   "metadata": {},
   "source": [
    "# IFNAR fl/fl baseline vs IFNAR fl/fl CpG D3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "375413e1-b48e-40d9-ac8c-61b4c7b23cc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrasts_vec <- c(\"IFNAR_fl_CpG_D3\", \"IFNAR_fl_Baseline_D0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fcea127-dafe-4067-b42c-5895b827185c",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_1 <- lapply(so, function(x) tryCatch({subset(x, subset=group %in% contrasts_vec)}, error=function(e) {NULL}))\n",
    "so_1 <- so_1[!lapply(so_1, is.null) %>% unlist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69badbaa-814c-4db4-98a4-6dede75bb151",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_1 <- lapply(so_1, function(so) {wilcox(\n",
    "    \n",
    "    so=so, \n",
    "    ident=\"group\", \n",
    "    ident_1=contrasts_vec[1], \n",
    "    ident_2=contrasts_vec[2], \n",
    "    only_pos=FALSE, \n",
    "    avg_log2FC_threshold=0, \n",
    "    pct_min=pct_min, \n",
    "    cnt_min=cnt_min, \n",
    "    cell_min=cell_min,   \n",
    "    test_use=\"wilcox\"\n",
    "    \n",
    ")\n",
    "                                  }\n",
    "               )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05067d7e-5bbc-44fb-b57e-bc4f34d461d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_1 <- res_1[!lapply(res_1, is.null) %>% unlist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5ea5aaa-3d56-4894-a428-2c31f2477962",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=6*6, repr.plot.height=6)\n",
    "\n",
    "vp_1 <- lapply(names(res_1), function(i) v_pl(res_1[[i]], title=i, log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=2, color_pos=unlist(color$sample_group[contrasts_vec[1]]), color_neg=unlist(color$sample_group[contrasts_vec[2]]), top_label=20) + theme_global_set(size_select=1))\n",
    "ggpubr::ggarrange(plotlist=vp_1, ncol=4, common.legend=TRUE, legend=\"bottom\") %>% print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dee4e1f-47ce-4232-a237-89b308dde631",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".pdf\"), onefile=TRUE, width=12, height=6)\n",
    "for (i in seq(length(vp_1))) {plot(vp_1[[i]])}\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0aa2d30-02de-4822-bb2d-1aff0e274317",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(res_1) <- make.names(names(res_1))\n",
    "openxlsx::write.xlsx(res_1, paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".xlsx\"))\n",
    "saveRDS(res_1, paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b6c2db0-15c0-4467-bf37-9a3f5e3935f7",
   "metadata": {},
   "source": [
    "# IFNAR fl/fl LysM cre baseline vs IFNAR fl/fl LysM cre CpG D1 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e96f193f-9c04-4ab9-9443-ba5fda3499f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrasts_vec <- c(\"IFNAR_fl_LysM_cre_CpG_D1\", \"IFNAR_fl_LysM_cre_Baseline_D0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b53f8b79-0feb-4ef9-94d9-00eecb82a43b",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_1 <- lapply(so, function(x) tryCatch({subset(x, subset=group %in% contrasts_vec)}, error=function(e) {NULL}))\n",
    "so_1 <- so_1[!lapply(so_1, is.null) %>% unlist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89116421-c6da-40bc-8cb3-515abe2c37e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_1 <- lapply(so_1, function(so) {wilcox(\n",
    "    \n",
    "    so=so, \n",
    "    ident=\"group\", \n",
    "    ident_1=contrasts_vec[1], \n",
    "    ident_2=contrasts_vec[2], \n",
    "    only_pos=FALSE, \n",
    "    avg_log2FC_threshold=0, \n",
    "    pct_min=pct_min, \n",
    "    cnt_min=cnt_min, \n",
    "    cell_min=cell_min,  \n",
    "    test_use=\"wilcox\"\n",
    "    \n",
    ")\n",
    "                                  }\n",
    "               )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50735107-7370-4480-a52d-940dc6934ab6",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_1 <- res_1[!lapply(res_1, is.null) %>% unlist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fae8530-0b7f-4ea8-9762-d5c9287caf35",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=6*6, repr.plot.height=6)\n",
    "\n",
    "vp_1 <- lapply(names(res_1), function(i) v_pl(res_1[[i]], title=i, log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=2, color_pos=unlist(color$sample_group[contrasts_vec[1]]), color_neg=unlist(color$sample_group[contrasts_vec[2]]), top_label=20) + theme_global_set(size_select=1))\n",
    "ggpubr::ggarrange(plotlist=vp_1, ncol=4, common.legend=TRUE, legend=\"bottom\") %>% print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "997c44f9-d0db-4801-bcd5-6f3defcb3b9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".pdf\"), onefile=TRUE, width=12, height=6)\n",
    "for (i in seq(length(vp_1))) {plot(vp_1[[i]])}\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4277fb1-73d9-483f-bd0f-17448af165b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(res_1) <- make.names(names(res_1))\n",
    "openxlsx::write.xlsx(res_1, paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".xlsx\"))\n",
    "saveRDS(res_1, paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98edbeb8-a3e5-4ce9-9a92-3c993fcfde63",
   "metadata": {},
   "source": [
    "# IFNAR fl/fl LysM cre baseline vs IFNAR fl/fl LysM cre CpG D3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be543836-71bb-4895-b24f-1d03a6d78a56",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrasts_vec <- c(\"IFNAR_fl_LysM_cre_CpG_D3\", \"IFNAR_fl_LysM_cre_Baseline_D0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34856369-bea9-44d2-8507-d8b53ac3c002",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_1 <- lapply(so, function(x) tryCatch({subset(x, subset=group %in% contrasts_vec)}, error=function(e) {NULL}))\n",
    "so_1 <- so_1[!lapply(so_1, is.null) %>% unlist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddb86149-a2a2-49df-8d11-63419a9b0e80",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_1 <- lapply(so_1, function(so) {wilcox(\n",
    "    \n",
    "    so=so, \n",
    "    ident=\"group\", \n",
    "    ident_1=contrasts_vec[1], \n",
    "    ident_2=contrasts_vec[2], \n",
    "    only_pos=FALSE, \n",
    "    avg_log2FC_threshold=0, \n",
    "    pct_min=pct_min, \n",
    "    cnt_min=cnt_min, \n",
    "    cell_min=cell_min,   \n",
    "    test_use=\"wilcox\"\n",
    "    \n",
    ")\n",
    "                                  }\n",
    "               )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "378c918e-20b1-4407-a364-90db3ba93c56",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_1 <- res_1[!lapply(res_1, is.null) %>% unlist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18d8f1d3-2997-4d95-ba26-5e7f868bef8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=6*6, repr.plot.height=6)\n",
    "\n",
    "vp_1 <- lapply(names(res_1), function(i) v_pl(res_1[[i]], title=i, log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=2, color_pos=unlist(color$sample_group[contrasts_vec[1]]), color_neg=unlist(color$sample_group[contrasts_vec[2]]), top_label=20) + theme_global_set(size_select=1))\n",
    "ggpubr::ggarrange(plotlist=vp_1, ncol=4, common.legend=TRUE, legend=\"bottom\") %>% print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94967694-24c2-4527-a966-b3a741a2afb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".pdf\"), onefile=TRUE, width=12, height=6)\n",
    "for (i in seq(length(vp_1))) {plot(vp_1[[i]])}\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5757eed7-4f87-45b9-bb8a-d4b5ffb0bb36",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(res_1) <- make.names(names(res_1))\n",
    "openxlsx::write.xlsx(res_1, paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".xlsx\"))\n",
    "saveRDS(res_1, paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb14c4f7-f1be-4ce8-84d7-868ca8aee468",
   "metadata": {},
   "source": [
    "# IFNAR fl/fl LysM cre baseline vs IFNAR fl/fl baseline "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1197d326-3763-48f9-9870-2681f5b2e0d6",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "contrasts_vec <- c(\"IFNAR_fl_LysM_cre_Baseline_D0\", \"IFNAR_fl_Baseline_D0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a10569f-bfb0-41bc-a9b9-bf925b3d157d",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_1 <- lapply(so, function(x) tryCatch({subset(x, subset=group %in% contrasts_vec)}, error=function(e) {NULL}))\n",
    "so_1 <- so_1[!lapply(so_1, is.null) %>% unlist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4119b567-78b9-4ffb-97f7-d72fc34dacfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_1 <- lapply(so_1, function(so) {wilcox(\n",
    "    \n",
    "    so=so, \n",
    "    ident=\"group\", \n",
    "    ident_1=contrasts_vec[1], \n",
    "    ident_2=contrasts_vec[2], \n",
    "    only_pos=FALSE, \n",
    "    avg_log2FC_threshold=0, \n",
    "    pct_min=pct_min, \n",
    "    cnt_min=cnt_min, \n",
    "    cell_min=cell_min,  \n",
    "    test_use=\"wilcox\"\n",
    "    \n",
    ")\n",
    "                                  }\n",
    "               )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8dfafc3-04d7-4d58-ad8b-680e4607e0c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_1 <- res_1[!lapply(res_1, is.null) %>% unlist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1ce64eb-225c-46f7-afe9-b011817c296d",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=6*6, repr.plot.height=6)\n",
    "\n",
    "vp_1 <- lapply(names(res_1), function(i) v_pl(res_1[[i]], title=i, log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=2, color_pos=unlist(color$sample_group[contrasts_vec[1]]), color_neg=unlist(color$sample_group[contrasts_vec[2]]), top_label=20) + theme_global_set(size_select=1))\n",
    "ggpubr::ggarrange(plotlist=vp_1, ncol=4, common.legend=TRUE, legend=\"bottom\") %>% print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80fcf081-6792-4972-b288-ee4de240f7b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".pdf\"), onefile=TRUE, width=12, height=6)\n",
    "for (i in seq(length(vp_1))) {plot(vp_1[[i]])}\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b38bd51a-aa6e-491e-a2c3-0bc68a2dddcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(res_1) <- make.names(names(res_1))\n",
    "openxlsx::write.xlsx(res_1, paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".xlsx\"))\n",
    "saveRDS(res_1, paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75a92dc6-884e-4573-a56d-2de2f0004d6d",
   "metadata": {},
   "source": [
    "# IFNAR fl/fl LysM cre D1 vs IFNAR fl/fl D1 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "479bbec2-627f-4b25-a18c-7555eadbbca4",
   "metadata": {},
   "outputs": [],
   "source": [
    "contrasts_vec <- c(\"IFNAR_fl_LysM_cre_CpG_D1\", \"IFNAR_fl_CpG_D1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7aa8d5ab-ed13-414a-8412-9dbd67c2cb96",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_1 <- lapply(so, function(x) tryCatch({subset(x, subset=group %in% contrasts_vec)}, error=function(e) {NULL}))\n",
    "so_1 <- so_1[!lapply(so_1, is.null) %>% unlist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c1ebff5-00fb-49a9-b72e-34f499ccdc70",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_1 <- lapply(so_1, function(so) {wilcox(\n",
    "    \n",
    "    so=so, \n",
    "    ident=\"group\", \n",
    "    ident_1=contrasts_vec[1], \n",
    "    ident_2=contrasts_vec[2], \n",
    "    only_pos=FALSE, \n",
    "    avg_log2FC_threshold=0, \n",
    "    pct_min=pct_min, \n",
    "    cnt_min=cnt_min, \n",
    "    cell_min=cell_min,   \n",
    "    test_use=\"wilcox\"\n",
    "    \n",
    ")\n",
    "                                  }\n",
    "               )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27611a3f-3f89-4f88-aa68-ba8616c8eace",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_1 <- res_1[!lapply(res_1, is.null) %>% unlist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba747b81-f6c6-4c78-96cd-e0bbc1c44537",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=6*6, repr.plot.height=6)\n",
    "\n",
    "vp_1 <- lapply(names(res_1), function(i) v_pl(res_1[[i]], title=i, log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=2, color_pos=unlist(color$sample_group[contrasts_vec[1]]), color_neg=unlist(color$sample_group[contrasts_vec[2]]), top_label=20) + theme_global_set(size_select=1))\n",
    "ggpubr::ggarrange(plotlist=vp_1, ncol=4, common.legend=TRUE, legend=\"bottom\") %>% print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f59a4494-08c3-44f3-b21b-a76c4c30200c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".pdf\"), onefile=TRUE, width=12, height=6)\n",
    "for (i in seq(length(vp_1))) {plot(vp_1[[i]])}\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afa1784a-49e2-4e6d-b8ac-6a2c638c84f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(res_1) <- make.names(names(res_1))\n",
    "openxlsx::write.xlsx(res_1, paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".xlsx\"))\n",
    "saveRDS(res_1, paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63e54e6f-11fa-42f5-8fc5-0e4802906467",
   "metadata": {},
   "source": [
    "# IFNAR fl/fl LysM cre D3 vs IFNAR fl/fl D3 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46041d80-d5ad-454b-b3af-ddbbc7a83e7a",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "contrasts_vec <- c(\"IFNAR_fl_LysM_cre_CpG_D3\", \"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca5c0376-7344-4260-ae60-de73242c3a66",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_1 <- lapply(so, function(x) tryCatch({subset(x, subset=group %in% contrasts_vec)}, error=function(e) {NULL}))\n",
    "so_1 <- so_1[!lapply(so_1, is.null) %>% unlist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96a9ebd7-911e-4e77-9356-ef3af0116925",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_1 <- lapply(so_1, function(so) {wilcox(\n",
    "    \n",
    "    so=so, \n",
    "    ident=\"group\", \n",
    "    ident_1=contrasts_vec[1], \n",
    "    ident_2=contrasts_vec[2], \n",
    "    only_pos=FALSE, \n",
    "    avg_log2FC_threshold=0, \n",
    "    pct_min=pct_min, \n",
    "    cnt_min=cnt_min, \n",
    "    cell_min=cell_min,  \n",
    "    test_use=\"wilcox\"\n",
    "    \n",
    ")\n",
    "                                  }\n",
    "               )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49cca866-231d-4034-b54d-b96097f44e02",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_1 <- res_1[!lapply(res_1, is.null) %>% unlist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfe911fa-8131-4a74-87f4-ec6b58cf719d",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=6*6, repr.plot.height=6)\n",
    "\n",
    "vp_1 <- lapply(names(res_1), function(i) v_pl(res_1[[i]], title=i, log2_thr=log2_thr, adj_pval_thr=adj_pval_thr, point_size=2, color_pos=unlist(color$sample_group[contrasts_vec[1]]), color_neg=unlist(color$sample_group[contrasts_vec[2]]), top_label=20) + theme_global_set(size_select=1))\n",
    "ggpubr::ggarrange(plotlist=vp_1, ncol=4, common.legend=TRUE, legend=\"bottom\") %>% print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1d7f8e4-1976-4f0e-b5ca-828a81a98ff8",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".pdf\"), onefile=TRUE, width=12, height=6)\n",
    "for (i in seq(length(vp_1))) {plot(vp_1[[i]])}\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19061c79-cf8d-4fac-ad8e-09008881b33d",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(res_1) <- make.names(names(res_1))\n",
    "openxlsx::write.xlsx(res_1, paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".xlsx\"))\n",
    "saveRDS(res_1, paste0(\"result/dea/scRNAseq/wilcox/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a5d1028-95b0-4ef3-a049-1711923e549a",
   "metadata": {},
   "source": [
    "# Session info "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09c6f204-af6a-4576-82aa-a6782a402c60",
   "metadata": {},
   "outputs": [],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "r.4.4.1-FD20200109SPLENO",
   "language": "R",
   "name": "r.4.4.1-fd20200109spleno"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
