{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3304709f-bb88-43b9-8055-ce19dcd51aef",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import sys\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ec24ffc-7b79-4ac2-89d8-0300f58ac4df",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Root directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a1e0304-27dc-4bed-8498-f6857ddbffd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/research/peer/fdeckert/FD20200109SPLENO')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f564f844-e25f-499b-bde2-3f06848b0718",
   "metadata": {},
   "source": [
    "## Custom modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "208e369b-9b08-41fe-99d1-8eab8f6d9fc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.append('bin/')\n",
    "from adata_qc import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1097d920-ff3b-46d4-9ae4-56a0f09d497f",
   "metadata": {},
   "source": [
    "## Figures "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f1117a1-60b7-44b1-bd22-6e81544f9aa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(figsize=(7.5, 7.5), dpi_save=1200, fontsize=12, frameon=False, facecolor='white')\n",
    "mpl.rcParams['figure.facecolor'] = 'white'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c205ad6-8ac1-43f3-9fe2-90cd63b58eef",
   "metadata": {},
   "source": [
    "# Import AnnData and write bar codes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aab957e6-e8ad-474f-847c-9817fe2c47fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad('data/scRNAseq/object/pp_1.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cea8adc-d2f7-40fb-bce9-580f280b3562",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(adata, target_sum=10000)\n",
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01192f06-8f5d-4ffe-ae82-f1ed2faec10c",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['celltype_low', 'Ifnar1', 'Lyz2', 'Tlr9'], size=50, wspace=0.1, legend_loc='on data', use_raw=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "672bb82b-0bf3-44be-96f7-8c514e7381cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adata[adata.obs.celltype_low.isin(['cMo', 'intMo', 'ncMo'])].obs_names.str.replace(r\"(-1).*\", r\"\\1\", regex=True).to_series().to_csv(\"data/scRNAseq/bamcoverage/celltype/bc_1.tsv\", index=False, header=False)\n",
    "# adata[adata.obs.celltype_low.isin(['RPM'])].obs_names.str.replace(r\"(-1).*\", r\"\\1\", regex=True).to_series().to_csv(\"data/scRNAseq/bamcoverage/celltype/bc_2.tsv\", index=False, header=False)\n",
    "# adata[adata.obs.celltype_low.isin(['MEP', 'Proerythroblast', 'Erythroblast'])].obs_names.str.replace(r\"(-1).*\", r\"\\1\", regex=True).to_series().to_csv(\"data/scRNAseq/bamcoverage/celltype/bc_3.tsv\", index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edf54720-67a0-4e12-a4df-e588615e25ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%bash\n",
    "# module load SAMtools\n",
    "\n",
    "# # Set working directory \n",
    "# cd data/scRNAseq/bamcoverage/celltype/\n",
    "\n",
    "# # Subset BAM by cell barcodes \n",
    "# /nobackup/peer/fdeckert/miniconda3/envs/p.3.10.16-FD20200109SPLENO/bin/subset-bam -b /nobackup/peer/fdeckert/FD20200109SPLENO/data/VBC_R18518_22WYJ2LT3/22WYJ2LT3_R18518/outs/possorted_genome_bam.bam -c bc_1.tsv -o bc_1.bam --cores 32\n",
    "# /nobackup/peer/fdeckert/miniconda3/envs/p.3.10.16-FD20200109SPLENO/bin/subset-bam -b /nobackup/peer/fdeckert/FD20200109SPLENO/data/VBC_R18518_22WYJ2LT3/22WYJ2LT3_R18518/outs/possorted_genome_bam.bam -c bc_2.tsv -o bc_2.bam --cores 32\n",
    "# /nobackup/peer/fdeckert/miniconda3/envs/p.3.10.16-FD20200109SPLENO/bin/subset-bam -b /nobackup/peer/fdeckert/FD20200109SPLENO/data/VBC_R18518_22WYJ2LT3/22WYJ2LT3_R18518/outs/possorted_genome_bam.bam -c bc_3.tsv -o bc_3.bam --cores 32\n",
    "\n",
    "# # Index BAM \n",
    "# samtools index bc_1.bam -@ 32\n",
    "# samtools index bc_2.bam -@ 32\n",
    "# samtools index bc_3.bam -@ 32"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5238f6cf-d0e5-4fab-a55b-f8e4f5aaffd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%bash\n",
    "# module load deepTools\n",
    "\n",
    "# # Set working directory\n",
    "# cd data/scRNAseq/bamcoverage/celltype/\n",
    "\n",
    "# # BAM coverage BigWig\n",
    "# bamCoverage -b bc_1.bam -o bc_1.bw --normalizeUsing CPM --binSize 10 --smoothLength 20 --effectiveGenomeSize 2652783500 -p 32\n",
    "# bamCoverage -b bc_2.bam -o bc_2.bw --normalizeUsing CPM --binSize 10 --smoothLength 20 --effectiveGenomeSize 2652783500 -p 32\n",
    "# bamCoverage -b bc_3.bam -o bc_3.bw --normalizeUsing CPM --binSize 10 --smoothLength 20 --effectiveGenomeSize 2652783500 -p 32"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "442854d7-2aba-414f-8136-09a64c820491",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%bash \n",
    "\n",
    "# cd data/scRNAseq/bamcoverage/celltype/\n",
    "\n",
    "# awk '$3 == \"exon\" && $0 ~ /gene_name \"Ifnar1\"/' /nobackup/peer/fdeckert/cellranger/mm10-2020-A/mm10/genes/genes.gtf | \\\n",
    "# awk -F'\\t' '{print $0}' | \\\n",
    "# awk 'BEGIN{OFS=\"\\t\"} {print $1, $4-1, $5, $9}' | \\\n",
    "# awk 'NR==3' > ifnar1_exon_3.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d3bdd4c-a684-4e86-95f1-d30b2bba683d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load deepTools\n",
    "module load BEDTools\n",
    "\n",
    "# Set working directory\n",
    "cd result/bamcoverage/scRNAseq/celltype/\n",
    "\n",
    "# Visualization tracks\n",
    "/nobackup/peer/fdeckert/miniconda3/envs/p.3.10.16-FD20200109SPLENO/bin/pyGenomeTracks --tracks track_ifnar1.ini --fontSize 8 --trackLabelFraction 0.1 --dpi 300 --region chr16:91482000-91512000 -o track_ifnar1.pdf\n",
    "/nobackup/peer/fdeckert/miniconda3/envs/p.3.10.16-FD20200109SPLENO/bin/pyGenomeTracks --tracks track_lyzm2.ini --fontSize 8 --trackLabelFraction 0.1 --dpi 300 --region chr10:117270000-117300000 -o track_lyzm2.pdf\n",
    "/nobackup/peer/fdeckert/miniconda3/envs/p.3.10.16-FD20200109SPLENO/bin/pyGenomeTracks --tracks track_tlr9.ini --fontSize 8 --trackLabelFraction 0.1 --dpi 300 --region chr9:106220598-106229883 -o track_tlr9.pdf"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "p.3.10.16",
   "language": "python",
   "name": "p.3.10.16"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
