{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cd04ed1a-d1ae-441f-9d5e-88c0aef441ca",
   "metadata": {},
   "source": [
    "# QC and filter seurat object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e9ed315-3005-46a2-b6bc-bed3a2094875",
   "metadata": {},
   "outputs": [],
   "source": [
    "library_load <- suppressMessages(\n",
    "    \n",
    "    list(\n",
    "        # Seurat \n",
    "        library(Seurat),\n",
    "\n",
    "        # Single R\n",
    "        library(SingleCellExperiment),\n",
    "        library(SingleR), \n",
    "\n",
    "        # Data \n",
    "        library(dplyr),\n",
    "        library(reshape2),\n",
    "        library(Matrix),\n",
    "\n",
    "        # Source \n",
    "        library(msigdbr),\n",
    "\n",
    "        # Plotting\n",
    "        library(ggplot2),\n",
    "        library(ComplexHeatmap),\n",
    "        library(gridExtra),\n",
    "        library(patchwork),\n",
    "        library(RColorBrewer),\n",
    "        library(viridis)\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6ab1590-bda7-46f7-b50a-5fdde52a7846",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)\n",
    "ht_opt$message=FALSE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac325532-cdf7-4143-8775-463c6852b68e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Set working directory to project root\n",
    "setwd(\"/research/peer/fdeckert/FD20200109SPLENO\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44a53f47-22fb-424c-868c-8bc2fe748580",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Source files\n",
    "source(\"plotting_global.R\")\n",
    "source(\"bin/SeuratQC.R\")\n",
    "\n",
    "# SingleRPlot\n",
    "source(\"bin/SingleRQC.R\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dfe8c56-8c9d-4dc0-a5d4-d6d075b84a27",
   "metadata": {},
   "source": [
    "# Parameter settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1e02234-e351-4946-b71d-40c09cceb047",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filtering Parameter\n",
    "nFeature_RNA_min_m <- 300\n",
    "nFeature_RNA_min_p <- 300\n",
    "\n",
    "nFeature_RNA_max_m <- 3000\n",
    "nFeature_RNA_max_p <- 3000\n",
    "\n",
    "nCount_RNA_min_m <- 1000\n",
    "nCount_RNA_min_p <- 1000\n",
    "\n",
    "nCount_RNA_max_m <- 20000\n",
    "nCount_RNA_max_p <- 20000\n",
    "\n",
    "pMt_RNA_max_m <- 5\n",
    "pMt_RNA_max_p <- 7.5\n",
    "\n",
    "# Files \n",
    "so_raw_file <- \"data/object/seurat_raw.rds\"\n",
    "\n",
    "so_qc_file <- \"data/object/seurat.rds\"\n",
    "\n",
    "# Plotting Theme\n",
    "ggplot2::theme_set(theme_global_set()) # From project global source()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "354a4c0c-2393-4815-a7a8-dd8beb9e3127",
   "metadata": {},
   "source": [
    "# Import Seurat object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b406cedc-cfeb-4376-a330-9305d17c3e62",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_raw <- readRDS(so_raw_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d160faf4-e9c2-49a7-abd7-ff86af3ff4cc",
   "metadata": {},
   "source": [
    "# Add filter meta data to Seurat object "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa2f990f-8f91-4539-9676-0433435c2a04",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_raw$nFeature_RNA_max <- ifelse(so_raw$tissue == \"Myeloid\", nFeature_RNA_max_m, nFeature_RNA_max_p)\n",
    "so_raw$nFeature_RNA_min <- ifelse(so_raw$tissue == \"Myeloid\", nFeature_RNA_min_m, nFeature_RNA_min_p)\n",
    "\n",
    "so_raw$nCount_RNA_max <- ifelse(so_raw$tissue == \"Myeloid\", nCount_RNA_max_m, nCount_RNA_max_p)\n",
    "so_raw$nCount_RNA_min <- ifelse(so_raw$tissue == \"Myeloid\", nCount_RNA_min_m, nCount_RNA_min_p)\n",
    "\n",
    "so_raw$pMt_RNA_max <- ifelse(so_raw$tissue == \"Myeloid\", pMt_RNA_max_m, pMt_RNA_max_p)\n",
    "\n",
    "# QC \n",
    "so_raw$qc_class <- ifelse(\n",
    "    so_raw$cellranger_class == \"Cell\" &\n",
    "    so_raw$nFeature_RNA <= so_raw$nFeature_RNA_max & \n",
    "    so_raw$nFeature_RNA >= so_raw$nFeature_RNA_min & \n",
    "    so_raw$nCount_RNA <= so_raw$nCount_RNA_max & \n",
    "    so_raw$nCount_RNA >= so_raw$nCount_RNA_min &\n",
    "    so_raw$pMt_RNA <= so_raw$pMt_RNA_max, \n",
    "  \"pass\", \"fail\"\n",
    "  )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b069077-d33d-458c-8ecd-6bcdad56379a",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_raw$tissue <- factor(so_raw$tissue, levels=c(\"Myeloid\", \"Progenitor\"))\n",
    "so_raw$treatment <- factor(so_raw$treatment, levels=c(\"NaCl\", \"CpG\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72d58dc5-d02c-44be-91b1-10b983728b17",
   "metadata": {},
   "source": [
    "# Filter cells by rank plot\n",
    "\n",
    "Empty droplets were determined with CellRanger V3.0.2 Lun et al., 2019 EmptyDrop heuristic. RNAse activity of granulocytes might be wrongly identified as empty cells by CellRanger.\n",
    "\n",
    "**Typical Sample** A steep drop-off is indicative of good separation between the cell-associated barcodes and the barcodes associated with empty GEMs. A ideal barcode rank plot has a distincitve shape, which is referred to as a \"cliff and knee\".  \n",
    "**Heterogeneous Sample** Heterogeneous populations of cells in a sample result in two \"cliff and knee\" distributions. However, there should still be clear separation between the bacodes.   \n",
    "**Compromised Sample** Round curve and lack of steep cliff may indicate low sample quality or loss of single-cell behavior. This can be due to a wetting failure, premature cell lysis, or low cell viability.   \n",
    "**Compromised Sample** Defined cliff and knee, but the total number of barcodes detected may be lower than expected. This can be caused by a sample clog or inaccurate cell count.   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48f4c35b-90b9-4a08-b8d7-99c86c7a8ddf",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=5, repr.plot.height=5)\n",
    "rank_plot <- ggplot(so_raw@meta.data, aes(x=log10(nCount_RNA_rank), y=log10(nCount_RNA), color=cellranger_class)) + \n",
    "  geom_point() + \n",
    "  scale_color_manual(values=color$cellranger_class) +\n",
    "  ggtitle(\"Barcode rank plot\") +\n",
    "  xlab(\"log10(cell barcode rank)\") + ylab(\"log10(cell UMI counts)\") + \n",
    "  facet_grid(tissue~treatment) + \n",
    "  theme(aspect.ratio=1, legend.position=\"bottom\")\n",
    "rank_plot\n",
    "ggsave(rank_plot, filename=\"result/plot/seurat/rank_plot.png\", width=5, height=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "295859a3-b6c3-44de-88da-311555aff740",
   "metadata": {},
   "source": [
    "# Filter by CellRanger class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04d17536-94d8-4229-bdab-36ea8bfc78b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_qc <- subset(so_raw, subset=cellranger_class == \"Cell\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "207a9009-9445-44c8-9908-7e1b60e63251",
   "metadata": {},
   "source": [
    "# Filter by UMI and Feature count \n",
    "\n",
    "The Progenitor NaCl sample show a high MT%. It would be interessting to see whether those are apoptoitic cells or if they have a increased energy demand. That could be done by correlating the MT% with complementary nuclear mitochondrial gene expression from the OXPHOS system. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fa41dec-0eea-44f5-8377-15f8e551705f",
   "metadata": {},
   "outputs": [],
   "source": [
    "qc_1 <- ggplot(so_qc@meta.data, aes(x=log10(nCount_RNA), fill=tissue)) + \n",
    "  geom_density() + \n",
    "  ggtitle(\"Density plot UMI count\") + xlab(\"log10(UMI count)\") + ylab(\"Density\") +\n",
    "  geom_vline(aes(xintercept=log10(nCount_RNA_min)), color=\"red\", linetype=\"longdash\") +\n",
    "  geom_vline(aes(xintercept=log10(nCount_RNA_max)), color=\"red\", linetype=\"longdash\") + \n",
    "  scale_x_continuous(breaks=integer_breaks()) + \n",
    "  scale_fill_manual(values=color$tissue) +\n",
    "  facet_grid(tissue~treatment+sample_rep) + \n",
    "  theme(legend.position=\"bottom\", aspect.ratio=1) \n",
    "\n",
    "qc_2 <- ggplot(so_qc@meta.data, aes(x=log10(nFeature_RNA), fill=tissue)) + \n",
    "  geom_density() + \n",
    "  ggtitle(\"Density plot Feature count\") + xlab(\"log10(Feature count)\") + ylab(\"Density\") +\n",
    "  geom_vline(aes(xintercept=log10(nFeature_RNA_min)), color=\"red\", linetype=\"longdash\") +\n",
    "  geom_vline(aes(xintercept=log10(nFeature_RNA_max)), color=\"red\", linetype=\"longdash\") + \n",
    "  scale_x_continuous(breaks=integer_breaks()) + \n",
    "  scale_fill_manual(values=color$tissue) +\n",
    "  facet_grid(tissue~treatment+sample_rep) + \n",
    "  theme(legend.position=\"bottom\", aspect.ratio=1)\n",
    "\n",
    "qc_3 <- ggplot(so_qc@meta.data, aes(x=pMt_RNA, fill=tissue)) + \n",
    "  geom_density() + \n",
    "  ggtitle(\"Density plot Mt %\") + xlab(\"Mt [%]\") + ylab(\"Density\") +\n",
    "  geom_vline(aes(xintercept=pMt_RNA_max), color=\"red\", linetype=\"longdash\") +\n",
    "  scale_x_continuous(breaks=integer_breaks()) +\n",
    "  scale_fill_manual(values=color$tissue) +\n",
    "  facet_grid(tissue~treatment+sample_rep)  + \n",
    "  theme(legend.position=\"bottom\", aspect.ratio=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e516954c-2bea-40c3-be9b-df383277c160",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=15, repr.plot.height=10)\n",
    "qc_1 + qc_2 + qc_3 + plot_layout(ncol=2) & theme(legend.position=\"bottom\")\n",
    "ggsave(qc_1, filename=\"result/plot/seurat/density_umi.png\", width=5, height=5)\n",
    "ggsave(qc_2, filename=\"result/plot/seurat/density_feature.png\", width=5, height=5)\n",
    "ggsave(qc_3, filename=\"result/plot/seurat/density_mt.png\", width=5, height=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a73ac64-2e0a-4709-b4f1-d24dc94332af",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc_1 <- ggplot(so_qc@meta.data, aes(x=log10(nCount_RNA), y=log10(nFeature_RNA), color=pMt_RNA)) + \n",
    "  geom_point() + ggtitle(\"Mitochondrial gene percentage\") + ylab(\"log10(feature count)\") + xlab(\"log10(umi count)\") + \n",
    "  geom_vline(aes(xintercept=log10(nCount_RNA_min)), color=\"red\", linetype=\"longdash\") +\n",
    "  geom_vline(aes(xintercept=log10(nCount_RNA_max)), color=\"red\", linetype=\"longdash\") +\n",
    "  geom_hline(aes(yintercept=log10(nFeature_RNA_min)), color=\"red\", linetype=\"longdash\") + \n",
    "  geom_hline(aes(yintercept=log10(nFeature_RNA_max)), color=\"red\", linetype=\"longdash\") + \n",
    "  facet_grid(tissue~treatment+sample_rep) + theme(aspect.ratio=1, legend.position=\"bottom\") + \n",
    "  scale_size(guide=guide_legend(direction=\"vertical\"))\n",
    "\n",
    "sc_2 <- ggplot(so_qc@meta.data, aes(x=log10(nCount_RNA), y=log10(nFeature_RNA), color=pHb_RNA)) + \n",
    "  geom_point() + ggtitle(\"Hemoglobin gene percentage\") + ylab(\"log10(feature count)\") + xlab(\"log10(umi count)\") +\n",
    "  geom_vline(aes(xintercept=log10(nCount_RNA_min)), color=\"red\", linetype=\"longdash\") +\n",
    "  geom_vline(aes(xintercept=log10(nCount_RNA_max)), color=\"red\", linetype=\"longdash\") +\n",
    "  geom_hline(aes(yintercept=log10(nFeature_RNA_min)), color=\"red\", linetype=\"longdash\") + \n",
    "  geom_hline(aes(yintercept=log10(nFeature_RNA_max)), color=\"red\", linetype=\"longdash\") + \n",
    "  facet_grid(tissue~treatment+sample_rep) + theme(aspect.ratio=1, legend.position=\"bottom\") + \n",
    "  scale_size(guide=guide_legend(direction=\"vertical\"))\n",
    "\n",
    "sc_3 <- ggplot(so_qc@meta.data, aes(x=log10(nCount_RNA), y=log10(nFeature_RNA), color=pRp_RNA)) + \n",
    "  geom_point() + ggtitle(\"Ribsonmal gene percentage\") + ylab(\"log10(feature count)\") + xlab(\"log10(umi count)\") +\n",
    "  geom_vline(aes(xintercept=log10(nCount_RNA_min)), color=\"red\", linetype=\"longdash\") +\n",
    "  geom_vline(aes(xintercept=log10(nCount_RNA_max)), color=\"red\", linetype=\"longdash\") +\n",
    "  geom_hline(aes(yintercept=log10(nFeature_RNA_min)), color=\"red\", linetype=\"longdash\") + \n",
    "  geom_hline(aes(yintercept=log10(nFeature_RNA_max)), color=\"red\", linetype=\"longdash\") + \n",
    "  facet_grid(tissue~treatment+sample_rep) + theme(aspect.ratio=1, legend.position=\"bottom\") + \n",
    "  scale_size(guide=guide_legend(direction=\"vertical\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d53bcb99-0012-49db-afa4-4f0b46ff3974",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=15, repr.plot.height=10)\n",
    "sc_1 + sc_2 + sc_3 + plot_layout(ncol=2) & theme(legend.position=\"bottom\")\n",
    "ggsave(sc_1, filename=\"result/plot/seurat/sc_mt.png\", width=5, height=5)\n",
    "ggsave(sc_2, filename=\"result/plot/seurat/sc_hg.png\", width=5, height=5)\n",
    "ggsave(sc_3, filename=\"result/plot/seurat/sc_rb.png\", width=5, height=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92568f4e-f83f-4772-8627-b961c30a67be",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_qc <- subset(so_qc, subset=nFeature_RNA >= nFeature_RNA_min & nFeature_RNA <= nFeature_RNA_max & nCount_RNA >= nCount_RNA_min & nCount_RNA <= nCount_RNA_max)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5309164f-aeb4-4999-a06a-f2cc6994fb4c",
   "metadata": {},
   "source": [
    "# Correlate percent Mt with OXPHOS nDNA model \n",
    "\n",
    "Frederick et al., 2020 https://www.nature.com/articles/s41598-020-63448-z"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73380ebe-fd38-4617-838d-2228fb9db530",
   "metadata": {},
   "outputs": [],
   "source": [
    "go_bp <- msigdbr(species=\"mouse\", category=\"C5\", subcategory=\"BP\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59b4ff17-41e4-4798-8a49-3929b10c5d2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Positive regulation of glycolytic process (GO:0045821)\n",
    "go_bp_gly <- go_bp[go_bp$gs_exact_source == \"GO:0045821\", ]$gene_symbol\n",
    "\n",
    "# Positive regulation of oxidative phosphorylation (GO:1903862)\n",
    "# Not found in msigdbr getting genes manually from http://www.informatics.jax.org/go/term/GO:1903862\n",
    "go_bp_op <- c(\"Gadd45gip1\", \"Lexm\", \"Myc\", \"Myog\", \"Nupr1\", \"Vcp\")\n",
    "go_bp_op <- go_bp_op[go_bp_op %in% rownames(so_qc)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57670932-7639-420e-b395-03bd59e9b468",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)\n",
    "so_qc <- AddModuleScore(so_qc, features=list(go_bp_gly), assays=\"RNA\", slot=\"data\", ctrl=100, nbin=25, name=\"msGly_RNA\")\n",
    "so_qc <- AddModuleScore(so_qc, features=list(go_bp_op), assays=\"RNA\", slot=\"data\", ctrl=100, nbin=25, name=\"msOx_RNA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d891fc34-e087-424c-9e36-84c1e745b595",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc_1 <- ggplot(so_qc@meta.data, aes(x=pMt_RNA, y=msGly_RNA1, color=log10(nFeature_RNA))) + \n",
    "    geom_point() + ggtitle(\"Glycolytic modele score vs %MT\") + ylab(\"Module score Glycolysis\") + xlab(\"%Mt\") +\n",
    "    geom_vline(aes(xintercept=pMt_RNA_max), color=\"red\", linetype=\"longdash\") +\n",
    "    facet_grid(tissue~treatment+sample_rep) + theme(aspect.ratio=1, legend.position=\"bottom\") + \n",
    "    scale_size(guide=guide_legend(direction=\"vertical\"))\n",
    "\n",
    "sc_2 <- ggplot(so_qc@meta.data, aes(x=pMt_RNA, y=msOx_RNA1, color=log10(nFeature_RNA))) + \n",
    "    geom_point() + ggtitle(\"OXPHOS module score vs %MT\") + ylab(\"Module score OXPHOS\") + xlab(\"%Mt\") +\n",
    "    geom_vline(aes(xintercept=pMt_RNA_max), color=\"red\", linetype=\"longdash\") +\n",
    "    facet_grid(tissue~treatment+sample_rep) + theme(aspect.ratio=1, legend.position=\"bottom\") + \n",
    "    scale_size(guide=guide_legend(direction=\"vertical\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f61e103-ddc4-4b45-acb7-e335dce6af2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=15, repr.plot.height=5)\n",
    "ggsave(sc_1, filename=\"result/plot/seurat/sc_mt_glycolytic.png\", width=5, height=5)\n",
    "ggsave(sc_2, filename=\"result/plot/seurat/sc_mt_oxphos.png\", width=5, height=5)\n",
    "sc_1 + sc_2 + plot_layout(ncol=2) & theme(legend.position=\"bottom\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70f9234b-3b10-48d7-ad75-9b3c0a101ca1",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_qc <- subset(so_qc, subset=pMt_RNA <= pMt_RNA_max)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d45e4c7b-79bc-4789-8d86-c44a126fd491",
   "metadata": {},
   "source": [
    "# Box plots "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7751f3ad-a731-41d6-b096-c48d2b4c09fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "qc_vln_FUN <- function(data, y, fill, ylab=\"\", ymin, ymax) {\n",
    "  \n",
    "  vln_plot_1 <- ggplot(data, aes(x=treatment, y={{y}}, color={{fill}})) + \n",
    "    geom_jitter(alpha=0.2, shape=16, color=\"gray\") + \n",
    "    geom_boxplot(alpha=1.0) + xlab(\"\") + \n",
    "    ylim(ymin, ymax) +\n",
    "    scale_color_manual(values=color$tissue) + \n",
    "    ggtitle(\"GEM\") + \n",
    "    facet_wrap(~tissue, scales=\"free_x\") + ylab(ylab) +\n",
    "    theme(\n",
    "      plot.title=element_text(size=12, face=\"bold\", margin=margin(t=0, r=0, b=5, l=0)), \n",
    "      axis.text.x=element_text(angle=45, vjust=1, hjust=1), \n",
    "      strip.text=element_blank()\n",
    "      )\n",
    "  \n",
    "  vln_plot_2 <- ggplot(data[data$qc_class == \"pass\", ], aes(x=treatment, y={{y}}, color={{fill}})) +\n",
    "    geom_jitter(alpha=0.2, shape=16, color=\"gray\") +\n",
    "    geom_boxplot(alpha=1.0) + xlab(\"\") + \n",
    "    ylim(ymin, ymax) +\n",
    "    scale_color_manual(values=color$tissue) +\n",
    "    ggtitle(\"Filtered\") +\n",
    "    facet_wrap(~tissue, scales=\"free_x\") + ylab(ylab) + \n",
    "    theme(\n",
    "      plot.title=element_text(size=12, face=\"bold\", margin=margin(t=0, r=0, b=5, l=0)), \n",
    "      axis.text.x=element_text(angle=45, vjust=1, hjust=1), \n",
    "      strip.text=element_blank()\n",
    "      )\n",
    "\n",
    "  return(list(vln_plot_1, vln_plot_2))\n",
    "}\n",
    "\n",
    "data <- dplyr::filter(so_raw@meta.data, cellranger_class == \"Cell\")\n",
    "\n",
    "qc_vln_1 <- qc_vln_FUN(data, nCount_RNA, tissue, \"UMI [count]\",  ymin=0, ymax=max(so_raw$nCount_RNA)+100)\n",
    "qc_vln_2 <- qc_vln_FUN(data, nFeature_RNA, tissue, \"Feature [count]\", ymin=0, ymax=max(so_raw$nFeature_RNA)+100)\n",
    "qc_vln_3 <- qc_vln_FUN(data, pMt_RNA, tissue, \"Mt [%]\", ymin=0, ymax=100)\n",
    "qc_vln_4 <- qc_vln_FUN(data, pHb_RNA, tissue, \"Hb [%]\", ymin=0, ymax=100)\n",
    "qc_vln_5 <- qc_vln_FUN(data, pRp_RNA, tissue, \"Rbl [%]\", ymin=0, ymax=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75689b90-0cdf-43f8-99c0-f2eb246db1a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=15, repr.plot.height=10)\n",
    "qc_vln_1[[1]] + qc_vln_1[[2]] + qc_vln_2[[1]]  + qc_vln_2[[2]] + \n",
    "qc_vln_3[[1]] + qc_vln_3[[2]] + qc_vln_4[[1]]  + qc_vln_4[[2]] + \n",
    "qc_vln_5[[1]] + qc_vln_5[[2]] + plot_spacer() + plot_spacer() + plot_layout(guides=\"collect\", ncol=4) & theme(legend.position=\"bottom\")\n",
    "\n",
    "ggsave(qc_vln_1[[1]] + qc_vln_1[[2]] + plot_layout(guides=\"collect\", ncol=2) & theme(legend.position=\"bottom\"), filename=\"result/plot/seurat/qc_vln_1.png\", width=5, height=2.5)\n",
    "ggsave(qc_vln_2[[1]] + qc_vln_2[[2]] + plot_layout(guides=\"collect\", ncol=2) & theme(legend.position=\"bottom\"), filename=\"result/plot/seurat/qc_vln_2.png\", width=5, height=2.5)\n",
    "ggsave(qc_vln_3[[1]] + qc_vln_3[[2]] + plot_layout(guides=\"collect\", ncol=2) & theme(legend.position=\"bottom\"), filename=\"result/plot/seurat/qc_vln_3.png\", width=5, height=2.5)\n",
    "ggsave(qc_vln_4[[1]] + qc_vln_4[[2]] + plot_layout(guides=\"collect\", ncol=2) & theme(legend.position=\"bottom\"), filename=\"result/plot/seurat/qc_vln_4.png\", width=5, height=2.5)\n",
    "ggsave(qc_vln_5[[1]] + qc_vln_5[[2]] + plot_layout(guides=\"collect\", ncol=2) & theme(legend.position=\"bottom\"), filename=\"result/plot/seurat/qc_vln_5.png\", width=5, height=2.5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "011464ba-92dd-4896-a79e-dc1fbf15b121",
   "metadata": {},
   "source": [
    "# SingleR annotation\n",
    "SingleR identifies marker genes from the reference and uses them to compute assignment scores (based on the Spearman correlation across markers) for each cell in the test dataset against each label in the reference. The label with the highest score is the assigned to the test cell, possibly with further fine-tuning to resolve closely related labels.  \n",
    "\n",
    "first.labels: Labels before fine-tuning  \n",
    "labels: Labels after fine-tuning  \n",
    "pruning: labels after pruning  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26806a43-2235-404f-a4c6-c40ba3944dea",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref <- readRDS(\"data/haemosphere/se_haemosphere.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59158a8c-f089-40c9-a5c5-44fd044931ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Seurat object to SingleCellExperiment\n",
    "sce <- SingleCellExperiment(list(counts=so_qc@assays$RNA@counts))\n",
    "    \n",
    "# Predict labels\n",
    "label_main <- SingleR::SingleR(test=sce, ref=ref, labels=ref$label.main, assay.type.test=\"counts\", de.method=\"classic\")\n",
    "label_fine <- SingleR::SingleR(test=sce, ref=ref, labels=ref$label.fine, assay.type.test=\"counts\", de.method=\"classic\")\n",
    "\n",
    "# Save predicted labels \n",
    "saveRDS(label_main, \"result/label_main.rds\")\n",
    "saveRDS(label_fine, \"result/label_fine.rds\")\n",
    "\n",
    "# Add labels to Seurat object\n",
    "label_main_meta <- as.data.frame(label_main) %>% dplyr::select(pruned.labels, tuning.scores.first) %>% dplyr::rename(main_labels=pruned.labels, main_delta_score=tuning.scores.first)\n",
    "label_fine_meta <- as.data.frame(label_fine) %>% dplyr::select(pruned.labels, tuning.scores.first) %>% dplyr::rename(fine_labels=pruned.labels, fine_delta_score=tuning.scores.first)\n",
    "\n",
    "so_qc <- AddMetaData(so_qc, cbind(label_main_meta, label_fine_meta))\n",
    "\n",
    "# Set factor level for labels \n",
    "so_qc$main_labels <- factor(so_qc$main_labels, levels=names(color$main_labels_haemosphere))\n",
    "so_qc$fine_labels <- factor(so_qc$fine_labels, levels=names(color$fine_labels_haemosphere))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e35aebc6-1437-4f71-a1e4-12c0fb271b62",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_qc$main_labels <- factor(so_qc$main_labels, levels=names(color$main_labels_haemosphere))\n",
    "so_qc$fine_labels <- factor(so_qc$fine_labels, levels=names(color$fine_labels_haemosphere))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe6dd8a3-df97-4e07-9684-904f13709348",
   "metadata": {},
   "source": [
    "# Ery score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ffe36fc-9571-4d7d-b4ec-4624e59acf3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_qc$ery_score=label_main$scores[, \"Ery\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18d80cc7-baea-49d5-965d-00bcb571dfee",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_qc$main_labels <- factor(so_qc$main_labels, levels=names(color$main_labels_haemosphere))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6438898-6414-4d73-a505-6a41b482b103",
   "metadata": {},
   "outputs": [],
   "source": [
    "boxplot_ery_score <- ggplot(so_qc@meta.data, aes(x=ery_score, color=main_labels)) + \n",
    "    geom_boxplot() + \n",
    "    scale_color_manual(values=color$main_labels_haemosphere) + \n",
    "    guides(colour=guide_legend(reverse=TRUE)) + \n",
    "    facet_grid(~tissue, scales=\"free_y\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "840cf28d-d516-4eaa-8818-2457b25e3d14",
   "metadata": {},
   "outputs": [],
   "source": [
    "count_module <- function(so, pattern, assay=\"RNA\", slot=\"counts\", stat=\"sum\") {\n",
    "    \n",
    "    genes <- rownames(so)[grep(pattern, rownames(so))]\n",
    "    cnt <- GetAssayData(so, assay=assay, slot=slot)\n",
    "    cnt <- as.matrix(cnt)\n",
    "    cnt <- cnt[rownames(cnt) %in% genes, ]\n",
    "    \n",
    "    if(stat==\"sum\"){cnt_stat <- colSums(cnt)}\n",
    "\n",
    "    return(cnt_stat)\n",
    "}\n",
    "so_qc$cHb_RNA <- count_module(so_qc, pattern=\"Hba-|Hbb-|Hbq1b|Hbq1a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b79d128-038e-424a-8ae9-4fd53f4fd7bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "boxplot_ery_count_module <- ggplot(so_qc@meta.data, aes(x=log10(cHb_RNA+1), color=main_labels)) + \n",
    "    geom_boxplot() + \n",
    "    scale_color_manual(values=color$main_labels_haemosphere) + \n",
    "    guides(colour=guide_legend(reverse=TRUE)) +\n",
    "    facet_grid(~tissue, scales=\"free_y\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bfa8879-3f7d-4db1-826a-441ce8e3779f",
   "metadata": {},
   "outputs": [],
   "source": [
    "boxplot_ery_count_feature <- ggplot(so_qc@meta.data, aes(x=nFeature_RNA, y=log10(cHb_RNA), color=main_labels)) + \n",
    "    geom_point() + \n",
    "    geom_hline(aes(yintercept=log10(10)), color=\"red\", linetype=\"longdash\") +\n",
    "    scale_color_manual(values=color$main_labels_haemosphere) + \n",
    "    guides(colour=guide_legend(reverse=TRUE)) +\n",
    "    facet_grid(~tissue, scales=\"free_y\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ad974da-9ae0-4ebc-a2a0-985d770a3c58",
   "metadata": {},
   "outputs": [],
   "source": [
    "boxplot_ery_count_umi <- ggplot(so_qc@meta.data, aes(x=nCount_RNA, y=log10(cHb_RNA), color=main_labels)) + \n",
    "    geom_point() + \n",
    "    geom_hline(aes(yintercept=log10(10)), color=\"red\", linetype=\"longdash\") +\n",
    "    scale_color_manual(values=color$main_labels_haemosphere) + \n",
    "    guides(colour=guide_legend(reverse=TRUE)) +\n",
    "    facet_grid(~tissue, scales=\"free_y\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57b89752-ba0d-4f79-815c-8f2d1a68a9e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10, repr.plot.height=10, warn=-1)\n",
    "boxplot_ery_score + boxplot_ery_count_module + boxplot_ery_count_feature + boxplot_ery_count_umi + plot_layout(guides=\"collect\", ncol=2) & theme(legend.position=\"bottom\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b826d61a-51e9-4fa5-abc5-40263256927f",
   "metadata": {},
   "outputs": [],
   "source": [
    "main_labels_split <- split(label_main, f=so_qc$sample_group)\n",
    "main_labels_score_hm <- list()\n",
    "for(i in names(main_labels_split)) {main_labels_score_hm[[i]] <- singler_score_hm(main_labels_split[[i]], color=color$main_labels_haemosphere, main=i)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ae9ed63-3991-4906-bee5-8d5fc7cdf4b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=20, repr.plot.height=20)\n",
    "main_labels_score_hm_grid <- gridExtra::arrangeGrob(grobs=main_labels_score_hm, ncol=1) \n",
    "ggsave(main_labels_score_hm_grid, filename=\"result/plot/seurat/main_labels_score_hm_grid.png\", width=20, height=20)\n",
    "grid::grid.draw(main_labels_score_hm_grid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ef17af6-f4a4-4bc8-b0db-c9e6ac148286",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_qc$ery_qc <- ifelse(!(so_qc$tissue == \"Myeloid\" & so_qc$main_labels == \"Ery\"), \"pass\", \"fail\")\n",
    "so_qc <- subset(so_qc, subset = ery_qc == \"pass\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ab1daea-dc01-4bb1-9079-b509cd80aee1",
   "metadata": {},
   "source": [
    "# Compute module scores "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "508bea14-7e97-4ff4-9934-80b8bd23409b",
   "metadata": {},
   "source": [
    "## Histocompatibility Class I\n",
    "H2-K1 - Class I histocompatibility antigen, kappa-B alpha chain   \n",
    "H2-D1 - Class I histocompatibility antigen, D-B alpha chain   \n",
    "H2-L1 - not found   \n",
    "\n",
    "## Histocompatibility Class II\n",
    "H2-Aa - Class II histocompatibility antigen, A-B alpha chain   \n",
    "H2-Ab1 - Class II histocompatibility antigen, A beta chain   \n",
    "\n",
    "H2-Eb1 - Class II histocompatibility antigen, I-E beta chain   \n",
    "H2-Eb2 - Class II histocompatibility antigen, E beta chain  \n",
    "\n",
    "Info: The antigens presented by class I peptides are derived from cytosolic proteins while calss II derived from extracellular proteins. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac2af910-596b-4945-918f-05aa7a2fb212",
   "metadata": {},
   "outputs": [],
   "source": [
    "mhcI_genes <- c(\"H2-K1\", \"H2-D1\")\n",
    "mhcII_genes <- c(\"H2-Aa\", \"H2-Ab1\", \"H2-Eb1\", \"H2-Eb2\")\n",
    "\n",
    "so_qc <- AddModuleScore(so_qc, features=list(mhcI_genes), assays=\"RNA\", slot=\"data\", ctrl=100, nbin=25, name=\"msMHCI_RNA\")\n",
    "so_qc <- AddModuleScore(so_qc, features=list(mhcII_genes), assays=\"RNA\", slot=\"data\", ctrl=100, nbin=25, name=\"msMHCII_RNA\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c07929c2-1725-44d2-b5e3-103447bf6bb8",
   "metadata": {},
   "source": [
    "## Surface marker \n",
    "\n",
    "Cd4 - T-cell surface glycoprotein CD4, Co-receptor for MHC class II, active on T-helper cells and triggers differentiation of monocytes into functional mature macrophages.     \n",
    "Cd8 - T-cell surface glycoprotein CD8, Co-receptor for MHC class I, CTL activattion and thymic selection of Cd8+ T cells. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94d60396-1353-4ded-a979-04758af8c316",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd4_genes <- c(\"Cd4\") \n",
    "cd8_genes <- c(\"Cd8a\", \"Cd8b1\")\n",
    "\n",
    "so_qc <- AddModuleScore(so_qc, features=list(cd4_genes), assays=\"RNA\", slot=\"data\", ctrl=100, nbin=25, name=\"msCd4_RNA\")\n",
    "so_qc <- AddModuleScore(so_qc, features=list(cd8_genes), assays=\"RNA\", slot=\"data\", ctrl=100, nbin=25, name=\"msCd8_RNA\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b350ba5-d6cd-49f1-a50e-a306f7140e4a",
   "metadata": {},
   "source": [
    "## T-cell TCR-CD3 complex \n",
    "Trac - T cell receptor alpha constant  \n",
    "Trbc1 - T cell receptor beta constant 1  \n",
    "Trbc2 - T cell receptor beta constant 2  \n",
    "Trdc - T cell receptor delta constant  \n",
    "Trgc1 - T cell receptor gamma constant 1  \n",
    "Trgc2 - T cell receptor gamma constant 2  \n",
    "\n",
    "Cd247 - T-cell surface glycoprotein Cd3 zeta chain (Cd3z)  \n",
    "Cd3g - T-cell surface glycoprotein Cd3 gamma chain   \n",
    "Cd3e - T-cell surface glycoprotein Cd3 epsilon chain   \n",
    "Cd3d - T-cell surface glycoprotein Cd3 delta chain   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d79c1b5c-3c54-4492-8259-bf1fc003dfa9",
   "metadata": {},
   "outputs": [],
   "source": [
    "tcr_genes <- c(\"Trac\", \"Trbc1\", \"Trbc2\", \"Trdc\", \"Trgc1\", \"Trgc2\")\n",
    "tcr_cd3_genes <- c(\"Cd247\", \"Cd3g\", \"Cd3e\", \"Cd3d\")\n",
    "\n",
    "so_qc <- AddModuleScore(so_qc, features=list(tcr_genes), assays=\"RNA\", slot=\"data\", ctrl=100, nbin=25, name=\"msTcr_RNA\")\n",
    "so_qc <- AddModuleScore(so_qc, features=list(tcr_cd3_genes), assays=\"RNA\", slot=\"data\", ctrl=100, nbin=25, name=\"msTcr_cd3_RNA\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6aaeee5a-9e5b-480d-932e-4c1749f4bc91",
   "metadata": {},
   "source": [
    "## B-cell Immonogloblin complex \n",
    "Naive B-cells produce the following Ig classes:  \n",
    "Ighm - Immunoglobulin heavy constant mu (naive B-cells)  \n",
    "Ighd - Immunoglobulin heavy constant delta (naive B-cells)  \n",
    "\n",
    "Through isotope switching the following Ig classes can be produced:   \n",
    "Ighg1 - Immunoglobulin heavy constant gamma (Mouse with Igh1b have Igg2c isotope instead of Igg2a)  \n",
    "Ighg2a - Immunoglobulin heavy constant gamma (NA)  \n",
    "Ighg2b - Immunoglobulin heavy constant gamma  \n",
    "Ighg2c - Immunoglobulin heavy constant gamma   \n",
    "Ighg3 - Immunoglobulin heavy constant gamma   \n",
    "Igha - Immunoglobulin heavy constant alpha   \n",
    "Ighe - Immunoglobulin heavy constant epsilon (NA)  \n",
    "\n",
    "Igkc - Immunoglobulin kappa constant (light chain)  \n",
    "Iglc - Immunoglobulin lambda constant (light chain)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bc6fd08-af4d-473a-a137-e2a7c6410afa",
   "metadata": {},
   "outputs": [],
   "source": [
    "Ighm_genes <- c(\"Ighm\")\n",
    "Ighd_genes <- c(\"Ighd\")\n",
    "Ighg_genes <- c(\"Ighg1\", \"Ighg2a\", \"Ighg2b\", \"Ighg2c\", \"Ighg3\")\n",
    "Igha_genes <- c(\"Igha\")\n",
    "Ighe_genes <- c(\"Ighe\") \n",
    "Igkc_genes <- c(\"Igkc\")\n",
    "Iglc_genes <- c(\"Iglc1\", \"Iglc2\", \"Iglc3\")\n",
    "\n",
    "so_qc <- AddModuleScore(so_qc, features=list(Ighm_genes), assays=\"RNA\", slot=\"data\", ctrl=100, nbin=25, name=\"msIghm_RNA\")\n",
    "so_qc <- AddModuleScore(so_qc, features=list(Ighd_genes), assays=\"RNA\", slot=\"data\", ctrl=100, nbin=25, name=\"msIghd_RNA\")\n",
    "so_qc <- AddModuleScore(so_qc, features=list(Ighg_genes), assays=\"RNA\", slot=\"data\", ctrl=100, nbin=25, name=\"msIghg_RNA\")\n",
    "so_qc <- AddModuleScore(so_qc, features=list(Igha_genes), assays=\"RNA\", slot=\"data\", ctrl=100, nbin=25, name=\"msIgha_RNA\")\n",
    "#so_qc <- AddModuleScore(so_qc, features=list(Ighe_genes), assays=\"RNA\", slot=\"data\", ctrl=100, nbin=25, name=\"msIghe_RNA\")\n",
    "so_qc <- AddModuleScore(so_qc, features=list(Igkc_genes), assays=\"RNA\", slot=\"data\", ctrl=100, nbin=25, name=\"msIgkc_RNA\")\n",
    "so_qc <- AddModuleScore(so_qc, features=list(Iglc_genes), assays=\"RNA\", slot=\"data\", ctrl=100, nbin=25, name=\"msIglc_RNA\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0dd3a0ba-447f-4dfc-8598-237d6e7a4993",
   "metadata": {},
   "source": [
    "# Compute cell cycle score "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "797eae62-410a-419e-b9cb-9856ad2a01a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "cc_genes_seurat_s <- cc.genes.updated.2019$s.genes\n",
    "cc_genes_seurat_g2m <- cc.genes.updated.2019$g2m.genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9acdc28-b1ce-4a07-8bc6-758b9bac94f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Translate human to mouse \n",
    "library(biomaRt) \n",
    "\n",
    "# Set ssl for biomart  \n",
    "httr::set_config(httr::config(ssl_verifypeer=FALSE))\n",
    "\n",
    "ensembl_hgnc <- useMart(\"ensembl\", dataset=\"hsapiens_gene_ensembl\") \n",
    "ensembl_mm <- useMart(\"ensembl\", dataset=\"mmusculus_gene_ensembl\")\n",
    "\n",
    "cc_genes_seurat_s <- getLDS(attributes=c(\"hgnc_symbol\"), filters=\"hgnc_symbol\", values=cc_genes_seurat_s, mart=ensembl_hgnc, attributesL=c(\"mgi_symbol\"), martL=ensembl_mm, uniqueRows=T)\n",
    "cc_genes_seurat_s <- cc_genes_seurat_s[, 2]\n",
    "\n",
    "cc_genes_seurat_g2m <- getLDS(attributes=c(\"hgnc_symbol\"), filters=\"hgnc_symbol\", values=cc_genes_seurat_g2m, mart=ensembl_hgnc, attributesL=c(\"mgi_symbol\"), martL=ensembl_mm, uniqueRows=T)\n",
    "cc_genes_seurat_g2m <- cc_genes_seurat_g2m[, 2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54c3d955-6559-46a0-bd00-c6a4e7c5e1bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_qc <- CellCycleScoring(so_qc, s.features=cc_genes_seurat_s, g2m.features=cc_genes_seurat_g2m, set.ident=FALSE)\n",
    "\n",
    "colnames(so_qc@meta.data) <- gsub(\"Phase\", \"cc_phase_class\", colnames(so_qc@meta.data))\n",
    "colnames(so_qc@meta.data) <- gsub(\"S.Score\", \"msS_RNA\", colnames(so_qc@meta.data))\n",
    "colnames(so_qc@meta.data) <- gsub(\"G2M.Score\", \"msG2M_RNA\", colnames(so_qc@meta.data)) \n",
    "    \n",
    "so_qc$msCC_diff_RNA <- so_qc$msS_RNA - so_qc$msG2M_RNA\n",
    "so_qc$cc_phase_class <- factor(so_qc$cc_phase_class, levels=names(color$cc_phase_class))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf402e9d-9cff-4d24-a165-d81884be0291",
   "metadata": {},
   "outputs": [],
   "source": [
    "# saveRDS(so_qc, \"data/object/qc/seurat_qc.rds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60a697ff-3260-4a07-8a17-b3aa00a89269",
   "metadata": {},
   "source": [
    "# SCTransform and dimensional reduction (sample_group)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbc6112c-d3b3-4f04-92b8-b78aa8fedf07",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "options(warn=-1)\n",
    "so_qc_sample_group <- SplitObject(so_qc, split.by=\"sample_group\")\n",
    "so_qc_sample_group <- lapply(so_qc_sample_group, function(so) {\n",
    "    \n",
    "    so <- SCTransform(so, assay=\"RNA\", verbose=FALSE)\n",
    "    so <- RunPCA(so, npcs=100, verbose=FALSE)\n",
    "    so <- FindNeighbors(so, dims=1:100, k.param=10, verbose=FALSE)\n",
    "    so <- FindClusters(so, verbose=FALSE)\n",
    "    so <- RunUMAP(so, dims=1:100, verbose=FALSE)\n",
    "    \n",
    "    return(so)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a755f46-2f68-4a2c-9cbb-4a021b2194cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# saveRDS(so_qc_sample_group, \"data/object/qc/so_qc_sample_group.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54d08d58-6066-471e-aaa0-4a60ef73c1e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=20, repr.plot.height=20)\n",
    "dim_plot_sample_group <- lapply(so_qc_sample_group, dim_plot, cluster=\"SCT_snn_res.0.8\")\n",
    "\n",
    "ggsave(dim_plot_sample_group[[1]], filename=\"result/plot/seurat/dim_plot_sample_group_m_cpg.png\", width=20, height=20)\n",
    "ggsave(dim_plot_sample_group[[2]], filename=\"result/plot/seurat/dim_plot_sample_group_p_cpg.png\", width=20, height=20)\n",
    "ggsave(dim_plot_sample_group[[3]], filename=\"result/plot/seurat/dim_plot_sample_group_m_nacl.png\", width=20, height=20)\n",
    "ggsave(dim_plot_sample_group[[4]], filename=\"result/plot/seurat/dim_plot_sample_group_p_nacl.png\", width=20, height=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c82661ea-f8df-473d-8413-b79d0cbe9eda",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=20, repr.plot.height=10)\n",
    "\n",
    "so_qc_sample_group_pc_annotation <- lapply(so_qc_sample_group, pc_annotation, dims=100, ident=\"sample_group\")\n",
    "so_qc_sample_group_pc_annotation_plot <- lapply(so_qc_sample_group_pc_annotation, pc_annotation_plot)\n",
    "\n",
    "so_qc_sample_group_pc_annotation_plot <- \n",
    "    so_qc_sample_group_pc_annotation_plot[[1]] + \n",
    "    so_qc_sample_group_pc_annotation_plot[[2]] + \n",
    "    so_qc_sample_group_pc_annotation_plot[[3]] + \n",
    "    so_qc_sample_group_pc_annotation_plot[[4]] + \n",
    "    plot_layout(ncol=2, guides=\"collect\")\n",
    "\n",
    "so_qc_sample_group_pc_annotation_plot\n",
    "ggsave(so_qc_sample_group_pc_annotation_plot, filename=\"result/plot/seurat/so_qc_sample_group_pc_loadings_plot.png\", width=10, height=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96434c79-b809-4b40-b5c6-5d13db6dee39",
   "metadata": {},
   "source": [
    "# Identify outlier cluster"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5eb9613e-c001-4d2c-927c-ca2d3c320f18",
   "metadata": {},
   "source": [
    "## Sample replicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e63016c-adfc-4473-bd67-8d4f722d1a2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10, repr.plot.height=10, warn=-1)\n",
    "cluster_sample_rep <- function(so) {\n",
    "    \n",
    "    box_plot <- ggplot(so@meta.data, aes(x=SCT_snn_res.0.8, fill=tissue, alpha=sample_rep)) + \n",
    "        geom_bar(stat=\"count\", position=\"fill\") + \n",
    "        scale_fill_manual(values=color$tissue) +\n",
    "        scale_alpha_manual(values=c(1, 0.7)) + \n",
    "        ggtitle(paste0(so$tissue[1], so$treatment[1])) + xlab(\"Cluster\") + ylab(\"Cell frequency\") + \n",
    "        facet_grid(~tissue)\n",
    "    \n",
    "    return(box_plot)\n",
    "}\n",
    "\n",
    "cluster_sample_rep_plot <- lapply(so_qc_sample_group, cluster_sample_rep)\n",
    "cluster_sample_rep_plot <- \n",
    "    cluster_sample_rep_plot[[1]] + \n",
    "    cluster_sample_rep_plot[[2]] + \n",
    "    cluster_sample_rep_plot[[3]] + \n",
    "    cluster_sample_rep_plot[[4]] + \n",
    "    plot_layout(ncol=2, guides=\"collect\") & theme(legend.position=\"bottom\")\n",
    "\n",
    "cluster_sample_rep_plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba92e29e-d434-4bdf-bb7c-59119a81ff22",
   "metadata": {},
   "source": [
    "## Myeloid CpG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8c35ebb-f39d-450a-94ba-000f7fa8f502",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Myeloid_CpG\n",
    "so_qc_sample_group_m_cpg <- so_qc_sample_group[[1]]\n",
    "\n",
    "# Cluster 16: B-cell \n",
    "deg_cluster_16 <- FindMarkers(so_qc_sample_group_m_cpg, ident.1=16, only.pos=FALSE, group.by=\"SCT_snn_res.0.8\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f68b77d-b0cd-4dfc-9b5f-e8ae07a69ff8",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=20, repr.plot.height=5)\n",
    "deg_vp_sample_group_m_cpg <- \n",
    "    deg_volcano_plot(deg_cluster_16, title=\"Myeloid (CpG) - Cluster 16\") + \n",
    "    plot_spacer() + plot_spacer() + plot_spacer() +\n",
    "    plot_layout(ncol=4)\n",
    "deg_vp_sample_group_m_cpg\n",
    "ggsave(deg_vp_sample_group_m_cpg, filename=\"result/plot/seurat/deg_vp_sample_group_m_cpg.png\", width=20, height=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db4464f9-0fd9-4853-bcb2-849f8296f88b",
   "metadata": {},
   "source": [
    "## Progenitor CpG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d2f0880-7d0f-4e05-bc4e-7ad29a57d3da",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Progenitor CpG\n",
    "so_qc_sample_group_p_cpg <- so_qc_sample_group[[2]]\n",
    "\n",
    "# Cluster 14: B-cell \n",
    "deg_cluster_14 <- FindMarkers(so_qc_sample_group_p_cpg, ident.1=14, only.pos=FALSE, group.by=\"SCT_snn_res.0.8\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab5d2551-799b-4e3e-9365-128fffbcfad4",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=20, repr.plot.height=5)\n",
    "deg_vp_sample_group_p_cpg <- \n",
    "    deg_volcano_plot(deg_cluster_14, title=\"Progenitor (CpG) - Cluster 14\") +\n",
    "    plot_spacer() + plot_spacer() + plot_spacer() +\n",
    "    plot_layout(ncol=4)\n",
    "deg_vp_sample_group_p_cpg\n",
    "ggsave(deg_vp_sample_group_p_cpg, filename=\"result/plot/seurat/deg_vp_sample_group_p_cpg.png\", width=20, height=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32b7f90b-7d8d-46fe-8ce5-fce93f2dda5a",
   "metadata": {},
   "source": [
    "## Myeloid NaCl "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c534888-2688-4bd6-a8ba-1d469d3c653b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Myeloid NaCl\n",
    "so_qc_sample_group_m_nacl <- so_qc_sample_group[[3]]\n",
    "\n",
    "# Cluster 17: B-cells \n",
    "deg_cluster_17 <- FindMarkers(so_qc_sample_group_m_nacl, ident.1=17, only.pos=FALSE, group.by=\"SCT_snn_res.0.8\")\n",
    "\n",
    "# Cluster 10, 12, 13, 15, 16: Unknown\n",
    "deg_cluster_10 <- FindMarkers(so_qc_sample_group_m_nacl, ident.1=10, only.pos=FALSE, group.by=\"SCT_snn_res.0.8\")\n",
    "deg_cluster_12 <- FindMarkers(so_qc_sample_group_m_nacl, ident.1=12, only.pos=FALSE, group.by=\"SCT_snn_res.0.8\")\n",
    "deg_cluster_13 <- FindMarkers(so_qc_sample_group_m_nacl, ident.1=13, only.pos=FALSE, group.by=\"SCT_snn_res.0.8\")\n",
    "deg_cluster_15 <- FindMarkers(so_qc_sample_group_m_nacl, ident.1=15, only.pos=FALSE, group.by=\"SCT_snn_res.0.8\")\n",
    "deg_cluster_16 <- FindMarkers(so_qc_sample_group_m_nacl, ident.1=16, only.pos=FALSE, group.by=\"SCT_snn_res.0.8\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21617c0f-d7dc-4395-a6a2-84477d830ff7",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=20, repr.plot.height=10)\n",
    "deg_vp_sample_group_m_nacl <- \n",
    "    deg_volcano_plot(deg_cluster_10, title=\"Myeloid (NaCl) - Cluster 10\") +\n",
    "    deg_volcano_plot(deg_cluster_12, title=\"Myeloid (NaCl) - Cluster 12\") +\n",
    "    deg_volcano_plot(deg_cluster_13, title=\"Myeloid (NaCl) - Cluster 13\") +\n",
    "    deg_volcano_plot(deg_cluster_15, title=\"Myeloid (NaCl) - Cluster 15\") +\n",
    "    deg_volcano_plot(deg_cluster_16, title=\"Myeloid (NaCl) - Cluster 16\") +\n",
    "    deg_volcano_plot(deg_cluster_17, title=\"Myeloid (NaCl) - Cluster 17\") +\n",
    "    plot_spacer() + plot_spacer() +    \n",
    "    plot_layout(ncol=4)\n",
    "deg_vp_sample_group_m_nacl\n",
    "ggsave(deg_vp_sample_group_m_nacl, filename=\"result/plot/seurat/deg_vp_sample_group_m_nacl.png\", width=20, height=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "096c6ef2-4f15-4055-a3c5-6aebd153d827",
   "metadata": {},
   "source": [
    "## Progenitor NaCl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5c26b44-d714-483f-a9d2-2f9fa7f2bc91",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Myeloid NaCl\n",
    "so_qc_sample_group_p_nacl <- so_qc_sample_group[[4]]\n",
    "\n",
    "# Cluster 11: T-cell  \n",
    "deg_cluster_11 <- FindMarkers(so_qc_sample_group_p_nacl, ident.1=11, only.pos=FALSE, group.by=\"SCT_snn_res.0.8\")\n",
    "\n",
    "# Cluster 15: B-cell  \n",
    "deg_cluster_15 <- FindMarkers(so_qc_sample_group_p_nacl, ident.1=15, only.pos=FALSE, group.by=\"SCT_snn_res.0.8\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f98f717-ed77-43c2-8fa2-0098afcc6dab",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=20, repr.plot.height=5)\n",
    "deg_vp_sample_group_p_nacl <- \n",
    "    deg_volcano_plot(deg_cluster_11, title=\"Progenitor (NaCl) - Cluster 11\") +\n",
    "    deg_volcano_plot(deg_cluster_15, title=\"Progenitor (NaCl) - Cluster 15\") +\n",
    "    plot_spacer() + plot_spacer() + \n",
    "    plot_layout(ncol=4)\n",
    "deg_vp_sample_group_p_nacl\n",
    "ggsave(deg_vp_sample_group_p_nacl, filename=\"result/plot/seurat/deg_vp_sample_group_p_nacl.png\", width=20, height=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abaf728f-5710-4399-9342-c96ddc5d424d",
   "metadata": {},
   "source": [
    "## Remove cluster from so_qc (sample group)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3671a881-ef8b-4424-80ef-402d771ecfef",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_qc <- merge(so_qc_sample_group[[1]], so_qc_sample_group[2:4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5416ac47-46af-4baa-a9dc-d568be65b6ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_qc <- subset(so_qc, subset=!(\n",
    "    (tissue == \"Myeloid\" & treatment == \"CpG\" &  SCT_snn_res.0.8 %in% c(16)) | \n",
    "    (tissue == \"Progenitor\" & treatment == \"CpG\" & SCT_snn_res.0.8 %in% c(14)) | \n",
    "    (tissue == \"Myeloid\" & treatment == \"NaCl\" & SCT_snn_res.0.8 %in% c(15, 13, 17)) | \n",
    "    (tissue == \"Progenitor\" & treatment == \"NaCl\" & SCT_snn_res.0.8 %in% c(11, 15))\n",
    ")\n",
    "      )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b58650a-0e2f-4ed5-bc4b-300e3da14495",
   "metadata": {},
   "source": [
    "# SCTransform and dimensional reduction (treatment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9c09460-6545-4c80-9330-9f0094b48935",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)\n",
    "so_qc_treatment <- SplitObject(so_qc, split.by=\"treatment\")\n",
    "so_qc_treatment <- lapply(so_qc_treatment, function(so) {\n",
    "    \n",
    "    so <- SCTransform(so, assay=\"RNA\", verbose=FALSE)\n",
    "    so <- RunPCA(so, npcs=100, verbose=FALSE)\n",
    "    so <- FindNeighbors(so, dims=1:100, verbose=FALSE)\n",
    "    so <- FindClusters(so, verbose=FALSE)\n",
    "    so <- RunUMAP(so, dims=1:100, verbose=FALSE)\n",
    "    \n",
    "    return(so)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a2acf2b-aa6a-403e-af8e-19fe1860cbf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# saveRDS(so_qc_treatment, \"data/object/qc/so_qc_treatment.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d343d826-264b-4a41-bf8f-31a788958761",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=20, repr.plot.height=20)\n",
    "dim_plot_treatment <- lapply(so_qc_treatment, dim_plot, cluster=\"SCT_snn_res.0.8\")\n",
    "\n",
    "ggsave(dim_plot_treatment[[1]], filename=\"result/plot/seurat/dim_plot_treatment_cpg.png\", width=20, height=20)\n",
    "ggsave(dim_plot_treatment[[2]], filename=\"result/plot/seurat/dim_plot_treatment_nacl.png\", width=20, height=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63454f1a-2d8d-4a10-b96b-fa49356d9d60",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=20, repr.plot.height=5)\n",
    "\n",
    "so_qc_treatment_pc_annotation <- lapply(so_qc_treatment, pc_annotation, dims=100, ident=\"treatment\")\n",
    "so_qc_treatment_pc_annotation_plot <- lapply(so_qc_treatment_pc_annotation, pc_annotation_plot)\n",
    "\n",
    "so_qc_treatment_pc_annotation_plot <- \n",
    "    so_qc_treatment_pc_annotation_plot[[1]] + \n",
    "    so_qc_treatment_pc_annotation_plot[[2]] + \n",
    "    plot_layout(ncol=2, guides=\"collect\")\n",
    "\n",
    "so_qc_treatment_pc_annotation_plot\n",
    "ggsave(so_qc_treatment_pc_annotation_plot, filename=\"result/plot/seurat/so_qc_treatment_pca_loadings_plot.png\", width=10, height=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1924fdf7-059b-4ae5-b66a-e03a8768eff6",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=20, repr.plot.height=5, warn=-1)\n",
    "cluster_sample_rep <- function(so) {\n",
    "    \n",
    "    box_plot <- ggplot(so@meta.data, aes(x=SCT_snn_res.0.8, fill=tissue, alpha=sample_rep)) + \n",
    "        geom_bar(stat=\"count\", position=\"fill\") + \n",
    "        scale_fill_manual(values=color$tissue) +\n",
    "        scale_alpha_manual(values=c(1, 0.7)) + \n",
    "        ggtitle(so$treatment[1]) + xlab(\"Cluster\") + ylab(\"Cell frequency\") + \n",
    "        facet_grid(~tissue)\n",
    "    \n",
    "    return(box_plot)\n",
    "}\n",
    "\n",
    "cluster_sample_rep_plot <- lapply(so_qc_treatment, cluster_sample_rep)\n",
    "cluster_sample_rep_plot <- \n",
    "    cluster_sample_rep_plot[[1]] + \n",
    "    cluster_sample_rep_plot[[2]] + \n",
    "    plot_layout(ncol=2, guides=\"collect\") & theme(legend.position=\"bottom\")\n",
    "cluster_sample_rep_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26ffd0a8-10a9-46b7-a96c-cda7030049c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10, repr.plot.height=5)\n",
    "\n",
    "cluster_tissue_1 <- ggplot(so_qc_treatment[[1]]@meta.data, aes(x=SCT_snn_res.0.8, fill=tissue)) + \n",
    "      geom_bar(stat=\"count\", position=\"fill\") + \n",
    "      scale_fill_manual(values=color$tissue) +\n",
    "      ggtitle(\"CpG sample\") + xlab(\"Cluster\") + ylab(\"Cell frequency\") + \n",
    "      theme(legend.position=\"bottom\")\n",
    "\n",
    "cluster_tissue_2 <- ggplot(so_qc_treatment[[2]]@meta.data, aes(x=SCT_snn_res.0.8, fill=tissue)) + \n",
    "      geom_bar(stat=\"count\", position=\"fill\") + \n",
    "      scale_fill_manual(values=color$tissue) +\n",
    "      ggtitle(\"NaCl sample\") + xlab(\"Cluster\") + ylab(\"Cell frequency\") + \n",
    "      theme(legend.position=\"bottom\")\n",
    "\n",
    "cluster_tissue <- cluster_tissue_1 + cluster_tissue_2\n",
    "cluster_tissue\n",
    "ggsave(cluster_tissue, filename=\"result/plot/seurat/cluster_tissue_treatment.png\", width=10, height=2.5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1aff7c85-a060-4737-9c5c-8bd33649e08a",
   "metadata": {},
   "source": [
    "# SCTransform + Cell cycle Regression and dimensional reduction (treatment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f74ee6fe-fb3b-4820-b0d9-5679b30a411f",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)\n",
    "so_qc_treatment_reg <- SplitObject(so_qc, split.by=\"treatment\")\n",
    "so_qc_treatment_reg <- lapply(so_qc_treatment_reg, function(so) {\n",
    "    \n",
    "    so <- SCTransform(so, assay=\"RNA\", vars.to.regress=c(\"msCC_diff_RNA\"), verbose=FALSE)\n",
    "    so <- RunPCA(so, npcs=100, verbose=FALSE)\n",
    "    so <- FindNeighbors(so, dims=1:100, verbose=FALSE)\n",
    "    so <- FindClusters(so, verbose=FALSE)\n",
    "    so <- RunUMAP(so, dims=1:100, verbose=FALSE)\n",
    "    \n",
    "    return(so)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67a8f26d-c9d5-415d-bd11-21468ef216ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(so_qc_treatment_reg, \"data/object/qc/seurat_sct_treatment.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "058a548b-b456-40b6-ae55-ba2fae7c79bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=20, repr.plot.height=20)\n",
    "dim_plot_treatment_reg <- lapply(so_qc_treatment_reg, dim_plot, cluster=\"SCT_snn_res.0.8\")\n",
    "\n",
    "ggsave(dim_plot_treatment_reg[[1]], filename=\"result/plot/seurat/dim_plot_treatment_reg_cpg.png\", width=20, height=20)\n",
    "ggsave(dim_plot_treatment_reg[[2]], filename=\"result/plot/seurat/dim_plot_treatment_reg_nacl.png\", width=20, height=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "793e572f-5b72-4417-96ac-c83c9e8c5af3",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=20, repr.plot.height=5)\n",
    "\n",
    "so_qc_treatment_reg_pc_annotation <- lapply(so_qc_treatment_reg, pc_annotation, dims=100, ident=\"treatment\")\n",
    "so_qc_treatment_reg_pc_annotation_plot <- lapply(so_qc_treatment_reg_pc_annotation, pc_annotation_plot)\n",
    "\n",
    "so_qc_treatment_reg_pc_annotation_plot <- \n",
    "    so_qc_treatment_reg_pc_annotation_plot[[1]] + \n",
    "    so_qc_treatment_reg_pc_annotation_plot[[2]] + \n",
    "    plot_layout(ncol=2, guides=\"collect\")\n",
    "\n",
    "so_qc_treatment_reg_pc_annotation_plot\n",
    "ggsave(so_qc_treatment_reg_pc_annotation_plot, filename=\"result/plot/seurat/so_qc_treatment_reg_pca_loadings_plot.png\", width=10, height=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "758bb3a7-5751-4b06-a885-fb37ccde7c2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=20, repr.plot.height=5, warn=-1)\n",
    "cluster_sample_rep <- function(so) {\n",
    "    \n",
    "    box_plot <- ggplot(so@meta.data, aes(x=SCT_snn_res.0.8, fill=tissue, alpha=sample_rep)) + \n",
    "        geom_bar(stat=\"count\", position=\"fill\") + \n",
    "        scale_fill_manual(values=color$tissue) +\n",
    "        scale_alpha_manual(values=c(1, 0.7)) + \n",
    "        ggtitle(so$treatment[1]) + xlab(\"Cluster\") + ylab(\"Cell frequency\") + \n",
    "        facet_grid(~tissue)\n",
    "    \n",
    "    return(box_plot)\n",
    "}\n",
    "\n",
    "cluster_sample_rep_plot <- lapply(so_qc_treatment_reg, cluster_sample_rep)\n",
    "cluster_sample_rep_plot <- \n",
    "    cluster_sample_rep_plot[[1]] + \n",
    "    cluster_sample_rep_plot[[2]] + \n",
    "    plot_layout(ncol=2, guides=\"collect\") & theme(legend.position=\"bottom\")\n",
    "cluster_sample_rep_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b6f6e08-3be1-4295-8a21-a0d44127921a",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10, repr.plot.height=5)\n",
    "cluster_tissue_1 <- ggplot(so_qc_treatment_reg[[1]]@meta.data, aes(x=SCT_snn_res.0.8, fill=tissue)) + \n",
    "      geom_bar(stat=\"count\", position=\"fill\") + \n",
    "      scale_fill_manual(values=color$tissue) +\n",
    "      ggtitle(\"CpG sample\") + xlab(\"Cluster\") + ylab(\"Cell frequency\") + \n",
    "      theme(legend.position=\"bottom\")\n",
    "\n",
    "cluster_tissue_2 <- ggplot(so_qc_treatment_reg[[2]]@meta.data, aes(x=SCT_snn_res.0.8, fill=tissue)) + \n",
    "      geom_bar(stat=\"count\", position=\"fill\") + \n",
    "      scale_fill_manual(values=color$tissue) +\n",
    "      ggtitle(\"NaCl sample\") + xlab(\"Cluster\") + ylab(\"Cell frequency\") + \n",
    "      theme(legend.position=\"bottom\")\n",
    "\n",
    "cluster_tissue <- cluster_tissue_1 + cluster_tissue_2\n",
    "cluster_tissue\n",
    "ggsave(cluster_tissue, filename=\"result/plot/seurat/cluster_tissue_treatment_reg.png\", width=10, height=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed8123e8-cb8c-40c6-98fd-5bbcbd7ccff2",
   "metadata": {},
   "source": [
    "# Save output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe6c7f21-739f-47ba-8b24-cf4315cd8507",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_out <- merge(so_qc_treatment_reg[[1]], so_qc_treatment_reg[[2]])\n",
    "saveRDS(so_out, so_qc_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06263b4a-b9c3-4a92-b013-a5916ec76cbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "source(\"bin/SeuratFacility.R\")\n",
    "seurat2dir(so=so_out, dir=\"data/object/seurat/\", overwrite=TRUE)\n",
    "seurat2dir(so=so_qc_treatment_reg[[\"NaCl\"]], dir=\"data/object/seurat_sct_nacl/\", overwrite=TRUE)\n",
    "seurat2dir(so=so_qc_treatment_reg[[\"CpG\"]], dir=\"data/object/seurat_sct_cpg/\", overwrite=TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3573893c-81fa-4f9d-aef4-764b80bc18c6",
   "metadata": {},
   "source": [
    "# Session info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8650431-fd79-4f68-bc8f-57c8d4decdef",
   "metadata": {},
   "outputs": [],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "r.4.1.0-FD20200109SPLENO",
   "language": "R",
   "name": "r.4.1.0-fd20200109spleno"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
