{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3304709f-bb88-43b9-8055-ce19dcd51aef",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import sys\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ec24ffc-7b79-4ac2-89d8-0300f58ac4df",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Root directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a1e0304-27dc-4bed-8498-f6857ddbffd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/research/peer/fdeckert/FD20200109SPLENO')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2ca4e28-4241-46fa-a0e5-059494a6f77a",
   "metadata": {},
   "source": [
    "# Merge BAM files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52c8c203-9714-4db4-9707-a4775abca170",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%bash\n",
    "# module load SAMtools\n",
    "\n",
    "# meta_csv=\"/research/peer/fdeckert/FD20200109SPLENO/data/bulkRNAseq/object/meta.csv\"\n",
    "# output_dir=\"/research/peer/fdeckert/FD20200109SPLENO/data/bulkRNAseq/bamcoverage/dpi\"\n",
    "\n",
    "# genotypes=(\"IFNAR_fl\" \"IFNAR_fl_LysM_cre\")\n",
    "# celltypes=(\"cMo\" \"PreRPM\" \"RPM\" \"EB\")\n",
    "# dpis=(\"D0\" \"D1\" \"D3\" \"D6\")\n",
    "\n",
    "# for genotype in \"${genotypes[@]}\"; do\n",
    "#   for celltype in \"${celltypes[@]}\"; do\n",
    "#     for dpi in \"${dpis[@]}\"; do\n",
    "\n",
    "#       bam_list=($(awk -F, -v g=\"$genotype\" -v c=\"$celltype\" -v d=\"$dpi\" '\n",
    "#       BEGIN { OFS=\",\" }\n",
    "#         NR > 1 {\n",
    "#           gsub(/^\"|\"$/, \"\", $5);   # dpi\n",
    "#           gsub(/^\"|\"$/, \"\", $3);   # celltype\n",
    "#           gsub(/^\"|\"$/, \"\", $6);   # genotype\n",
    "#           gsub(/^\"|\"$/, \"\", $8);   # bam_path\n",
    "#           if ($6 == g && $3 == c && $5 == d) print $8\n",
    "#         }' \"$meta_csv\"))\n",
    "                                                                \n",
    "#       if [ \"${#bam_list[@]}\" -gt 0 ]; then\n",
    "                                                                \n",
    "#         output_path=\"$output_dir/${genotype}_${celltype}_${dpi}.bam\"                                          \n",
    "                                                                \n",
    "#         samtools merge -f \"$output_path\" \"${bam_list[@]}\" -@ 32\n",
    "#         samtools index \"$output_path\" -@ 32\n",
    "#       fi\n",
    "\n",
    "#     done\n",
    "#   done\n",
    "# done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5238f6cf-d0e5-4fab-a55b-f8e4f5aaffd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%bash\n",
    "# module load deepTools\n",
    "\n",
    "# # Set working directory\n",
    "# cd /research/peer/fdeckert/FD20200109SPLENO/data/bulkRNAseq/bamcoverage/dpi/\n",
    "\n",
    "# # BAM coverage BigWig\n",
    "# bamCoverage -b IFNAR_fl_RPM_D0.bam -o IFNAR_fl_RPM_D0.bw --normalizeUsing CPM --binSize 10 --smoothLength 20 --effectiveGenomeSize 2150570000 -p 32\n",
    "# bamCoverage -b IFNAR_fl_RPM_D1.bam -o IFNAR_fl_RPM_D1.bw --normalizeUsing CPM --binSize 10 --smoothLength 20 --effectiveGenomeSize 2150570000 -p 32\n",
    "# bamCoverage -b IFNAR_fl_RPM_D3.bam -o IFNAR_fl_RPM_D3.bw --normalizeUsing CPM --binSize 10 --smoothLength 20 --effectiveGenomeSize 2150570000 -p 32\n",
    "# bamCoverage -b IFNAR_fl_RPM_D6.bam -o IFNAR_fl_RPM_D6.bw --normalizeUsing CPM --binSize 10 --smoothLength 20 --effectiveGenomeSize 2150570000 -p 32\n",
    "\n",
    "# bamCoverage -b IFNAR_fl_LysM_cre_RPM_D0.bam -o IFNAR_fl_LysM_cre_RPM_D0.bw --normalizeUsing CPM --binSize 10 --smoothLength 20 --effectiveGenomeSize 2150570000 -p 32\n",
    "# bamCoverage -b IFNAR_fl_LysM_cre_RPM_D1.bam -o IFNAR_fl_LysM_cre_RPM_D1.bw --normalizeUsing CPM --binSize 10 --smoothLength 20 --effectiveGenomeSize 2150570000 -p 32\n",
    "# bamCoverage -b IFNAR_fl_LysM_cre_RPM_D3.bam -o IFNAR_fl_LysM_cre_RPM_D3.bw --normalizeUsing CPM --binSize 10 --smoothLength 20 --effectiveGenomeSize 2150570000 -p 32\n",
    "# bamCoverage -b IFNAR_fl_LysM_cre_RPM_D6.bam -o IFNAR_fl_LysM_cre_RPM_D6.bw --normalizeUsing CPM --binSize 10 --smoothLength 20 --effectiveGenomeSize 2150570000 -p 32"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "442854d7-2aba-414f-8136-09a64c820491",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%bash \n",
    "\n",
    "# cd /research/peer/fdeckert/FD20200109SPLENO/data/bulkRNAseq/bamcoverage/dpi/\n",
    "\n",
    "# zcat /research/peer/fdeckert/reference/genome/GRCm39/gencode.vM33.annotation.gtf.gz | \\\n",
    "# awk '$3 == \"exon\" && $0 ~ /gene_name \"Ifnar1\"/' | \\\n",
    "# awk -F'\\t' '{print $0}' | \\\n",
    "# awk 'BEGIN{OFS=\"\\t\"} {print $1, $4-1, $5, $9}' | \\\n",
    "# awk 'NR==3' > ifnar1_exon_3.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d3bdd4c-a684-4e86-95f1-d30b2bba683d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load deepTools\n",
    "module load BEDTools\n",
    "\n",
    "# Set working directory\n",
    "cd /research/peer/fdeckert/FD20200109SPLENO/result/bamcoverage/bulkRNAseq/dpi/\n",
    "\n",
    "# Visualization tracks\n",
    "/nobackup/peer/fdeckert/miniconda3/envs/p.3.10.16-FD20200109SPLENO/bin/pyGenomeTracks --tracks track_ifnar1_extend.ini --fontSize 8 --trackLabelFraction 0.1 --dpi 300 --region chr16:91181126-91355329 -o track_ifnar1_extend.pdf\n",
    "/nobackup/peer/fdeckert/miniconda3/envs/p.3.10.16-FD20200109SPLENO/bin/pyGenomeTracks --tracks track_ifnar1.ini --fontSize 8 --trackLabelFraction 0.1 --dpi 300 --region chr16:91281126-91305329 -o track_ifnar1.pdf\n",
    "/nobackup/peer/fdeckert/miniconda3/envs/p.3.10.16-FD20200109SPLENO/bin/pyGenomeTracks --tracks track_ifna_ifnb.ini --fontSize 8 --trackLabelFraction 0.1 --dpi 300 --region chr4:88390262-88818893 -o track_ifna_ifnb.pdf\n",
    "/nobackup/peer/fdeckert/miniconda3/envs/p.3.10.16-FD20200109SPLENO/bin/pyGenomeTracks --tracks track_lyzm2.ini --fontSize 8 --trackLabelFraction 0.1 --dpi 300 --region chr10:117113036-117118426 -o track_lyzm2.pdf"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "p.3.10.16-FD20200109SPLENO",
   "language": "python",
   "name": "p.3.10.16-fd20200109spleno"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
