{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "41388f0f-6943-4543-928c-ce4559fe82c7",
   "metadata": {},
   "source": [
    "# Leiden annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ec2fb9e-9c29-4f4e-9753-095eb18ae052",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scvi\n",
    "import scanpy as sc\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import sys\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2393e82-4794-42ec-a5ff-979dd30437d8",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Root directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ebd3302-386b-4c8d-a48e-fb6d83c6091b",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/research/peer/fdeckert/FD20200109SPLENO')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf429f20-7d6f-44d3-9cf9-416484428b1d",
   "metadata": {},
   "source": [
    "## Custom modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64810afd-a8eb-4e32-9097-d03b8ece92dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.append('bin/')\n",
    "from adata_qc import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87ab45eb-e633-48d0-aae2-8d3cf37088c4",
   "metadata": {},
   "source": [
    "## Settup rpy2 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16ee2d7d-3c22-475b-a71a-2eb20af68ce4",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.environ['R_HOME'] = '/nobackup/peer/fdeckert/miniconda3/envs/r.4.4.1-FD20200109SPLENO/lib/R'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70a464bf-3459-418c-b27a-94fc28f0ffb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import rpy2\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b0cfe67-ae12-4b6a-9d37-a84d01d6f5e9",
   "metadata": {},
   "source": [
    "## Figures "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c392965f-3b1d-4d07-bc24-9545b44e85ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(figsize=(5, 5), dpi_save=1200, fontsize=12, frameon=False, facecolor='white')\n",
    "mpl.rcParams['figure.facecolor'] = 'white'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39f6292c-8429-4f4e-932e-8db2c1fe5241",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load color \n",
    "import rpy2.robjects as robjects\n",
    "color_load = robjects.r.source('plotting_global.R')\n",
    "color = dict()\n",
    "for i in range(len(color_load[0])):\n",
    "    color[color_load[0].names[i]] = {key : color_load[0][i].rx2(key)[0] for key in color_load[0][i].names}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6eda87be-d152-4684-81f2-45a8414dc5f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set color function \n",
    "def set_color(categories): \n",
    "    \n",
    "    categories = [x for x in categories if x in list(adata.obs.columns)]\n",
    "\n",
    "    for category in categories: \n",
    "\n",
    "        adata.obs[category] = pd.Series(adata.obs[category], dtype='category').cat.remove_unused_categories()\n",
    "        \n",
    "        keys = list(color[category].keys())\n",
    "        keys = [x for x in keys if x in list(adata.obs[category])]        \n",
    "\n",
    "        adata.obs[category] = adata.obs[category].cat.reorder_categories(keys)\n",
    "        adata.uns[category+'_colors'] = np.array([color[category].get(key) for key in keys], dtype=object)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8588514-4db9-49ef-aa8a-9245f8f88c4d",
   "metadata": {},
   "source": [
    "## Cell type marker genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c879f34-8a1a-4dc6-907e-1d6f8d9c8b93",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype_genes_dict = {\n",
    "    \n",
    "    'Cycling': ['Mki67'], \n",
    "    'Immmune': ['Ptprc'], \n",
    "    'MHC-I': ['H2-K1'], \n",
    "    'MHC-II': ['H2-Ab1'], \n",
    "    'Migration': ['Cx3cr1', 'Ccr2', 'Ccr5', 'Cxcr2', 'Ccr7', 'Ccr1'], # Cx3cr1:RP Mo, Ccr2:RP Mo, Ccr5:RP DC, Cxcr2:RP Neu, Ccr7:WP cDC, MG\n",
    "    \n",
    "    'MHC': ['H2-D1', 'H2-Ab1'], \n",
    "    \n",
    "    'CD4': ['Cd4'], \n",
    "    'CD8': ['Cd8a', 'Cd8b1'], \n",
    "    \n",
    "    'Progenitor': ['Cd34', 'Kit', 'Flt3', 'Slamf1', 'Ly6a', 'Cd48', 'Il7r', 'Cebpa', 'Irf8', 'Csf1r'], \n",
    "    'Lin': ['Cd3e', 'Cd19', 'Ly6g', 'Ly6c1', 'Ly6c2', 'Itgam', 'Gypa'], \n",
    "    \n",
    "    'Neutrophil': ['Elane', 'Gfi1'],  \n",
    "    'Eo': ['Prg2', 'Prg3'],  \n",
    "    'Basophil': ['Mcpt8', 'Prss34', 'Fcer1a', 'Cd200r3', 'Clec12a'], # https://www.nature.com/articles/s41467-023-38356-1\n",
    "    'Mast': ['Gzmb', 'Cma1', 'Mcpt1', 'Mcpt4'], \n",
    "    \n",
    "    'Mo': ['Csf1r', 'Ccr2', 'Ly6c2', 'Cx3cr1'],\n",
    "    'RPM': ['Adgre1', 'Spic', 'Siglec1'], \n",
    "\n",
    "    'cDC': ['Zbtb46', 'Flt3', 'Itgax'], \n",
    "    'cDC1': ['Xcr1', 'Clec9a', 'Itgae'],\n",
    "    'cDC2': ['Itgam', 'Sirpa', 'Clec10a', 'Esam', 'Cxcr1', 'Cxcl13'],\n",
    "    'pDC': ['Siglech', 'Il3ra', 'Tcf4', 'Clec4b1', 'Ly6d'],\n",
    "\n",
    "    'Meg': ['Pf4', 'Itga2b'],\n",
    "\n",
    "    'Erythroid': ['Gata2', 'Gata1', 'Klf1', 'Epor', 'Hbb-bt', 'Hba-x'], \n",
    "    \n",
    "    'T synapse': ['Trac', 'Trbc1', 'Trbc2', 'Trdc', 'Cd3e', 'Cd3d', 'Cd3g', 'Cd247'], \n",
    "    \n",
    "    'B cell': ['Ms4a1', 'Ighm', 'Ighd', 'Igha', 'Ly6k']\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46eca5df-ef95-4922-a9a5-c3832538f17d",
   "metadata": {},
   "source": [
    "# Import QC data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8298bc89-398b-43ab-8429-8cb517b65662",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad('data/scRNAseq/object/qc.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "010fc6c7-b29f-42bb-88e6-6a594fdda829",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.leiden(adata, resolution=0.10, restrict_to=('leiden', ['3']), flavor='igraph', n_iterations=2, key_added='leiden_restrict')\n",
    "sc.tl.leiden(adata, resolution=0.25, restrict_to=('leiden_restrict', ['10']), flavor='igraph', n_iterations=2, key_added='leiden_restrict')\n",
    "sc.tl.leiden(adata, resolution=0.10, restrict_to=('leiden_restrict', ['20']), flavor='igraph', n_iterations=2, key_added='leiden_restrict')\n",
    "\n",
    "adata.obs['leiden'] = adata.obs['leiden_restrict']\n",
    "del adata.obs['leiden_restrict']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cb57500-1928-4d52-a1a4-2e6f8ccfc3e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['leiden', 'label_main_immgen', 'label_main_haemopedia', 'solo_pred_doublet', 'solo_label', 'dm_class', 'log1p_total_counts', 'log1p_n_genes_by_counts', 'pct_counts_MT', 'pct_counts_RP', 'pct_counts_HB', 'log1p_total_counts_XIST', 'Adgre1', 'Spic', 'Pf4', 'Cd34', 'Trac', 'Elane'], frameon=False, ncols=6, wspace=0.1, size=30, legend_loc='on data', use_raw=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0b684c8-d437-48f7-8acc-9f3d78c17bc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[(~(adata.obs.leiden.isin(['3,1', '3,2', '5', '6', '8', '10,3', '11', '13', '20,2', '21', '22', '23', '25', '26']))) & (adata.obs.solo_pred_doublet <= 0.5)]\n",
    "adata = adata[(adata.obs.leiden.isin(['12', '15', '16', '17'])) | (adata.obs.pct_counts_HB <= 5.0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff6f5789-168f-46e6-8429-a27b3fdaa6c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['leiden', 'label_main_immgen', 'label_main_haemopedia', 'solo_pred_doublet', 'solo_label', 'dm_class', 'log1p_total_counts', 'log1p_n_genes_by_counts', 'pct_counts_MT', 'pct_counts_RP', 'pct_counts_HB', 'log1p_total_counts_XIST', 'Adgre1', 'Spic', 'Pf4', 'Cd34', 'Trac', 'Elane'], frameon=False, ncols=6, wspace=0.1, size=30, legend_loc='on data', use_raw=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f82a8fd0-0d27-4927-9a39-0f0b6f40c6d6",
   "metadata": {},
   "source": [
    "## SCVI model "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c450fda5-c895-4a9d-a52d-f926b0e02f9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[:, (adata.X>=3).sum(axis=0)>=3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38ee5f4b-f15e-4170-aa84-9208d049bbc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89953e54-d064-4b0b-a746-ecc3bc0b8c65",
   "metadata": {},
   "outputs": [],
   "source": [
    "cache_scvi = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa07d13d-1ca5-4086-8c54-bf1e77ffa909",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not cache_scvi: \n",
    "    \n",
    "    scvi.model.SCVI.setup_anndata(\n",
    "    \n",
    "        adata, \n",
    "        batch_key='sample_group_rep', \n",
    "        categorical_covariate_keys=['facility'], \n",
    "        continuous_covariate_keys=['S_score', 'G2M_score']\n",
    "        \n",
    "    )\n",
    "    \n",
    "    model = scvi.model.SCVI(\n",
    "    \n",
    "        adata, \n",
    "        n_latent=30, \n",
    "        n_hidden=128, \n",
    "        n_layers=2, \n",
    "        gene_likelihood='nb'\n",
    "\n",
    "    )\n",
    "    \n",
    "    max_epochs = int(np.min([round((20000 / adata.n_obs) * 400), 400]))\n",
    "    \n",
    "    model.train(max_epochs=max_epochs, check_val_every_n_epoch=1)\n",
    "    \n",
    "    model.save('data/scRNAseq/object/scvi/annotation/model/', overwrite=True)\n",
    "    \n",
    "else: \n",
    "    \n",
    "    model = scvi.model.SCVI.load('data/scRNAseq/object/scvi/annotation/model/', adata=adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d700f857-b316-41ed-b1dd-20bd732f9565",
   "metadata": {},
   "outputs": [],
   "source": [
    "def model_history(model): \n",
    "    \n",
    "    # Plot model history \n",
    "    fig, axes=plt.subplots(1, 2, figsize=(10, 5))\n",
    "    \n",
    "    axes[0].plot(model.history['reconstruction_loss_train']['reconstruction_loss_train'], label='train')\n",
    "    axes[0].plot(model.history['reconstruction_loss_validation']['reconstruction_loss_validation'], label='validation')\n",
    "    axes[0].set_title('Reconstruction Loss')\n",
    "    axes[0].set_xlabel('Epoch')\n",
    "    axes[0].set_ylabel('Loss')\n",
    "    axes[0].legend()\n",
    "    \n",
    "    axes[1].plot(model.history['elbo_train']['elbo_train'], label='train')\n",
    "    axes[1].plot(model.history['elbo_validation']['elbo_validation'], label='validation')\n",
    "    axes[1].set_title('ELBO')\n",
    "    axes[1].set_xlabel('Epoch')\n",
    "    axes[1].set_ylabel('Loss')\n",
    "    axes[1].legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a189bc1-102f-46e0-a8f4-440e605889eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_history(model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b319f22-a8b8-481f-ace6-6e63f12258f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "cache_dim = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d460b953-de6b-4366-b90b-2a39adb728e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not cache_dim: \n",
    "    \n",
    "    adata.obsm['latent'] = model.get_latent_representation()\n",
    "    \n",
    "    sc.pp.neighbors(adata, n_neighbors=30, use_rep='latent')\n",
    "    sc.tl.leiden(adata, resolution=1, flavor='igraph', n_iterations=2, key_added='leiden')\n",
    "    sc.tl.umap(adata, min_dist=1)\n",
    "    \n",
    "    adata.write_h5ad('data/scRNAseq/object/scvi/annotation/adata.h5ad')\n",
    "    \n",
    "else: \n",
    "    \n",
    "    adata = sc.read_h5ad('data/scRNAseq/object/scvi/annotation/adata.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bff8833-56cd-493e-bff7-156cd7c18227",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata.raw.to_adata()\n",
    "adata.raw = adata.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d88f6a24-d21c-4070-b8d0-32af85b8fac9",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(adata, target_sum=10000)\n",
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b257aae-e905-42d7-943c-0c8d5a0608b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.leiden(adata, resolution=0.25, restrict_to=('leiden', ['0']), flavor='igraph', n_iterations=2, key_added='leiden_restrict')\n",
    "sc.tl.leiden(adata, resolution=0.50, restrict_to=('leiden_restrict', ['8']), flavor='igraph', n_iterations=2, key_added='leiden_restrict')\n",
    "sc.tl.leiden(adata, resolution=0.25, restrict_to=('leiden_restrict', ['10']), flavor='igraph', n_iterations=2, key_added='leiden_restrict')\n",
    "sc.tl.leiden(adata, resolution=0.50, restrict_to=('leiden_restrict', ['11']), flavor='igraph', n_iterations=2, key_added='leiden_restrict')\n",
    "\n",
    "adata.obs['leiden'] = adata.obs['leiden_restrict']\n",
    "del adata.obs['leiden_restrict']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f977565-0f82-43d1-8d5e-c5e5b7371598",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['leiden', 'facility', 'label_main_immgen', 'label_main_haemopedia', 'S_score', 'G2M_score', 'dm_class', 'log1p_total_counts', 'log1p_n_genes_by_counts', 'pct_counts_MT', 'pct_counts_RP', 'pct_counts_HB', 'log1p_total_counts_XIST'], frameon=False, ncols=6, wspace=0.1, size=30, legend_loc='on data', use_raw=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93fd05a7-003e-4d8a-8234-81d81325ea7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dp = sc.pl.dotplot(adata, celltype_genes_dict, groupby='leiden', standard_scale='var', use_raw=False, dendrogram=False, return_fig=True)\n",
    "dp.add_totals().style(dot_edge_color='black', dot_edge_lw=0.5, cmap='Reds').show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09c7b5d9-20da-4139-9a0b-966ab9f1b6e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype_leiden = {\n",
    "\n",
    "    '0,0': 'intMo (0,0)',\n",
    "    '0,1': 'cMo (0,1)',\n",
    "    '0,2': 'cMo (0,2)',\n",
    "    '0,3': 'intMo (0,3)',\n",
    "    '1': 'RPM (1)',\n",
    "    '2': 'ncMo (2)',\n",
    "    '3': 'moDC (3)',\n",
    "    '4': 'cDC2 (4)',\n",
    "    '5': 'cDC mig. (5)',\n",
    "    '6': 'cDC1 (6)',\n",
    "    '7': 'Proerythroblast (7)',\n",
    "    '8,0': 'Erythroblast (8,0)',\n",
    "    '8,1': 'Erythroblast (8,1)',\n",
    "    '8,2': 'Proerythroblast (8,2)',\n",
    "    '8,3': 'Erythroblast (8,3)',\n",
    "    '8,4': 'Erythroblast (8,4)',\n",
    "    '9': 'Erythroblast (9)',\n",
    "    '10,0': 'MEP (10,0)',\n",
    "    '10,1': 'MEP (10,1)',\n",
    "    '10,2': 'MegP (10,2)',\n",
    "    '11,0': 'NeuP (11,0)',\n",
    "    '11,1': 'BasoP (11,1)',\n",
    "    '11,2': 'GMP (11,2)',\n",
    "    '11,3': 'BasoP (11,3)',\n",
    "    '11,4': 'MastP (11,4)',\n",
    "    '11,5': 'MegP (11,5)',\n",
    "    '12': 'Basophil (12)',\n",
    "    '13': 'Plasma cell (13)',\n",
    "    '14': 'B cell (14)',\n",
    "    '15': 'T cell/ILC (15)',\n",
    "    '16': 'T cell/ILC (16)'\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81ce96f1-be1f-4a42-bd64-8ffcc07ac537",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mapper \n",
    "celltype_leiden = pd.DataFrame({'celltype_leiden': adata.obs['leiden'].map(lambda x: celltype_leiden.get(x, x)).astype('category')})\n",
    "adata.obs = adata.obs.merge(celltype_leiden, left_index=True, right_index=True, how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fffa5aab-e0dc-4372-906b-85ec91a889c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['celltype_leiden'] = adata.obs['celltype_leiden'].cat.reorder_categories([\n",
    "\n",
    "    'NeuP (11,0)',\n",
    "    'BasoP (11,1)',\n",
    "    'BasoP (11,3)',\n",
    "    'Basophil (12)',\n",
    "    'MastP (11,4)',\n",
    "    'cMo (0,2)',   \n",
    "    'cMo (0,1)',\n",
    "    'intMo (0,0)',\n",
    "    'intMo (0,3)',\n",
    "    'ncMo (2)',\n",
    "    \n",
    "    'RPM (1)',\n",
    "    \n",
    "    'moDC (3)',\n",
    "    'cDC1 (6)',\n",
    "    'cDC2 (4)',\n",
    "    'cDC mig. (5)',\n",
    "\n",
    "    'MegP (10,2)',\n",
    "    'MegP (11,5)',\n",
    "    'MEP (10,0)',\n",
    "    'MEP (10,1)',\n",
    "    \n",
    "    'Proerythroblast (7)',\n",
    "    'Proerythroblast (8,2)',\n",
    "    'Erythroblast (8,0)',\n",
    "    'Erythroblast (8,1)',\n",
    "    'Erythroblast (8,3)',\n",
    "    'Erythroblast (8,4)',\n",
    "    'Erythroblast (9)', \n",
    "    \n",
    "    'GMP (11,2)',\n",
    "    \n",
    "    'T cell/ILC (15)',\n",
    "    'T cell/ILC (16)',\n",
    "    \n",
    "    'B cell (14)',\n",
    "    'Plasma cell (13)'\n",
    "    \n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9445f254-d36b-4a3b-9a82-c80b3169e5c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "dp = sc.pl.dotplot(adata, celltype_genes_dict, groupby='celltype_leiden', standard_scale='var', use_raw=False, dendrogram=False, return_fig=True)\n",
    "dp.add_totals().style(dot_edge_color='black', dot_edge_lw=0.5, cmap='Reds').show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0adb6bd7-2aca-4593-817b-b7afc0c8997b",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype_low = {\n",
    "    \n",
    "    '0,0': 'intMo',\n",
    "    '0,1': 'cMo',\n",
    "    '0,2': 'cMo',\n",
    "    '0,3': 'intMo',\n",
    "    '1': 'RPM',\n",
    "    '2': 'ncMo',\n",
    "    '3': 'moDC',\n",
    "    '4': 'cDC2',\n",
    "    '5': 'cDC mig.',\n",
    "    '6': 'cDC1',\n",
    "    '7': 'Proerythroblast',\n",
    "    '8,0': 'Erythroblast',\n",
    "    '8,1': 'Erythroblast',\n",
    "    '8,2': 'Proerythroblast',\n",
    "    '8,3': 'Erythroblast',\n",
    "    '8,4': 'Erythroblast',\n",
    "    '9': 'Erythroblast',\n",
    "    '10,0': 'MEP',\n",
    "    '10,1': 'MEP',\n",
    "    '10,2': 'MegP',\n",
    "    '11,0': 'NeuP',\n",
    "    '11,1': 'BasoP',\n",
    "    '11,2': 'GMP',\n",
    "    '11,3': 'BasoP',\n",
    "    '11,4': 'MastP',\n",
    "    '11,5': 'MegP',\n",
    "    '12': 'Basophil',\n",
    "    '13': 'Plasma cell',\n",
    "    '14': 'B cell',\n",
    "    '15': 'T cell/ILC',\n",
    "    '16': 'T cell/ILC'\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d02b998e-53ef-4d83-aee2-73f24876fac9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mapper \n",
    "celltype_low = pd.DataFrame({'celltype_low': adata.obs['leiden'].map(lambda x: celltype_low.get(x, x)).astype('category')})\n",
    "adata.obs = adata.obs.merge(celltype_low, left_index=True, right_index=True, how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f44ed491-a419-411b-9d6d-5d7ae41f277a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['celltype_low'] = adata.obs['celltype_low'].cat.reorder_categories([\n",
    "\n",
    "    'GMP',\n",
    "    \n",
    "    'NeuP',\n",
    "    \n",
    "    'BasoP',\n",
    "    'Basophil',\n",
    "    \n",
    "    'MastP',\n",
    "    \n",
    "    'cMo',\n",
    "    'intMo',\n",
    "    'ncMo',\n",
    "    \n",
    "    'RPM',\n",
    "    \n",
    "    'moDC',\n",
    "    'cDC1',\n",
    "    'cDC2',\n",
    "    'cDC mig.',\n",
    "    \n",
    "    'MegP',\n",
    "    'MEP',\n",
    "    'Proerythroblast',\n",
    "    'Erythroblast',\n",
    "    \n",
    "    'T cell/ILC',\n",
    "    'B cell',\n",
    "    'Plasma cell'\n",
    "\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed34afed-ecf8-4c29-9f42-17b1cfec119e",
   "metadata": {},
   "outputs": [],
   "source": [
    "dp = sc.pl.dotplot(adata, celltype_genes_dict, groupby='celltype_low', standard_scale='var', use_raw=False, dendrogram=False, return_fig=True)\n",
    "dp.add_totals().style(dot_edge_color='black', dot_edge_lw=0.5, cmap='Reds').show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e1d1506-9f49-4c07-811d-729f64f5171f",
   "metadata": {},
   "source": [
    "# Remove some artefacts and lymphoid cells "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8b542d5-752a-4088-a866-58259298a3b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.leiden(adata, resolution=0.5, restrict_to=('leiden', ['11,0']), flavor='igraph', n_iterations=2, key_added='leiden_tmp')\n",
    "sc.tl.leiden(adata, resolution=0.5, restrict_to=('leiden_tmp', ['11,2']), flavor='igraph', n_iterations=2, key_added='leiden_tmp')\n",
    "sc.tl.leiden(adata, resolution=0.5, restrict_to=('leiden_tmp', ['6']), flavor='igraph', n_iterations=2, key_added='leiden_tmp')\n",
    "adata = adata[~adata.obs.leiden_tmp.isin(['11,0,2', '11,2,0', '6,1'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "336ec133-e125-45f2-86e1-5f38cb3ae5ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[~adata.obs.celltype_low.isin(['Plasma cell', 'T cell/ILC', 'B cell'])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37cf8012-c3b3-4543-af52-5e6b8590a634",
   "metadata": {},
   "source": [
    "# Set color"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea070f65-3579-4c0b-9d92-8a2bdeed9064",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set colors\n",
    "set_color(list(color.keys()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "534bd343-b7de-4b24-a7eb-577d29f69ab2",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['leiden', 'leiden_tmp', 'celltype_low', 'label_main_immgen', 'label_main_haemopedia', 'facility', 'infection', 'genotype', 'dpi', 'S_score_scale', 'G2M_score_scale', 'log1p_total_counts', 'log1p_n_genes_by_counts', 'log1p_total_counts_MT', 'log1p_total_counts_RP', 'pct_counts_HB', 'log1p_total_counts_XIST'], frameon=False, ncols=6, wspace=0.1, size=20, legend_loc='on data', use_raw=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce327c95-9ecc-4e47-93b4-c4b701d20428",
   "metadata": {},
   "source": [
    "# Remove some ambagious cells "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "988e8cd2-127a-457e-8481-fd6c179099ff",
   "metadata": {},
   "source": [
    "# Set raw "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5abae22-43bf-414e-8e3a-787597f276f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "133dde08-f41d-4d1c-be12-16a5dd740106",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata.raw.to_adata()\n",
    "adata.raw = adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e1696b9-c623-4fb3-8853-57f31a2c4b88",
   "metadata": {},
   "source": [
    "# Save results "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69273051-6794-4690-839e-6e2ad85ebb7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write_h5ad('data/scRNAseq/object/pp.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16954566-27b3-4767-a128-aba8a5d91a4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "cnt = adata.X.T.todense()\n",
    "cell_id = adata.obs_names\n",
    "gene_id = adata.var_names\n",
    "meta = adata.obs\n",
    "umap = adata.obsm['X_umap']\n",
    "latent = adata.obsm['latent']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cac395c4-0fd1-41f3-a5f1-bdf1a8b76100",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i cnt -i cell_id -i gene_id -i meta -i umap -i latent\n",
    "\n",
    "colnames(cnt) <- cell_id\n",
    "rownames(cnt) <- gene_id\n",
    "\n",
    "so <- Seurat::CreateSeuratObject(cnt, meta.data=meta, assay='RNA')\n",
    "\n",
    "colnames(umap) <- paste0('UMAP_', 1:ncol(umap))\n",
    "rownames(umap) <- cell_id\n",
    "so[['umap']] <- Seurat::CreateDimReducObject(embeddings=as.matrix(umap), key='UMAP_', assay='RNA')\n",
    "\n",
    "colnames(latent) <- paste0('LATENT_', 1:ncol(latent))\n",
    "rownames(latent) <- cell_id\n",
    "so[['latent']] <- Seurat::CreateDimReducObject(embeddings=as.matrix(latent), key='LATENT_', assay='RNA')\n",
    "\n",
    "saveRDS(so, 'data/scRNAseq/object/pp.rds')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scvi-tools.1.2.2",
   "language": "python",
   "name": "scvi-tools.1.2.2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
