{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "50e4fbf7-815e-4b90-9012-a0afabdd59a5",
   "metadata": {},
   "source": [
    "# Figure panel 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cfaec12-a6e4-450e-a15e-e4a93255590b",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "800de35d-86e3-4b71-8dd3-911e631758f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "library_load <- suppressMessages(\n",
    "    \n",
    "    suppressWarnings(\n",
    "        \n",
    "        list(\n",
    "        \n",
    "            # Seurat \n",
    "            library(Seurat), \n",
    "\n",
    "            # TradeSeq\n",
    "            library(tradeSeq),\n",
    "\n",
    "            # CCI \n",
    "            library(CellChat), \n",
    "\n",
    "            # Data \n",
    "            library(tidyverse), \n",
    "            library(data.table), \n",
    "\n",
    "            # STRING\n",
    "            library(STRINGdb), \n",
    "            library(igraph), \n",
    "            library(tidygraph), \n",
    "            library(ggraph), \n",
    "\n",
    "            # Plotting \n",
    "            library(ggplot2), \n",
    "            library(patchwork), \n",
    "            library(cowplot), \n",
    "            library(ComplexHeatmap), \n",
    "\n",
    "            # Pyhton compatibility\n",
    "            library(reticulate)\n",
    "\n",
    "        )\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15b0ec44-5e03-4e48-82cd-1042b31d1a5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Configure reticulate \n",
    "# use_condaenv(condaenv=\"p.3.10.16\", conda=\"C:/miniconda3/bin/conda\", required=NULL)\n",
    "# py_config()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45028df8-0d7a-4a09-865f-5d769c33cf3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_seed <- 42\n",
    "set.seed(random_seed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ae344ae-fb5c-42da-af0e-e5559e7b3e09",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set working directory to project root\n",
    "setwd(\"C:/Users/Florian Deckert/Projects/FD20200109SPLENO/\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b52cea3-0401-43b9-a99a-a97fc2a66c4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Source\n",
    "source(\"bin/so_pl.R\")\n",
    "source(\"bin/dea_pp.R\")\n",
    "source(\"bin/dea_pl.R\")\n",
    "source(\"bin/gsea_pp.R\")\n",
    "source(\"bin/gsea_pl.R\")\n",
    "source(\"bin/cci_pp.R\")\n",
    "source(\"bin/cci_pl.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ab8c554-ca21-4e65-8932-2a87d05f1f30",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting Theme\n",
    "source(\"plotting_global.R\")\n",
    "ggplot2::theme_set(theme_global_set(size_select=4)) # From project global source()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7d45dcf-0f7c-422e-815d-c982dd7d9e37",
   "metadata": {},
   "source": [
    "# Import scRNAseq data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04ba5c21-247b-4af4-ac1b-37c0bddd7b69",
   "metadata": {},
   "outputs": [],
   "source": [
    "so <- readRDS(\"data/scRNAseq/object/pp_1.rds\")\n",
    "so <- NormalizeData(so)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be7d13cf-4977-45e9-a028-89ea7e5d7350",
   "metadata": {},
   "source": [
    "# Differential expression analysis (DEA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b711403-c917-4872-9c8d-331b39b4cd30",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Response gene DEA \n",
    "contrasts_vec <- c(\"IFNAR_fl_CpG_D1\", \"IFNAR_fl_Baseline_D0\")\n",
    "dea_res_1 <- readRDS(paste0(\"result/dea/scRNAseq/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))\n",
    "\n",
    "contrasts_vec <- c(\"IFNAR_fl_LysM_cre_CpG_D1\", \"IFNAR_fl_LysM_cre_Baseline_D0\")\n",
    "dea_res_2 <- readRDS(paste0(\"result/dea/scRNAseq/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))\n",
    "\n",
    "contrasts_vec <- c(\"IFNAR_fl_CpG_D3\", \"IFNAR_fl_Baseline_D0\")\n",
    "dea_res_3 <- readRDS(paste0(\"result/dea/scRNAseq/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))\n",
    "\n",
    "contrasts_vec <- c(\"IFNAR_fl_LysM_cre_CpG_D3\", \"IFNAR_fl_LysM_cre_Baseline_D0\")\n",
    "dea_res_4 <- readRDS(paste0(\"result/dea/scRNAseq/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))\n",
    "\n",
    "# Genotype DEA\n",
    "contrasts_vec <- c(\"IFNAR_fl_LysM_cre_Baseline_D0\", \"IFNAR_fl_Baseline_D0\")\n",
    "dea_res_5 <- readRDS(paste0(\"result/dea/scRNAseq/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))\n",
    "\n",
    "contrasts_vec <- c(\"IFNAR_fl_LysM_cre_CpG_D1\", \"IFNAR_fl_CpG_D1\")\n",
    "dea_res_6 <- readRDS(paste0(\"result/dea/scRNAseq/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))\n",
    "\n",
    "contrasts_vec <- c(\"IFNAR_fl_LysM_cre_CpG_D3\", \"IFNAR_fl_CpG_D3\")\n",
    "dea_res_7 <- readRDS(paste0(\"result/dea/scRNAseq/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8af20ea3-e859-4451-8822-8ed4bc6a06c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "adj_pval_thr <- 0.05\n",
    "log2FC_thr <- 0.50"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f9c804f-bc7b-49c4-87a1-c6e5fa94e653",
   "metadata": {},
   "source": [
    "# DEG count overview"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d0f4dd5-fd88-44f2-9388-42752bfe03aa",
   "metadata": {},
   "source": [
    "## Response count D1 and D3 vs Ctl (Heatmap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e334cd1-b03d-49da-84f9-cd76cdb3033c",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- lapply(dea_res_1, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2FC_thr) %>% dplyr::pull(gene) %>% length}) %>% data.frame() %>% t() \n",
    "mat_2 <- lapply(dea_res_2, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2FC_thr) %>% dplyr::pull(gene) %>% length}) %>% data.frame() %>% t() \n",
    "mat_3 <- lapply(dea_res_3, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2FC_thr) %>% dplyr::pull(gene) %>% length}) %>% data.frame() %>% t() \n",
    "mat_4 <- lapply(dea_res_4, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2FC_thr) %>% dplyr::pull(gene) %>% length}) %>% data.frame() %>% t() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f059b523-aebe-4707-a779-0340e2a6fed1",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat <- list(mat_1, mat_2, mat_3, mat_4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7788e6c-8792-455f-9cb9-5530da268f45",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat <- lapply(mat, function(x) {\n",
    "    \n",
    "    x <- as.data.frame(x)\n",
    "    set_diff <- setdiff(make.names(levels(so$celltype_low)), rownames(x))\n",
    "    x[set_diff, ] <- 0\n",
    "    x[make.names(levels(so$celltype_low)), , drop=FALSE]\n",
    "\n",
    "}\n",
    "             )\n",
    "\n",
    "mat <- do.call(cbind, mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2a8afae-68d4-461f-a2ff-ca5464cf8341",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(mat) <- c(\"+/+\", \"cre/+\", \"+/+\", \"cre/+\")\n",
    "col_split <- c(\"D1\", \"D1\", \"D3\", \"D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea688900-a1d9-4611-a03c-f6cdccb65b5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/hm_deg_count_d1_d3_vs_ctl_cg_and_ko.pdf\", width=1.5, height=3)\n",
    "\n",
    "draw(dea_count_hm(as.matrix(mat), col_split=col_split, fontsize_select=2))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef419035-3b29-4615-8552-e10d0855634b",
   "metadata": {},
   "source": [
    "## Genotype (Heatmap) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "beea45c3-e72f-4452-ba49-109e43f17f22",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype_select <- levels(so$celltype_low) %>% make.names()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec2bf1d0-8f11-4263-98eb-9efe0984c5ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- rbind(\n",
    "        \n",
    "    dplyr::bind_rows(dea_res_5, .id=\"celltype\") %>% dplyr::mutate(dpi=\"Ctl\"), \n",
    "    dplyr::bind_rows(dea_res_6, .id=\"celltype\") %>% dplyr::mutate(dpi=\"D1\"), \n",
    "    dplyr::bind_rows(dea_res_7, .id=\"celltype\") %>% dplyr::mutate(dpi=\"D3\")\n",
    "\n",
    ") %>% \n",
    "    dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2FC_thr) %>% \n",
    "    dplyr::group_by(celltype, dpi) %>% dplyr::summarise(n=n()) %>% tidyr::pivot_wider(., names_from=dpi, values_from=n, values_fill=0) %>% \n",
    "    tibble::column_to_rownames(\"celltype\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eaf50e28-76a3-4720-8c31-e72a0f067e8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "set_diff <- setdiff(celltype_select, rownames(mat_1))\n",
    "mat_1[set_diff, ] <- 0\n",
    "mat_1 <- mat_1[celltype_select, c(\"Ctl\", \"D1\", \"D3\"), drop=FALSE]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a1db0f8-b61a-4b5f-aa6c-79be5c683476",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/hm_deg_count_cg_vs_ko.pdf\", width=1.5, height=3)\n",
    "\n",
    "draw(dea_count_hm(as.matrix(mat_1), col_split=NULL, fontsize_select=2))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0344bc81-1f55-4b48-b364-f874bae6917a",
   "metadata": {},
   "source": [
    "## Genotype (jitter plot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d6c7bca-0eb9-4852-a03a-53ae5aa1da94",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype_select <- c(\"MEP\", \"Proerythroblast\", \"Erythroblast\", \"cMo\", \"intMo\", \"RPM\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "686ec949-7fba-46cf-b8b7-8c7856addc2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- rbind(\n",
    "    \n",
    "    dplyr::bind_rows(dea_res_6, .id=\"celltype\") %>% dplyr::mutate(dpi=\"D1\"), \n",
    "    dplyr::bind_rows(dea_res_7, .id=\"celltype\") %>% dplyr::mutate(dpi=\"D3\")\n",
    "\n",
    "    ) %>% dplyr::filter(celltype %in% celltype_select) %>% \n",
    "\n",
    "    dplyr::mutate(sig=ifelse(p_val_adj<=adj_pval_thr & avg_log2FC<=-log2FC_thr, \"+/+\", ifelse(p_val_adj<=adj_pval_thr & avg_log2FC>=+log2FC_thr, \"cre/+\", \"n.s.\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f189327-c1e2-4c18-ae06-10ec7dd7a8ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_2 <- mat_1 %>% dplyr::filter(sig!=\"n.s.\") %>% dplyr::group_by(dpi, celltype, sig) %>% dplyr::summarise(n=n()) %>% dplyr::arrange(celltype, dpi, sig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d649c7b-141b-48d0-9556-6a50ff28c12e",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- mat_1 %>% dplyr::mutate(sign_log10_padj=-sign(avg_log2FC)*log10(p_val_adj)) %>% dplyr::mutate(sign_log10_padj=ifelse(abs(sign_log10_padj)>60, sign(avg_log2FC)*60, sign_log10_padj)) %>% \n",
    "    dplyr::mutate(avg_log2FC=ifelse(abs(avg_log2FC)>5, sign(avg_log2FC)*5, avg_log2FC))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ea1fda1-9fed-45e9-823a-fb663152da0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "jp_1 <- ggplot(mat_1 %>% dplyr::filter(celltype==celltype_select[1]), aes(y=sign_log10_padj, x=dpi, color=sig)) + \n",
    "    geom_jitter(data=mat_1 %>% dplyr::filter(celltype==celltype_select[1] & sig==\"n.s.\"), size=0.1) + \n",
    "    geom_jitter(data=mat_1 %>% dplyr::filter(celltype==celltype_select[1] & sig!=\"n.s.\"), size=0.1) + \n",
    "    geom_hline(yintercept=0) +\n",
    "    ylim(-60, +60) + \n",
    "    scale_color_manual(values=rev(c(\"+/+\"=\"#cd34b5\", \"cre/+\"=\"#FFAC1E\", \"n.s.\"=\"#7f7f7f\"))) + \n",
    "    ylab(\"\") + xlab(\"\") + \n",
    "    facet_grid(~dpi, scales=\"free\") + theme(aspect.ratio=2) + theme_global_set(4)\n",
    "\n",
    "jp_2 <- ggplot(mat_1 %>% dplyr::filter(celltype==celltype_select[2]), aes(y=sign_log10_padj, x=dpi, color=sig)) + \n",
    "    geom_jitter(data=mat_1 %>% dplyr::filter(celltype==celltype_select[2] & sig==\"n.s.\"), size=0.1) + \n",
    "    geom_jitter(data=mat_1 %>% dplyr::filter(celltype==celltype_select[2] & sig!=\"n.s.\"), size=0.1) + \n",
    "    geom_hline(yintercept=0) +\n",
    "    ylim(-60, +60) + \n",
    "    scale_color_manual(values=c(\"+/+\"=\"#cd34b5\", \"cre/+\"=\"#FFAC1E\", \"n.s.\"=\"#7f7f7f\")) + \n",
    "    ylab(\"\") + xlab(\"\") + \n",
    "    facet_grid(~dpi, scales=\"free\") + theme(aspect.ratio=2) + theme_global_set(4)\n",
    "\n",
    "jp_3 <- ggplot(mat_1 %>% dplyr::filter(celltype==celltype_select[3]), aes(y=sign_log10_padj, x=dpi, color=sig)) + \n",
    "    geom_jitter(data=mat_1 %>% dplyr::filter(celltype==celltype_select[3] & sig==\"n.s.\"), size=0.1) + \n",
    "    geom_jitter(data=mat_1 %>% dplyr::filter(celltype==celltype_select[3] & sig!=\"n.s.\"), size=0.1) + \n",
    "    geom_hline(yintercept=0) +\n",
    "    ylim(-60, +60) + \n",
    "    scale_color_manual(values=c(\"+/+\"=\"#cd34b5\", \"cre/+\"=\"#FFAC1E\", \"n.s.\"=\"#7f7f7f\")) + \n",
    "    ylab(\"\") + xlab(\"\") + \n",
    "    facet_grid(~dpi, scales=\"free\") + theme(aspect.ratio=2) + theme_global_set(4)\n",
    "\n",
    "jp_4 <- ggplot(mat_1 %>% dplyr::filter(celltype==celltype_select[4]), aes(y=sign_log10_padj, x=dpi, color=sig)) + \n",
    "    geom_jitter(data=mat_1 %>% dplyr::filter(celltype==celltype_select[4] & sig==\"n.s.\"), size=0.1) + \n",
    "    geom_jitter(data=mat_1 %>% dplyr::filter(celltype==celltype_select[4] & sig!=\"n.s.\"), size=0.1) + \n",
    "    geom_hline(yintercept=0) +\n",
    "    ylim(-60, +60) + \n",
    "    scale_color_manual(values=c(\"+/+\"=\"#cd34b5\", \"cre/+\"=\"#FFAC1E\", \"n.s.\"=\"#7f7f7f\")) + \n",
    "    ylab(\"\") + xlab(\"\") + \n",
    "    facet_grid(~dpi, scales=\"free\") + theme(aspect.ratio=2) + theme_global_set(4)\n",
    "\n",
    "jp_5 <- ggplot(mat_1 %>% dplyr::filter(celltype==celltype_select[5]), aes(y=sign_log10_padj, x=dpi, color=sig)) + \n",
    "    geom_jitter(data=mat_1 %>% dplyr::filter(celltype==celltype_select[5] & sig==\"n.s.\"), size=0.1) + \n",
    "    geom_jitter(data=mat_1 %>% dplyr::filter(celltype==celltype_select[5] & sig!=\"n.s.\"), size=0.1) + \n",
    "    geom_hline(yintercept=0) +\n",
    "    ylim(-60, +60) + \n",
    "    scale_color_manual(values=c(\"+/+\"=\"#cd34b5\", \"cre/+\"=\"#FFAC1E\", \"n.s.\"=\"#7f7f7f\")) + \n",
    "    ylab(\"\") + xlab(\"\") + \n",
    "    facet_grid(~dpi, scales=\"free\") + theme(aspect.ratio=2) + theme_global_set(4)\n",
    "\n",
    "jp_6 <- ggplot(mat_1 %>% dplyr::filter(celltype==celltype_select[6]), aes(y=sign_log10_padj, x=dpi, color=sig)) + \n",
    "    geom_jitter(data=mat_1 %>% dplyr::filter(celltype==celltype_select[6] & sig==\"n.s.\"), size=0.1) + \n",
    "    geom_jitter(data=mat_1 %>% dplyr::filter(celltype==celltype_select[6] & sig!=\"n.s.\"), size=0.1) +  \n",
    "    geom_hline(yintercept=0) +\n",
    "    ylim(-60, +60) + \n",
    "    scale_color_manual(values=c(\"+/+\"=\"#cd34b5\", \"cre/+\"=\"#FFAC1E\", \"n.s.\"=\"#7f7f7f\")) + \n",
    "    ylab(\"\") + xlab(\"\") + \n",
    "    facet_grid(~dpi, scales=\"free\") + theme(aspect.ratio=2) + theme_global_set(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f1009d3-5573-4d8c-a0b6-ca9fe8100d11",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/jp_dea_count_genotype.pdf\", width=3, height=6)\n",
    "\n",
    "    gridExtra::grid.arrange(\n",
    "            \n",
    "        jp_1 %>% egg::set_panel_size(., width=unit(0.5, \"cm\"), height=unit(1.5, \"cm\")), \n",
    "        jp_2 %>% egg::set_panel_size(., width=unit(0.5, \"cm\"), height=unit(1.5, \"cm\")), \n",
    "        jp_3 %>% egg::set_panel_size(., width=unit(0.5, \"cm\"), height=unit(1.5, \"cm\")), \n",
    "        jp_4 %>% egg::set_panel_size(., width=unit(0.5, \"cm\"), height=unit(1.5, \"cm\")), \n",
    "        jp_5 %>% egg::set_panel_size(., width=unit(0.5, \"cm\"), height=unit(1.5, \"cm\")), \n",
    "        jp_6 %>% egg::set_panel_size(., width=unit(0.5, \"cm\"), height=unit(1.5, \"cm\")), \n",
    "        nrow=6\n",
    "        \n",
    "    )\n",
    "\n",
    "dev.off() "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e0d441e-f38d-43a0-8ff0-99c3b4fc6a0a",
   "metadata": {},
   "source": [
    "# Volcano plots  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b85b9f1b-242e-4315-878c-b8a5a85f9ff6",
   "metadata": {},
   "source": [
    "## CtlKO vs Ctl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11e942bf-4f62-43a3-b4af-af58db021e57",
   "metadata": {},
   "outputs": [],
   "source": [
    "labels <- list(\n",
    "    \n",
    "    \"intMo\"=c(\"Icam1\", \"Itga4\", \"Itgb1\", \"Vcam1\", \"Gdf15\", \"Ccl2\", \"Tnf\", \"Il6\", \"Tgfb1\", \"Stat1\", \"Stat2\", \"Spic\", \"Irf8\", \"Siglec1\", \"Slc40a1\", \"Hmox1\", \"Blvrb\", \"Treml4\", \"Bach1\", \"Il1a\", \"Il1b\", \"Il1rn\", \"Il18\", \"Il33\")\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8f42ed7-cdf9-4104-84e3-bf1f27a2c398",
   "metadata": {},
   "outputs": [],
   "source": [
    "vp_5 <- lapply(names(dea_res_5), function(i) {\n",
    "\n",
    "    v_pl_i <- v_pl(dea_res_5[[i]], title=i, log2FC_thr=log2FC_thr, adj_pval_thr=adj_pval_thr, point_size=0.5, color_pos=unlist(color$sample_group[\"IFNAR_fl_LysM_cre_Baseline_D0\"]), color_neg=unlist(color$sample_group[\"IFNAR_fl_Baseline_D0\"]), label=if(i %in% names(labels)) {labels[[i]]} else {label <- NULL}, label_size=2, y_limit=4, aspect_ratio=0.8, top_label=5, label_merge=TRUE) + theme(legend.position=\"none\") + theme_global_set(size_select=4)\n",
    "\n",
    "}\n",
    "                 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69728dce-5683-4d9d-83b1-ce39b72a321c",
   "metadata": {},
   "outputs": [],
   "source": [
    "vp_5 <- lapply(vp_5, function(p) {egg::set_panel_size(p, width=unit(3, \"cm\"), height=unit(2.5, \"cm\"))})\n",
    "vp_5 <- do.call(gridExtra::arrangeGrob, c(vp_5, ncol=5, nrow=ceiling(length(vp_5)/5)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e873235a-2be6-45db-9c85-846a837129bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/vp_ctlko_vs_ctl.pdf\", width=2*5, height=2*ceiling(length(vp_5)/5))\n",
    "\n",
    "grid::grid.draw(vp_5)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "440d7e4f-abfe-46c8-9cb2-d28cbea7957b",
   "metadata": {},
   "source": [
    "## D1KO vs D1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed9fefd9-8d33-4c0a-ae65-0a3ac221deb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "labels <- list(\n",
    "    \n",
    "    \"intMo\"=c(\"Icam1\", \"Itga4\", \"Itgb1\", \"Vcam1\", \"Gdf15\", \"Ccl2\", \"Tnf\", \"Il6\", \"Tgfb1\", \"Stat1\", \"Stat2\", \"Spic\", \"Irf8\", \"Siglec1\", \"Slc40a1\", \"Hmox1\", \"Blvrb\", \"Treml4\", \"Bach1\", \"Il1a\", \"Il1b\", \"Il1rn\", \"Il18\", \"Il33\")\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96869801-b7fb-423e-8a7a-139c1c68fe53",
   "metadata": {},
   "outputs": [],
   "source": [
    "vp_6 <- lapply(names(dea_res_6), function(i) {\n",
    "\n",
    "    v_pl_i <- v_pl(dea_res_6[[i]], title=i, log2FC_thr=log2FC_thr, adj_pval_thr=adj_pval_thr, point_size=0.5, color_pos=unlist(color$sample_group[\"IFNAR_fl_LysM_cre_CpG_D1\"]), color_neg=unlist(color$sample_group[\"IFNAR_fl_CpG_D1\"]), label=if(i %in% names(labels)) {labels[[i]]} else {label <- NULL}, label_size=2, y_limit=4, aspect_ratio=0.8, top_label=5, label_merge=TRUE) + theme(legend.position=\"none\") + theme_global_set(size_select=4)\n",
    "\n",
    "}\n",
    "                 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f2c7a25-61a7-4e38-ae44-617f4ace31d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "vp_6 <- lapply(vp_6, function(p) {egg::set_panel_size(p, width=unit(3, \"cm\"), height=unit(2.5, \"cm\"))})\n",
    "vp_6 <- do.call(gridExtra::arrangeGrob, c(vp_6, ncol=5, nrow=ceiling(length(vp_6)/5)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0d922b1-99c3-4a12-a56c-0c644437e176",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/vp_d1ko_vs_d1.pdf\", width=2*5, height=2*ceiling(length(vp_6)/5))\n",
    "\n",
    "grid::grid.draw(vp_6)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ee54a3a-1232-4e26-a180-489f09cfd187",
   "metadata": {},
   "source": [
    "## D3KO vs D3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5da1dab4-cb86-4195-a8f8-9bab537045ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "labels <- list(\n",
    "    \n",
    "    \"intMo\"=c(\"Icam1\", \"Itga4\", \"Itgb1\", \"Vcam1\", \"Gdf15\", \"Ccl2\", \"Tnf\", \"Il6\", \"Tgfb1\", \"Stat1\", \"Stat2\", \"Spic\", \"Irf8\", \"Siglec1\", \"Slc40a1\", \"Hmox1\", \"Blvrb\", \"Treml4\", \"Bach1\", \"Il1a\", \"Il1b\", \"Il1rn\", \"Il18\", \"Il33\")\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da1feeed-6cb3-4479-bee6-336cfed1c139",
   "metadata": {},
   "outputs": [],
   "source": [
    "vp_7 <- lapply(names(dea_res_7), function(i) {\n",
    "\n",
    "    v_pl_i <- v_pl(dea_res_7[[i]], title=i, log2FC_thr=log2FC_thr, adj_pval_thr=adj_pval_thr, point_size=0.5, color_pos=unlist(color$sample_group[\"IFNAR_fl_LysM_cre_CpG_D3\"]), color_neg=unlist(color$sample_group[\"IFNAR_fl_CpG_D3\"]), label=if(i %in% names(labels)) {labels[[i]]} else {label <- NULL}, label_size=2, y_limit=4, aspect_ratio=0.8, top_label=5, label_merge=TRUE) + theme(legend.position=\"none\") + theme_global_set(size_select=4)\n",
    "\n",
    "}\n",
    "                 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ce97425-0a76-4c2a-9490-3c5451fa0334",
   "metadata": {},
   "outputs": [],
   "source": [
    "vp_7 <- lapply(vp_7, function(p) {egg::set_panel_size(p, width=unit(3, \"cm\"), height=unit(2.5, \"cm\"))})\n",
    "vp_7 <- do.call(gridExtra::arrangeGrob, c(vp_7, ncol=5, nrow=ceiling(length(vp_7)/5)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b8ced5d-8c7d-4e11-b999-86ab69a1602e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/vp_d3ko_vs_d3.pdf\", width=2*5, height=2*ceiling(length(vp_7)/5))\n",
    "\n",
    "grid::grid.draw(vp_7)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33376363-6359-4d3c-a5b9-d0b6d204db8f",
   "metadata": {},
   "source": [
    "# DEA result heatmap with bulk expression matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62bf275e-cfbf-4d23-baa2-4d3590fccc35",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_dea_fetch <- function(celltype, row_genes, col_samples, col_split, col_label=NULL, scale_celltype=NULL, scale_mode=\"all\") {\n",
    "\n",
    "    # Subset by sample group \n",
    "    so_i <- so[, so$sample_group %in% col_samples & so$celltype_low %in% celltype, drop=FALSE] # Filtering celltype_low for performance but will not work with scale_mode all\n",
    "    \n",
    "    # Get pseudobulk counts \n",
    "    if(scale_mode==\"all\" & !is.null(scale_celltype)) {\n",
    "\n",
    "        so_i <- so[, so$celltype_low %in% scale_celltype, drop=FALSE]\n",
    "        \n",
    "    } \n",
    "    \n",
    "    # Get pseudobulk counts \n",
    "    mat_1 <- GetAssayData(so_i, assay=\"RNA\", layer=\"counts\")\n",
    "    mat_1 <- split(t(mat_1) %>% as.data.frame(), f=paste(so_i$celltype_low, so_i$sample_group))\n",
    "    mat_1 <- lapply(mat_1, colSums) %>% do.call(rbind, .) %>% t() \n",
    "    mat_1 <- t(t(mat_1) / colSums(mat_1))\n",
    "    \n",
    "    if(scale_mode==\"all\") {\n",
    "\n",
    "        mat_1 <- apply(mat_1, 1, scales::rescale) %>% t() # It apply transforms the data frame\n",
    "        mat_1 <- mat_1[unlist(row_genes), paste(celltype, col_samples)]\n",
    "\n",
    "        \n",
    "    } else if (scale_mode==\"sample\") {\n",
    "\n",
    "        mat_1 <- mat_1[unlist(row_genes), paste(celltype, col_samples)]\n",
    "        mat_1 <- apply(mat_1, 1, scales::rescale) %>% t() # It apply transforms the data frame\n",
    "        \n",
    "    }\n",
    "\n",
    "    if(!is.null(col_label)) {\n",
    "\n",
    "        colnames(mat_1) <- col_label\n",
    "        \n",
    "    }\n",
    "    \n",
    "\n",
    "    row_split <- setNames(unlist(row_genes, use.names=FALSE), rep(names(row_genes), sapply(row_genes, length)))\n",
    "    row_split <- row_split[row_split %in% rownames(mat_1)]\n",
    "\n",
    "    # DEA log2FC \n",
    "    mat_2 <- purrr::reduce(\n",
    "        \n",
    "        list(\n",
    "            \n",
    "            dea_res_5[[celltype]] %>% dplyr::mutate(Ctl=avg_log2FC) %>% dplyr::select(gene, Ctl), \n",
    "            dea_res_6[[celltype]] %>% dplyr::mutate(D1=avg_log2FC) %>% dplyr::select(gene, D1), \n",
    "            dea_res_7[[celltype]] %>% dplyr::mutate(D3=avg_log2FC) %>% dplyr::select(gene, D3)\n",
    "        \n",
    "        ), full_join, by=\"gene\"\n",
    "    \n",
    "    ) %>% tibble::column_to_rownames(\"gene\")\n",
    "\n",
    "    mat_2[setdiff(row_split, rownames(mat_2)), ] <- 0\n",
    "    mat_2 <- mat_2[row_split, ]\n",
    "    mat_2[is.na(mat_2)] <- 0\n",
    "    \n",
    "    # DEA adj p-values\n",
    "    mat_3 <- purrr::reduce(\n",
    "        \n",
    "        list(\n",
    "            \n",
    "            dea_res_5[[celltype]] %>% dplyr::mutate(Ctl=p_val_adj) %>% dplyr::select(gene, Ctl), \n",
    "            dea_res_6[[celltype]] %>% dplyr::mutate(D1=p_val_adj) %>% dplyr::select(gene, D1), \n",
    "            dea_res_7[[celltype]] %>% dplyr::mutate(D3=p_val_adj) %>% dplyr::select(gene, D3)\n",
    "        \n",
    "        ), full_join, by=\"gene\"\n",
    "    \n",
    "    ) %>% tibble::column_to_rownames(\"gene\")\n",
    "\n",
    "    mat_3[setdiff(row_split, rownames(mat_3)), ] <- 1\n",
    "    mat_3 <- mat_3[row_split, ]\n",
    "    mat_3[is.na(mat_3)] <- 1\n",
    "    mat_3[(mat_3>adj_pval_thr | abs(mat_2)<log2FC_thr)] <- 1\n",
    "\n",
    "    # Row split\n",
    "    row_split=names(row_split)\n",
    "    row_split=factor(row_split, levels=names(row_genes))\n",
    "    \n",
    "    hm_1 <- dea_exp_hm(mat_1, col_split=col_split, row_split=row_split, scale_width=1, legend_title=\"Exp.\", cluster_rows=TRUE, fontsize_select=2)\n",
    "    hm_2 <- dea_res_hm(mat_2, mat_3, p_val_adj_thr=adj_pval_thr, row_split=row_split, scale_width=1, color_min=\"#cd34b5\", color_max=\"#FFAC1E\", cluster_rows=TRUE, fontsize_select=2)\n",
    "\n",
    "    hm <- hm_2 + hm_1\n",
    "\n",
    "    return(hm)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c85301c-3f1c-4d54-9ce1-3ee431db88db",
   "metadata": {},
   "outputs": [],
   "source": [
    "adj_pval_thr <- 0.05\n",
    "log2FC_thr <- 0.50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5666b180-e755-429d-bf7b-d3aa2495c329",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_samples <- c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_LysM_cre_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_LysM_cre_CpG_D1\", \"IFNAR_fl_CpG_D3\", \"IFNAR_fl_LysM_cre_CpG_D3\")\n",
    "col_split <- c(\"Ctl\", \"Ctl\", \"D1\", \"D1\", \"D3\", \"D3\")\n",
    "col_label <- c(\"+/+\", \"cre/+\", \"+/+\", \"cre/+\", \"+/+\", \"cre/+\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b339cfb6-5599-4a3b-b53d-9a1e89fdff1f",
   "metadata": {},
   "source": [
    "## cMo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a8e6b00-1540-4330-9797-d8ab0f872aa9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# genes_1 <- dea_res_6[[\"cMo\"]] %>% dplyr::filter(abs(avg_log2FC)>=log2FC_thr & p_val_adj<=adj_pval_thr) %>% pull(gene)\n",
    "# genes_2 <- dea_res_7[[\"cMo\"]] %>% dplyr::filter(abs(avg_log2FC)>=log2FC_thr & p_val_adj<=adj_pval_thr) %>% pull(gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d08e287-0744-4bca-99da-92bb52db9ce3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# row_genes <- list(\"cMo\"=c(genes_1, genes_2) %>% unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a135c421-3be7-4b8e-8d5d-709a12b78fc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# hm_1 <- hm_dea_fetch(celltype=\"cMo\", row_genes=row_genes, col_samples=col_samples, col_split=col_split, col_label=col_label, scale_mode=\"sample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f18fcac-4d8f-4e2f-b64b-e7db1a11ca83",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pdf(\"result/figures/1_scRNAseq/panel_5/hm_dea_ctl_d1_d3_genotype_cmo_pb.pdf\", width=5, height=50)\n",
    "\n",
    "# hm_1\n",
    "\n",
    "# dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c99776e3-8d18-4c24-aae4-53d0664dd21b",
   "metadata": {},
   "source": [
    "## intMo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a60728e4-12e8-4d1f-9b53-5eab5377fc0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# genes_1 <- dea_res_6[[\"intMo\"]] %>% dplyr::filter(abs(avg_log2FC)>=log2FC_thr & p_val_adj<=adj_pval_thr) %>% pull(gene)\n",
    "# genes_2 <- dea_res_7[[\"intMo\"]] %>% dplyr::filter(abs(avg_log2FC)>=log2FC_thr & p_val_adj<=adj_pval_thr) %>% pull(gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b50879c5-c92e-4dd9-90fb-1f3df4bf8d42",
   "metadata": {},
   "outputs": [],
   "source": [
    "# row_genes <- list(\"intMo\"=c(genes_1, genes_2) %>% unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89a7fb2f-0a3f-4b0c-9915-93023c5be879",
   "metadata": {},
   "outputs": [],
   "source": [
    "# hm_1 <- hm_dea_fetch(celltype=\"intMo\", row_genes=row_genes, col_samples=col_samples, col_split=col_split, col_label=col_label, scale_mode=\"sample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "069fb535-a436-4195-a103-0f2357e47c45",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pdf(\"result/figures/1_scRNAseq/panel_5/hm_dea_ctl_d1_d3_genotype_intmo_pb.pdf\", width=5, height=50)\n",
    "\n",
    "# hm_1\n",
    "\n",
    "# dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fbe225b-3e7c-4437-9fb7-fa58d376ba78",
   "metadata": {},
   "source": [
    "## cMo (selected)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67838cfb-c791-4cd7-a0da-d28792cf97b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_genes <- list(\n",
    "\n",
    "    \"G1\"=c(\"Stat1\", \"Stat2\", \"Irf1\", \"Irf7\", \"Ifit1\", \"Ifit2\", \"Ifit3\", \"Isg15\", \"Ddx60\", \"Usp18\", \"Socs1\", \"Ccl2\", \"Cxcl10\"), # Interferon stimulation \n",
    "    \"G2\"=c(\"Spic\", \"Hmox1\", \"Maf\", \"Mafb\", \"Axl\", \"C5ar1\", \"Fcgr1\"), # Macrophage development\n",
    "    \"G3\"=c(\"Ftl1\", \"Fth1\", \"Nr1h3\", \"Blvrb\"), # Iron metabolism\n",
    "    \"G4\"=c(\"Dusp1\", \"Nlrp3\", \"Atf3\", \"Jun\", \"Fos\", \"Ccl4\", \"Ccl3\", \"Tnf\", \"Il1b\", \"Tnfaip3\", \"Nfkbia\") # MAPK activation (Dusp1), Nfkb inhibition (Tnfaip3)\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "739b697a-1132-4348-a735-e600c84ad673",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_1 <- hm_dea_fetch(celltype=\"cMo\", row_genes=row_genes, col_samples=col_samples, col_split=col_split, col_label=col_label, scale_mode=\"sample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bec07b3-ab51-401c-984b-b4cd37852072",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/hm_dea_ctl_d1_d3_genotype_cmo_pb_selected.pdf\", width=5, height=10)\n",
    "\n",
    "hm_1\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d970fde5-6e01-4896-9e67-903f5f8819de",
   "metadata": {},
   "source": [
    "## intMo (selected)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4710035-bc57-4887-ae24-b0d0e6948110",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_genes <- list(\n",
    "\n",
    "    \"G1\"=c(\"Stat1\", \"Stat2\", \"Irf7\", \"Ifit1\", \"Ifit2\", \"Ifit3\", \"Isg15\", \"Ddx60\", \"Usp18\", \"Socs1\", \"Ccl2\", \"Cxcl10\", \"Gbp2\", \"Gbp2b\", \"Gbp3\", \"Gbp5\", \"Gbp7\", \"Cxcl9\", \"Ciita\"), # Interferon stimulation \n",
    "    \"G2\"=c(\"Spic\", \"Hmox1\", \"Maf\", \"Mafb\", \"Vcam1\", \"Msr1\", \"Axl\", \"C5ar1\", \"Fcgr1\"), # Macrophage development\n",
    "    \"G3\"=c(\"Slc40a1\", \"Slc48a1\", \"Ftl1\", \"Fth1\", \"Prdx1\", \"Nr1h3\", \"Blvrb\", \"Hfe\"), # Iron metabolism\n",
    "    \"G4\"=c(\"Dusp1\", \"Atf3\", \"Jun\", \"Fos\", \"Ccl4\", \"Ccl3\", \"Il18\", \"Tnfaip3\", \"Nfkbia\"), # MAPK activation (Dusp1), Nfkb inhibition (Tnfaip3)\n",
    "    \"G5\"=c(\"Eif2ak2\", \"Atf4\", \"Ppp1r15a\", \"Slc7a11\", \"Gdf15\") # Integrated stress response. Eif2ak2 is PKR, a kinase activator of \n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cca6070-5b20-469e-8de0-3f34f61dd63e",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_1 <- hm_dea_fetch(celltype=\"intMo\", row_genes=row_genes, col_samples=col_samples, col_split=col_split, col_label=col_label, scale_mode=\"sample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e3131cf-f068-4cb6-ba7b-b3cb0d620137",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/hm_dea_ctl_d1_d3_genotype_intmo_pb_selected.pdf\", width=5, height=10)\n",
    "\n",
    "hm_1\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee504804-12f0-400f-a9b6-d25866f1fcbb",
   "metadata": {},
   "source": [
    "# Protein-Protein interaction "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08ada2c9-372d-4260-a1e3-6b1ff2059eb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create STRING DB for mouse \n",
    "string_db <- STRINGdb::STRINGdb$new(version=\"12.0\", species=10090, score_threshold=400, network_type=\"full\", link_data=\"combined_only\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9026235-d41f-4521-a26d-b09e62c75940",
   "metadata": {},
   "source": [
    "## intMo D1 KO UP "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4052f96-9f23-4349-8312-1dc6cb50d35c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mapp DEA results to STRING DB\n",
    "string_db_dea <- dea_res_6[[\"intMo\"]] %>% dplyr::select(gene, avg_log2FC, p_val_adj) %>% dplyr::rename(logFC=avg_log2FC, padj=p_val_adj) %>% na.omit() %>% dplyr::filter(padj<=adj_pval_thr & logFC>=+log2FC_thr) %>% string_db$map(., \"gene\", removeUnmappedRows=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4338cc0c-0d86-4b71-bca2-eafbbcc3cca6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Labe PPI vertices \n",
    "ppi_v <- string_db_dea %>% dplyr::distinct() %>% dplyr::rename(name=STRING_id) %>% \n",
    "    dplyr::group_by(name) %>% dplyr::mutate(n=n()) %>% dplyr::filter(n==1) %>% dplyr::select(name, gene, logFC, padj) %>% \n",
    "    dplyr::mutate(node_group=ifelse(padj<=0.05 & logFC<=-0.50, \"neg\", ifelse(padj<=0.05 & logFC>=0.50, \"pos\", \"n.s.\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0ac67f9-6f44-4a4e-8e14-b8a4a423b685",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create PPI edges from DEA PPI\n",
    "ppi_e <- string_db$get_interactions(string_db_dea$STRING_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbe527f4-b495-4ac4-a506-d683f6e66592",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute igraph object\n",
    "g <- igraph::graph_from_data_frame(\n",
    "    \n",
    "    d=ppi_e,\n",
    "    vertices=ppi_v, \n",
    "    directed=FALSE\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45195992-2a06-4e68-b4f7-0f42fff7bf90",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute GO enrichment \n",
    "enrichment <- string_db$get_enrichment(V(g)$name, category=\"Process\", methodMT=\"fdr\", iea=TRUE)\n",
    "openxlsx::write.xlsx(enrichment %>% dplyr::arrange(fdr), \"result/string/scRNAseq/enrichment_intmo_d1_ko_up.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1e86120-9dcc-4b01-a731-9c5f3be98876",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Connected components \n",
    "min_component_size <- 4  # keep components with >= 4 vertices\n",
    "\n",
    "g_components <- igraph::components(g)\n",
    "g <- igraph::delete_vertices(g, which(g_components$csize[g_components$membership] < min_component_size))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a3bbdb8-20f7-4299-b48e-de0881055716",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute Louvain cluster on combined score \n",
    "E(g)$weight <- E(g)$combined_score\n",
    "\n",
    "louvain <- igraph::cluster_louvain(g, weights=E(g)$weight)\n",
    "V(g)$louvain <- membership(louvain)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "603317d6-2982-4fde-8091-1ac7a809c0d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute GO enrichment for Louvain \n",
    "enrichment_louvain <- tibble::tibble(name=V(g)$name, louvain=V(g)$louvain) %>% group_by(louvain) %>%\n",
    "    \n",
    "    group_modify(~{\n",
    "        \n",
    "        enrichment <- string_db$get_enrichment(.x$name, category=\"Process\", methodMT=\"fdr\", iea=TRUE) %>% arrange(fdr)\n",
    "        if (is.null(enrichment) | nrow(enrichment)==0) {\n",
    "            \n",
    "            return(tibble::tibble())\n",
    "        \n",
    "        } else {\n",
    "\n",
    "            return(enrichment %>% arrange(fdr))\n",
    "            \n",
    "        }\n",
    "        \n",
    "    }\n",
    "                ) %>% ungroup()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2b75a6b-8e3e-4722-bcf8-481d94ddb158",
   "metadata": {},
   "outputs": [],
   "source": [
    "openxlsx::write.xlsx(enrichment_louvain %>% dplyr::filter(fdr<=0.01), \"result/string/scRNAseq/enrichment_louvain_intmo_d1_ko_up.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "874808da-4f2e-497d-b0f1-84603cbca9de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fetch GO enrichment \n",
    "enrichment_label <- enrichment_louvain %>% group_by(louvain) %>% arrange(fdr) %>% slice(1) %>% ungroup() %>% transmute(louvain, label=description)\n",
    "V(g)$enrichment_label <- enrichment_label$label[match(V(g)$louvain, enrichment_label$louvain)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64e869f9-1e63-4193-bef2-01683d5b5356",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove vertices without enrichment label \n",
    "g <- igraph::delete_vertices(g, which(is.na(V(g)$enrichment_label)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c19f597b-c7b7-40a0-9eeb-95c38c2e42b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot graph \n",
    "g_1 <- tidygraph::as_tbl_graph(g) %>% ggraph::ggraph(., layout=\"lgl\") +\n",
    "    geom_edge_link(alpha=0.2, width=0.1, color=\"#000000\") +\n",
    "    geom_node_point(aes(fill=factor(enrichment_label), size=abs(logFC)), color=\"black\", shape=21, stroke=0.1) +\n",
    "    geom_node_text(aes(label=ifelse(gene %in% c(\"Slc40a1\", \"Slc48a1\", \"Nr1h3\", \"Hmox1\", \"Stat2\", \"Irf7\", \"Eif2ak2\", \"Gdf15\"), gene, NA)), repel=FALSE, size=2.1, color=\"black\") + \n",
    "    scale_fill_manual(values=c(\"#6d6875\", \"#4f6d7a\", \"#4a6fa5\", \"#161b33\")) +\n",
    "    scale_size_continuous(range=c(0.5, 3)) + \n",
    "    theme(aspect.ratio=1) + theme_global_set(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0eeb6ef4-bfc0-4ca3-b2f8-6939565ef862",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/string_intmo_d1_ko_up.pdf\", width=10, height=10)\n",
    "\n",
    "gridExtra::grid.arrange(egg::set_panel_size(g_1, width=unit(5, \"cm\"), height=unit(5, \"cm\")))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09832274-4c09-4a36-9c6e-89f15686f6f6",
   "metadata": {},
   "source": [
    "## intMo D1 KO DOWN "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "405b302a-c856-422e-b05a-24a718dc2f86",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mapp DEA results to STRING DB\n",
    "string_db_dea <- dea_res_6[[\"intMo\"]] %>% dplyr::select(gene, avg_log2FC, p_val_adj) %>% dplyr::rename(logFC=avg_log2FC, padj=p_val_adj) %>% na.omit() %>% dplyr::filter(padj<=adj_pval_thr & logFC<=-log2FC_thr) %>% string_db$map(., \"gene\", removeUnmappedRows=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fef5e59a-abcd-476f-a21e-fa72162c0c75",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Labe PPI vertices \n",
    "ppi_v <- string_db_dea %>% dplyr::distinct() %>% dplyr::rename(name=STRING_id) %>% \n",
    "    dplyr::group_by(name) %>% dplyr::mutate(n=n()) %>% dplyr::filter(n==1) %>% dplyr::select(name, gene, logFC, padj) %>% \n",
    "    dplyr::mutate(node_group=ifelse(padj<=0.05 & logFC<=-0.50, \"neg\", ifelse(padj<=0.05 & logFC>=0.50, \"pos\", \"n.s.\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e1cc99e-0123-4594-9b92-98e67e029e74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create PPI edges from DEA PPI\n",
    "ppi_e <- string_db$get_interactions(string_db_dea$STRING_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "090b1c62-d814-4a83-9f4a-4ecd912e7972",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute igraph object\n",
    "g <- igraph::graph_from_data_frame(\n",
    "    \n",
    "    d=ppi_e,\n",
    "    vertices=ppi_v, \n",
    "    directed=FALSE\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f45b125-aeec-48a0-a19a-a6a3bb34c346",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute GO enrichment \n",
    "enrichment <- string_db$get_enrichment(V(g)$name, category=\"Process\", methodMT=\"fdr\", iea=TRUE)\n",
    "openxlsx::write.xlsx(enrichment %>% dplyr::arrange(fdr), \"result/string/scRNAseq/enrichment_intmo_d1_ko_down.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d3e44a3-c30c-4d33-ba4a-de5fb378b312",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Connected components \n",
    "min_component_size <- 4  # keep components with >= 4 vertices\n",
    "\n",
    "g_components <- igraph::components(g)\n",
    "g <- igraph::delete_vertices(g, which(g_components$csize[g_components$membership] < min_component_size))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74dca835-4b0b-4a00-826e-648fe46d7455",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute Louvain cluster on combined score \n",
    "E(g)$weight <- E(g)$combined_score\n",
    "\n",
    "louvain <- igraph::cluster_louvain(g, weights=E(g)$weight, resolution=1.0)  # or weights=NULL\n",
    "V(g)$louvain <- membership(louvain)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cb2a9ce-9604-4b48-a371-969605d63a4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute GO enrichment for Louvain \n",
    "enrichment_louvain <- tibble::tibble(name=V(g)$name, louvain=V(g)$louvain) %>% group_by(louvain) %>%\n",
    "    \n",
    "    group_modify(~{\n",
    "        \n",
    "        enrichment <- string_db$get_enrichment(.x$name, category=\"Process\", methodMT=\"fdr\", iea=TRUE) %>% arrange(fdr)\n",
    "        if (is.null(enrichment) | nrow(enrichment)==0) {\n",
    "            \n",
    "            return(tibble::tibble())\n",
    "        \n",
    "        } else {\n",
    "\n",
    "            return(enrichment %>% arrange(fdr))\n",
    "            \n",
    "        }\n",
    "        \n",
    "    }\n",
    "                ) %>% ungroup()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b049279-77dc-421e-8c57-02b53f87664c",
   "metadata": {},
   "outputs": [],
   "source": [
    "openxlsx::write.xlsx(enrichment_louvain %>% dplyr::filter(fdr<=0.01), \"result/string/scRNAseq/enrichment_louvain_intmo_d1_ko_down.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4af2d87-c33f-4497-b704-db490d16f799",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fetch GO enrichment \n",
    "enrichment_label <- enrichment_louvain %>% group_by(louvain) %>% arrange(fdr) %>% slice(1) %>% ungroup() %>% transmute(louvain, label=description)\n",
    "V(g)$enrichment_label <- enrichment_label$label[match(V(g)$louvain, enrichment_label$louvain)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e0c215b-7bf0-4a8f-9347-173f0b1fd7dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove vertices without enrichment label \n",
    "g <- igraph::delete_vertices(g, which(is.na(V(g)$enrichment_label)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bd6dc02-791d-4ae7-842f-43becc961eba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot graph \n",
    "g_1 <- tidygraph::as_tbl_graph(g) %>% ggraph::ggraph(., layout=\"lgl\") +\n",
    "    geom_edge_link(alpha=0.2, width=0.1, color=\"#000000\") +\n",
    "    geom_node_point(aes(fill=factor(enrichment_label), size=abs(logFC)), color=\"black\", shape=21, stroke=0.1) +\n",
    "    geom_node_text(aes(label=ifelse(gene %in% c(\"Slc40a1\", \"Slc48a1\", \"Nr1h3\", \"Hmox1\", \"Stat2\", \"Irf7\", \"Eif2ak2\", \"Gdf15\"), gene, NA)), repel=FALSE, size=2.1, color=\"black\") + \n",
    "    scale_fill_manual(values=c(\"#ffcdb2\", \"#ffb4a2\", \"#e5989b\", \"#b5838d\", \"#cfe1b9\", \"#97a97c\", \"#718355\", \"#c0d6df\", \"#474973\")) +\n",
    "    scale_size_continuous(range=c(0.5, 3)) + \n",
    "    theme(aspect.ratio=1) + theme_global_set(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae3009d8-c637-42a0-940c-70fba04acf38",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/string_intmo_d1_ko_down.pdf\", width=10, height=10)\n",
    "\n",
    "gridExtra::grid.arrange(egg::set_panel_size(g_1, width=unit(5, \"cm\"), height=unit(5, \"cm\")))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cca0dcc6-1a31-441f-8bb0-fc94f1f47a2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "e <- igraph::as_data_frame(g, what=\"edges\")\n",
    "v <- igraph::as_data_frame(g, what=\"vertices\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23f10665-d367-47a7-aec5-ea89648f9892",
   "metadata": {},
   "outputs": [],
   "source": [
    "v_1 <- v %>% dplyr::filter(louvain==2) %>% dplyr::select(name, gene, louvain) %>% dplyr::rename(target_name=name, target_gene=gene, target_louvain=louvain)\n",
    "e_1 <- e %>% dplyr::select(from, to) %>% dplyr::filter(from %in% v_1$target_name | to %in% v_1$target_name) %>% dplyr::filter(!(from %in% v_1$target_name & to %in% v_1$target_name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d32d7a57-ca4c-4cf1-b4b9-60af8e3c1b82",
   "metadata": {},
   "outputs": [],
   "source": [
    "v_2 <- v %>% dplyr::filter(louvain!=2) %>% dplyr::select(name, gene, louvain) %>% dplyr::rename(source_name=name, source_gene=gene, source_louvain=louvain)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37157031-ce63-4468-92f5-b0f94cf77372",
   "metadata": {},
   "outputs": [],
   "source": [
    "connect <- rbind(\n",
    "    \n",
    "    dplyr::left_join(v_1, e_1, by=c(\"target_name\"=\"from\")) %>% dplyr::mutate(from=target_name), \n",
    "    dplyr::left_join(v_1, e_1, by=c(\"target_name\"=\"to\"))  %>% dplyr::mutate(to=target_name)\n",
    "    \n",
    ") %>% na.omit() %>% dplyr::distinct() %>% dplyr::mutate(source_name=ifelse(target_name!=to, to, from)) %>% dplyr::select(-from, -to) %>% dplyr::left_join(., v_2, by = join_by(source_name)) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11b0b43e-1473-48ac-98ad-f91f0ae8a8fd",
   "metadata": {},
   "source": [
    "## intMo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43c8b539-3f48-4103-86ea-b75a5dfe57bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mapp DEA results to STRING DB\n",
    "string_db_dea <- dea_res_6[[\"intMo\"]] %>% dplyr::select(gene, avg_log2FC, p_val_adj) %>% dplyr::rename(logFC=avg_log2FC, padj=p_val_adj) %>% na.omit() %>% dplyr::filter(padj<=adj_pval_thr & abs(logFC)>=log2FC_thr) %>% string_db$map(., \"gene\", removeUnmappedRows=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff1fc5bc-452d-457b-8e4d-a9c218e2bee7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Labe PPI vertices \n",
    "ppi_v <- string_db_dea %>% dplyr::distinct() %>% dplyr::rename(name=STRING_id) %>% \n",
    "    dplyr::group_by(name) %>% dplyr::mutate(n=n()) %>% dplyr::filter(n==1) %>% dplyr::select(name, gene, logFC, padj) %>% \n",
    "    dplyr::mutate(node_group=ifelse(padj<=0.05 & logFC<=-0.50, \"IFNAR_fl_CpG_D1\", ifelse(padj<=0.05 & logFC>=0.50, \"IFNAR_fl_LysM_cre_CpG_D1\", \"n.s.\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8ca949a-2431-4b87-aac2-373a44496887",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create PPI edges from DEA PPI\n",
    "ppi_e <- string_db$get_interactions(string_db_dea$STRING_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfc1f5ca-6548-4a31-be0e-6ee6877960d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute igraph object\n",
    "g <- igraph::graph_from_data_frame(\n",
    "    \n",
    "    d=ppi_e,\n",
    "    vertices=ppi_v, \n",
    "    directed=FALSE\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d446b6e1-a041-49ee-831b-6fd1f02665aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Connected components \n",
    "min_component_size <- 4  # keep components with >= 4 vertices\n",
    "\n",
    "g_components <- igraph::components(g)\n",
    "g <- igraph::delete_vertices(g, which(g_components$csize[g_components$membership] < min_component_size))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9274464-3715-4dcb-88c3-89a2e1f8126f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute Louvain cluster on combined score \n",
    "E(g)$weight <- E(g)$combined_score\n",
    "\n",
    "louvain <- igraph::cluster_louvain(g, weights=E(g)$weight, resolution=1.0)  # or weights=NULL\n",
    "V(g)$louvain <- membership(louvain)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06cd418f-c248-41c6-8a5e-4fe9b52b8fce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute GO enrichment for Louvain \n",
    "enrichment_louvain <- tibble::tibble(name=V(g)$name, louvain=V(g)$louvain) %>% group_by(louvain) %>%\n",
    "    \n",
    "    group_modify(~{\n",
    "        \n",
    "        enrichment <- string_db$get_enrichment(.x$name, category=\"Process\", methodMT=\"fdr\", iea=TRUE) %>% arrange(fdr)\n",
    "        if (is.null(enrichment) | nrow(enrichment)==0) {\n",
    "            \n",
    "            return(tibble::tibble())\n",
    "        \n",
    "        } else {\n",
    "\n",
    "            return(enrichment %>% arrange(fdr))\n",
    "            \n",
    "        }\n",
    "        \n",
    "    }\n",
    "                ) %>% ungroup()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ece16bec-00a8-43fc-b159-11f29a119133",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fetch GO enrichment \n",
    "enrichment_label <- enrichment_louvain %>% group_by(louvain) %>% arrange(fdr) %>% slice(1) %>% ungroup() %>% transmute(louvain, label=description)\n",
    "V(g)$enrichment_label <- enrichment_label$label[match(V(g)$louvain, enrichment_label$louvain)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d2dffb4-9cd2-4192-b7ce-58dbe690f3cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove vertices without enrichment label \n",
    "g <- igraph::delete_vertices(g, which(is.na(V(g)$enrichment_label)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25d07a3f-bf03-4a22-84c4-355ba40b1af1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot graph \n",
    "g_1 <- tidygraph::as_tbl_graph(g) %>% ggraph::ggraph(., layout=\"lgl\") +\n",
    "    geom_edge_link(alpha=0.2, width=0.1, color=\"#000000\") +\n",
    "    geom_node_point(aes(fill=node_group, size=abs(logFC)), color=\"black\", shape=21, stroke=0.1) +\n",
    "    geom_node_text(aes(label=ifelse(gene %in% c(\"Slc40a1\", \"Slc48a1\", \"Nr1h3\", \"Hmox1\", \"Stat2\", \"Irf7\", \"Eif2ak2\", \"Gdf15\"), gene, NA)), repel=FALSE, size=2.1, color=\"black\") + \n",
    "    scale_fill_manual(values=color$sample_group) +\n",
    "    scale_size_continuous(range=c(0.5, 3)) + \n",
    "    theme(aspect.ratio=1) + theme_global_set(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93a61076-e063-4b2f-968b-52f878409efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/string_intmo_d1.pdf\", width=10, height=10)\n",
    "\n",
    "gridExtra::grid.arrange(egg::set_panel_size(g_1, width=unit(5, \"cm\"), height=unit(5, \"cm\")))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55510124-e19d-4d86-b9b3-9c6fc6cad538",
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- rbind(\n",
    "    \n",
    "    openxlsx::read.xlsx(\"result/string/scRNAseq/enrichment_intmo_d1_ko_down.xlsx\") %>% dplyr::mutate(fill=\"IFNAR_fl_CpG_D1\"), \n",
    "    openxlsx::read.xlsx(\"result/string/scRNAseq/enrichment_intmo_d1_ko_up.xlsx\") %>% dplyr::mutate(fill=\"IFNAR_fl_LysM_cre_CpG_D1\")\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "279da1e0-7128-4aee-8f59-6e73af6d7d77",
   "metadata": {},
   "outputs": [],
   "source": [
    "go_select <- c(\n",
    "\n",
    "    \"Response to interferon-alpha\", \n",
    "    \"Response to interferon-beta\", \n",
    "\n",
    "    \"Positive regulation of tumor necrosis factor production\", \n",
    "\n",
    "    \"Response to stress\", \n",
    "    \n",
    "    \"Cellular iron ion homeostasis\", \n",
    "    \"Iron ion transport\", \n",
    "\n",
    "    \"Response to interferon-gamma\", \n",
    "    \"Peptide antigen assembly with MHC class II protein complex\"\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed2f9a15-0f0b-4c72-a628-4948779a0cf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "data %>% dplyr::filter(description %in% go_select)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10fe65d4-cf61-4f71-95f8-fb054b0dc3b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- data %>% dplyr::filter(description %in% go_select) %>% dplyr::mutate(score=-log10(fdr)) %>% dplyr::group_by(fill) %>% dplyr::arrange(score) %>% dplyr::ungroup() %>% \n",
    "    dplyr::mutate(description=factor(description, levels=rev(description)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32fec27f-0974-4919-96af-168894530ac5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# grn_p_1 <- ggplot(data, aes(x=score, y=description, fill=fill)) + \n",
    "#     geom_bar(stat=\"identity\", width=0.8, color=\"black\", linewidth=0.1) + \n",
    "#     scale_fill_manual(values=c(color$sample_group, \"n.s.\"=\"#7f7f7f\")) + theme_global_set(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f17a00a8-3b8d-4723-bd73-cd6f1dd7d927",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pdf(\"result/figures/1_scRNAseq/panel_5/enrichment_intmo_d1.pdf\", width=4, height=8)\n",
    "\n",
    "# set_height <- 2.00 \n",
    "# set_width <- 1\n",
    "\n",
    "# grn_p_1 + ggh4x::force_panelsizes(cols=unit(set_width, \"cm\"), total_height=unit(set_height, \"cm\"))\n",
    "\n",
    "# dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc03ee44-7fc1-4793-8bf4-b1f5a01eaf54",
   "metadata": {},
   "source": [
    "## Genotype comparison - M5 GO:BP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c9289eb-8412-4d07-b34e-8ea15840dce8",
   "metadata": {},
   "outputs": [],
   "source": [
    "compute_gsea <- FALSE\n",
    "celltype_select <- c(\"intMo\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "885824cd-7f1d-4363-975d-14d058961671",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(compute_gsea) {\n",
    "\n",
    "    gsea_res <- lapply(celltype_select, function(i) {\n",
    "    \n",
    "        gsea_res_1 <- gsea(dea_res_5[[i]], category=\"M5\", subcategory=\"GO:BP\", gene_set=NULL) %>% dplyr::mutate(col_split=\"Ctl\", col_label=i)\n",
    "        gsea_res_2 <- gsea(dea_res_6[[i]], category=\"M5\", subcategory=\"GO:BP\", gene_set=NULL) %>% dplyr::mutate(col_split=\"D1\", col_label=i)\n",
    "        gsea_res_3 <- gsea(dea_res_7[[i]], category=\"M5\", subcategory=\"GO:BP\", gene_set=NULL) %>% dplyr::mutate(col_split=\"D3\", col_label=i)\n",
    "    \n",
    "        gsea_res <- rbind(gsea_res_1, gsea_res_2, gsea_res_3)\n",
    "    \n",
    "        return(gsea_res)\n",
    "        \n",
    "    }\n",
    "          )\n",
    "\n",
    "    saveRDS(gsea_res, \"result/gsea/scRNAseq/gsea_ctl_d1_d3_wt_vs_ko_m5_gobp.rds\")\n",
    "\n",
    "} else {\n",
    "\n",
    "    gsea_res <- readRDS(\"result/gsea/scRNAseq/gsea_ctl_d1_d3_wt_vs_ko_m5_gobp.rds\")\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c2a8bf5-2982-4e06-85bc-c8b458c138e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "is.na(gsea_res$leadingEdge)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba957b2c-78b2-4ba5-8a1a-e5ea52c90a21",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res <- do.call(rbind, gsea_res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4ec549f-a660-49fa-8ec4-51f0d7ff23f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res[gsea_res$pathway %in% grep(\"IRON\", gsea_res$pathway, value=TRUE), ] %>% dplyr::filter(pval<=0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3869b7e0-1870-4bc2-b6b0-fd4b6aeb5329",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1b2cc42-6b52-42b9-865d-7398eb673076",
   "metadata": {},
   "outputs": [],
   "source": [
    "grep(\"IRON\", gsea_res %>% dplyr::pull(pathway), value=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63b1abc0-9358-418e-812d-de21ea45cb7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs_name_select <- gsea_res %>% dplyr::filter(padj<=0.05 & abs(NES)>=1) %>% dplyr::pull(pathway)\n",
    "# gsea_res <- gsea_res %>% dplyr::filter(pathway %in% gs_name_select)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d952f82-1080-4577-9e36-cdbd6c1f2bef",
   "metadata": {},
   "outputs": [],
   "source": [
    "grep(\"IRON\", gsea_res$pathway, value=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6139a685-8339-42e2-ae7d-7cbc061f55a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- gsea_res %>% dplyr::filter(pathway %in% gs_name_select) %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=padj,\n",
    "        values_fill=1\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1a51c94-5778-4d10-97ae-767a69147b47",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_2 <- gsea_res %>% dplyr::filter(pathway %in% gs_name_select) %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=NES,\n",
    "        values_fill=0\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55514ba2-b2b2-449c-8f67-509d0da3df7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_label <- mat_1$col_label\n",
    "col_split <- mat_1$col_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3f5482c-2d04-419f-afac-fa1776a3f428",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- mat_1 %>% dplyr::select(-col_label, -col_split) %>% t()\n",
    "mat_2 <- mat_2 %>% dplyr::select(-col_label, -col_split) %>% t()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ad8479d-4a47-4e4f-9972-8061ac80c91e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# mat_1 <- mat_1[gs_name_order, ]\n",
    "# mat_2 <- mat_2[gs_name_order, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57c3d582-4c80-42a8-a057-df8780f9a190",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_label[col_label==\"cDC.mig.\"] <- \"cDC mig.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abdc1726-c16c-4468-96d5-003e05e0aeb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/hm_gsea_ctl_d1_d3_wt_vs_ko_m5_gobp.pdf\", width=8, height=100)\n",
    "\n",
    "source(\"bin/gsea_pl.R\")\n",
    "gsea_hm(mat_1, mat_2, col_label=col_label, col_split=col_split, row_split=NULL, pathway_suffix=\"HALLMARK\", color_neg=color$sample_group[\"IFNAR_fl_CpG_D1\"], color_pos=color$sample_group[\"IFNAR_fl_LysM_cre_CpG_D1\"], fontsize_select=2)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc5432fa-6bdf-40c1-a391-3185dd3bed3f",
   "metadata": {},
   "source": [
    "## Violine gene expression plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4168949f-ccb8-43c3-a8dd-da784c206c3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_select <- c(\"Ly6c2\", \"Spic\", \"Cd163\", \"Gdf15\", \"Ccl2\", \"Slfn4\", \"Bcl2l1\", \"Hmox1\", \"Jun\")\n",
    "celltype_select <- c(\"cMo\", \"intMo\", \"RPM\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb9f59b6-a237-4468-892b-2d1d19272111",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat <- cbind(so@meta.data[, c(\"sample_group\", \"celltype_low\"), drop=FALSE], t(GetAssayData(so, assay=\"RNA\", slot=\"data\")[gene_select, ])) %>% \n",
    "    pivot_longer(cols=-c(sample_group, celltype_low), names_to=\"gene\", values_to=\"expression\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06a030b5-a3e0-4e6b-a25a-1367fe96696e",
   "metadata": {},
   "outputs": [],
   "source": [
    "map <- data.frame(\n",
    "    \n",
    "    sample_group=c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_LysM_cre_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_LysM_cre_CpG_D1\", \"IFNAR_fl_CpG_D3\", \"IFNAR_fl_LysM_cre_CpG_D3\"),\n",
    "    time_point=c(\"Ctl\", \"Ctl\", \"D1\", \"D1\", \"D3\", \"D3\"),\n",
    "    genotype=c(\"+/+\", \"cre/+\", \"+/+\", \"cre/+\", \"+/+\", \"cre/+\")\n",
    ")\n",
    "\n",
    "mat <- mat %>% dplyr::left_join(., map, by=join_by(sample_group)) %>% \n",
    "    dplyr::mutate(gene=factor(gene, levels=gene_select))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7ee6260-001a-485d-84db-7619e4a3d9cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "genotype <- c(\"#732E8F\", \"#D390BF\", \"#FA8775\")\n",
    "names(genotype) <- c(\"WT\", \"+/+\", \"cre/+\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d442739b-0a5a-4ce6-a930-1ba8def20c61",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=5, repr.plot.height=7.5)\n",
    "\n",
    "exp_p <- ggplot(mat %>% dplyr::filter(celltype_low %in% celltype_select), aes(x=time_point, y=expression, color=sample_group, fill=sample_group)) + \n",
    "\n",
    "        geom_jitter(position=position_jitterdodge(jitter.width=0.20, dodge.width=1.0), shape=21, stroke=0.1, size=0.75, show.legend=FALSE, color=\"black\") + \n",
    "        # geom_boxplot(position=position_dodge(width=1.0), width=0.20, color=\"black\", alpha=1, outlier.color=NA) + \n",
    "        \n",
    "        xlab(\"\") + ylab(\"log(CPTT)\") + \n",
    "        ggtitle(\"Spic\") + \n",
    "        facet_grid(gene~celltype_low, scales=\"free\") + \n",
    "        scale_color_manual(values=color$sample_group) + \n",
    "        scale_fill_manual(values=color$sample_group)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21cc23ae-0b66-47e4-86ec-efc515519adf",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/exp_myeloid.pdf\", width=3, height=8)\n",
    "\n",
    "set_height <- 1 * length(gene_select)\n",
    "set_width <- 3 * length(celltype_select)\n",
    "\n",
    "exp_p + ggh4x::force_panelsizes(cols=unit(set_width, \"mm\"), total_height=unit(set_height, \"cm\"))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ab3da78-0778-49d1-818b-06668d452dc8",
   "metadata": {},
   "source": [
    "# Gene regulatory network analysis "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bec7dcc9-9441-4d05-a97d-6b754c4885de",
   "metadata": {},
   "outputs": [],
   "source": [
    "auc_mtx_3 <- read.csv(\"result/scenic/scRNAseq/res_3/auc_mtx.csv\", row.names=1) %>% t()\n",
    "rownames(auc_mtx_3) <- gsub(\"\\\\.\", \"\", rownames(auc_mtx_3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acb535e7-c927-486b-b2bc-6125cc46d963",
   "metadata": {},
   "outputs": [],
   "source": [
    "so_3 <- subset(so, subset=cell_id %in% colnames(auc_mtx_3))\n",
    "auc_mtx_3 <- auc_mtx_3[, colnames(so_3)]\n",
    "so_3[[\"AUC\"]] <- CreateAssayObject(data=auc_mtx_3[, colnames(so_3)], min.cells=0, min.features=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41874325-b4c4-4d58-8037-aaeaf4470467",
   "metadata": {},
   "source": [
    "## Jitter plot "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ce707c7-b1a2-4aff-a946-8cfd0842756e",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat <- cbind(so_3@meta.data, t(auc_mtx_3[\"Spic\", , drop=FALSE]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ff8d13e-9981-4338-b246-e8da01e283a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "auc_p_1 <- ggplot(mat %>% dplyr::filter(celltype_low %in% \"intMo\"), aes(x=genotype, y=Spic, color=sample_group, fill=sample_group)) + \n",
    "         \n",
    "        geom_jitter(position=position_jitterdodge(jitter.width=0.75, dodge.width=1.0), shape=21, stroke=0.1, size=0.75, show.legend=FALSE, color=\"black\") + \n",
    "        geom_boxplot(color=\"black\", width=0.50, box.linewidth=0.1, whisker.linewidth=0.1, median.linewidth=0.1, outliers=FALSE) +\n",
    "        xlab(\"\") + ylab(\"AUC\") +\n",
    "        ylim(c(0, 0.30)) + \n",
    "        ggtitle(\"Spic\") + \n",
    "        facet_grid(celltype_low~dpi) + \n",
    "        scale_color_manual(values=color$sample_group) + \n",
    "        scale_fill_manual(values=color$sample_group) + theme_global_set(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12229dfc-0728-4dd1-88c4-2f98756af129",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/auc_spic_myeloid.pdf\", width=4, height=8)\n",
    "\n",
    "set_height <- 1.5 * length(celltype_select)\n",
    "set_width <- 0.75\n",
    "\n",
    "exp_p + ggh4x::force_panelsizes(cols=unit(set_width, \"cm\"), total_height=unit(set_height, \"cm\"))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72737235-ab53-4c3d-92df-59a2048c8fdc",
   "metadata": {},
   "source": [
    "## Bar plot differential AUC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60b1a5a3-d323-4082-b275-4456dbea46be",
   "metadata": {},
   "outputs": [],
   "source": [
    "grn_wilcox_res <- readRDS(\"result/scenic/scRNAseq/auc_wilcox_ery_ifnar_fl_lysm_cre_cpg_d1_vs_ifnar_fl_cpg_d1.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0eff06a9-27b8-48bc-bcf2-c3d48af078d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- grn_wilcox_res[[\"intMo\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55afe7cd-d216-4495-8781-eeab7a914022",
   "metadata": {},
   "outputs": [],
   "source": [
    "label <- c(\n",
    "    \n",
    "    data %>% arrange(p_val_adj) %>% dplyr::filter(sign(avg_log2FC)==-1 & rank_1 <= 50 & p_val_adj<=0.01) %>% dplyr::slice_head(n=5) %>% dplyr::pull(module), \n",
    "    data %>% arrange(p_val_adj) %>% dplyr::filter(sign(avg_log2FC)==+1 & rank_2 <= 50 & p_val_adj<=0.01) %>% dplyr::slice_head(n=5) %>% dplyr::pull(module)\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2980e6e-2bae-4035-8a48-bc512c14e30d",
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- data %>% dplyr::filter(module %in% label) %>% dplyr::mutate(score=-sign(avg_log2FC)*log10(p_val_adj)) %>% dplyr::arrange(score) %>% \n",
    "    dplyr::mutate(fill=ifelse(sign(score)==+1, \"IFNAR_fl_LysM_cre_CpG_D1\", \"IFNAR_fl_CpG_D1\")) %>% \n",
    "    dplyr::mutate(module=factor(module, levels=rev(module)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b901dfb-671f-4f6d-8dac-e1eed2e79f1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "grn_p_1 <- ggplot(data, aes(x=score, y=module, fill=fill)) + \n",
    "    geom_bar(stat=\"identity\", width=0.8, color=\"black\", linewidth=0.1) + \n",
    "    scale_fill_manual(values=c(color$sample_group, \"n.s.\"=\"#7f7f7f\")) + theme_global_set(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a23341ce-880b-455a-bfc8-1574c87904fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/auc_diff_intmo_d1.pdf\", width=4, height=8)\n",
    "\n",
    "set_height <- 2.00 * length(celltype_select)\n",
    "set_width <- 1\n",
    "\n",
    "grn_p_1 + ggh4x::force_panelsizes(cols=unit(set_width, \"cm\"), total_height=unit(set_height, \"cm\"))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff664abc-1793-481d-8b25-cefbb946187e",
   "metadata": {},
   "source": [
    "# CCI DEA "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec59b755-f507-4b45-aa40-e32141627ea4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load source \n",
    "source(\"bin/cci_pp.R\")\n",
    "source(\"bin/cci_pl.R\")\n",
    "\n",
    "# Parameter \n",
    "pval_prob_thr <- 0.05\n",
    "p_val_adj_dea <- 0.05\n",
    "pathway_name_thr <- c(\"ICAM\", \"VCAM\", \"SIRP\", \"KIT\", \"TGFb\", \"TNF\", \"CXCL\", \"CCL\")\n",
    "\n",
    "# CellChat results\n",
    "lr_res <- read.csv(\"result/cci/lr_res.csv\", row.names=1)\n",
    "\n",
    "# Add reversed Cell-cell interactions \n",
    "db <- CellChat::CellChatDB.mouse\n",
    "\n",
    "# Impute custom interactions\n",
    "options(stringsAsFactors=FALSE)\n",
    "interaction <- read.csv(\"data/reference/cellchat/interaction.csv\", row.names=1, na.strings=c(\"\"))\n",
    "interaction[is.na(interaction)] <- \"\"\n",
    "db$interaction <- rbind(db$interaction, interaction)\n",
    "\n",
    "# Get cell-cell contact interaction \n",
    "db <- CellChat::subsetDB(db, search=c(\"Cell-Cell Contact\"), key=\"annotation\")\n",
    "\n",
    "# Subset lr results for cell-cell contact forward \n",
    "lr_res_ccc_fwd <- lr_res %>% dplyr::filter(lr_res$interaction_name %in% db$interaction$interaction_name)\n",
    "\n",
    "# Reverse cell-cell contact \n",
    "lr_res_ccc_rev <- data.frame(\n",
    "\n",
    "    source=lr_res_ccc_fwd$target, \n",
    "    target=lr_res_ccc_fwd$source, \n",
    "    interaction_name=paste0(lr_res_ccc_fwd$receptor_symbol, \"_\", lr_res_ccc_fwd$ligand_symbol), \n",
    "    prob=lr_res_ccc_fwd$prob, \n",
    "    pval=lr_res_ccc_fwd$pval, \n",
    "    pathway_name=lr_res_ccc_fwd$pathway_name, \n",
    "    ligand_symbol=lr_res_ccc_fwd$receptor_symbol, \n",
    "    receptor_symbol=lr_res_ccc_fwd$ligand_symbol, \n",
    "    sample_group=lr_res_ccc_fwd$sample_group\n",
    "    \n",
    ")\n",
    "\n",
    "# Combine results \n",
    "lr_res_ccc <- rbind(lr_res_ccc_fwd, lr_res_ccc_rev)\n",
    "\n",
    "# Assemble LR results \n",
    "lr_res <- rbind(lr_res[!lr_res$interaction_name %in% lr_res_ccc$interaction_name, ], lr_res_ccc)\n",
    "\n",
    "# Subset by interaction pathway\n",
    "lr_res <- lr_res[lr_res$pathway_name %in% pathway_name_thr, ]\n",
    "lr_res$pathway_name <- factor(lr_res$pathway_name, levels=pathway_name_thr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67410385-21c6-4e7a-8671-1fdbdadf36b5",
   "metadata": {},
   "source": [
    "# DEA results "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d60c5142-d878-4dfe-957d-426e2a819a7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "dea_mat_5 <- dea_res_5[c(\"cMo\", \"intMo\", \"RPM\", \"cDC2\", \"Basophil\", \"MEP\", \"Proerythroblast\", \"Erythroblast\")]\n",
    "dea_mat_5 <- lapply(dea_mat_5, function(x) {list(\"Ctl\"=x)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1347d3d0-1bde-44f0-882f-3102ba6ba5ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "dea_mat_6<- dea_res_6[c(\"cMo\", \"intMo\", \"RPM\", \"cDC2\", \"Basophil\", \"MEP\", \"Proerythroblast\", \"Erythroblast\")]\n",
    "dea_mat_6 <- lapply(dea_mat_6, function(x) {list(\"D1\"=x)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f3261de-4089-4970-807f-50ace3e7ae72",
   "metadata": {},
   "outputs": [],
   "source": [
    "dea_mat_7 <- dea_res_7[c(\"cMo\", \"intMo\", \"RPM\", \"cDC2\", \"Basophil\", \"MEP\", \"Proerythroblast\", \"Erythroblast\")]\n",
    "dea_mat_7 <- lapply(dea_mat_7, function(x) {list(\"D3\"=x)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fa13ead-4344-4d8d-91bf-9fe99ff9bcd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "dea_mat <- Map(c, dea_mat_5, dea_mat_6, dea_mat_7)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c48a72b-db65-4fd1-a47a-81423d4d9799",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## PTPG B (+/+) vs D1 (+/+) and D3 (+/+)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f08fc2ce-da91-4161-80a9-ed3281425407",
   "metadata": {},
   "source": [
    "## cMo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "003fb3e9-c960-4a2d-8241-b743963df5d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"cMo\")\n",
    "target <- c(\"MEP\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "462e97b1-b38a-4f27-a5bd-7716913ac0a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80d21cd2-cd95-4e0b-95b0-c629a4b37417",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/cci_cmo_mep_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18e4ce5b-ea26-4e93-936a-0048d8f7e2d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"cMo\")\n",
    "target <- c(\"Proerythroblast\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8526ccc3-5d19-42bc-8f8a-9febc3749351",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d1a01fd-b574-4c76-8901-092caa18b404",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/cci_cmo_proeb_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d290092c-21c1-431a-ab30-1cf2562eff89",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"cMo\")\n",
    "target <- c(\"Erythroblast\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3fcc361-927a-4b1b-a265-80fc3cc8d5d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e2f70a2-0f05-46ea-a89b-f1974e85494e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/cci_cmo_eb_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f691ebc-9fc2-4997-bb21-bac8cbb90db6",
   "metadata": {},
   "source": [
    "## intMo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b65a947-44dc-4f2a-a16c-105d000e247a",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"intMo\")\n",
    "target <- c(\"MEP\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e643f889-0815-49f9-9034-eb6ba03f5bef",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aac4fd3a-632f-4bd9-8eef-2cade700a7d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/cci_intmo_mep_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f12cfed-1372-4a14-b341-6e776b765d6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"intMo\")\n",
    "target <- c(\"Proerythroblast\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16f44f38-859b-4337-a545-5c2e05da9178",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71a43bc6-400e-47b1-b23a-66a4e944d4ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/cci_intmo_proeb_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69480d57-2a72-41d9-8385-52117e0012a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"intMo\")\n",
    "target <- c(\"Erythroblast\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16af9b97-b869-4128-a6bb-194a597f5d5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba316e76-2191-48ab-a35a-1fe2a7d33a71",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/cci_intmo_eb_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "074532ad-a423-4100-b9b0-1a3d1904063b",
   "metadata": {},
   "source": [
    "## RPM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c89da626-687b-49c2-94b0-1cddcd97f8d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"RPM\")\n",
    "target <- c(\"MEP\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bba0449-c718-42f4-a8d4-eb16a204f4f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc02eac0-37bc-486e-9f4a-422f17f6604e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/cci_rpm_mep_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c90c12bd-e06c-4fa7-b3c8-e8842916db4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"RPM\")\n",
    "target <- c(\"Proerythroblast\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3225ebbc-095d-4d91-ae0d-e5d6b6d3f6a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a6c0dec-fcd0-4ef0-b368-a2892ae33c01",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/cci_rpm_proeb_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58ba8d68-224a-4ce3-9412-4623d6fbf1f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"RPM\")\n",
    "target <- c(\"Erythroblast\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72af504a-8e9e-4139-94af-1e93e722fc0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39388a7c-74b7-42d3-a138-248cd8600c7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/cci_rpm_eb_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62c60458-8f68-4833-8c7e-967ef4efe998",
   "metadata": {},
   "source": [
    "# Interaction plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca3919c0-95db-40cf-99bb-7fd54e463343",
   "metadata": {},
   "outputs": [],
   "source": [
    "sources <- c(\"RPM\", \"intMo\", \"cMo\")\n",
    "targets <- c(\"MEP\", \"Proerythroblast\", \"Erythroblast\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0037cc6c-331c-4ab5-a9d9-16b1cc3c9416",
   "metadata": {},
   "outputs": [],
   "source": [
    "cco_1 <- readRDS(\"data/scRNAseq/cci/cco_ifnar_fl_baseline_d0.rds\")\n",
    "cco_2 <- readRDS(\"data/scRNAseq/cci/cco_ifnar_fl_cpg_d1.rds\")\n",
    "cco_3 <- readRDS(\"data/scRNAseq/cci/cco_ifnar_fl_cpg_d3.rds\")\n",
    "cco_4 <- readRDS(\"data/scRNAseq/cci/cco_ifnar_fl_lysm_cre_baseline_d0.rds\")\n",
    "cco_5 <- readRDS(\"data/scRNAseq/cci/cco_ifnar_fl_lysm_cre_cpg_d1.rds\")\n",
    "cco_6 <- readRDS(\"data/scRNAseq/cci/cco_ifnar_fl_lysm_cre_cpg_d3.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3011b27-7789-4eca-8e3d-ab0ad506eafb",
   "metadata": {},
   "outputs": [],
   "source": [
    "cco_1 <- subsetCellChat(object=cco_1, idents.use=c(sources, targets))\n",
    "cco_2 <- subsetCellChat(object=cco_2, idents.use=c(sources, targets))\n",
    "cco_3 <- subsetCellChat(object=cco_3, idents.use=c(sources, targets))\n",
    "cco_4 <- subsetCellChat(object=cco_4, idents.use=c(sources, targets))\n",
    "cco_5 <- subsetCellChat(object=cco_5, idents.use=c(sources, targets))\n",
    "cco_6 <- subsetCellChat(object=cco_6, idents.use=c(sources, targets))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f3e2c0c-fef7-4539-880e-c30cca6baea8",
   "metadata": {},
   "outputs": [],
   "source": [
    "levels(cco_1@idents) <- c(targets, sources)\n",
    "levels(cco_2@idents) <- c(targets, sources)\n",
    "levels(cco_3@idents) <- c(targets, sources)\n",
    "levels(cco_4@idents) <- c(targets, sources)\n",
    "levels(cco_5@idents) <- c(targets, sources)\n",
    "levels(cco_6@idents) <- c(targets, sources)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b7bf337-daf2-4d4f-a30e-8507c6974e97",
   "metadata": {},
   "outputs": [],
   "source": [
    "color_use <- color$celltype_low[names(color$celltype_low) %in% c(sources, targets)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10ec20f9-774b-462e-ac37-991ccb3d2020",
   "metadata": {},
   "outputs": [],
   "source": [
    "cco_list <- list(cco_1, cco_2, cco_3, cco_4, cco_5, cco_6)\n",
    "\n",
    "q_prob_thr <- quantile(c(cco_1@net$prob, cco_2@net$prob, cco_3@net$prob), 0.90)\n",
    "for (i in seq_along(cco_list)) {\n",
    "    \n",
    "    prob_arr <- cco_list[[i]]@net$prob\n",
    "    prob_arr[prob_arr > q_prob_thr] <- q_prob_thr\n",
    "    cco_list[[i]]@net$prob <- prob_arr\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb06d56a-6d68-4be4-9277-a21e7987ef04",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/cp_cci_net_ctl_ko.pdf\", width=2.0, height=2.0)\n",
    "\n",
    "    replayPlot(netVisual_chord_gene(cco_list[[4]], slot.name=\"net\", sources.use=sources, targets.use=targets, signaling=pathway_name_thr, color.use=color_use, title.name=NULL, show.legend=FALSE, small.gap=1, big.gap=5, lab.cex=0.5))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4deee3b9-f1f4-4764-b9d8-ea6b10dd66d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/cp_cci_net_d1_ko.pdf\", width=2.0, height=2.0)\n",
    "\n",
    "    replayPlot(netVisual_chord_gene(cco_list[[5]], slot.name=\"net\", sources.use=sources, targets.use=targets, signaling=pathway_name_thr, color.use=color_use, title.name=NULL, show.legend=FALSE, small.gap=1, big.gap=5, lab.cex=0.5))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11ad2ca6-235a-462a-b090-a0804e3e3092",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/cp_cci_net_d3_ko.pdf\", width=2.0, height=2.0)\n",
    "\n",
    "    replayPlot(netVisual_chord_gene(cco_list[[6]], slot.name=\"net\", sources.use=sources, targets.use=targets, signaling=pathway_name_thr, color.use=color_use, title.name=NULL, show.legend=FALSE, small.gap=1, big.gap=5, lab.cex=0.5))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2043893a-a025-4809-8ed5-515302b8b149",
   "metadata": {},
   "source": [
    "# Feature plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "694dcd94-473c-4cd9-96ed-0c31148daeb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/fp_selected_genes.pdf\", width=2*2.5, height=5*2.5)\n",
    "\n",
    "gridExtra::grid.arrange(\n",
    "    \n",
    "    fplot(so, features=\"Ly6c2\", slot=\"data\", restrict=c(\"sample_group\", \"IFNAR_fl_CpG_D1\"), pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    fplot(so, features=\"Ly6c2\", slot=\"data\", restrict=c(\"sample_group\", \"IFNAR_fl_LysM_cre_CpG_D1\"), pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    fplot(so, features=\"Spic\", slot=\"data\", restrict=c(\"sample_group\", \"IFNAR_fl_CpG_D1\"), pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    fplot(so, features=\"Spic\", slot=\"data\", restrict=c(\"sample_group\", \"IFNAR_fl_LysM_cre_CpG_D1\"), pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    fplot(so, features=\"Vcam1\", slot=\"data\", restrict=c(\"sample_group\", \"IFNAR_fl_CpG_D1\"), pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    fplot(so, features=\"Vcam1\", slot=\"data\", restrict=c(\"sample_group\", \"IFNAR_fl_LysM_cre_CpG_D1\"), pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    fplot(so, features=\"Gdf15\", slot=\"data\", restrict=c(\"sample_group\", \"IFNAR_fl_CpG_D1\"), pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    fplot(so, features=\"Gdf15\", slot=\"data\", restrict=c(\"sample_group\", \"IFNAR_fl_LysM_cre_CpG_D1\"), pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    fplot(so, features=\"Ccl2\", slot=\"data\", restrict=c(\"sample_group\", \"IFNAR_fl_CpG_D1\"), pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    fplot(so, features=\"Ccl2\", slot=\"data\", restrict=c(\"sample_group\", \"IFNAR_fl_LysM_cre_CpG_D1\"), pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), ncol=2\n",
    "    \n",
    ")\n",
    "\n",
    "dev.off()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R [conda env:r.4.4.1]",
   "language": "R",
   "name": "conda-env-r.4.4.1-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
