{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "50e4fbf7-815e-4b90-9012-a0afabdd59a5",
   "metadata": {},
   "source": [
    "# Figure panel 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6cfaec12-a6e4-450e-a15e-e4a93255590b",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "800de35d-86e3-4b71-8dd3-911e631758f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "library_load <- suppressMessages(\n",
    "    \n",
    "    suppressWarnings(\n",
    "        \n",
    "        list(\n",
    "        \n",
    "            # Seurat \n",
    "            library(Seurat), \n",
    "\n",
    "            # TradeSeq\n",
    "            library(tradeSeq),\n",
    "\n",
    "            # CCI \n",
    "            library(CellChat), \n",
    "\n",
    "            # Data \n",
    "            library(tidyverse), \n",
    "            library(data.table), \n",
    "\n",
    "            # miloR\n",
    "            library(miloR), \n",
    "            library(ggbeeswarm), \n",
    "\n",
    "            # Plotting \n",
    "            library(ggplot2), \n",
    "            library(patchwork), \n",
    "            library(cowplot), \n",
    "            library(ComplexHeatmap), \n",
    "\n",
    "            # Pyhton compatibility\n",
    "            library(reticulate)\n",
    "\n",
    "        )\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "15b0ec44-5e03-4e48-82cd-1042b31d1a5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Configure reticulate \n",
    "# use_condaenv(condaenv=\"p.3.10.16-FD20200109SPLENO\", conda=\"/nobackup/peer/fdeckert/miniconda3/bin/conda\", required=NULL)\n",
    "# py_config()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "45028df8-0d7a-4a09-865f-5d769c33cf3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_seed <- 42\n",
    "set.seed(random_seed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5ae344ae-fb5c-42da-af0e-e5559e7b3e09",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set working directory to project root\n",
    "setwd(\"/research/peer/fdeckert/FD20200109SPLENO\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2b52cea3-0401-43b9-a99a-a97fc2a66c4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Source\n",
    "source(\"bin/so_pl.R\")\n",
    "source(\"bin/dea_pp.R\")\n",
    "source(\"bin/dea_pl.R\")\n",
    "source(\"bin/gsea_pp.R\")\n",
    "source(\"bin/gsea_pl.R\")\n",
    "source(\"bin/cci_pp.R\")\n",
    "source(\"bin/cci_pl.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2ab8c554-ca21-4e65-8932-2a87d05f1f30",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting Theme\n",
    "source(\"plotting_global.R\")\n",
    "ggplot2::theme_set(theme_global_set(size_select=4)) # From project global source()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7d45dcf-0f7c-422e-815d-c982dd7d9e37",
   "metadata": {},
   "source": [
    "# Import scRNAseq data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "04ba5c21-247b-4af4-ac1b-37c0bddd7b69",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Normalizing layer: counts\n",
      "\n"
     ]
    }
   ],
   "source": [
    "so <- readRDS(\"data/scRNAseq/object/pp_1.rds\")\n",
    "so <- NormalizeData(so)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be7d13cf-4977-45e9-a028-89ea7e5d7350",
   "metadata": {},
   "source": [
    "# Differential expression analysis (DEA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5b711403-c917-4872-9c8d-331b39b4cd30",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Response gene DEA \n",
    "contrasts_vec <- c(\"IFNAR_fl_CpG_D1\", \"IFNAR_fl_Baseline_D0\")\n",
    "dea_res_1 <- readRDS(paste0(\"result/dea/scRNAseq/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))\n",
    "\n",
    "contrasts_vec <- c(\"IFNAR_fl_LysM_cre_CpG_D1\", \"IFNAR_fl_LysM_cre_Baseline_D0\")\n",
    "dea_res_2 <- readRDS(paste0(\"result/dea/scRNAseq/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))\n",
    "\n",
    "contrasts_vec <- c(\"IFNAR_fl_CpG_D3\", \"IFNAR_fl_Baseline_D0\")\n",
    "dea_res_3 <- readRDS(paste0(\"result/dea/scRNAseq/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))\n",
    "\n",
    "contrasts_vec <- c(\"IFNAR_fl_LysM_cre_CpG_D3\", \"IFNAR_fl_LysM_cre_Baseline_D0\")\n",
    "dea_res_4 <- readRDS(paste0(\"result/dea/scRNAseq/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))\n",
    "\n",
    "# Genotype DEA\n",
    "contrasts_vec <- c(\"IFNAR_fl_LysM_cre_Baseline_D0\", \"IFNAR_fl_Baseline_D0\")\n",
    "dea_res_5 <- readRDS(paste0(\"result/dea/scRNAseq/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))\n",
    "\n",
    "contrasts_vec <- c(\"IFNAR_fl_LysM_cre_CpG_D1\", \"IFNAR_fl_CpG_D1\")\n",
    "dea_res_6 <- readRDS(paste0(\"result/dea/scRNAseq/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))\n",
    "\n",
    "contrasts_vec <- c(\"IFNAR_fl_LysM_cre_CpG_D3\", \"IFNAR_fl_CpG_D3\")\n",
    "dea_res_7 <- readRDS(paste0(\"result/dea/scRNAseq/sample_group/\", contrasts_vec[1], \"_vs_\", contrasts_vec[2], \".rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8af20ea3-e859-4451-8822-8ed4bc6a06c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "adj_pval_thr <- 0.05\n",
    "log2FC_thr <- 0.25"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f9c804f-bc7b-49c4-87a1-c6e5fa94e653",
   "metadata": {},
   "source": [
    "# DEG count overview"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d0f4dd5-fd88-44f2-9388-42752bfe03aa",
   "metadata": {},
   "source": [
    "## Response count D1 and D3 vs Ctl "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9e334cd1-b03d-49da-84f9-cd76cdb3033c",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- lapply(dea_res_1, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2FC_thr) %>% dplyr::pull(gene) %>% length}) %>% data.frame() %>% t() \n",
    "mat_2 <- lapply(dea_res_2, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2FC_thr) %>% dplyr::pull(gene) %>% length}) %>% data.frame() %>% t() \n",
    "mat_3 <- lapply(dea_res_3, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2FC_thr) %>% dplyr::pull(gene) %>% length}) %>% data.frame() %>% t() \n",
    "mat_4 <- lapply(dea_res_4, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2FC_thr) %>% dplyr::pull(gene) %>% length}) %>% data.frame() %>% t() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f059b523-aebe-4707-a779-0340e2a6fed1",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat <- list(mat_1, mat_2, mat_3, mat_4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d7788e6c-8792-455f-9cb9-5530da268f45",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat <- lapply(mat, function(x) {\n",
    "    \n",
    "    x <- as.data.frame(x)\n",
    "    set_diff <- setdiff(make.names(levels(so$celltype_low)), rownames(x))\n",
    "    x[set_diff, ] <- 0\n",
    "    x[make.names(levels(so$celltype_low)), , drop=FALSE]\n",
    "\n",
    "}\n",
    "             )\n",
    "\n",
    "mat <- do.call(cbind, mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f2a8afae-68d4-461f-a2ff-ca5464cf8341",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(mat) <- c(\"+/+\", \"cre/+\", \"+/+\", \"cre/+\")\n",
    "col_split <- c(\"D1\", \"D1\", \"D3\", \"D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ea688900-a1d9-4611-a03c-f6cdccb65b5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/hm_deg_count_d1_d3_vs_ctl_cg_and_ko.pdf\", width=1.5, height=3)\n",
    "source(\"bin/dea_pl.R\")\n",
    "draw(dea_count_hm(as.matrix(mat), col_split=col_split, fontsize_select=2))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef419035-3b29-4615-8552-e10d0855634b",
   "metadata": {},
   "source": [
    "## Genotype comparison "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "af472cc2-f9a8-4dae-8260-b655d0ebe33b",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- lapply(dea_res_5, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2FC_thr) %>% dplyr::pull(gene) %>% length}) %>% data.frame() %>% t() \n",
    "mat_2 <- lapply(dea_res_6, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2FC_thr) %>% dplyr::pull(gene) %>% length}) %>% data.frame() %>% t() \n",
    "mat_3 <- lapply(dea_res_7, function(x) {x %>% dplyr::filter(p_val_adj<=adj_pval_thr & abs(avg_log2FC)>=log2FC_thr) %>% dplyr::pull(gene) %>% length}) %>% data.frame() %>% t() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5bb300fd-dcd3-4ce8-80b8-50413c34b18c",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat <- list(mat_1, mat_2, mat_3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "51d9b947-9c97-4038-a5bf-d50085fb9ceb",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat <- lapply(mat, function(x) {\n",
    "    \n",
    "    x <- as.data.frame(x)\n",
    "    set_diff <- setdiff(make.names(levels(so$celltype_low)), rownames(x))\n",
    "    x[set_diff, ] <- 0\n",
    "    x[make.names(levels(so$celltype_low)), , drop=FALSE]\n",
    "\n",
    "}\n",
    "             )\n",
    "\n",
    "mat <- do.call(cbind, mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4b5c9ebc-f2e1-4b3e-8b73-58daf037702e",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(mat) <- c(\"+/+ vs cre/+\", \"+/+ vs cre/+\", \"+/+ vs cre/+\")\n",
    "col_split <- c(\"Ctl\", \"D1\", \"D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7a1db0f8-b61a-4b5f-aa6c-79be5c683476",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/hm_deg_count_cg_vs_ko.pdf\", width=1.5, height=3)\n",
    "\n",
    "draw(dea_count_hm(as.matrix(mat), col_split=col_split, fontsize_select=2))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e0d441e-f38d-43a0-8ff0-99c3b4fc6a0a",
   "metadata": {},
   "source": [
    "# Volcano plots  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b85b9f1b-242e-4315-878c-b8a5a85f9ff6",
   "metadata": {},
   "source": [
    "## CtlKO vs Ctl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "11e942bf-4f62-43a3-b4af-af58db021e57",
   "metadata": {},
   "outputs": [],
   "source": [
    "labels <- list(\n",
    "    \n",
    "    \"intMo\"=c(\"Icam1\", \"Itga4\", \"Itgb1\", \"Vcam1\", \"Gdf15\", \"Ccl2\", \"Tnf\", \"Il6\", \"Tgfb1\", \"Stat1\", \"Stat2\", \"Spic\", \"Irf8\", \"Siglec1\", \"Slc40a1\", \"Hmox1\", \"Blvrb\", \"Treml4\", \"Bach1\", \"Il1a\", \"Il1b\", \"Il1rn\", \"Il18\", \"Il33\")\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e8f42ed7-cdf9-4104-84e3-bf1f27a2c398",
   "metadata": {},
   "outputs": [],
   "source": [
    "vp_5 <- lapply(names(dea_res_5), function(i) {\n",
    "\n",
    "    v_pl_i <- v_pl(dea_res_5[[i]], title=i, log2FC_thr=log2FC_thr, adj_pval_thr=adj_pval_thr, point_size=0.5, color_pos=unlist(color$sample_group[\"IFNAR_fl_LysM_cre_Baseline_D0\"]), color_neg=unlist(color$sample_group[\"IFNAR_fl_Baseline_D0\"]), label=if(i %in% names(labels)) {labels[[i]]} else {label <- NULL}, label_size=2, y_limit=4, aspect_ratio=0.8, top_label=5, label_merge=TRUE) + theme(legend.position=\"none\") + theme_global_set(size_select=4)\n",
    "\n",
    "}\n",
    "                 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "69728dce-5683-4d9d-83b1-ce39b72a321c",
   "metadata": {},
   "outputs": [],
   "source": [
    "vp_5 <- lapply(vp_5, function(p) {egg::set_panel_size(p, width=unit(3, \"cm\"), height=unit(2.5, \"cm\"))})\n",
    "vp_5 <- do.call(gridExtra::arrangeGrob, c(vp_5, ncol=5, nrow=ceiling(length(vp_5)/5)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e873235a-2be6-45db-9c85-846a837129bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/vp_ctlko_vs_ctl.pdf\", width=2*5, height=2*ceiling(length(vp_5)/5))\n",
    "\n",
    "grid::grid.draw(vp_5)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "440d7e4f-abfe-46c8-9cb2-d28cbea7957b",
   "metadata": {},
   "source": [
    "## D1KO vs D1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "ed9fefd9-8d33-4c0a-ae65-0a3ac221deb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "labels <- list(\n",
    "    \n",
    "    \"intMo\"=c(\"Icam1\", \"Itga4\", \"Itgb1\", \"Vcam1\", \"Gdf15\", \"Ccl2\", \"Tnf\", \"Il6\", \"Tgfb1\", \"Stat1\", \"Stat2\", \"Spic\", \"Irf8\", \"Siglec1\", \"Slc40a1\", \"Hmox1\", \"Blvrb\", \"Treml4\", \"Bach1\", \"Il1a\", \"Il1b\", \"Il1rn\", \"Il18\", \"Il33\")\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "96869801-b7fb-423e-8a7a-139c1c68fe53",
   "metadata": {},
   "outputs": [],
   "source": [
    "vp_6 <- lapply(names(dea_res_6), function(i) {\n",
    "\n",
    "    v_pl_i <- v_pl(dea_res_6[[i]], title=i, log2FC_thr=log2FC_thr, adj_pval_thr=adj_pval_thr, point_size=0.5, color_pos=unlist(color$sample_group[\"IFNAR_fl_LysM_cre_CpG_D1\"]), color_neg=unlist(color$sample_group[\"IFNAR_fl_CpG_D1\"]), label=if(i %in% names(labels)) {labels[[i]]} else {label <- NULL}, label_size=2, y_limit=4, aspect_ratio=0.8, top_label=5, label_merge=TRUE) + theme(legend.position=\"none\") + theme_global_set(size_select=4)\n",
    "\n",
    "}\n",
    "                 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "0f2c7a25-61a7-4e38-ae44-617f4ace31d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "vp_6 <- lapply(vp_6, function(p) {egg::set_panel_size(p, width=unit(3, \"cm\"), height=unit(2.5, \"cm\"))})\n",
    "vp_6 <- do.call(gridExtra::arrangeGrob, c(vp_6, ncol=5, nrow=ceiling(length(vp_6)/5)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "b0d922b1-99c3-4a12-a56c-0c644437e176",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/vp_d1ko_vs_d1.pdf\", width=2*5, height=2*ceiling(length(vp_6)/5))\n",
    "\n",
    "grid::grid.draw(vp_6)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ee54a3a-1232-4e26-a180-489f09cfd187",
   "metadata": {},
   "source": [
    "## D3KO vs D3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "5da1dab4-cb86-4195-a8f8-9bab537045ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "labels <- list(\n",
    "    \n",
    "    \"intMo\"=c(\"Icam1\", \"Itga4\", \"Itgb1\", \"Vcam1\", \"Gdf15\", \"Ccl2\", \"Tnf\", \"Il6\", \"Tgfb1\", \"Stat1\", \"Stat2\", \"Spic\", \"Irf8\", \"Siglec1\", \"Slc40a1\", \"Hmox1\", \"Blvrb\", \"Treml4\", \"Bach1\", \"Il1a\", \"Il1b\", \"Il1rn\", \"Il18\", \"Il33\")\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "da1feeed-6cb3-4479-bee6-336cfed1c139",
   "metadata": {},
   "outputs": [],
   "source": [
    "vp_7 <- lapply(names(dea_res_7), function(i) {\n",
    "\n",
    "    v_pl_i <- v_pl(dea_res_7[[i]], title=i, log2FC_thr=log2FC_thr, adj_pval_thr=adj_pval_thr, point_size=0.5, color_pos=unlist(color$sample_group[\"IFNAR_fl_LysM_cre_CpG_D3\"]), color_neg=unlist(color$sample_group[\"IFNAR_fl_CpG_D3\"]), label=if(i %in% names(labels)) {labels[[i]]} else {label <- NULL}, label_size=2, y_limit=4, aspect_ratio=0.8, top_label=5, label_merge=TRUE) + theme(legend.position=\"none\") + theme_global_set(size_select=4)\n",
    "\n",
    "}\n",
    "                 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "4ce97425-0a76-4c2a-9490-3c5451fa0334",
   "metadata": {},
   "outputs": [],
   "source": [
    "vp_7 <- lapply(vp_7, function(p) {egg::set_panel_size(p, width=unit(3, \"cm\"), height=unit(2.5, \"cm\"))})\n",
    "vp_7 <- do.call(gridExtra::arrangeGrob, c(vp_7, ncol=5, nrow=ceiling(length(vp_7)/5)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "7b8ced5d-8c7d-4e11-b999-86ab69a1602e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/vp_d3ko_vs_d3.pdf\", width=2*5, height=2*ceiling(length(vp_7)/5))\n",
    "\n",
    "grid::grid.draw(vp_7)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33376363-6359-4d3c-a5b9-d0b6d204db8f",
   "metadata": {},
   "source": [
    "# Heat map log2FC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "3a8e6b00-1540-4330-9797-d8ab0f872aa9",
   "metadata": {},
   "outputs": [],
   "source": [
    "perturbed_genes_1 <- dea_res_6[[\"intMo\"]] %>% dplyr::filter(avg_log2FC<=-0.25 & p_val_adj<=0.05) %>% pull(gene)\n",
    "perturbed_genes_2 <- dea_res_7[[\"intMo\"]] %>% dplyr::filter(avg_log2FC<=-0.25 & p_val_adj<=0.05) %>% pull(gene)\n",
    "perturbed_genes_3 <- dea_res_6[[\"cMo\"]] %>% dplyr::filter(avg_log2FC<=-0.25 & p_val_adj<=0.05) %>% pull(gene)\n",
    "perturbed_genes_4 <- dea_res_7[[\"cMo\"]] %>% dplyr::filter(avg_log2FC<=-0.25 & p_val_adj<=0.05) %>% pull(gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "cd9305b5-1af5-4729-87a0-a4f5e5c46dfd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "1536"
      ],
      "text/latex": [
       "1536"
      ],
      "text/markdown": [
       "1536"
      ],
      "text/plain": [
       "[1] 1536"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "diff_genes <- c(perturbed_genes_1, perturbed_genes_2, perturbed_genes_3, perturbed_genes_4) %>% unique()\n",
    "length(diff_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "5666b180-e755-429d-bf7b-d3aa2495c329",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_samples <- c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_LysM_cre_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_LysM_cre_CpG_D1\", \"IFNAR_fl_CpG_D3\", \"IFNAR_fl_LysM_cre_CpG_D3\")\n",
    "col_split <- c(\"Ctl\", \"Ctl\", \"D1\", \"D1\", \"D3\", \"D3\")\n",
    "col_label <- c(\"+/+\", \"cre/+\", \"+/+\", \"cre/+\", \"+/+\", \"cre/+\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "97d2f22b-636a-49d0-bfd6-77010ddf53ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_genes <- list(\n",
    "\n",
    "    \"G1\"=c(\"Ccr2\", \"Cd163\", \"Adgre1\", \"Spic\", \"Slfn4\"), \n",
    "    \n",
    "    \"G2\"=c(\"Gdf15\", \"Ccl2\", \"Ccl12\", \"Cxcl10\", \"Il1b\", \"Il18bp\", \"Ccl4\", \"Cxcl2\", \"Tnf\"), # Inflamasome Il1b, Il18\n",
    "    \n",
    "    \"G3\"=c(\"Slc48a1\", \"Ctsb\", \"Ctsd\", \"Slc40a1\", \"Ftl1\", \"Fth1\", \"Hmox1\", \"Blvra\"), # Slc48a1 (Hrg1), Nfe2l2 is constitutively expressed in intMo\n",
    "    \n",
    "    \"G4\"=c(\"Scarf1\",  \"Tmem41b\", \"Vps11\"),\n",
    "    \n",
    "    \"G5\"=c(\"Stat1\", \"Stat2\", \"Irf1\", \"Irf9\", \"Irf7\", \"Ifih1\", \"Isg15\", \"Isg20\", \"Ifi44\", \"Ifit1\", \"Ifit2\", \"Ifit3\", \"Ifit3b\", \"Ifitm1\", \"Ifitm3\", \"Usp18\", \"Oasl1\", \"Oasl2\", \"Ddx60\", \"Cd274\", \"Bst2\"), \n",
    "    \n",
    "    \"G6\"=c(\"H2-Aa\", \"H2-Ab1\", \"H2-Eb1\", \"Cd74\", \"Ciita\"), \n",
    "    \n",
    "    \"G7\"=c(\"Gbp2\", \"Gbp2b\", \"Gbp3\", \"Gbp4\", \"Gbp5\", \"Gbp6\", \"Gbp7\", \"Gbp8\"), \n",
    "    \n",
    "    \"G9\"=c(\"Maf\", \"Mafb\", \"Trib1\", \"Egr1\", \"Fos\", \"Fosb\", \"Jun\"), \n",
    "\n",
    "    \"G10\"=c(\"Lcn2\"), \n",
    "    \n",
    "    \"G11\"=c(\"Bcl2\", \"Bcl2l1\", \"Bcl2l11\"), # Cre LysM Bim fl/fl\n",
    "\n",
    "    \"G12\"=c(\"Atf4\", \"Gdf15\", \"Eif2ak2\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "60e01263-97a1-4cbf-9dd0-b69b841609d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_genes <- list(\"KO UP\"=diff_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "2635f9c9-1f8c-45ae-ac5f-1450ec0f534d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [],
      "text/latex": [],
      "text/markdown": [],
      "text/plain": [
       "named character(0)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "row_split <- setNames(unlist(row_genes, use.names=FALSE), rep(names(row_genes), sapply(row_genes, length)))\n",
    "row_split[!row_split %in% rownames(so)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "62bf275e-cfbf-4d23-baa2-4d3590fccc35",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_dea_fetch <- function(celltype, scale_mode=\"all\") {\n",
    "    \n",
    "    # Get pseudobulk counts \n",
    "    so_i <- so[, (so$celltype_low %in% celltype_select) & (so$sample_group %in% col_samples), drop=FALSE]\n",
    "\n",
    "    # Get pseudobulk counts \n",
    "    mat_1 <- GetAssayData(so_i, assay=\"RNA\", layer=\"counts\")\n",
    "    mat_1 <- split(t(mat_1) %>% as.data.frame(), f=paste(so_i$celltype_low, so_i$sample_group))\n",
    "    mat_1 <- lapply(mat_1, colSums) %>% do.call(rbind, .) %>% t() \n",
    "    mat_1 <- t(t(mat_1) / colSums(mat_1))\n",
    "    \n",
    "    if(scale_mode==\"all\") {\n",
    "\n",
    "        mat_1 <- apply(mat_1, 1, scales::rescale) %>% t() # It apply transforms the data frame\n",
    "        mat_1 <- mat_1[row_split, paste(celltype, col_samples)]\n",
    "\n",
    "        \n",
    "    } else if (scale_mode==\"sample\") {\n",
    "\n",
    "        mat_1 <- mat_1[row_split, paste(celltype, col_samples)]\n",
    "        mat_1 <- apply(mat_1, 1, scales::rescale) %>% t() # It apply transforms the data frame\n",
    "        \n",
    "    }\n",
    "\n",
    "    colnames(mat_1) <- col_label\n",
    "\n",
    "    row_split <- setNames(unlist(row_genes, use.names=FALSE), rep(names(row_genes), sapply(row_genes, length)))\n",
    "    row_split <- row_split[row_split %in% rownames(mat_1)]\n",
    "\n",
    "    # DEA log2FC \n",
    "    mat_2 <- purrr::reduce(\n",
    "        \n",
    "        list(\n",
    "            \n",
    "            dea_res_5[[celltype]] %>% dplyr::mutate(Ctl=avg_log2FC) %>% dplyr::select(gene, Ctl), \n",
    "            dea_res_6[[celltype]] %>% dplyr::mutate(D1=avg_log2FC) %>% dplyr::select(gene, D1), \n",
    "            dea_res_7[[celltype]] %>% dplyr::mutate(D3=avg_log2FC) %>% dplyr::select(gene, D3)\n",
    "        \n",
    "        ), full_join, by=\"gene\"\n",
    "    \n",
    "    ) %>% tibble::column_to_rownames(\"gene\")\n",
    "\n",
    "    mat_2[setdiff(row_split, rownames(mat_2)), ] <- 0\n",
    "    mat_2 <- mat_2[row_split, ]\n",
    "    mat_2[is.na(mat_2)] <- 0\n",
    "    \n",
    "    # DEA adj p-values\n",
    "    mat_3 <- purrr::reduce(\n",
    "        \n",
    "        list(\n",
    "            \n",
    "            dea_res_5[[celltype]] %>% dplyr::mutate(Ctl=p_val_adj) %>% dplyr::select(gene, Ctl), \n",
    "            dea_res_6[[celltype]] %>% dplyr::mutate(D1=p_val_adj) %>% dplyr::select(gene, D1), \n",
    "            dea_res_7[[celltype]] %>% dplyr::mutate(D3=p_val_adj) %>% dplyr::select(gene, D3)\n",
    "        \n",
    "        ), full_join, by=\"gene\"\n",
    "    \n",
    "    ) %>% tibble::column_to_rownames(\"gene\")\n",
    "\n",
    "    mat_3[setdiff(row_split, rownames(mat_3)), ] <- 1\n",
    "    mat_3 <- mat_3[row_split, ]\n",
    "    mat_3[is.na(mat_3)] <- 1\n",
    "    mat_3[(mat_3>adj_pval_thr | abs(mat_2)<log2FC_thr)] <- 1\n",
    "\n",
    "    # Row split\n",
    "    row_split=names(row_split)\n",
    "    row_split=factor(row_split, levels=names(row_genes))\n",
    "    \n",
    "    hm_1 <- dea_exp_hm(mat_1, col_split=col_split, row_split=row_split, scale_width=1, legend_title=\"Exp.\", cluster_rows=TRUE, fontsize_select=2)\n",
    "    hm_2 <- dea_res_hm(mat_2, mat_3, p_val_adj_thr=adj_pval_thr, row_split=row_split, scale_width=1, color_min=\"#cd34b5\", color_max=\"#FFAC1E\", cluster_rows=TRUE, fontsize_select=2)\n",
    "\n",
    "    hm <- hm_2 + hm_1\n",
    "\n",
    "    return(hm)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "f82ae7b3-d093-4eba-805d-823264e47628",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype_select <- c(\"cMo\", \"intMo\", \"RPM\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "6d56e2df-3b1d-47d9-8b7a-685fb82c28b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "hm_1 <- hm_dea_fetch(celltype=\"cMo\")\n",
    "hm_2 <- hm_dea_fetch(celltype=\"intMo\")\n",
    "hm_3 <- hm_dea_fetch(celltype=\"RPM\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "b3bcb71d-6d89-41d9-87ae-15c7fb9b510f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>pdf:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{pdf:} 2"
      ],
      "text/markdown": [
       "**pdf:** 2"
      ],
      "text/plain": [
       "pdf \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/hm_dea_ctl_d1_d3_genotype_ko_down.pdf\", width=5, height=200)\n",
    "\n",
    "hm_1 + hm_2 + hm_3\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3997e093-0946-4fd9-87d3-23c2aacc8bd0",
   "metadata": {},
   "source": [
    "# Gene set enrichment analysis (GSEA)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a56decd3-ed0f-461c-bcbb-25b73b15e214",
   "metadata": {},
   "source": [
    "## Genotype comparison - M7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 316,
   "id": "ab320482-f659-41b4-bf29-a7a83ab484e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "compute_gsea <- FALSE\n",
    "celltype_select <- c(\"MEP\", \"Proerythroblast\", \"Erythroblast\", \"cMo\", \"intMo\", \"ncMo\", \"RPM\", \"cDC1\", \"cDC2\", \"moDC\", \"cDC.mig.\", \"Basophil\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1359afa3-7e3c-478a-a60e-0be197bbb19a",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(compute_gsea) {\n",
    "\n",
    "    # Collapse gene set\n",
    "    gene_set <- msigdbr::msigdbr(species=\"mouse\", db_species=\"MM\", category=\"M7\", subcategory=NULL)\n",
    "    gene_set <- gene_set[grep(\"_UP\", gene_set$gs_name), ]\n",
    "    gene_set$gs_name <- sapply(strsplit(gene_set$gs_name, \"_\"), function(x) {paste(x[max(1, length(x)-3):(length(x)-2)], collapse=\"_\")})\n",
    "    gene_set$gs_name <- gsub(\"CELL_|CD4_|CD8_|MONOCYTE_|MACROPHAGE_|TREG_|LANGERHANS_|NEUTROPHIL_|CDC1_|CDC2_|MIGDC_|ETAC_|GD_|ILC_|PDC_\", \"\", gene_set$gs_name)\n",
    "    gene_set <- gene_set %>% dplyr::select(gene_symbol, gs_name) %>% dplyr::distinct()\n",
    "\n",
    "    gsea_res <- lapply(celltype_select, function(i) {\n",
    "    \n",
    "        gsea_res_1 <- gsea(dea_res_5[[i]], gene_set=gene_set) %>% dplyr::mutate(col_split=\"Ctl\", col_label=i)\n",
    "        gsea_res_2 <- gsea(dea_res_6[[i]], gene_set=gene_set) %>% dplyr::mutate(col_split=\"D1\", col_label=i)\n",
    "        gsea_res_3 <- gsea(dea_res_7[[i]], gene_set=gene_set) %>% dplyr::mutate(col_split=\"D3\", col_label=i)\n",
    "    \n",
    "        gsea_res <- rbind(gsea_res_1, gsea_res_2, gsea_res_3)\n",
    "    \n",
    "        return(gsea_res)\n",
    "        \n",
    "    }\n",
    "          )\n",
    "\n",
    "    saveRDS(gsea_res, \"result/gsea/scRNAseq/gsea_ctl_d1_d3_wt_vs_ko_m7.rds\")\n",
    "    \n",
    "} else {\n",
    "\n",
    "    gsea_res <- readRDS(\"result/gsea/scRNAseq/gsea_ctl_d1_d3_wt_vs_ko_m7.rds\")\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42d9360e-dc81-4ab3-8d8c-1a919d1664ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res <- do.call(rbind, gsea_res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "685414aa-068b-4b11-8f97-eaf9867f8b7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs_name_select <- gsea_res %>% dplyr::filter(padj<=0.05 & abs(NES)>=1) %>% dplyr::pull(pathway)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1dd31e51-eee3-462b-9486-4bdb61a8954e",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- gsea_res %>% dplyr::filter(pathway %in% gs_name_select) %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=padj,\n",
    "        values_fill=1\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6006c1c-5934-4bba-a3ab-254f2d5edce0",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_2 <- gsea_res %>% dplyr::filter(pathway %in% gs_name_select) %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=NES,\n",
    "        values_fill=0\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08115c33-eb78-4629-8e66-9abe5f18eb8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_split <- read.csv(\"data/reference/immune_dictionary/immune_dictionary.csv\") %>% tibble::column_to_rownames(\"gs_name\")\n",
    "col_label <- mat_1$col_label\n",
    "col_split <- mat_1$col_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3df90dc3-3d27-4b5c-a683-de6ad1aefe01",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- mat_1 %>% dplyr::select(-col_label, -col_split) %>% t()\n",
    "mat_2 <- mat_2 %>% dplyr::select(-col_label, -col_split) %>% t()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5971db3-87ca-4018-920f-397ea543608c",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_split <- row_split[rownames(mat_1), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3f20be2-86af-46b0-86da-7db847e27bc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "rownames(mat_1) <- row_split$norm\n",
    "rownames(mat_2) <- row_split$norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6f4092b-a24f-475e-819a-73c1583b7884",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "row_split <- row_split %>% dplyr::select(norm, family) %>% dplyr::rename(row_split_label=family) %>% dplyr::distinct()\n",
    "rownames(row_split) <- row_split$norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11f01b9c-6b82-45cb-a6ff-5ad736dd936e",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_label[col_label==\"cDC.mig.\"] <- \"cDC mig.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e581fe24-9fba-48ee-8747-206d8a1dda6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/hm_gsea_ctl_d1_d3_wt_vs_ko_m7.pdf\", width=8, height=12)\n",
    "\n",
    "gsea_hm(mat_1, mat_2, col_label=col_label, col_split=col_split, row_split=row_split, fontsize_select=2)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc03ee44-7fc1-4793-8bf4-b1f5a01eaf54",
   "metadata": {},
   "source": [
    "## Genotype comparison - HM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c9289eb-8412-4d07-b34e-8ea15840dce8",
   "metadata": {},
   "outputs": [],
   "source": [
    "compute_gsea <- FALSE\n",
    "celltype_select <- c(\"MEP\", \"Proerythroblast\", \"Erythroblast\", \"cMo\", \"intMo\", \"ncMo\", \"RPM\", \"cDC1\", \"cDC2\", \"moDC\", \"cDC.mig.\", \"Basophil\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "885824cd-7f1d-4363-975d-14d058961671",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(compute_gsea) {\n",
    "\n",
    "    gsea_res <- lapply(celltype_select, function(i) {\n",
    "    \n",
    "        gsea_res_1 <- gsea(dea_res_5[[i]], category=\"MH\", subcategory=NULL, gene_set=NULL) %>% dplyr::mutate(col_split=\"Ctl\", col_label=i)\n",
    "        gsea_res_2 <- gsea(dea_res_6[[i]], category=\"MH\", subcategory=NULL, gene_set=NULL) %>% dplyr::mutate(col_split=\"D1\", col_label=i)\n",
    "        gsea_res_3 <- gsea(dea_res_7[[i]], category=\"MH\", subcategory=NULL, gene_set=NULL) %>% dplyr::mutate(col_split=\"D3\", col_label=i)\n",
    "    \n",
    "        gsea_res <- rbind(gsea_res_1, gsea_res_2, gsea_res_3)\n",
    "    \n",
    "        return(gsea_res)\n",
    "        \n",
    "    }\n",
    "          )\n",
    "\n",
    "    saveRDS(gsea_res, \"result/gsea/scRNAseq/gsea_ctl_d1_d3_wt_vs_ko_mh.rds\")\n",
    "    \n",
    "} else {\n",
    "\n",
    "    gsea_res <- readRDS(\"result/gsea/scRNAseq/gsea_ctl_d1_d3_wt_vs_ko_mh.rds\")\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba957b2c-78b2-4ba5-8a1a-e5ea52c90a21",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res <- do.call(rbind, gsea_res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cf9a1f3-2f0d-4397-82a8-7dc37371e87f",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res <- gsea_res[gsea_res$col_label %in% c(\"MEP\", \"Proerythroblast\", \"Erythroblast\", \"cMo\", \"intMo\", \"RPM\"), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63b1abc0-9358-418e-812d-de21ea45cb7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs_name_select <- gsea_res %>% dplyr::filter(padj<=0.05 & abs(NES)>=1) %>% dplyr::pull(pathway)\n",
    "gs_name_select <- gs_name_select[grep(\"E2F|MYC|HYPOXIA|STAT3|INTERFERON|TNF|HEME\", gs_name_select)]\n",
    "gsea_res <- gsea_res %>% dplyr::filter(pathway %in% gs_name_select)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ea92895-6429-4435-aba3-6cc87736381a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs_name_order <- c(\"HALLMARK_INTERFERON_ALPHA_RESPONSE\", \"HALLMARK_INTERFERON_GAMMA_RESPONSE\", \"HALLMARK_IL6_JAK_STAT3_SIGNALING\", \"HALLMARK_TNFA_SIGNALING_VIA_NFKB\", \"HALLMARK_HYPOXIA\", \"HALLMARK_HEME_METABOLISM\", \"HALLMARK_MYC_TARGETS_V1\", \"HALLMARK_MYC_TARGETS_V2\", \"HALLMARK_E2F_TARGETS\")\n",
    "gsea_res$gs_name <- factor(gsea_res$gs_name, levels=gs_name_order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6139a685-8339-42e2-ae7d-7cbc061f55a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- gsea_res %>% dplyr::filter(pathway %in% gs_name_select) %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=padj,\n",
    "        values_fill=1\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1a51c94-5778-4d10-97ae-767a69147b47",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_2 <- gsea_res %>% dplyr::filter(pathway %in% gs_name_select) %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=NES,\n",
    "        values_fill=0\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55514ba2-b2b2-449c-8f67-509d0da3df7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_label <- mat_1$col_label\n",
    "col_split <- mat_1$col_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3f5482c-2d04-419f-afac-fa1776a3f428",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- mat_1 %>% dplyr::select(-col_label, -col_split) %>% t()\n",
    "mat_2 <- mat_2 %>% dplyr::select(-col_label, -col_split) %>% t()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ad8479d-4a47-4e4f-9972-8061ac80c91e",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- mat_1[gs_name_order, ]\n",
    "mat_2 <- mat_2[gs_name_order, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57c3d582-4c80-42a8-a057-df8780f9a190",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_label[col_label==\"cDC.mig.\"] <- \"cDC mig.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abdc1726-c16c-4468-96d5-003e05e0aeb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/hm_gsea_ctl_d1_d3_wt_vs_ko_mh.pdf\", width=8, height=10)\n",
    "\n",
    "source(\"bin/gsea_pl.R\")\n",
    "gsea_hm(mat_1, mat_2, col_label=col_label, col_split=col_split, row_split=NULL, pathway_suffix=\"HALLMARK\", color_neg=color$sample_group[\"IFNAR_fl_CpG_D1\"], color_pos=color$sample_group[\"IFNAR_fl_LysM_cre_CpG_D1\"], fontsize_select=2)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a96e2ea-b027-410b-a7cf-e59959ea9e1c",
   "metadata": {},
   "source": [
    "## Genotype comparison - GO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dd64847-8400-48f0-af90-99d8f0e02ef7",
   "metadata": {},
   "outputs": [],
   "source": [
    "compute_gsea <- FALSE\n",
    "celltype_select <- c(\"MEP\", \"Proerythroblast\", \"Erythroblast\", \"cMo\", \"intMo\", \"ncMo\", \"RPM\", \"cDC1\", \"cDC2\", \"moDC\", \"cDC.mig.\", \"Basophil\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e7734f6-65c0-4467-8e25-b7b7d4e8a7c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(compute_gsea) {\n",
    "\n",
    "    gsea_res <- lapply(celltype_select, function(i) {\n",
    "    \n",
    "        gsea_res_1 <- gsea_cp(dea_res_5[[i]]) %>% dplyr::mutate(col_split=\"Ctl\", col_label=i)\n",
    "        gsea_res_2 <- gsea_cp(dea_res_6[[i]]) %>% dplyr::mutate(col_split=\"D1\", col_label=i)\n",
    "        gsea_res_3 <- gsea_cp(dea_res_7[[i]]) %>% dplyr::mutate(col_split=\"D3\", col_label=i)\n",
    "    \n",
    "        gsea_res <- rbind(gsea_res_1, gsea_res_2, gsea_res_3)\n",
    "    \n",
    "        return(gsea_res)\n",
    "        \n",
    "    }\n",
    "          )\n",
    "\n",
    "    saveRDS(gsea_res, \"result/gsea/scRNAseq/gsea_ctl_d1_d3_wt_vs_ko_gobp.rds\")\n",
    "    \n",
    "} else {\n",
    "\n",
    "    gsea_res <- readRDS(\"result/gsea/scRNAseq/gsea_ctl_d1_d3_wt_vs_ko_gobp.rds\")\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7b5f832-67cc-443c-8b9d-cabf9347966c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res <- do.call(rbind, gsea_res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1777f5d7-18bb-42cf-a416-b6bf86a6343a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs_name_select <- gsea_res %>% dplyr::filter(p.adjust<=0.05 & abs(NES)>=1) %>% dplyr::pull(ID)\n",
    "gsea_res <- gsea_res %>% dplyr::filter(ID %in% gs_name_select)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39b94e20-83be-416e-92e6-cf3543f724f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res <- gsea_res[!gsea_res$Description %in% grep(\"T cell|B cell|CD4|CD8|lymphocyte|endothelial|epithelial\", gsea_res$Description, value=TRUE), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af296a68-40e3-41ae-8be2-1dbf6f56d93e",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(compute_gsea) {\n",
    "\n",
    "    gs_reduce <- gsea_cp_reduce(gsea_res)\n",
    "\n",
    "    saveRDS(gs_reduce, \"result/gsea/scRNAseq/gsea_reduce_ctl_d1_d3_wt_vs_ko_gobp.rds\")\n",
    "    \n",
    "} else {\n",
    "\n",
    "    gs_reduce <- readRDS(\"result/gsea/scRNAseq/gsea_reduce_ctl_d1_d3_wt_vs_ko_gobp.rds\")\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adddea3f-2873-4b63-a260-9d0c923dce87",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_res <- gsea_res %>% dplyr::rename(pathway=Description, padj=p.adjust) %>% dplyr::filter(ID %in% gs_reduce$go)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9341e76-d801-4f73-83e2-1e48074ad4b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "gs_name_select <- gsea_res %>% dplyr::filter(padj<=0.05 & abs(NES)>=2) %>% dplyr::pull(ID)\n",
    "gsea_res <- gsea_res %>% dplyr::filter(ID %in% gs_name_select)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d12e8ed-a07f-4588-acc7-9cca036764e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- gsea_res %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=padj,\n",
    "        values_fill=1\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd5f1776-fdd8-428b-aed5-fe46b1419df6",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_2 <- gsea_res %>% \n",
    "\n",
    "    tidyr::pivot_wider(\n",
    "        \n",
    "        id_cols=c(col_label, col_split),\n",
    "        names_from=pathway,\n",
    "        values_from=NES,\n",
    "        values_fill=0\n",
    "    \n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47e01794-d392-4f49-88b3-520251c6b6cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_label <- mat_1$col_label\n",
    "col_split <- mat_1$col_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a369b354-e980-4542-ab09-6becb9c9d7d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat_1 <- mat_1 %>% dplyr::select(-col_label, -col_split) %>% t()\n",
    "mat_2 <- mat_2 %>% dplyr::select(-col_label, -col_split) %>% t()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cb59cfb-091f-490f-bd12-bc886ee98807",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_split <- gs_reduce %>% dplyr::select(term, parentTerm) %>% dplyr::rename(row_split_label=parentTerm) %>% dplyr::distinct()\n",
    "rownames(row_split) <- gs_reduce$term"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f4cbe1d-cbc0-48fb-b76d-45650164ee86",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_label[col_label==\"cDC.mig.\"] <- \"cDC mig.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c2c3e03-1886-4da8-8fe0-fdc9ca96bfa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/hm_gsea_ctl_d1_d3_wt_vs_ko_gobp.pdf\", width=10, height=50)\n",
    "\n",
    "gsea_hm(mat_1, mat_2, col_label=col_label, col_split=col_split, row_split=row_split, pathway_suffix=\"WP\", fontsize_select=2)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc5432fa-6bdf-40c1-a391-3185dd3bed3f",
   "metadata": {},
   "source": [
    "## Violine gene expression plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4168949f-ccb8-43c3-a8dd-da784c206c3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_select <- c(\"Ly6c2\", \"Spic\", \"Cd163\", \"Gdf15\", \"Ccl2\", \"Slfn4\", \"Bcl2l1\", \"Hmox1\", \"Jun\")\n",
    "celltype_select <- c(\"cMo\", \"intMo\", \"RPM\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb9f59b6-a237-4468-892b-2d1d19272111",
   "metadata": {},
   "outputs": [],
   "source": [
    "mat <- cbind(so@meta.data[, c(\"sample_group\", \"celltype_low\"), drop=FALSE], t(GetAssayData(so, assay=\"RNA\", slot=\"data\")[gene_select, ])) %>% \n",
    "    pivot_longer(cols=-c(sample_group, celltype_low), names_to=\"gene\", values_to=\"expression\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06a030b5-a3e0-4e6b-a25a-1367fe96696e",
   "metadata": {},
   "outputs": [],
   "source": [
    "map <- data.frame(\n",
    "    \n",
    "    sample_group=c(\"IFNAR_fl_Baseline_D0\", \"IFNAR_fl_LysM_cre_Baseline_D0\", \"IFNAR_fl_CpG_D1\", \"IFNAR_fl_LysM_cre_CpG_D1\", \"IFNAR_fl_CpG_D3\", \"IFNAR_fl_LysM_cre_CpG_D3\"),\n",
    "    time_point=c(\"Ctl\", \"Ctl\", \"D1\", \"D1\", \"D3\", \"D3\"),\n",
    "    genotype=c(\"+/+\", \"cre/+\", \"+/+\", \"cre/+\", \"+/+\", \"cre/+\")\n",
    ")\n",
    "\n",
    "mat <- mat %>% dplyr::left_join(., map, by=join_by(sample_group)) %>% \n",
    "    dplyr::mutate(gene=factor(gene, levels=gene_select))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7ee6260-001a-485d-84db-7619e4a3d9cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "genotype <- c(\"#732E8F\", \"#D390BF\", \"#FA8775\")\n",
    "names(genotype) <- c(\"WT\", \"+/+\", \"cre/+\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d442739b-0a5a-4ce6-a930-1ba8def20c61",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=5, repr.plot.height=7.5)\n",
    "\n",
    "exp_p <- ggplot(mat %>% dplyr::filter(celltype_low %in% celltype_select), aes(x=time_point, y=expression, color=sample_group, fill=sample_group)) + \n",
    "\n",
    "        geom_jitter(position=position_jitterdodge(jitter.width=0.20, dodge.width=1.0), shape=21, stroke=0.1, size=0.75, show.legend=FALSE, color=\"black\") + \n",
    "        # geom_boxplot(position=position_dodge(width=1.0), width=0.20, color=\"black\", alpha=1, outlier.color=NA) + \n",
    "        \n",
    "        xlab(\"\") + ylab(\"log(CPTT)\") + \n",
    "        ggtitle(\"Spic\") + \n",
    "        facet_grid(gene~celltype_low, scales=\"free\") + \n",
    "        scale_color_manual(values=color$sample_group) + \n",
    "        scale_fill_manual(values=color$sample_group)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21cc23ae-0b66-47e4-86ec-efc515519adf",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/exp_myeloid.pdf\", width=3, height=8)\n",
    "\n",
    "set_height <- 1 * length(gene_select)\n",
    "set_width <- 3 * length(celltype_select)\n",
    "\n",
    "exp_p + ggh4x::force_panelsizes(cols=unit(set_width, \"mm\"), total_height=unit(set_height, \"cm\"))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff664abc-1793-481d-8b25-cefbb946187e",
   "metadata": {},
   "source": [
    "# CCI DEA "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec59b755-f507-4b45-aa40-e32141627ea4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load source \n",
    "source(\"bin/cci_pp.R\")\n",
    "source(\"bin/cci_pl.R\")\n",
    "\n",
    "# Parameter \n",
    "pval_prob_thr <- 0.05\n",
    "p_val_adj_dea <- 0.05\n",
    "pathway_name_thr <- c(\"ICAM\", \"VCAM\", \"SIRP\", \"KIT\", \"TGFb\", \"TNF\", \"CXCL\", \"CCL\")\n",
    "\n",
    "# CellChat results\n",
    "lr_res <- read.csv(\"result/cci/lr_res.csv\", row.names=1)\n",
    "\n",
    "# Add reversed Cell-cell interactions \n",
    "db <- CellChat::CellChatDB.mouse\n",
    "\n",
    "# Impute custom interactions\n",
    "options(stringsAsFactors=FALSE)\n",
    "interaction <- read.csv(\"data/reference/cellchat/interaction.csv\", row.names=1, na.strings=c(\"\"))\n",
    "interaction[is.na(interaction)] <- \"\"\n",
    "db$interaction <- rbind(db$interaction, interaction)\n",
    "\n",
    "# Get cell-cell contact interaction \n",
    "db <- CellChat::subsetDB(db, search=c(\"Cell-Cell Contact\"), key=\"annotation\")\n",
    "\n",
    "# Subset lr results for cell-cell contact forward \n",
    "lr_res_ccc_fwd <- lr_res %>% dplyr::filter(lr_res$interaction_name %in% db$interaction$interaction_name)\n",
    "\n",
    "# Reverse cell-cell contact \n",
    "lr_res_ccc_rev <- data.frame(\n",
    "\n",
    "    source=lr_res_ccc_fwd$target, \n",
    "    target=lr_res_ccc_fwd$source, \n",
    "    interaction_name=paste0(lr_res_ccc_fwd$receptor_symbol, \"_\", lr_res_ccc_fwd$ligand_symbol), \n",
    "    prob=lr_res_ccc_fwd$prob, \n",
    "    pval=lr_res_ccc_fwd$pval, \n",
    "    pathway_name=lr_res_ccc_fwd$pathway_name, \n",
    "    ligand_symbol=lr_res_ccc_fwd$receptor_symbol, \n",
    "    receptor_symbol=lr_res_ccc_fwd$ligand_symbol, \n",
    "    sample_group=lr_res_ccc_fwd$sample_group\n",
    "    \n",
    ")\n",
    "\n",
    "# Combine results \n",
    "lr_res_ccc <- rbind(lr_res_ccc_fwd, lr_res_ccc_rev)\n",
    "\n",
    "# Assemble LR results \n",
    "lr_res <- rbind(lr_res[!lr_res$interaction_name %in% lr_res_ccc$interaction_name, ], lr_res_ccc)\n",
    "\n",
    "# Subset by interaction pathway\n",
    "lr_res <- lr_res[lr_res$pathway_name %in% pathway_name_thr, ]\n",
    "lr_res$pathway_name <- factor(lr_res$pathway_name, levels=pathway_name_thr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67410385-21c6-4e7a-8671-1fdbdadf36b5",
   "metadata": {},
   "source": [
    "# DEA results "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d60c5142-d878-4dfe-957d-426e2a819a7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "dea_mat_5 <- dea_res_5[c(\"cMo\", \"intMo\", \"RPM\", \"cDC2\", \"Basophil\", \"MEP\", \"Proerythroblast\", \"Erythroblast\")]\n",
    "dea_mat_5 <- lapply(dea_mat_5, function(x) {list(\"Ctl\"=x)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1347d3d0-1bde-44f0-882f-3102ba6ba5ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "dea_mat_6<- dea_res_6[c(\"cMo\", \"intMo\", \"RPM\", \"cDC2\", \"Basophil\", \"MEP\", \"Proerythroblast\", \"Erythroblast\")]\n",
    "dea_mat_6 <- lapply(dea_mat_6, function(x) {list(\"D1\"=x)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f3261de-4089-4970-807f-50ace3e7ae72",
   "metadata": {},
   "outputs": [],
   "source": [
    "dea_mat_7 <- dea_res_7[c(\"cMo\", \"intMo\", \"RPM\", \"cDC2\", \"Basophil\", \"MEP\", \"Proerythroblast\", \"Erythroblast\")]\n",
    "dea_mat_7 <- lapply(dea_mat_7, function(x) {list(\"D3\"=x)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fa13ead-4344-4d8d-91bf-9fe99ff9bcd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "dea_mat <- Map(c, dea_mat_5, dea_mat_6, dea_mat_7)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c48a72b-db65-4fd1-a47a-81423d4d9799",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## PTPG B (+/+) vs D1 (+/+) and D3 (+/+)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f08fc2ce-da91-4161-80a9-ed3281425407",
   "metadata": {},
   "source": [
    "## cMo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "003fb3e9-c960-4a2d-8241-b743963df5d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"cMo\")\n",
    "target <- c(\"MEP\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "462e97b1-b38a-4f27-a5bd-7716913ac0a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80d21cd2-cd95-4e0b-95b0-c629a4b37417",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/cci_cmo_mep_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18e4ce5b-ea26-4e93-936a-0048d8f7e2d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"cMo\")\n",
    "target <- c(\"Proerythroblast\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8526ccc3-5d19-42bc-8f8a-9febc3749351",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d1a01fd-b574-4c76-8901-092caa18b404",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/cci_cmo_proeb_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d290092c-21c1-431a-ab30-1cf2562eff89",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"cMo\")\n",
    "target <- c(\"Erythroblast\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3fcc361-927a-4b1b-a265-80fc3cc8d5d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e2f70a2-0f05-46ea-a89b-f1974e85494e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/cci_cmo_eb_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f691ebc-9fc2-4997-bb21-bac8cbb90db6",
   "metadata": {},
   "source": [
    "## intMo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b65a947-44dc-4f2a-a16c-105d000e247a",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"intMo\")\n",
    "target <- c(\"MEP\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e643f889-0815-49f9-9034-eb6ba03f5bef",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aac4fd3a-632f-4bd9-8eef-2cade700a7d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/cci_intmo_mep_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f12cfed-1372-4a14-b341-6e776b765d6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"intMo\")\n",
    "target <- c(\"Proerythroblast\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16f44f38-859b-4337-a545-5c2e05da9178",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71a43bc6-400e-47b1-b23a-66a4e944d4ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/cci_intmo_proeb_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69480d57-2a72-41d9-8385-52117e0012a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"intMo\")\n",
    "target <- c(\"Erythroblast\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16af9b97-b869-4128-a6bb-194a597f5d5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba316e76-2191-48ab-a35a-1fe2a7d33a71",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/cci_intmo_eb_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "074532ad-a423-4100-b9b0-1a3d1904063b",
   "metadata": {},
   "source": [
    "## RPM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c89da626-687b-49c2-94b0-1cddcd97f8d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"RPM\")\n",
    "target <- c(\"MEP\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bba0449-c718-42f4-a8d4-eb16a204f4f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc02eac0-37bc-486e-9f4a-422f17f6604e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/cci_rpm_mep_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c90c12bd-e06c-4fa7-b3c8-e8842916db4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"RPM\")\n",
    "target <- c(\"Proerythroblast\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3225ebbc-095d-4d91-ae0d-e5d6b6d3f6a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a6c0dec-fcd0-4ef0-b368-a2892ae33c01",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/cci_rpm_proeb_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58ba8d68-224a-4ce3-9412-4623d6fbf1f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "source <- c(\"RPM\")\n",
    "target <- c(\"Erythroblast\")\n",
    "sample_group <- c(\"Ctr\"=\"IFNAR_fl_Baseline_D0\", \"D1\"=\"IFNAR_fl_CpG_D1\", \"D3\"=\"IFNAR_fl_CpG_D3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72af504a-8e9e-4139-94af-1e93e722fc0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "lr_res_i <- lr_res_parse(lr_res, source, target, sample_group, pval_prob_thr)\n",
    "lr_res_i <- lr_res_stretch(lr_res_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39388a7c-74b7-42d3-a138-248cd8600c7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/cci_rpm_eb_ko_vs_cg.pdf\", width=5, height=5)\n",
    "\n",
    "dea_lr_hm(lr_res_i, dea_mat, source, target, fontsize_select=2, lr_de_check=TRUE)\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62c60458-8f68-4833-8c7e-967ef4efe998",
   "metadata": {},
   "source": [
    "# Interaction plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca3919c0-95db-40cf-99bb-7fd54e463343",
   "metadata": {},
   "outputs": [],
   "source": [
    "sources <- c(\"RPM\", \"intMo\", \"cMo\")\n",
    "targets <- c(\"MEP\", \"Proerythroblast\", \"Erythroblast\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0037cc6c-331c-4ab5-a9d9-16b1cc3c9416",
   "metadata": {},
   "outputs": [],
   "source": [
    "cco_1 <- readRDS(\"data/scRNAseq/cci/cco_ifnar_fl_baseline_d0.rds\")\n",
    "cco_2 <- readRDS(\"data/scRNAseq/cci/cco_ifnar_fl_cpg_d1.rds\")\n",
    "cco_3 <- readRDS(\"data/scRNAseq/cci/cco_ifnar_fl_cpg_d3.rds\")\n",
    "cco_4 <- readRDS(\"data/scRNAseq/cci/cco_ifnar_fl_lysm_cre_baseline_d0.rds\")\n",
    "cco_5 <- readRDS(\"data/scRNAseq/cci/cco_ifnar_fl_lysm_cre_cpg_d1.rds\")\n",
    "cco_6 <- readRDS(\"data/scRNAseq/cci/cco_ifnar_fl_lysm_cre_cpg_d3.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3011b27-7789-4eca-8e3d-ab0ad506eafb",
   "metadata": {},
   "outputs": [],
   "source": [
    "cco_1 <- subsetCellChat(object=cco_1, idents.use=c(sources, targets))\n",
    "cco_2 <- subsetCellChat(object=cco_2, idents.use=c(sources, targets))\n",
    "cco_3 <- subsetCellChat(object=cco_3, idents.use=c(sources, targets))\n",
    "cco_4 <- subsetCellChat(object=cco_4, idents.use=c(sources, targets))\n",
    "cco_5 <- subsetCellChat(object=cco_5, idents.use=c(sources, targets))\n",
    "cco_6 <- subsetCellChat(object=cco_6, idents.use=c(sources, targets))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f3e2c0c-fef7-4539-880e-c30cca6baea8",
   "metadata": {},
   "outputs": [],
   "source": [
    "levels(cco_1@idents) <- c(targets, sources)\n",
    "levels(cco_2@idents) <- c(targets, sources)\n",
    "levels(cco_3@idents) <- c(targets, sources)\n",
    "levels(cco_4@idents) <- c(targets, sources)\n",
    "levels(cco_5@idents) <- c(targets, sources)\n",
    "levels(cco_6@idents) <- c(targets, sources)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b7bf337-daf2-4d4f-a30e-8507c6974e97",
   "metadata": {},
   "outputs": [],
   "source": [
    "color_use <- color$celltype_low[names(color$celltype_low) %in% c(sources, targets)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10ec20f9-774b-462e-ac37-991ccb3d2020",
   "metadata": {},
   "outputs": [],
   "source": [
    "cco_list <- list(cco_1, cco_2, cco_3, cco_4, cco_5, cco_6)\n",
    "\n",
    "q_prob_thr <- quantile(c(cco_1@net$prob, cco_2@net$prob, cco_3@net$prob), 0.90)\n",
    "for (i in seq_along(cco_list)) {\n",
    "    \n",
    "    prob_arr <- cco_list[[i]]@net$prob\n",
    "    prob_arr[prob_arr > q_prob_thr] <- q_prob_thr\n",
    "    cco_list[[i]]@net$prob <- prob_arr\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb06d56a-6d68-4be4-9277-a21e7987ef04",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/cp_cci_net_ctl_ko.pdf\", width=2.0, height=2.0)\n",
    "\n",
    "    replayPlot(netVisual_chord_gene(cco_list[[4]], slot.name=\"net\", sources.use=sources, targets.use=targets, signaling=pathway_name_thr, color.use=color_use, title.name=NULL, show.legend=FALSE, small.gap=1, big.gap=5, lab.cex=0.5))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4deee3b9-f1f4-4764-b9d8-ea6b10dd66d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/cp_cci_net_d1_ko.pdf\", width=2.0, height=2.0)\n",
    "\n",
    "    replayPlot(netVisual_chord_gene(cco_list[[5]], slot.name=\"net\", sources.use=sources, targets.use=targets, signaling=pathway_name_thr, color.use=color_use, title.name=NULL, show.legend=FALSE, small.gap=1, big.gap=5, lab.cex=0.5))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11ad2ca6-235a-462a-b090-a0804e3e3092",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/cp_cci_net_d3_ko.pdf\", width=2.0, height=2.0)\n",
    "\n",
    "    replayPlot(netVisual_chord_gene(cco_list[[6]], slot.name=\"net\", sources.use=sources, targets.use=targets, signaling=pathway_name_thr, color.use=color_use, title.name=NULL, show.legend=FALSE, small.gap=1, big.gap=5, lab.cex=0.5))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2043893a-a025-4809-8ed5-515302b8b149",
   "metadata": {},
   "source": [
    "# Feature plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "694dcd94-473c-4cd9-96ed-0c31148daeb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"result/figures/1_scRNAseq/panel_5/fp_selected_genes.pdf\", width=2*2.5, height=5*2.5)\n",
    "\n",
    "gridExtra::grid.arrange(\n",
    "    \n",
    "    fplot(so, features=\"Ly6c2\", slot=\"data\", restrict=c(\"sample_group\", \"IFNAR_fl_CpG_D1\"), pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    fplot(so, features=\"Ly6c2\", slot=\"data\", restrict=c(\"sample_group\", \"IFNAR_fl_LysM_cre_CpG_D1\"), pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    fplot(so, features=\"Spic\", slot=\"data\", restrict=c(\"sample_group\", \"IFNAR_fl_CpG_D1\"), pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    fplot(so, features=\"Spic\", slot=\"data\", restrict=c(\"sample_group\", \"IFNAR_fl_LysM_cre_CpG_D1\"), pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    fplot(so, features=\"Vcam1\", slot=\"data\", restrict=c(\"sample_group\", \"IFNAR_fl_CpG_D1\"), pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    fplot(so, features=\"Vcam1\", slot=\"data\", restrict=c(\"sample_group\", \"IFNAR_fl_LysM_cre_CpG_D1\"), pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    fplot(so, features=\"Gdf15\", slot=\"data\", restrict=c(\"sample_group\", \"IFNAR_fl_CpG_D1\"), pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    fplot(so, features=\"Gdf15\", slot=\"data\", restrict=c(\"sample_group\", \"IFNAR_fl_LysM_cre_CpG_D1\"), pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    fplot(so, features=\"Ccl2\", slot=\"data\", restrict=c(\"sample_group\", \"IFNAR_fl_CpG_D1\"), pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), \n",
    "    fplot(so, features=\"Ccl2\", slot=\"data\", restrict=c(\"sample_group\", \"IFNAR_fl_LysM_cre_CpG_D1\"), pt_size=0.25, size_select=4) %>% egg::set_panel_size(., width=unit(2.5, \"cm\"), height=unit(2.5, \"cm\")), ncol=2\n",
    "    \n",
    ")\n",
    "\n",
    "dev.off()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "r.4.4.1-FD20200109SPLENO",
   "language": "R",
   "name": "r.4.4.1-fd20200109spleno"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
