{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8e0dccd7-3dbb-4154-9cea-6ee0b97537c2",
   "metadata": {},
   "source": [
    "# Transcription factor binding site "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "88f4a271-d5db-469d-bf43-d1545cb23bb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "library_load <- suppressMessages(\n",
    "    \n",
    "    suppressWarnings(\n",
    "    \n",
    "        list(\n",
    "            \n",
    "            # Genomic tools \n",
    "            library(rtracklayer), \n",
    "            library(GenomicRanges), \n",
    "            \n",
    "            # Data \n",
    "            library(tidyverse),\n",
    "            \n",
    "            # Plot \n",
    "            library(ggplot2), \n",
    "            \n",
    "            # Pyhton compatibility\n",
    "            library(reticulate)\n",
    "            \n",
    "        )\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "eb23d461-29c7-4257-a156-f366f3af41c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "python:         /nobackup/peer/fdeckert/miniconda3/envs/p.3.10.16-FD20200109SPLENO/bin/python\n",
       "libpython:      /nobackup/peer/fdeckert/miniconda3/envs/p.3.10.16-FD20200109SPLENO/lib/libpython3.10.so\n",
       "pythonhome:     /nobackup/peer/fdeckert/miniconda3/envs/p.3.10.16-FD20200109SPLENO:/nobackup/peer/fdeckert/miniconda3/envs/p.3.10.16-FD20200109SPLENO\n",
       "version:        3.10.16 | packaged by conda-forge | (main, Apr  8 2025, 20:53:32) [GCC 13.3.0]\n",
       "numpy:          /nobackup/peer/fdeckert/miniconda3/envs/p.3.10.16-FD20200109SPLENO/lib/python3.10/site-packages/numpy\n",
       "numpy_version:  1.26.4\n",
       "\n",
       "NOTE: Python version was forced by use_python() function"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Configure reticulate \n",
    "use_condaenv(condaenv='p.3.10.16-FD20200109SPLENO', conda=\"/nobackup/peer/fdeckert/miniconda3/bin/conda\", required=NULL)\n",
    "py_config()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "366ed9e3-e6bf-4a36-8985-4d330b13c4a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn=-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "39d93a5c-dbd5-4b30-bb56-59eb214c6ab8",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_seed <- 42\n",
    "set.seed(random_seed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7d64285a-2310-4080-b5c0-b6e0de840caf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set working directory to project root\n",
    "setwd(\"/research/peer/fdeckert/FD20200109SPLENO\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "350935ba-956a-4bfe-a797-f36fd725a34e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Source files\n",
    "source(\"plotting_global.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e0c5215b-09fc-40bd-b269-7950c24db0de",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf <- rtracklayer::import(\"/research/peer/fdeckert/reference/genome/GRCm38/gencode.vM25.annotation.gtf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "709208b4-4f51-4f03-aab6-8416fab09bb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes <- c(\"Ly6a\", \"Adarb1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "14a6348d-5cbb-468d-ac47-2364b8d47f6f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[1]]\n",
       "BEDFile object\n",
       "resource: script/1_scRNAseq/8_tfbs/ly6a.bed \n",
       "\n",
       "[[2]]\n",
       "BEDFile object\n",
       "resource: script/1_scRNAseq/8_tfbs/adarb1.bed \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "lapply(genes, function(i) {\n",
    "    \n",
    "    gn <- gtf[gtf$type==\"gene\" & gtf$gene_name==i]\n",
    "    pr <- GenomicRanges::promoters(gn, upstream=100, downstream=500)\n",
    "    rtracklayer::export(as.data.frame(gn)[, c(\"seqnames\", \"start\", \"end\", \"strand\")], paste0(\"script/1_scRNAseq/8_tfbs/\", tolower(i), \".bed\"))\n",
    "\n",
    "}\n",
    "      )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4706633c-3b18-4ad1-83e2-7dec3750c31e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "133"
      ],
      "text/latex": [
       "133"
      ],
      "text/markdown": [
       "133"
      ],
      "text/plain": [
       "[1] 133"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "read.table(\"script/1_scRNAseq/8_tfbs/adarb1_tfbs.bed\") %>% dplyr::pull(V7) %>% toupper() %>% unique() %>% length()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "9e058753-c870-4b68-a4ad-6505a1dc6738",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "72"
      ],
      "text/latex": [
       "72"
      ],
      "text/markdown": [
       "72"
      ],
      "text/plain": [
       "[1] 72"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "read.table(\"script/1_scRNAseq/8_tfbs/ly6a_tfbs.bed\") %>% dplyr::pull(V7) %>% toupper() %>% unique() %>% length()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f3a0070-e84a-444f-a7e2-1ca812b857c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# module load BEDTools\n",
    "# module load Kent_tools\n",
    "\t\n",
    "# /nobackup/peer/fdeckert/JASPAR/tools/jaspar_tfbs_extraction/bin/extract_TFBSs_JASPAR.sh \\\n",
    "# -i /research/peer/fdeckert/FD20200109SPLENO/script/1_scRNAseq/8_tfbs/adarb1.bed \\\n",
    "# -b /nobackup/peer/fdeckert/JASPAR/database/JASPAR_TFBSs/2024/JASPAR2024_mm10.bb \\\n",
    "# -o /research/peer/fdeckert/FD20200109SPLENO/script/1_scRNAseq/8_tfbs/adarb1_tfbs.bed \\\n",
    "# -s 200\n",
    "\n",
    "# /nobackup/peer/fdeckert/JASPAR/tools/jaspar_tfbs_extraction/bin/extract_TFBSs_JASPAR.sh \\\n",
    "# -i /research/peer/fdeckert/FD20200109SPLENO/script/1_scRNAseq/8_tfbs/ly6a.bed \\\n",
    "# -b /nobackup/peer/fdeckert/JASPAR/database/JASPAR_TFBSs/2024/JASPAR2024_mm10.bb \\\n",
    "# -o /research/peer/fdeckert/FD20200109SPLENO/script/1_scRNAseq/8_tfbs/ly6a_tfbs.bed \\\n",
    "# -s 200"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "r.4.4.1-FD20200109SPLENO",
   "language": "R",
   "name": "r.4.4.1-fd20200109spleno"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
